{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[nltk_data] Downloading package stopwords to /home/omai.r/nltk_data...\n",
      "[nltk_data]   Package stopwords is already up-to-date!\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from RAKE.rake import Rake\n",
    "import json\n",
    "import ast\n",
    "import nltk\n",
    "from nltk.corpus import stopwords \n",
    "from nltk.tokenize import word_tokenize \n",
    "import json\n",
    "import numpy as np\n",
    "from tqdm import tqdm_notebook as tqdm\n",
    "from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer\n",
    "\n",
    "nltk.download('stopwords')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "matched_dfs = pd.read_csv('data/intermediate/matched_dfs.csv', converters = {'clean_text': ast.literal_eval})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#matched_dfs = pd.read_csv('data/intermediate/matched_dfs_merged.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "nature2references   = pd.read_hdf('data/nature/preprocessed/Nature2References.hdf')\n",
    "nature_references_info = pd.read_hdf('data/nature/preprocessed/NatureReferencesInformation.hdf').set_index('DOI')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "nature2references = pd.merge(nature2references, nature_references_info[['ArticleID']].reset_index(), right_on = 'ArticleID', left_on = 'CitingArticleID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#matched_dfs = pd.merge(nature2references, matched_dfs,\n",
    "#                       right_on = 'ArticleID', left_on = 'CitedArticleID')\n",
    "\n",
    "matched_dfs = pd.merge(matched_dfs, nature_references_info[['ArticleID', 'Title']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ArticleID</th>\n",
       "      <th>text</th>\n",
       "      <th>clean_text</th>\n",
       "      <th>count</th>\n",
       "      <th>Title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pub.1000000277</td>\n",
       "      <td>[['A detailed comparison of residues involved ...</td>\n",
       "      <td>[A detailed comparison of residues involved in...</td>\n",
       "      <td>1</td>\n",
       "      <td>Molecular Determinants of Agonist Discriminati...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pub.1000000280</td>\n",
       "      <td>[['Also, the epigenetic stability of human ES ...</td>\n",
       "      <td>[Also, the epigenetic stability of human ES ce...</td>\n",
       "      <td>1</td>\n",
       "      <td>Temporal and parental-specific expression of i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pub.1000000588</td>\n",
       "      <td>[['Hospital and clinic HTC had similar costs (...</td>\n",
       "      <td>[Hospital and clinic HTC had similar costs (&amp;d...</td>\n",
       "      <td>1</td>\n",
       "      <td>What Is the Cost of Providing Outpatient HIV C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pub.1000000718</td>\n",
       "      <td>[['&lt;b&gt;e&lt;/b&gt;, Allele-based pleiotropy of &lt;i&gt;IL6...</td>\n",
       "      <td>[&lt;b&gt;e&lt;/b&gt;, Allele-based pleiotropy of &lt;i&gt;IL6R&lt;...</td>\n",
       "      <td>2</td>\n",
       "      <td>Identification of multiple risk variants for a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pub.1000000866</td>\n",
       "      <td>[['However, structural&lt;bibr rid=\"b39 b40 b41 b...</td>\n",
       "      <td>[However, structural&lt;bibr rid=\"b39 b40 b41 b42...</td>\n",
       "      <td>1</td>\n",
       "      <td>Crystal Structure of an α1,4-N-Acetylhexosamin...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304443</th>\n",
       "      <td>pub.1121819253</td>\n",
       "      <td>[['Lineage-affiliated transcription factors&lt;bi...</td>\n",
       "      <td>[Lineage-affiliated transcription factors&lt;bibr...</td>\n",
       "      <td>4</td>\n",
       "      <td>GATA-1 interacts with the myeloid PU.1 transcr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304444</th>\n",
       "      <td>pub.1121819356</td>\n",
       "      <td>[['\\r\\n&lt;p&gt;In order to investigate a yolk-sac c...</td>\n",
       "      <td>[\\r\\n&lt;p&gt;In order to investigate a yolk-sac con...</td>\n",
       "      <td>1</td>\n",
       "      <td>VEGF-C signaling pathways through VEGFR-2 and ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304445</th>\n",
       "      <td>pub.1121824901</td>\n",
       "      <td>[['Despite the clear upper limit in observed a...</td>\n",
       "      <td>[Despite the clear upper limit in observed axi...</td>\n",
       "      <td>1</td>\n",
       "      <td>Sea-Floor Tectonics and Submarine Hydrothermal...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304446</th>\n",
       "      <td>pub.1122267945</td>\n",
       "      <td>[['The so-called photon-dominated or photodiss...</td>\n",
       "      <td>[The so-called photon-dominated or photodissoc...</td>\n",
       "      <td>7</td>\n",
       "      <td>Physics of the Interstellar and Intergalactic ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304447</th>\n",
       "      <td>pub.1122288893</td>\n",
       "      <td>[['Instead, transformations of reactive N in t...</td>\n",
       "      <td>[Instead, transformations of reactive N in the...</td>\n",
       "      <td>2</td>\n",
       "      <td>Ocean Biogeochemical Dynamics</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>304448 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             ArticleID                                               text  \\\n",
       "0       pub.1000000277  [['A detailed comparison of residues involved ...   \n",
       "1       pub.1000000280  [['Also, the epigenetic stability of human ES ...   \n",
       "2       pub.1000000588  [['Hospital and clinic HTC had similar costs (...   \n",
       "3       pub.1000000718  [['<b>e</b>, Allele-based pleiotropy of <i>IL6...   \n",
       "4       pub.1000000866  [['However, structural<bibr rid=\"b39 b40 b41 b...   \n",
       "...                ...                                                ...   \n",
       "304443  pub.1121819253  [['Lineage-affiliated transcription factors<bi...   \n",
       "304444  pub.1121819356  [['\\r\\n<p>In order to investigate a yolk-sac c...   \n",
       "304445  pub.1121824901  [['Despite the clear upper limit in observed a...   \n",
       "304446  pub.1122267945  [['The so-called photon-dominated or photodiss...   \n",
       "304447  pub.1122288893  [['Instead, transformations of reactive N in t...   \n",
       "\n",
       "                                               clean_text  count  \\\n",
       "0       [A detailed comparison of residues involved in...      1   \n",
       "1       [Also, the epigenetic stability of human ES ce...      1   \n",
       "2       [Hospital and clinic HTC had similar costs (&d...      1   \n",
       "3       [<b>e</b>, Allele-based pleiotropy of <i>IL6R<...      2   \n",
       "4       [However, structural<bibr rid=\"b39 b40 b41 b42...      1   \n",
       "...                                                   ...    ...   \n",
       "304443  [Lineage-affiliated transcription factors<bibr...      4   \n",
       "304444  [\\r\\n<p>In order to investigate a yolk-sac con...      1   \n",
       "304445  [Despite the clear upper limit in observed axi...      1   \n",
       "304446  [The so-called photon-dominated or photodissoc...      7   \n",
       "304447  [Instead, transformations of reactive N in the...      2   \n",
       "\n",
       "                                                    Title  \n",
       "0       Molecular Determinants of Agonist Discriminati...  \n",
       "1       Temporal and parental-specific expression of i...  \n",
       "2       What Is the Cost of Providing Outpatient HIV C...  \n",
       "3       Identification of multiple risk variants for a...  \n",
       "4       Crystal Structure of an α1,4-N-Acetylhexosamin...  \n",
       "...                                                   ...  \n",
       "304443  GATA-1 interacts with the myeloid PU.1 transcr...  \n",
       "304444  VEGF-C signaling pathways through VEGFR-2 and ...  \n",
       "304445  Sea-Floor Tectonics and Submarine Hydrothermal...  \n",
       "304446  Physics of the Interstellar and Intergalactic ...  \n",
       "304447                      Ocean Biogeochemical Dynamics  \n",
       "\n",
       "[304448 rows x 5 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matched_dfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ArticleID</th>\n",
       "      <th>text</th>\n",
       "      <th>clean_text</th>\n",
       "      <th>count</th>\n",
       "      <th>Title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pub.1000000277</td>\n",
       "      <td>[['A detailed comparison of residues involved ...</td>\n",
       "      <td>[A detailed comparison of residues involved in...</td>\n",
       "      <td>1</td>\n",
       "      <td>Molecular Determinants of Agonist Discriminati...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pub.1000000280</td>\n",
       "      <td>[['Also, the epigenetic stability of human ES ...</td>\n",
       "      <td>[Also, the epigenetic stability of human ES ce...</td>\n",
       "      <td>1</td>\n",
       "      <td>Temporal and parental-specific expression of i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pub.1000000588</td>\n",
       "      <td>[['Hospital and clinic HTC had similar costs (...</td>\n",
       "      <td>[Hospital and clinic HTC had similar costs (&amp;d...</td>\n",
       "      <td>1</td>\n",
       "      <td>What Is the Cost of Providing Outpatient HIV C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pub.1000000718</td>\n",
       "      <td>[['&lt;b&gt;e&lt;/b&gt;, Allele-based pleiotropy of &lt;i&gt;IL6...</td>\n",
       "      <td>[&lt;b&gt;e&lt;/b&gt;, Allele-based pleiotropy of &lt;i&gt;IL6R&lt;...</td>\n",
       "      <td>2</td>\n",
       "      <td>Identification of multiple risk variants for a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pub.1000000866</td>\n",
       "      <td>[['However, structural&lt;bibr rid=\"b39 b40 b41 b...</td>\n",
       "      <td>[However, structural&lt;bibr rid=\"b39 b40 b41 b42...</td>\n",
       "      <td>1</td>\n",
       "      <td>Crystal Structure of an α1,4-N-Acetylhexosamin...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        ArticleID                                               text  \\\n",
       "0  pub.1000000277  [['A detailed comparison of residues involved ...   \n",
       "1  pub.1000000280  [['Also, the epigenetic stability of human ES ...   \n",
       "2  pub.1000000588  [['Hospital and clinic HTC had similar costs (...   \n",
       "3  pub.1000000718  [['<b>e</b>, Allele-based pleiotropy of <i>IL6...   \n",
       "4  pub.1000000866  [['However, structural<bibr rid=\"b39 b40 b41 b...   \n",
       "\n",
       "                                          clean_text  count  \\\n",
       "0  [A detailed comparison of residues involved in...      1   \n",
       "1  [Also, the epigenetic stability of human ES ce...      1   \n",
       "2  [Hospital and clinic HTC had similar costs (&d...      1   \n",
       "3  [<b>e</b>, Allele-based pleiotropy of <i>IL6R<...      2   \n",
       "4  [However, structural<bibr rid=\"b39 b40 b41 b42...      1   \n",
       "\n",
       "                                               Title  \n",
       "0  Molecular Determinants of Agonist Discriminati...  \n",
       "1  Temporal and parental-specific expression of i...  \n",
       "2  What Is the Cost of Providing Outpatient HIV C...  \n",
       "3  Identification of multiple risk variants for a...  \n",
       "4  Crystal Structure of an α1,4-N-Acetylhexosamin...  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matched_dfs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ArticleID</th>\n",
       "      <th>PubYear</th>\n",
       "      <th>Title</th>\n",
       "      <th>Volume</th>\n",
       "      <th>Issue</th>\n",
       "      <th>Cites</th>\n",
       "      <th>Journal</th>\n",
       "      <th>NatureID</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DOI</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10.1038/166444b0</th>\n",
       "      <td>pub.1007899685</td>\n",
       "      <td>1950</td>\n",
       "      <td>Detection of Sugars on Paper Chromatograms</td>\n",
       "      <td>166</td>\n",
       "      <td>4219</td>\n",
       "      <td>3143</td>\n",
       "      <td>Nature</td>\n",
       "      <td>166444b0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10.1038/166117b0</th>\n",
       "      <td>pub.1026639700</td>\n",
       "      <td>1950</td>\n",
       "      <td>The Soil-Plate Method for Isolation of Fungi f...</td>\n",
       "      <td>166</td>\n",
       "      <td>4211</td>\n",
       "      <td>394</td>\n",
       "      <td>Nature</td>\n",
       "      <td>166117b0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10.1038/165514a0</th>\n",
       "      <td>pub.1039717355</td>\n",
       "      <td>1950</td>\n",
       "      <td>Reversible Dilation and Contraction by Changin...</td>\n",
       "      <td>165</td>\n",
       "      <td>4196</td>\n",
       "      <td>359</td>\n",
       "      <td>Nature</td>\n",
       "      <td>165514a0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10.1038/166780a0</th>\n",
       "      <td>pub.1011096023</td>\n",
       "      <td>1950</td>\n",
       "      <td>Cell Specificity of Histones</td>\n",
       "      <td>166</td>\n",
       "      <td>4227</td>\n",
       "      <td>280</td>\n",
       "      <td>Nature</td>\n",
       "      <td>166780a0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10.1038/166288a0</th>\n",
       "      <td>pub.1000041177</td>\n",
       "      <td>1950</td>\n",
       "      <td>Synthesis of Peptides in Enzymic Reactions inv...</td>\n",
       "      <td>166</td>\n",
       "      <td>4216</td>\n",
       "      <td>259</td>\n",
       "      <td>Nature</td>\n",
       "      <td>166288a0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       ArticleID PubYear  \\\n",
       "DOI                                        \n",
       "10.1038/166444b0  pub.1007899685    1950   \n",
       "10.1038/166117b0  pub.1026639700    1950   \n",
       "10.1038/165514a0  pub.1039717355    1950   \n",
       "10.1038/166780a0  pub.1011096023    1950   \n",
       "10.1038/166288a0  pub.1000041177    1950   \n",
       "\n",
       "                                                              Title Volume  \\\n",
       "DOI                                                                          \n",
       "10.1038/166444b0         Detection of Sugars on Paper Chromatograms    166   \n",
       "10.1038/166117b0  The Soil-Plate Method for Isolation of Fungi f...    166   \n",
       "10.1038/165514a0  Reversible Dilation and Contraction by Changin...    165   \n",
       "10.1038/166780a0                       Cell Specificity of Histones    166   \n",
       "10.1038/166288a0  Synthesis of Peptides in Enzymic Reactions inv...    166   \n",
       "\n",
       "                 Issue Cites Journal  NatureID  \n",
       "DOI                                             \n",
       "10.1038/166444b0  4219  3143  Nature  166444b0  \n",
       "10.1038/166117b0  4211   394  Nature  166117b0  \n",
       "10.1038/165514a0  4196   359  Nature  165514a0  \n",
       "10.1038/166780a0  4227   280  Nature  166780a0  \n",
       "10.1038/166288a0  4216   259  Nature  166288a0  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nature_references_info.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['However, structural and biochemical studies of several retaining glycosyltransferases in the presence of substrate analogues failed to identify an active-site nucleophile that could be involved in forming the enzyme–sugar adduct.']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matched_dfs.loc[4]['clean_text']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "rake = Rake(\"RAKE/SmartStoplist.txt\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test on a few papers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def top_trigram(text):\n",
    "    #print(text)\n",
    "    \n",
    "    \n",
    "    stop_words = set(stopwords.words('english')) \n",
    "    word_tokens = word_tokenize(' '.join(text))\n",
    "    filtered_sentence = [w for w in word_tokens if (w not in stop_words) and (w.isalpha())]\n",
    "    \n",
    "    distribution = nltk.FreqDist(nltk.trigrams(filtered_sentence))\n",
    "    distribution =  pd.DataFrame.from_dict(distribution, orient = 'index')\n",
    "    distribution['freq'] = distribution[0] / distribution[0].sum()\n",
    "    \n",
    "    \n",
    "    \n",
    "    return distribution.sort_values('freq', ascending = False).iloc[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Subjects were assigned a location in a network of 20 nodes, generated by a preferential attachment model; the network structure was created de novo for each session by attaching new nodes (each with two links) to existing nodes; and subjects were placed into the resulting networks at random.',\n",
       " 'A soft thresholding power was chosen on the basis of the criterion of approximate scale-free topology.',\n",
       " 'ÁngelesSerranoM\\nMariánBoguñáM\\nDmitriKrioukovD\\nDepartment of Electrical Engineering, Computer Engineering and Informatics, Cyprus University of Technology, 33 Saripolou Street, 3036 Limassol, Cyprus\\nCooperative Association for Internet Data Analysis (CAIDA), University of California, San Diego (UCSD), La Jolla, California 92093, USA\\nDepartament de Física Fonamental, Universitat de Barcelona, Martí i Franquès 1, 08028 Barcelona, Spain\\nf.papadopoulos@cut.ac.cy\\ndima@ucsd.edu\\nA framework is developed in which new connections to a growing network optimize geometric trade-offs between popularity and similarity, instead of simply preferring popular nodes; this approach accurately describes the large-scale evolution of various networks.&e120927-5;\\r\\n\\n\\nThe principle that ‘popularity is attractive’ underlies preferential attachment, which is a common explanation for the emergence of scaling in growing networks.',\n",
       " 'The networks grown as described may seem to have nothing in common with preferential attachment (PA).',\n",
       " 'We also explore evolution on random and scale-free networks.',\n",
       " 'This procedure is applied to several different real networks: (1) a part of the WWW composed of 325,729 web pages that are connected if there is a URL link from one page to another (http://www.nd.edu/~networks); (2) a social network where the nodes are 392,340 actors linked if they were cast together in at least one film; (3) the biological networks of protein–protein interactions found in Escherichia coli (429 proteins) and Homo sapiens (946 proteins) linked if there is a physical binding between them (database available via the Database of Interacting Proteins, other PINs are discussed in the Supplementary Information), and (4) the cellular networks compiled by ref.',\n",
       " '12 in the context of scale-free networks, such as social networks, protein interaction networks or metabolic networks.',\n",
       " '15: social networks, the world wide web, metabolic networks, and yeast protein interaction networks (PIN).',\n",
       " '',\n",
       " 'As is typical of networks, the degree distributions have significantly fatter tails than a similarly sized group composed of individuals randomly forming the same number of social ties (Kolmogorov–Smirnov test, P\\u2009<\\u200910−15 for all comparisons).',\n",
       " 'These findings imply that the observed flux distribution is a generic feature of flux conservation on a scale-free network, as it is independent of the optimal or non-optimal nature of the growth rate or the growth conditions.',\n",
       " '18 and Supplementary Information), coupled with the presence of a power law in both the optimal and non-optimal flux states, suggests that the metabolic flux organization is a direct consequence of the scale-free topology of the network.',\n",
       " '(vicsek@angel.elte.hu).\\r\\n\\r\\n\\r\\n&e050609-17;\\r\\n\\r\\n\\nMany complex systems in nature and society can be described in terms of networks capturing the intricate web of connections among the units they are made of.',\n",
       " 'It is well known that the nodes of large real networks have a power-law degree distribution.',\n",
       " 'Here, a better fit is achieved with a scale-free network; however, neither model simultaneously fits both distributions.',\n",
       " '(buldyrev@yu.edu).\\r\\n\\r\\n&e100415-7; &nature08932-s1;\\r\\n\\n\\nComplex networks have been studied intensively for a decade, but research still focuses on the limited case of a single, non-interacting network.',\n",
       " 'For two interdependent scale-free networks with power-law degree distributions, PA(k) = PB(k)\\u2009∝\\u2009k-&lgr;, we find that the existence criteria for the giant component are quite different from those in a single network.',\n",
       " '1) are as follows: 100,000 Erdös–Rényi (ER) with edge probability 0\\u2009<\\u2009p\\u2009<\\u20091; 100,000 small world (SW) with initial connection distance 1\\u2009≤\\u2009d\\u2009≤\\u20095 and edge creation probability 0\\u2009<\\u2009p\\u2009<\\u20090.05; 100,000 Barabási–Albert (BA) with linking number 1\\u2009≤\\u2009m\\u2009≤\\u200910; 100,000 random recursive (RR) (like Barabási–Albert except that edges are added uniformly instead of preferentially) with linking number 2\\u2009≤\\u2009m\\u2009≤\\u20098; 200,000 Holme–Kim (HK) with linking number 2\\u2009≤\\u2009m\\u2009≤\\u20094 and triad formation parameter 0\\u2009<\\u2009p\\u2009<\\u20090.15; 200,000 Klemm–Eguiluz (KE) with linking number 3\\u2009≤\\u2009m\\u2009≤\\u20095 and deactivation parameter 0\\u2009<\\u2009μ\\u2009<\\u20090.15; 200,000 shifted-linear preferential attachment with linking number 1\\u2009≤\\u2009m\\u2009≤\\u20097 and shift 0\\u2009<\\u2009θ\\u2009<\\u200940; 100,000 forest fire (FF) with parameters 0\\u2009<\\u2009pf\\u2009<\\u2009pb\\u2009<\\u20090.15; 100,000 Island Barabási–Albert; and 100,000 Island Erdös–Rényi.',\n",
       " '2a, b shows for two canonical network models (Erd&odblac;s–Rényi and scale-free), the fraction of driver nodes is significantly higher among low-k nodes than among the hubs.',\n",
       " '2c), indicating scale-free network topology.',\n",
       " 'Previous studies have shown that highly connected proteins within a network tend to be more highly conserved, a consequence of either functional constraints or preferential interaction of new proteins with existing highly connected proteins.',\n",
       " 'As demonstrated by Barabasi and colleagues and by several other researchers, power laws describe the distribution of various quantities in numerous biological, physical and social contexts; such distributions can seem to be fundamentally different (for example, the number of links between documents in the Internet, the population of towns and the number of reactions in which a given metabolite is involved).',\n",
       " 'Lower-tier nodes are more sparsely connected and the network structure has a scale-free power-law degree distribution.',\n",
       " 'Here we describe a surprisingly simple rule that is a good approximation for all graphs that we have analysed, including cycles, spatial lattices, random regular graphs, random graphs and scale-free networks: natural selection favours cooperation, if the benefit of the altruistic act, b, divided by the cost, c, exceeds the average number of neighbours, k, which means b/c > k.',\n",
       " 'Moreover, we find that the rule also holds for random graphs and scale-free networks, where individuals differ in the number of their neighbours.',\n",
       " '(vicsek@angel.elte.hu).\\r\\n\\r\\n&e070405-11;\\r\\n\\n\\nThe rich set of interactions between individuals in society results in complex community structure, capturing highly connected circles of friends, families or professional cliques in a social network.',\n",
       " 'The analogy to networks is also noteworthy: if poorly connected nodes (interacting cells) are damaged then the system will repair itself, whereas if strongly connected nodes are damaged the self-organization will disappear.',\n",
       " 'When social ties follow a scale-free distribution, cooperation is enhanced whenever all individuals are expected to contribute a fixed amount irrespective of the plethora of public goods games in which they engage.',\n",
       " 'Such strongly heterogeneous populations were obtained by means of the Barabási–Albert scale-free model based on growth and preferential attachment.',\n",
       " 'This implies that the smallest group in this population has three individuals.',\n",
       " 'b, Scale-free graphs in which small-world effects coexist with a large heterogeneity in neighbourhood size.']"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matched_dfs.iloc[45845]['clean_text']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>freq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>(networks, degree, distributions)</th>\n",
       "      <td>2</td>\n",
       "      <td>0.003317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(highly, connected, proteins)</th>\n",
       "      <td>2</td>\n",
       "      <td>0.003317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(networks, social, networks)</th>\n",
       "      <td>2</td>\n",
       "      <td>0.003317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(random, graphs, networks)</th>\n",
       "      <td>2</td>\n",
       "      <td>0.003317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(protein, interaction, networks)</th>\n",
       "      <td>2</td>\n",
       "      <td>0.003317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(degree, distribution, Here)</th>\n",
       "      <td>2</td>\n",
       "      <td>0.003317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(Internet, Data, Analysis)</th>\n",
       "      <td>1</td>\n",
       "      <td>0.001658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(attachment, This, implies)</th>\n",
       "      <td>1</td>\n",
       "      <td>0.001658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(PB, k, lgr)</th>\n",
       "      <td>1</td>\n",
       "      <td>0.001658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(spatial, lattices, random)</th>\n",
       "      <td>1</td>\n",
       "      <td>0.001658</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   0      freq\n",
       "(networks, degree, distributions)  2  0.003317\n",
       "(highly, connected, proteins)      2  0.003317\n",
       "(networks, social, networks)       2  0.003317\n",
       "(random, graphs, networks)         2  0.003317\n",
       "(protein, interaction, networks)   2  0.003317\n",
       "(degree, distribution, Here)       2  0.003317\n",
       "(Internet, Data, Analysis)         1  0.001658\n",
       "(attachment, This, implies)        1  0.001658\n",
       "(PB, k, lgr)                       1  0.001658\n",
       "(spatial, lattices, random)        1  0.001658"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_trigram(matched_dfs.iloc[45845]['clean_text'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_df = matched_dfs.sort_values('count', ascending = False).iloc[:300].sample(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\n",
      "Paper Title: A Bivalent Chromatin Structure Marks Key Developmental Genes in Embryonic Stem Cells \n",
      "\n",
      "Paper Keywords: 10.1016/j.cell.2006.02.041 \n",
      "\n",
      " PaperID: [('primary human fibroblast cell types', 21.5), ('actively silence differentiation-affiliated transcription factors', 20.7), ('h3k27me pathway disrupt cardiac development', 19.3), ('chromatin regulatory networks controlling pluripotency', 18.77894736842105), ('bivalent domains mark developmentally important loci', 17.57016632016632), ('systematically analysed chromatin-modifying proteins', 15.0), ('non-expressed target genes adopts', 14.45945945945946), ('maps show close agreement', 13.5), ('mouse cage tag clusters', 13.4), ('previously reported chip-chip data', 13.0), ('properly restore bivalent domains', 12.793243243243243), ('human frontal cortex h3k4me3', 12.666666666666666), ('regulating key developmental genes', 12.316602316602317), ('poised h3k4me3+h3k27me3+ bivalent status', 11.925061425061426), ('mark active enhancers specifically', 11.876923076923077), ('histone h3 lysine 4 trimethylation', 11.704761904761906), ('induced pluripotent stem cells', 11.589147286821705), ('poise developmental regulator genes', 11.566602316602317), ('differentiating embryonic stem cells', 11.547480620155039), ('silent chromatin configuration characteristic', 11.412280701754387), ('chromatin marks characterizes ‘poised', 11.094098883572569), ('transcriptionally repress developmental genes', 10.816602316602317), ('transcription initiation mark h3k4me3', 10.443589743589744), ('found 563 potential ‘poised enhancers', 10.181818181818182), ('developmentally regulated loci', 9.95), ('developmentally regulated genes', 9.20945945945946), ('mark developmental loci', 9.134065934065934), ('subsequent lineage decisions', 9.0), ('integrating sequence conservation', 9.0), ('pre-proximal paused tsss', 9.0), ('play key parts', 9.0), ('h1 frequently displayed', 9.0), ('pcg proteins function', 9.0), ('advances include insights', 9.0), ('putative enhancers captured', 9.0), ('developmentally important genes', 8.95945945945946), ('histone h3 lysine 27', 8.904761904761905), ('histone h3 lysine 4', 8.904761904761905), ('progenitor cell marks', 8.833333333333334), ('‘multicellular organismal development', 8.8), ('embryonic stem cell', 8.791666666666666), ('h3k27me3-establishing polycomb proteins', 8.75), ('lineage regulators harbour', 8.666666666666666), ('additional chromatin proteins', 8.578947368421053), ('developmental regulatory genes', 8.516602316602317), ('core component suz12', 8.5), ('proteins promote self-renewal', 8.5), ('reflect enhancer potential', 8.5), ('trophoblast marker genes', 8.45945945945946), ('pluripotency marker genes', 8.45945945945946), ('tumour suppressor genes', 8.45945945945946), ('poised regulatory sequences', 8.381818181818183), ('histone 3 lysine 27 trimethylation', 8.37142857142857), ('histone 3 lysine 4 trimethylation', 8.37142857142857), ('critical functional role', 8.333333333333334), ('key developmental genes', 8.316602316602317), ('es cell regulators', 8.238095238095237), ('evolutionarily conserved regions', 8.2), ('‘transcription factor activity', 8.2), ('transcription factor activity', 8.2), ('shared chromatin landscape', 8.078947368421053), ('histone modification pattern', 8.047619047619047), ('display transcriptional pausing', 8.0), ('recent evidence suggests', 8.0), ('sequential chip experiments', 8.0), ('generally highly expressed', 8.0), ('adult stem cells', 7.922480620155039), ('transiently pluripotent cells', 7.922480620155039), ('permanently pluripotent cells', 7.922480620155039), ('histone lysine methylation', 7.821428571428571), ('termed ‘bivalent domains', 7.793243243243244), ('lack dna methylation', 7.75), ('stem-cell pluripotency regulation', 7.75), ('bivalent genes constitute', 7.7027027027027035), ('activating histone modifications', 7.685714285714285), ('poised transcriptional status', 7.681818181818182), ('ipscs occupy enhancers', 7.666666666666667), ('modifications previously reported', 7.571428571428571), ('repressive histone modifications', 7.558441558441558), ('embryonic stem cells', 7.547480620155039), ('repressive polycomb complexes', 7.522727272727273), ('preserve distinct patterns', 7.5), ('activating mark found', 7.476923076923077), ('monovalent inactive genes', 7.45945945945946), ('stap stem cells', 7.422480620155039), ('bivalent chromatin domains', 7.372190611664296), ('recent studies showed', 7.333333333333334), ('previous evidence implicating', 7.333333333333334), ('wild-type es cells', 7.327242524916944), ('es cells carry', 7.327242524916944), ('differentiated somatic cells', 7.2558139534883725), ('chromatin state characterized', 7.245614035087719), ('activating h3k4me3 mark', 7.143589743589743), ('including hox genes', 7.126126126126126), ('bivalent pattern genes', 7.036036036036037), ('polycomb repressive complex', 7.022727272727273), ('pluripotent es cells', 6.99390919158361), ('human gene promoters', 6.958333333333334), ('repressive mark h3k27me3', 6.94965034965035), ('repressive h3k27me3 mark', 6.94965034965035), ('exhibits bivalent h3k4me3', 6.909909909909911), ('embryonic hox expression', 6.891666666666666), ('bivalent domains found', 6.793243243243243), ('repressive h3k27 methylation', 6.411616161616162), ('h3k27me3 bivalent domains', 6.393243243243243), ('bivalent state thought', 6.30990990990991), ('cardiac development', 6.3), ('key regulatory', 6.2), ('mouse es h3k4me3', 6.1380952380952385), ('active h3k4 methylation', 5.8277777777777775), ('transcription factors', 5.7), ('transcription initiation', 5.7), ('mark genes', 5.536382536382536), ('developmental regulators', 5.523809523809524), ('methylation mark', 5.326923076923077), ('developmental genes', 5.316602316602317), ('embryonic stem', 5.291666666666666), ('development genes', 5.2594594594594595), ('polycomb complexes', 5.25), ('chromatin modifications', 5.150375939849624), ('chromatin domains', 5.128947368421052), ('hox genes', 5.126126126126126), ('histone marks', 5.0476190476190474), ('gene transcription', 5.033333333333333), ('show enrichment', 5.0), ('cage tags', 5.0), ('identify proteins', 5.0), ('transcriptionally inactive', 5.0), ('human brain', 5.0), ('enhancer activity', 5.0), ('potential mechanisms', 5.0), ('genes reported', 4.95945945945946), ('genes expressed', 4.95945945945946), ('stem cells', 4.922480620155039), ('chromatin marks', 4.912280701754386), ('bivalent state', 4.90990990990991), ('previously showed', 4.833333333333334), ('generally silent', 4.833333333333334), ('bivalent domains', 4.793243243243243), ('chromatin regions', 4.778947368421052), ('somatic cells', 4.7558139534883725), ('differentiated cells', 4.7558139534883725), ('stap cells', 4.7558139534883725), ('h3k4me3 mark', 4.743589743589744), ('activating marks', 4.733333333333333), ('histone tails', 4.714285714285714), ('poise genes', 4.70945945945946), ('prevent transcription', 4.7), ('h3k27 trimethylation', 4.688888888888888), ('human lincrnas', 4.666666666666667), ('chromatin immunoprecipitation', 4.578947368421053), ('chromatin organization', 4.578947368421053), ('chromatin targets', 4.578947368421053), ('h3k4 trimethylation', 4.5777777777777775), ('post-translational modifications', 4.571428571428571), ('h3k4me3/h3k27me3 domains', 4.55), ('regulate development', 4.55), ('active transcription', 4.5), ('human tissues', 4.5), ('pioneer factors', 4.5), ('‘dna binding', 4.5), ('accumulation patterns', 4.5), ('transcriptional activity', 4.5), ('lineage-restricted genes', 4.45945945945946), ('housekeeping genes', 4.45945945945946), ('prc2-targeted genes', 4.45945945945946), ('inactive genes', 4.45945945945946), ('regulate transcription', 4.45), ('mouse genome', 4.4), ('developmental processes', 4.357142857142858), ('previous studies', 4.333333333333334), ('analysis showed', 4.333333333333334), ('previous reports', 4.333333333333334), ('gene ontology', 4.333333333333334), ('marks coexist', 4.333333333333334), ('critical roles', 4.333333333333334), ('pluripotent escs', 4.333333333333333), ('es cells', 4.327242524916944), ('active genes', 4.2594594594594595), ('multipotent cells', 4.2558139534883725), ('resident cells', 4.2558139534883725), ('mammalian cells', 4.2558139534883725), ('cells found', 4.2558139534883725), ('ips cells', 4.2558139534883725), ('epigenetic landscape', 4.25), ('bivalent conformation', 4.243243243243244), ('bivalent nature', 4.243243243243244), ('bivalent domain', 4.243243243243244), ('h3k27me3 modifications', 4.171428571428572), ('h3k27 methylation', 4.138888888888889), ('mouse escs', 4.066666666666666), ('h3k4 methylation', 4.027777777777778), ('terminal ch3', 4.0), ('rlu@stanford', 4.0), ('high levels', 4.0), ('faulty restoration', 4.0), ('chip-seq analyses', 4.0), ('methionine substituted', 4.0), ('intrinsic mechanisms', 4.0), ('coordinated interaction', 4.0), ('sequentially accessible', 4.0), ('largely devoid', 4.0), ('ideal system', 4.0), ('unmethylated cpgs', 4.0), ('finding reveals', 4.0), ('supplementary table 2', 4.0), ('cancers arise', 4.0), ('‘hoxa transcript', 4.0), ('core region', 4.0), ('initially believed', 4.0), ('rarely detected', 4.0), ('19 distinct cpgs', 4.0), ('ezh1 knockdown', 4.0), ('genome-wide studies', 4.0), ('inherent property', 4.0), ('translational mechanisms', 4.0), ('embryoid bodies', 4.0), ('rnapii-bound interactions', 4.0), ('possibly due', 4.0), ('lower frequency', 4.0), ('cell-fate specification', 4.0), ('preferentially localize', 4.0), ('es-cell specific', 4.0), ('distal tip', 4.0), ('niche interactions', 4.0), ('real-time pcr', 4.0), ('integral part', 4.0), ('findings establish', 4.0), ('bivalent sites', 3.9932432432432434), ('h3k27me3 marks', 3.9333333333333336), ('bivalent promoters', 3.8682432432432434), ('group 2 regions', 3.866666666666667), ('inactivating marks', 3.8333333333333335), ('critical residue', 3.8333333333333335), ('promote differentiation', 3.833333333333333), ('active cis-elements', 3.8), ('quantify h3k4', 3.7777777777777777), ('adjacent h3k4', 3.7777777777777777), ('nmolecular regulation', 3.75), ('epigenetic plasticity', 3.75), ('epigenetic reprogramming', 3.75), ('genomic sites', 3.75), ('binding sites', 3.75), ('promoter regions', 3.7), ('large set', 3.666666666666667), ('h3k4me3 peaks', 3.666666666666667), ('overlapping h3k4me3', 3.666666666666667), ('separate group', 3.666666666666667), ('h3k4me3-bound promoters', 3.625), ('h3k27me3 enrichment', 3.6), ('h3k27me3 peaks', 3.6), ('h3k27me3 extends', 3.6), ('weak h3k27me3', 3.6), ('methylated h3k27', 3.5555555555555554), ('striking overlap', 3.5), ('fate involves', 3.5), ('vivo evidence', 3.5), ('transcriptional repression', 3.5), ('299 hcps sampled', 3.5), ('methylated h3k4', 3.4444444444444446), ('regulate expression', 3.35), ('facilitate activation', 3.333333333333333), ('prevent differentiation', 3.333333333333333), ('transcriptional activation', 3.333333333333333), ('non-methylated cgis', 3.25), ('loci', 3.2), ('lincrnas interact', 3.166666666666667), ('simultaneously enriched', 3.166666666666667), ('promoters embedded', 3.125), ('enhancers', 3.0), ('h3k27me', 3.0), ('pluripotency', 3.0), ('lysine 27', 2.857142857142857), ('important', 2.75), ('histone 3', 2.7142857142857144), ('pluripotent', 2.6666666666666665), ('chromatin', 2.5789473684210527), ('domains', 2.55), ('repress', 2.5), ('agreement', 2.5), ('characteristic', 2.5), ('previously', 2.5), ('reported', 2.5), ('genes', 2.4594594594594597), ('mouse', 2.4), ('activating', 2.4), ('pattern', 2.3333333333333335), ('silent', 2.3333333333333335), ('‘repressive', 2.272727272727273), ('repressive', 2.272727272727273), ('cells', 2.255813953488372), ('poise', 2.25), ('methylation', 2.25), ('‘bivalent', 2.2432432432432434), ('bivalent', 2.2432432432432434), ('‘poised', 2.1818181818181817), ('poised', 2.1818181818181817), ('es', 2.0714285714285716), ('transcriptional', 2.0), ('including', 2.0), ('complex', 2.0), ('chip', 2.0), ('characterized', 2.0), ('inactive', 2.0), ('found', 2.0), ('evidence', 2.0), ('studies', 2.0), ('distinct', 2.0), ('h3k27', 1.8888888888888888), ('‘active', 1.8), ('active', 1.8), ('h3k4', 1.7777777777777777), ('regulation', 1.75), ('sites', 1.75), ('regulate', 1.75), ('epigenetic', 1.75), ('escs', 1.6666666666666667), ('methylated', 1.6666666666666667), ('set', 1.6666666666666667), ('lincrnas', 1.6666666666666667), ('h3k4me3', 1.6666666666666667), ('group', 1.6666666666666667), ('ipscs', 1.6666666666666667), ('promoters', 1.625), ('expression', 1.6), ('h3k27me3', 1.6), ('promoter', 1.5), ('‘pioneer', 1.5), ('vivo', 1.5), ('repression', 1.5), ('inactivating', 1.5), ('fate', 1.5), ('interact', 1.5), ('tissues', 1.5), ('region', 1.5), ('residue', 1.5), ('processes', 1.5), ('overlap', 1.5), ('hcps', 1.5), ('embedded', 1.5), ('thought', 1.4), ('activation', 1.3333333333333333), ('differentiation', 1.3333333333333333), ('cgis', 1.25), ('enriched', 1.1666666666666667), ('mimicked', 1.0), ('commonly', 1.0), ('control', 1.0), ('lung', 1.0), ('arrows', 1.0), ('studied', 1.0), ('culture', 1.0), ('restriction', 1.0), ('esc', 1.0), ('interestingly', 1.0), ('suppressed', 1.0), ('notably', 1.0), ('foot', 1.0), ('&nature08575-s1', 1.0), ('considered', 1.0), ('recently', 1.0), ('shown', 1.0), ('locus', 1.0), ('hottip', 1.0), ('zebrafish', 1.0), ('marked', 1.0), ('instil', 1.0), ('molecular', 1.0), ('shank3', 1.0), ('2b', 1.0), ('represent', 1.0), ('embryo', 1.0), ('essentially', 1.0), ('searched', 1.0), ('itga7', 1.0), ('embryos', 1.0), ('defined', 1.0), ('resembled', 1.0), ('drosophila', 1.0), ('hallmark', 1.0), ('tssbiv', 1.0), ('defining', 1.0), ('fraction', 1.0), ('albeit', 1.0), ('hypothesized', 1.0), ('work', 1.0), ('&e091119-13', 1.0), ('resemble', 1.0), ('level', 1.0), ('sox2', 1.0), ('combination', 1.0), ('explain', 1.0), ('map', 1.0), ('keeping', 1.0), ('fig', 1.0), ('results', 1.0), ('strategies', 1.0), ('oct4', 1.0), ('observed', 1.0), ('annotated', 1.0), ('u200910-12', 1.0), ('irrespective', 1.0), ('fails', 1.0), ('u2009kb', 1.0), ('involved', 1.0), ('capacity', 1.0), ('silencing', 1.0), ('repressed', 1.0), ('ecrs', 1.0), ('similarity', 1.0), ('existence', 1.0), ('virtually', 1.0), ('means', 1.0), ('u200910-13', 1.0), ('figs 1', 1.0), ('locations', 1.0), ('possibility', 1.0), ('carried', 1.0), ('determine', 1.0), ('recruitment', 1.0), ('correlate', 1.0), ('raising', 1.0), ('exist', 1.0), ('half', 1.0), ('subset', 1.0), ('integrate', 1.0), ('reprpcwk', 1.0), ('exit', 1.0), ('implicated', 1.0), ('similar', 1.0), ('p-values', 1.0), ('minority', 1.0), ('‘univalent', 1.0), ('mutations', 1.0), ('nkx6-2', 1.0), ('co-occupied', 1.0), ('formation', 1.0), ('5′ element', 1.0), ('h3k4me', 1.0), ('coincide', 1.0), ('program', 1.0), ('earlier', 1.0), ('chd1', 1.0), ('terms', 1.0), ('average', 1.0), ('study', 1.0), ('cdx2', 1.0), ('promoters/enhancers', 1.0), ('h3k27ac', 1.0), ('providing', 1.0), ('presence', 1.0), ('depletion', 1.0), ('turn', 1.0), ('interest', 1.0), ('consistent', 1.0), ('gata2', 1.0), ('mediated', 1.0), ('reminiscent', 1.0), ('prc2', 1.0), ('eomes', 1.0), ('silenced', 1.0), ('brachyury', 1.0), ('present', 1.0), ('position', 1.0), ('postulated', 1.0), ('connecting', 1.0), ('surprising', 1.0), ('finally', 1.0), ('fescs', 1.0), ('positions', 1.0), ('signalling', 1.0), ('molecule', 1.0), ('u2009×', 1.0), ('‘wide', 1.0), ('u200910-6', 1.0), ('loss', 1.0), ('suggest', 1.0), ('variants', 1.0), ('carrying', 1.0), ('forming', 1.0), ('2a', 1.0), ('nanog', 1.0), ('[', 0), ('∼2', 0), (']', 0), ('8', 0), ('5%', 0), ('= 564', 0), ('3', 0), ('9', 0), ('5', 0), ('4', 0), ('1', 0), ('10', 0)]\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Paper Title: Comprehensive genomic characterization defines human glioblastoma genes and core pathways \n",
      "\n",
      "Paper Keywords: 10.1038/nature07385 \n",
      "\n",
      " PaperID: [('tcga projects including acute myelogenous leukaemia', 28.166666666666664), ('recently published cancer genome sequencing studies identified', 26.27449370552819), ('ahr target gene cyp1b1 compared', 19.0), ('recently published cancer genome studies', 18.99671592775041), ('homozygous deletions targeting genes involved', 16.72222222222222), ('pik3ca interaction domain similar', 16.0), ('dna alkylating agent temozolomide', 16.0), ('technological limitations initially constrained', 16.0), ('mutant idh repression signature', 15.0), ('systematic polymerase chain reaction', 14.666666666666666), ('redundant housekeeping gene deleted', 14.666666666666666), ('demonstrate identical genomic profiles', 14.6), ('array comparative genomic hybridization', 14.6), ('suggested specific therapeutic approaches', 14.25), ('high-resolution copy number arrays', 14.033333333333335), ('recent exhaustive mutational analyses', 14.0), ('comparative genomic hybridization data', 13.838095238095239), ('international cancer genome consortium —', 13.830049261083744), ('systematic genome characterization efforts', 13.771428571428572), ('copy number quantification reaction', 13.7), ('cancer genome atlas network', 13.580049261083744), ('22 adult gbm samples sequenced previously', 13.36111111111111), ('large-cell neuroendocrine lung tumours', 13.333333333333334), ('atlas-tcga data set includes 77', 13.071428571428571), ('tcga copy number aberrations', 12.7), ('recent large cancer studies', 12.42528735632184), ('average background mutation frequency', 12.25), ('cancer genome atlas data', 11.818144499178981), ('including 13 significantly mutated genes', 11.322222222222223), ('polymerase chain reaction', 10.666666666666666), ('tcga gbm data set', 10.515873015873016), ('ahr targets cyp1b1', 10.333333333333334), ('massive genomic studies', 9.933333333333334), ('large international consortia —', 9.833333333333334), ('sequenced large numbers', 9.666666666666668), ('cancer genome atlas', 9.580049261083744), ('large-scale studies published', 9.166666666666668), ('highest copy number', 9.033333333333335), ('copy number alteration', 9.033333333333333), ('adult gbm databases', 9.027777777777779), ('improving human health', 9.0), ('detailed methodological support', 9.0), ('single nucleotide polymorphism', 9.0), ('essential cell activities', 9.0), ('carry huge numbers', 9.0), ('olfactory receptor or2l13', 9.0), ('fewer chromosome arm-level', 9.0), ('lung sqcc versus', 9.0), ('entire coding genome', 8.821428571428571), ('comprehensive genomic view', 8.8), ('178 lung sqccs identified', 8.777777777777779), ('consensus clustering approaches', 8.75), ('including ovarian cancer', 8.591954022988507), ('adult gbm subtypes', 8.527777777777779), ('specific exons selected', 8.5), ('rank sum test', 8.5), ('endothelial progenitor derivatives', 8.5), ('collectively termed ‘point', 8.5), ('-based re-sequencing studies', 8.333333333333334), ('renal cell carcinoma', 8.333333333333334), ('lung squamous carcinoma', 8.333333333333334), ('apparent false-positive findings', 8.333333333333334), ('major cancer type', 8.258620689655173), ('reveal molecular taxonomies', 8.25), ('important genes targeted', 8.222222222222221), ('potentially spurious genes', 8.222222222222221), ('critical signalling pathways', 8.0), ('find base substitutions', 8.0), ('1p36 tumour-suppressor locus', 8.0), ('protein-altering mutations identified', 7.977777777777778), ('gbm data set', 7.84920634920635), ('recurrently mutated genes', 7.822222222222222), ('frequently mutated genes', 7.822222222222222), ('identified transcriptome-based subgroups', 7.777777777777778), ('significant functional mutations', 7.7), ('cd133+ tumour cells', 7.666666666666668), ('17-gene expression signature', 7.666666666666667), ('pik3r1 contained 14 mutations', 7.533333333333334), ('previous tcga findings', 7.5), ('576 somatic mutations identified', 7.477777777777778), ('somatic cancer mutations', 7.458620689655173), ('performed quantitative pcr', 7.3), ('germline disease mutations', 7.2), ('previously identified mutations', 7.144444444444445), ('glioblastoma data set', 6.738095238095239), ('extended data fig', 6.738095238095238), ('underlie glioblastoma development', 6.666666666666667), ('tcga projects', 6.666666666666666), ('recurrent pik3r1 mutations', 6.533333333333334), ('copy number', 6.033333333333333), ('adult gbm', 6.027777777777778), ('sanger sequencing', 6.0), ('significantly mutated', 5.6), ('published data', 5.571428571428571), ('600 genes sequenced', 5.555555555555555), ('comprehensive studies', 5.533333333333333), ('large-scale analyses', 5.5), ('cyp1b1 derived', 5.333333333333334), ('genetic studies', 5.333333333333334), ('published literature', 5.333333333333334), ('cancer type', 5.258620689655173), ('previous approaches', 5.25), ('0 array data', 5.238095238095238), ('ovarian cancer', 5.0919540229885065), ('data set', 5.071428571428571), ('homozygous deletion', 5.0), ('signature typical', 5.0), ('homozygously deleted', 5.0), ('eno1 gene', 5.0), ('base substitutions', 5.0), ('lung adenocarcinoma', 5.0), ('1p36 locus', 5.0), ('recent demonstration', 5.0), ('-based re-sequencing', 5.0), ('tcga tumours', 5.0), ('recent years', 5.0), ('paediatric gbm', 4.777777777777778), ('endometrial cancer', 4.758620689655173), ('breast cancer', 4.758620689655173), ('colon cancer', 4.758620689655173), ('cancer subtype', 4.758620689655173), ('analytical approaches', 4.75), ('point mutations', 4.7), ('small number', 4.7), ('somatic mutations', 4.7), ('larger number', 4.7), ('expression profiles', 4.666666666666667), ('359 gbm samples', 4.611111111111111), ('cancer genomes', 4.508620689655173), ('exact test', 4.5), ('methylation subtypes', 4.5), ('endothelial progenitors', 4.5), ('quantitative fish', 4.5), ('coding genes', 4.472222222222222), ('‘comprehensive characterization', 4.4), ('comprehensive characterization', 4.4), ('tumour biology', 4.333333333333334), ('colorectal carcinoma', 4.333333333333334), ('breast carcinoma', 4.333333333333334), ('tumour types', 4.333333333333334), ('cd105+ cells', 4.333333333333334), ('molecular sub-classification', 4.25), ('high frequency', 4.25), ('coding exons', 4.25), ('molecular basis', 4.25), ('mutation data', 4.238095238095238), ('source data', 4.238095238095238), ('seq data', 4.238095238095238), ('mutagenesis data', 4.238095238095238), ('genes harbouring', 4.222222222222222), ('protein-coding genes', 4.222222222222222), ('cancer-causing genes', 4.222222222222222), ('recurrent mutations', 4.2), ('extensive characterization', 4.2), ('kinase mutations', 4.2), ('identify mutations', 4.2), ('indicator mutations', 4.2), ('comprehensive understanding', 4.2), ('egfr exons', 4.0), ('altered pathways', 4.0), ('nr1d2 status', 4.0), ('tdo-derived kyn', 4.0), ('cd133+/cd144− population', 4.0), ('kinase-signalling pathways', 4.0), ('survival probabilities', 4.0), ('intermediate prognosis', 4.0), ('egfr primers', 4.0), ('appreciable progress', 4.0), ('identify subgroups', 4.0), ('focal scnas', 4.0), ('inherited diseases', 4.0), ('functional relevance', 4.0), ('great power', 4.0), ('early detection', 4.0), ('today identify', 4.0), ('small proportion', 4.0), ('detecting insertions/deletions', 4.0), ('designed based', 4.0), ('rev-erb status', 4.0), ('limited overlap', 4.0), ('egfr amplicon', 4.0), ('mutation load', 4.0), ('structural variation', 4.0), ('near-complete absence', 4.0), ('grade iv', 4.0), ('potent drivers', 4.0), ('identify opportunities', 4.0), ('scnas stands', 4.0), ('small insertions', 4.0), ('aberrant pathways', 4.0), ('identify ‘driver', 4.0), ('mature endothelium', 4.0), ('gene-coding regions', 4.0), ('potential contributions', 4.0), ('largely made', 4.0), ('glioblastoma data', 3.904761904761905), ('glioma cells', 3.8333333333333335), ('pcr products', 3.8), ('real-time pcr', 3.8), ('h3f3a mutations', 3.7), ('supplementary table 8', 3.666666666666667), ('glioblastoma project', 3.666666666666667), ('high expression', 3.666666666666667), ('glioblastoma multiforme', 3.666666666666667), ('supplementary table 1', 3.666666666666667), ('previously reported', 3.6666666666666665), ('low frequency', 3.583333333333333), ('genome', 3.5714285714285716), ('supplementary fig', 3.5), ('small indels', 3.5), ('detect variations', 3.5), ('nby examining', 3.5), ('rates observed', 3.5), ('including', 3.5), ('focused primarily', 3.5), ('cancer-specific events', 3.5), ('high rate', 3.5), ('analyse mrna', 3.5), ('striking contrast', 3.5), ('studies', 3.3333333333333335), ('published', 3.3333333333333335), ('led directly', 3.333333333333333), ('combined analysis', 3.25), ('deletions', 3.0), ('involved', 3.0), ('compared', 3.0), ('set', 2.8333333333333335), ('identified 5', 2.7777777777777777), ('gbm', 2.7777777777777777), ('identified', 2.7777777777777777), ('cancer', 2.7586206896551726), ('number', 2.7), ('tcga', 2.6666666666666665), ('mutated', 2.6), ('findings', 2.3333333333333335), ('tumours', 2.3333333333333335), ('pik3r1', 2.3333333333333335), ('frequency', 2.25), ('data', 2.238095238095238), ('845 genes', 2.2222222222222223), ('genes', 2.2222222222222223), ('mutations', 2.2), ('characterization', 2.2), ('previously', 2.1666666666666665), ('disease', 2.0), ('mutation', 2.0), ('subgroups', 2.0), ('development', 2.0), ('exons 4', 2.0), ('recurrent', 2.0), ('pathways', 2.0), ('92 samples', 1.8333333333333333), ('samples', 1.8333333333333333), ('21 samples', 1.8333333333333333), ('pcr', 1.8), ('genomes', 1.75), ('glioblastoma', 1.6666666666666667), ('expression', 1.6666666666666667), ('table 1', 1.6666666666666667), ('analyse', 1.5), ('events', 1.5), ('contrast', 1.5), ('glioma', 1.5), ('rate', 1.5), ('fig', 1.5), ('observed', 1.5), ('h3f3a', 1.5), ('primarily', 1.5), ('detect', 1.5), ('relevance', 1.5), ('deletion', 1.5), ('indels', 1.5), ('reported', 1.5), ('sanger', 1.5), ('examining', 1.5), ('low', 1.3333333333333333), ('led', 1.3333333333333333), ('eno1', 1.3333333333333333), ('analysis', 1.25), ('wide', 1.0), ('expansion', 1.0), ('scope', 1.0), ('context', 1.0), ('tek', 1.0), ('il8', 1.0), ('mrd', 1.0), ('genomics', 1.0), ('ncbi', 1.0), ('qpcr', 1.0), ('1 february 2012', 1.0), ('daxx', 1.0), ('t-test', 1.0), ('box 2', 1.0), ('order', 1.0), ('patients', 1.0), ('snp', 1.0), ('part', 1.0), ('resistance', 1.0), ('melanoma', 1.0), ('identification', 1.0), ('pattern', 1.0), ('prkd1', 1.0), ('considered', 1.0), ('regulated', 1.0), ('shown', 1.0), ('dp', 1.0), ('pclo', 1.0), ('application', 1.0), ('wilcoxon', 1.0), ('il6', 1.0), ('dlbcl', 1.0), ('best-prognosis', 1.0), ('problem', 1.0), ('gov/gwastudies', 1.0), ('mapping', 1.0), ('u20091', 1.0), ('elucidated', 1.0), ('org', 1.0), ('summary', 1.0), ('heterogeneous', 1.0), ('higher', 1.0), ('relapse', 1.0), ('nrg3', 1.0), ('finally', 1.0), ('mb', 1.0), ('ttn', 1.0), ('tlr4', 1.0), ('abundant', 1.0), ('provided', 1.0), ('insights', 1.0), ('class', 1.0), ('typically', 1.0), ('common', 1.0), ('nominated', 1.0), ('publically', 1.0), ('abundance', 1.0), ('oncogenesis', 1.0), ('icgc', 1.0), ('cbioportal', 1.0), ('muc4', 1.0), ('catalogues', 1.0), ('included', 1.0), ('role', 1.0), ('muc16', 1.0), ('cancer-', 1.0), ('u200910−16', 1.0), ('nr1d1', 1.0), ('173 patients', 1.0), ('mirna', 1.0), ('occur', 1.0), ('atrx', 1.0), ('u2009×', 1.0), ('supported', 1.0), ('156 segments', 1.0), ('segments', 1.0), ('pinpoint', 1.0), ('uk/perl/genetics/cgp/cosmic', 1.0), ('recognizing', 1.0), ('red', 1.0), ('unaltered', 1.0), ('alterations', 1.0), ('http', 1.0), ('optimism', 1.0), ('fuelled', 1.0), ('coupled', 1.0), ('477 segments', 1.0), ('predictors', 1.0), ('//www', 1.0), ('efficient', 1.0), ('clustered', 1.0), ('prkdc', 1.0), ('reviewed', 1.0), ('spop', 1.0), ('pathogenesis', 1.0), ('gov/omim', 1.0), ('worst-', 1.0), ('nlm', 1.0), ('additionally', 1.0), ('her2', 1.0), ('ac', 1.0), ('cancers', 1.0), ('acgh', 1.0), ('classifier', 1.0), ('comparison', 1.0), ('gbms', 1.0), ('org/', 1.0), ('agents', 1.0), ('program', 1.0), ('green', 1.0), ('prevalence', 1.0), ('prevention', 1.0), ('csmd3', 1.0), ('study', 1.0), ('recur', 1.0), ('u20092', 1.0), ('notch4', 1.0), ('neno1', 1.0), ('enabled', 1.0), ('algorithms', 1.0), ('chip', 1.0), ('cosmic', 1.0), ('cases', 1.0), ('aspects', 1.0), ('u2009<', 1.0), ('resides', 1.0), ('282 segments', 1.0), ('fisher', 1.0), ('expected', 1.0), ('present', 1.0), ('thousands', 1.0), ('candidates', 1.0), ('323 segments', 1.0), ('demonstrates', 1.0), ('found', 1.0), ('il1b', 1.0), ('nih', 1.0), ('gliomas', 1.0), ('213 segments', 1.0), ('conducted', 1.0), ('‘druggable', 1.0), ('substantiates', 1.0), ('contribute', 1.0), ('oligodendrogliomas', 1.0), ('therapies', 1.0), ('treatment', 1.0), ('transcriptomes', 1.0), ('9d', 1.0), ('lrp1b', 1.0), ('u200910−15', 1.0), ('applied', 1.0), ('outcome', 1.0), ('diversity', 1.0), ('glioblastomas', 1.0), ('361', 0), ('= 362', 0), ('9', 0), ('[', 0), ('56', 0), ('5%', 0), ('∼1%', 0), (']', 0), ('1', 0), ('449', 0), ('43%', 0), ('2', 0), ('3', 0), ('6', 0), ('0', 0), ('= 173', 0), ('95%', 0), ('11', 0)]\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Paper Title: Core Transcriptional Regulatory Circuitry in Human Embryonic Stem Cells \n",
      "\n",
      "Paper Keywords: 10.1016/j.cell.2005.08.020 \n",
      "\n",
      " PaperID: [('data confirmed previous studies showing limited overlap', 34.96666666666667), ('sequence-specific autoregulatory transcription factors targeting self-renewal genes', 32.952380952380956), ('es cell master transcription factors oct4', 21.76638655462185), ('regulatory circuitry underlying pluripotency reveal', 20.4), ('nanog-/- embryonic stem cells calls', 20.38095238095238), ('similar core transcription factor circuitry', 19.258333333333333), ('key pluripotency transcription factors—oct4', 17.49019607843137), ('nanog null embryonic stem cells', 16.815734989648035), ('mouse pluripotent embryonic stem cells', 16.78095238095238), ('mature sperm overlap greatly', 16.666666666666668), ('recursive triple feedback system', 16.0), ('developmental pcg target genes identified', 15.785714285714285), ('form specialized regulatory circuitry', 15.399999999999999), ('repressed lineage-specific developmental regulators', 15.0), ('core transcription factors based', 14.791666666666666), ('co-occupy rest promoter/enhancer sequences', 14.0), ('increased cell division rate', 13.809523809523808), ('mouse es cells showed', 13.775), ('genes encoding transcriptional activators', 13.035714285714285), ('turn targets regulatory genes', 12.55238095238095), ('recent chromatin occupancy studies', 12.2), ('maintains human es cells', 11.174999999999999), ('sox2 promoter binding information', 11.085714285714285), ('core transcription factors', 10.791666666666666), ('transcription factors collaborate', 10.666666666666666), ('embryonic stem cell', 10.482142857142858), ('embryonic stem cells', 10.380952380952381), ('embryonic stem cell-', 10.339285714285715), ('transcription factor repertoire', 10.333333333333334), ('mouse es cells', 9.775), ('core transcriptional circuitry', 9.675), ('∼7-fold greater overlap', 9.666666666666668), ('adult stem cells', 9.666666666666666), ('recent chromatin-precipitation studies', 9.6), ('specific regulatory subnetworks', 9.6), ('recent chromatin studies', 9.2), ('murine es cells', 9.174999999999999), ('resulting individual cell', 9.142857142857142), ('partially reprogrammed cells', 9.041666666666666), ('genome-wide location analysis', 9.0), ('genetic screens identifying', 9.0), ('advances include insights', 9.0), ('interconnected auto-regulatory loop', 9.0), ('previously validated algorithm', 9.0), ('experimental conditions surveyed', 9.0), ('transcriptional housekeeping machinery', 8.75), ('dna-binding protein rest', 8.7), ('genes encoding proteins', 8.685714285714285), ('human es cells', 8.674999999999999), ('supplementary table s2', 8.666666666666666), ('supplementary tables s1', 8.666666666666666), ('osn cooperatively bind', 8.5), ('specific expression pattern', 8.5), ('reprogram human fibroblasts', 8.5), ('genomic sites occupied', 8.333333333333334), ('core pluripotency network', 8.325), ('transcriptional effector proteins', 8.15), ('additional conserved instances', 8.0), ('dna binding annotations', 8.0), ('core regulators oct4', 7.948529411764706), ('pluripotent transcriptional network', 7.75), ('direct transcriptional target', 7.75), ('transcription factors', 7.666666666666666), ('apparently directly downstream', 7.5), ('embryonic stem', 7.339285714285714), ('developmental promoters bound', 7.25), ('oct4 protein occupancy', 7.023529411764706), ('esc fate alteration', 7.0), ('self-renewal network', 6.7), ('nanog protein binds', 6.6347826086956525), ('recent studies', 6.6), ('previous assumptions', 6.5), ('es cells', 6.175), ('core regulators', 6.125), ('pcg proteins', 5.9), ('cell division', 5.809523809523809), ('molecular circuitry', 5.8), ('supplementary information', 5.666666666666666), ('chromatin occupancy', 5.6), ('regulatory loci', 5.6), ('important factor', 5.333333333333334), ('direct binding', 5.3), ('target genes', 5.285714285714286), ('ips cell', 5.142857142857142), ('core element', 5.125), ('daughter cells', 5.041666666666666), ('ps cells', 5.041666666666666), ('undifferentiated cells', 5.041666666666666), ('driving cells', 5.041666666666666), ('establishing pluripotency', 5.0), ('pluripotent network', 5.0), ('stabilizes pluripotency', 5.0), ('recent chip-on-chip', 4.8), ('pluripotent state', 4.8), ('neuronal genes', 4.785714285714286), ('haematopoietic-specific genes', 4.785714285714286), ('transcriptional repressor', 4.75), ('chromatin immunoprecipitation', 4.6), ('chromatin structure', 4.6), ('continuing expression', 4.5), ('directly regulated', 4.5), ('abbreviated osn', 4.5), ('foxd3 proteins', 4.4), ('interacting proteins', 4.4), ('genomic maps', 4.333333333333334), ('genomic locations', 4.333333333333334), ('otx2 binding', 4.3), ('nanog genes', 4.220496894409938), ('dna hypomethylation', 4.2), ('-enriched dna', 4.2), ('human escs', 4.166666666666667), ('alleged targets', 4.166666666666666), ('sox2 genes', 4.071428571428571), ('differentiated fates', 4.0), ('rlu@stanford', 4.0), ('well-characterized enhancers', 4.0), ('polycomb repressor', 4.0), ('dot1l inhibition', 4.0), ('highly expressed', 4.0), ('showing', 4.0), ('systems-level analyses', 4.0), ('esc-specific regulation', 4.0), ('conservation-based approaches', 4.0), ('coordinated interaction', 4.0), ('nmolecular regulation', 4.0), ('signalling pathways', 4.0), ('intrinsic mechanisms', 4.0), ('hierarchical grn', 4.0), ('secondary role', 4.0), ('autoregulatory', 4.0), ('higher coverage', 4.0), ('reprogramming process', 4.0), ('fate involves', 4.0), ('challenge due', 4.0), ('data', 4.0), ('haematopoietic phenotype', 4.0), ('chip-pet analyses', 4.0), ('histone modifications', 4.0), ('substantive role', 4.0), ('independent probability', 4.0), ('essential role', 4.0), ('early embryos', 4.0), ('chip-on-chip technology', 4.0), ('differentiation states', 4.0), ('amplification step', 4.0), ('activator complex', 4.0), ('translational mechanisms', 4.0), ('called nrsf', 4.0), ('highly enriched', 4.0), ('niche interactions', 4.0), ('lin28 upregulated', 4.0), ('primary regulation', 4.0), ('poised state', 4.0), ('oct4 targets', 3.9901960784313726), ('oct3/4 targets', 3.833333333333333), ('dna-hypomethylated promoters', 3.75), ('undifferentiated escs', 3.666666666666667), ('factors', 3.6666666666666665), ('epigenetic marks', 3.5), ('combinatorial control', 3.5), ('links nanog', 3.4347826086956523), ('putative nanog', 3.4347826086956523), ('called pou5f1', 3.2), ('es', 3.1333333333333333), ('cells', 3.0416666666666665), ('pluripotency', 3.0), ('identified', 3.0), ('find nanog', 2.9347826086956523), ('genes', 2.7857142857142856), ('transcriptional', 2.75), ('target', 2.5), ('maintains', 2.5), ('human', 2.5), ('co-occupy', 2.5), ('network', 2.2), ('dna', 2.2), ('protein', 2.2), ('targets', 2.1666666666666665), ('greater', 2.0), ('esc', 2.0), ('instances', 2.0), ('fate', 2.0), ('apparently', 2.0), ('oct4', 1.8235294117647058), ('promoters', 1.75), ('escs', 1.6666666666666667), ('oct3/4', 1.6666666666666667), ('control', 1.5), ('otx2', 1.5), ('epigenetic', 1.5), ('bound', 1.5), ('find', 1.5), ('nanog—', 1.434782608695652), ('nanog', 1.434782608695652), ('sox2', 1.2857142857142858), ('pou5f1', 1.2), ('number', 1.0), ('facilitate', 1.0), ('noteworthy', 1.0), ('smad1', 1.0), ('ref', 1.0), ('linked', 1.0), ('forming', 1.0), ('serve', 1.0), ('variety', 1.0), ('functional', 1.0), ('s4', 1.0), ('missed', 1.0), ('mini-interactome', 1.0), ('intersection', 1.0), ('klf4', 1.0), ('size', 1.0), ('exist', 1.0), ('reflecting', 1.0), ('capable', 1.0), ('constitute', 1.0), ('mediator', 1.0), ('part', 1.0), ('progress', 1.0), ('transcript', 1.0), ('acquisition', 1.0), ('accelerate', 1.0), ('pou5f1/oct4', 1.0), ('conjunction', 1.0), ('considered', 1.0), ('shown', 1.0), ('re-establishment', 1.0), ('regulate', 1.0), ('publications', 1.0), ('genome', 1.0), ('handful', 1.0), ('restricted', 1.0), ('detected', 1.0), ('hypothesis', 1.0), ('repressors', 1.0), ('headed', 1.0), ('study', 1.0), ('mice', 1.0), ('ability', 1.0), ('existence', 1.0), ('vivo', 1.0), ('super-enhancers', 1.0), ('properties', 1.0), ('peaks', 1.0), ('u2009<', 1.0), ('question', 1.0), ('ecat', 1.0), ('chip', 1.0), ('smad2/3', 1.0), ('active', 1.0), ('contradict', 1.0), ('hescs', 1.0), ('slightly', 1.0), ('&e091119-13', 1.0), ('substantially', 1.0), ('understanding', 1.0), ('complexity', 1.0), ('prdm14', 1.0), ('/', 1.0), ('pitx2', 1.0), ('progressing', 1.0), ('drive', 1.0), ('contrary', 1.0), ('expected', 1.0), ('u20090', 1.0), ('concert', 1.0), ('notably', 1.0), ('ipscs', 1.0), ('recently', 1.0), ('thousands', 1.0), ('intrigued', 1.0), ('humans', 1.0), ('fig', 1.0), ('nuclear', 1.0), ('nodes', 1.0), ('suggests', 1.0), ('tcf3', 1.0), ('demonstrates', 1.0), ('reinforcement', 1.0), ('found', 1.0), ('possibilities', 1.0), ('designated', 1.0), ('linkages', 1.0), ('kinetics', 1.0), ('suggested', 1.0), ('&nature08575-s1', 1.0), ('combination', 1.0), ('reported', 1.0), ('co-regulating', 1.0), ('reconfiguration', 1.0), ('oct-4', 1.0), ('abundant', 1.0), ('pscs', 1.0), ('including', 1.0), ('code', 1.0), ('amplify', 1.0), ('episcs', 1.0), ('18', 0), ('[', 0), ('2', 0), ('4', 0), ('01', 0), ('56%', 0), (']', 0), ('1', 0)]\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Paper Title: Defining the Epithelial Stem Cell Niche in Skin \n",
      "\n",
      "Paper Keywords: 10.1126/science.1092436 \n",
      "\n",
      " PaperID: [('hair follicle stem cells express elevated nuclear factor', 34.29519331243469), ('special properties underlying bulge stem cells', 24.906304423545805), ('labelled cell undergoes cell division', 18.0), ('gfp protein remains stably expressed', 17.75), ('epithelial hair follicle stem cells', 15.211859979101359), ('transforming growth factor-β pathways —', 15.0), ('upper outer root sheath', 14.999999999999998), ('diminished gfp signal intensity', 14.583333333333334), ('follicle outer root sheath', 14.333333333333332), ('epidermal stem cell niches', 14.310344827586206), ('identifying live stem cells', 14.128526645768025), ('murine epidermal stem cells', 14.128526645768025), ('maintaining normal interfollicular epidermis', 13.75), ('female cd-1 mice transgenic', 13.529411764705882), ('keratin-5 promoter drives expression', 13.0), ('ensuring long-term tissue regeneration', 12.75), ('green fluorescent protein retention', 12.666666666666666), ('bulge cells remain quiescent', 12.429292929292929), ('show nuclear localization', 11.5), ('outer root sheath', 11.0), ('special quiescent nature', 10.833333333333334), ('bulge skin stem cells', 10.656304423545803), ('express proteins typically', 10.333333333333332), ('t-cell-specific transcription factor 3', 10.0), ('purity remains uncertain', 10.0), ('green fluorescent protein', 9.666666666666666), ('quiescent stem cells', 9.461859979101359), ('hair follicle—termed', 9.333333333333334), ('stem cell compartment', 9.310344827586206), ('hair stem cells', 9.128526645768025), ('melanocyte stem cells', 9.128526645768025), ('individual stem cells', 9.128526645768025), ('marking stem cells', 9.128526645768025), ('adult stem cells', 9.128526645768025), ('expended stem cells', 9.128526645768025), ('hair follicle scs', 9.083333333333334), ('sustain cyclical bouts', 9.0), ('retain dna labels', 9.0), ('segregate chromosomes asymmetrically', 9.0), ('differentiated cell types', 9.0), ('physiological hair regeneration', 9.0), ('requires micro-environmental stimuli', 9.0), ('examples include haematopoietic', 9.0), ('complex signalling interactions', 9.0), ('mouse hair follicle', 8.933333333333334), ('epithelial stem cells', 8.878526645768025), ('replenishes cells lost', 8.818181818181818), ('stem cell niche', 8.64367816091954), ('hair-follicle stem cells', 8.628526645768025), ('including stem cells', 8.628526645768025), ('transgenic mouse line', 8.6), ('gfp fluorescence decreases', 8.583333333333334), ('hair follicle cycle', 8.533333333333335), ('ptreh2bgfp transgenic mice', 8.529411764705882), ('tre-mcmv-h2b-gfp transgenic mice', 8.529411764705882), ('h2bmcherry transgenic mice', 8.529411764705882), ('intercrossed tetop-tert mice', 8.529411764705882), ('wnt receptor proteins', 8.5), ('basal cells increased', 8.484848484848484), ('proliferating basal cells', 8.484848484848484), ('bulge stem cells', 8.406304423545803), ('stable histone 2b', 8.4), ('gfp fusion protein', 8.25), ('putative ife scs', 8.25), ('histone label-retaining cells', 8.218181818181819), ('gfp transgenic mice', 8.112745098039216), ('resting bulge cells', 8.095959595959595), ('small number residing', 7.833333333333334), ('extended data fig', 7.75), ('teto-h2b-gfp transgenic mouse', 7.6), ('gfp label retention', 7.583333333333334), ('anagen cycle depends', 6.95), ('monitoring niche behaviour', 6.733333333333334), ('cell division', 6.5), ('histone h2b mice', 6.429411764705883), ('hair follicle', 6.333333333333334), ('express additional', 6.333333333333333), ('stem cells', 6.128526645768025), ('hair regeneration', 6.0), ('bulge undergoes', 5.777777777777778), ('gfp protein', 5.75), ('fusion protein', 5.666666666666666), ('transgenic mouse', 5.6), ('epithelial cells', 5.568181818181818), ('growth cycle', 5.2), ('genes previously shown', 5.166666666666667), ('gfp mice', 5.112745098039216), ('bulge cells', 5.095959595959596), ('skin cells', 5.068181818181818), ('keratin 14 promoter', 5.0), ('label retention', 5.0), ('fluorescent labelling', 5.0), ('hair cycling', 5.0), ('hair follicles', 5.0), ('selectively expressed', 5.0), ('preferentially expressed', 5.0), ('follicle base', 4.833333333333334), ('small number', 4.833333333333334), ('daughter cells', 4.818181818181818), ('axin2+ cells', 4.818181818181818), ('label cells', 4.818181818181818), ('cp cells', 4.818181818181818), ('hair-follicle bulge', 4.777777777777778), ('slow-cycling scs', 4.75), ('epithelial nuclei', 4.75), ('basal layer', 4.666666666666666), ('mouse lines', 4.6), ('enhanced gfp', 4.583333333333334), ('dct-rtta mice', 4.529411764705882), ('ptre-h2bgfp mice', 4.529411764705882), ('scltta mice', 4.529411764705882), ('ntransgenic mice', 4.529411764705882), ('ife lie', 4.5), ('wnt inhibition', 4.5), ('including hscs', 4.5), ('niche hf-scs', 4.333333333333334), ('niche influences', 4.333333333333334), ('significant number', 4.333333333333334), ('bulge ablation', 4.277777777777778), ('bulge signature', 4.277777777777778), ('axin2-rtta2s*m2', 4.0), ('skin epidermis', 4.0), ('molecular markers', 4.0), ('permanent portion', 4.0), ('microarray profiling', 4.0), ('improved form', 4.0), ('visualize directly', 4.0), ('unique system', 4.0), ('reasonable reliability', 4.0), ('bi-transgenic idct', 4.0), ('divide slowly', 4.0), ('genetic markers', 4.0), ('tetracycline transactivator', 4.0), ('factor', 4.0), ('wound area', 4.0), ('non-differentiated state', 4.0), ('sebaceous glands', 4.0), ('tritiated thymidine', 4.0), ('anatomically distinct', 4.0), ('two-photon microscope', 4.0), ('lineage tracing', 4.0), ('established markers', 4.0), ('drug treatment', 4.0), ('jackson laboratory', 4.0), ('dox administration', 4.0), ('stem-cell systems', 4.0), ('past decade', 4.0), ('switch rapidly', 4.0), ('wound healing', 4.0), ('exclusively responsible', 4.0), ('mesenchyme influences', 4.0), ('self-contained mini-organ', 4.0), ('anagen cycle', 3.95), ('histone h2b', 3.9), ('bulge relative', 3.7777777777777777), ('supplementary fig', 3.75), ('damaged epidermis', 3.75), ('epidermis uniformly', 3.75), ('ifnar−/−', 3.6), ('fanca−/−', 3.6), ('sca-1−/−', 3.6), ('researchers lack', 3.5), ('tert exerts', 3.5), ('progenitors capable', 3.5), ('differentiation state', 3.5), ('activity inhibited', 3.5), ('populations differed', 3.5), ('nmethods summary', 3.5), ('maturation states', 3.5), ('briefly self-renew', 3.5), ('facs purification', 3.5), ('clonal progeny', 3.4), ('fuchs laboratory', 3.4), ('lineage-restricted progeny', 3.4), ('repair wounds', 3.333333333333333), ('brdu pulse', 3.333333333333333), ('cell', 3.0), ('cells', 2.8181818181818183), ('gfp', 2.5833333333333335), ('mice', 2.5294117647058822), ('histone', 2.4), ('niche', 2.3333333333333335), ('bulge', 2.2777777777777777), ('bulge—', 2.2777777777777777), ('skin', 2.25), ('cycle', 2.2), ('shown', 2.0), ('teto-h2b-gfp', 2.0), ('expression', 2.0), ('label', 2.0), ('fig', 1.75), ('anagen', 1.75), ('epidermis', 1.75), ('tissue', 1.75), ('150 genes', 1.6666666666666667), ('genes —', 1.6666666666666667), ('/', 1.6), ('populations', 1.5), ('maturation', 1.5), ('h2b', 1.5), ('relative', 1.5), ('purification', 1.5), ('self-renew', 1.5), ('activity', 1.5), ('previously', 1.5), ('researchers', 1.5), ('nmethods', 1.5), ('tert', 1.5), ('base', 1.5), ('differentiation', 1.5), ('progenitors', 1.5), ('behaviour', 1.4), ('fuchs', 1.4), ('progeny', 1.4), ('brdu', 1.3333333333333333), ('repair', 1.3333333333333333), ('activated', 1.0), ('doxycycline', 1.0), ('enriched', 1.0), ('addressed', 1.0), ('play', 1.0), ('re-populated', 1.0), ('nk14-cre-er', 1.0), ('phospho-smad2', 1.0), ('knowledge', 1.0), ('krt14-h2b-gfp', 1.0), ('initiation', 1.0), ('analyse', 1.0), ('krt14-h2b', 1.0), ('k14-gfpactin', 1.0), ('turned', 1.0), ('generating', 1.0), ('7cos/', 1.0), ('represents', 1.0), ('gas6', 1.0), ('part', 1.0), ('proliferate', 1.0), ('addition', 1.0), ('identification', 1.0), ('contribution', 1.0), ('vivo', 1.0), ('monitored', 1.0), ('confirm', 1.0), ('129s-wlstm1', 1.0), ('location', 1.0), ('affecting', 1.0), ('reside', 1.0), ('reversibly', 1.0), ('visualization', 1.0), ('resulting', 1.0), ('k14h2bgfp/lef1rfp', 1.0), ('stages', 1.0), ('test', 1.0), ('well-established', 1.0), ('provide', 1.0), ('u2009c', 1.0), ('reflect', 1.0), ('type', 1.0), ('defined', 1.0), ('marks', 1.0), ('hallmark', 1.0), ('mobilized', 1.0), ('wnt-responsive', 1.0), ('home', 1.0), ('proposed', 1.0), ('feed', 1.0), ('round', 1.0), ('phospho-smad1', 1.0), ('tacs', 1.0), ('divided', 1.0), ('surrounded', 1.0), ('hypothesized', 1.0), ('successfully', 1.0), ('provided', 1.0), ('-egfp', 1.0), ('cg-tg', 1.0), ('located', 1.0), ('b6', 1.0), ('exceptions', 1.0), ('combination', 1.0), ('5-bromo-2-deoxyuridine', 1.0), ('e15', 1.0), ('isthmus', 1.0), ('mid-anagen', 1.0), ('importantly', 1.0), ('determined', 1.0), ('activation', 1.0), ('k14h2bgfp', 1.0), ('replenishing', 1.0), ('population', 1.0), ('persist', 1.0), ('understand', 1.0), ('1g', 1.0), ('vicinity', 1.0), ('kinetics', 1.0), ('process', 1.0), ('epithelium', 1.0), ('wlsflox', 1.0), ('prompted', 1.0), ('47efu/', 1.0), ('1lan/', 1.0), ('fzds', 1.0), ('rosa26-mtmgflox', 1.0), ('learned', 1.0), ('axin2-rtta', 1.0), ('means', 1.0), ('luo/', 1.0), ('tcf4', 1.0), ('quiescence', 1.0), ('driven', 1.0), ('nfib', 1.0), ('tg', 1.0), ('mammals', 1.0), ('tissues', 1.0), ('molecularly', 1.0), ('tenascin', 1.0), ('integral', 1.0), ('led', 1.0), ('1a', 1.0), ('job', 1.0), ('searching', 1.0), ('nmice', 1.0), ('constitute', 1.0), ('nk14h2bgfp', 1.0), ('26sortm4', 1.0), ('k14', 1.0), ('expressing', 1.0), ('effects', 1.0), ('junction', 1.0), ('onset', 1.0), ('expresses', 1.0), ('k5-tta', 1.0), ('h2bgfp', 1.0), ('generated', 1.0), ('xa0days', 1.0), ('obtained', 1.0), ('intriguingly', 1.0), ('egfp', 1.0), ('p18', 1.0), ('runx1', 1.0), ('h2b-gfp', 1.0), ('transition', 1.0), ('ability', 1.0), ('sox9', 1.0), ('tcf-3', 1.0), ('tta', 1.0), ('rare', 1.0), ('numerous', 1.0), ('k5tta', 1.0), ('study', 1.0), ('contribute', 1.0), ('2c', 1.0), ('ephrin', 1.0), ('bmp', 1.0), ('pigmentation', 1.0), ('cases', 1.0), ('lef1-rfp', 1.0), ('performed', 1.0), ('subsets', 1.0), ('mediated', 1.0), ('teto-hist1h2bj/gfp', 1.0), ('venture', 1.0), ('evident', 1.0), ('tetracyclines', 1.0), ('identified', 1.0), ('determining', 1.0), ('tcf3', 1.0), ('subsequently', 1.0), ('germline', 1.0), ('bred', 1.0), ('present', 1.0), ('position', 1.0), ('threefold', 1.0), ('inactive', 1.0), ('rosa', 1.0), ('generate', 1.0), ('gradually', 1.0), ('u2009a4', 1.0), ('tre-h2b', 1.0), ('nk14-h2bgfp', 1.0), ('studies', 1.0), ('gt', 1.0), ('injury', 1.0), ('e13', 1.0), ('k14-h2bgfp', 1.0), ('issue', 1.0), ('turn', 1.0), ('effect', 1.0), ('actb-tdtomato', 1.0), ('finding', 1.0), ('segment', 1.0), ('retained', 1.0), ('[', 0), ('32', 0), (']', 0), ('5', 0), ('30', 0), ('19', 0)]\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Paper Title: MUSCLE: multiple sequence alignment with high accuracy and high throughput \n",
      "\n",
      "Paper Keywords: 10.1093/nar/gkh340 \n",
      "\n",
      " PaperID: [('maximum likelihood tree-building pipeline workflow referred', 31.5), ('100 randomly selected full-length genomic copies', 22.5), ('characterized full-length lanb proteins', 14.833333333333334), ('13 mitochondrial protein coding genes', 12.285714285714285), ('maximum likelihood estimation', 12.0), ('ha sequence data sets', 11.88888888888889), ('aligning amino acid sequences', 11.008771929824562), ('0b36 tree-building tool', 10.5), ('muscle protein-based codon alignments', 10.373106060606059), ('amino acid translations', 9.666666666666668), ('consensus secondary structure', 9.0), ('u2009mm isopropyl-&bgr', 9.0), ('implementing clustal omega', 9.0), ('amino acid alignment', 9.0), ('custom shell scripts', 9.0), ('selected pyrin domains', 9.0), ('ncbi flu database', 9.0), ('snake venom toxins', 9.0), ('spab c-terminal domains', 9.0), ('multiple sequence comparison', 8.797979797979798), ('tc1-mariner transposable elements', 8.5), ('arb software package', 8.5), ('delimited coding sequences', 8.342105263157894), ('ete3 build -', 8.333333333333334), ('genetics computer group', 8.333333333333334), ('approximate maximum-likelihood trees', 8.25), ('multiple alignment biases', 8.242424242424242), ('individual cluster-reference alignments', 8.227272727272727), ('construct multiple alignments', 8.136363636363637), ('multiple sequence alignment', 8.131313131313131), ('multiple sequence alignments', 8.025252525252524), ('cross-species phylogenetic analysis', 7.9523809523809526), ('nucleotide sequence based', 7.888888888888889), ('manually adjusted based', 7.75), ('orthogroup amino-acid alignments', 7.727272727272727), ('initial contigs generated', 7.7), ('geneious pro version 8', 7.25), ('generated multiple alignments', 6.836363636363636), ('multiple alignments', 5.136363636363637), ('sequence alignments', 5.116161616161616), ('codon positions', 5.0), ('bootstrap analysis', 4.666666666666666), ('phylogenetic trees', 4.535714285714286), ('protein alignments', 4.512987012987013), ('orthologous genes', 4.5), ('geneious r6', 4.5), ('paml package', 4.5), ('transcripts based', 4.5), ('40 tc1-mariner families', 4.5), ('amino-acid sequences', 4.342105263157895), ('nucleotide sequences', 4.342105263157895), ('fasta -', 4.333333333333334), ('multiple-sequence alignment', 4.333333333333334), ('proteins present', 4.333333333333334), ('reference alignment', 4.333333333333334), ('phylogenetic reconstruction', 4.285714285714286), ('manually corrected', 4.25), ('manually refined', 4.25), ('large alignments', 4.227272727272727), ('edited based', 4.166666666666667), ('protein sequences', 4.12781954887218), ('manual editing', 4.0), ('functional rbps', 4.0), ('substitution rates', 4.0), ('obtaining nt', 4.0), ('ks estimates', 4.0), ('121 gut genomes', 4.0), ('hky model', 4.0), ('disulphide bonds', 4.0), ('pairwise comparisons', 4.0), ('subsequently retrieved', 4.0), ('intergenic regions', 4.0), ('subsequently merged', 4.0), ('na genes', 4.0), ('transcript assemblies', 4.0), ('ancestral reconstruction', 4.0), ('assemblies produced', 4.0), ('historical activity', 4.0), ('default setting', 4.0), ('default settings', 4.0), ('checked visually', 4.0), ('structural superposition', 4.0), ('default parameters', 4.0), ('manually edited', 3.916666666666667), ('translated sequences', 3.8421052631578947), ('ucp1 sequences', 3.8421052631578947), ('peptide sequences', 3.8421052631578947), ('collected sequences', 3.8421052631578947), ('eggnog41 -', 3.8333333333333335), ('phylogenetic tree', 3.7857142857142856), ('gene family', 3.75), ('unigenes generated', 3.7), ('100 bootstrap replicates', 3.666666666666667), ('prank alignments', 3.5606060606060606), ('ete v3', 3.5), ('tree building', 3.5), ('codeml program', 3.5), ('presented structures', 3.5), ('mafft v5', 3.5), ('muscle algorithm', 3.145833333333333), ('muscle plugin', 3.145833333333333), ('selected', 3.0), ('muscle version 3', 2.895833333333333), ('muscle program', 2.645833333333333), ('data', 2.5), ('ha', 2.5), ('aligning', 2.5), ('group', 2.3333333333333335), ('alignment', 2.3333333333333335), ('proteins', 2.3333333333333335), ('alignments', 2.227272727272727), ('sequences', 1.8421052631578947), ('version 3', 1.75), ('family', 1.75), ('generated', 1.7), ('edited', 1.6666666666666667), ('000 replicates', 1.6666666666666667), ('6 program', 1.5), ('tree', 1.5), ('eggnog41', 1.5), ('na', 1.5), ('program', 1.5), ('mafft', 1.5), ('v3', 1.5), ('structures', 1.5), ('prank', 1.3333333333333333), ('muscle 3', 1.1458333333333333), ('muscle', 1.1458333333333333), ('0 method', 1.0), ('request', 1.0), ('curation', 1.0), ('ref', 1.0), ('estscan', 1.0), ('basis', 1.0), ('sites', 1.0), ('clustalw', 1.0), ('homologues', 1.0), ('m-coffee', 1.0), ('accelrys', 1.0), ('removed', 1.0), ('gaps', 1.0), ('induced', 1.0), ('effects', 1.0), ('one-to-', 1.0), ('replicated', 1.0), ('created', 1.0), ('authors', 1.0), ('aac', 1.0), ('trimai', 1.0), ('obtained', 1.0), ('gametologues', 1.0), ('phylogeny', 1.0), ('v4', 1.0), ('guide', 1.0), ('frameshifts', 1.0), ('dnaml', 1.0), ('taxon', 1.0), ('expression', 1.0), ('test', 1.0), ('berghei', 1.0), ('pairs', 1.0), ('errors', 1.0), ('set', 1.0), ('columns', 1.0), ('classes', 1.0), ('reconstructed', 1.0), ('cleaned', 1.0), ('study', 1.0), ('performed', 1.0), ('2015 outbreak', 1.0), ('downloaded', 1.0), ('aline', 1.0), ('executed', 1.0), ('templates', 1.0), ('concatenated', 1.0), ('align', 1.0), ('hmmer', 1.0), ('u2009h', 1.0), ('class', 1.0), ('cogs', 1.0), ('identified', 1.0), ('my_sequences', 1.0), ('mrbayes 3', 1.0), ('u2009°', 1.0), ('ebov', 1.0), ('gblocks', 1.0), ('rpl9', 1.0), ('validated', 1.0), ('phylip', 1.0), ('iptg', 1.0), ('generate', 1.0), ('/', 1.0), ('annotated', 1.0), ('log-expectation', 1.0), ('4c', 1.0), ('constructed', 1.0), ('conducted', 1.0), ('segments 4', 1.0), ('viewed', 1.0), ('-d-thiogalactoside', 1.0), ('phyml 3', 1.0), ('cog', 1.0), ('analysed', 1.0), ('jalview', 1.0), ('gamma-distribution', 1.0), ('106 iterations', 1.0), ('confirmed', 1.0), ('mega 4', 1.0), ('aligned', 1.0), ('knowlesi', 1.0), ('extracted', 1.0), ('estimate', 1.0), ('fasttree', 1.0), ('results/', 1.0), ('mrbayes', 1.0), ('31', 0), ('>90%', 0), ('2014', 0), ('5', 0), ('18', 0), ('[', 0), ('8', 0), ('1', 0), ('16', 0), ('7', 0), ('6', 0), ('2', 0), ('0', 0), (']', 0), ('19', 0)]\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for index, paper in test_df.iterrows():\n",
    "    \n",
    "    print('\\n\\n')\n",
    "    print('Paper Title: %s \\n\\nPaper DOI: %s \\n\\n Paper Keywords: %s\\n\\n' % (paper['Title'],paper['DOI'],\n",
    "                                                                          str(rake.run(paper['clean_text']))))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['[\\'The 5′ element, for which we suggest the name HOTTIP for ‘HOXA transcript at the distal tip’, exhibits bivalent H3K4me3 and H3K27me3, a histone modification pattern associated with poised regulatory sequences.\\', \\'The combination of both activating (H3K4 methylation) and inactivating (H3K27 methylation) chromatin marks characterizes ‘poised’ promoters and enhancers, which regulate expression of key developmental genes.\\', \\'H3K4me is an activating mark found in promoters/enhancers of key developmental genes.\\', \\'These findings establish that mutations in many genes in the H3K4me–H3K27me pathway disrupt cardiac development and are consistent with previous evidence implicating these chromatin marks in regulating key developmental genes, including those involved in cardiac development.\\', \\'Lysine 27 is a critical residue of histone 3 and its variants, and methylation at this position (H3K27me), which may be mimicked by the terminal CH3 of methionine substituted at this residue, is commonly associated with transcriptional repression.\\', \\'These advances include insights into the intrinsic mechanisms and the niche interactions that regulate the fate of embryonic stem (ES) cells and adult stem cells.\\', \\'The maps show close agreement with our previously reported ChIP-chip data from ∼2.5% of the mouse genome (Fig.\\', \\'Recently, we and others observed that some promoters in ES cells carry both H3K4me3 and H3K27me3.\\', \\'A minority of these (n = 564) are ‘wide’ bivalent sites in which H3K27me3 extends over a region of at least 5\\\\u2009kb and resemble those described previously.\\', \\'\\\\n\\', \\'\\\\n\\', \\'2b), raising the possibility that H3K4me3 may poise these promoters for activation.\\', \\'The first group is associated with bivalent domains, a chromatin state characterized by the presence of both H3K4me3 and H3K27me3.\\', \\'First, we searched for in vivo evidence of promoters embedded within SHANK3 by integrating sequence conservation (ECRs), evidence of transcription initiation in both mouse and human tissues (CAGE tags), the presence of H3K4me3 in the human brain as well as overlapping H3K4me3 and H3K27me3 peaks from ChIP-seq analyses of embryonic stem (ES) cells.\\', \\'Evolutionarily conserved regions (ECRs) overlap with mouse CAGE tag clusters (arrows), mouse ES H3K4me3 and H3K27me3 bivalent domains and human frontal cortex H3K4me3.\\', \\'4).\\', \\'Repression of the resident cells’ program in turn might be mediated by the capacity of ES cell regulators to actively silence differentiation-affiliated transcription factors, such as through recruitment of polycomb complexes and formation of bivalent chromatin domains.\\', \\'A subset of less active genes is marked by a combination of H3K4me3 and H3K27me3, defining them as bivalent (Fig.\\', \\'(rlu@stanford.edu).\\\\r\\\\n\\\\r\\\\n&e091119-13; &nature08575-s1;\\\\r\\\\n\\\\n\\\\nMolecular regulation of embryonic stem cell (ESC) fate involves a coordinated interaction between epigenetic, transcriptional and translational mechanisms.\\', \\'Bivalent domains, simultaneously enriched for both the activating H3K4me3 mark and repressive H3K27me3 mark, are largely devoid of Chd1 (Fig.\\', \\'ES cells have been reported to be ‘poised’ for differentiation by the presence of bivalent domains (marked by H3K4me3 and H3K27me3) in developmental regulatory genes.\\', \\'Recent evidence suggests that PcG proteins function to transcriptionally repress developmental genes in ES cells, the expression of which would otherwise promote differentiation.\\', \\'ES cells and induced pluripotent stem cells (iPS cells) show enrichment of TssBiv, consistent with previous studies, and a depletion of ReprPCWk (defined by weak H3K27me3), possibly due to restriction of H3K27me3-establishing Polycomb proteins to promoter regions.\\', \\'\\\\n\\', \\'In differentiating embryonic stem cells, chromatin at lineage-restricted genes becomes sequentially accessible, probably by means of ‘pioneer’ transcription factor activity, but tissues may use other strategies in vivo.\\', \\'As H3K4 is methylated at both poised and active cis-elements in embryonic stem (ES) cells, the striking overlap in progenitor cell marks might reflect enhancer potential, distinct from enhancer activity; H3K27ac is thought to mark active enhancers specifically.\\', \\'Faulty restoration of bivalent domains, which mark developmental loci with both active and repressive histone modifications, and loss of pioneer factors, which in fESCs and iPSCs occupy enhancers of genes expressed only in differentiated cells, represent two other potential mechanisms.\\', \\'In ES cells, HCPs at housekeeping genes are enriched with the transcription initiation mark H3K4me3 (‘univalent’) and are generally highly expressed, whereas those at developmental genes are enriched with both H3K4me3 and the repressive mark H3K27me3 (‘bivalent’) and are generally silent.\\', \\'Out of the 10,299 HCPs sampled (on average, 19 distinct CpGs per promoter), we found that virtually all contain a core region of unmethylated CpGs, regardless of their level of expression or H3K27me3 enrichment (Figs 1 and 2a).\\', \"These \\'bivalent domains\\' correlate with locations of genes for transcription factors that regulate development.\", \\'STAP stem cells (as well as STAP cells) had accumulation patterns of H3K4 and H3K27 trimethylation that resembled those of ES cells at the loci of pluripotency marker genes (Oct4, Nanog, Sox2), bivalent pattern genes (Gata2, brachyury, Nkx6-2) and trophoblast marker genes (Cdx2, Eomes, Itga7).\\', \\'Notably, the chromatin at many of these non-expressed target genes adopts a bivalent conformation in ES cells, carrying both the ‘active’ histone H3 lysine 4 (H3K4) methylation mark and the ‘repressive’ histone H3 lysine 27 (H3K27) methylation mark.\\', \\'We used chromatin immunoprecipitation (ChIP) and real-time PCR to quantify H3K4 and H3K27 methylation for a set of genes reported to be bivalent in pluripotent ES cells.\\', \\'For example, bivalent chromatin domains (both histone 3 lysine 4 trimethylation and histone 3 lysine 27 trimethylation) mark genes that are poised to play key parts in subsequent lineage decisions.\\', \\'Interestingly, among all the putative enhancers captured by RNAPII-bound interactions, we found 563 potential ‘poised enhancers’ connecting to the bivalent promoters.\\', \\'We hypothesized that these molecular changes upon EZH1 knockdown were mediated by bivalent, or poised, chromatin domains, often implicated in the control of developmentally regulated genes.\\', \\'This epigenetic plasticity may be an inherent property of the stem cells from which cancers arise, for example the bivalent nature of adjacent H3K4 and H3K27 methylation that is seen at many genomic sites in stem cells but not in somatic cells after differentiation.\\', \\'In ES cells, these proteins promote self-renewal and also work with repressive polycomb complexes (PRC2; containing core component SUZ12) to help repress a large set of developmental regulators (including HOX genes) to prevent differentiation.\\', \\'Although none of the histone marks taken alone explain the silencing of the locus, when taken together they are reminiscent of the silent chromatin configuration characteristic of ES cells and known as ‘bivalent’ in which repressive (H3K27me3) and active (H3K4me3) marks coexist in the same molecule.\\', \\'About half of the bivalent domains in ES cells are associated with binding sites for Oct4, Sox2 or Nanog.\\', \\'Sequential ChIP experiments were carried out essentially as described earlier with modifications previously reported.\\', \\'Bivalent domains, which are co-occupied by the H3K4me3 and H3K27me3 marks, were first described in ES cells and were postulated to poise developmental regulator genes for later transcriptional activation upon differentiation.\\', \\'The PRE-proximal paused TSSs have some similarity to the ‘bivalent’ genes in mammalian cells, which also display transcriptional pausing of key regulatory and developmental genes.\\', \\'Most human gene promoters are embedded within CGIs that lack DNA methylation and coincide with sites of histone H3 lysine 4 trimethylation (H3K4me3), irrespective of transcriptional activity.\\', \\'Agreement between these results and genome-wide studies of chromatin modifications indicated that this fraction could be used to identify proteins that preferentially localize to non-methylated CGIs.\\', \\'The positions of CGIs and H3K4me3 peaks at this locus in wild-type ES cells are shown below the map.\\', \\'CGIs and H3K4me3 peaks at this locus in wild-type ES cells are shown below.\\', \"Notably, the H3K4me3 mark, often associated with active transcription, was present at most, if not all, PRC2-targeted genes in ES cells, forming the \\'bivalent domain\\'.\", \\'Although this pattern was initially believed to be ES-cell specific, bivalent domains have been found in differentiated somatic cells, albeit at a lower frequency; they were also found in zebrafish but are rarely detected in Drosophila.\\', \\'Furthermore, group 2 regions in H1 frequently displayed both H3K4me3 and H3K27me3 modifications, characteristic of the bivalent state that is thought to instil a suppressed but poised transcriptional status.\\', \\'Histone lysine methylation is an integral part of the post-translational modifications of histone tails that are important for chromatin organization and regulation of gene transcription.\\', \\'The signalling, transcriptional and chromatin regulatory networks controlling pluripotency have been well characterized, providing an ideal system to determine how lincRNAs may integrate into these processes.\\', \\'We previously showed that many human lincRNAs can interact with the polycomb repressive complex, a complex that has a critical functional role in the regulation of ES cells.\\', \\'To determine if lincRNAs interact with additional chromatin proteins, we systematically analysed chromatin-modifying proteins that have been shown to have critical roles in ES cells.\\', \\'H3K27 methylation has been shown to regulate transcription of tumour suppressor genes, and may also be important for stem-cell pluripotency regulation.\\', \\'We studied two primary human fibroblast cell types (foot and lung) that preserve distinct patterns of embryonic HOX expression, as well as mouse ESCs, in which HOX genes are silent.\\', \\'Recent studies showed that a large set of developmentally important genes in ESCs are enriched for both H3K27me3 and H3K4me3 (bivalent domains), which represent the repressive and activating histone modifications, respectively.\\', \\'Such bivalent chromatin domains also exist in embryonic stem cells and are thought to poise genes for activation while keeping them repressed.\\', \\'The inactive genes were of particular interest, because studies in embryonic stem (ES) cells found that developmental regulatory genes are often inactive and associated with H3K4me3/H3K27me3 domains.\\', \\'Gene Ontology (GO) analysis showed that these inactive genes were enriched for those annotated as ‘DNA binding’, ‘transcription factor activity’ and ‘multicellular organismal development’, with P-values of 3.3\\\\u2009×\\\\u200910-13, 8.9\\\\u2009×\\\\u200910-12 and 4.1\\\\u2009×\\\\u200910-6, respectively, similar to that seen for ES cells.\\', \\'Such ‘bivalent’ domains were found in ES cells and have been postulated to poise genes for activation (H3K4me3) while keeping them repressed (H3K27me3).\\', \\'The existence of such monovalent inactive genes was surprising because H3K27me3 had been postulated to prevent transcription from H3K4me3-bound promoters.\\', \\'Second, our study provides evidence in embryos for the bivalent domains found in ES cells.\\', \\'This finding reveals a shared chromatin landscape between transiently pluripotent cells in the embryo and permanently pluripotent cells in culture.\\', \\'The promoters of many lineage regulators harbour both active (H3K4me3) and repressive (H3K27me3) histone marks, a bivalent state thought to facilitate activation of development genes upon exit from pluripotency.\\', \\'Bivalent genes constitute a separate group of chromatin targets that are marked by both active H3K4 methylation and repressive H3K27 methylation in iPSCs or ES cells.\\', \\'In ES cells, genes with ‘bivalent’ H3K27 and H3K4 trimethylation are transcriptionally inactive but poised for expression upon differentiation to embryoid bodies.\\', \"Interestingly, some developmental genes are present within \\'bivalent\\' chromatin regions, which contain both inactivating marks (methylated H3K27) and activating marks (methylated H3K4).\", \\'Chromatin regions containing high levels of both H3K4me3 and H3K27me3 have been termed ‘bivalent domains’ and are silenced but thought to be ‘poised’ for activation.\\', \\'Bivalent domains mark developmentally important loci in pluripotent and multipotent cells.\\', \\'\\\\n\\', \\'The poised H3K4me3+H3K27me3+ bivalent status of developmentally regulated loci is considered to be a hallmark of the epigenetic landscape of pluripotent ESCs and iPSCs.\\', \\'These were enriched for the GO terms associated with cell-fate specification and developmental processes (Supplementary Table 2), which is consistent with previous reports.\\', \\'Finally, epigenetic reprogramming sometimes fails to properly restore bivalent domains, which mark developmental loci with active and repressive histone modifications.\\']'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_df[test_df['DOI'] == '10.1016/j.cell.2006.02.041']['clean_text'].values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Investigating Keywords for Scale Free"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ArticleID</th>\n",
       "      <th>text</th>\n",
       "      <th>clean_text</th>\n",
       "      <th>count</th>\n",
       "      <th>Title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>45845</th>\n",
       "      <td>pub.1010080128</td>\n",
       "      <td>[['Subjects were assigned a location in a netw...</td>\n",
       "      <td>[Subjects were assigned a location in a networ...</td>\n",
       "      <td>31</td>\n",
       "      <td>Emergence of Scaling in Random Networks</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            ArticleID                                               text  \\\n",
       "45845  pub.1010080128  [['Subjects were assigned a location in a netw...   \n",
       "\n",
       "                                              clean_text  count  \\\n",
       "45845  [Subjects were assigned a location in a networ...     31   \n",
       "\n",
       "                                         Title  \n",
       "45845  Emergence of Scaling in Random Networks  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matched_dfs[matched_dfs['Title'].str.contains(r'Scaling.*Random.*')]\\\n",
    "            .sort_values('count', ascending = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#rake = Rake('RAKE/SmartStoplist.txt')\n",
    "#rake.run(matched_dfs.iloc[45845]['clean_text'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>freq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>(networks, degree, distributions)</th>\n",
       "      <td>2</td>\n",
       "      <td>0.003317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(highly, connected, proteins)</th>\n",
       "      <td>2</td>\n",
       "      <td>0.003317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(networks, social, networks)</th>\n",
       "      <td>2</td>\n",
       "      <td>0.003317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(random, graphs, networks)</th>\n",
       "      <td>2</td>\n",
       "      <td>0.003317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(protein, interaction, networks)</th>\n",
       "      <td>2</td>\n",
       "      <td>0.003317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(degree, distribution, Here)</th>\n",
       "      <td>2</td>\n",
       "      <td>0.003317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(Internet, Data, Analysis)</th>\n",
       "      <td>1</td>\n",
       "      <td>0.001658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(attachment, This, implies)</th>\n",
       "      <td>1</td>\n",
       "      <td>0.001658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(PB, k, lgr)</th>\n",
       "      <td>1</td>\n",
       "      <td>0.001658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(spatial, lattices, random)</th>\n",
       "      <td>1</td>\n",
       "      <td>0.001658</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   0      freq\n",
       "(networks, degree, distributions)  2  0.003317\n",
       "(highly, connected, proteins)      2  0.003317\n",
       "(networks, social, networks)       2  0.003317\n",
       "(random, graphs, networks)         2  0.003317\n",
       "(protein, interaction, networks)   2  0.003317\n",
       "(degree, distribution, Here)       2  0.003317\n",
       "(Internet, Data, Analysis)         1  0.001658\n",
       "(attachment, This, implies)        1  0.001658\n",
       "(PB, k, lgr)                       1  0.001658\n",
       "(spatial, lattices, random)        1  0.001658"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_trigram(matched_dfs.iloc[45845]['clean_text'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ArticleID</th>\n",
       "      <th>text</th>\n",
       "      <th>clean_text</th>\n",
       "      <th>count</th>\n",
       "      <th>Title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>92203</th>\n",
       "      <td>pub.1020281098</td>\n",
       "      <td>[['This model was combined into a single prote...</td>\n",
       "      <td>[This model was combined into a single protein...</td>\n",
       "      <td>520</td>\n",
       "      <td>Coot: model-building tools for molecular graphics</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            ArticleID                                               text  \\\n",
       "92203  pub.1020281098  [['This model was combined into a single prote...   \n",
       "\n",
       "                                              clean_text  count  \\\n",
       "92203  [This model was combined into a single protein...    520   \n",
       "\n",
       "                                                   Title  \n",
       "92203  Coot: model-building tools for molecular graphics  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matched_dfs[matched_dfs['Title'].str.contains(r'.*Coot:.*')]\\\n",
    "            .sort_values('count', ascending = False)\n",
    "#[['Title', 'DOI', 'ArticleID', 'count', 'Cites']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#matched_dfs[matched_dfs['Title'].str.contains(r'.*Coot:.*')]\\\n",
    "#            .sort_values('count', ascending = False)\n",
    "#[['Title', 'DOI', 'ArticleID', 'count', 'Cites']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ArticleID</th>\n",
       "      <th>text</th>\n",
       "      <th>clean_text</th>\n",
       "      <th>count</th>\n",
       "      <th>Title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>183344</th>\n",
       "      <td>pub.1040343773</td>\n",
       "      <td>[['In addition, the design of nucleic acids th...</td>\n",
       "      <td>[In addition, the design of nucleic acids that...</td>\n",
       "      <td>20</td>\n",
       "      <td>Molecular Structure of Nucleic Acids: A Struct...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10716</th>\n",
       "      <td>pub.1002346037</td>\n",
       "      <td>[['Yet in 1953 &lt;i&gt;Nature&lt;/i&gt; published seven p...</td>\n",
       "      <td>[Yet in 1953 Nature published seven papers on ...</td>\n",
       "      <td>2</td>\n",
       "      <td>Molecular Structure of Nucleic Acids: Molecula...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             ArticleID                                               text  \\\n",
       "183344  pub.1040343773  [['In addition, the design of nucleic acids th...   \n",
       "10716   pub.1002346037  [['Yet in 1953 <i>Nature</i> published seven p...   \n",
       "\n",
       "                                               clean_text  count  \\\n",
       "183344  [In addition, the design of nucleic acids that...     20   \n",
       "10716   [Yet in 1953 Nature published seven papers on ...      2   \n",
       "\n",
       "                                                    Title  \n",
       "183344  Molecular Structure of Nucleic Acids: A Struct...  \n",
       "10716   Molecular Structure of Nucleic Acids: Molecula...  "
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matched_dfs[matched_dfs['Title'].str.contains(r'.*MOLECULAR STRUCTURE OF NUCLEIC ACIDS.*', case = False)]\\\n",
    "            .sort_values('count', ascending = False)\n",
    "#[['Title', 'DOI', 'ArticleID', 'count', 'Cites']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>freq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>(Watson, Francis, Crick)</th>\n",
       "      <td>3</td>\n",
       "      <td>0.007958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(James, Watson, Francis)</th>\n",
       "      <td>3</td>\n",
       "      <td>0.007958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(DNA, double, helix)</th>\n",
       "      <td>3</td>\n",
       "      <td>0.007958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(immediately, suggests, possible)</th>\n",
       "      <td>2</td>\n",
       "      <td>0.005305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(suggests, possible, copying)</th>\n",
       "      <td>2</td>\n",
       "      <td>0.005305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(possible, copying, mechanism)</th>\n",
       "      <td>2</td>\n",
       "      <td>0.005305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(copying, mechanism, genetic)</th>\n",
       "      <td>2</td>\n",
       "      <td>0.005305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(DNA, Until, however)</th>\n",
       "      <td>1</td>\n",
       "      <td>0.002653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(adenine, forms, hydrogen)</th>\n",
       "      <td>1</td>\n",
       "      <td>0.002653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(seven, papers, structure)</th>\n",
       "      <td>1</td>\n",
       "      <td>0.002653</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   0      freq\n",
       "(Watson, Francis, Crick)           3  0.007958\n",
       "(James, Watson, Francis)           3  0.007958\n",
       "(DNA, double, helix)               3  0.007958\n",
       "(immediately, suggests, possible)  2  0.005305\n",
       "(suggests, possible, copying)      2  0.005305\n",
       "(possible, copying, mechanism)     2  0.005305\n",
       "(copying, mechanism, genetic)      2  0.005305\n",
       "(DNA, Until, however)              1  0.002653\n",
       "(adenine, forms, hydrogen)         1  0.002653\n",
       "(seven, papers, structure)         1  0.002653"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_trigram(matched_dfs.iloc[183344]['clean_text'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"In addition, the design of nucleic acids that bind to other nucleic acids with high affinity and specificity follows the simple and long-established rules of Watson–Crick base pairing. The discovery of the double helix in 1953 immediately raised questions about how biological information is encoded in DNA. As implied by James Watson and Francis Crick in their landmark paper, base pairing suggests a template-copying mechanism that accounts for the fidelity in copying of genetic material during DNA replication (see article in this issue by Alberts, page 431). A DNA duplex for redundancy\\nWatson and Crick noted, with infamous prophetic understatement, “it has not escaped our notice that the specific [base] pairing we have postulated immediately suggests a possible copying mechanism for the genetic material”. Following the discovery of DNA as the molecular substrate of inheritance and determination of its structure, this speculation was supported by the demonstration that agents that damage DNA and generate mutations also cause cancer. The information gleaned from this scientific endeavour continues to have a profound influence on our understanding of biological systems. The five files comprised all 154 of Shakespeare’s sonnets (ASCII text), a classic scientific paper (PDF format), a medium-resolution colour photograph of the European Bioinformatics Institute (JPEG 2000 format), a 26-s excerpt from Martin Luther King’s 1963 ‘I have a dream’ speech (MP3 format) and a Huffman code used in this study to convert bytes to base-3 digits (ASCII text), giving a total of 757,051 bytes or a Shannon information of 5.2\\u2009×\\u2009106 bits (see Supplementary Information and Supplementary Table 1 for full details). \\n The 50-year journey of the DNA molecule from the reticent line diagram in Watson and Crick's seminal article (Fig. Yet in 1953 Nature published seven papers on the structure and function of DNA, but only one national British newspaper — the News Chronicle — referred to the double helix (see facsimile below). Chief among its “novel features” of “considerable biological interest”, Watson and Crick described the pairing of the bases, where adenine forms hydrogen bonds with thymine, and guanine with cytosine. This pairing, they wrote, “immediately suggests a possible copying mechanism for the genetic material.” Expanding on this in a subsequent paper appearing in Nature a month later, they wrote of DNA: “Until now, however, no evidence has been presented to show how it might carry out the essential operation required of a genetic material, that of exact self-duplication.” And what better emblem or mascot is there for molecular biology than the double helix, and its spartan yet elegant representation in the original paper from the pen of Odile Crick, Francis's wife, fifty years ago? 1953 James Watson and Francis Crick, Rosalind Franklin and Raymond Gosling, Maurice Wilkins, W. E. Seeds, Alec Stokes and Herbert Wilson, and Bertil Jacobson all publish on the structure of DNA. In DNA, the double helix arises from the existence of two competing length scales, one set by the distance between monomers in the sugar backbone, and the other set by the stacking of the base pairs. The model for the DNA double helix proposed by James Watson and Francis Crick is based on two paired DNA strands that are complementary in their nucleotide sequence. The essential properties of the human genome sequence — accuracy and completeness — reflect intrinsic features of the molecular structure of DNA. Recognition of DNA as the hereditary material, determination of its structure, elucidation of the genetic code, development of recombinant DNA technologies, and establishment of increasingly automatable methods for DNA sequencing set the stage for the Human Genome Project (HGP) to begin in 1990 (see also http://www.nature.com/nature/DNA50). \\n The replication process exploits the complementarity of the four-letter code of bases in the linear DNA double helix, where adenine (A) pairs with thymine (T) and guanine (G) pairs with cytosine (C).\""
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "' '.join(matched_dfs.iloc[183344]['clean_text'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## RAKE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('experimentally phased ncs averaged electron density map permitted',\n",
       "  32.08662587412587),\n",
       " ('protein data bank accession number 4r12', 26.850215432824125),\n",
       " ('protein data bank accession number 4hyc', 26.850215432824125),\n",
       " ('protein data bank accession number 3mku', 26.850215432824125),\n",
       " ('resulting maps showed clear unbiased density', 24.68044871794872),\n",
       " ('selenium anomalous difference fourier peak', 22.0),\n",
       " ('highly reliable initial rigid body fit', 21.57804459691252),\n",
       " ('multi-domain averaged electron density maps', 20.925641025641024),\n",
       " ('iterative process involving careful model building', 20.778171702225674),\n",
       " ('two-fold symmetry averaging produced experimental maps',\n",
       "  19.948484848484846),\n",
       " ('2\\u2009å resolution sigmaa-weighted difference fourier maps',\n",
       "  18.833333333333332),\n",
       " ('composite annealed omit maps calculated', 18.409523809523808),\n",
       " ('coot applying ten-fold ncs operations', 18.004032258064516),\n",
       " ('resulting phase-combined electron density map', 16.836625874125872),\n",
       " ('pdb accession number 3a1b', 16.833333333333332),\n",
       " ('pdb accession number 2qrv', 16.833333333333332),\n",
       " ('native eif3d data set processed', 16.57647757647758),\n",
       " ('maximum likelihood restrained positional refinement', 16.15317778587921),\n",
       " ('fo\\u2009−\\u2009fc electron density maps', 16.116550116550115),\n",
       " ('poorly modelled loop spanning residues 821', 16.101754385964913),\n",
       " ('simulated annealing composite omit maps', 16.035897435897436),\n",
       " ('van der waals contacts', 16.0),\n",
       " ('default tight non-crystallographic symmetry restraints',\n",
       "  15.712121212121211),\n",
       " ('manually placing idealized poly-ala helices', 15.481481481481481),\n",
       " ('final models displaying excellent geometry', 14.884578223708658),\n",
       " ('initial experimental electron density map', 14.852170471038397),\n",
       " ('rcsb protein data bank', 14.816882099490794),\n",
       " ('protein data bank 2hue', 14.816882099490794),\n",
       " ('x-ray holographic image reconstruction', 14.666666666666666),\n",
       " ('employed nucleic acid scaffold', 14.25),\n",
       " ('rigid bodies deviated substantially', 14.0),\n",
       " ('fo\\u2009−\\u2009fc omit maps', 13.624242424242425),\n",
       " ('extended data figs 6b', 13.527027027027028),\n",
       " ('solvent-flattened electron density map', 13.274125874125874),\n",
       " ('canonical duplex b-form dna', 13.166666666666666),\n",
       " ('resultant density-modified experimental map', 13.113636363636363),\n",
       " ('highly idealized alanine &agr', 13.066666666666666),\n",
       " ('high-resolution p42 native data', 12.884169884169886),\n",
       " ('apo-smoδc polypeptide chain', 12.75),\n",
       " ('experimentally phased map', 12.681818181818182),\n",
       " ('excellent electron density map', 12.674125874125872),\n",
       " ('molecular dynamic flexible fitting', 12.634615384615385),\n",
       " ('fourfold multidomain ncs averaging', 12.479166666666666),\n",
       " ('experimental electron density map', 12.455944055944055),\n",
       " ('phenix applying anisotropic scaling', 12.397435897435896),\n",
       " ('model closed promoter dna', 12.209562109025915),\n",
       " ('disulphide bond restraint added', 12.153846153846153),\n",
       " ('extended data table 1a', 12.142411642411643),\n",
       " ('crystallographic refinement involved repeated cycles', 11.97343723657344),\n",
       " ('fimd usher translocation domain', 11.933333333333334),\n",
       " ('secondary structure elements aligned', 11.885416666666666),\n",
       " ('resulting electron density map', 11.836625874125874),\n",
       " ('initial electron density map', 11.670352289220213),\n",
       " ('derived tarp homology model', 11.54289544235925),\n",
       " ('sbsb molecules  encompassing residues 202', 11.511278195488721),\n",
       " ('electron density map manually', 11.338940688940689),\n",
       " ('individual atomic displacement parameters', 11.336764705882352),\n",
       " ('crystallographic refinement software package phenix', 10.967405225106646),\n",
       " ('simulated-annealing composite omit', 10.933333333333334),\n",
       " ('protein data bank', 10.816882099490794),\n",
       " ('molecular replacement program balbes', 10.813186813186814),\n",
       " ('higher resolution apo structure', 10.802083333333334),\n",
       " ('molecular replacement solution consisting', 10.741758241758241),\n",
       " ('computer graphic program coot', 10.679627496159753),\n",
       " ('pdb accession 3zmr', 10.633333333333333),\n",
       " ('interactive graphics program coot', 10.429627496159753),\n",
       " ('tls refinement techniques adopting', 10.268562401263823),\n",
       " ('obtained head module model', 10.158280057743864),\n",
       " ('canonical duplex b-dna', 10.0),\n",
       " ('two-fold non-crystallographic symmetry', 9.712121212121211),\n",
       " ('computer graphics program coot', 9.679627496159753),\n",
       " ('real space model building', 9.637430961484933),\n",
       " ('cerevisiae head module', 9.5),\n",
       " ('nitrogen bond length', 9.5),\n",
       " ('automated real-space refinement cycles', 9.427982691118896),\n",
       " ('electron density maps', 9.425641025641026),\n",
       " ('excellent geometrical characteristics', 9.4),\n",
       " ('final search model based', 9.349054862649105),\n",
       " ('subsequent manual model building', 9.291816926397212),\n",
       " ('experimental electron density', 9.274125874125874),\n",
       " ('electron density map', 9.274125874125874),\n",
       " ('iterative manual model building', 9.19922433380462),\n",
       " ('three-fold symmetry-averaged map', 9.181818181818182),\n",
       " ('map sharpening utility', 9.181818181818182),\n",
       " ('map sharpening utilities', 9.181818181818182),\n",
       " ('individual restrained b-factor refinement', 9.156965664667087),\n",
       " ('tfg1 charged helix', 9.0),\n",
       " ('saccharomyces cerevisiae sam50', 9.0),\n",
       " ('ligand-bound fluoride riboswitch', 9.0),\n",
       " ('deformable elastic network', 9.0),\n",
       " ('2fo\\u2009−\\u2009fc maps', 8.995670995670997),\n",
       " ('em density map', 8.966433566433565),\n",
       " ('electron density interpretation', 8.892307692307693),\n",
       " ('high-resolution native data', 8.884169884169886),\n",
       " ('yielding excellent geometry', 8.844444444444445),\n",
       " ('molecular graphics program coot', 8.81424288077514),\n",
       " ('pdb code 4gwp', 8.8),\n",
       " ('pdb accessions 4iuj', 8.8),\n",
       " ('pdb code 4oud', 8.8),\n",
       " ('resolved helical density', 8.784615384615385),\n",
       " ('native-1 data set', 8.71933471933472),\n",
       " ('test data set', 8.71933471933472),\n",
       " ('atomic displacement parameters', 8.711764705882352),\n",
       " ('density-modified maps produced', 8.683333333333334),\n",
       " ('grade web server', 8.666666666666666),\n",
       " ('weighted 2fo\\u2009−\\u2009fc', 8.662337662337663),\n",
       " ('-weighted 2fo\\u2009−\\u2009fc', 8.662337662337663),\n",
       " ('global 6d search', 8.625),\n",
       " ('3\\u2009å native data set', 8.576477576477576),\n",
       " ('native data set', 8.576477576477576),\n",
       " ('0\\u2009å native data set', 8.576477576477576),\n",
       " ('custom solvent mask', 8.571428571428571),\n",
       " ('000 adomet radical enzymes', 8.5),\n",
       " ('density-modified electron-density map', 8.431818181818182),\n",
       " ('protein model built manually', 8.427063854711578),\n",
       " ('secondary structural elements', 8.369047619047619),\n",
       " ('pdb 1u7g placing', 8.3),\n",
       " ('bulk solvent scaling', 8.238095238095237),\n",
       " ('electron density calculated', 8.235164835164834),\n",
       " ('non-crystallographic six-fold symmetry', 8.212121212121211),\n",
       " ('ago ternary complex', 8.1875),\n",
       " ('graphical map inspection', 8.181818181818182),\n",
       " ('rigid body fit', 8.181818181818182),\n",
       " ('added step-by-step mg8', 8.153846153846153),\n",
       " ('extended data table 1', 8.142411642411643),\n",
       " ('686 amino acid residues', 8.118421052631579),\n",
       " ('electron density inspection', 8.092307692307692),\n",
       " ('avoid model bias', 8.04289544235925),\n",
       " ('minimize model bias', 8.04289544235925),\n",
       " ('single protein chain', 8.039855072463768),\n",
       " ('bridge helix region', 8.0),\n",
       " ('metagenome sequence d1jbk2-4', 8.0),\n",
       " ('eisenberg hydrophobicity scale', 8.0),\n",
       " ('cryo-em map density', 7.966433566433566),\n",
       " ('original protein sequence', 7.956521739130435),\n",
       " ('nucleic acid models', 7.9423076923076925),\n",
       " ('initial cα model', 7.939121857453589),\n",
       " ('readily interpretable map', 7.848484848484849),\n",
       " ('electron density allowed', 7.842307692307692),\n",
       " ('single tls body', 7.833333333333334),\n",
       " ('partially proteolysed proteins', 7.833333333333333),\n",
       " ('component proteins trafct', 7.833333333333333),\n",
       " ('gsh bound structure', 7.802083333333333),\n",
       " ('flat density present', 7.7846153846153845),\n",
       " ('torsion-angle simulated annealing', 7.76923076923077),\n",
       " ('riboflavin-binding protein structure', 7.758605072463768),\n",
       " ('multiple translation-libration-screw rotation', 7.75),\n",
       " ('rna hydrogen bonds', 7.714285714285714),\n",
       " ('bulk solvent correction', 7.7142857142857135),\n",
       " ('heavy non-crystallographic symmetry', 7.712121212121211),\n",
       " ('automated model building', 7.704097628151599),\n",
       " ('monomeric psi complex', 7.6875),\n",
       " ('conjugate-gradient energy minimization', 7.666666666666668),\n",
       " ('asymmetric unit cell', 7.666666666666667),\n",
       " ('ssm superposition routine', 7.666666666666667),\n",
       " ('diffraction anisotropy server', 7.666666666666666),\n",
       " ('improved electron density', 7.654807692307692),\n",
       " ('final sbsb structure', 7.6499094202898545),\n",
       " ('visual molecular dynamics', 7.634615384615385),\n",
       " ('subsequently re-built automatically', 7.6190476190476195),\n",
       " ('remaining structural elements', 7.619047619047619),\n",
       " ('jelly body refinement', 7.601895734597156),\n",
       " ('rigid body refinement', 7.601895734597156),\n",
       " ('secondary structure matching', 7.552083333333333),\n",
       " ('h4 tetramer complex', 7.520833333333334),\n",
       " ('higher resolution structure', 7.468749999999999),\n",
       " ('protein side chains', 7.456521739130435),\n",
       " ('nbkb6 crystal collected', 7.444444444444445),\n",
       " ('crude helical model', 7.44289544235925),\n",
       " ('initial output model', 7.439121857453589),\n",
       " ('intermittent model building', 7.370764294818265),\n",
       " ('built de novo', 7.3628318584070795),\n",
       " ('meta\\u2009ii models', 7.358974358974359),\n",
       " ('topo ii&dgr', 7.333333333333333),\n",
       " ('four-fold ncs restraints', 7.3125),\n",
       " ('three-fold ncs restraints', 7.3125),\n",
       " ('additional missing residues', 7.285087719298245),\n",
       " ('maximum-likelihood density modification', 7.2846153846153845),\n",
       " ('extended data fig', 7.277027027027027),\n",
       " ('local real-space refinement', 7.268562401263822),\n",
       " ('reciprocal space refinement', 7.268562401263822),\n",
       " ('sf3a crystal structure', 7.246527777777778),\n",
       " ('dna-pkcs crystal structure', 7.246527777777778),\n",
       " ('izumo1 crystal structure', 7.246527777777778),\n",
       " ('lsq model fit”', 7.224713624177431),\n",
       " ('coot depended critically', 7.191532258064516),\n",
       " ('neuronal core complex', 7.1875),\n",
       " ('complete intasome model', 7.185752585216392),\n",
       " ('missing side chains', 7.166666666666666),\n",
       " ('target dna strands', 7.166666666666666),\n",
       " ('iterative manual building', 7.156328891445371),\n",
       " ('side chain information', 7.083333333333334),\n",
       " ('iterative manual adjustment', 7.078460038986355),\n",
       " ('improve protein geometry', 7.067632850241546),\n",
       " ('initial search model', 7.064121857453589),\n",
       " ('‘secondary structure restrains', 7.052083333333333),\n",
       " ('group b-factor refinements', 7.045454545454545),\n",
       " ('secondary structural features', 7.035714285714286),\n",
       " ('manually build protein', 7.0213365539452495),\n",
       " ('manual corrections made', 7.0210526315789465),\n",
       " ('substantial model rebuilding', 6.974713624177431),\n",
       " ('including tls parameters', 6.966666666666666),\n",
       " ('chimera software packages', 6.958333333333334),\n",
       " ('entire atomic model', 6.954660148241603),\n",
       " ('automatic model building', 6.942192866246836),\n",
       " ('scsam50 potra domain', 6.933333333333334),\n",
       " ('initial structure refinement performed', 6.897766458634585),\n",
       " ('fab homology model', 6.876228775692583),\n",
       " ('alternating manual building', 6.873921484037963),\n",
       " ('subsequent model building', 6.870764294818265),\n",
       " ('real space refinement', 6.868562401263823),\n",
       " ('tfiif n-terminal residues', 6.868421052631579),\n",
       " ('initial atomic model', 6.850886563335942),\n",
       " ('phenix software package', 6.820054945054945),\n",
       " ('independent search models', 6.8173076923076925),\n",
       " ('manual model building', 6.791816926397212),\n",
       " ('atomic model building', 6.7825290007006185),\n",
       " ('iterative model building', 6.778171702225674),\n",
       " ('individual b-factor refinement', 6.772350280051701),\n",
       " ('additional residues added', 6.772267206477732),\n",
       " ('initial model building', 6.766990709912605),\n",
       " ('molecular replacement solution', 6.741758241758241),\n",
       " ('slightly adjusted manually', 6.731481481481482),\n",
       " ('initial structural model', 6.724836143167875),\n",
       " ('ccp4 program suite', 6.7247262541380195),\n",
       " ('manual model fitting', 6.713948073938196),\n",
       " ('lower resolution models', 6.6923076923076925),\n",
       " ('fo\\u2009−\\u2009fc', 6.690909090909091),\n",
       " ('computer program coot', 6.679627496159754),\n",
       " ('software suite cns', 6.658549783549783),\n",
       " ('complete atomic model', 6.597517291098745),\n",
       " ('single fab molecule', 6.595238095238096),\n",
       " ('automatic model rebuilding', 6.546142195606002),\n",
       " ('anomalous data 0', 6.527027027027027),\n",
       " ('initial model manually', 6.503936672268404),\n",
       " ('graphical program coot', 6.429627496159754),\n",
       " ('graphics program coot', 6.429627496159754),\n",
       " ('high-resolution structure obtained', 6.417467948717949),\n",
       " ('iterative process', 6.407407407407407),\n",
       " ('models solvent molecules', 6.406593406593407),\n",
       " ('manual model rebuilding', 6.395766255756378),\n",
       " ('simulated annealing refinement', 6.3711265038279254),\n",
       " ('experimental map', 6.363636363636363),\n",
       " ('iteratively building regions', 6.331492040864814),\n",
       " ('final model rebuilding', 6.3225397111339525),\n",
       " ('initial structure model', 6.241205190786922),\n",
       " ('phenix yielded r-factors', 6.230769230769231),\n",
       " ('adjusted modeller model', 6.209562109025916),\n",
       " ('phenix program package', 6.183150183150183),\n",
       " ('density maps', 6.1179487179487175),\n",
       " ('stereochemical refinement tools', 6.101895734597156),\n",
       " ('electron density', 6.092307692307692),\n",
       " ('ccp4 suite improved', 6.0491310160427805),\n",
       " ('phenix program suite', 6.014319014319014),\n",
       " ('careful inspection', 6.0),\n",
       " ('rigid bodies', 6.0),\n",
       " ('clear bumps', 6.0),\n",
       " ('rigid body', 6.0),\n",
       " ('tight restraint', 6.0),\n",
       " ('eloc complex structure', 5.989583333333333),\n",
       " ('restrained tls refinement', 5.986511119212541),\n",
       " ('automatic structure refinement', 5.975407639359061),\n",
       " ('density map', 5.966433566433566),\n",
       " ('final tls refinement', 5.949721821553678),\n",
       " ('program suite refmac', 5.931697931697932),\n",
       " ('native data', 5.884169884169884),\n",
       " ('ccp4 program buccaneer', 5.845938375350141),\n",
       " ('excellent geometry', 5.844444444444445),\n",
       " ('electron-density maps', 5.833333333333334),\n",
       " ('iterative structure refinement', 5.811386475337897),\n",
       " ('including tls refinement', 5.768562401263822),\n",
       " ('resulting map', 5.744318181818182),\n",
       " ('automatically built model', 5.6914415864806145),\n",
       " ('subsequently completed manually', 5.689814814814815),\n",
       " ('2fo\\u2009−\\u2009fc', 5.662337662337663),\n",
       " ('2fo\\u2009− fc', 5.662337662337663),\n",
       " ('tls refinement restraints', 5.601895734597156),\n",
       " ('fo\\xa0×\\xa0exp', 5.6),\n",
       " ('diffraction data', 5.527027027027027),\n",
       " ('sigmaa-weighted 3fobs', 5.5),\n",
       " ('model building performed', 5.468325270428021),\n",
       " ('protein geometry', 5.400966183574879),\n",
       " ('molecular graphics', 5.384615384615385),\n",
       " ('ncs-averaged maps', 5.333333333333334),\n",
       " ('ccp4 program refmac5', 5.285654687406878),\n",
       " ('real space', 5.266666666666667),\n",
       " ('alanine helices', 5.25),\n",
       " ('polypeptide chain', 5.25),\n",
       " ('additional restraint', 5.25),\n",
       " ('crystallographic symmetry', 5.212121212121212),\n",
       " ('non-crystallographic symmetry', 5.212121212121212),\n",
       " ('individual b-factor', 5.170454545454545),\n",
       " ('apo structure', 5.135416666666667),\n",
       " ('idealized &agr', 5.066666666666666),\n",
       " ('group b-factor', 5.045454545454545),\n",
       " ('data sets', 5.027027027027027),\n",
       " ('data processing', 5.027027027027027),\n",
       " ('experimental phases', 5.0),\n",
       " ('nucleic acids', 5.0),\n",
       " ('tarp tms', 5.0),\n",
       " ('16 selenium atoms', 5.0),\n",
       " ('protein model', 4.999417181489684),\n",
       " ('computer programs', 4.964285714285714),\n",
       " ('25 protein sequences', 4.956521739130435),\n",
       " ('protein interfaces', 4.956521739130435),\n",
       " ('subsequently extended', 4.833333333333334),\n",
       " ('single ensemble', 4.833333333333334),\n",
       " ('fcalc maps', 4.833333333333334),\n",
       " ('repeated cycles', 4.826086956521739),\n",
       " ('ncs restrains', 4.8125),\n",
       " ('electron densities', 4.8076923076923075),\n",
       " ('high-resolution structure', 4.802083333333333),\n",
       " ('pdb 2bvg', 4.8),\n",
       " ('tls parameters', 4.8),\n",
       " ('pdb 1ngq', 4.8),\n",
       " ('b3 density', 4.7846153846153845),\n",
       " ('density modification', 4.7846153846153845),\n",
       " ('b7 density', 4.7846153846153845),\n",
       " ('cryo-em density', 4.7846153846153845),\n",
       " ('density consistent', 4.7846153846153845),\n",
       " ('-helical density', 4.7846153846153845),\n",
       " ('simulated annealing', 4.769230769230769),\n",
       " ('protein structure', 4.758605072463768),\n",
       " ('heavy chain', 4.75),\n",
       " ('secondary structure/ncs', 4.75),\n",
       " ('manual building', 4.748921484037964),\n",
       " ('improved map', 4.744318181818182),\n",
       " ('repeated rounds', 4.742857142857143),\n",
       " ('program modeller', 4.738095238095238),\n",
       " ('resulting solution', 4.729166666666666),\n",
       " ('solvent molecules', 4.714285714285714),\n",
       " ('manual adjustment', 4.671052631578947),\n",
       " ('manual fitting', 4.671052631578947),\n",
       " ('search model', 4.667895442359249),\n",
       " ('readily interpretable', 4.666666666666667),\n",
       " ('asymmetric unit', 4.666666666666667),\n",
       " ('energy minimization', 4.666666666666667),\n",
       " ('pol ii', 4.666666666666666),\n",
       " ('symmetry expanded', 4.666666666666666),\n",
       " ('ccp4 package', 4.65546218487395),\n",
       " ('anisotropy correction', 4.642857142857142),\n",
       " ('supplementary table 1', 4.615384615384615),\n",
       " ('supplementary table 2', 4.615384615384615),\n",
       " ('supplementary table 3', 4.615384615384615),\n",
       " ('resulting model', 4.605395442359249),\n",
       " ('structure interpretation', 4.602083333333333),\n",
       " ('crystallographic model', 4.588349987813794),\n",
       " ('molecular replacement', 4.575091575091575),\n",
       " ('solvent flattening', 4.571428571428571),\n",
       " ('extended manually', 4.564814814814815),\n",
       " ('manual correction', 4.56390977443609),\n",
       " ('maps generated', 4.555555555555555),\n",
       " ('secondary structure', 4.552083333333333),\n",
       " ('hkl2000 suite', 4.545454545454545),\n",
       " ('b-factor restraints', 4.545454545454545),\n",
       " ('entire model', 4.542895442359249),\n",
       " ('homology model', 4.542895442359249),\n",
       " ('software buccaneer', 4.541666666666667),\n",
       " ('missing residues', 4.535087719298246),\n",
       " ('combined experimental', 4.515151515151515),\n",
       " ('automatic rebuilding', 4.503246753246754),\n",
       " ('independent effort', 4.5),\n",
       " ('permitted', 4.5),\n",
       " ('side chains', 4.5),\n",
       " ('cyanobacterial psi', 4.5),\n",
       " ('extracellular side', 4.5),\n",
       " ('1 tls group', 4.5),\n",
       " ('dual-cofactor enzymes', 4.5),\n",
       " ('ccp4 suite', 4.4866310160427805),\n",
       " ('structure missing', 4.46875),\n",
       " ('x-ray structure', 4.46875),\n",
       " ('protein backbone', 4.456521739130435),\n",
       " ('atomic model', 4.454660148241603),\n",
       " ('defined geometry', 4.444444444444445),\n",
       " ('good geometry', 4.444444444444445),\n",
       " ('ideal geometry', 4.444444444444445),\n",
       " ('crude model', 4.44289544235925),\n",
       " ('initial model', 4.439121857453589),\n",
       " ('manual examination', 4.421052631578947),\n",
       " ('manual alteration', 4.421052631578947),\n",
       " ('manual inspection', 4.421052631578947),\n",
       " ('manual docking', 4.421052631578947),\n",
       " ('iterative steps', 4.407407407407407),\n",
       " ('iterative manner', 4.407407407407407),\n",
       " ('loop region', 4.4),\n",
       " ('final model', 4.390721529315771),\n",
       " ('molecular substitution', 4.384615384615385),\n",
       " ('resulting phases', 4.380681818181818),\n",
       " ('model building', 4.370764294818265),\n",
       " ('resulting structure', 4.364583333333333),\n",
       " ('native  sbsb32', 4.357142857142858),\n",
       " ('protein subunits', 4.356521739130434),\n",
       " ('dna complex', 4.354166666666666),\n",
       " ('manual rebuilding', 4.3528708133971294),\n",
       " ('final position', 4.3478260869565215),\n",
       " ('iterative rebuilding', 4.339225589225589),\n",
       " ('supplementary information', 4.333333333333334),\n",
       " ('structural model', 4.328609728073535),\n",
       " ('search models', 4.3173076923076925),\n",
       " ('loop regions', 4.316666666666666),\n",
       " ('manually fitting', 4.314814814814815),\n",
       " ('torsion ncs', 4.3125),\n",
       " ('ncs restraints', 4.3125),\n",
       " ('refine module', 4.3),\n",
       " ('individual site', 4.291666666666667),\n",
       " ('structural units', 4.285714285714286),\n",
       " ('refinement adopting', 4.268562401263823),\n",
       " ('real-space refinement', 4.268562401263823),\n",
       " ('programs crystallographic', 4.259740259740259),\n",
       " ('light chain', 4.25),\n",
       " ('visual inspection', 4.25),\n",
       " ('water fitting', 4.25),\n",
       " ('fit manually', 4.2466329966329965),\n",
       " ('manually fit', 4.2466329966329965),\n",
       " ('crystal structure', 4.246527777777778),\n",
       " ('building regions', 4.244535519125683),\n",
       " ('subsequent rounds', 4.242857142857143),\n",
       " ('program sloop', 4.238095238095238),\n",
       " ('iterative cycles', 4.233494363929147),\n",
       " ('added manually', 4.218660968660968),\n",
       " ('manually added', 4.218660968660968),\n",
       " ('initial phases', 4.214408233276158),\n",
       " ('initial structure', 4.198309748427673),\n",
       " ('homology models', 4.1923076923076925),\n",
       " ('trimeric complex', 4.1875),\n",
       " ('fimh complex', 4.1875),\n",
       " ('cdnmt3l-h3 complex', 4.1875),\n",
       " ('syt-ii complex', 4.1875),\n",
       " ('complete model', 4.185752585216392),\n",
       " ('model correction', 4.185752585216392),\n",
       " ('convincing fit', 4.181818181818182),\n",
       " ('final cycles', 4.173913043478261),\n",
       " ('automatic refinement', 4.173324306025728),\n",
       " ('stereochemical validation', 4.166666666666667),\n",
       " ('dna sequence', 4.166666666666666),\n",
       " ('protospacer dna', 4.166666666666666),\n",
       " ('dna binding', 4.166666666666666),\n",
       " ('standard positional', 4.166666666666666),\n",
       " ('including addition', 4.166666666666666),\n",
       " ('x-ray crystallography', 4.166666666666666),\n",
       " ('upstream dna', 4.166666666666666),\n",
       " ('iterative rounds', 4.15026455026455),\n",
       " ('additional &agr', 4.15),\n",
       " ('crystallographic refinement', 4.147350280051702),\n",
       " ('b-factor refinement', 4.147350280051702),\n",
       " ('water molecules', 4.142857142857142),\n",
       " ('initial rounds', 4.139083557951483),\n",
       " ('individual domains', 4.125),\n",
       " ('phenix software', 4.105769230769231),\n",
       " ('atomic models', 4.104072398190045),\n",
       " ('restraint refinement', 4.101895734597156),\n",
       " ('subsequent refinement', 4.101895734597156),\n",
       " ('translation-libration-screw refinement', 4.101895734597156),\n",
       " ('maximum-likelihood refinement', 4.101895734597156),\n",
       " ('initial models', 4.088534107402032),\n",
       " ('manual adjustments', 4.087719298245614),\n",
       " ('iteratively relaxed', 4.086956521739131),\n",
       " ('software coot', 4.066532258064516),\n",
       " ('manually exchanged', 4.064814814814815),\n",
       " ('manually assigned', 4.064814814814815),\n",
       " ('manually positioning', 4.064814814814815),\n",
       " ('checked manually', 4.064814814814815),\n",
       " ('manually checked', 4.064814814814815),\n",
       " ('manually readjusted', 4.064814814814815),\n",
       " ('manually updated', 4.064814814814815),\n",
       " ('assigned manually', 4.064814814814815),\n",
       " ('build manually', 4.064814814814815),\n",
       " ('manually build', 4.064814814814815),\n",
       " ('manually extending', 4.064814814814815),\n",
       " ('b-factor sharpened', 4.045454545454545),\n",
       " ('crystallographic analysis', 4.045454545454545),\n",
       " ('model improvements', 4.042895442359249),\n",
       " ('nbkb6 model', 4.042895442359249),\n",
       " ('preliminary model', 4.042895442359249),\n",
       " ('opsin model', 4.042895442359249),\n",
       " ('polyalanine model', 4.042895442359249),\n",
       " ('auto-built model', 4.042895442359249),\n",
       " ('current model', 4.042895442359249),\n",
       " ('nmr model', 4.042895442359249),\n",
       " ('model improvement', 4.042895442359249),\n",
       " ('mbh model', 4.042895442359249),\n",
       " ('e1 model', 4.042895442359249),\n",
       " ('final models', 4.040133779264214),\n",
       " ('iterative refinement', 4.009303142004564),\n",
       " ('iterative modelling', 4.007407407407408),\n",
       " ('slow cooling', 4.0),\n",
       " ('rcrane plugin', 4.0),\n",
       " ('quiescent state', 4.0),\n",
       " ('high agreement', 4.0),\n",
       " ('ramachandran outliers', 4.0),\n",
       " ('leut portion', 4.0),\n",
       " ('place vps29', 4.0),\n",
       " ('subunit nqo8', 4.0),\n",
       " ('assemblies service', 4.0),\n",
       " ('rat glut5', 4.0),\n",
       " ('stereochemistry information', 4.0),\n",
       " ('ramachandran favoured/outliers', 4.0),\n",
       " ('subunit arrangement', 4.0),\n",
       " ('backrub tool', 4.0),\n",
       " ('water picking', 4.0),\n",
       " ('randomly selected', 4.0),\n",
       " ('large subset', 4.0),\n",
       " ('least-squares algorithm', 4.0),\n",
       " ('respective templates', 4.0),\n",
       " ('ir tails', 4.0),\n",
       " ('coli bama', 4.0),\n",
       " ('top500 library', 4.0),\n",
       " ('c-terminal barrel', 4.0),\n",
       " ('sm ring', 4.0),\n",
       " ('chemical environment', 4.0),\n",
       " ('delano scientific', 4.0),\n",
       " ('phased', 4.0),\n",
       " ('identical orientations', 4.0),\n",
       " ('transmembrane region', 4.0),\n",
       " ('electrostatic potentials', 4.0),\n",
       " ('nc&dot', 4.0),\n",
       " ('taking advantage', 4.0),\n",
       " ('trial rotamer', 4.0),\n",
       " ('duck 2card', 4.0),\n",
       " ('cutoff applied', 4.0),\n",
       " ('covalent dnmt3a', 4.0),\n",
       " ('potential absence', 4.0),\n",
       " ('ramachandran plot', 4.0),\n",
       " ('subsequently constructed', 4.0),\n",
       " ('screw-motion parameterization', 4.0),\n",
       " ('nmr system', 4.0),\n",
       " ('respective resolutions', 4.0),\n",
       " ('transcribing state', 4.0),\n",
       " ('mutant form', 4.0),\n",
       " ('90-å resolution', 4.0),\n",
       " ('selenomethionine peaks', 4.0),\n",
       " ('ramachandran diagram', 4.0),\n",
       " ('acceptable levels', 4.0),\n",
       " ('tls refinements', 4.0),\n",
       " ('llg metrics', 4.0),\n",
       " ('full-length glua2cryst', 4.0),\n",
       " ('viral clone', 4.0),\n",
       " ('c-terminal bundles', 4.0),\n",
       " ('apelota/dom34 aabce1/rli1', 4.0),\n",
       " ('originally designed', 4.0),\n",
       " ('tls options', 4.0),\n",
       " ('multiple rounds', 3.992857142857143),\n",
       " ('restrained refinement', 3.9865111192125413),\n",
       " ('molecular modelling', 3.9846153846153847),\n",
       " ('structural models', 3.978021978021978),\n",
       " ('model rebuilding', 3.974713624177431),\n",
       " ('alternating cycles', 3.9510869565217392),\n",
       " ('phenix package', 3.9450549450549453),\n",
       " ('rfree set', 3.9423076923076925),\n",
       " ('head/crown domain', 3.9333333333333336),\n",
       " ('u2 3′ domain', 3.9333333333333336),\n",
       " ('domain type', 3.9333333333333336),\n",
       " ('transmembrane domain', 3.9333333333333336),\n",
       " ('n-terminal domain', 3.9333333333333336),\n",
       " ('morc1bah domain', 3.9333333333333336),\n",
       " ('extracellular domain', 3.9333333333333336),\n",
       " ('minor rebuilding', 3.9318181818181817),\n",
       " ('modelling building', 3.9278688524590164),\n",
       " ('manual model-building', 3.9210526315789473),\n",
       " ('manual placement', 3.9210526315789473),\n",
       " ('manual re-building', 3.9210526315789473),\n",
       " ('disallowed regions', 3.916666666666667),\n",
       " ('best-defined regions', 3.916666666666667),\n",
       " ('program buccaneer', 3.904761904761905),\n",
       " ('rna complex', 3.9017857142857144),\n",
       " ('structural refinement', 3.887610020311442),\n",
       " ('good quality', 3.875),\n",
       " ('terminal residues 1', 3.8684210526315788),\n",
       " ('snupn residues 1', 3.8684210526315788),\n",
       " ('selenomethionine residues', 3.8684210526315788),\n",
       " ('crm1 residues 53', 3.8684210526315788),\n",
       " ('contained residues 330', 3.8684210526315788),\n",
       " ('alternating rounds', 3.867857142857143),\n",
       " ('missing loops', 3.8666666666666663),\n",
       " ('model phases', 3.8610772605410677),\n",
       " ('coot/buster-tnt led', 3.857142857142857),\n",
       " ('hydroxylamine complex', 3.854166666666667),\n",
       " ('final rwork', 3.847826086956522),\n",
       " ('final rwork\\u2009=\\u200919', 3.847826086956522),\n",
       " ('structure model', 3.8449787756925824),\n",
       " ('partial model', 3.842895442359249),\n",
       " ('modelled chains', 3.833333333333333),\n",
       " ('engineered proteins', 3.833333333333333),\n",
       " ('stepwise cycles', 3.8260869565217392),\n",
       " ('recursive cycles', 3.8260869565217392),\n",
       " ('alternate cycles', 3.8260869565217392),\n",
       " ('numerous cycles', 3.8260869565217392),\n",
       " ('successive cycles', 3.8260869565217392),\n",
       " ('real-space refined', 3.8192090395480225),\n",
       " ('subcomplex structure', 3.802083333333333),\n",
       " ('previous structure', 3.802083333333333),\n",
       " ('cryo-em structure', 3.802083333333333),\n",
       " ('kinase structure', 3.802083333333333),\n",
       " ('143h282 structure', 3.802083333333333),\n",
       " ('remaining structure', 3.802083333333333),\n",
       " ('traced automatically', 3.7857142857142856),\n",
       " ('molprobity server', 3.7777777777777777),\n",
       " ('positional refinement', 3.7685624012638232),\n",
       " ('rna model', 3.7571811566449638),\n",
       " ('rwork/rfree values', 3.75),\n",
       " ('ucsf chimera', 3.75),\n",
       " ('allowed extension', 3.75),\n",
       " ('supplementary fig', 3.75),\n",
       " ('successive rounds', 3.742857142857143),\n",
       " ('cyclic rounds', 3.742857142857143),\n",
       " ('manually constructed', 3.731481481481482),\n",
       " ('manually adjusted', 3.731481481481482),\n",
       " ('adjusted manually', 3.731481481481482),\n",
       " ('rna nucleotides', 3.7142857142857144),\n",
       " ('model validation', 3.7095621090259163),\n",
       " ('crystal structures', 3.7076023391812867),\n",
       " ('refmac5 produced', 3.706382978723404),\n",
       " ('fitted automatically', 3.7023809523809526),\n",
       " ('models giving', 3.6923076923076925),\n",
       " ('predicted models', 3.6923076923076925),\n",
       " ('ten models', 3.6923076923076925),\n",
       " ('searching models', 3.6923076923076925),\n",
       " ('atomic structures', 3.674922600619195),\n",
       " ('manual tracing', 3.6710526315789473),\n",
       " ('factors converged', 3.666666666666667),\n",
       " ('making adjustments', 3.666666666666667),\n",
       " ('good stereochemistry', 3.666666666666667),\n",
       " ('published fragments', 3.666666666666667),\n",
       " ('active site', 3.666666666666667),\n",
       " ('r-free calculation', 3.666666666666667),\n",
       " ('dna complexes', 3.6666666666666665),\n",
       " ('averaging', 3.6666666666666665),\n",
       " ('top solution', 3.6666666666666665),\n",
       " ('modified manually', 3.664814814814815),\n",
       " ('manually modified', 3.664814814814815),\n",
       " ('obtained model', 3.6582800577438648),\n",
       " ('ccp4 programs', 3.6554621848739495),\n",
       " ('iteratively improved', 3.6494565217391304),\n",
       " ('built automatically', 3.648546144121365),\n",
       " ('model refinement', 3.6447911769564056),\n",
       " ('manually improved', 3.627314814814815),\n",
       " ('partial phases', 3.618181818181818),\n",
       " ('restraints obtained', 3.6153846153846154),\n",
       " ('improved model', 3.6053954423592494),\n",
       " ('temperature-factor refinement', 3.6018957345971563),\n",
       " ('full refinement', 3.6018957345971563),\n",
       " ('refinement algorithms', 3.6018957345971563),\n",
       " ('tls refinement', 3.6018957345971563),\n",
       " ('standard refinement', 3.6018957345971563),\n",
       " ('refinement options', 3.6018957345971563),\n",
       " ('fragment-based refinement', 3.6018957345971563),\n",
       " ('refinement cycle', 3.6018957345971563),\n",
       " ('representation made', 3.6),\n",
       " ('omit', 3.6),\n",
       " ('sufficient quality', 3.5892857142857144),\n",
       " ('iteratively rebuilt', 3.5869565217391304),\n",
       " ('restrained translation', 3.5846153846153843),\n",
       " ('manually adjusting', 3.564814814814815),\n",
       " ('rebuilt manually', 3.564814814814815),\n",
       " ('remodelled manually', 3.564814814814815),\n",
       " ('manually rebuilt', 3.564814814814815),\n",
       " ('traced manually', 3.564814814814815),\n",
       " ('manually edited', 3.564814814814815),\n",
       " ('phenix packages', 3.5641025641025643),\n",
       " ('slowly improved', 3.5625),\n",
       " ('subsequently subjected', 3.555555555555556),\n",
       " ('model monitored', 3.5428954423592494),\n",
       " ('coot based', 3.52486559139785),\n",
       " ('phasing power', 3.5),\n",
       " ('accurately reflect', 3.5),\n",
       " ('conserved β-sheet', 3.5),\n",
       " ('ide-substrate complexes', 3.5),\n",
       " ('trimer formation', 3.5),\n",
       " ('s1 domains', 3.5),\n",
       " ('starting temperature', 3.5),\n",
       " ('adding waters', 3.5),\n",
       " ('judged successful', 3.5),\n",
       " ('multiple iterations', 3.5),\n",
       " ('wild-type uvr8', 3.5),\n",
       " ('visible rods', 3.5),\n",
       " ('fitted manually', 3.481481481481482),\n",
       " ('manually fitted', 3.481481481481482),\n",
       " ('program cns', 3.4761904761904763),\n",
       " ('program phenix', 3.468864468864469),\n",
       " ('structure visualization', 3.46875),\n",
       " ('structure determination', 3.46875),\n",
       " ('phase determination', 3.466666666666667),\n",
       " ('phase calculation', 3.466666666666667),\n",
       " ('built iteratively', 3.44978838014621),\n",
       " ('iteratively built', 3.44978838014621),\n",
       " ('program resolve', 3.4380952380952383),\n",
       " ('add domain', 3.4333333333333336),\n",
       " ('program coot', 3.429627496159754),\n",
       " ('coot program', 3.429627496159754),\n",
       " ('ncul5 molecule', 3.428571428571429),\n",
       " ('refinement cycles', 3.4279826911188955),\n",
       " ('manually built', 3.4276466732218944),\n",
       " ('built manually', 3.4276466732218944),\n",
       " ('accomplished iteratively', 3.420289855072464),\n",
       " ('structure obtained', 3.417467948717949),\n",
       " ('fitted rigidly', 3.416666666666667),\n",
       " ('built model', 3.405727300766329),\n",
       " ('program arp/warp', 3.404761904761905),\n",
       " ('structure refinement', 3.4039790679304898),\n",
       " ('haem &agr', 3.4),\n",
       " ('rrp4 subunits', 3.4),\n",
       " ('refmac program', 3.386243386243386),\n",
       " ('program refmac', 3.386243386243386),\n",
       " ('model superpositions', 3.3762287756925824),\n",
       " ('residues cys\\u2009690', 3.3684210526315788),\n",
       " ('changed residues', 3.3684210526315788),\n",
       " ('models converged', 3.358974358974359),\n",
       " ('completed manually', 3.356481481481482),\n",
       " ('manually completed', 3.356481481481482),\n",
       " ('phaser program', 3.3557422969187676),\n",
       " ('program phaser', 3.3557422969187676),\n",
       " ('program refmac5', 3.3444782168186427),\n",
       " ('maps', 3.3333333333333335),\n",
       " ('semet positions', 3.333333333333333),\n",
       " ('polyalanine trace', 3.333333333333333),\n",
       " ('/~ssa1/gorgon', 3.333333333333333),\n",
       " ('receptor chains', 3.333333333333333),\n",
       " ('phenix iteratively', 3.317725752508361),\n",
       " ('optimized manually', 3.314814814814815),\n",
       " ('generated manually', 3.287037037037037),\n",
       " ('coot manually', 3.256347072879331),\n",
       " ('rfree values', 3.25),\n",
       " ('favoured regions', 3.25),\n",
       " ('refined iteratively', 3.2394988946204863),\n",
       " ('iteratively refined', 3.2394988946204863),\n",
       " ('undertaken manually', 3.231481481481482),\n",
       " ('manually corrected', 3.231481481481482),\n",
       " ('manually refined', 3.217357187696171),\n",
       " ('rigid-body refinement', 3.201895734597157),\n",
       " ('coot constrained', 3.191532258064516),\n",
       " ('coot toolkit', 3.191532258064516),\n",
       " ('coot independently', 3.191532258064516),\n",
       " ('experimental', 3.1818181818181817),\n",
       " ('map', 3.1818181818181817),\n",
       " ('starting fragments', 3.166666666666667),\n",
       " ('hydroxylamine complexes', 3.166666666666667),\n",
       " ('free r-factor', 3.166666666666667),\n",
       " ('performed manually', 3.162375790424571),\n",
       " ('refined individually', 3.152542372881356),\n",
       " ('screw refinement', 3.1018957345971563),\n",
       " ('model carried', 3.0651176645814715),\n",
       " ('coot led', 3.0486751152073737),\n",
       " ('phenix phase', 3.0307692307692307),\n",
       " ('data', 3.027027027027027),\n",
       " ('rigid-body fitted', 3.0166666666666666),\n",
       " ('refmac led', 3.0052910052910056),\n",
       " ('factor refinement', 3.0018957345971566),\n",
       " ('processed', 3.0),\n",
       " ('phase combination', 3.0),\n",
       " ('involved', 3.0),\n",
       " ('graphics', 3.0),\n",
       " ('protein', 2.9565217391304346),\n",
       " ('structure refined', 2.9546257062146895),\n",
       " ('cns programs', 2.9523809523809526),\n",
       " ('programs cns', 2.9523809523809526),\n",
       " ('phenix programs', 2.9450549450549453),\n",
       " ('programs phenix', 2.9450549450549453),\n",
       " ('programs coot', 2.9058179723502304),\n",
       " ('refmac programs', 2.8624338624338623),\n",
       " ('cns refinement', 2.8399909726923944),\n",
       " ('solomon resulted', 2.833333333333333),\n",
       " ('rigid-body refining', 2.8),\n",
       " ('pdb', 2.8),\n",
       " ('density', 2.7846153846153845),\n",
       " ('autobuster alternated', 2.75),\n",
       " ('density-modified', 2.75),\n",
       " ('phenix autobuild', 2.730769230769231),\n",
       " ('buster refinement', 2.7268957345971563),\n",
       " ('refinement performed', 2.6994567102069125),\n",
       " ('set', 2.6923076923076925),\n",
       " ('scaling', 2.6666666666666665),\n",
       " ('refined initially', 2.652542372881356),\n",
       " ('individual', 2.625),\n",
       " ('produced', 2.6),\n",
       " ('phenix refine', 2.5307692307692307),\n",
       " ('extended', 2.5),\n",
       " ('subsequent', 2.5),\n",
       " ('placing', 2.5),\n",
       " ('multidomain', 2.5),\n",
       " ('refmac version 5', 2.4814814814814814),\n",
       " ('geometry', 2.4444444444444446),\n",
       " ('coot guided', 2.441532258064516),\n",
       " ('manual', 2.4210526315789473),\n",
       " ('loop 6', 2.4),\n",
       " ('restrained', 2.3846153846153846),\n",
       " ('based', 2.3333333333333335),\n",
       " ('fab', 2.3333333333333335),\n",
       " ('h4', 2.3333333333333335),\n",
       " ('building', 2.3278688524590163),\n",
       " ('refined coordinates', 2.3192090395480225),\n",
       " ('ncs', 2.3125),\n",
       " ('chain', 2.25),\n",
       " ('helices', 2.25),\n",
       " ('fitting', 2.25),\n",
       " ('replacement', 2.1904761904761907),\n",
       " ('complex', 2.1875),\n",
       " ('fit', 2.1818181818181817),\n",
       " ('including', 2.1666666666666665),\n",
       " ('solution', 2.1666666666666665),\n",
       " ('dna', 2.1666666666666665),\n",
       " ('positional', 2.1666666666666665),\n",
       " ('added', 2.1538461538461537),\n",
       " ('complete', 2.142857142857143),\n",
       " ('calculated', 2.142857142857143),\n",
       " ('molecules', 2.142857142857143),\n",
       " ('manually', 2.064814814814815),\n",
       " ('model', 2.0428954423592494),\n",
       " ('n-terminal', 2.0),\n",
       " ('4\\u2009å resolution', 2.0),\n",
       " ('potra 1', 2.0),\n",
       " ('52\\u2009å resolution', 2.0),\n",
       " ('core', 2.0),\n",
       " ('build', 2.0),\n",
       " ('sequence', 2.0),\n",
       " ('eloc', 2.0),\n",
       " ('hydrophobicity', 2.0),\n",
       " ('present', 2.0),\n",
       " ('remaining', 2.0),\n",
       " ('modification', 2.0),\n",
       " ('r-factors', 2.0),\n",
       " ('nbkb6', 2.0),\n",
       " ('0\\u2009å resolution', 2.0),\n",
       " ('3 å resolution', 2.0),\n",
       " ('region', 2.0),\n",
       " ('15\\u2009å resolution', 2.0),\n",
       " ('chains', 2.0),\n",
       " ('0 å resolution', 2.0),\n",
       " ('inspection', 2.0),\n",
       " ('refinements', 2.0),\n",
       " ('cryo-em', 2.0),\n",
       " ('tools', 2.0),\n",
       " ('6\\u2009å resolution', 2.0),\n",
       " ('restraints', 2.0),\n",
       " ('fimd', 2.0),\n",
       " ('features', 2.0),\n",
       " ('strands', 2.0),\n",
       " ('18 chains', 2.0),\n",
       " ('tls', 2.0),\n",
       " ('ccp4', 1.9411764705882353),\n",
       " ('domain', 1.9333333333333333),\n",
       " ('rebuilding', 1.9318181818181819),\n",
       " ('regions', 1.9166666666666667),\n",
       " ('quality', 1.875),\n",
       " ('residues 327', 1.868421052631579),\n",
       " ('residues', 1.868421052631579),\n",
       " ('025 residues', 1.868421052631579),\n",
       " ('residues  295', 1.868421052631579),\n",
       " ('32 residues', 1.868421052631579),\n",
       " ('residues 374', 1.868421052631579),\n",
       " ('residues 79', 1.868421052631579),\n",
       " ('residues 4', 1.868421052631579),\n",
       " ('led', 1.8571428571428572),\n",
       " ('modelled', 1.8333333333333333),\n",
       " ('proteins', 1.8333333333333333),\n",
       " ('10 cycles', 1.826086956521739),\n",
       " ('cycles', 1.826086956521739),\n",
       " ('phases', 1.8181818181818181),\n",
       " ('structure', 1.8020833333333333),\n",
       " ('partial', 1.8),\n",
       " ('phase', 1.8),\n",
       " ('rwork/rfree', 1.75),\n",
       " ('fig', 1.75),\n",
       " ('allowed', 1.75),\n",
       " ('chimera', 1.75),\n",
       " ('rounds', 1.7428571428571429),\n",
       " ('programs', 1.7142857142857142),\n",
       " ('sufficient', 1.7142857142857142),\n",
       " ('rna', 1.7142857142857142),\n",
       " ('models', 1.6923076923076923),\n",
       " ('validation', 1.6666666666666667),\n",
       " ('converged', 1.6666666666666667),\n",
       " ('determination', 1.6666666666666667),\n",
       " ('6% r-factor', 1.6666666666666667),\n",
       " ('fragments', 1.6666666666666667),\n",
       " ('improve', 1.6666666666666667),\n",
       " ('buccaneer', 1.6666666666666667),\n",
       " ('stereochemistry', 1.6666666666666667),\n",
       " ('adjustments', 1.6666666666666667),\n",
       " ('constructed', 1.6666666666666667),\n",
       " ('site', 1.6666666666666667),\n",
       " ('visualization', 1.6666666666666667),\n",
       " ('calculation', 1.6666666666666667),\n",
       " ('adjusted', 1.6666666666666667),\n",
       " ('hydroxylamine', 1.6666666666666667),\n",
       " ('superposition', 1.6666666666666667),\n",
       " ('&dgr', 1.6666666666666667),\n",
       " ('obtained', 1.6153846153846154),\n",
       " ('refinement', 1.6018957345971565),\n",
       " ('made', 1.6),\n",
       " ('rigid-body', 1.6),\n",
       " ('modelling', 1.6),\n",
       " ('modified', 1.6),\n",
       " ('improved', 1.5625),\n",
       " ('re-building', 1.5),\n",
       " ('rwork 23', 1.5),\n",
       " ('fcalc', 1.5),\n",
       " ('crystallography', 1.5),\n",
       " ('conserved', 1.5),\n",
       " ('cys\\u2009700', 1.5),\n",
       " ('autobuster', 1.5),\n",
       " ('formation', 1.5),\n",
       " ('judged', 1.5),\n",
       " ('waters', 1.5),\n",
       " ('adjusting', 1.5),\n",
       " ('visible', 1.5),\n",
       " ('monitored', 1.5),\n",
       " ('changed', 1.5),\n",
       " ('rebuilt', 1.5),\n",
       " ('reflect', 1.5),\n",
       " ('autobuild', 1.5),\n",
       " ('sharpened', 1.5),\n",
       " ('sbsb32', 1.5),\n",
       " ('edited', 1.5),\n",
       " ('screw', 1.5),\n",
       " ('model-building', 1.5),\n",
       " ('initially', 1.5),\n",
       " ('solomon', 1.5),\n",
       " ('remodelled', 1.5),\n",
       " ('starting', 1.5),\n",
       " ('atoms', 1.5),\n",
       " ('uvr8', 1.5),\n",
       " ('traced', 1.5),\n",
       " ('placement', 1.5),\n",
       " ('phasing', 1.5),\n",
       " ('complexes', 1.5),\n",
       " ('domains', 1.5),\n",
       " ('densities', 1.5),\n",
       " ('rwork', 1.5),\n",
       " ('analysis', 1.5),\n",
       " ('add', 1.5),\n",
       " ('top', 1.5),\n",
       " ('backbone', 1.5),\n",
       " ('free', 1.5),\n",
       " ('molecule 1', 1.4285714285714286),\n",
       " ('molecule 2', 1.4285714285714286),\n",
       " ('molecule', 1.4285714285714286),\n",
       " ('fitted', 1.4166666666666667),\n",
       " ('subunits', 1.4),\n",
       " ('factor', 1.4),\n",
       " ('&agr', 1.4),\n",
       " ('built', 1.3628318584070795),\n",
       " ('trace', 1.3333333333333333),\n",
       " ('receptor', 1.3333333333333333),\n",
       " ('positions', 1.3333333333333333),\n",
       " ('favoured', 1.3333333333333333),\n",
       " ('superpositions', 1.3333333333333333),\n",
       " ('combined', 1.3333333333333333),\n",
       " ('resulted', 1.3333333333333333),\n",
       " ('/', 1.3333333333333333),\n",
       " ('version 0', 1.3333333333333333),\n",
       " ('accomplished', 1.3333333333333333),\n",
       " ('version 1', 1.3333333333333333),\n",
       " ('refine', 1.3),\n",
       " ('completed', 1.2916666666666667),\n",
       " ('structures', 1.263157894736842),\n",
       " ('rfree 26', 1.25),\n",
       " ('iterations', 1.25),\n",
       " ('alternated', 1.25),\n",
       " ('rfree\\u2009=\\u200924', 1.25),\n",
       " ('tracing', 1.25),\n",
       " ('rfree', 1.25),\n",
       " ('guided', 1.25),\n",
       " ('optimized', 1.25),\n",
       " ('cns 1', 1.2380952380952381),\n",
       " ('cns', 1.2380952380952381),\n",
       " ('phenix', 1.2307692307692308),\n",
       " ('subjected', 1.2222222222222223),\n",
       " ('generated', 1.2222222222222223),\n",
       " ('translation', 1.2),\n",
       " ('resolve', 1.2),\n",
       " ('combination', 1.2),\n",
       " ('loops', 1.2),\n",
       " ('refining', 1.2),\n",
       " ('coot', 1.1915322580645162),\n",
       " ('‘coot', 1.1915322580645162),\n",
       " ('arp/warp', 1.1666666666666667),\n",
       " ('coordinates', 1.1666666666666667),\n",
       " ('undertaken', 1.1666666666666667),\n",
       " ('corrected', 1.1666666666666667),\n",
       " ('refined', 1.152542372881356),\n",
       " ('refmac 5', 1.1481481481481481),\n",
       " ('refmac', 1.1481481481481481),\n",
       " ('buster', 1.125),\n",
       " ('phaser', 1.1176470588235294),\n",
       " ('molprobity', 1.1111111111111112),\n",
       " ('refmac5', 1.1063829787234043),\n",
       " ('performed', 1.0975609756097562),\n",
       " ('carried', 1.0222222222222221),\n",
       " ('m27', 1.0),\n",
       " ('translation/libration/screw', 1.0),\n",
       " ('ccp4-scala', 1.0),\n",
       " ('/rfree', 1.0),\n",
       " ('retain', 1.0),\n",
       " ('&psgr', 1.0),\n",
       " ('rsref', 1.0),\n",
       " ('2fcalc', 1.0),\n",
       " ('providing', 1.0),\n",
       " ('mg7', 1.0),\n",
       " ('superfamily', 1.0),\n",
       " ('m1', 1.0),\n",
       " ...]"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rake.run(' '.join(matched_dfs.iloc[92203]['clean_text']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>freq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>(The, model, built)</th>\n",
       "      <td>25</td>\n",
       "      <td>0.004197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(manual, model, building)</th>\n",
       "      <td>23</td>\n",
       "      <td>0.003861</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(model, built, using)</th>\n",
       "      <td>19</td>\n",
       "      <td>0.003190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(model, building, refinement)</th>\n",
       "      <td>18</td>\n",
       "      <td>0.003022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(performed, using, COOT)</th>\n",
       "      <td>15</td>\n",
       "      <td>0.002518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(Model, building, performed)</th>\n",
       "      <td>15</td>\n",
       "      <td>0.002518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(model, building, COOT)</th>\n",
       "      <td>15</td>\n",
       "      <td>0.002518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(initial, model, built)</th>\n",
       "      <td>15</td>\n",
       "      <td>0.002518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(building, refinement, performed)</th>\n",
       "      <td>15</td>\n",
       "      <td>0.002518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(Model, building, refinement)</th>\n",
       "      <td>12</td>\n",
       "      <td>0.002014</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    0      freq\n",
       "(The, model, built)                25  0.004197\n",
       "(manual, model, building)          23  0.003861\n",
       "(model, built, using)              19  0.003190\n",
       "(model, building, refinement)      18  0.003022\n",
       "(performed, using, COOT)           15  0.002518\n",
       "(Model, building, performed)       15  0.002518\n",
       "(model, building, COOT)            15  0.002518\n",
       "(initial, model, built)            15  0.002518\n",
       "(building, refinement, performed)  15  0.002518\n",
       "(Model, building, refinement)      12  0.002014"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_trigram(matched_dfs.iloc[92203]['clean_text'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_trigrams = top_trigram(matched_dfs.iloc[92208]['clean_text'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Build TF-IDF based key words system"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.5/dist-packages/ipykernel_launcher.py:2: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0\n",
      "Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`\n",
      "  \n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "3593741f52954ed59f580b40b9b66be4",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=0, max=304448), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "#clean_text_corpus = [' '.join(ast.literal_eval(i)) for i in tqdm(matched_dfs['clean_text'].tolist())]\n",
    "clean_text_corpus = [' '.join(i) for i in tqdm(matched_dfs['clean_text'].tolist())]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['A detailed comparison of residues involved in glycine binding to NR1 subunits<bibr rid=\"b8 b9 b10\"/> and glutamate binding to NR2 subunits<bibr rid=\"b11 b12\"/> revealed that NR3B (or NR3A) could potentially share ligand specificity with either or both of these classes of NMDAR subunit.',\n",
       " 'Also, the epigenetic stability of human ES cells seems to be greater than mouse ES cells, as shown here and previously by their extent of monoallelic imprinted gene expression<bibr rid=\"b23 b24\"/>.',\n",
       " 'Hospital and clinic HTC had similar costs (&dollar;12.56 and &dollar;12.32, respectively)<bibr rid=\"b81 b88 b93 b140 b142 b143 b144 b145 b146 b147\"/> (Supplementary Table S3 and Figure S62).',\n",
       " '<b>e</b>, Allele-based pleiotropy of <i>IL6R</i> 358Asp (rs2228145 (A))<bibr rid=\"b5\"/> on multiple disease phenotypes, including increased risk of RA, ankylosing spondylitis and coronary heart disease (asterisks indicate associations obtained from the literature<bibr rid=\"b29 b30\"/>) and protection from asthma, as well as levels of biomarkers (increased C-reactive protein (CRP) and fibrinogen but decreased soluble interleukin-6 receptor (sIL6R)). For ankylosing spondylitis<bibr rid=\"b51\"/>, atopic dermatitis<bibr rid=\"b52\"/>, primary sclerosing cholangitis<bibr rid=\"b53\"/>, juvenile idiopathic arthritis<bibr rid=\"b54\"/>, and psoriasis<bibr rid=\"b55\"/>, Immunochip studies had been previously been published, but only the lead SNPs from associated Immunochip regions were available.',\n",
       " 'However, structural<bibr rid=\"b39 b40 b41 b42\"/> and biochemical<bibr rid=\"b43 b44\"/> studies of several retaining glycosyltransferases in the presence of substrate analogues failed to identify an active-site nucleophile that could be involved in forming the enzyme&ndash;sugar adduct.',\n",
       " 'Although this effect seems to reflect the maintenance of object representations from the memory array, it is necessary to rule out the possibility that it reflects executive processes<bibr rid=\"b18\"/> involved in performing the task, or even more general processes such as increased effort or arousal<bibr rid=\"b19 b20 b21\"/>.',\n",
       " '\\r\\n<sec level=\"1\">\\r\\n<sectitle>Results</sectitle>\\r\\n<sec level=\"2\">\\r\\n<sectitle>Identification of the host factor</sectitle>\\r\\n<p>To identify the host factor required for cleavage of TcdB, we used TcdB preparations<bibr rid=\"b32\"/> that were Cy3-labelled as described in the literature<bibr rid=\"b29\"/>.',\n",
       " 'enterica</i> effector SseL was recently shown to catalyse the removal of polyubiquitin chains that had been attached to host proteins during infection<bibr rid=\"b67\"/>.',\n",
       " 'This model may be relevant to haem-containing globins in a broad range of NOS-containing somatic cells<bibr rid=\"b7 b8 b9 b10 b11 b12 b13\"/>. 8c, d</sir>), consistent with a previous report<bibr rid=\"b11\"/>.',\n",
       " '\\r\\n<p>An important consideration for siRNA-mediated inhibition of gene expression is whether the observed effects are specific and not due to nonspecific &ldquo;off target&rdquo; effects<bibr rid=\"b15\"/> and potential interferon responses<bibr rid=\"b16\"/>, which have been reported with siRNAs <i>in vitro</i> and other oligonucleotide-based approaches <i>in vivo</i>.',\n",
       " 'cruzi</i> parasites, hydrophilic acylated surface proteins<bibr rid=\"b10\"/> and flagellar calcium-binding protein<bibr rid=\"b11\"/>, respectively.',\n",
       " 'The mass models for this cluster<bibr rid=\"b14 b15\"/> suggest a relatively flat mass distribution profile and a large area of high magnification, making it one of the most powerful cosmic lenses known.',\n",
       " 'However, it should be possible to exceed the catalysis threshold by improving the photon-collection efficiency through the use of nearby optical collection elements or optical cavities<bibr rid=\"b29\"/>.',\n",
       " 'The delivery of nutrients to the surface ocean and the partitioning of trophic resources between shelf and open ocean settings have also been linked to the abundance and diversification of planktonic foraminifera<bibr rid=\"b9\"/>.',\n",
       " 'In the template-dependent DNA polymerases, nucleotide substrates are selected in the catalytically active closed conformation<bibr rid=\"b1\"/>. The open-to-closed transition that is triggered by the incoming NTP is quite similar to that observed in conventional DNA polymerases<bibr rid=\"b1\"/>. This is in striking contrast to the continuous polymerization that is catalysed by conventional template-dependent DNA/RNA polymerases<bibr rid=\"b1 b2 b3\"/>.',\n",
       " 'We imaged flies bearing the <i>GH146-Gal4</i> (also known as <i>P{GAL4}GH146</i>) and <i>UAS-GCaMP</i> (<i>P{UAS-G-CaMP}</i>) transgenes, which have G-CaMP expression in 83 out of 150 projection neurons<bibr rid=\"b9 b28 b29\"/>. \\r\\n</sec>\\r\\n<sec level=\"2\">\\r\\n<sectitle>Fly stocks</sectitle>\\r\\n<p>The GH146-Gal4 and MZ19-Gal4 driver lines used to label PNs have been described previously<bibr rid=\"b10 b11 b12 b32 b33\"/>. Many interneurons that are immuno-positive for GABA (&ggr;-aminobutyric acid) form connections between glomeruli<bibr rid=\"b9 b13 b14\"/>, but several recent studies have failed to observe any interglomerular inhibition<bibr rid=\"b7 b11 b12\"/>. Many individual GABAergic interneurons innervate all glomeruli<bibr rid=\"b9 b12 b13\"/>, suggesting that lateral inhibitory input to each glomerulus might reflect pooled input from all ORNs. Similarly, flies expressing UAS-<i>Shi</i><super><i>ts</i></super> and another Gal4 driver, GH146-Gal4, which is expressed in about two-thirds of antennal lobe projection neurons<bibr rid=\"b5 b17\"/> (but not in those innervating V), also showed robust CO<sub>2</sub> avoidance at the non-permissive temperature (<figr rid=\"f3\">Fig. \\r\\n<online-methods>\\r\\n<sec level=\"2\">\\r\\n<sectitle>Fly stocks</sectitle>\\r\\n<p>All fly transgenic lines (synaptobrevin<super>GAL4</super>, OK107<super>GAL4</super>, GH146<super>GAL4</super>, Fruitless<super>GAL4</super>, UAS&ndash;C3PA&ndash;GFP, UAS&ndash;SPA&ndash;GFP and UAS&ndash;GCaMP3) have been described previously<bibr rid=\"b25 b26 b27 b28\"/>.',\n",
       " 'Recently, nucleosynthesis in the low-mass population III stars was studied extensively<bibr rid=\"b12 b13\"/>.',\n",
       " 'Patterned hydrogenation on graphene already shows its bandgap opening<bibr rid=\"b31\"/>. In order to manipulate both the Fermi level and the bandgap of graphene, patterned doping has to be achieved<bibr rid=\"b31\"/>. This value is similar to that predicted previously for spin&#x02013;orbit-coupled 3<i>d</i> transition metals in kagome lattices<bibr rid=\"b7\"/>, but smaller than that observed in magnetically doped topological insulators (about 50&#x02009;meV)<bibr rid=\"b26 b27\"/> and in hydrogen-decorated graphene (at least 0.5&#x02009;eV)<bibr rid=\"b28\"/>.',\n",
       " 'This effect of social context on acoustic communication in songbirds is in accordance with neurophysiological data: birds recorded in isolation or in visual and auditory contact with a female show different levels of electrophysiological activity<bibr rid=\"b21\"/> and ZENK gene expression<bibr rid=\"b22\"/> within auditory brain nuclei. The act of singing induces gene expression in the male song control nuclei, and these patterns of gene activation also vary with the context of the experience<bibr rid=\"b18\"/>. To examine this possibility, we compared song produced by birds in two behavioural conditions in which the neural activity in the AFP is known to differ: when a male bird sings alone (&lsquo;undirected&rsquo; song), activity in LMAN is greater in magnitude and more variable in pattern across renditions than when it sings to another bird (&lsquo;directed&rsquo; song)<bibr rid=\"b12 b20\"/> (<figr rid=\"f3\">Fig.',\n",
       " 'Regulatory proteins that sense NO mediate the primary response to NO and nitrosative stress<bibr rid=\"b3 b4 b5 b6 b7 b8 b9\"/>.',\n",
       " 'Such work has led to a series of additional PARP inhibitors being tested in clinical trials<bibr rid=\"b39\"/>. In fact, the clinical observations with potent catalytic-site PARP inhibitors have been promising when used in appropriate populations<bibr rid=\"b39\"/>. In the clinical setting, patients with triple-negative breast cancer do not show a frequent response to either olaparib or veliparib<bibr rid=\"b39\"/>, although the number of patients tested is still small.',\n",
       " 'These results are consistent with reports that reduced apoptosis or increased survival of unexposed cells is one of the endpoints of RIBEs<bibr rid=\"b3 b7 b8 b9\"/>.',\n",
       " 'The <i>c</i> value is strongly debated<bibr rid=\"b4 b5 b6 b7 b8 b9 b10\"/>, as researchers<bibr rid=\"b11 b12\"/> claim that <i>c</i> is close to zero and larger values are, at least partially, artefacts caused by lower data completeness after a main shock, especially during the main-shock coda. A similar relationship has been reported for individual earthquakes in California and Japan<bibr rid=\"b5 b9\"/>.',\n",
       " 'This effect is expected to be more pronounced in the AID-transfected N89 clones in which mutation rates in the leader sequence (that is, where the nonsense codon is located) are lower<bibr rid=\"b12 b13\"/>.',\n",
       " '\\r\\n</sec>\\r\\n<sec level=\"2\">\\r\\n<sectitle>Adenoviruses</sectitle>\\r\\n<p>Type 5, E1/E3-deleted adenoviruses expressing murine HSP110 (AdHSP110) or <i>Escherichia coli</i> &bgr;-galactosidase (AdLacZ) under the control of the cytomegalovirus promoter, were constructed according to the <i>in vitro</i> ligation method<bibr rid=\"b33 b34\"/>.',\n",
       " '\\n</fp>\\n<p>Over half of all humans are infected by <i>Helicobacter pylori</i>, a Gram-negative bacterium that can cause peptic ulcers and constitutes a risk factor for stomach cancer<bibr rid=\"b8\"/>. \\r\\n</abs>&e100311-6; &nature08756-s1;</fm>\\r\\n<bdy>\\r\\n<p>About 50% of all humans are infected with <i>Helicobacter pylori</i>, a microaerophilic, Gram-negative &epsi;-proteobacterium that thrives in the acidic environment of the stomach, and is associated with severe inflammation, peptic ulcer disease and gastric cancer<bibr rid=\"b1 b2\"/>.',\n",
       " 'Positions in the alignments with consensus scores of &lt;6 from M-Coffee were removed; filtered alignments were then run through the TCS tool<bibr rid=\"b101\"/> via the T-Coffee web server and any positions with a reliability score of &lt;6 were removed.',\n",
       " 'reported automated sequential solution-phase combinatorial synthesis to perform a 2&times;2 synthesis using the Knoevenagel condensation of 1,3-diketones and aldehydes<bibr rid=\"b44\"/>.',\n",
       " 'However, when the degree of evidence behind each candidate is taken into account, using the gene scoring module in SFARI Gene 2.0<bibr rid=\"b6\"/> (as of 1 April 2014), it remains clear that numerous strong ASD candidate genes are very long (&gt;200 kilobases). They also showed that the list of large genes affected by TOP1 inhibition is enriched with candidate genes for autism spectrum disorders (ASD); however, the list of candidate genes that was used contains many genes with limited evidence for association with ASD<bibr rid=\"b2\"/>. To address this concern, a gene scoring module was developed in SFARI Gene 2.0 to estimate the evidence level for individual genes<bibr rid=\"b2\"/>.',\n",
       " 'In this respect, it is intriguing that PKM2, in addition to sensing serine<bibr rid=\"b25 b26\"/>, is inactivated by oxidative stress<bibr rid=\"b27\"/>. \\r\\n<p>Serine starvation activates <i>de novo</i> serine synthesis, diverting glycolytic intermediates away from energy production<bibr rid=\"b11 b12\"/>. Serine activates PKM2<bibr rid=\"b19\"/> and decreased PKM2 activity following serine starvation causes an accumulation of upstream glycolytic intermediates for diversion to the SSP<bibr rid=\"b20\"/>. To balance lower glycolysis following PKM2 inhibition, cells increase flux of pyruvate to the tricarboxylic acid (TCA) cycle, requiring cells depleted of PKM to display increased O<sub>2</sub> consumption to support elevated oxidative phosphorylation<bibr rid=\"b20\"/>. Exchange rate (uptake or secretion) measurements of key metabolites and stable isotope labelling were performed as described previously<bibr rid=\"b28 b29 b30\"/>. Mass spectrometry data were analysed by LCquan (Thermo Scientific), and quantifications of intracellular and extracellular metabolites were performed by the standard dilution method as described previously<bibr rid=\"b30\"/>. \\r\\n<p>A reduction in the activity of the rate-limiting glycolytic enzymes PFK1 and PKM2 in cancer cells re-directs glycolytic intermediates into the pentose phosphate (PPP) and serine synthesis pathways<bibr rid=\"b16 b17 b18 b19\"/>. While a link between SHMT2 and FBP is less clear, we nonetheless noted that the suppression of SHMT2 significantly increases levels of all three known activators of pyruvate kinase isoform M2 (PKM2)&mdash;serine, FBP and SAICAR<bibr rid=\"b21 b22 b23\"/>&mdash;raising the possibility that SHMT2 antagonizes PKM2 activity by decreasing the levels of its activators.',\n",
       " 'No significant activation was found when the analyses applied to the entorhinal ROI were performed to control ROIs in the posterior right hippocampal location associated with memory performance in this task<bibr rid=\"b11\"/> or the visual cortical location showing adaptation to running direction (below) (see <sir rid=\"s1\">Supplementary Fig. These results outline a circuit for navigation, consistent with suggestions that medial and lateral temporal, posterior and medial parietal and medial prefrontal areas cooperate to support spatial cognition<bibr rid=\"b2 b8 b11 b21 b22\"/>, and implicate a specific type of underlying neural representation. The current report includes data from a previous study<bibr rid=\"b11\"/> where we report the results of unrelated analyses.',\n",
       " '\\r\\n<p>The experiment was conducted using <i>Bromus inermis</i> (a perennial C3 grass) and its associated mycorrhizal community from the Long-Term Mycorrhiza Research Site in Guelph, Ontario, Canada<bibr rid=\"b16\"/>. inermis</i> was chosen because it is a common plant at the site, shows significant growth responses when associated with AMF<bibr rid=\"b16\"/>, and preliminary studies have demonstrated that the plant significantly increases its dependency on AMF for growth in an elevated atmospheric [CO<sub>2</sub>] environment (data not shown).',\n",
       " 'Of note, single-cell RNA sequencing (RNA-seq) approaches<bibr rid=\"b6 b7\"/> and more refined barcoding lineage-tracking experiments<bibr rid=\"b8\"/> indicate pre-existing lineage specifications in common myeloid progenitors, and a large degree of molecular and functional heterogeneity in all myeloid progenitor populations. They also provide an explanation for the recently reported molecular heterogeneity in GMPs<bibr rid=\"b6 b7 b26\"/>, with the co-existence in variable proportions of distinct steady-state and self-renewing GMPs, and are essential for interpreting the results of lineage-tracking experiments showing major contributions of MPPs and myeloid progenitors to myeloid output<bibr rid=\"b1 b2\"/>. \\r\\n<p>Single-cell index sorting<bibr rid=\"b27\"/> was performed to assess lineage reconstitution as a function of <i>Vwf</i>-tdTomato expression in LSK CD34<super>&#x02212;</super>CD150<super>+</super>CD48<super>&#x02212;</super> BM cells (<figr rid=\"f4\">Fig.',\n",
       " 'Further calculations<bibr rid=\"b35 b36 b37\"/>, performed with larger values of the quark mass, up as high as the charm quark mass, show the same pattern of hybrid mesons, and they are found to be consistently 1.3&#x02009;GeV heavier than the lightest <i>J</i><super><i>PC</i></super>&#x02009;=&#x02009;1<super>&#x02212;&#x02212;</super> meson.',\n",
       " 'Positive candidates obtained in the screen (<sir rid=\"s1\">Supplementary Table 1</sir>) include Na<super>&plus;</super>/Ca<super>2&plus;</super> exchangers and SERCA Ca<super>2&plus;</super> ATPases, the knockdown of which would be expected to increase basal levels of intracellular free Ca<super>2&plus;</super> ([Ca<super>2&plus;</super>]<sub>i</sub>)<bibr rid=\"b12\"/>; the scaffold protein Homer, which has been linked to Ca<super>2&plus;</super> influx and Ca<super>2&plus;</super> homeostasis<bibr rid=\"b13\"/>; stromal interaction molecule (STIM), a recently identified regulator of store-operated Ca<super>2&plus;</super> influx<bibr rid=\"b14 b15 b16\"/> (<sir rid=\"s1\">Supplementary Fig.',\n",
       " '</edsumm></fm>\\r\\n<bdy>\\r\\n<fp>\\r\\n<p>It is well known that the microstructures of the transition-metal oxides<bibr rid=\"b1 b2 b3\"/>, including the high-transition-temperature (high-<i>T</i><sub>c</sub>) copper oxide superconductors<bibr rid=\"b4 b5 b6 b7\"/>, are complex.',\n",
       " 'Recently, ultra-dense resistive memory arrays built from various two-terminal semiconductor or insulator thin film devices have been demonstrated<bibr rid=\"b7 b8 b9 b10 b11 b12\"/>. (<email>b.de.boer@rug.nl</email>).</caff>\\r\\n</aug>\\r\\n<hst><re year=\"2005\" month=\"9\" day=\"6\"/><acc year=\"2006\" month=\"3\" day=\"2\"/></hst>\\r\\n&e060504-7;&nature04699-s1;</fm>\\r\\n<bdy>\\r\\n<fp><p>Electronic transport through single molecules has been studied extensively by academic<bibr rid=\"b1 b2 b3 b4 b5 b6 b7 b8\"/> and industrial<bibr rid=\"b9 b10\"/> research groups. We also reported on the use of bistable &lsqb;2&rsqb;rotaxane molecular monolayers as the storage elements within crossbar memories, using micrometre-scale wiring<bibr rid=\"b20\"/>. \\r\\n<p>The solid-state switching signature of the bistable &lsqb;2&rsqb;rotaxanes that were used here has been shown to originate from electrochemically addressable, molecular mechanical switching for certain device structures<bibr rid=\"b10 b20\"/>, but not for metal wire/molecule/metal wire junctions<bibr rid=\"b25\"/>. Thus, our results are consistent with a molecular mechanism for the switching operation<bibr rid=\"b10 b20\"/>.',\n",
       " '(<email>krasnow@cmgm.stanford.edu</email>)</caff>\\r\\n</aug>\\r\\n<hst><re year=\"2005\" month=\"10\" day=\"30\"/><acc year=\"2006\" month=\"4\" day=\"19\"/></hst>\\r\\n&nature04829-s1;</fm>\\r\\n<bdy>\\r\\n<fp><p>Many organs are composed of tubular networks that arise by branching morphogenesis in which cells bud from an epithelium and organize into a tube<bibr rid=\"b1 b2 b3\"/>.',\n",
       " 'Dominant females develop high testosterone levels in late pregnancy and their offspring are more competitive<bibr rid=\"b21\"/> and achieve higher status than those of subordinates<bibr rid=\"b22\"/>.',\n",
       " 'It is widely thought that interaction of AvrPto with Pto stimulates Pto kinase activity and thereby initiates resistance<bibr rid=\"b5 b13 b14 b15\"/>. The importance of p-Pto(T199) for AvrPto recognition is consistent with previous findings that Pto mutants abolishing the kinase activity also eliminated their interaction with AvrPto<bibr rid=\"b6 b7 b14 b15 b23\"/>. The kinase-deficient mutant Pto(D164N) still interacts with AvrPto, as shown in a previous<bibr rid=\"b14\"/> and this (<figr rid=\"f2\">Fig. \\r\\n<p>Previous studies identified a number of constitutive gain-of-function (CGF) Pto mutants that elicit AvrPto-independent but Prf-dependent hypersensitive response<bibr rid=\"b14 b15\"/>. However, Pto kinase activity is required for AvrPto-triggered resistance<bibr rid=\"b6 b7 b14 b15 b23\"/>. <i>tomato</i> resistance) kinase<bibr rid=\"b17\"/>.',\n",
       " 'This is also supported by satellite observations of sea surface height<bibr rid=\"b225\"/>.',\n",
       " '\\r\\n<p>Genomic resources for studying the model plant <i>Arabidopsis thaliana</i> have begun to provide insight into the epigenetic \\'landscape\\' of this organism<bibr rid=\"b11 b12\"/>. High-resolution mapping of cytosine methylation by using whole-genome microarrays has confirmed previous reports, showing that this modification co-localizes with repeat sequences and with the centromeric regions<bibr rid=\"b11 b12 b15\"/>. Fewer than 5&percnt; of expressed genes were shown to have methylated promoters, although about one-third of genes were methylated in their open reading frame<bibr rid=\"b11 b12\"/>. The significance of methylation in the body of a gene is not fully understood, but such methylation was found to correlate with genes that are both highly transcribed and constitutively expressed<bibr rid=\"b11 b12\"/>. By contrast, genes with methylated promoters had lower expression levels and frequently had tissue-specific expression patterns<bibr rid=\"b11 b12\"/>. thaliana</i><bibr rid=\"b11 b12 b15 b20\"/>. In plants<bibr rid=\"b3 b4 b5\"/>, fission yeast<bibr rid=\"b6 b7 b8\"/>, ciliates<bibr rid=\"b9 b10\"/>, flies<bibr rid=\"b11\"/> and mammalian cells<bibr rid=\"b12 b13\"/>, short interfering RNAs (siRNAs) also induce DNA or chromatin modifications at the homologous genomic locus, which can result in transcriptional silencing or sequence elimination<bibr rid=\"b14\"/>. Finally, in <i>Tetrahymena</i>, an Ago homologue and siRNAs are found in a complex implicated in a silencing-related mechanism that leads to genome rearrangement<bibr rid=\"b16 b17\"/>. Finally, more recently, the RNAi machinery was linked to chromatin regulation in yeast<bibr rid=\"b36\"/>, and to chromosomal rearrangement during development of the somatic macronucleus in <i>Tetrahymena</i><bibr rid=\"b37\"/>. In addition, DNA elimination is associated with Dicer-produced small RNAs, called scan RNAs (scnRNAs), giving rise to the idea that a scnRNA RITS-like complex targets sequences destined for elimination into heterochromatin<bibr rid=\"b40\"/>. melanogaster</i> Ago1 PAZ domain are deposited in the Protein Data Bank under ID code 1R4K.</caff>\\r\\n</aug>\\r\\n<hst><re year=\"2003\" month=\"8\" day=\"4\"/><acc year=\"2003\" month=\"10\" day=\"16\"/><pubdate type=\"aop\" year=\"2003\" month=\"11\" day=\"16\"/></hst>\\r\\n</fm>\\r\\n<bdy>\\r\\n<fp><p>The discovery of RNA-mediated gene-silencing pathways, including RNA interference<bibr rid=\"b1 b2 b3\"/>, highlights a fundamental role of short RNAs in eukaryotic gene regulation<bibr rid=\"b4 b5 b6 b7 b8 b9 b10\"/> and antiviral defence<bibr rid=\"b11 b12\"/>. 3</figr>)<bibr rid=\"b76\"/>; consequent histone methylation then groups these sequences for collective deletion<bibr rid=\"b77\"/>. In <i>Tetrahymena</i>, germ-line-specific small RNAs called scan RNAs derive from the micronuclear-limited sequence and target DNA for elimination<bibr rid=\"b15\"/>.',\n",
       " '\\r\\n<p>Mixing of subducted slabs with ambient mantle is not instantaneous, but is thought to have an exponential dependence on time<bibr rid=\"b15 b16\"/>. Convective motions stretch and juxtapose recycled slab and ambient mantle at increasingly smaller length scales over a characteristic mixing time, <i>&tgr;</i>, which may range between a few hundred million years to a few billion years<bibr rid=\"b15 b16 b17\"/>. This would imply re-mixing of heterogeneities at rates of the order of 1&thinsp;Gyr, which are 10 times longer than previously estimated<bibr rid=\"b27 b28\"/>, but in agreement with recent <super>182</super>W findings<bibr rid=\"b29\"/>.',\n",
       " 'Our hypothesis is that animals understand the space of possible behaviours and their value from previous experience<bibr rid=\"b19\"/>, and that animals adapt by intelligently selecting tests that validate or invalidate whole families of promising compensatory behaviours. Damage recovery in animals may occur without learning&mdash;for instance, due to the built-in robustness of evolved control loops<bibr rid=\"b24\"/>&mdash;but if such pre-programmed robustness fails, many animals turn to learning<bibr rid=\"b19\"/>.',\n",
       " 'All references that we have used to compare with our data are listed<bibr rid=\"b33 b39 b47 b48 b49 b50 b51 b52 b53 b54 b55 b56 b57 b58 b59 b60 b61 b62 b63 b64 b65 b66 b67 b68 b69 b70 b71 b72 b73 b74 b75 b76 b77 b78 b79 b80 b81 b82 b83 b84 b85 b86 b87 b88 b89 b90 b91 b92 b93 b94 b95 b96 b97 b98 b99 b100 b101 b102 b103 b104 b105 b106 b107 b108 b109 b110 b111 b112 b113 b114 b115 b116 b117 b118 b119 b120 b121 b122 b123 b124 b125 b126 b127\"/>.',\n",
       " 'These halogenated compounds range from peptides, polyketides, indoles, terpenes, acetogenins and phenols to volatile halogenated hydrocarbons (for example bromoform, chloroform and dibromomethane) that are produced on a very large scale<bibr rid=\"b1 b2 b3\"/>.',\n",
       " 'As a result, PT symmetry has been observed only in elemental two-component systems<bibr rid=\"b18 b19 b20 b21\"/>.',\n",
       " '</abs>\\n<websumm type=\"toc\">Repeated contacts between the endoplasmic reticulum (ER) and a subset of endosomes called late endosomes (LEs) is shown to promote microtubule-dependent translocation of LEs to the cell periphery and their subsequent fusion with the plasma membrane to induce outgrowth of neuronal protrusions.</websumm>\\n&e150409-14;</fm>\\n<bdy>\\n<p>The unusual presence of a PtdIns(3)P-binding FYVE domain in the ER protein protrudin<bibr rid=\"b7 b8 b9 b10\"/> begged the question as to whether this protein interacts with endosomes, the major PtdIns(3)P-containing organelles in cells<bibr rid=\"b11\"/>. 1g</figr>), as visualized with a 2&times;FYVE probe expressed at a low level<bibr rid=\"b11\"/>. There was no sensor recruitment for a disabled variant with mutations in\\r\\n            the PtdIns3P binding site<bibr rid=\"b17\"/> (<figr rid=\"f2\">Fig. The PH domain of\\r\\n                            PLC&#x003B4;1<bibr rid=\"b12\"/>, the P4M domain of DrrA<bibr\\r\\n                            rid=\"b18\"/>, the tandem FYVE domains of Hrs<bibr rid=\"b17\"/>, the tandem\\r\\n                        PH domains of TAPP1<bibr rid=\"b23\"/>, the triple repeat PHD domains of\\r\\n                            ING2<bibr rid=\"b32\"/>, and the PH domain of Btk<bibr rid=\"b33\"/> were\\r\\n                        amplified by PCR from the corresponding expression vectors, which were gifts\\r\\n                        from T. Balla, H. Stenmark, T. Takenawa and O. Gozani.',\n",
       " '</fp>\\r\\n<p>With the exception of its four centromeric regions, the needle-like, extremely condensed <i>Drosophila</i> sperm nucleus is devoid of the core histones (H2A, H2B, H3 and H4) that form the nucleosome particle<bibr rid=\"b6 b7\"/>. Indeed, in elongating spermatids, histones are replaced by sperm-specific chromosomal proteins including protamines<bibr rid=\"b7\"/>. A recent study has confirmed that in <i>ssm</i> eggs, the paternal chromatin remains associated with protamines<bibr rid=\"b7\"/>.',\n",
       " 'Interestingly, despite the local diversity of orientation preference and receptive field structure in mouse visual cortex, adjacent neurons exhibit specificity in the overlap of their ON and/or OFF subfields that is predictive of connectivity<bibr rid=\"b29 b30\"/>, consistent with the idea that the topology of ON and OFF inputs shapes the organization of cortical circuits even in the absence of cortical columns. RFs of nearby cortical neurons were highly diverse within each imaged population in L2/3<bibr rid=\"b12 b13 b14 b15\"/> (example region in <figr rid=\"f2\">Fig. The final parameters chosen were: <i>&agr;</i>&thinsp;&equals;&thinsp;1; <i>&bgr;</i>&thinsp;&equals;&thinsp;baseline; <i>&dgr;</i>&thinsp;&equals;&thinsp;0.33&thinsp;s; <i>&sgr;</i>&thinsp;&equals;&thinsp;0.16; <i>&phiv;</i>&thinsp;&equals;&thinsp;5&thinsp;Hz; <i>&tgr;</i>&thinsp;&equals;&thinsp;2&thinsp;s. These values were not exactly the same as, but were comparable to, values reported by others in cortical regions<bibr rid=\"b49 b55 b62\"/>. \\n<p>Dendritic regenerative events provide a mechanism by which a relatively small number of inputs can drive spike output<bibr rid=\"b24 b25\"/>, changing the effective connectivity between local functional groups of neurons<bibr rid=\"b26\"/> or mitigating the noise in cortical circuits<bibr rid=\"b27\"/>, by ensuring that variable synaptic input can result in a more reliable postsynaptic response. Here we investigate these questions in networks of layer&thinsp;2/3 (L2/3) pyramidal cells in mouse primary visual cortex (V1)&mdash;where neighbouring neurons exhibit a diversity of preference for visual features<bibr rid=\"b5 b15 b16 b17\"/>&mdash;by determining how local synaptic connectivity relates to visual response properties during development. To characterize the spatial RF structure of L2/3 neurons in V1 at eye opening (postnatal days (P) 14&ndash;15) and in more mature mice (P28&ndash;35), we used <i>in vivo</i> two-photon calcium imaging<bibr rid=\"b23\"/> in monocular V1 to obtain spatial RFs by reverse correlation<bibr rid=\"b16 b17 b24\"/> of static natural images and spiking responses inferred from calcium signals<bibr rid=\"b25\"/> (Methods and <figr rid=\"f1\">Fig.']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clean_text_corpus[:50]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "def tf_idfer(clean_text_corpus):\n",
    "\n",
    "    print('Getting stop words.. ')\n",
    "    stop_words = set(stopwords.words('english')) \n",
    "\n",
    "    print('Tokenizing the sentences..')\n",
    "    clean_text_corpus = [word_tokenize(sent) for sent in clean_text_corpus]\n",
    "\n",
    "    print('Cleaning the sentences..')\n",
    "    cleaner = lambda sent: [w for w in sent if (w not in stop_words) and (w.isalpha())]\n",
    "    clean_text_corpus = list(map(cleaner, clean_text_corpus))\n",
    "\n",
    "    print('Joining the words back to make a sentence..')\n",
    "    clean_text_corpus = [' '.join(i) for i in clean_text_corpus]\n",
    "\n",
    "    print('Vectorizing the sentence..')\n",
    "    vectorizer = CountVectorizer(analyzer='word', ngram_range=(2, 3))\n",
    "    filtered_corpus_vector = vectorizer.fit_transform(clean_text_corpus)\n",
    "    \n",
    "    #print(len(vectorizer.get_feature_names()))\n",
    "    #print(filtered_corpus_vector.toarray())\n",
    "    \n",
    "    return vectorizer, filtered_corpus_vector\n",
    "    "
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "stop_words = set(stopwords.words('english')) \n",
    "\n",
    "clean_text_corpus = [word_tokenize(sent) for sent in clean_text_corpus]\n",
    "\n",
    "cleaner = lambda sent: [w for w in sent if (w not in stop_words) and (w.isalpha())]\n",
    "\n",
    "clean_text_corpus = list(map(cleaner, clean_text_corpus))\n",
    "\n",
    "clean_text_corpus = [' '.join(i) for i in clean_text_corpus]\n",
    "\n",
    "vectorizer = CountVectorizer(analyzer='word', ngram_range=(3, 3))\n",
    "filtered_corpus_vector = vectorizer.fit_transform(clean_text_corpus)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Getting stop words.. \n",
      "Tokenizing the sentences..\n",
      "Cleaning the sentences..\n",
      "Joining the words back to make a sentence..\n",
      "Vectorizing the sentence..\n",
      "CPU times: user 5min 3s, sys: 3.74 s, total: 5min 7s\n",
      "Wall time: 5min 7s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "vectorizer, filtered_corpus_vector = tf_idfer(clean_text_corpus)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### The one with no mapping has a lot of sentences. This should be removed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[242, 176, 130, 654, 229, 284, 183, 147, 146, 302]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[len(i) for i in clean_text_corpus[:10]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "trigrams = vectorizer.get_feature_names()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "transformer = TfidfTransformer(smooth_idf=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "tf_idf = transformer.fit_transform(filtered_corpus_vector)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "#tf_idf = tf_idf[1:]\n",
    "#filtered_corpus_vector = filtered_corpus_vector[1:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(304448, 8566042)"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_corpus_vector.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(304448, 8566042)"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tf_idf.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "304448"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(matched_dfs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Looking into the tf-idf for coot to see if it makes sense."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ArticleID                                        pub.1020281098\n",
       "text          [['This model was combined into a single prote...\n",
       "clean_text    [This model was combined into a single protein...\n",
       "count                                                       520\n",
       "Title         Coot: model-building tools for molecular graphics\n",
       "Name: 92203, dtype: object"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matched_dfs.loc[92203]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'using coot'"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trigrams[np.argmax(tf_idf[92203])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Coot: model-building tools for molecular graphics'"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matched_dfs.loc[92203]['Title']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "using coot\n",
      "the model\n",
      "program coot\n",
      "model built\n",
      "model building\n",
      "coot the\n",
      "coot refinement\n",
      "coot refined\n",
      "coot phenix\n",
      "building coot\n"
     ]
    }
   ],
   "source": [
    "for i in sorted(np.argsort(tf_idf[92203].todense())[:5].tolist()[0][-10:], reverse = True):\n",
    "    print(trigrams[i])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Woo! It makes sense! Trying on other papers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\n",
      "Paper Quantum phase transition from a superfluid to a Mott insulator in a gas of ultracold atoms \n",
      "Key Phrases:\n",
      "\n",
      "transition superfluid\n",
      "quantum phase transition\n",
      "quantum phase\n",
      "phase transition\n",
      "optical lattice\n",
      "mott insulator\n",
      "mott insulating state\n",
      "mott insulating\n",
      "lattice potential\n",
      "insulating state\n",
      "\n",
      "\n",
      "\n",
      "Paper Sequence and comparative analysis of the chicken genome provide unique perspectives on vertebrate evolution \n",
      "Key Phrases:\n",
      "\n",
      "ucd line\n",
      "red jungle fowl\n",
      "red jungle\n",
      "partly inbred ucd\n",
      "partly inbred\n",
      "jungle fowl\n",
      "inbred ucd line\n",
      "inbred ucd\n",
      "female partly inbred\n",
      "chicken genome\n",
      "\n",
      "\n",
      "\n",
      "Paper Role of Tet proteins in 5mC to 5hmC conversion, ES-cell self-renewal and inner cell mass specification \n",
      "Key Phrases:\n",
      "\n",
      "tet proteins\n",
      "tet family\n",
      "stem es cells\n",
      "stem es\n",
      "mouse es cells\n",
      "mouse es\n",
      "massachusetts usa\n",
      "es cells\n",
      "embryonic stem es\n",
      "embryonic stem\n",
      "\n",
      "\n",
      "\n",
      "Paper Structure of the TRPV1 ion channel determined by electron cryo-microscopy \n",
      "Key Phrases:\n",
      "\n",
      "trp channels\n",
      "trp channel structures\n",
      "previously described\n",
      "in vgics\n",
      "distinct trp\n",
      "detergents amphipols\n",
      "channels contain\n",
      "channel structures\n",
      "channel gating\n",
      "amphipathic polymer\n",
      "\n",
      "\n",
      "\n",
      "Paper Oncogenic ras Provokes Premature Cell Senescence Associated with Accumulation of p53 and p16INK4a \n",
      "Key Phrases:\n",
      "\n",
      "transformation phenotypes mefs\n",
      "transformation phenotypes\n",
      "primary cells\n",
      "premature senescence\n",
      "phenotypes mefs\n",
      "oncogenic ras\n",
      "oncogenic mutations\n",
      "neoplastic transformation phenotypes\n",
      "expression oncogenic\n",
      "cellular senescence\n"
     ]
    }
   ],
   "source": [
    "test_df = matched_dfs.sort_values('count', ascending = False).iloc[:300].sample(5)\n",
    "\n",
    "for index, paper in test_df.iterrows():\n",
    "    \n",
    "    print('\\n\\n')\n",
    "    print('Paper %s ' % paper['Title'])\n",
    "    print('Key Phrases:\\n')\n",
    "    for i in sorted(np.argsort(tf_idf[index].todense())[:5].tolist()[0][-10:], reverse = True):\n",
    "        print(trigrams[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Subjects were assigned a location in a network of 20 nodes, generated by a preferential attachment model; the network structure was created de novo for each session by attaching new nodes (each with two links) to existing nodes; and subjects were placed into the resulting networks at random.',\n",
       " 'A soft thresholding power was chosen on the basis of the criterion of approximate scale-free topology.',\n",
       " 'ÁngelesSerranoM\\nMariánBoguñáM\\nDmitriKrioukovD\\nDepartment of Electrical Engineering, Computer Engineering and Informatics, Cyprus University of Technology, 33 Saripolou Street, 3036 Limassol, Cyprus\\nCooperative Association for Internet Data Analysis (CAIDA), University of California, San Diego (UCSD), La Jolla, California 92093, USA\\nDepartament de Física Fonamental, Universitat de Barcelona, Martí i Franquès 1, 08028 Barcelona, Spain\\nf.papadopoulos@cut.ac.cy\\ndima@ucsd.edu\\nA framework is developed in which new connections to a growing network optimize geometric trade-offs between popularity and similarity, instead of simply preferring popular nodes; this approach accurately describes the large-scale evolution of various networks.&e120927-5;\\r\\n\\n\\nThe principle that ‘popularity is attractive’ underlies preferential attachment, which is a common explanation for the emergence of scaling in growing networks.',\n",
       " 'The networks grown as described may seem to have nothing in common with preferential attachment (PA).',\n",
       " 'We also explore evolution on random and scale-free networks.',\n",
       " 'This procedure is applied to several different real networks: (1) a part of the WWW composed of 325,729 web pages that are connected if there is a URL link from one page to another (http://www.nd.edu/~networks); (2) a social network where the nodes are 392,340 actors linked if they were cast together in at least one film; (3) the biological networks of protein–protein interactions found in Escherichia coli (429 proteins) and Homo sapiens (946 proteins) linked if there is a physical binding between them (database available via the Database of Interacting Proteins, other PINs are discussed in the Supplementary Information), and (4) the cellular networks compiled by ref.',\n",
       " '12 in the context of scale-free networks, such as social networks, protein interaction networks or metabolic networks.',\n",
       " '15: social networks, the world wide web, metabolic networks, and yeast protein interaction networks (PIN).',\n",
       " '',\n",
       " 'As is typical of networks, the degree distributions have significantly fatter tails than a similarly sized group composed of individuals randomly forming the same number of social ties (Kolmogorov–Smirnov test, P\\u2009<\\u200910−15 for all comparisons).',\n",
       " 'These findings imply that the observed flux distribution is a generic feature of flux conservation on a scale-free network, as it is independent of the optimal or non-optimal nature of the growth rate or the growth conditions.',\n",
       " '18 and Supplementary Information), coupled with the presence of a power law in both the optimal and non-optimal flux states, suggests that the metabolic flux organization is a direct consequence of the scale-free topology of the network.',\n",
       " '(vicsek@angel.elte.hu).\\r\\n\\r\\n\\r\\n&e050609-17;\\r\\n\\r\\n\\nMany complex systems in nature and society can be described in terms of networks capturing the intricate web of connections among the units they are made of.',\n",
       " 'It is well known that the nodes of large real networks have a power-law degree distribution.',\n",
       " 'Here, a better fit is achieved with a scale-free network; however, neither model simultaneously fits both distributions.',\n",
       " '(buldyrev@yu.edu).\\r\\n\\r\\n&e100415-7; &nature08932-s1;\\r\\n\\n\\nComplex networks have been studied intensively for a decade, but research still focuses on the limited case of a single, non-interacting network.',\n",
       " 'For two interdependent scale-free networks with power-law degree distributions, PA(k) = PB(k)\\u2009∝\\u2009k-&lgr;, we find that the existence criteria for the giant component are quite different from those in a single network.',\n",
       " '1) are as follows: 100,000 Erdös–Rényi (ER) with edge probability 0\\u2009<\\u2009p\\u2009<\\u20091; 100,000 small world (SW) with initial connection distance 1\\u2009≤\\u2009d\\u2009≤\\u20095 and edge creation probability 0\\u2009<\\u2009p\\u2009<\\u20090.05; 100,000 Barabási–Albert (BA) with linking number 1\\u2009≤\\u2009m\\u2009≤\\u200910; 100,000 random recursive (RR) (like Barabási–Albert except that edges are added uniformly instead of preferentially) with linking number 2\\u2009≤\\u2009m\\u2009≤\\u20098; 200,000 Holme–Kim (HK) with linking number 2\\u2009≤\\u2009m\\u2009≤\\u20094 and triad formation parameter 0\\u2009<\\u2009p\\u2009<\\u20090.15; 200,000 Klemm–Eguiluz (KE) with linking number 3\\u2009≤\\u2009m\\u2009≤\\u20095 and deactivation parameter 0\\u2009<\\u2009μ\\u2009<\\u20090.15; 200,000 shifted-linear preferential attachment with linking number 1\\u2009≤\\u2009m\\u2009≤\\u20097 and shift 0\\u2009<\\u2009θ\\u2009<\\u200940; 100,000 forest fire (FF) with parameters 0\\u2009<\\u2009pf\\u2009<\\u2009pb\\u2009<\\u20090.15; 100,000 Island Barabási–Albert; and 100,000 Island Erdös–Rényi.',\n",
       " '2a, b shows for two canonical network models (Erd&odblac;s–Rényi and scale-free), the fraction of driver nodes is significantly higher among low-k nodes than among the hubs.',\n",
       " '2c), indicating scale-free network topology.',\n",
       " 'Previous studies have shown that highly connected proteins within a network tend to be more highly conserved, a consequence of either functional constraints or preferential interaction of new proteins with existing highly connected proteins.',\n",
       " 'As demonstrated by Barabasi and colleagues and by several other researchers, power laws describe the distribution of various quantities in numerous biological, physical and social contexts; such distributions can seem to be fundamentally different (for example, the number of links between documents in the Internet, the population of towns and the number of reactions in which a given metabolite is involved).',\n",
       " 'Lower-tier nodes are more sparsely connected and the network structure has a scale-free power-law degree distribution.',\n",
       " 'Here we describe a surprisingly simple rule that is a good approximation for all graphs that we have analysed, including cycles, spatial lattices, random regular graphs, random graphs and scale-free networks: natural selection favours cooperation, if the benefit of the altruistic act, b, divided by the cost, c, exceeds the average number of neighbours, k, which means b/c > k.',\n",
       " 'Moreover, we find that the rule also holds for random graphs and scale-free networks, where individuals differ in the number of their neighbours.',\n",
       " '(vicsek@angel.elte.hu).\\r\\n\\r\\n&e070405-11;\\r\\n\\n\\nThe rich set of interactions between individuals in society results in complex community structure, capturing highly connected circles of friends, families or professional cliques in a social network.',\n",
       " 'The analogy to networks is also noteworthy: if poorly connected nodes (interacting cells) are damaged then the system will repair itself, whereas if strongly connected nodes are damaged the self-organization will disappear.',\n",
       " 'When social ties follow a scale-free distribution, cooperation is enhanced whenever all individuals are expected to contribute a fixed amount irrespective of the plethora of public goods games in which they engage.',\n",
       " 'Such strongly heterogeneous populations were obtained by means of the Barabási–Albert scale-free model based on growth and preferential attachment.',\n",
       " 'This implies that the smallest group in this population has three individuals.',\n",
       " 'b, Scale-free graphs in which small-world effects coexist with a large heterogeneity in neighbourhood size.']"
      ]
     },
     "execution_count": 189,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matched_dfs.iloc[45845]['clean_text']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
