{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scispacy\n",
    "import spacy\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pickle as pkl\n",
    "import zipfile\n",
    "import os\n",
    "import xmltodict\n",
    "import json\n",
    "import zlib\n",
    "import seaborn as sns\n",
    "import resource\n",
    "import sys\n",
    "import regex as re\n",
    "from bs4 import BeautifulSoup\n",
    "from tqdm import tqdm_notebook as tqdm\n",
    "import nltk.data\n",
    "import time\n",
    "import json\n",
    "from multiprocessing import Pool\n",
    "from nltk.tokenize import word_tokenize\n",
    "from numpy.testing import assert_array_equal\n",
    "import threading\n",
    "import networkx as nx\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "from sklearn.metrics.pairwise import cosine_similarity\n",
    "from scipy import sparse\n",
    "from mlxtend.preprocessing import TransactionEncoder\n",
    "from datasketch import MinHash, MinHashLSH\n",
    "#nltk.download('punkt')\n",
    "sent_detector = nltk.data.load('tokenizers/punkt/english.pickle')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Increase Recursion Depth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(8388608, -1)\n",
      "3000\n"
     ]
    }
   ],
   "source": [
    "print (resource.getrlimit(resource.RLIMIT_STACK))\n",
    "print (sys.getrecursionlimit())\n",
    "\n",
    "max_rec = 0x100000\n",
    "\n",
    "resource.setrlimit(resource.RLIMIT_STACK, [0x100 * max_rec, resource.RLIM_INFINITY])\n",
    "sys.setrecursionlimit(max_rec)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1: Explore the Dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "article_df = pd.read_hdf('data/nature/preprocessed/article_df.hdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ArticleID</th>\n",
       "      <th>DOI</th>\n",
       "      <th>Issue</th>\n",
       "      <th>Volume</th>\n",
       "      <th>PageEnd</th>\n",
       "      <th>PageStart</th>\n",
       "      <th>PubYear</th>\n",
       "      <th>SortDate</th>\n",
       "      <th>Title</th>\n",
       "      <th>Type</th>\n",
       "      <th>Journal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>048485a0</td>\n",
       "      <td>10.1038/048485a0</td>\n",
       "      <td>1247.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>508.0</td>\n",
       "      <td>485.0</td>\n",
       "      <td>1893</td>\n",
       "      <td>1893-09-21</td>\n",
       "      <td>The British Association</td>\n",
       "      <td>ed</td>\n",
       "      <td>Nature</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>048508a0</td>\n",
       "      <td>10.1038/048508a0</td>\n",
       "      <td>1247.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>509.0</td>\n",
       "      <td>508.0</td>\n",
       "      <td>1893</td>\n",
       "      <td>1893-09-21</td>\n",
       "      <td>The Thieving of Assyrian Antiquities\\r\\n</td>\n",
       "      <td>lt</td>\n",
       "      <td>Nature</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>048509a0</td>\n",
       "      <td>10.1038/048509a0</td>\n",
       "      <td>1247.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>509.0</td>\n",
       "      <td>509.0</td>\n",
       "      <td>1893</td>\n",
       "      <td>1893-09-21</td>\n",
       "      <td>Bishop's Ring\\r\\n</td>\n",
       "      <td>lt</td>\n",
       "      <td>Nature</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>048509b0</td>\n",
       "      <td>10.1038/048509b0</td>\n",
       "      <td>1247.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>509.0</td>\n",
       "      <td>509.0</td>\n",
       "      <td>1893</td>\n",
       "      <td>1893-09-21</td>\n",
       "      <td>Spring and Autumn of 1893\\r\\n</td>\n",
       "      <td>lt</td>\n",
       "      <td>Nature</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>048509c0</td>\n",
       "      <td>10.1038/048509c0</td>\n",
       "      <td>1247.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>512.0</td>\n",
       "      <td>509.0</td>\n",
       "      <td>1893</td>\n",
       "      <td>1893-09-21</td>\n",
       "      <td>Notes</td>\n",
       "      <td>misc</td>\n",
       "      <td>Nature</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395498</th>\n",
       "      <td>149145b0</td>\n",
       "      <td>10.1038/149145b0</td>\n",
       "      <td>3770.0</td>\n",
       "      <td>149.0</td>\n",
       "      <td>146.0</td>\n",
       "      <td>145.0</td>\n",
       "      <td>1942</td>\n",
       "      <td>1942-01-31</td>\n",
       "      <td>NATIVE SUBSISTENCE ON THE AMERICAN CENTRAL PLAINS</td>\n",
       "      <td>nw</td>\n",
       "      <td>Nature</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395499</th>\n",
       "      <td>149146a0</td>\n",
       "      <td>10.1038/149146a0</td>\n",
       "      <td>3770.0</td>\n",
       "      <td>149.0</td>\n",
       "      <td>146.0</td>\n",
       "      <td>146.0</td>\n",
       "      <td>1942</td>\n",
       "      <td>1942-01-31</td>\n",
       "      <td>EARTHQUAKE ORIGINS IN THE NEW ZEALAND REGION</td>\n",
       "      <td>nw</td>\n",
       "      <td>Nature</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395500</th>\n",
       "      <td>149146b0</td>\n",
       "      <td>10.1038/149146b0</td>\n",
       "      <td>3770.0</td>\n",
       "      <td>149.0</td>\n",
       "      <td>146.0</td>\n",
       "      <td>146.0</td>\n",
       "      <td>1942</td>\n",
       "      <td>1942-01-31</td>\n",
       "      <td>FORTHCOMING EVENTS</td>\n",
       "      <td>misc</td>\n",
       "      <td>Nature</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395501</th>\n",
       "      <td>149146c0</td>\n",
       "      <td>10.1038/149146c0</td>\n",
       "      <td>3770.0</td>\n",
       "      <td>149.0</td>\n",
       "      <td>146.0</td>\n",
       "      <td>146.0</td>\n",
       "      <td>1942</td>\n",
       "      <td>1942-01-31</td>\n",
       "      <td>APPOINTMENTS VACANT</td>\n",
       "      <td>misc</td>\n",
       "      <td>Nature</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395502</th>\n",
       "      <td>149146d0</td>\n",
       "      <td>10.1038/149146d0</td>\n",
       "      <td>3770.0</td>\n",
       "      <td>149.0</td>\n",
       "      <td>146.0</td>\n",
       "      <td>146.0</td>\n",
       "      <td>1942</td>\n",
       "      <td>1942-01-31</td>\n",
       "      <td>REPORTS and Other PUBLICATIONS</td>\n",
       "      <td>misc</td>\n",
       "      <td>Nature</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>395503 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       ArticleID               DOI   Issue  Volume  PageEnd  PageStart  \\\n",
       "0       048485a0  10.1038/048485a0  1247.0    48.0    508.0      485.0   \n",
       "1       048508a0  10.1038/048508a0  1247.0    48.0    509.0      508.0   \n",
       "2       048509a0  10.1038/048509a0  1247.0    48.0    509.0      509.0   \n",
       "3       048509b0  10.1038/048509b0  1247.0    48.0    509.0      509.0   \n",
       "4       048509c0  10.1038/048509c0  1247.0    48.0    512.0      509.0   \n",
       "...          ...               ...     ...     ...      ...        ...   \n",
       "395498  149145b0  10.1038/149145b0  3770.0   149.0    146.0      145.0   \n",
       "395499  149146a0  10.1038/149146a0  3770.0   149.0    146.0      146.0   \n",
       "395500  149146b0  10.1038/149146b0  3770.0   149.0    146.0      146.0   \n",
       "395501  149146c0  10.1038/149146c0  3770.0   149.0    146.0      146.0   \n",
       "395502  149146d0  10.1038/149146d0  3770.0   149.0    146.0      146.0   \n",
       "\n",
       "        PubYear    SortDate  \\\n",
       "0          1893  1893-09-21   \n",
       "1          1893  1893-09-21   \n",
       "2          1893  1893-09-21   \n",
       "3          1893  1893-09-21   \n",
       "4          1893  1893-09-21   \n",
       "...         ...         ...   \n",
       "395498     1942  1942-01-31   \n",
       "395499     1942  1942-01-31   \n",
       "395500     1942  1942-01-31   \n",
       "395501     1942  1942-01-31   \n",
       "395502     1942  1942-01-31   \n",
       "\n",
       "                                                    Title  Type Journal  \n",
       "0                                 The British Association    ed  Nature  \n",
       "1                The Thieving of Assyrian Antiquities\\r\\n    lt  Nature  \n",
       "2                                       Bishop's Ring\\r\\n    lt  Nature  \n",
       "3                           Spring and Autumn of 1893\\r\\n    lt  Nature  \n",
       "4                                                   Notes  misc  Nature  \n",
       "...                                                   ...   ...     ...  \n",
       "395498  NATIVE SUBSISTENCE ON THE AMERICAN CENTRAL PLAINS    nw  Nature  \n",
       "395499       EARTHQUAKE ORIGINS IN THE NEW ZEALAND REGION    nw  Nature  \n",
       "395500                                 FORTHCOMING EVENTS  misc  Nature  \n",
       "395501                                APPOINTMENTS VACANT  misc  Nature  \n",
       "395502                     REPORTS and Other PUBLICATIONS  misc  Nature  \n",
       "\n",
       "[395503 rows x 11 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "article_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7ff906033668>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA3cAAAJCCAYAAABj8z68AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvnQurowAAIABJREFUeJzt3X2QpVddJ/DvjwwvMdEkCE6FJO7EIuhGRhCmAru66wTKEIi1YUtENAUJZjdba6hFN5YMlhqXl61BeREXxU2ZSPBtjAiVLBOM2WgX5UsUwtsQEBlhgIyRlEwcHF43ePaP+ww0k56Z7unb9z59+vOp6urnnvvce8/tXz/d93vPec6t1loAAABY3x4y7w4AAACwesIdAABAB4Q7AACADgh3AAAAHRDuAAAAOiDcAQAAdEC4AwAA6IBwBwAA0AHhDgAAoAOb5t2BY3nUox7VtmzZMu9ubEif+9zncsopp8y7GxxBXcZJXcZJXcZJXcZJXcZJXcZp1nW56667/rG19ujl7DvqcLdly5a8+93vnnc3NqSFhYVs37593t3gCOoyTuoyTuoyTuoyTuoyTuoyTrOuS1V9Yrn7mpYJAADQAeEOAACgA8IdAABAB4Q7AACADgh3AAAAHRDuAAAAOiDcAQAAdEC4AwAA6IBwBwAA0AHhDgAAoAPCHQAAQAeEOwAAgA4IdwAAAB0Q7gAAADog3AEAAHRAuAMAAOiAcAcAANAB4Q4AAKADwh0AAEAHhDsAAIAOCHcAAAAdEO4AAAA6INwBAAB0QLgDAADowKZ5dwAAYDm27Ni9ZPu+nZfMuCcA42TkDgAAoAPCHQAAQAeEOwAAgA445w4AWNeWOhfPeXjARiTcAQAbghAI9M60TAAAgA4IdwAAAB0Q7gAAADog3AEAAHRAuAMAAOiAcAcAANAB4Q4AAKADwh0AAEAHhDsAAIAObJp3BwAAjrRlx+55dwFg3TFyBwAA0AHhDgAAoAOmZQIAc2UKJsB0CHcAQHcERmAjMi0TAACgA8IdAABAB4Q7AACADgh3AAAAHRDuAAAAOiDcAQAAdEC4AwAA6IBwBwAA0AHhDgAAoAPCHQAAQAc2zbsDAADzsmXH7ge17dt5yRx6ArB6Ru4AAAA6INwBAAB0QLgDAADogHAHAADQAeEOAACgA8IdAABAB4Q7AACADgh3AAAAHRDuAAAAOiDcAQAAdEC4AwAA6IBwBwAA0AHhDgAAoAPCHQAAQAeEOwAAgA4IdwAAAB0Q7gAAADog3AEAAHRg07w7AABsHHv2H8wVO3bPuxtTsWWJ57Fv5yVz6AnAhJE7AACADgh3AAAAHRDuAAAAOiDcAQAAdEC4AwAA6IBwBwAA0AHhDgAAoAM+5w4AYBGfXwesV0buAAAAOiDcAQAAdGBZ4a6q9lXVnqp6X1W9e2h7ZFXdXlUfHb6fMbRXVf1KVe2tqg9U1ZMW3c/lw/4frarL1+YpAQAAbDwrGbm7sLX2xNbatuHyjiR3tNbOS3LHcDlJnpnkvOHrqiRvTCZhMMm1SZ6S5IIk1x4OhAAAAKzOaqZlXprkxmH7xiTPXtT+5jZxZ5LTq+rMJM9Icntr7UBr7f4ktye5eBWPDwAAwGC5q2W2JH9cVS3J/26tXZdkc2vt3uH6f0iyedg+K8mnFt32nqHtaO0AAKO21AqaAGOz3HD3va21/VX1LUlur6q/WXxla60NwW/VquqqTKZzZvPmzVlYWJjG3bJChw4d8rMfIXUZJ3UZJ3UZp80nJ9dsfWDe3Vgz6/V3zvEyTuoyTmOuy7LCXWtt//D9vqp6WybnzH26qs5srd07TLu8b9h9f5JzFt387KFtf5LtR7QvLPFY1yW5Lkm2bdvWtm/ffuQuzMDCwkL87MdHXcZJXcZJXcbpf/3OzXnNnn4/ZnffZdvn3YUT4ngZJ3UZpzHX5bjn3FXVKVX1jYe3k1yU5INJbklyeMXLy5PcPGzfkuQFw6qZT01ycJi+eVuSi6rqjGEhlYuGNgAAAFZpOW+dbU7ytqo6vP/vttb+qKreleSmqroyySeSPHfY/9Ykz0qyN8nnk7wwSVprB6rq5UneNez3stbagak9EwAAgA3suOGutfaxJE9Yov0zSZ6+RHtLcvVR7uuGJDesvJsAAAAcy2o+CgEAAICREO4AAAA6INwBAAB0QLgDAADogHAHAADQAeEOAACgA8IdAABAB4Q7AACADgh3AAAAHRDuAAAAOiDcAQAAdEC4AwAA6IBwBwAA0AHhDgAAoAPCHQAAQAeEOwAAgA4IdwAAAB3YNO8OAAB92rJj94Partk6h44AbBBG7gAAADog3AEAAHRAuAMAAOiAcAcAANAB4Q4AAKADwh0AAEAHhDsAAIAOCHcAAAAdEO4AAAA6INwBAAB0QLgDAADogHAHAADQgU3z7gAAsP5t2bF73l0A2PCM3AEAAHRAuAMAAOiAcAcAANAB4Q4AAKADwh0AAEAHhDsAAIAOCHcAAAAdEO4AAAA6INwBAAB0QLgDAADogHAHAADQAeEOAACgA8IdAABAB4Q7AACADgh3AAAAHRDuAAAAOiDcAQAAdEC4AwAA6IBwBwAA0AHhDgAAoAPCHQAAQAeEOwAAgA4IdwAAAB0Q7gAAADog3AEAAHRAuAMAAOiAcAcAANAB4Q4AAKADwh0AAEAHhDsAAIAOCHcAAAAdEO4AAAA6INwBAAB0QLgDAADogHAHAADQAeEOAACgA8IdAABAB4Q7AACADgh3AAAAHRDuAAAAOiDcAQAAdEC4AwAA6IBwBwAA0AHhDgAAoAPCHQAAQAeEOwAAgA4IdwAAAB0Q7gAAADog3AEAAHRAuAMAAOiAcAcAANCBZYe7qjqpqt5bVW8fLp9bVX9VVXur6ver6mFD+8OHy3uH67csuo+XDu0fqapnTPvJAAAAbFQrGbl7cZIPL7r8qiSva609Nsn9Sa4c2q9Mcv/Q/rphv1TV+Umel+Q7k1yc5Neq6qTVdR8AAIBkmeGuqs5OckmS3xguV5KnJXnLsMuNSZ49bF86XM5w/dOH/S9Nsqu19qXW2seT7E1ywTSeBAAAwEa33JG7X07y00n+Zbj8zUn+qbX2wHD5niRnDdtnJflUkgzXHxz2/2r7ErcBAABgFTYdb4eq+oEk97XW7qqq7Wvdoaq6KslVSbJ58+YsLCys9UOyhEOHDvnZj5C6jJO6jJO6zNY1Wx84/k5JNp+8/H3Xo/X6O+d4GSd1Gacx1+W44S7J9yT5D1X1rCSPSPJNSV6f5PSq2jSMzp2dZP+w//4k5yS5p6o2JTktyWcWtR+2+DZf1Vq7Lsl1SbJt27a2ffv2E3harNbCwkL87MdHXcZJXcZJXWbrih27l7XfNVsfyGv2LOflx/q077Lt8+7CCXG8jJO6jNOY63Lcv66ttZcmeWmSDCN3P9Vau6yq/iDJc5LsSnJ5kpuHm9wyXP7L4fo/aa21qrolye9W1WuTPCbJeUn+erpPBwCA9W7LEm8W7Nt5yRx6AuvLat46e0mSXVX1iiTvTXL90H59kt+qqr1JDmSyQmZaa3dX1U1JPpTkgSRXt9a+sorHBwAAYLCicNdaW0iyMGx/LEusdtla+2KSHzrK7V+Z5JUr7SQAAADHtpLPuQMAAGCkhDsAAIAO9LtcFQAAwBJ6XbRHuAMAGIleX3ACs2FaJgAAQAeEOwAAgA4IdwAAAB0Q7gAAADog3AEAAHTAapkAACNmBU1guYzcAQAAdEC4AwAA6IBpmQAAjJ7pqXB8Ru4AAAA6INwBAAB0wLRMAAC6YfomG5mROwAAgA4YuQMAYF1aapQONjLhDgBYES+oAcbJtEwAAIAOGLkDAJgSi3kA8yTcAQAAJ8QbGuNiWiYAAEAHhDsAAIAOCHcAAAAdEO4AAAA6YEEVAADmxucmwvQIdwDAUXnhDbB+mJYJAADQASN3AAAdGNvnjY2tP7ARGLkDAADogHAHAADQAeEOAACgA865AwBYZ8a2iunY+gMblXAHAMCGZNEXemNaJgAAQAeM3AEAMBOmb8LaEu4AADpl2iFsLKZlAgAAdMDIHQAAXVvJdFCjnaxnwh0A64oXXgCwNOEOAGCD27P/YK444o0Tb5rA+uOcOwAAgA4YuQMA4EFMgYb1x8gdAABAB4Q7AACADpiWCQCwgSw13fKarXPoCN0ypXd+hDsAgBlbyeeuMX/CCuuFaZkAAAAdEO4AAAA6YFomADN1tOlopjjB+JlOCuNm5A4AAKADwh0AAEAHhDsAAIAOOOcOgNFyfg9AH5xvPRvCHQBsMF5kAfTJtEwAAIAOGLkDYBRMwQSA1TFyBwAA0AHhDgAAoAOmZQKw7i01pfNNF58yh54AfL2l/j5ZvIi1ItwBAMAI7dl/MFccEQ4FQ45FuAMAgDlbaoTvmq1z6MhgniOORjtPnHAHwJqxAiYA87aR/hcJdwAAMEMbKWwwW8IdAACskIDGGAl3ADAizjXpjxAAzIpwBwAAG9g834CY9mNv9DdThDsAAFgnjhZejPCTJA+ZdwcAAABYPSN3AKyY88LWj40+RQlgIzFyBwAA0AHhDgAAoAOmZQLAnJgyCcA0CXcAQBJhE2C9My0TAACgA8IdAABAB4Q7AACADjjnDgBGzucKArAcwh0AU2ExjtkS+AA4knAHAACsO95UfDDhDgA64YUOwMZ23HBXVY9I8s4kDx/2f0tr7dqqOjfJriTfnOSuJM9vrX25qh6e5M1JnpzkM0l+uLW2b7ivlya5MslXkvy31tpt039KAABAT7x5tTzLGbn7UpKntdYOVdVDk/xZVb0jyX9P8rrW2q6q+vVMQtsbh+/3t9YeW1XPS/KqJD9cVecneV6S70zymCT/t6oe11r7yho8LwAA4AhCUt+O+1EIbeLQcPGhw1dL8rQkbxnab0zy7GH70uFyhuufXlU1tO9qrX2ptfbxJHuTXDCVZwEAALDBVWvt+DtVnZTJ1MvHJvnVJL+U5M7W2mOH689J8o7W2uOr6oNJLm6t3TNc93dJnpLkF4bb/PbQfv1wm7cc8VhXJbkqSTZv3vzkXbt2TeN5skKHDh3KqaeeOu9ucAR1GaeNWJc9+w/OuwvHde5pJ42+Luvh5zhtm09OPv2FefeCI6nLOK2kLlvPOm1Z+63m785Sj9HT37Hl/gxn/X//wgsvvKu1tm05+y5rQZVh6uQTq+r0JG9L8h2r6N/xHuu6JNclybZt29r27dvX6qE4hoWFhfjZj4+6jNNGrMsV62Baz5suPmX0dVkPP8dpu2brA3nNHuu5jY26jNNK6rLvsu3L2m81f3eWeoye/o4t92c45v/7x52WuVhr7Z+S/GmSf5Pk9Ko6/Nt2dpL9w/b+JOckyXD9aZksrPLV9iVuAwAAwCocN9xV1aOHEbtU1clJvj/JhzMJec8Zdrs8yc3D9i3D5QzX/0mbzP28Jcnzqurhw0qb5yX562k9EQAAgI1sOeO8Zya5cTjv7iFJbmqtvb2qPpRkV1W9Isl7k1w/7H99kt+qqr1JDmSyQmZaa3dX1U1JPpTkgSRXWykTAADWhpUxN57jhrvW2geSfPcS7R/LEqtdtta+mOSHjnJfr0zyypV3EwAAgGNx5iwAAKxzRulIVrigCgAAAONk5A4Apmypd9D37bxkDj0BYCMxcgcAANAB4Q4AAKADwh0AAEAHhDsAAIAOCHcAAAAdEO4AAAA6INwBAAB0QLgDAADogA8xBwAAjmvLjt3z7gLHYeQOAACgA8IdAABAB4Q7AACADgh3AAAAHbCgCgBd2rP/YK444uT/fTsvmVNvAGDtGbkDAADogHAHAADQAeEOAACgA8IdAABAB4Q7AACADgh3AAAAHRDuAAAAOiDcAQAAdEC4AwAA6IBwBwAA0AHhDgAAoAPCHQAAQAc2zbsDALARbNmxe95dAKBzRu4AAAA6INwBAAB0QLgDAADogHAHAADQAeEOAACgA8IdAABAB4Q7AACADgh3AAAAHRDuAAAAOiDcAQAAdEC4AwAA6IBwBwAA0AHhDgAAoAPCHQAAQAeEOwAAgA4IdwAAAB0Q7gAAADog3AEAAHRAuAMAAOiAcAcAANAB4Q4AAKADwh0AAEAHhDsAAIAOCHcAAAAdEO4AAAA6INwBAAB0YNO8OwAA68GWHbsf1LZv5yVz6AkALM3IHQAAQAeEOwAAgA4IdwAAAB0Q7gAAADog3AEAAHRAuAMAAOiAj0IAgBO01McjAMC8GLkDAADogHAHAADQAeEOAACgA8IdAABAB4Q7AACADlgtE4ANY7mrW+7becka9wQAps/IHQAAQAeM3AFwTD7LDQDWByN3AAAAHRDuAAAAOiDcAQAAdEC4AwAA6IBwBwAA0AGrZQLMyFKrTq7F56mt5nGsjDnh5wDAeiTcAczRrALfch8bAFi/TMsEAADogHAHAADQAdMyAUZmudMlZzV9EwBYH44b7qrqnCRvTrI5SUtyXWvt9VX1yCS/n2RLkn1Jnttau7+qKsnrkzwryeeTXNFae89wX5cn+dnhrl/RWrtxuk8HYONwzhwAsNhypmU+kOSa1tr5SZ6a5OqqOj/JjiR3tNbOS3LHcDlJnpnkvOHrqiRvTJIhDF6b5ClJLkhybVWdMcXnAgAAsGEdN9y11u49PPLWWvvnJB9OclaSS5McHnm7Mcmzh+1Lk7y5TdyZ5PSqOjPJM5Lc3lo70Fq7P8ntSS6e6rMBAADYoKq1tvydq7YkeWeSxyf5ZGvt9KG9ktzfWju9qt6eZGdr7c+G6+5I8pIk25M8orX2iqH955J8obX26iMe46pMRvyyefPmJ+/atWs1z48TdOjQoZx66qnz7gZHUJdxWm5d9uw/OIPeLG3rWac9qG2e/ZmFzScnn/7CvHvBkdRlnNRlnNRltpb6X7mUWb8eu/DCC+9qrW1bzr7LXlClqk5N8odJfqK19tlJnptorbWqWn5KPIbW2nVJrkuSbdu2te3bt0/jblmhhYWF+NmPj7qM03LrcsUcz5Hbd9n2B7XNsz+zcM3WB/KaPdYNGxt1GSd1GSd1ma2l/lcuZcyvx5b1UQhV9dBMgt3vtNbeOjR/ephumeH7fUP7/iTnLLr52UPb0doBAABYpeOGu2HK5fVJPtxae+2iq25JcvmwfXmSmxe1v6AmnprkYGvt3iS3Jbmoqs4YFlK5aGgDAABglZYzzvs9SZ6fZE9VvW9o+5kkO5PcVFVXJvlEkucO192ayccg7M3koxBemCSttQNV9fIk7xr2e1lr7cBUngUAx+RjEwCgf8cNd8PCKHWUq5++xP4tydVHua8bktywkg4CAABwfMs65w4AAIBxE+4AAAA6YG1VgFXas//ggz5WYN/OS+bUGwBgozJyBwAA0AHhDgAAoAOmZQKjsdRy/aY3AgAsj3AHsAZ8rhwAMGumZQIAAHRAuAMAAOiAcAcAANAB4Q4AAKADwh0AAEAHhDsAAIAOCHcAAAAdEO4AAAA6INwBAAB0QLgDAADogHAHAADQAeEOAACgA8IdAABAB4Q7AACADgh3AAAAHRDuAAAAOiDcAQAAdGDTvDsAbExbduyedxcAALpi5A4AAKADwh0AAEAHhDsAAIAOCHcAAAAdEO4AAAA6INwBAAB0QLgDAADogHAHAADQAeEOAACgA8IdAABAB4Q7AACADmyadweA/m3ZsXveXQAA6J6ROwAAgA4IdwAAAB0Q7gAAADog3AEAAHRAuAMAAOiAcAcAANAB4Q4AAKADwh0AAEAHhDsAAIAOCHcAAAAd2DTvDgAcy5Ydux/Utm/nJXPoCQDAuBm5AwAA6IBwBwAA0AHTMlnSnv0Hc8UR0+FMhWM5lppGCQDA2jNyBwAA0AEjd7BB9LQwSU/PBQBgWoQ76NByp0YKSQAA/RDugC4IqgDARifcwRoRNuZPDQCAjcSCKgAAAB0wcse6dLRzyozKAACwUQl3G9zRQtI1W2fcEQAAYFWEO6bOeU798cHkAADjJ9yxKtN+0S9EAADAiRHuWLbVBC+jeQAAsLaEuzU2plAztlGxsfUHAADWM+EOZmglYX+pfd908SlT79NyHhcAgPET7ujKcsPTtEdUpz1lFQAAVkq4mxIv0AEAgHkS7uhe78G79+cHAMDyCHdwDIJTf8a0yBEAwDQJdzAQ5AAAWM8eMu8OAAAAsHrCHQAAQAeEOwAAgA4IdwAAAB2woEqnLA4Cq2dlTQBgPRHu5mC5wcuLSI60Z//BXCG4T91K3gxZat9rtk6zNwAAJ8a0TAAAgA4IdwAAAB0Q7gAAADog3AEAAHRAuAMAAOiA1TI74GMPAAAAI3cAAAAdOO7IXVXdkOQHktzXWnv80PbIJL+fZEuSfUme21q7v6oqyeuTPCvJ55Nc0Vp7z3Cby5P87HC3r2it3Tjdp7IxGKUDAACWspyRuzclufiIth1J7mitnZfkjuFykjwzyXnD11VJ3ph8NQxem+QpSS5Icm1VnbHazgMAADBx3HDXWntnkgNHNF+a5PDI241Jnr2o/c1t4s4kp1fVmUmekeT21tqB1tr9SW7PgwMjAAAAJ6haa8ffqWpLkrcvmpb5T62104ftSnJ/a+30qnp7kp2ttT8brrsjyUuSbE/yiNbaK4b2n0vyhdbaq5d4rKsyGfXL5s2bn7xr167VPseZ2LP/4Ly7MFWbT04+/YV594Ijqcs4qcs4qcs4qcs4qcs4qctsbT3rtGXtd+jQoZx66qlr3JuvufDCC+9qrW1bzr6rXi2ztdaq6vgJcfn3d12S65Jk27Ztbfv27dO66zV1RWfnwl2z9YG8Zo/FVMdGXcZJXcZJXcZJXcZJXcZJXWZr32Xbl7XfwsJCxppRTnS1zE8P0y0zfL9vaN+f5JxF+509tB2tHQAAgCk40XB3S5LLh+3Lk9y8qP0FNfHUJAdba/cmuS3JRVV1xrCQykVDGwAAAFOwnI9C+L1Mzpl7VFXdk8mqlzuT3FRVVyb5RJLnDrvfmsnHIOzN5KMQXpgkrbUDVfXyJO8a9ntZa+3IRVoAAAA4QccNd621HznKVU9fYt+W5Oqj3M8NSW5YUe8AAABYlhOdlgkAAMCICHcAAAAdEO4AAAA6INwBAAB0QLgDAADogHAHAADQAeEOAACgA8IdAABAB4Q7AACADgh3AAAAHRDuAAAAOiDcAQAAdEC4AwAA6IBwBwAA0AHhDgAAoAPCHQAAQAeEOwAAgA4IdwAAAB0Q7gAAADog3AEAAHRAuAMAAOiAcAcAANAB4Q4AAKADwh0AAEAHhDsAAIAOCHcAAAAdEO4AAAA6INwBAAB0QLgDAADogHAHAADQAeEOAACgA8IdAABAB4Q7AACADgh3AAAAHRDuAAAAOiDcAQAAdEC4AwAA6IBwBwAA0AHhDgAAoAPCHQAAQAeEOwAAgA4IdwAAAB0Q7gAAADog3AEAAHRAuAMAAOiAcAcAANAB4Q4AAKADwh0AAEAHhDsAAIAOCHcAAAAdEO4AAAA6INwBAAB0QLgDAADogHAHAADQAeEOAACgA8IdAABAB4Q7AACADgh3AAAAHRDuAAAAOiDcAQAAdEC4AwAA6IBwBwAA0AHhDgAAoAPCHQAAQAeEOwAAgA4IdwAAAB3YNO8OrEdbduyedxcAAAC+jpE7AACADgh3AAAAHRDuAAAAOiDcAQAAdEC4AwAA6IBwBwAA0AHhDgAAoAPCHQAAQAeEOwAAgA4IdwAAAB0Q7gAAADog3AEAAHRAuAMAAOiAcAcAANCBmYe7qrq4qj5SVXurasesHx8AAKBHMw13VXVSkl9N8swk5yf5kao6f5Z9AAAA6NGsR+4uSLK3tfax1tqXk+xKcumM+wAAANCdWYe7s5J8atHle4Y2AAAAVqFaa7N7sKrnJLm4tfafhsvPT/KU1tqLFu1zVZKrhovfnuQjM+sgiz0qyT/OuxM8iLqMk7qMk7qMk7qMk7qMk7qM06zr8q9aa49ezo6b1ronR9if5JxFl88e2r6qtXZdkutm2SkerKre3VrbNu9+8PXUZZzUZZzUZZzUZZzUZZzUZZzGXJdZT8t8V5LzqurcqnpYkucluWXGfQAAAOjOTEfuWmsPVNWLktyW5KQkN7TW7p5lHwAAAHo062mZaa3dmuTWWT8uK2Zq7DipyzipyzipyzipyzipyzipyziNti4zXVAFAACAtTHrc+4AAABYA8LdBlFVN1TVfVX1wUVtT6yqO6vqfVX17qq6YGg/rar+T1W9v6rurqoXLrrN5VX10eHr8nk8l54cpS5PqKq/rKo9Qx2+adF1L62qvVX1kap6xqL2i4e2vVW1Y9bPozcrqUtVfX9V3TW031VVT1t0mycP7Xur6leqqubxfHqx0uNluP5bq+pQVf3UojbHy5SdwN+y7xquu3u4/hFDu2Nmilb4t+yhVXXj0P7hqnrpots4Zqaoqs6pqj+tqg8Nx8CLh/ZHVtXtw2us26vqjKG9huNhb1V9oKqetOi+vC6bkhOoy2VDPfZU1V9U1RMW3df8jpnWmq8N8JXk3yd5UpIPLmr74yTPHLaflWRh2P6ZJK8ath+d5ECShyV5ZJKPDd/PGLbPmPdzW89fR6nLu5J837D9Y0lePmyfn+T9SR6e5Nwkf5fJwkQnDdvfNtTp/UnOn/dzW89fK6zLdyd5zLD9+CT7F93mr5M8NUklecfh483X2tdl0fVvSfIHSX5quOx4mXNtMjnf/wNJnjBc/uYkJw3bjpn51eVHk+watr8hyb4kWxwza1KXM5M8adj+xiR/O/yP/8UkO4b2Hfnaa7FnDcdDDcfHXw3tXpfNty7/9vDPO8kzF9VlrseMkbsNorX2zkxC2tc1Jzn8TuppSf5+Ufs3Du+Ynjrc7oEkz0hye2vtQGvt/iS3J7l4rfves6PU5XFJ3jls357kB4ftSzP5x/ul1trHk+xNcsHwtbe19rHW2peT7Br25QStpC6ttfe21g4fO3cnObmqHl5VZyb5ptbanW3y1/7NSZ699r3v1wqPl1TVs5N8PJO6HOZ4WQMrrM1FST7QWnv/cNvPtNa+4piZvhXWpSU5pao2JTk5yZeTfDaOmalrrd3bWnvPsP3PST6c5KxMfq43DrvdmK/9/l+a5M1t4s4kpw/Hi9eFQ3RNAAAEb0lEQVRlU7TSurTW/mL4uSfJnZl8fncy52NGuNvYfiLJL1XVp5K8OsnhKRhvSPKvMwl7e5K8uLX2L5n8gn9q0e3vGdqYrrvztT8CP5TknGH7aD9/dZmNo9VlsR9M8p7W2pcyqcE9i65Tl7WxZF2q6tQkL0nyP47Y3/EyO0c7Zh6XpFXVbVX1nqr66aHdMTMbR6vLW5J8Lsm9ST6Z5NWttQNxzKypqtqSyQyQv0qyubV273DVPyTZPGz7/z9jy6zLYldmMrqazLkuwt3G9l+T/GRr7ZwkP5nk+qH9GUnel+QxSZ6Y5A1HnsfCmvqxJD9eVXdlMi3gy3PuDxPHrEtVfWeSVyX5L3Po20Z2tLr8QpLXtdYOzatjHLU2m5J8b5LLhu//saqePp8ubkhHq8sFSb6Syf/+c5NcU1XfNp8ubgzDm1B/mOQnWmufXXzdMHptSfs5WGldqurCTMLdS2bWyWOY+efcMSqXJ3nxsP0HSX5j2H5hkp3DL/Deqvp4ku9Isj/J9kW3PzvJwkx6uoG01v4mk2lLqarHJblkuGp/vn606OyhLcdoZ0qOUZdU1dlJ3pbkBa21vxua9+drUzQSdVkTx6jLU5I8p6p+McnpSf6lqr6Y5K44XmbiGLW5J8k7W2v/OFx3aybnhf12HDNr7hh1+dEkf9Ra+39J7quqP0+yLZMRCMfMlFXVQzMJEL/TWnvr0PzpqjqztXbvMO3yvqH9aP//vS6bshXWJVX1XZm8fn5ma+0zQ/OxXq+tOSN3G9vfJ/m+YftpST46bH8yydOTpKo2J/n2TE7SvS3JRVV1xrBS0EVDG1NUVd8yfH9Ikp9N8uvDVbcked5wPte5Sc7LZPGBdyU5r6rOraqHJXnesC9TdLS6VNXpSXZncrL1nx/ef5jC8dmqeupw/uoLktw884537mh1aa39u9baltbaliS/nOR/ttbeEMfLzBzjb9ltSbZW1TcM53d9X5IPOWZm4xh1+WQmrwVSVadksnDH38QxM3XD7/f1ST7cWnvtoqtuyeSN9wzfb17U/oJh1cynJjk4HC9el03RSutSVd+a5K1Jnt9a+9tF+8/1mDFyt0FU1e9l8u7Oo6rqniTXJvnPSV4//HP9YpKrht1fnuRNVbUnk5WZXrLoHdaXZ/JLmyQvG+bjc4KOUpdTq+rqYZe3JvnNJGmt3V1VNyX5UCYL3FzdWvvKcD8vyuQP+klJbmit3R1O2ErqkuRFSR6b5Oer6ueHtotaa/cl+fEkb8pkcYJ35Gvz8TkBK6zLklprDzhepm+Ff8vur6rXZvK/pCW5tbW2e9jPMTNFKzxmfjXJb1bV3Zn87//N1toHhvtxzEzX9yR5fpI9VfW+oe1nkuxMclNVXZnkE0meO1x3ayYrZu5N8vlMZliltXbA67KpWmldfj6T1X5/bZIL80Brbdu8/8/UsGQnAAAA65hpmQAAAB0Q7gAAADog3AEAAHRAuAMAAOiAcAcAANAB4Q4AAKADwh0AAEAHhDsAAIAO/H+DBPL3iOYO5QAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 1080x720 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "article_df['PubYear'].hist(bins = (article_df['PubYear'].max() - article_df['PubYear'].min()), figsize = (15, 10))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = list(article_df[['ArticleID','PubYear']].groupby('PubYear').count()['ArticleID'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "w = np.fft.fft(x)\n",
    "freqs = np.fft.fftfreq(len(w))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(395503+0j) * exp(2 pi i t * 0.0)\n",
      "(-47741.08173227605+108296.56565066085j) * exp(2 pi i t * 0.006666666666666667)\n",
      "(-2916.647505104697-4446.201373113057j) * exp(2 pi i t * 0.013333333333333334)\n",
      "(10539.743735805234+19397.577300762736j) * exp(2 pi i t * 0.02)\n",
      "(-25020.935120043745+20768.783330945193j) * exp(2 pi i t * 0.02666666666666667)\n",
      "(12143.577118135774-11115.604172927051j) * exp(2 pi i t * 0.03333333333333333)\n",
      "(-7739.549818821184+29193.817441610612j) * exp(2 pi i t * 0.04)\n",
      "(-11897.168711957263+2992.0775355719984j) * exp(2 pi i t * 0.04666666666666667)\n",
      "(-2124.4129891429643+1883.9264314710495j) * exp(2 pi i t * 0.05333333333333334)\n",
      "(-2058.8926500048374+8760.953164608516j) * exp(2 pi i t * 0.060000000000000005)\n",
      "(-12443.556051734498-732.4449024497588j) * exp(2 pi i t * 0.06666666666666667)\n",
      "(7775.474178371135+8722.553310698155j) * exp(2 pi i t * 0.07333333333333333)\n",
      "(-10919.90605381349-3971.8651346329198j) * exp(2 pi i t * 0.08)\n",
      "(-2641.354099406623+6462.622597416276j) * exp(2 pi i t * 0.08666666666666667)\n",
      "(-527.0369496018116-2573.6465704460134j) * exp(2 pi i t * 0.09333333333333334)\n",
      "(194.09999338930152+2907.7059372984213j) * exp(2 pi i t * 0.1)\n",
      "(-6849.589185707154+3686.634866559948j) * exp(2 pi i t * 0.10666666666666667)\n",
      "(-1360.8904498737645+272.78985259690035j) * exp(2 pi i t * 0.11333333333333334)\n",
      "(-2303.8750480253802+1354.8487619933553j) * exp(2 pi i t * 0.12000000000000001)\n",
      "(-6811.388270367468+2068.8530769069152j) * exp(2 pi i t * 0.12666666666666668)\n",
      "(-5665.529243176205-2296.7989073599233j) * exp(2 pi i t * 0.13333333333333333)\n",
      "(-2555.1210834681096+2883.130051773979j) * exp(2 pi i t * 0.14)\n",
      "(-5980.845285328575+24.78815456550251j) * exp(2 pi i t * 0.14666666666666667)\n",
      "(-4005.5058075535458-1467.9736598095678j) * exp(2 pi i t * 0.15333333333333335)\n",
      "(-3130.021315728716-1900.5634002267052j) * exp(2 pi i t * 0.16)\n",
      "(-5993.000000000002+4177.706547856132j) * exp(2 pi i t * 0.16666666666666669)\n",
      "(-3553.587202821709-2434.92390162413j) * exp(2 pi i t * 0.17333333333333334)\n",
      "(-4668.897384773285-3434.064807311437j) * exp(2 pi i t * 0.18000000000000002)\n",
      "(-2507.426531640384+1690.9022297090512j) * exp(2 pi i t * 0.18666666666666668)\n",
      "(-4866.594597561747-3606.6489920911918j) * exp(2 pi i t * 0.19333333333333336)\n",
      "(-850.5535821593367-1871.0509499891796j) * exp(2 pi i t * 0.2)\n",
      "(-3461.529726274238+331.72586471501586j) * exp(2 pi i t * 0.2066666666666667)\n",
      "(-3374.491842485894-845.8046413647328j) * exp(2 pi i t * 0.21333333333333335)\n",
      "(-2071.0087936710506-3891.2642135502656j) * exp(2 pi i t * 0.22)\n",
      "(-2156.1493597257277+171.6062159068349j) * exp(2 pi i t * 0.22666666666666668)\n",
      "(-4885.554896889578+327.02184235713503j) * exp(2 pi i t * 0.23333333333333334)\n",
      "(-3331.6542050048765-4363.418177125062j) * exp(2 pi i t * 0.24000000000000002)\n",
      "(-2226.9362938502004-1381.709148524596j) * exp(2 pi i t * 0.24666666666666667)\n",
      "(-4815.703799447792-1511.77451373902j) * exp(2 pi i t * 0.25333333333333335)\n",
      "(-2217.3486715402673-4633.1333811069835j) * exp(2 pi i t * 0.26)\n",
      "(-441.1416160431836-1871.3625445776709j) * exp(2 pi i t * 0.26666666666666666)\n",
      "(-1574.8232117895773-3187.6777810856993j) * exp(2 pi i t * 0.2733333333333334)\n",
      "(-2096.0275559708807-3211.6261075371913j) * exp(2 pi i t * 0.28)\n",
      "(1722.7031445648163-1496.6480898074533j) * exp(2 pi i t * 0.2866666666666667)\n",
      "(-2927.0131214866533-516.4185345251922j) * exp(2 pi i t * 0.29333333333333333)\n",
      "(-1318.599993389303-3229.6783790649324j) * exp(2 pi i t * 0.30000000000000004)\n",
      "(-608.7916668942169-2258.4290875965435j) * exp(2 pi i t * 0.3066666666666667)\n",
      "(-43.3469471793378-227.32130318274994j) * exp(2 pi i t * 0.31333333333333335)\n",
      "(-3128.1534332994843-2892.994032775878j) * exp(2 pi i t * 0.32)\n",
      "(-1611.9620435165584-665.7788290648109j) * exp(2 pi i t * 0.32666666666666666)\n",
      "(-1004-2405.818571713171j) * exp(2 pi i t * 0.33333333333333337)\n",
      "(-1821.1498109392824-4401.051652321263j) * exp(2 pi i t * 0.34)\n",
      "(-215.66156740081078-990.2632335195303j) * exp(2 pi i t * 0.3466666666666667)\n",
      "(-1014.1308439546638-3253.6601618822942j) * exp(2 pi i t * 0.35333333333333333)\n",
      "(-337.4595881468067-2722.5702559703022j) * exp(2 pi i t * 0.36000000000000004)\n",
      "(1665.634998464012-2753.5817881222947j) * exp(2 pi i t * 0.3666666666666667)\n",
      "(-128.91367074552727-1527.7695351832463j) * exp(2 pi i t * 0.37333333333333335)\n",
      "(1756.71441668633-2732.920440753898j) * exp(2 pi i t * 0.38)\n",
      "(-530.3445017465281-663.8265964491811j) * exp(2 pi i t * 0.3866666666666667)\n",
      "(187.71897755678947-1408.320622370862j) * exp(2 pi i t * 0.39333333333333337)\n",
      "(-285.9464178406415-1102.4856427838815j) * exp(2 pi i t * 0.4)\n",
      "(-296.0397035642808-1764.4788175011568j) * exp(2 pi i t * 0.4066666666666667)\n",
      "(1966.48877722399-964.6587410026473j) * exp(2 pi i t * 0.4133333333333334)\n",
      "(-3269.026148969573-2337.260170732659j) * exp(2 pi i t * 0.42000000000000004)\n",
      "(2596.5619104921498-1206.8735864009693j) * exp(2 pi i t * 0.4266666666666667)\n",
      "(726.8427802897929-2436.1126147100217j) * exp(2 pi i t * 0.43333333333333335)\n",
      "(906.4121854045527-2239.756598098442j) * exp(2 pi i t * 0.44)\n",
      "(2761.6119356684476-729.1028820978051j) * exp(2 pi i t * 0.4466666666666667)\n",
      "(2738.973103104046-241.0019738084177j) * exp(2 pi i t * 0.45333333333333337)\n",
      "(1017.4863908748388-825.2477009024828j) * exp(2 pi i t * 0.46)\n",
      "(2166.72691095389+105.2246205857341j) * exp(2 pi i t * 0.4666666666666667)\n",
      "(2026.2798768487576+278.31554476270486j) * exp(2 pi i t * 0.4733333333333334)\n",
      "(1377.7348334062744+508.539131106382j) * exp(2 pi i t * 0.48000000000000004)\n",
      "(-416.03567388463455-247.2568721075613j) * exp(2 pi i t * 0.4866666666666667)\n",
      "(3135.5265085040073+687.2266777275122j) * exp(2 pi i t * 0.49333333333333335)\n",
      "(-2531+2.2737367544323206e-13j) * exp(2 pi i t * -0.5)\n",
      "(3135.526508504006-687.2266777275104j) * exp(2 pi i t * -0.49333333333333335)\n",
      "(-416.03567388463387+247.25687210756564j) * exp(2 pi i t * -0.4866666666666667)\n",
      "(1377.7348334062744-508.53913110638155j) * exp(2 pi i t * -0.48000000000000004)\n",
      "(2026.2798768487564-278.31554476271003j) * exp(2 pi i t * -0.4733333333333334)\n",
      "(2166.72691095389-105.22462058573285j) * exp(2 pi i t * -0.4666666666666667)\n",
      "(1017.4863908748381+825.247700902484j) * exp(2 pi i t * -0.46)\n",
      "(2738.973103104045+241.00197380841712j) * exp(2 pi i t * -0.45333333333333337)\n",
      "(2761.611935668453+729.102882097805j) * exp(2 pi i t * -0.4466666666666667)\n",
      "(906.4121854045532+2239.756598098441j) * exp(2 pi i t * -0.44)\n",
      "(726.8427802897982+2436.1126147100204j) * exp(2 pi i t * -0.43333333333333335)\n",
      "(2596.561910492149+1206.873586400968j) * exp(2 pi i t * -0.4266666666666667)\n",
      "(-3269.026148969574+2337.260170732661j) * exp(2 pi i t * -0.42000000000000004)\n",
      "(1966.4887772239902+964.658741002648j) * exp(2 pi i t * -0.4133333333333334)\n",
      "(-296.0397035642791+1764.4788175011618j) * exp(2 pi i t * -0.4066666666666667)\n",
      "(-285.9464178406415+1102.4856427838815j) * exp(2 pi i t * -0.4)\n",
      "(187.71897755679575+1408.3206223708662j) * exp(2 pi i t * -0.39333333333333337)\n",
      "(-530.3445017465292+663.8265964491818j) * exp(2 pi i t * -0.3866666666666667)\n",
      "(1756.7144166863293+2732.9204407538987j) * exp(2 pi i t * -0.38)\n",
      "(-128.91367074552454+1527.7695351832442j) * exp(2 pi i t * -0.37333333333333335)\n",
      "(1665.6349984640058+2753.5817881222933j) * exp(2 pi i t * -0.3666666666666667)\n",
      "(-337.4595881468045+2722.570255970306j) * exp(2 pi i t * -0.36000000000000004)\n",
      "(-1014.1308439546633+3253.6601618822933j) * exp(2 pi i t * -0.35333333333333333)\n",
      "(-215.66156740081215+990.2632335195306j) * exp(2 pi i t * -0.3466666666666667)\n",
      "(-1821.1498109392778+4401.051652321263j) * exp(2 pi i t * -0.34)\n",
      "(-1003.9999999999998+2405.818571713171j) * exp(2 pi i t * -0.33333333333333337)\n",
      "(-1611.96204351656+665.7788290648061j) * exp(2 pi i t * -0.32666666666666666)\n",
      "(-3128.1534332994843+2892.9940327758786j) * exp(2 pi i t * -0.32)\n",
      "(-43.346947179333824+227.32130318275196j) * exp(2 pi i t * -0.31333333333333335)\n",
      "(-608.7916668942164+2258.4290875965435j) * exp(2 pi i t * -0.3066666666666667)\n",
      "(-1318.5999933893038+3229.678379064936j) * exp(2 pi i t * -0.30000000000000004)\n",
      "(-2927.0131214866547+516.4185345251913j) * exp(2 pi i t * -0.29333333333333333)\n",
      "(1722.70314456482+1496.6480898074517j) * exp(2 pi i t * -0.2866666666666667)\n",
      "(-2096.0275559708807+3211.6261075371913j) * exp(2 pi i t * -0.28)\n",
      "(-1574.8232117895732+3187.6777810856975j) * exp(2 pi i t * -0.2733333333333334)\n",
      "(-441.1416160431835+1871.362544577673j) * exp(2 pi i t * -0.26666666666666666)\n",
      "(-2217.348671540269+4633.1333811069835j) * exp(2 pi i t * -0.26)\n",
      "(-4815.703799447792+1511.7745137390193j) * exp(2 pi i t * -0.25333333333333335)\n",
      "(-2226.936293850197+1381.709148524601j) * exp(2 pi i t * -0.24666666666666667)\n",
      "(-3331.6542050048783+4363.418177125063j) * exp(2 pi i t * -0.24000000000000002)\n",
      "(-4885.554896889576-327.0218423571307j) * exp(2 pi i t * -0.23333333333333334)\n",
      "(-2156.149359725726-171.60621590683309j) * exp(2 pi i t * -0.22666666666666668)\n",
      "(-2071.0087936710497+3891.2642135502647j) * exp(2 pi i t * -0.22)\n",
      "(-3374.4918424858924+845.8046413647323j) * exp(2 pi i t * -0.21333333333333335)\n",
      "(-3461.529726274247-331.7258647150102j) * exp(2 pi i t * -0.2066666666666667)\n",
      "(-850.5535821593367+1871.0509499891796j) * exp(2 pi i t * -0.2)\n",
      "(-4866.594597561755+3606.6489920911845j) * exp(2 pi i t * -0.19333333333333336)\n",
      "(-2507.426531640385-1690.9022297090503j) * exp(2 pi i t * -0.18666666666666668)\n",
      "(-4668.897384773286+3434.064807311438j) * exp(2 pi i t * -0.18000000000000002)\n",
      "(-3553.58720282171+2434.92390162413j) * exp(2 pi i t * -0.17333333333333334)\n",
      "(-5992.999999999998-4177.706547856134j) * exp(2 pi i t * -0.16666666666666669)\n",
      "(-3130.021315728716+1900.5634002267066j) * exp(2 pi i t * -0.16)\n",
      "(-4005.5058075535394+1467.9736598095687j) * exp(2 pi i t * -0.15333333333333335)\n",
      "(-5980.845285328576-24.788154565501827j) * exp(2 pi i t * -0.14666666666666667)\n",
      "(-2555.121083468109-2883.1300517739814j) * exp(2 pi i t * -0.14)\n",
      "(-5665.529243176207+2296.7989073599224j) * exp(2 pi i t * -0.13333333333333333)\n",
      "(-6811.388270367468-2068.8530769069166j) * exp(2 pi i t * -0.12666666666666668)\n",
      "(-2303.8750480253802-1354.8487619933558j) * exp(2 pi i t * -0.12000000000000001)\n",
      "(-1360.8904498737647-272.78985259690046j) * exp(2 pi i t * -0.11333333333333334)\n",
      "(-6849.589185707154-3686.6348665599476j) * exp(2 pi i t * -0.10666666666666667)\n",
      "(194.09999338930254-2907.705937298425j) * exp(2 pi i t * -0.1)\n",
      "(-527.0369496018122+2573.6465704460115j) * exp(2 pi i t * -0.09333333333333334)\n",
      "(-2641.354099406622-6462.622597416279j) * exp(2 pi i t * -0.08666666666666667)\n",
      "(-10919.90605381349+3971.8651346329193j) * exp(2 pi i t * -0.08)\n",
      "(7775.474178371134-8722.553310698151j) * exp(2 pi i t * -0.07333333333333333)\n",
      "(-12443.5560517345+732.4449024497585j) * exp(2 pi i t * -0.06666666666666667)\n",
      "(-2058.892650004832-8760.953164608516j) * exp(2 pi i t * -0.060000000000000005)\n",
      "(-2124.4129891429634-1883.9264314710504j) * exp(2 pi i t * -0.05333333333333334)\n",
      "(-11897.16871195726-2992.0775355720007j) * exp(2 pi i t * -0.04666666666666667)\n",
      "(-7739.5498188211795-29193.817441610616j) * exp(2 pi i t * -0.04)\n",
      "(12143.57711813577+11115.604172927055j) * exp(2 pi i t * -0.03333333333333333)\n",
      "(-25020.935120043745-20768.783330945196j) * exp(2 pi i t * -0.02666666666666667)\n",
      "(10539.743735805238-19397.577300762736j) * exp(2 pi i t * -0.02)\n",
      "(-2916.6475051046973+4446.2013731130555j) * exp(2 pi i t * -0.013333333333333334)\n",
      "(-47741.08173227605-108296.56565066084j) * exp(2 pi i t * -0.006666666666666667)\n"
     ]
    }
   ],
   "source": [
    "for coef,freq in zip(w,freqs):\n",
    "    if coef:\n",
    "        print('{c:>6} * exp(2 pi i t * {f})'.format(c=coef,f=freq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.collections.PathCollection at 0x7f0143f39da0>"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAY0AAAD8CAYAAACLrvgBAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvnQurowAAE0VJREFUeJzt3X+sX3V9x/Hn2xa0bmIRGgJt8aI2mDqdyB3UkCwbTltgoc0kBjalGrTZhMxFw2wnCZuaUEcik0zZiBjLtvBjjJRO0aYDjNkSftwK0hWCXBDWFpUKFDbpUOp7f3w/l317vfd+P997v9977o/nI7npOZ/zOefz+fR773l9z/d8zr2RmUiSVONVTXdAkjR7GBqSpGqGhiSpmqEhSapmaEiSqhkakqRqhoYkqZqhIUmqZmhIkqotbLoDvXbsscfmwMBA092QpFll586dP83MJZ3qzbnQGBgYYGhoqOluSNKsEhFP1tTz4ylJUjVDQ5JUzdCQJFWrDo2IWBAR90fEN8r6SRFxT0QMR8RNEXFkKX91WR8u2wfajrGplD8SEavbyteUsuGI2NhWPmYbkqRmdHOl8Qng4bb1LwBXZeZbgOeAi0r5RcBzpfyqUo+IWAmcD7wNWAN8pQTRAuDLwFnASuCCUneiNiRJDagKjYhYBpwDfLWsB3AmcEupsgVYV5bXlnXK9veU+muBGzPzpcz8ITAMnFa+hjPz8cz8OXAjsLZDG3239f59nLH5Tk7a+E3O2HwnW+/fN11NS9KMVXul8TfAnwO/LOvHAAcy8+WyvhdYWpaXAnsAyvbnS/1XykftM175RG0cJiI2RMRQRAzt37+/ckjj23r/Pjbduot9Bw6SwL4DB9l06y6DQ9K81zE0IuL3gaczc+c09GdSMvPazBzMzMElSzo+m9LRldsf4eAvDh1WdvAXh7hy+yNTPrYkzWY1D/edAZwbEWcDrwGOAr4ELI6IheVKYBkw8jZ8H7Ac2BsRC4HXA8+0lY9o32es8mcmaKOvnjpwsKtySZovOl5pZOamzFyWmQO0bmTfmZl/BNwFnFeqrQduK8vbyjpl+52ZmaX8/DK76iRgBXAvcB+wosyUOrK0sa3sM14bfXXC4kVdlUvSfDGV5zQ+DXwyIoZp3X+4rpRfBxxTyj8JbATIzN3AzcBDwLeBizPzULmKuATYTmt21s2l7kRt9NWlq09m0RELDitbdMQCLl198nQ0L0kzVrTe0M8dg4OD2YvfPbX1/n1cuf0RnjpwkBMWL+LS1Sez7pQx78NL0qwXETszc7BTvTn3Cwt7Zd0pSw0JSRrFXyMiSapmaEiSqhkakqRqhoYkqZqhIUmqZmhIkqoZGpKkaoaGJKmaoSFJqmZoSJKqGRqSpGqGhiSpmqEhSapmaEiSqhkakqRqhoYkqZqhIUmqZmhIkqoZGpKkaoaGJKmaoSFJqmZoSJKqGRqSpGqGhiSpmqEhSapmaEiSqhkakqRqhoYkqZqhIUmqZmhIkqoZGpKkaoaGJKmaoSFJqmZoSJKqdQyNiHhNRNwbEd+PiN0R8Vel/KSIuCcihiPipog4spS/uqwPl+0DbcfaVMofiYjVbeVrStlwRGxsKx+zDUlSM2quNF4CzszM3wTeCayJiFXAF4CrMvMtwHPARaX+RcBzpfyqUo+IWAmcD7wNWAN8JSIWRMQC4MvAWcBK4IJSlwnakCQ1oGNoZMv/lNUjylcCZwK3lPItwLqyvLasU7a/JyKilN+YmS9l5g+BYeC08jWcmY9n5s+BG4G1ZZ/x2pAkNaDqnka5IngAeBrYATwGHMjMl0uVvcDSsrwU2ANQtj8PHNNePmqf8cqPmaANSVIDqkIjMw9l5juBZbSuDN7a1151KSI2RMRQRAzt37+/6e5I0pzV1eypzDwA3AW8G1gcEQvLpmXAvrK8D1gOULa/HnimvXzUPuOVPzNBG6P7dW1mDmbm4JIlS7oZkiSpCzWzp5ZExOKyvAh4L/AwrfA4r1RbD9xWlreVdcr2OzMzS/n5ZXbVScAK4F7gPmBFmSl1JK2b5dvKPuO1IUlqwMLOVTge2FJmOb0KuDkzvxERDwE3RsTngfuB60r964B/iIhh4FlaIUBm7o6Im4GHgJeBizPzEEBEXAJsBxYAX8vM3eVYnx6nDUlSA6L1hn7uGBwczKGhoaa7IUmzSkTszMzBTvV8IlySVM3QkCRVMzQkSdUMDUlSNUNDklTN0JAkVTM0JEnVDA1JUjVDQ5JUzdCQJFUzNCRJ1QwNSVI1Q0OSVM3QkCRVMzQkSdUMDUlSNUNDklTN0JAkVTM0JEnVDA1JUjVDQ5JUzdCQJFUzNCRJ1QwNSVI1Q0OSVM3QkCRVMzQkSdUMDUlSNUNDklTN0JAkVTM0JEnVDA1JUjVDQ5JUzdCQJFUzNCRJ1QwNSVK1jqEREcsj4q6IeCgidkfEJ0r5GyJiR0Q8Wv49upRHRFwdEcMR8WBEvKvtWOtL/UcjYn1b+akRsavsc3VExERtSJKaUXOl8TLwqcxcCawCLo6IlcBG4I7MXAHcUdYBzgJWlK8NwDXQCgDgcuB04DTg8rYQuAb4WNt+a0r5eG1IkhrQMTQy80eZ+b2y/N/Aw8BSYC2wpVTbAqwry2uB67PlbmBxRBwPrAZ2ZOazmfkcsANYU7YdlZl3Z2YC14861lhtSJIa0NU9jYgYAE4B7gGOy8wflU0/Bo4ry0uBPW277S1lE5XvHaOcCdoY3a8NETEUEUP79+/vZkiSpC5Uh0ZE/DrwL8CfZeYL7dvKFUL2uG+HmaiNzLw2Mwczc3DJkiX97IYkzWtVoRERR9AKjH/KzFtL8U/KR0uUf58u5fuA5W27LytlE5UvG6N8ojYkSQ2omT0VwHXAw5n5xbZN24CRGVDrgdvayi8ss6hWAc+Xj5i2A++LiKPLDfD3AdvLthciYlVp68JRxxqrDUlSAxZW1DkD+BCwKyIeKGV/AWwGbo6Ii4AngQ+UbbcDZwPDwIvARwAy89mI+BxwX6n32cx8tix/HPg6sAj4VvligjYkSQ2I1q2CuWNwcDCHhoaa7oYkzSoRsTMzBzvV84lwSVI1Q0OSVM3QkCRVMzQkSdUMDUlSNUNDklTN0JAkVTM0JEnVDA1JUjVDQ5JUzdCQJFUzNCRJ1QwNSVI1Q0OSVM3QkCRVMzQkSdUMDUlSNUNDklTN0JAkVTM0JEnVDA1JUjVDQ5JUzdCQJFUzNCRJ1QwNSVI1Q0OSVM3QkCRVMzQkSdUMDUlSNUNDklTN0JAkVTM0JEnVDA1JUjVDQ5JUzdCQJFXrGBoR8bWIeDoi/rOt7A0RsSMiHi3/Hl3KIyKujojhiHgwIt7Vts/6Uv/RiFjfVn5qROwq+1wdETFRG5Kk5tRcaXwdWDOqbCNwR2auAO4o6wBnASvK1wbgGmgFAHA5cDpwGnB5WwhcA3ysbb81HdqQJDWkY2hk5neBZ0cVrwW2lOUtwLq28uuz5W5gcUQcD6wGdmTms5n5HLADWFO2HZWZd2dmAtePOtZYbUiSGjLZexrHZeaPyvKPgePK8lJgT1u9vaVsovK9Y5RP1IYkqSFTvhFerhCyB32ZdBsRsSEihiJiaP/+/f3siiTNa5MNjZ+Uj5Yo/z5dyvcBy9vqLStlE5UvG6N8ojZ+RWZem5mDmTm4ZMmSSQ5JktTJZENjGzAyA2o9cFtb+YVlFtUq4PnyEdN24H0RcXS5Af4+YHvZ9kJErCqzpi4cdayx2pAkNWRhpwoRcQPwO8CxEbGX1iyozcDNEXER8CTwgVL9duBsYBh4EfgIQGY+GxGfA+4r9T6bmSM31z9Oa4bWIuBb5YsJ2pAkNSRatwvmjsHBwRwaGmq6G5I0q0TEzswc7FTPJ8IlSdUMDUlSNUNDklTN0JAkVTM0JEnVDA1JUjVDQ5JUzdCQJFUzNCRJ1QwNSVI1Q0OSVM3QkCRVMzQkSdUMDUlSNUNDklTN0JAkVTM0JEnVDA1JUjVDQ5JUzdCQJFUzNCRJ1QwNSVI1Q0OSVM3QkCRVMzQkSdUMDUlSNUNDklTN0JAkVTM0JEnVDA1JUjVDQ5JUbWHTHZgvLtu6ixvu2cOhTBZEcMHpy/n8urc33S1J6oqh0SMThcJlW3fxj3f/1yt1D2W+sm5wSJpNDI0e6BQKN9yzZ8z9brhnT09Cw6sYSdPFexo9MFEoQCtExjJeeTdGAmvkWCOBddnWXVM+tiSN5pVGD3QKhQURY9ZZEDHltvt9FSPNdd1eqY+u/6Ylr+Xx/S/25Up/Jn6KMONDIyLWAF8CFgBfzczNDXfpV3QKhQtOX37Yx1cjLjh9+ZTb7udVjGa2mXhCqdWp770c22TuN46UjQ6FANp/sg5l8ujTP5vU/p0CZ6beC42cwSeXiFgA/AB4L7AXuA+4IDMfGm+fwcHBHBoamqYetox+cUd8cNWJh30D9OMH/M2bbh83sB674uxx+9vNO6XZfHIabSaPpZvXZTLfcxOdvDq95qP7MpX10Sfe0brZPtm2Rv6fxvv5mckm+tmeiojYmZmDHevN8NB4N/CXmbm6rG8CyMwrxttnukKj35eotcfv9uTR6Qdy9P7jHX/EZC7nR5/8ehlg3byrHGss3Z4MezWW2tel1lSO1+u+qPcCOGHxIi5dfTLrTlnam2POkdA4D1iTmR8t6x8CTs/MS8bbZzpCo+ZE3Y/jj9feVE6U4xl5N1P7TmwyIdXtyWmq71D7qddjkWosOmIBV/zB23sSHPMqNCJiA7AB4MQTTzz1ySef7Gu/JvOR0GgTnehrT9Q17U335bcnQ2l6LV28iP/YeOaUj1MbGjP9Rvg+oP1u8bJSdpjMvBa4FlpXGv3u1FRvPne6wVV7nPHqtQfSdDMwNJt1+6an2/s1/fDUgYPT2t5MD437gBURcRKtsDgf+MNmuzT1KbTjTZNtn3VR682bbu/Jx1FSU2bCx47tV+1TvV/ZzQSEbm7an7H5TvaNERAnLF405fF3Y0aHRma+HBGXANtpTbn9WmbubrhbU55C28srgNqnz8cz+hu3k7n08dNMHku3J5Rub/K3H6/T/0OvZ09NdYJDN7Onau/xtf/sfn7d26d0b3Iq+0/0/3Dp6pPZdOsuDv7i0Cv1Fx2xgEtXnzzpvk7GjL6nMRlNzZ7qZrZUt/cZak7sI++UBjZ+s+OxJuprzf2afn381c+pme1qp6V2Wu/lWKY6C61m+0Rm8lTkXpvNY916/z6u3P4ITx046OypXmniOY1udfMR0hObzzlsfaJQeGLzOVO+Sd/tzLBup+RO9enbXk7Bnapej0VqkqExw9W8Wx/rRN8pFHoxHdiToTT/GBqzSDcn+iafPpc0dxkaDevnu3VDQVKvGRoN6vcT45LUa7Wh4d/T6INOf19DkmYrQ6MP/HXlkuYqQ6MPxnsyvBd/dEmSmmRo9MF4T4b34o8uSVKTZvSvEZmtRm52O8NJ0lzj7ClJkrOnJEm9Z2hIkqp5T6OST2FLkqFRpdNf2pOk+cKPpyr4hLcktRgaFXzCW5JaDI0KPuEtSS2GRgWf8JakFm+EV/AJb0lq8YlwSZJPhEuSes/QkCRVMzQkSdUMDUlSNUNDklRtzs2eioj9wJM9POSxwE97eLyZzLHOTfNlrPNlnNCfsb4xM5d0qjTnQqPXImKoZhraXOBY56b5Mtb5Mk5odqx+PCVJqmZoSJKqGRqdXdt0B6aRY52b5stY58s4ocGxek9DklTNKw1JUjVDo4iINRHxSEQMR8TGMba/OiJuKtvviYiB6e9lb1SM9bcj4nsR8XJEnNdEH3ulYqyfjIiHIuLBiLgjIt7YRD+nqmKcfxwRuyLigYj494hY2UQ/e6HTWNvqvT8iMiJm7Yyqitf1wxGxv7yuD0TER/veqcyc91/AAuAx4E3AkcD3gZWj6nwc+LuyfD5wU9P97uNYB4B3ANcD5zXd5z6P9XeB15blP5mNr2vlOI9qWz4X+HbT/e7XWEu91wHfBe4GBpvudx9f1w8Dfzud/fJKo+U0YDgzH8/MnwM3AmtH1VkLbCnLtwDviZiVf7qv41gz84nMfBD4ZRMd7KGasd6VmS+W1buBZdPcx16oGecLbau/BszWm5k1P6sAnwO+APzvdHaux2rHOq0MjZalwJ629b2lbMw6mfky8DxwzLT0rrdqxjpXdDvWi4Bv9bVH/VE1zoi4OCIeA/4a+NNp6luvdRxrRLwLWJ6Z35zOjvVB7ffv+8vHq7dERN//nKihIQER8UFgELiy6b70S2Z+OTPfDHwauKzp/vRDRLwK+CLwqab7Mk3+FRjIzHcAO/j/T0P6xtBo2Qe0J/SyUjZmnYhYCLweeGZaetdbNWOdK6rGGhG/B3wGODczX5qmvvVSt6/pjcC6vvaofzqN9XXAbwDfiYgngFXAtll6M7zj65qZz7R9z34VOLXfnTI0Wu4DVkTESRFxJK0b3dtG1dkGrC/L5wF3ZrkTNcvUjHWu6DjWiDgF+HtagfF0A33shZpxrmhbPQd4dBr710sTjjUzn8/MYzNzIDMHaN2nOjczZ+PfgK55XY9vWz0XeLjvvWp6hsBM+QLOBn5Aa7bCZ0rZZ2l9wwG8BvhnYBi4F3hT033u41h/i9bnpz+jdTW1u+k+93Gs/wb8BHigfG1rus99GueXgN1ljHcBb2u6z/0a66i632GWzp6qfF2vKK/r98vr+tZ+98knwiVJ1fx4SpJUzdCQJFUzNCRJ1QwNSVI1Q0OSVM3QkCRVMzQkSdUMDUlStf8DjYfq6LzLyV8AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.scatter(abs(freqs), w)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<matplotlib.lines.Line2D at 0x7f0143bf9828>]"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX0AAAD8CAYAAACb4nSYAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvnQurowAAIABJREFUeJzt3Xl8VNXdx/HPLxshYcvGGgiL7CpbZBMEd6QKyuMCakXBorYq1i5q26f62Fq7WNe6FhFtrYhUK1XUugGiKARkkT3shC3skEAgyXn+yMWOMSGDTHJnMt/36zWvzNx7buaXy/CdO+eeOdecc4iISHSI8bsAERGpOQp9EZEootAXEYkiCn0RkSii0BcRiSIKfRGRKKLQFxGJIgp9EZEootAXEYkicX4XUF56erpr3bq132WIiESU+fPn73TOZVTVLuxCv3Xr1uTk5PhdhohIRDGzDcG0U/eOiEgUUeiLiEQRhb6ISBRR6IuIRBGFvohIFFHoi4hEEYW+iEgUCSr0zWyIma00s1wzu7uC9Vlm9qGZLTazGWaWGbCuxMwWerdpoSw+UGmp44G3l7FhV0F1PYWISMSrMvTNLBZ4ErgI6AKMMrMu5Zo9BLzknDsduB94MGDdIedcd+82LER1f8v6XQW8Om8T33t8Nm8uzKuupxERiWjBHOn3BnKdc2udc0eAycDwcm26AB959z+uYH21a5tRj+njB9KpaX3GT17Iz15bREFRcU2XISIS1oIJ/RbApoDHm71lgRYBI7z7lwH1zSzNe5xoZjlm9rmZXXpS1VYhMyWJyeP6cvs5pzB1wWYu+ctslm7ZV51PKSISUUJ1IvenwCAz+xIYBOQBJd66LOdcNnA18KiZtSu/sZmN894YcvLz80+qkLjYGO68oCP/uLEvBUXFXPbkZ7zw6Tqccyf1e0VEaoNgQj8PaBnwONNb9jXn3Bbn3AjnXA/gl96yvd7PPO/nWmAG0KP8EzjnnnPOZTvnsjMyqpwkLij92qXxzvizGNg+nf/79zJ+8FIOuwuOhOR3i4hEqmBCfx7Q3szamFkCMBL4xigcM0s3s2O/6x5gorc8xczqHGsDnAksC1XxVUlNTmDC6GzuvaQLs1bt5KLHZjFnza6aenoRkbBTZeg754qBW4H3gOXAFOfcUjO738yOjcYZDKw0s1VAE+ABb3lnIMfMFlF2gvf3zrkaC30AM+OGM9vw+g/7k5wQx9UTPufP/1lJcUlpTZYhIhIWLNz6urOzs111zadfUFTMvdOWMnX+ZrKzUnhsVA9aNKpbLc8lIlKTzGy+d/70uKLqG7nJdeJ46IpuPDayOyu2HeCiR2fx7ldb/S5LRKTGRFXoHzO8ewvevn0ArdOTufnvC/jlG0s4fLSk6g1FRCJcVIY+QFZaMlNv7s9NZ7Xl5S82Muwvs1m1/YDfZYmIVKuoDX2AhLgY7hnamRfH9GZ3wREueWI2L3+xQWP6RaTWiurQP2ZQhwymjx9I7zap/PKNr/jjeyv9LklEpFoo9D2N6yfy4g29ubpPK56esYYnP871uyQRkZCL87uAcBITY/x2+KkUFhXzp/dWUq9OHKP7t/a7LBGRkFHolxMTY/zpim4UHCnh3mlLSa4Tx+W9MqveUEQkAqh7pwLxsTE8MaoHA05J5+dTFzF9icbyi0jtoNCvRGJ8LM9d14serVIYP/lLPl65w++SREROmkL/OJIS4ph4/Rl0aFKfm/82n8/XarI2EYlsCv0qNKwbz0tjepOZUpcbX8xh0aa9fpckIvKdKfSDkFavDi/f2JeU5HhGvzCXldv0zV0RiUwK/SA1bZjIy2P7UicuhmsmfMG6nQV+lyQicsIU+iegVVoSfx/bh1LnuHbCF2zZe8jvkkRETohC/wS1b1Kfl8b0Zv+ho1w74QvyDxT5XZKISNAU+t/BqS0a8sINZ7B132G+//wX7Cs86ndJIiJBUeh/R9mtU3nuul6szS9g9AtzOVhU7HdJIiJVUuifhIHtM3ji6h4sydvHD17M0YVYRCTsKfRP0oVdm/LQFafz+bpd/PDlBRzVBddFJIwp9EPgsh6Z/PbSU/loxQ5+/OpCSkp1ERYRCU+aZTNErumTRUFRMb+bvoLkhDgeHHEaMTHmd1kiIt+g0A+hcWe14+DhYh7/KJe4WOM3w09V8ItIWFHoh9iPz+/A0VLH0zPWUOrggUsV/CISPhT6IWZm/PzCjhjw1Iw1gOOBS9XVIyLhQaFfDcyMn13YkRgz/vJxLqWlqI9fRMKCQr+amBk/uaADMQaPf5RLqXP84X9OV/CLiK+CGrJpZkPMbKWZ5ZrZ3RWszzKzD81ssZnNMLPMgHWjzWy1dxsdyuLDnZlx5wUdGX9ue16bv5mf/3OxhnOKiK+qPNI3s1jgSeB8YDMwz8ymOeeWBTR7CHjJOfeimZ0DPAh838xSgXuBbMAB871t94T6DwlnPz6/A2bw6AerKXWOP13ejVgd8YuID4I50u8N5Drn1jrnjgCTgeHl2nQBPvLufxyw/kLgfefcbi/o3weGnHzZkeeO8zpw5/kdeH1BHj99bZGO+EXEF8GEfgtgU8Djzd6yQIuAEd79y4D6ZpYW5LaY2TgzyzGznPz8/GBrjzi3n9uen17QgTe+zOPOKQsp1pQNIlLDQjUNw0+BQWb2JTAIyAOCnn3MOfeccy7bOZedkZERopLC063ntOdnF3bkzYVbuHPKIgW/iNSoYEbv5AEtAx5nesu+5pzbgnekb2b1gP9xzu01szxgcLltZ5xEvbXCj84+hRgz/vDuCkqd49GruhMXq2mQRKT6BRP684D2ZtaGsrAfCVwd2MDM0oHdzrlS4B5gorfqPeB3ZpbiPb7AWx/1bhncjhiDB99ZgQMeU/CLSA2oMvSdc8VmditlAR4LTHTOLTWz+4Ec59w0yo7mHzQzB8wCfuRtu9vMfkPZGwfA/c653dXwd0Skmwa1I8aMB6YvBwePjuxOvIJfRKqRORdeo0iys7NdTk6O32XUqAmfrOW3by/nolOb8vioHgp+ETlhZjbfOZddVTulSxi4cWBb/vfiLrzz1TZu/ccCjhTr5K6IVA+FfpgYO6AN917ShfeWblfwi0i1UeiHkRvObMP/DevKf5Zt16UXRaRaKPTDzOj+rbl/eFc+WL6du/65mHA75yIikU2zbIah6/q1Zm/hUR5+fxVNGiRy15BOfpckIrWEQj9M3XbOKWzbf5inZ6yhaYNERvdv7XdJIlILKPTDlFnZNXbzDxRx37+XklG/DkNPa+Z3WSIS4dSnH8ZiY4zHR/agR8tG3PHqQr5Yu8vvkkQkwin0w1zdhFieH30GLVPq8oOXcli1/YDfJYlIBFPoR4CU5AReHNObxPhYRk+cy5a9h/wuSUQilEI/QmSmJDHpht4cOFzM9S/MZV/hUb9LEpEIpNCPIF2aN+C57/di3c4CfvC3HA4fDfqSBSIigEI/4vQ/JZ0/X9mduet28+NXF+qyiyJyQhT6EWhYt+b86nudeeerbdz/76X61q6IBE3j9CPUjQPbsm3fYSbMXkfThnW5ZXA7v0sSkQig0I9gvxjame0HivjDuyto0qAOI3pm+l2SiIQ5hX4Ei4kxHrridHYeKOLnUxeTVq8OgzrU7gvLi8jJUZ9+hKsTF8uz1/XilMb1uOXv81myeZ/fJYlIGFPo1wINEuN5cUxvUpISuGHSXDbsKvC7JBEJUwr9WqJJg0ReHNOb4lLH6Ilz2XmwyO+SRCQMKfRrkVMa1+P50dls3XeYsZPmUVBU7HdJIhJmFPq1TK+sVJ4Y1YMlefv40T8WUKxLLopIAIV+LXRB16b85tJTmbEyn/v05S0RCaAhm7XUNX2y2LirkGdnraV1WjI3Dmzrd0kiEgYU+rXYXUM6sWFXIQ9MX06r1CQu6NrU75JExGfq3qnFYmKMR67qzuktGjJ+8kKN4ReR4ELfzIaY2UozyzWzuytY38rMPjazL81ssZkN9Za3NrNDZrbQuz0T6j9Ajq9uQix/HZ1NanICY1+cpwuwiES5KkPfzGKBJ4GLgC7AKDPrUq7Zr4ApzrkewEjgqYB1a5xz3b3bzSGqW05A4/qJTLz+DAqPlDBm0jwOaiinSNQK5ki/N5DrnFvrnDsCTAaGl2vjgAbe/YbAltCVKKHQsWl9nrymJ6t3HOQ2DeUUiVrBhH4LYFPA483eskD3Adea2WZgOnBbwLo2XrfPTDMbeDLFyskZ1CGD/xvWlY9X5vPbt5f7XY6I+CBUJ3JHAZOcc5nAUOBvZhYDbAVaed0+dwL/MLMG5Tc2s3FmlmNmOfn5+SEqSSpybd8sbhzQhkmfrWfSp+v8LkdEalgwoZ8HtAx4nOktCzQWmALgnJsDJALpzrki59wub/l8YA3QofwTOOeec85lO+eyMzI0NXB1u2doZ87v0oT731rGRyu2+12OiNSgYEJ/HtDezNqYWQJlJ2qnlWuzETgXwMw6Uxb6+WaW4Z0IxszaAu2BtaEqXr6b2BjjsZHd6dq8Ibf+40uWbtFQTpFoUWXoO+eKgVuB94DllI3SWWpm95vZMK/ZT4AfmNki4BXgelf23f+zgMVmthCYCtzsnNtdHX+InJikhDgmjM6mYd14xk7KYdu+w36XJCI1wMJtXpbs7GyXk5PjdxlRY/nW/Vz+9Ge0Tk9myk39SK6jL2mLRCIzm++cy66qnb6RG+U6N2vAX67uyfKt+xk/+UtKSsPrIEBEQkuhL5zdqTH3DevKB8t38ICGcorUavosLwBc168163YWMPHTdbRJT+L7/Vr7XZKIVAOFvnztV9/rwqbdhdw7bSmZqUmc3bGx3yWJSIipe0e+VjaUswedmzXg1pcXsHzrfr9LEpEQU+jLNyTXieP50WdQPzGesZPmsWO/hnKK1CYKffmWpg0Tef76bPYeOsoNk+ax62CR3yWJSIgo9KVCXZs35KlrepK74yCXPzOHjbsK/S5JREJAoS+VGtyxMS/f2IfdBUcY8fRnfJWn6RpEIp1CX44ru3Uq/7ylH3XiYrjq2TnMWqVZUEUimUJfqnRK4/q8/sP+tEpLZsykefxz/ma/SxKR70ihL0Fp0iCRKTf1pU/bVH7y2iKempFLuM3bJCJVU+hL0OonxvPC9b0Z3r05f3x3JfdOW6q5ekQijL6RKyckIS6GR67sTtMGiTw7ay079hfx6MjuJMbH+l2aiARBR/pywmJijHuGdubXF3fhvWXb+P7zX7C38IjfZYlIEBT68p2NGdCGJ0b1YNGmfVz+zBzy9h7yuyQRqYJCX07Kxac356Wxvdm+/zAjnvpU8/WIhDmFvpy0vm3TmHpzfwzjymfm8NmanX6XJCKVUOhLSHRsWjaWv1mjRK6fOI9/L9rid0kiUgGFvoRM80Z1ee2m/nRv1YjbXvmSCZ+s9bskESlHoS8h1TApnpfG9GboaU357dvL+e1byyjVWH6RsKFx+hJyifGxPDGqJ43rL2PC7HXsOFDEQ1d0IyFOxxgiflPoS7WIjTHuvaQLTRok8od3V3Dg8FGeuqYXdRP0JS4RP+nQS6qNmXHL4HY8OOI0ZqzKZ/QLczlw+KjfZYlENYW+VLtRvVvx+MgeLNiwh1F//VxX4hLxkUJfasQl3Zrz1+uyWb39IFc+O4et+/TtXRE/KPSlxpzdqTEvjenN9v1FXP70HNbvLPC7JJGoE1Tom9kQM1tpZrlmdncF61uZ2cdm9qWZLTazoQHr7vG2W2lmF4ayeIk8fdqm8coP+nLoaAmXPzOHFds0bYNITaoy9M0sFngSuAjoAowysy7lmv0KmOKc6wGMBJ7ytu3iPe4KDAGe8n6fRLHTMhsy5aa+xMUYVz37OQs27vG7JJGoEcyRfm8g1zm31jl3BJgMDC/XxgENvPsNgWPfwR8OTHbOFTnn1gG53u+TKHdK4/q8dnM/GiXFc+2EL5i9WvP1iNSEYEK/BbAp4PFmb1mg+4BrzWwzMB247QS2xczGmVmOmeXk5+vC29GiZWoSr93cj1apSYyZNI/3lm7zuySRWi9UJ3JHAZOcc5nAUOBvZhb073bOPeecy3bOZWdkZISoJIkEjesnMnlcX7q2aMAPX16gi66LVLNggjkPaBnwONNbFmgsMAXAOTcHSATSg9xWolyjpAT+PrYPfb2Lrk/6dJ3fJYnUWsGE/jygvZm1MbMEyk7MTivXZiNwLoCZdaYs9PO9diPNrI6ZtQHaA3NDVbzUHsl14nh+9Blc0KUJ9/17GU98uBrnNFGbSKhVGfrOuWLgVuA9YDllo3SWmtn9ZjbMa/YT4Admtgh4BbjelVlK2SeAZcC7wI+ccyXV8YdI5EuMj+Wpa3oyomcL/vz+Kn43fbmCXyTELNz+U2VnZ7ucnBy/yxAflZY67n9rGZM+W89V2S353YjTiI0xv8sSCWtmNt85l11VO82yKWEnxpuhs0FiHI9/lMvBomIeuaq7pmYWCQGFvoQlM+POCzrSoG48v317ObsLjvDoyO40aZDod2kiEU2HThLWbhzYloev7MbCTXsZ8ugsjeUXOUkKfQl7I3pm8tbtA2iRUpeb/jafe15fQuGRYr/LEolICn2JCO0y6vH6LWdy06C2TJ63kYufmM1Xefv8Lksk4ij0JWIkxMVwz0WdeXlsHwqKirnsqU95btYaXXhd5AQo9CXi9D8lnXfHn8U5nRrzu+kruG7iXLbvP+x3WSIRQaEvESklOYFnru3FgyNOY/6GPQx5dBb/0UlekSop9CVimRmjerf6+iTvuL/N5xdvLOHQEX3pW6QyCn2JeIEneV+Zu5GLn/hEJ3lFKqHQl1oh8CTvQe8k719nrdVJXpFyFPpSqwSe5H1g+nJGv6CTvCKBFPpS6wSe5M1Zr5O8IoEU+lIrVXaSV9/klWin0Jda7euTvGcdO8k7myWbdZJXopdCX2q9hLgY7hnamZdv7ENhUQmXPfUpT83IpUQneSUKKfQlavRvl867dwzkwq5N+eO7K7n6r5+Tt/eQ32WJ1CiFvkSVRkkJ/OXqHjx0RTe+ytvHkEdnMW3RFr/LEqkxCn2JOmbG5b0ymT5+IO0b1+P2V77kzlcXcuDwUb9LE6l2Cn2JWllpyUy5qR93nNeeNxdt4aLHPiFn/W6/yxKpVgp9iWpxsTHccV4HptzUjxgzrnx2Dg//ZyVHS0r9Lk2kWij0RYBeWSlMHz+QET0zefyjXC5/Zg7rdxb4XZZIyCn0RTz16sTx0BXdePLqnqzfWcDQxz/h1XkbcU5DO6X2UOiLlPO905vx7h0D6d6yEXf9cwm3/H0BewqO+F2WSEgo9EUq0KxhXf4+tg+/GNqJD1dsZ8hjs5i9eqffZYmctDi/CxAJVzExxriz2tG/XTp3vLqQa5//guysFFqnJ5OVmkSrtCRapyWTlZZEo6QEv8sVCYoF019pZkOAx4BYYIJz7vfl1j8CnO09TAIaO+caeetKgCXeuo3OuWHHe67s7GyXk5NzQn+ESHU7dKSEv3y8mnnr97BhVwHb9xd9Y33DuvFkpSXRKrXsjSDwDaFx/TqYmU+VS7Qws/nOuewq21UV+mYWC6wCzgc2A/OAUc65ZZW0vw3o4Zwb4z0+6JyrF2zhCn2JBIeOlLBxdyEbdhWwYVchG3Z7P3cVkrf30Dfm9UmMjyEr9dgbQRLDu7fg1BYNfaxeaqNgQz+Y7p3eQK5zbq33iycDw4EKQx8YBdwbbKEikahuQiwdm9anY9P631p3tKSULXsPsX5XIRt3FbDeezPYsKuAmavymTB7HVf3bsXPLuyobiGpccGEfgtgU8DjzUCfihqaWRbQBvgoYHGimeUAxcDvnXP/+o61ikSE+NgYstKSyUpLBjK+sW7foaM88v4qXpqznulLtnLXkE5cmd2SmBh1/0jNCPXonZHAVOdcScCyLO8jx9XAo2bWrvxGZjbOzHLMLCc/Pz/EJYmEj4Z147lvWFfeum0gpzSux92vL2HE059pjn+pMcGEfh7QMuBxpresIiOBVwIXOOfyvJ9rgRlAj/IbOeeec85lO+eyMzIyyq8WqXW6NG/AlJv68fCV3di85xDDnpzNL99Ywt5CfR9AqlcwoT8PaG9mbcwsgbJgn1a+kZl1AlKAOQHLUsysjnc/HTiTys8FiEQVM2NEz0w++ukgru/fmlfmbuTsh2Ywee5GSnWBF6kmVYa+c64YuBV4D1gOTHHOLTWz+80scPjlSGCy++ZwoM5AjpktAj6mrE9foS8SoEFiPPde0pW3b/9vl89lT3/G4s17/S5NaqGgxunXJA3ZlGjmnOONL/P43fQV7CooYlTvVvzsgo6kJGuUjxxfsEM2NQ2DSBgp3+Xz6rxNnPPnGbyiLh8JEYW+SBg61uXz1m0DOKVxPe5Rl4+EiEJfJIx1blY2yueRq7qRt+cQw5/8lF9olI+cBIW+SJgzMy7rUdblc0P/Nl6Xz0ymzt+suf7lhCn0RSJEg8R4fn1JF966bQBt0pP56WuLuOq5z1m9/YDfpUkEUeiLRJjOzRrw2k39+P2I01i57QAXPfYJf3h3BYeOlFS9sUQ9hb5IBIqJMUb2bsVHPxnEpT1a8PSMNZz38Ew+XL7d79IkzCn0RSJYWr06PHRFN14d15ekhFjGvpjDuJdyyNt7yO/SJEwp9EVqgT5t03j79oHcNaQTs1bnc/7DM3lu1hqOlpT6XZqEGYW+SC2REBfDLYPb8f6PB9GvbRq/m76CS56YTc763X6XJmFEoS9Sy7RMTWLC6Gye/X4v9h86yuXPzOGuqYvZU6Cx/aLQF6mVzIwLuzbl/TsHMe6stkxdsJlz/jyDKTmbNLY/yin0RWqx5Dpx/GJoZ966bQBtM+rx86mLuerZz1mlsf1RS7NsikSJ0lLHa/M38eA7Kzh4uJjBHRszvHtzzuvchLoJsX6XJycplBdGF5FaICbGuOqMVpzfpSnPzFzDmwvz+GD5dpISYrmwa1OGdW/OgFPSiY9VB0BtpiN9kShVUur4Yt0upi3cwvQlW9l/uJjU5AS+d1ozhnVvTq9WKbpgewQJ9khfoS8iFBWXMHNlPtMWbeGD5ds5fLSUFo3qckm35gzv3pxOTetjpjeAcKbQF5Hv5GBRMe8v28abC7fwyeqdlJQ6OjSpx/DuLRjWrTktU5P8LlEqoNAXkZO262AR05ds5c2FW8jZsAeAnq0aMbx7C4ae1oyM+nV8rlCOUeiLSEht2l3IvxdvYdrCLazYdoC4GOOqM1py+7ntadIg0e/yop5CX0SqzcptB/j75xt4Ze5G4mKN6/u34ZZB7WiYFO93aVFLoS8i1W7jrkIe+WAV/1qYR/06cdw8uB039G+jcf8+UOiLSI1ZvnU/D723kg9X7CCjfh1uP7c9I89oqTH/NSjY0Ne/iIictM7NGvD89Wcw9eZ+tE5L4n//9RXnPTyTNxfmUVoaXgeW0U6hLyIhk906lSk39eOF68+gbnws4ycv5HtPzObjFTs00VuYUOiLSEiZGWd3asz02wfy2MjuFBQVc8OkeVz57BzmaW5/3yn0RaRaxMQYw7u34IM7B/GbS09l/a5CrnhmDmMnzWP51v1+lxe1ggp9MxtiZivNLNfM7q5g/SNmttC7rTKzvQHrRpvZau82OpTFi0j4S4iL4ft9s5j5s8H8fEhH5q3fzdDHP+GOyV+ycVeh3+VFnSpH75hZLLAKOB/YDMwDRjnnllXS/jagh3NujJmlAjlANuCA+UAv59yeyp5Po3dEard9hUd5ZtYaXvh0HUXFpZye2YhBHTIY1CGD7i0bEatJ3r6TUE6t3BvIdc6t9X7xZGA4UGHoA6OAe737FwLvO+d2e9u+DwwBXgnieUWkFmqYFM9dQzpxff/WTJ67iZmrdvCXj1bz+IeraVg3ngHt079+E9A3fUMvmNBvAWwKeLwZ6FNRQzPLAtoAHx1n2xYVbDcOGAfQqlWrIEoSkUjXpEEi489rz/jz2rO38Aizc3cyc2U+M1fl8/birQB0alqfQR0zGNyhMb2yUkiI02nIkxXqi6iMBKY650pOZCPn3HPAc1DWvRPimkQkzDVKSuDi05tz8enNcc6xYtsBZq7KZ+bKfCbOXsezM9eSnBBL/1P++ylAs31+N8GEfh7QMuBxpresIiOBH5XbdnC5bWcEX56IRBszo3OzBnRu1oCbB7XjYFExn+XuLHsTWJXP+8u2A9A2I5lBHTI4p1NjzmyXrgu+BCmYE7lxlJ3IPZeyEJ8HXO2cW1quXSfgXaCN836pdyJ3PtDTa7aAshO5lQ7W1YlcEamMc461Owu+7gb6fO0uiopLaZOezA1ntuZ/emaSXCc6rwIb0rl3zGwo8CgQC0x0zj1gZvcDOc65aV6b+4BE59zd5bYdA/zCe/iAc+6F4z2XQl9EgnX4aAnvLd3GxNnrWLR5H/UT4xh5Rkuu69c66rp/NOGaiEQN5xwLNu5l4qfrePerbTjnuLBrU8YMaEN2VkpUXOoxlEM2RUTCmpnRKyuFXlkpbNl7iJfmlM31/85X2zitRUPGDGjN905rrtE/6EhfRGqpwiPFvL4gjxc+Xcea/AIy6tfhur5ZXN2nFWn1at9lHtW9IyIClJY6Zq3OZ+Kn65m1Kp+EuBgu7d6cG85sQ+dmDfwuL2TUvSMiQtnEb4M7NmZwx8bk7jjAC5+u558LNjMlZzP926Ux5sw2nNOpcdQM+dSRvohEnb2FR3hl7iZemrOerfsO06JRXc7p1JjBHTPo1y6NpITIOx5W946ISBWOlpTy7lfbeHNhHp/m7uLQ0RISYmPo0zaVQR0yGNyxMe0ykiNi9I9CX0TkBBQVlzBv3R5mrNzBxyt3sCa/AIDMlLqc3TH8PwUo9EVETsKm3YXMWJXPzJU7IuJTgEJfRCREAj8FzFiVT+6Og0DZp4DB3iygfdqmUj8x3rcaFfoiItVk0+5CZq7KZ8bKfD5bs5PCI2UTC6clJ9AyNYmstCRapSbRMrXsZ6vUJJo2SKzWEUIKfRGRGnDsU8CizXvZvKeQjbvLblv2Hqak9L/5mhAbQ2ZKXVql/feNIPBN4WQnitM4fRGRGlAnLpYB7dMZ0D79G8uPlpSyde9hNu4uZMPuAjbuLmST94Ywf8MeDhwu/kb79HpaVJvaAAAFj0lEQVQJ9GuXzhOjelRrvQp9EZFqEB8bU3ZUn5bEANK/tX5f4dGvPxUcu6UmV/85AYW+iIgPGibFc1pSQ07LbFijz6sp50REoohCX0Qkiij0RUSiiEJfRCSKKPRFRKKIQl9EJIoo9EVEoohCX0QkioTd3Dtmlg9sOIlfkQ7sDFE51Ul1hlak1AmRU6vqDL3qrDXLOZdRVaOwC/2TZWY5wUw65DfVGVqRUidETq2qM/TCoVZ174iIRBGFvohIFKmNof+c3wUESXWGVqTUCZFTq+oMPd9rrXV9+iIiUrnaeKQvIiKViMjQN7MhZrbSzHLN7O4K1tcxs1e99V+YWeuarxLMrKWZfWxmy8xsqZmNr6DNYDPbZ2YLvduvfap1vZkt8Wr41vUqrczj3j5dbGY9faixY8B+Wmhm+83sjnJtfNufZjbRzHaY2VcBy1LN7H0zW+39TKlk29Fem9VmNtqHOv9kZiu8f9s3zKxRJdse93VSA3XeZ2Z5Af++QyvZ9rgZUUO1vhpQ53ozW1jJtjW2TwFwzkXUDYgF1gBtgQRgEdClXJsfAs9490cCr/pUazOgp3e/PrCqgloHA2+FwX5dD6QfZ/1Q4B3AgL7AF2HwOthG2djksNifwFlAT+CrgGV/BO727t8N/KGC7VKBtd7PFO9+Sg3XeQEQ593/Q0V1BvM6qYE67wN+GsRr47gZURO1llv/Z+DXfu9T51xEHun3BnKdc2udc0eAycDwcm2GAy9696cC55pZ9V2GvhLOua3OuQXe/QPAcqBFTdcRIsOBl1yZz4FGZtbMx3rOBdY4507mi3wh5ZybBewutzjwtfgicGkFm14IvO+c2+2c2wO8DwypyTqdc/9xzh27aOvnQGZ1PX+wKtmfwQgmI0LqeLV62XMl8Ep11hCsSAz9FsCmgMeb+XaQft3GeyHvA9JqpLpKeF1MPYAvKljdz8wWmdk7Zta1Rgv7Lwf8x8zmm9m4CtYHs99r0kgq/08UDvvzmCbOua3e/W1AkwrahNu+HUPZp7qKVPU6qQm3et1QEyvpLgu3/TkQ2O6cW13J+hrdp5EY+hHHzOoB/wTucM7tL7d6AWVdFN2AJ4B/1XR9ngHOuZ7ARcCPzOwsn+qokpklAMOA1ypYHS7781tc2Wf5sB4uZ2a/BIqBlytp4vfr5GmgHdAd2EpZt0m4G8Xxj/JrdJ9GYujnAS0DHmd6yypsY2ZxQENgV41UV46ZxVMW+C87514vv945t985d9C7Px2IN7P0Gi4T51ye93MH8AZlH5EDBbPfa8pFwALn3PbyK8JlfwbYfqwbzPu5o4I2YbFvzex64GLgGu8N6luCeJ1UK+fcdudciXOuFPhrJc8fFvsTvs6fEcCrlbWp6X0aiaE/D2hvZm28I76RwLRybaYBx0ZAXA58VNmLuDp5fXnPA8udcw9X0qbpsfMNZtabsn+TGn2DMrNkM6t/7D5lJ/W+KtdsGnCdN4qnL7AvoNuiplV65BQO+7OcwNfiaODNCtq8B1xgZiled8UF3rIaY2ZDgJ8Dw5xzhZW0CeZ1Uq3KnUe6rJLnDyYjasp5wArn3OaKVvqyT2vqjHEob5SNJFlF2Rn6X3rL7qfsBQuQSNlH/1xgLtDWpzoHUPZxfjGw0LsNBW4Gbvba3AospWyEwedAfx/qbOs9/yKvlmP7NLBOA5709vkSINunfZpMWYg3DFgWFvuTsjeircBRyvqRx1J2LulDYDXwAZDqtc0GJgRsO8Z7veYCN/hQZy5l/eDHXqfHRr81B6Yf73VSw3X+zXv9LaYsyJuVr9N7/K2MqOlaveWTjr02A9r6tk+dc/pGrohINInE7h0REfmOFPoiIlFEoS8iEkUU+iIiUUShLyISRRT6IiJRRKEvIhJFFPoiIlHk/wFvBT8Be7PfawAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "def autocorr(x, max_lag, steps):\n",
    "    \n",
    "    results = []\n",
    "    \n",
    "    for i in range(1, max_lag, steps):\n",
    "        \n",
    "        x_hat = np.append(np.zeros(i), x[:len(x)-i])\n",
    "        results.append(np.corrcoef(x, x_hat)[1][0])\n",
    "        \n",
    "    return results\n",
    "\n",
    "plt.plot(autocorr(x, 20, 1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ArticleID                   048485a0\n",
       "DOI                 10.1038/048485a0\n",
       "Issue                           1247\n",
       "Volume                            48\n",
       "PageEnd                          508\n",
       "PageStart                        485\n",
       "PubYear                         1893\n",
       "SortDate                  1893-09-21\n",
       "Title        The British Association\n",
       "Type                              ed\n",
       "Journal                       Nature\n",
       "Name: 0, dtype: object"
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "article_df.loc[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 2: Reading and Preprocessign XML Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('data/nature/preprocessed/abstract_df.pickle', 'rb') as f:\n",
    "        abstract_df = pickle.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ArticleID</th>\n",
       "      <th>Abstract</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>048485a0</td>\n",
       "      <td>THE Nottingham meeting of the British Associat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>048508a0</td>\n",
       "      <td>I HAD hoped that the British Museum slander ca...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>048509a0</td>\n",
       "      <td>BISHOP'S RING still continues very conspicuous...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>048509b0</td>\n",
       "      <td>As the peculiarities of this season are receiv...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>048509c0</td>\n",
       "      <td>MR. W. SAVILLE-KENT, at present engaged as Com...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395498</th>\n",
       "      <td>149145b0</td>\n",
       "      <td>IN a recent paper, W. R. Wedel dealt with man'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395499</th>\n",
       "      <td>149146a0</td>\n",
       "      <td>R. C. HAYES, acting director of the Dominion O...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395500</th>\n",
       "      <td>149146b0</td>\n",
       "      <td>(\\nMeeting marked with an asterisk is open to ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395501</th>\n",
       "      <td>149146c0</td>\n",
       "      <td>APPLICATIONS are invited for the following app...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395502</th>\n",
       "      <td>149146d0</td>\n",
       "      <td>(\\nnot included in the monthly Books Supplemen...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>395503 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       ArticleID                                           Abstract\n",
       "0       048485a0  THE Nottingham meeting of the British Associat...\n",
       "1       048508a0  I HAD hoped that the British Museum slander ca...\n",
       "2       048509a0  BISHOP'S RING still continues very conspicuous...\n",
       "3       048509b0  As the peculiarities of this season are receiv...\n",
       "4       048509c0  MR. W. SAVILLE-KENT, at present engaged as Com...\n",
       "...          ...                                                ...\n",
       "395498  149145b0  IN a recent paper, W. R. Wedel dealt with man'...\n",
       "395499  149146a0  R. C. HAYES, acting director of the Dominion O...\n",
       "395500  149146b0  (\\nMeeting marked with an asterisk is open to ...\n",
       "395501  149146c0  APPLICATIONS are invited for the following app...\n",
       "395502  149146d0  (\\nnot included in the monthly Books Supplemen...\n",
       "\n",
       "[395503 rows x 2 columns]"
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "abstract_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [],
   "source": [
    "directory = 'data/nature/raw xml/'\n",
    "zfile = zipfile.ZipFile(os.path.join(directory, os.listdir(directory)[0]))\n",
    "\n",
    "for finfo in zfile.infolist():\n",
    "        ifile = zfile.open(finfo)\n",
    "        line_list = ifile.readlines()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'nature_v173_n4402_xml.zip'"
      ]
     },
     "execution_count": 171,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.listdir(directory)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "b'<iss>4402</iss>\\r\\n'"
      ]
     },
     "execution_count": 169,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "line_list[np.argmax(['4402' in str(line) for line in line_list])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'nature_v173_n4402_xml.zip'"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.listdir(directory)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[b'<?xml version=\"1.0\"?>\\r\\n',\n",
       " b'<!DOCTYPE headerx PUBLIC \"-//NPG//DTD headerx//EN\" \"headerx.dtd\">\\r\\n',\n",
       " b'<headerx>\\r\\n',\n",
       " b'<ArticleIdList>\\r\\n',\n",
       " b'<ArticleId>173508a0</ArticleId></ArticleIdList>\\r\\n',\n",
       " b'<pubfm>\\r\\n',\n",
       " b'<jtl>Nature</jtl>\\r\\n',\n",
       " b'<vol>173</vol>\\r\\n',\n",
       " b'<iss>4402</iss>\\r\\n',\n",
       " b'<idt>19540313</idt>\\r\\n',\n",
       " b'<categ id=\"categtxt\"/>\\r\\n',\n",
       " b'<categtxt>Reports and other Publications</categtxt>\\r\\n',\n",
       " b'<pp><spn>508</spn><epn>508</epn></pp>\\r\\n',\n",
       " b'<issn>0028-0836</issn>\\r\\n',\n",
       " b'<cpg><cpy>1954</cpy><cpn>Nature Publishing Group</cpn></cpg>\\r\\n',\n",
       " b'<doi>10.1038/173508a0</doi>\\r\\n',\n",
       " b'</pubfm>\\r\\n',\n",
       " b'<fm>\\r\\n',\n",
       " b'<atl>Reports and other Publications</atl>\\r\\n',\n",
       " b'</fm>\\r\\n',\n",
       " b'<bdy>(not included in the monthly Books Supplement)\\r\\n',\n",
       " b'Great Britain and Ireland\\r\\n',\n",
       " b'Air Ministry : Meteorological Office. Averages of Temperature for Great Britain and Northern Ireland 1921-50. Pp. ii &plus; 36. (London : H.M. Stationery Office, 1953.) 2s. net. [121\\r\\n',\n",
       " b'Ministry of Fuel and Power. Reports of H.M. Inspectors of Mines, under the Coal Mines Act, 1911, for the year 1952. Scottish Division. By H. R. Houston. Pp. ii &plus; 26. (London: H.M. Stationery Office, 1953.) Is. 3d. net. [121\\r\\n',\n",
       " b'Orgonomic Functionalism. Vol. 1, No. 1 (January, 1954). Pp. 46. (Nottingham: The Ritter Press, Alston, Alexandra Park, 1954.) 2s. Qd. [121\\r\\n',\n",
       " b'Eighth Annual Report and Accounts for the year ended 30th September, 1953. Pp. 12. (London : Negretti and .Zambra, Ltd., 1954.) [121\\r\\n',\n",
       " b'Department of Scientific and Industrial Research. Summary of Progress of the Geological Survey of Great Britain and the Museum of Practical Geology for the year 1952. With Report of the Geological Survey Board and Report of the Director. Pp. iv &plus; 72. (London: H.M. Stationery Office, 1953.) 2s. 6d. net. [121\\r\\n',\n",
       " b'Aero Research Technical Notes. Bulletin No. 132 : \"Araldite\" Miscellany. Pp. 8. (Duxford: Aero Research, Ltd., 1953.) [121\\r\\n',\n",
       " b'Communications of the Dublin Institute for Advanced Studies. Series A, No. 10 : Papers read at Cosmic Ray Colloquium, September, 1951. Pp. 195 (19 plates). (Dublin : Dublin Institute for Advanced Studies, 1952.) [121\\r\\n',\n",
       " b'Philosophical Transactions of the Royal Society of London. Series A. Mathematical and Physical Sciences. No. 914, Vol. 246 (12 January 1954): Spectral Theory for the Differential Equations of Simple Birth and Death Processes. By W. Ledermann and G. E. H. Reuter. Pp. 321-369. lls. No. 915, Vol. 246 (12 January 1954): The North Sea Storm Surge of 31 January and 1 February 1953. By J. R. Rossiter. Pp. 371-400. 7s. (London: Cambridge University Press, 1954.) [121\\r\\n',\n",
       " b'Ministry of Fuel and Power. Reports of H.M. Inspectors of Mines, amder the Coal Mines Act, 1911, for the year 1952. North Eastern Division. By C. W. Scott. Pp. ii &plus; 30. (London: H.M. Stationery Office, 1953.) Is. 9d. net. -. [181\\r\\n',\n",
       " b'Ministry of Fuel and Power. Reports of H.M. Inspectors of Mines, under the Coal Mines Act, 1911, for the year 1952. South Western Division. By T. A. Jones. Pp. ii &plus; 20. (London: H.M. Stationery Office, 1953.) Is. 3d. net. [191\\r\\n',\n",
       " b'London School of Hygiene and Tropical Medicine. (University of London.) Incorporating the Ross Institute. Report on the work of the School for the year 1952-1953. Pp. 126. (London : London School of Hygiene and Tropical Medicine, 1954.) J191\\r\\n',\n",
       " b'British Welding Journal. Vol. 1, No. 1 (January 1954). Pp. 36. (London : Institute of Welding, 1954.) 10s. [191\\r\\n',\n",
       " b'Other Countries\\r\\n',\n",
       " b'Smithsonian Institution. United States National Museum. Contributions from the United States National Herbarium. Vol. 30, Part 5 : Studies of Pacific Islands Plants, 15, The Genus Elaeocarpus in the New Hebrides, Fiji, Samoa, and Tonga. By A. C. Smith. Vol. 31,, Part 2: The Acanthaceae of Colombia, II. By Emery C. Leonard. (Washington, D.C. : Government Printing Office, 1953.) [3012\\r\\n',\n",
       " b'Sudan Government: Ministry of Agriculture. Annual Report of the Research Division, 1949/1950. Pp. ii &plus; 142. (Khartoum: Ministry of Agriculture, 1952.) [3012\\r\\n',\n",
       " b'United States Department of Commerce. National Bureau of Standards. Tables of Coefficients for the Numerical Calculation of Laplace Transforms. By Herbert E. Salzer. (Applied Mathematics Series-30.) Pp. ii &plus; 36. 25 cents. Hydraulic Research in the United States. Edited by Helen K. Middleton and Sonya W. Matchett. (Miscellaneous Publication-208.) Pp. xii &plus; 204. 1.25 dollars. (Washington, D.C. : Government Printing Office, 1953.) [3012\\r\\n',\n",
       " b'Veroffentlichungen des Fraunhofer Instituts, Freiburg I.B. Sonnen-Zirkular f iir die monate Juli, August, September 1953. Pp. 45. (Freiburg : Fraunhofer Institut, 1953.) [3012\\r\\n',\n",
       " b'British Honduras. Forest Department. Annual Report for 1952 (with Statistics in Standard Form for 1950 and 1951). Pp. 42. (British Honduras : Government Printer, 1953.) [3012\\r\\n',\n",
       " b'Nigeria : Agricultural Department. Special Bulletin No. 5 : Notes on the Main Types of Nigerian Soils. By Dr. H. Vine. Pp. 6 &plus; 1 map. (Lagos: Government Printer, 1953.) 2s. 6d. [61\\r\\n',\n",
       " b'Durban Museum and Art Gallery. Annual Report for the Municipal Year 1952-53. Pp. 12. (Durban : Durban Museum and Art Gallery, 1953.) [61\\r\\n',\n",
       " b'Report of Her Majesty\\'s Astronomer at the Cape of Good Hope to the Secretary of the Admiralty for the year 1952. Pp\". 16. (Observatory, C. P. : Royal Observatory, 1953.) [61\\r\\n',\n",
       " b'Metropolitan Life Insurance Company. Statistical Bulletin, Vol. 34, No. 11 : The Chances of Living To and Beyond Age 65. Pp. 12. (New York: Metropolitan Life Insurance Company, 1953.) [81\\r\\n',\n",
       " b'Nigeria. Annual Report on the Forest Administration of the Western Region of Nigeria for the year 1951-52. Pp. 48. (Lagos: Government Printer; London: Crown Agents for the Colonies, 1953.) 9rf.net. [81\\r\\n',\n",
       " b'World Health Organization. Technical Report Series No. 72 : Joint FAO/WHO Expert Committee on Nutrition. Third Report. Pp. 30. (Geneva: World Health Organization; London: H.M. Stationery Office, 1953.) Is. 6d.; 0.20 dollars; 65 French francs; 0.80 Swiss francs. [81\\r\\n',\n",
       " b'Linen Flax Fibre Production in New Zealand. By J. W. Hadfleld. Pp. xviii &plus; 126. (Christchurch, N.Z. : Linen Flax Corporation of New Zealand, and the N.Z. Departments of Agriculture, Industries and Commerce, and Scientific and Industrial Research, 1953.) [81\\r\\n',\n",
       " b'Current Biological Research in the Alaskan Arctic. Edited by Ira L. Wiggins. (Stanford University Publications: University Series. Biological Sciences, Vol. 11.) Pp. viii &plus; 55. (Stanford, California: Stanford University Press ; London : Oxford University Press, 1953.) 8s. net. [81\\r\\n',\n",
       " b'United States Department of Agriculture. Circular No. 919: Aiding Blister Rust Control by Silvicultural Measures on the Western White Pine Type. By Virgil D. Moss and Charles A. Wellner. Pp. 32. 10 cents. Circular No. 920 : The Hall Scale Eradication Project. By E. H. Fosen, A. W. Cressman and H. M. Armitage. Pp. 16. 10 cents. (Washington, D.C : Government Printing Office, 1953.) [81\\r\\n',\n",
       " b'Composite Wood. Vol. 1, No. 1 (October 1953). Pp. 26 &plus; 10. (Dehra Dun : Forest Research Institute, 1953.) 2/8/- rupees. [121\\r\\n',\n",
       " b'Bibliography of Scientific Publications of South Asia (India, Burma, Ceylon). No. 8 : July-December, 1952. Pp. vi &plus; 157. (New Delhi: UNESCO, 1953.) [121\\r\\n',\n",
       " b\"Suomen Geodeettisen Laitoksen Julkaisuja. Veroffentlichungen des Finnischen Geodatischen Institutes. No. 42 : On the Recent Uplift of the Earth's Crust in Finland. By Erkki Kaariainen. Pp. 106. No. 43: Astronomische Ortsbestimmungen in den Jahren 1946-1948. Von Pentti Kalaja. Pp. 146. (Helsinki: Suomen Gdtti Litk Jlkija, 1953.) [121\\r\\n\",\n",
       " b' Cl d Si M\"i\\r\\n',\n",
       " b'19461948. Von Pentt Kaja. Geodeettisen Laitoksen Julkaisuja, Adi Rl d Bli\\r\\n',\n",
       " b'eodeettisen Laitoksen Julkaisuja, 1953.) [\\r\\n',\n",
       " b'Academie Royale de Belgique. Classe des Sciences. Me\"moires, Tome 28, Fascicule 5 : Iiltude de I\\'lnteraction entre le champ propre d\\'une particle et un champ e\\'lectro-magne\\'tique homogene et constant. Par Marcel Demeur. Pp. 98. (Bruxelles: Palais des Academies, 1953.) 100 francs. [121\\r\\n',\n",
       " b'Tobacco Research Board of Southern Rhodesia. Interim Report No. 3 : How to Control Insects Attacking Tobacco. By P. W. Miles. Pp. 12. (Salisbury : Tobacco Research Board of Southern Rhodesia, 1953.) [121\\r\\n',\n",
       " b\"State of California. Department of Fish and Game: Marine Fisheries Branch. Fish Bulletin No. 93 : The Life History of the Cabezon, Scorpaenichthys marmoratus (Ayres). By Charles P. O'Con-nell. Pp. 77. Fish Bulletin No. 94 : The Behaviour and ^Reproduction of Salmonid Fishes in a Small Coastal Stream. By John C. Briggs. Pp. 62. (San Pedro, Calif.: California State Fisheries Laboratories, 1953.) [151</bdy></headerx>\"]"
      ]
     },
     "execution_count": 172,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "line_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7867"
      ]
     },
     "execution_count": 173,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(os.listdir(directory))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['nature_v173_n4402_xml.zip',\n",
       " 'nature_v18_n449_xml.zip',\n",
       " 'nature_v260_n5554_xml.zip',\n",
       " 'nature_v232_n5308_xml.zip',\n",
       " 'nature_v485_n7400test-jo_xml.zip',\n",
       " 'nature_v201_n4914_xml.zip',\n",
       " 'nature_v57_n1476_xml.zip',\n",
       " 'nature_v72_n1854_xml.zip',\n",
       " 'nature_v57_n1468_xml.zip',\n",
       " 'nature_v22_n554_xml.zip',\n",
       " 'nature_v127_n3205_xml.zip',\n",
       " 'nature_v78_n2034_xml.zip',\n",
       " 'nature_v87_n2175_xml.zip',\n",
       " 'nature_v12_n300_xml.zip',\n",
       " 'nature_v110_n2772_xml.zip',\n",
       " 'nature_v176_n4484_xml.zip',\n",
       " 'nature_v43_n1121_xml.zip',\n",
       " 'nature_v238_n5358_xml.zip',\n",
       " 'nature_v258_n5531_xml.zip',\n",
       " 'nature_v10_n237_xml.zip',\n",
       " 'nature_v67_n1748_xml.zip',\n",
       " 'nature_v158_n4018_xml.zip',\n",
       " 'nature_v20_n501_xml.zip',\n",
       " 'nature_v211_n5050_xml.zip',\n",
       " 'nature_v369_n6478_xml.zip',\n",
       " 'nature_v40_n1028_xml.zip',\n",
       " 'nature_v373_n6512_xml.zip',\n",
       " 'nature_v156_n3965_xml.zip',\n",
       " 'nature_v88_n2202_xml.zip',\n",
       " 'nature_v383_n6601_xml.zip',\n",
       " 'nature_v104_n2611_xml.zip',\n",
       " 'nature_v463_n7278_xml.zip',\n",
       " 'nature_v39_n992_xml.zip',\n",
       " 'nature_v109_n2738_xml.zip',\n",
       " 'nature_v146_n3696_xml.zip',\n",
       " 'nature_v27_n684_xml.zip',\n",
       " 'nature_v318_n6043_xml.zip',\n",
       " 'nature_v145_n3675_xml.zip',\n",
       " 'nature_v94_n2341_xml.zip',\n",
       " 'nature_v56_n1443_xml.zip',\n",
       " 'nature_v197_n4862_xml.zip',\n",
       " 'nature_v76_n1962_xml.zip',\n",
       " 'nature_v453_n7193_xml.zip',\n",
       " 'nature_v466_n7302_xml.zip',\n",
       " 'nature_v167_n4245_xml.zip',\n",
       " 'nature_v498_n7455testjo_xml.zip',\n",
       " 'nature_v192_n4801_xml.zip',\n",
       " 'nature_v288_n5787_xml.zip',\n",
       " 'nature_v329_n6138_xml.zip',\n",
       " 'nature_v228_n5276_xml.zip',\n",
       " 'nature_v550_n7674_supp_spot_xml.zip',\n",
       " 'nature_v456_n7219_xml.zip',\n",
       " 'nature_v63_n1625_xml.zip',\n",
       " 'nature_v112_n2825_xml.zip',\n",
       " 'nature_v75_n1942_xml.zip',\n",
       " 'nature_v390_n6660_xml.zip',\n",
       " 'nature_v144_n3639supp_xml.zip',\n",
       " 'nature_v5_n107_xml.zip',\n",
       " 'nature_v321_n6068_xml.zip',\n",
       " 'nature_v357_n6379_xml.zip',\n",
       " 'nature_v127_n3201_xml.zip',\n",
       " 'nature_v76_n1977_xml.zip',\n",
       " 'nature_v137_n3469_xml.zip',\n",
       " 'nature_v68_n1753_xml.zip',\n",
       " 'nature_v312_n5996_xml.zip',\n",
       " 'nature_v117_n2944supp_xml.zip',\n",
       " 'nature_v374_n6518_xml.zip',\n",
       " 'nature_v424_n6952_xml.zip',\n",
       " 'nature_v183_n4668_xml.zip',\n",
       " 'nature_v201_n4918_xml.zip',\n",
       " 'nature_v396_n6713_xml.zip',\n",
       " 'nature_v269_n5630_xml.zip',\n",
       " 'nature_v109_n2737_xml.zip',\n",
       " 'nature_v10_n247_xml.zip',\n",
       " 'nature_v481_n7380_xml.zip',\n",
       " 'nature_v203_n4942_xml.zip',\n",
       " 'nature_v497_n7449_xml.zip',\n",
       " 'nature_v467_n7318_xml.zip',\n",
       " 'nature_v14_n342_xml.zip',\n",
       " 'nature_v484_n7392_xml.zip',\n",
       " 'nature_v48_n1251_xml.zip',\n",
       " 'nature_v17_n424_xml.zip',\n",
       " 'nature_v484_n7393_xml.zip',\n",
       " 'nature_v4_n91_xml.zip',\n",
       " 'nature_v549_n7671_xml.zip',\n",
       " 'nature_v62_n1600_xml.zip',\n",
       " 'nature_v177_n4506_xml.zip',\n",
       " 'nature_v425_n6955_xml.zip',\n",
       " 'nature_v478_n7368_supp_xml.zip',\n",
       " 'nature_v31_n786_xml.zip',\n",
       " 'nature_v459_n7244_xml.zip',\n",
       " 'nature_v155_n3924_xml.zip',\n",
       " 'nature_v550_n7677_supp_out_xml.zip',\n",
       " 'nature_v122_n3077_xml.zip',\n",
       " 'nature_v124_n3120_xml.zip',\n",
       " 'nature_v162_n4112_xml.zip',\n",
       " 'nature_v533_n7602_supp_xml.zip',\n",
       " 'nature_v188_n4754_xml.zip',\n",
       " 'nature_v219_n5152_xml.zip',\n",
       " 'nature_v151_n3822_xml.zip',\n",
       " 'nature_v418_n6894_xml.zip',\n",
       " 'nature_v421_n6918_xml.zip',\n",
       " 'nature_v166_n4215_xml.zip',\n",
       " 'nature_v143_n3615_xml.zip',\n",
       " 'nature_v297_n5861_xml.zip',\n",
       " 'nature_v48_n1227_xml.zip',\n",
       " 'nature_v65_n1686_xml.zip',\n",
       " 'nature_v225_n5230_xml.zip',\n",
       " 'nature_v259_n5539_xml.zip',\n",
       " 'nature_v546_n7659_supp_xml.zip',\n",
       " 'nature_v516_n7529_supp_xml.zip',\n",
       " 'nature_v406_n6794_xml.zip',\n",
       " 'nature_v42_n1087_xml.zip',\n",
       " 'nature_v94_n2353_xml.zip',\n",
       " 'nature_v161_n4090_xml.zip',\n",
       " 'nature_v450_n7173_xml.zip',\n",
       " 'nature_v83_n2112_xml.zip',\n",
       " 'nature_v36_n927_xml.zip',\n",
       " 'nature_v46_n1188_xml.zip',\n",
       " 'nature_v118_n2973_xml.zip',\n",
       " 'nature_v46_n1196_xml.zip',\n",
       " 'nature_v81_n2077_xml.zip',\n",
       " 'nature_v172_n4380_xml.zip',\n",
       " 'nature_v175_n4462_xml.zip',\n",
       " 'nature_v164_n4175_xml.zip',\n",
       " 'nature_v468_n7322_xml.zip',\n",
       " 'nature_v132_n3341_xml.zip',\n",
       " 'nature_v235_n5338_xml.zip',\n",
       " 'nature_v167_n4243_xml.zip',\n",
       " 'nature_v57_n1462_xml.zip',\n",
       " 'nature_v145_n3663_xml.zip',\n",
       " 'nature_v129_n3263_xml.zip',\n",
       " 'nature_v58_n1504_xml.zip',\n",
       " 'nature_v289_n5799_xml.zip',\n",
       " 'nature_v432_n7020_xml.zip',\n",
       " 'nature_v10_n257_xml.zip',\n",
       " 'nature_v72_n1876_xml.zip',\n",
       " 'nature_v26_n671_xml.zip',\n",
       " 'nature_v71_n1832_xml.zip',\n",
       " 'nature_v178_n4547_xml.zip',\n",
       " 'nature_v29_n732_xml.zip',\n",
       " 'nature_v302_n5905_xml.zip',\n",
       " 'nature_v365_n6441_xml.zip',\n",
       " 'nature_v67_n1736_xml.zip',\n",
       " 'nature_v137_n3453_xml.zip',\n",
       " 'nature_v389_n6653_xml.zip',\n",
       " 'nature_v290_n5808_xml.zip',\n",
       " 'nature_v255_n5511_xml.zip',\n",
       " 'nature_v40_n1038_xml.zip',\n",
       " 'nature_v426_n6966_xml.zip',\n",
       " 'nature_v95_n2383_xml.zip',\n",
       " 'nature_v153_n3890_xml.zip',\n",
       " 'nature_v7_n158_xml.zip',\n",
       " 'nature_v76_n1957_xml.zip',\n",
       " 'nature_v16_n402_xml.zip',\n",
       " 'nature_v178_n4525_xml.zip',\n",
       " 'nature_v403_n6770_xml.zip',\n",
       " 'nature_v70_n1824_xml.zip',\n",
       " 'nature_v432_n7013_xml.zip',\n",
       " 'nature_v509_n7498_supp_xml.zip',\n",
       " 'nature_v316_n6027_xml.zip',\n",
       " 'nature_v372_n6501_xml.zip',\n",
       " 'nature_v469_n7330_xml.zip',\n",
       " 'nature_v150_n3808_xml.zip',\n",
       " 'nature_v37_n960_xml.zip',\n",
       " 'nature_v379_n6566_xml.zip',\n",
       " 'nature_v215_n5097_xml.zip',\n",
       " 'nature_v277_n5695_xml.zip',\n",
       " 'nature_v62_n1614_xml.zip',\n",
       " 'nature_v268_n5618_xml.zip',\n",
       " 'nature_v46_n1187_xml.zip',\n",
       " 'nature_v150_n3813_xml.zip',\n",
       " 'nature_v527_n7578_supp_xml.zip',\n",
       " 'nature_v146_n3694_xml.zip',\n",
       " 'nature_v157_n3999_xml.zip',\n",
       " 'nature_v84_n2123_xml.zip',\n",
       " 'nature_v7_n159_xml.zip',\n",
       " 'nature_v144_n3656supp_xml.zip',\n",
       " 'nature_v170_n4319_xml.zip',\n",
       " 'nature_v145_n3662_xml.zip',\n",
       " 'nature_v48_n1244_xml.zip',\n",
       " 'nature_v202_n4928_xml.zip',\n",
       " 'nature_v68_n1749_xml.zip',\n",
       " 'nature_v13_n327_xml.zip',\n",
       " 'nature_v139_n3506_xml.zip',\n",
       " 'nature_v173_n4396_xml.zip',\n",
       " 'nature_v110_n2766_xml.zip',\n",
       " 'nature_v326_n6108_xml.zip',\n",
       " 'nature_v55_n1429_xml.zip',\n",
       " 'nature_v96_n2403_xml.zip',\n",
       " 'nature_v94_n2342_xml.zip',\n",
       " 'nature_v115_n2889_xml.zip',\n",
       " 'nature_v182_n4641_xml.zip',\n",
       " 'nature_v276_n5686_xml.zip',\n",
       " 'nature_v102_n2556_xml.zip',\n",
       " 'nature_v168_n4269_xml.zip',\n",
       " 'nature_v158_n4021_xml.zip',\n",
       " 'nature_v101_n2523_xml.zip',\n",
       " 'nature_v86_n2169_xml.zip',\n",
       " 'nature_v171_n4347_xml.zip',\n",
       " 'nature_v460_n7259_xml.zip',\n",
       " 'nature_v221_n5177_xml.zip',\n",
       " 'nature_v68_n1765_xml.zip',\n",
       " 'nature_v239_n5374_xml.zip',\n",
       " 'nature_v186_n4723_xml.zip',\n",
       " 'nature_v190_n4780_xml.zip',\n",
       " 'nature_v148_n3764_xml.zip',\n",
       " 'nature_v280_n5717_xml.zip',\n",
       " 'nature_v177_n4516_xml.zip',\n",
       " 'nature_v101_n2545_xml.zip',\n",
       " 'nature_v85_n2150_xml.zip',\n",
       " 'nature_v23_n594_xml.zip',\n",
       " 'nature_v289_n5795_xml.zip',\n",
       " 'nature_v48_n1247_xml.zip',\n",
       " 'nature_v440_n7082_xml.zip',\n",
       " 'nature_v79_n2044_xml.zip',\n",
       " 'nature_v224_n5220_xml.zip',\n",
       " 'nature_v108_n2720_xml.zip',\n",
       " 'nature_v56_n1438_xml.zip',\n",
       " 'nature_v28_n718_xml.zip',\n",
       " 'nature_v38_n978_xml.zip',\n",
       " 'nature_v70_n1821_xml.zip',\n",
       " 'nature_v453_n7195_xml.zip',\n",
       " 'nature_v127_n3199_xml.zip',\n",
       " 'nature_v7_n163_xml.zip',\n",
       " 'nature_v225_n5236_xml.zip',\n",
       " 'nature_v296_n5855_xml.zip',\n",
       " 'nature_v104_n2612_xml.zip',\n",
       " 'nature_v114_n2867_xml.zip',\n",
       " 'nature_v534_n7609_xml.zip',\n",
       " 'nature_v63_n1636_xml.zip',\n",
       " 'nature_v63_n1629_xml.zip',\n",
       " 'nature_v248_n5449_xml.zip',\n",
       " 'nature_v378_n6558_xml.zip',\n",
       " 'nature_v166_n4213_xml.zip',\n",
       " 'nature_v41_n1057_xml.zip',\n",
       " 'nature_v21_n545_xml.zip',\n",
       " 'nature_v85_n2155_xml.zip',\n",
       " 'nature_v262_n5567_xml.zip',\n",
       " 'nature_v29_n734_xml.zip',\n",
       " 'nature_v550_n7675_supp_out_xml.zip',\n",
       " 'nature_v550_n7675_supp_xml.zip',\n",
       " 'nature_v16_n408_xml.zip',\n",
       " 'nature_v169_n4307_xml.zip',\n",
       " 'nature_v86_n2164_xml.zip',\n",
       " 'nature_v459_n7243_xml.zip',\n",
       " 'nature_v158_n4023_xml.zip',\n",
       " 'nature_v59_n1529_xml.zip',\n",
       " 'nature_v175_n4457_xml.zip',\n",
       " 'nature_v34_n885_xml.zip',\n",
       " 'nature_v349_n6306_xml.zip',\n",
       " 'nature_v113_n2851_xml.zip',\n",
       " 'nature_v409_n6823_xml.zip',\n",
       " 'nature_v204_n4964_xml.zip',\n",
       " 'nature_v553_n7687_xml.zip',\n",
       " 'nature_v122_n3063_xml.zip',\n",
       " 'nature_v39_n997_xml.zip',\n",
       " 'nature_v154_n3904_xml.zip',\n",
       " 'nature_v67_n1723_xml.zip',\n",
       " 'nature_v142_n3584_xml.zip',\n",
       " 'nature_v173_n4397_xml.zip',\n",
       " 'nature_v205_n4969_xml.zip',\n",
       " 'nature_v38_n975_xml.zip',\n",
       " 'nature_v132_n3327_xml.zip',\n",
       " 'nature_v14_n358_xml.zip',\n",
       " 'nature_v147_n3721_xml.zip',\n",
       " 'nature_v349_n6311_xml.zip',\n",
       " 'nature_v284_n5757_xml.zip',\n",
       " 'nature_v126_n3185_xml.zip',\n",
       " 'nature_v61_n1574_xml.zip',\n",
       " 'nature_v91_n2268_xml.zip',\n",
       " 'nature_v443_n7109_xml.zip',\n",
       " 'nature_v57_n1473_xml.zip',\n",
       " 'nature_v511_n7508_xml.zip',\n",
       " 'nature_v303_n5917_xml.zip',\n",
       " 'nature_v249_n5454_xml.zip',\n",
       " 'nature_v8_n187_xml.zip',\n",
       " 'nature_v399_n6733_xml.zip',\n",
       " 'nature_v212_n5065_xml.zip',\n",
       " 'nature_v36_n915_xml.zip',\n",
       " 'nature_v23_n583_xml.zip',\n",
       " 'nature_v446_n7132_xml.zip',\n",
       " 'nature_v38_n977_xml.zip',\n",
       " 'nature_v108_n2721_xml.zip',\n",
       " 'nature_v359_n6392_xml.zip',\n",
       " 'nature_v85_n2141_xml.zip',\n",
       " 'nature_v466_n7303_xml.zip',\n",
       " 'nature_v276_n5685_xml.zip',\n",
       " 'nature_v167_n4251_xml.zip',\n",
       " 'nature_v333_n6170_xml.zip',\n",
       " 'nature_v281_n5727_xml.zip',\n",
       " 'nature_v56_n1449_xml.zip',\n",
       " 'nature_v259_n5538_xml.zip',\n",
       " 'nature_v15_n390_xml.zip',\n",
       " 'nature_v340_n6232_xml.zip',\n",
       " 'nature_v179_n4561_xml.zip',\n",
       " 'nature_v363_n6431_xml.zip',\n",
       " 'nature_v198_n4877_xml.zip',\n",
       " 'nature_v469_n7328_xml.zip',\n",
       " 'nature_v459_n7246_xml.zip',\n",
       " 'nature_v46_n1179_xml.zip',\n",
       " 'nature_v116_n2921_xml.zip',\n",
       " 'nature_v225_n5229_xml.zip',\n",
       " 'nature_v21_n534_xml.zip',\n",
       " 'nature_v2_n48_xml.zip',\n",
       " 'nature_v290_n5802_xml.zip',\n",
       " 'nature_v3_n64_xml.zip',\n",
       " 'nature_v535_n7613_xml.zip',\n",
       " 'nature_v141_n3560_xml.zip',\n",
       " 'nature_v136_n3441_xml.zip',\n",
       " 'nature_v26_n674_xml.zip',\n",
       " 'nature_v465_n7294_xml.zip',\n",
       " 'nature_v80_n2058_xml.zip',\n",
       " 'nature_v69_n1777_xml.zip',\n",
       " 'nature_v55_n1431_xml.zip',\n",
       " 'nature_v299_n5883_xml.zip',\n",
       " 'nature_v52_n1343_xml.zip',\n",
       " 'nature_v421_n6922_xml.zip',\n",
       " 'nature_v240_n5377_xml.zip',\n",
       " 'nature_v183_n4664_xml.zip',\n",
       " 'nature_v16_n415_xml.zip',\n",
       " 'nature_v63_n1632_xml.zip',\n",
       " 'nature_v243_n5403_xml.zip',\n",
       " 'nature_v522_n7557_xml.zip',\n",
       " 'nature_v281_n5729_xml.zip',\n",
       " 'nature_v372_n6503_xml.zip',\n",
       " 'nature_v46_n1199_xml.zip',\n",
       " 'nature_v531_n7592_supp_xml.zip',\n",
       " 'nature_v132_n3336_xml.zip',\n",
       " 'nature_v333_n6172_xml.zip',\n",
       " 'nature_v157_n3977_xml.zip',\n",
       " 'nature_v128_n3220_xml.zip',\n",
       " 'nature_v550_n7675_xml.zip',\n",
       " 'nature_v130_n3296_xml.zip',\n",
       " 'nature_v53_n1378_xml.zip',\n",
       " 'nature_v388_n6643_xml.zip',\n",
       " 'nature_v88_n2193_xml.zip',\n",
       " 'nature_v19_n491_xml.zip',\n",
       " 'nature_v209_n5027_xml.zip',\n",
       " 'nature_v287_n5781_xml.zip',\n",
       " 'nature_v302_n5908_xml.zip',\n",
       " 'nature_v379_n6560_xml.zip',\n",
       " 'nature_v44_n1148_xml.zip',\n",
       " 'nature_v36_n934_xml.zip',\n",
       " 'nature_v173_n4409_xml.zip',\n",
       " 'nature_v186_n4725_xml.zip',\n",
       " 'nature_v9_n220_xml.zip',\n",
       " 'nature_v33_n854_xml.zip',\n",
       " 'nature_v14_n363_xml.zip',\n",
       " 'nature_v111_n2788_xml.zip',\n",
       " 'nature_v14_n356_xml.zip',\n",
       " 'nature_v129_n3258_xml.zip',\n",
       " 'nature_v82_n2095_xml.zip',\n",
       " 'nature_v258_n5533_xml.zip',\n",
       " 'nature_v60_n1550_xml.zip',\n",
       " 'nature_v291_n5815_xml.zip',\n",
       " 'nature_v342_n6251_xml.zip',\n",
       " 'nature_v20_n522_xml.zip',\n",
       " 'nature_v547_n7661_xml.zip',\n",
       " 'nature_v285_n5760_xml.zip',\n",
       " 'nature_v116_n2910_xml.zip',\n",
       " 'nature_v176_n4480_xml.zip',\n",
       " 'nature_v416_n6880_xml.zip',\n",
       " 'nature_v59_n1520_xml.zip',\n",
       " 'nature_v88_n2199_xml.zip',\n",
       " 'nature_v212_n5069_xml.zip',\n",
       " 'nature_v36_n923_xml.zip',\n",
       " 'nature_v501_n7465_xml.zip',\n",
       " 'nature_v17_n437_xml.zip',\n",
       " 'nature_v134_n3375_xml.zip',\n",
       " 'nature_v548_n7666_xml.zip',\n",
       " 'nature_v393_n6680_xml.zip',\n",
       " 'nature_v86_n2157_xml.zip',\n",
       " 'nature_v18_n467_xml.zip',\n",
       " 'nature_v30_n776_xml.zip',\n",
       " 'nature_v69_n1792_xml.zip',\n",
       " 'nature_v109_n2726_xml.zip',\n",
       " 'nature_v166_n4227_xml.zip',\n",
       " 'nature_v310_n5978_xml.zip',\n",
       " 'nature_v537_n7619_supp_xml.zip',\n",
       " 'nature_v71_n1841_xml.zip',\n",
       " 'nature_v39_n1006_xml.zip',\n",
       " 'nature_v75_n1932_xml.zip',\n",
       " 'nature_v79_n2038_xml.zip',\n",
       " 'nature_v52_n1336_xml.zip',\n",
       " 'nature_v98_n2466_xml.zip',\n",
       " 'nature_v451_n7179_xml.zip',\n",
       " 'nature_v216_n5110_xml.zip',\n",
       " 'nature_v326_n6110_xml.zip',\n",
       " 'nature_v449_n7159_xml.zip',\n",
       " 'nature_v101_n2531_xml.zip',\n",
       " 'nature_v146_n3709_xml.zip',\n",
       " 'nature_v292_n5825_xml.zip',\n",
       " 'nature_v418_n6893_xml.zip',\n",
       " 'nature_v16_n397_xml.zip',\n",
       " 'nature_v76_n1971_xml.zip',\n",
       " 'nature_v132_n3333_xml.zip',\n",
       " 'nature_v48_n1250_xml.zip',\n",
       " 'nature_v138_n3491_xml.zip',\n",
       " 'nature_v195_n4839_xml.zip',\n",
       " 'nature_v175_n4454_xml.zip',\n",
       " 'nature_v130_n3288_xml.zip',\n",
       " 'nature_v46_n1200_xml.zip',\n",
       " 'nature_v76_n1983_xml.zip',\n",
       " 'nature_v64_n1648_xml.zip',\n",
       " 'nature_v398_n6725_xml.zip',\n",
       " 'nature_v195_n4841_xml.zip',\n",
       " 'nature_v208_n5017_xml.zip',\n",
       " 'nature_v14_n357_xml.zip',\n",
       " 'nature_v141_n3582_xml.zip',\n",
       " 'nature_v19_n481_xml.zip',\n",
       " 'nature_v91_n2266_xml.zip',\n",
       " 'nature_v410_n6831_xml.zip',\n",
       " 'nature_v388_n6638_xml.zip',\n",
       " 'nature_v117_n2940_xml.zip',\n",
       " 'nature_v52_n1353_xml.zip',\n",
       " 'nature_v440_n7080_xml.zip',\n",
       " 'nature_v29_n745_xml.zip',\n",
       " 'nature_v242_n5392_xml.zip',\n",
       " 'nature_v100_n2514_xml.zip',\n",
       " 'nature_v248_n5451_xml.zip',\n",
       " 'nature_v380_n6571_xml.zip',\n",
       " 'nature_v101_n2525_xml.zip',\n",
       " 'nature_v91_n2274_xml.zip',\n",
       " 'nature_v448_n7156_xml.zip',\n",
       " 'nature_v69_n1788_xml.zip',\n",
       " 'nature_v123_n3099_xml.zip',\n",
       " 'nature_v207_n4996_xml.zip',\n",
       " 'nature_v155_n3944_xml.zip',\n",
       " 'nature_v144_n3657_xml.zip',\n",
       " 'nature_v111_n2775_xml.zip',\n",
       " 'nature_v19_n494_xml.zip',\n",
       " 'nature_v53_n1360_xml.zip',\n",
       " 'nature_v318_n6044_xml.zip',\n",
       " 'nature_v150_n3794_xml.zip',\n",
       " 'nature_v47_n1206_xml.zip',\n",
       " 'nature_v241_n5385_xml.zip',\n",
       " 'nature_v161_n4096_xml.zip',\n",
       " 'nature_v12_n290_xml.zip',\n",
       " 'nature_v19_n484_xml.zip',\n",
       " 'nature_v22_n555_xml.zip',\n",
       " 'nature_v493_n7432_xml.zip',\n",
       " 'nature_v198_n4886_xml.zip',\n",
       " 'nature_v83_n2122_xml.zip',\n",
       " 'nature_v68_n1773_xml.zip',\n",
       " 'nature_v402_n6759_xml.zip',\n",
       " 'nature_v113_n2833_xml.zip',\n",
       " 'nature_v219_n5151_xml.zip',\n",
       " 'nature_v75_n1954_xml.zip',\n",
       " 'nature_v112_n2824_xml.zip',\n",
       " 'nature_v54_n1385_xml.zip',\n",
       " 'nature_v91_n2264_xml.zip',\n",
       " 'nature_v352_n6335_xml.zip',\n",
       " 'nature_v347_n6293_xml.zip',\n",
       " 'nature_v254_n5497_xml.zip',\n",
       " 'nature_v43_n1119_xml.zip',\n",
       " 'nature_v76_n1964_xml.zip',\n",
       " 'nature_v223_n5209_xml.zip',\n",
       " 'nature_v468_n7327_xml.zip',\n",
       " 'nature_v102_n2562_xml.zip',\n",
       " 'nature_v167_n4259_xml.zip',\n",
       " 'nature_v314_n6012_xml.zip',\n",
       " 'nature_v104_n2615_xml.zip',\n",
       " 'nature_v183_n4672_xml.zip',\n",
       " 'nature_v91_n2278_xml.zip',\n",
       " 'nature_v501_n7467_xml.zip',\n",
       " 'nature_v110_n2769_xml.zip',\n",
       " 'nature_v165_n4188_xml.zip',\n",
       " 'nature_v176_n4490_xml.zip',\n",
       " 'nature_v49_n1255_xml.zip',\n",
       " 'nature_v180_n4584_xml.zip',\n",
       " 'nature_v119_n3004_xml.zip',\n",
       " 'nature_v447_n7140_xml.zip',\n",
       " 'nature_v29_n740_xml.zip',\n",
       " 'nature_v83_n2108_xml.zip',\n",
       " 'nature_v62_n1611_xml.zip',\n",
       " 'nature_v39_n1010_xml.zip',\n",
       " 'nature_v154_n3907_xml.zip',\n",
       " 'nature_v351_n6327_xml.zip',\n",
       " 'nature_v88_n2207_xml.zip',\n",
       " 'nature_v389_n6649_xml.zip',\n",
       " 'nature_v187_n4741_xml.zip',\n",
       " 'nature_v298_n5873_xml.zip',\n",
       " 'nature_v472_n7342_xml.zip',\n",
       " 'nature_v136_n3429_xml.zip',\n",
       " 'nature_v3_n69_xml.zip',\n",
       " 'nature_v132_n3332_xml.zip',\n",
       " 'nature_v29_n755_xml.zip',\n",
       " 'nature_v54_n1402_xml.zip',\n",
       " 'nature_v63_n1643_xml.zip',\n",
       " 'nature_v102_n2558_xml.zip',\n",
       " 'nature_v527_n7576_supp_xml.zip',\n",
       " 'nature_v81_n2081_xml.zip',\n",
       " 'nature_v313_n5997_xml.zip',\n",
       " 'nature_v78_n2027_xml.zip',\n",
       " 'nature_v34_n873_xml.zip',\n",
       " 'nature_v137_n3473_xml.zip',\n",
       " 'nature_v361_n6413_xml.zip',\n",
       " 'nature_v307_n5947_xml.zip',\n",
       " 'nature_v471_n7337_xml.zip',\n",
       " 'nature_v113_n2850_xml.zip',\n",
       " 'nature_v76_n1978_xml.zip',\n",
       " 'nature_v422_n6930_xml.zip',\n",
       " 'nature_v181_n4614_xml.zip',\n",
       " 'nature_v94_n2356_xml.zip',\n",
       " 'nature_v519_n7544_supp_xml.zip',\n",
       " 'nature_v58_n1496_xml.zip',\n",
       " 'nature_v98_n2456_xml.zip',\n",
       " 'nature_v184_n4679_xml.zip',\n",
       " 'nature_v520_n7546_xml.zip',\n",
       " 'nature_v366_n6452_xml.zip',\n",
       " 'nature_v344_n6268_xml.zip',\n",
       " 'nature_v109_n2735_xml.zip',\n",
       " 'nature_v541_n7638_xml.zip',\n",
       " 'nature_v227_n5256_xml.zip',\n",
       " 'nature_v472_n7343_xml.zip',\n",
       " 'nature_v103_n2595_xml.zip',\n",
       " 'nature_v134_n3398_xml.zip',\n",
       " 'nature_v316_n6028_xml.zip',\n",
       " 'nature_v126_n3167_xml.zip',\n",
       " 'nature_v110_n2771_xml.zip',\n",
       " 'nature_v310_n5975_xml.zip',\n",
       " 'nature_v53_n1361_xml.zip',\n",
       " 'nature_v209_n5020_xml.zip',\n",
       " 'nature_v117_n2948supp_xml.zip',\n",
       " 'nature_v8_n190_xml.zip',\n",
       " 'nature_v126_n3186_xml.zip',\n",
       " 'nature_v60_n1557_xml.zip',\n",
       " 'nature_v162_n4116_xml.zip',\n",
       " 'nature_v65_n1683_xml.zip',\n",
       " 'nature_v233_n5320_xml.zip',\n",
       " 'nature_v184_n4694_xml.zip',\n",
       " 'nature_v480_n7378_supp_xml.zip',\n",
       " 'nature_v122_n3071_xml.zip',\n",
       " 'nature_v202_n4935_xml.zip',\n",
       " 'nature_v262_n5569_xml.zip',\n",
       " 'nature_v172_n4378_xml.zip',\n",
       " 'nature_v365_n6443_xml.zip',\n",
       " 'nature_v106_n2663_xml.zip',\n",
       " 'nature_v459_n7248_xml.zip',\n",
       " 'nature_v448_n7155_xml.zip',\n",
       " 'nature_v538_n7623_xml.zip',\n",
       " 'nature_v408_n6811_xml.zip',\n",
       " 'nature_v78_n2021_xml.zip',\n",
       " 'nature_v349_n6312_xml.zip',\n",
       " 'nature_v555_n7696_xml.zip',\n",
       " 'nature_v77_n1986_xml.zip',\n",
       " 'nature_v295_n5847_xml.zip',\n",
       " 'nature_v175_n4447_xml.zip',\n",
       " 'nature_v110_n2768_xml.zip',\n",
       " 'nature_v59_n1535_xml.zip',\n",
       " 'nature_v250_n5466_xml.zip',\n",
       " 'nature_v41_n1063_xml.zip',\n",
       " 'nature_v15_n389_xml.zip',\n",
       " 'nature_v431_n7010_xml.zip',\n",
       " 'nature_v156_n3967_xml.zip',\n",
       " 'nature_v201_n4915_xml.zip',\n",
       " 'nature_v273_n5663_xml.zip',\n",
       " 'nature_v29_n733_xml.zip',\n",
       " 'nature_v24_n619_xml.zip',\n",
       " 'nature_v335_n6188_xml.zip',\n",
       " 'nature_v128_n3237_xml.zip',\n",
       " 'nature_v304_n5925_xml.zip',\n",
       " 'nature_v94_n2354_xml.zip',\n",
       " 'nature_v523_n7562_xml.zip',\n",
       " 'nature_v96_n2406_xml.zip',\n",
       " 'nature_v30_n764_xml.zip',\n",
       " 'nature_v144_n3653_xml.zip',\n",
       " 'nature_v362_n6415_xml.zip',\n",
       " 'nature_v468_n7326_xml.zip',\n",
       " 'nature_v231_n5301_xml.zip',\n",
       " 'nature_v227_n5255_xml.zip',\n",
       " 'nature_v273_n5660_xml.zip',\n",
       " 'nature_v214_n5091_xml.zip',\n",
       " 'nature_v253_n5488_xml.zip',\n",
       " 'nature_v93_n2319_xml.zip',\n",
       " 'nature_v116_n2907_xml.zip',\n",
       " 'nature_v300_n5888_xml.zip',\n",
       " 'nature_v462_n7271_xml.zip',\n",
       " 'nature_v89_n2230_xml.zip',\n",
       " 'nature_v99_n2484_xml.zip',\n",
       " 'nature_v4_n99_xml.zip',\n",
       " 'nature_v175_n4465_xml.zip',\n",
       " 'nature_v330_n6146_xml.zip',\n",
       " 'nature_v157_n3981_xml.zip',\n",
       " 'nature_v169_n4299_xml.zip',\n",
       " 'nature_v31_n794_xml.zip',\n",
       " 'nature_v291_n5810_xml.zip',\n",
       " 'nature_v190_n4770_xml.zip',\n",
       " 'nature_v216_n5116_xml.zip',\n",
       " 'nature_v176_n4491_xml.zip',\n",
       " 'nature_v431_n7012_xml.zip',\n",
       " 'nature_v173_n4404_xml.zip',\n",
       " 'nature_v218_n5145_xml.zip',\n",
       " 'nature_v448_n7157_xml.zip',\n",
       " 'nature_v135_n3405_xml.zip',\n",
       " 'nature_v478_n7367_xml.zip',\n",
       " 'nature_v62_n1606_xml.zip',\n",
       " 'nature_v45_n1166_xml.zip',\n",
       " 'nature_v166_n4226_xml.zip',\n",
       " 'nature_v107_n2683_xml.zip',\n",
       " 'nature_v1_n10_xml.zip',\n",
       " 'nature_v179_n4567_xml.zip',\n",
       " 'nature_v326_n6114_xml.zip',\n",
       " 'nature_v369_n6481_xml.zip',\n",
       " 'nature_v518_n7540_supp_xml.zip',\n",
       " 'nature_v78_n2015_xml.zip',\n",
       " 'nature_v6_n137_xml.zip',\n",
       " 'nature_v45_n1165_xml.zip',\n",
       " 'nature_v44_n1135_xml.zip',\n",
       " 'nature_v544_n7650_xml.zip',\n",
       " 'nature_v364_n6437_xml.zip',\n",
       " 'nature_v310_n5980_xml.zip',\n",
       " 'nature_v92_n2313_xml.zip',\n",
       " 'nature_v170_n4321_xml.zip',\n",
       " 'nature_v228_n5278_xml.zip',\n",
       " 'nature_v476_n7361_xml.zip',\n",
       " 'nature_v6_n132_xml.zip',\n",
       " 'nature_v24_n602_xml.zip',\n",
       " 'nature_v344_n6264_xml.zip',\n",
       " 'nature_v426_n6962_xml.zip',\n",
       " 'nature_v105_n2634_xml.zip',\n",
       " 'nature_v173_n4392_xml.zip',\n",
       " 'nature_v115_n2888_xml.zip',\n",
       " 'nature_v228_n5272_xml.zip',\n",
       " 'nature_v256_n5515_xml.zip',\n",
       " 'nature_v54_n1406_xml.zip',\n",
       " 'nature_v68_n1774_xml.zip',\n",
       " 'nature_v202_n4931_xml.zip',\n",
       " 'nature_v494_n7438_xml.zip',\n",
       " 'nature_v489_n7414_xml.zip',\n",
       " 'nature_v79_n2037_xml.zip',\n",
       " 'nature_v214_n5085_xml.zip',\n",
       " 'nature_v98_n2449_xml.zip',\n",
       " 'nature_v51_n1319_xml.zip',\n",
       " 'nature_v468_n7320_xml.zip',\n",
       " 'nature_v208_n5006_xml.zip',\n",
       " 'nature_v72_n1855_xml.zip',\n",
       " 'nature_v159_n4043_xml.zip',\n",
       " 'nature_v150_n3797_xml.zip',\n",
       " 'nature_v458_n7234_xml.zip',\n",
       " 'nature_v131_n3306_xml.zip',\n",
       " 'nature_v3_n53_xml.zip',\n",
       " 'nature_v71_n1842_xml.zip',\n",
       " 'nature_v183_n4653_xml.zip',\n",
       " 'nature_v162_n4128_xml.zip',\n",
       " 'nature_v9_n221_xml.zip',\n",
       " 'nature_v245_n5426_xml.zip',\n",
       " 'nature_v129_n3262_xml.zip',\n",
       " 'nature_v120_n3034_xml.zip',\n",
       " 'nature_v9_n212_xml.zip',\n",
       " 'nature_v339_n6219_xml.zip',\n",
       " 'nature_v521_n7550_xml.zip',\n",
       " 'nature_v83_n2107_xml.zip',\n",
       " 'nature_v145_n3664_xml.zip',\n",
       " 'nature_v387_n6628_xml.zip',\n",
       " 'nature_v180_n4588_xml.zip',\n",
       " 'nature_v243_n5404_xml.zip',\n",
       " 'nature_v176_n4478_xml.zip',\n",
       " 'nature_v304_n5926_xml.zip',\n",
       " 'nature_v317_n6034_xml.zip',\n",
       " 'nature_v31_n796_xml.zip',\n",
       " 'nature_v260_n5550_xml.zip',\n",
       " 'nature_v163_n4140_xml.zip',\n",
       " 'nature_v22_n558_xml.zip',\n",
       " 'nature_v339_n6225_xml.zip',\n",
       " 'nature_v190_n4778_xml.zip',\n",
       " 'nature_v45_n1152_xml.zip',\n",
       " 'nature_v334_n6184_xml.zip',\n",
       " 'nature_v142_n3604_xml.zip',\n",
       " 'nature_v130_n3270_xml.zip',\n",
       " 'nature_v409_n6821_xml.zip',\n",
       " 'nature_v397_n6715_xml.zip',\n",
       " 'nature_v254_n5495_xml.zip',\n",
       " 'nature_v97_n2418_xml.zip',\n",
       " 'nature_v140_n3543_xml.zip',\n",
       " 'nature_v239_n5371_xml.zip',\n",
       " 'nature_v121_n3037_xml.zip',\n",
       " 'nature_v162_n4124_xml.zip',\n",
       " 'nature_v93_n2321_xml.zip',\n",
       " 'nature_v341_n6244_xml.zip',\n",
       " 'nature_v528_n7581_xml.zip',\n",
       " 'nature_v82_n2093_xml.zip',\n",
       " 'nature_v260_n5553_xml.zip',\n",
       " 'nature_v416_n6876_xml.zip',\n",
       " 'nature_v80_n2054_xml.zip',\n",
       " 'nature_v77_n2002_xml.zip',\n",
       " 'nature_v423_n6940_xml.zip',\n",
       " 'nature_v394_n6695_xml.zip',\n",
       " 'nature_v143_n3617_xml.zip',\n",
       " 'nature_v121_n3052_xml.zip',\n",
       " 'nature_v186_n4730_xml.zip',\n",
       " 'nature_v127_n3207_xml.zip',\n",
       " 'nature_v67_n1725_xml.zip',\n",
       " 'nature_v10_n251_xml.zip',\n",
       " 'nature_v106_n2674_xml.zip',\n",
       " 'nature_v471_n7340_xml.zip',\n",
       " 'nature_v137_n3465_xml.zip',\n",
       " 'nature_v308_n5960_xml.zip',\n",
       " 'nature_v73_n1896_xml.zip',\n",
       " 'nature_v496_n7444_xml.zip',\n",
       " 'nature_v355_n6357_xml.zip',\n",
       " 'nature_v402_n6763_xml.zip',\n",
       " 'nature_v152_n3856_xml.zip',\n",
       " 'nature_v109_n2728_xml.zip',\n",
       " 'nature_v158_n4004_xml.zip',\n",
       " 'nature_v94_n2359_xml.zip',\n",
       " 'nature_v106_n2677_xml.zip',\n",
       " 'nature_v75_n1950_xml.zip',\n",
       " 'nature_v212_n5064_xml.zip',\n",
       " 'nature_v7_n168_xml.zip',\n",
       " 'nature_v376_n6538_xml.zip',\n",
       " 'nature_v2_n31_xml.zip',\n",
       " 'nature_v269_n5628_xml.zip',\n",
       " 'nature_v210_n5035_xml.zip',\n",
       " 'nature_v182_n4632_xml.zip',\n",
       " 'nature_v395_n6704_xml.zip',\n",
       " 'nature_v182_n4631_xml.zip',\n",
       " 'nature_v273_n5662_xml.zip',\n",
       " 'nature_v23_n596_xml.zip',\n",
       " 'nature_v54_n1394_xml.zip',\n",
       " 'nature_v136_n3428_xml.zip',\n",
       " 'nature_v371_n6499_xml.zip',\n",
       " 'nature_v34_n886_xml.zip',\n",
       " 'nature_v96_n2397_xml.zip',\n",
       " 'nature_v322_n6074_xml.zip',\n",
       " 'nature_v496_n7445_xml.zip',\n",
       " 'nature_v112_n2804_xml.zip',\n",
       " 'nature_v450_n7172_xml.zip',\n",
       " 'nature_v253_n5491_xml.zip',\n",
       " 'nature_v42_n1085_xml.zip',\n",
       " 'nature_v2_n44_xml.zip',\n",
       " 'nature_v542_n7640_xml.zip',\n",
       " 'nature_v195_n4842_xml.zip',\n",
       " 'nature_v412_n6849_xml.zip',\n",
       " 'nature_v41_n1061_xml.zip',\n",
       " 'nature_v87_n2187_xml.zip',\n",
       " 'nature_v265_n5592_xml.zip',\n",
       " 'nature_v381_n6584_xml.zip',\n",
       " 'nature_v513_n7517_xml.zip',\n",
       " 'nature_v262_n5563_xml.zip',\n",
       " 'nature_v166_n4233_xml.zip',\n",
       " 'nature_v180_n4598_xml.zip',\n",
       " 'nature_v85_n2142_xml.zip',\n",
       " 'nature_v464_n7286_xml.zip',\n",
       " 'nature_v243_n5405_xml.zip',\n",
       " 'nature_v3_n73_xml.zip',\n",
       " 'nature_v53_n1365_xml.zip',\n",
       " 'nature_v314_n6011_xml.zip',\n",
       " 'nature_v269_n5626_xml.zip',\n",
       " 'nature_v194_n4825_xml.zip',\n",
       " 'nature_v90_n2236_xml.zip',\n",
       " 'nature_v1_n15_xml.zip',\n",
       " 'nature_v131_n3314_xml.zip',\n",
       " 'nature_v473_n7348_xml.zip',\n",
       " 'nature_v357_n6380_xml.zip',\n",
       " 'nature_v19_n478_xml.zip',\n",
       " 'nature_v192_n4802_xml.zip',\n",
       " 'nature_v284_n5752_xml.zip',\n",
       " 'nature_v10_n261_xml.zip',\n",
       " 'nature_v223_n5210_xml.zip',\n",
       " 'nature_v315_n6020_xml.zip',\n",
       " 'nature_v140_n3544_xml.zip',\n",
       " 'nature_v76_n1979_xml.zip',\n",
       " 'nature_v118_n2966_xml.zip',\n",
       " 'nature_v147_n3733_xml.zip',\n",
       " 'nature_v354_n6353_xml.zip',\n",
       " 'nature_v79_n2046_xml.zip',\n",
       " 'nature_v223_n5213_xml.zip',\n",
       " 'nature_v215_n5101_xml.zip',\n",
       " 'nature_v143_n3613_xml.zip',\n",
       " 'nature_v37_n955_xml.zip',\n",
       " 'nature_v526_n7575_supp_xml.zip',\n",
       " 'nature_v479_n7372_xml.zip',\n",
       " 'nature_v13_n315_xml.zip',\n",
       " 'nature_v239_n5373_xml.zip',\n",
       " 'nature_v135_n3414_xml.zip',\n",
       " 'nature_v163_n4154_xml.zip',\n",
       " 'nature_v138_n3495_xml.zip',\n",
       " 'nature_v474_n7349_xml.zip',\n",
       " 'nature_v347_n6289_xml.zip',\n",
       " 'nature_v100_n2518_xml.zip',\n",
       " 'nature_v243_n5408_xml.zip',\n",
       " 'nature_v139_n3519_xml.zip',\n",
       " 'nature_v138_n3489_xml.zip',\n",
       " 'nature_v156_n3957_xml.zip',\n",
       " 'nature_v243_n5402_xml.zip',\n",
       " 'nature_v135_n3406_xml.zip',\n",
       " 'nature_v188_n4756_xml.zip',\n",
       " 'nature_v114_n2870_xml.zip',\n",
       " 'nature_v526_n7574_supp_xml.zip',\n",
       " 'nature_v176_n4487_xml.zip',\n",
       " 'nature_v164_n4164_xml.zip',\n",
       " 'nature_v159_n4039_xml.zip',\n",
       " 'nature_v250_n5465_xml.zip',\n",
       " 'nature_v175_n4451_xml.zip',\n",
       " 'nature_v24_n620_xml.zip',\n",
       " 'nature_v73_n1891_xml.zip',\n",
       " 'nature_v282_n5740_xml.zip',\n",
       " 'nature_v415_n6872_xml.zip',\n",
       " 'nature_v34_n887_xml.zip',\n",
       " 'nature_v32_n829_xml.zip',\n",
       " 'nature_v136_n3451_xml.zip',\n",
       " 'nature_v78_n2025_xml.zip',\n",
       " 'nature_v27_n694_xml.zip',\n",
       " 'nature_v101_n2528_xml.zip',\n",
       " 'nature_v63_n1642_xml.zip',\n",
       " 'nature_v281_n5733_xml.zip',\n",
       " 'nature_v204_n4957_xml.zip',\n",
       " 'nature_v181_n4626_xml.zip',\n",
       " 'nature_v535_n7610_xml.zip',\n",
       " 'nature_v461_n7265_xml.zip',\n",
       " 'nature_v20_n506_xml.zip',\n",
       " 'nature_v96_n2409_xml.zip',\n",
       " 'nature_v146_n3705_xml.zip',\n",
       " 'nature_v16_n410_xml.zip',\n",
       " 'nature_v62_n1595_xml.zip',\n",
       " 'nature_v100_n2517_xml.zip',\n",
       " 'nature_v437_n7062_xml.zip',\n",
       " 'nature_v148_n3763_xml.zip',\n",
       " 'nature_v294_n5837_xml.zip',\n",
       " 'nature_v237_n5349_xml.zip',\n",
       " 'nature_v66_n1706_xml.zip',\n",
       " 'nature_v190_n4775_xml.zip',\n",
       " 'nature_v122_n3074_xml.zip',\n",
       " 'nature_v117_n2952_xml.zip',\n",
       " 'nature_v176_n4494_xml.zip',\n",
       " 'nature_v184_n4681_xml.zip',\n",
       " 'nature_v162_n4117_xml.zip',\n",
       " 'nature_v65_n1673_xml.zip',\n",
       " 'nature_v280_n5723_xml.zip',\n",
       " 'nature_v517_n7533_xml.zip',\n",
       " 'nature_v169_n4313_xml.zip',\n",
       " 'nature_v149_n3771_xml.zip',\n",
       " 'nature_v10_n256_xml.zip',\n",
       " 'nature_v4_n98_xml.zip',\n",
       " 'nature_v71_n1852_xml.zip',\n",
       " 'nature_v146_n3699_xml.zip',\n",
       " 'nature_v333_n6168_xml.zip',\n",
       " 'nature_v399_n6738supp_xml.zip',\n",
       " 'nature_v215_n5098_xml.zip',\n",
       " 'nature_v156_n3971_xml.zip',\n",
       " 'nature_v172_n4383_xml.zip',\n",
       " 'nature_v212_n5061_xml.zip',\n",
       " 'nature_v63_n1634_xml.zip',\n",
       " 'nature_v517_n7532_xml.zip',\n",
       " 'nature_v410_n6829_xml.zip',\n",
       " 'nature_v279_n5711_xml.zip',\n",
       " 'nature_v156_n3953_xml.zip',\n",
       " 'nature_v100_n2519_xml.zip',\n",
       " 'nature_v34_n865_xml.zip',\n",
       " 'nature_v78_n2032_xml.zip',\n",
       " 'nature_v392_n6678_xml.zip',\n",
       " 'nature_v311_n5988_xml.zip',\n",
       " 'nature_v11_n281_xml.zip',\n",
       " 'nature_v332_n6160_xml.zip',\n",
       " 'nature_v349_n6307_xml.zip',\n",
       " 'nature_v8_n185_xml.zip',\n",
       " 'nature_v131_n3317_xml.zip',\n",
       " 'nature_v24_n601_xml.zip',\n",
       " 'nature_v372_n6505_xml.zip',\n",
       " 'nature_v151_n3836_xml.zip',\n",
       " 'nature_v164_n4163_xml.zip',\n",
       " 'nature_v172_n4376_xml.zip',\n",
       " 'nature_v422_n6934_xml.zip',\n",
       " 'nature_v493_n7430_xml.zip',\n",
       " 'nature_v59_n1514_xml.zip',\n",
       " 'nature_v412_n6844_xml.zip',\n",
       " 'nature_v55_n1414_xml.zip',\n",
       " 'nature_v89_n2212_xml.zip',\n",
       " 'nature_v78_n2031_xml.zip',\n",
       " 'nature_v107_n2685_xml.zip',\n",
       " 'nature_v60_n1545_xml.zip',\n",
       " 'nature_v112_n2810_xml.zip',\n",
       " 'nature_v342_n6249_xml.zip',\n",
       " 'nature_v22_n572_xml.zip',\n",
       " 'nature_v180_n4575_xml.zip',\n",
       " 'nature_v182_n4645_xml.zip',\n",
       " 'nature_v18_n458_xml.zip',\n",
       " 'nature_v224_n5219_xml.zip',\n",
       " 'nature_v45_n1163_xml.zip',\n",
       " 'nature_v198_n4885_xml.zip',\n",
       " 'nature_v59_n1526_xml.zip',\n",
       " 'nature_v8_n203_xml.zip',\n",
       " 'nature_v394_n6696_xml.zip',\n",
       " 'nature_v300_n5892_xml.zip',\n",
       " 'nature_v3_n58_xml.zip',\n",
       " 'nature_v38_n991_xml.zip',\n",
       " 'nature_v347_n6295_xml.zip',\n",
       " 'nature_v97_n2435_xml.zip',\n",
       " 'nature_v479_n7373_xml.zip',\n",
       " 'nature_v47_n1216_xml.zip',\n",
       " 'nature_v114_n2873_xml.zip',\n",
       " 'nature_v440_n7086_xml.zip',\n",
       " 'nature_v295_n5848_xml.zip',\n",
       " 'nature_v144_n3635_xml.zip',\n",
       " 'nature_v188_n4752_xml.zip',\n",
       " 'nature_v405_n6782_xml.zip',\n",
       " 'nature_v89_n2229_xml.zip',\n",
       " 'nature_v118_n2957supp_xml.zip',\n",
       " 'nature_v332_n6162_xml.zip',\n",
       " 'nature_v399_n6732_xml.zip',\n",
       " 'nature_v469_n7329_xml.zip',\n",
       " 'nature_v94_n2361_xml.zip',\n",
       " 'nature_v56_n1456_xml.zip',\n",
       " 'nature_v459_n7249_xml.zip',\n",
       " 'nature_v23_n600_xml.zip',\n",
       " 'nature_v26_n668_xml.zip',\n",
       " 'nature_v238_n5362_xml.zip',\n",
       " 'nature_v38_n969_xml.zip',\n",
       " 'nature_v13_n323_xml.zip',\n",
       " 'nature_v321_n6073_xml.zip',\n",
       " 'nature_v157_n3984_xml.zip',\n",
       " 'nature_v32_n815_xml.zip',\n",
       " 'nature_v71_n1846_xml.zip',\n",
       " 'nature_v324_n6097_xml.zip',\n",
       " 'nature_v358_n6388_xml.zip',\n",
       " 'nature_v49_n1261_xml.zip',\n",
       " 'nature_v145_n3672_xml.zip',\n",
       " 'nature_v8_n183_xml.zip',\n",
       " 'nature_v127_n3202_xml.zip',\n",
       " 'nature_v136_n3452_xml.zip',\n",
       " 'nature_v547_n7663_xml.zip',\n",
       " 'nature_v195_n4847_xml.zip',\n",
       " 'nature_v170_n4323_xml.zip',\n",
       " 'nature_v453_n7196_xml.zip',\n",
       " 'nature_v65_n1679_xml.zip',\n",
       " 'nature_v254_n5502_xml.zip',\n",
       " 'nature_v43_n1107_xml.zip',\n",
       " 'nature_v515_n7526_xml.zip',\n",
       " 'nature_v329_n6141_xml.zip',\n",
       " 'nature_v33_n847_xml.zip',\n",
       " 'nature_v185_n4706_xml.zip',\n",
       " 'nature_v163_n4141_xml.zip',\n",
       " 'nature_v163_n4144_xml.zip',\n",
       " 'nature_v73_n1880_xml.zip',\n",
       " 'nature_v407_n6803_xml.zip',\n",
       " 'nature_v63_n1618_xml.zip',\n",
       " 'nature_v11_n264_xml.zip',\n",
       " 'nature_v63_n1640_xml.zip',\n",
       " 'nature_v24_n615_xml.zip',\n",
       " 'nature_v188_n4751_xml.zip',\n",
       " 'nature_v409_n6820_xml.zip',\n",
       " 'nature_v74_n1926_xml.zip',\n",
       " 'nature_v250_n5463_xml.zip',\n",
       " 'nature_v45_n1150_xml.zip',\n",
       " 'nature_v329_n6142_xml.zip',\n",
       " 'nature_v191_n4790_xml.zip',\n",
       " 'nature_v98_n2469_xml.zip',\n",
       " 'nature_v74_n1916_xml.zip',\n",
       " 'nature_v72_n1877_xml.zip',\n",
       " 'nature_v362_n6418_xml.zip',\n",
       " 'nature_v227_n5254_xml.zip',\n",
       " 'nature_v343_n6255_xml.zip',\n",
       " 'nature_v159_n4042_xml.zip',\n",
       " 'nature_v99_n2488_xml.zip',\n",
       " 'nature_v111_n2791_xml.zip',\n",
       " 'nature_v184_n4689_xml.zip',\n",
       " 'nature_v165_n4198_xml.zip',\n",
       " 'nature_v6_n146_xml.zip',\n",
       " 'nature_v369_n6482_xml.zip',\n",
       " 'nature_v335_n6191_xml.zip',\n",
       " 'nature_v161_n4081_xml.zip',\n",
       " 'nature_v61_n1576_xml.zip',\n",
       " 'nature_v187_n4731_xml.zip',\n",
       " 'nature_v335_n6189_xml.zip',\n",
       " 'nature_v193_n4818_xml.zip',\n",
       " 'nature_v277_n5691_xml.zip',\n",
       " 'nature_v183_n4667_xml.zip',\n",
       " 'nature_v172_n4367_xml.zip',\n",
       " 'nature_v244_n5411_xml.zip',\n",
       " 'nature_v43_n1097_xml.zip',\n",
       " 'nature_v87_n2179_xml.zip',\n",
       " 'nature_v476_n7359_xml.zip',\n",
       " 'nature_v216_n5122_xml.zip',\n",
       " 'nature_v82_n2102_xml.zip',\n",
       " 'nature_v10_n254_xml.zip',\n",
       " 'nature_v514_n7524_supp_xml.zip',\n",
       " 'nature_v149_n3791_xml.zip',\n",
       " 'nature_v343_n6256_xml.zip',\n",
       " 'nature_v515_n7528_xml.zip',\n",
       " 'nature_v159_n4050_xml.zip',\n",
       " 'nature_v2_n46_xml.zip',\n",
       " 'nature_v177_n4508_xml.zip',\n",
       " 'nature_v275_n5680_xml.zip',\n",
       " 'nature_v377_n6544_xml.zip',\n",
       " 'nature_v221_n5184_xml.zip',\n",
       " 'nature_v198_n4876_xml.zip',\n",
       " 'nature_v97_n2442_xml.zip',\n",
       " 'nature_v178_n4534_xml.zip',\n",
       " 'nature_v101_n2534_xml.zip',\n",
       " 'nature_v261_n5561_xml.zip',\n",
       " 'nature_v526_n7572_supp_xml.zip',\n",
       " 'nature_v118_n2964_xml.zip',\n",
       " 'nature_v303_n5912_xml.zip',\n",
       " 'nature_v435_n7038_xml.zip',\n",
       " 'nature_v129_n3260_xml.zip',\n",
       " 'nature_v519_n7542_xml.zip',\n",
       " 'nature_v90_n2249_xml.zip',\n",
       " 'nature_v400_n6739_xml.zip',\n",
       " ...]"
      ]
     },
     "execution_count": 190,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.listdir(directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_reference_mapping(filename, content):\n",
    "    \n",
    "    \"\"\"\n",
    "    Given XML filename and XML file, extract rid mappings and attribute data\n",
    "    \"\"\"\n",
    "    \n",
    "    mappings  = {}\n",
    "    extracted = {}\n",
    "    references = {}\n",
    "    parsed    = xmltodict.parse(content.decode('UTF-8'))\n",
    "    soup      = BeautifulSoup(content)\n",
    "\n",
    "    if '@id' in parsed['article']:\n",
    "        extracted['id']       = str(parsed['article']['@id'])\n",
    "    if '@language' in parsed['article']:\n",
    "        extracted['language'] = str(parsed['article']['@language'])\n",
    "    if '@publish' in parsed['article']:\n",
    "        extracted['publish']  = str(parsed['article']['@publish'])\n",
    "    if '@relation' in parsed['article']:\n",
    "        extracted['relation'] = str(parsed['article']['@relation'])\n",
    "    \n",
    "    if ('article' in parsed):\n",
    "        if ('pubfm' in parsed['article']):\n",
    "            if ('jtl' in parsed['article']['pubfm']):\n",
    "                extracted['jtl']   = str(parsed['article']['pubfm']['jtl'])\n",
    "    \n",
    "    if ('article' in parsed):\n",
    "        if ('pubfm' in parsed['article']):\n",
    "            if ('vol' in parsed['article']['pubfm']):\n",
    "                extracted['vol']   = str(parsed['article']['pubfm']['vol'])\n",
    "    \n",
    "    if ('article' in parsed):\n",
    "        if ('pubfm' in parsed['article']):\n",
    "            if ('issue' in parsed['article']['pubfm']):\n",
    "                extracted['issue'] = str(parsed['article']['pubfm']['issue'])\n",
    "\n",
    "    if ('article' in parsed):\n",
    "        if ('pubfm' in parsed['article']):\n",
    "            if ('vol' in parsed['article']['pubfm']):\n",
    "                extracted['doi']   = str(parsed['article']['pubfm']['doi'])\n",
    "    \n",
    "    if ('article' in parsed):\n",
    "        if ('fm' in parsed['article']):\n",
    "            if ('atl' in parsed['article']['fm']):\n",
    "                extracted['title']   = str(parsed['article']['fm']['atl'])\n",
    "     \n",
    "    \n",
    "    del parsed\n",
    "    \n",
    "    for bib in soup.find_all(\"bib\"):\n",
    "        \n",
    "        try:\n",
    "            reference_attr = {}\n",
    "            \n",
    "            reference_attr['title']   = str(bib.atl.contents[0])\n",
    "            reference_attr['snm']     = str([i.contents[0] for i in bib.find_all('snm')])\n",
    "            reference_attr['fnm']     = str([i.contents[0] for i in bib.find_all('fnm')])\n",
    "            reference_attr['journal'] = str(bib.jtl.contents[0])\n",
    "            reference_attr['year']    = str(bib.find_all('cd')[0].contents[0])\n",
    "\n",
    "            references.update({bib.attrs['id']: reference_attr})\n",
    "        \n",
    "        except Exception as e:\n",
    "            \n",
    "            if DEBUG == True:\n",
    "            \n",
    "                print('='*50)\n",
    "                print('Something is wrong with BeatifulSoup Tags: %s' % str(bib))\n",
    "                for i in ['snm', 'fnm', 'jtl', 'year', 'atl']:\n",
    "                    if len(bib.find_all(i)) == 0:\n",
    "                        print('%s attribute is missing.' %i)\n",
    "            \n",
    "            else: pass\n",
    "            \n",
    "    \n",
    "    references = {'metadata': extracted, 'references': references}\n",
    "    \n",
    "    try:\n",
    "        with open('data/intermediate/references/%s' % (filename + '.json'), 'w') as f:\n",
    "            f.write(json.dumps(references))\n",
    "        return True\n",
    "    \n",
    "    except TypeError as e:\n",
    "        #print('Some contents of the file %s is not serializable' % filename)\n",
    "        raise e\n",
    "        \n",
    "        \n",
    "def get_reference_text(filename, content):\n",
    "    \n",
    "    \"\"\"\n",
    "    Given XML filename and XML file, extract referencing text and reference metadata\n",
    "    TODO: Compile all regex to make it faster\n",
    "    \n",
    "    Returns  {str(unique_paper_indentifier), list(preceeding_text)}\n",
    "    \"\"\"\n",
    "    \n",
    "    content   = content.decode('UTF-8')\n",
    "    bibid     = re.findall(r'<bibr\\srid=\\\"(.*?)\\\"\\s*\\/>', content)\n",
    "    #bibtext   = re.findall(r\"\\s.*?<bibr\\s\", content)\n",
    "    \n",
    "    bibtext_intermediate = [s for s in sent_detector.tokenize(content)\\\n",
    "                               if re.search(r'<bibr\\srid', s)]\n",
    "\n",
    "    bibtext_intermediate = [s.split('</p>') for s in bibtext_intermediate]\n",
    "    bibtext_intermediate = [item for sublist in bibtext_intermediate for item in sublist]\n",
    "\n",
    "    bibtext_intermediate = [s for s in bibtext_intermediate if re.search(r'<bibr\\srid', s)]\n",
    "    text_counter         = [len(re.findall(r'<bibr\\srid', s)) for s in bibtext_intermediate]\n",
    "\n",
    "    bibtext = []\n",
    "\n",
    "    for i, s in enumerate(bibtext_intermediate):\n",
    "        while text_counter[i] != 0:\n",
    "            bibtext.append(s)\n",
    "            text_counter[i] -= 1\n",
    "        \n",
    "    \n",
    "    assert(len(bibid) == len(bibtext)), \"The bibid's and preceeding text don't match: \" +\\\n",
    "                                        \"for article %s\\n\" % filename +\\\n",
    "                                        \"bibid: %s\\n\" %bibid +\\\n",
    "                                        \"bibtext: %s\\n\" %bibtext\n",
    "    \n",
    "    with open('data/intermediate/tex_ref_mappings/%s' % (filename + '.json'), 'w') as f:\n",
    "        f.write(json.dumps([{i[1]:i[0].split(' ')} for i in zip(bibid, bibtext)]))\n",
    "    \n",
    "    return \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_zips():\n",
    "    \n",
    "    \"\"\"\n",
    "    Function to extract relevant files from the filebase\n",
    "    \"\"\"\n",
    "    \n",
    "    citations = {}\n",
    "    directory = 'data/nature/raw xml/'\n",
    "    zip_list  = [i for i in os.listdir(directory) if 'supp_xml' not in i]\n",
    "    \n",
    "    errored   = []\n",
    "    #zfile     = zipfile.ZipFile(os.path.join(directory, zip_list[0]))\n",
    "    \n",
    "    for zfile in tqdm(zip_list):\n",
    "        \n",
    "        zfile     = zipfile.ZipFile(os.path.join(directory, zfile))\n",
    "\n",
    "        for finfo in zfile.infolist():\n",
    "\n",
    "            if 'nature' in finfo.filename:\n",
    "\n",
    "                try:\n",
    "                    ifile = zfile.open(finfo)\n",
    "                    content = ifile.read()\n",
    "                    tex_ref_map = get_reference_text(finfo.filename, content)\n",
    "                    ref_id_ref_map = get_reference_mapping(finfo.filename, content)\n",
    "\n",
    "                    citations[finfo.filename] = {'tex_ref_map': tex_ref_map,\n",
    "                                                'ref_id_ref_map': ref_id_ref_map}\n",
    "                except Exception as e:\n",
    "\n",
    "                    print('Filename: %s, Zip:%s, Content:%s' % (finfo.filename,\n",
    "                                                                zfile,\n",
    "                                                                content))\n",
    "                    errored.append({'Filename':finfo.filename,\n",
    "                                    'Zip':zfile ,\n",
    "                                    'Content': content})\n",
    "                    #return content\n",
    "                    #raise Exception\n",
    "            \n",
    "    return citations, errored\n",
    "        \n",
    "#papers, errored = get_zips()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 0 ns, sys: 0 ns, total: 0 ns\n",
      "Wall time: 6.44 µs\n"
     ]
    }
   ],
   "source": [
    "%time\n",
    "\n",
    "def get_zips_parallel_mapper(directory):\n",
    "    \n",
    "    \"\"\"\n",
    "    Function to extract relevant files from the filebase\n",
    "    \"\"\"\n",
    "    \n",
    "    try:\n",
    "        \n",
    "        zip_list  = [i for i in os.listdir(directory) if 'supp_xml' not in i]\n",
    "        errored   = []\n",
    "        finfos    = []\n",
    "        contents  = []\n",
    "\n",
    "        for zfile in tqdm(zip_list):\n",
    "\n",
    "            zfile     = zipfile.ZipFile(os.path.join(directory, zfile))\n",
    "\n",
    "            for finfo in zfile.infolist():\n",
    "\n",
    "                if 'nature' in finfo.filename:\n",
    "                    \n",
    "                    ifile = zfile.open(finfo)\n",
    "                    content = ifile.read()\n",
    "\n",
    "                    finfos.append(finfo)\n",
    "                    contents.append(zlib.compress(content))\n",
    "\n",
    "        return finfos, contents\n",
    "    \n",
    "    except Exception as e:\n",
    "        \n",
    "        print('Could not read file from zip %s, file %s' % (zfile, finfo.name))\n",
    "        \n",
    "def get_zips_parallel_reducer(arg):\n",
    "    \n",
    "    try:\n",
    "        finfo, content = arg[0], zlib.decompress(arg[1])\n",
    "        del arg\n",
    "\n",
    "        #citations = {}\n",
    "        #tex_ref_map = get_reference_text(finfo.filename, content)\n",
    "        #ref_id_ref_map = get_reference_mapping(finfo.filename, content)\n",
    "        #citations[finfo.filename] = {'tex_ref_map': tex_ref_map,\n",
    "        #                            'ref_id_ref_map': ref_id_ref_map}\n",
    "        \n",
    "        get_reference_text(finfo.filename, content)\n",
    "        get_reference_mapping(finfo.filename, content)\n",
    "        \n",
    "        return True\n",
    "    \n",
    "    except Exception as e:\n",
    "        \n",
    "        #print('Could not extract references from %s' % finfo)\n",
    "        return str(Exception)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.5/dist-packages/ipykernel_launcher.py:16: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0\n",
      "Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`\n",
      "  app.launch_new_instance()\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "9672f45a20dd4b17b7441463530d89ff",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=0, max=7727), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "##TEST on one\n",
    "DEBUG = False\n",
    "directory     = 'data/nature/raw xml/'\n",
    "finfos, contents = get_zips_parallel_mapper(directory)\n",
    "get_zips_parallel_reducer((finfos[2], contents[2]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 0 ns, sys: 0 ns, total: 0 ns\n",
      "Wall time: 7.39 µs\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.5/dist-packages/ipykernel_launcher.py:16: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0\n",
      "Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`\n",
      "  app.launch_new_instance()\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "be9bb0c5b5694e339c245d7c28832bee",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=0, max=7727), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "0.999939 of all processed files failed\n"
     ]
    }
   ],
   "source": [
    "%time\n",
    "finfos, contents = get_zips_parallel_mapper(directory)\n",
    "p = Pool(6)\n",
    "success = p.map(get_zips_parallel_reducer, zip(finfos, contents))\n",
    "error_rate = sum([1 if i == True else 0 for i in success])/len(success)\n",
    "print ('%f of all processed files Succeeded' % error_rate)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4320"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum([1 if i == True else 0 for i in success])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('data/intermediate/paper_citations.pkl', 'wb') as f:\n",
    "    pkl.dump(papers, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 3: Grouping References Together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def build_ref_index(tex_ref_mappings, reference):\n",
    "\n",
    "    inverse_tex_ref = {}\n",
    "    \n",
    "    for tex_refs in tex_ref_mappings:\n",
    "        \n",
    "        text, refs = list(tex_refs.items())[0]\n",
    "        \n",
    "        for ref in refs: \n",
    "            if ref in inverse_tex_ref:\n",
    "                inverse_tex_ref[ref].append(text)\n",
    "            else:\n",
    "                inverse_tex_ref[ref] = [text]\n",
    "    \n",
    "    return inverse_tex_ref\n",
    "    \n",
    "def read_files():\n",
    "    \n",
    "    directory = 'data/intermediate/'\n",
    "    tex_ref_mappings = []\n",
    "    references       = []\n",
    "    \n",
    "    tex_ref_files = os.listdir(os.path.join(directory, 'tex_ref_mappings'))\n",
    "    references_files = os.listdir(os.path.join(directory, 'references'))\n",
    "    \n",
    "    if len(tex_ref_files) != len(references_files):\n",
    "        print('Not all text files have a references counterpart.\\n'\n",
    "              'Continuing with the files that do have a mapping')\n",
    "    \n",
    "    #file = tex_ref_files[1]\n",
    "    \n",
    "    for file in tqdm(tex_ref_files):\n",
    "        #print(file)\n",
    "        #print(tex_ref_mappings)\n",
    "        \n",
    "        with open(os.path.join(directory, 'tex_ref_mappings', file), 'r', encoding = 'UTF-8') as f0:\n",
    "\n",
    "            try:\n",
    "                with open(os.path.join(directory, 'references', file), 'r',  encoding = 'UTF-8') as f1:\n",
    "\n",
    "                    tex_ref_mappings.append(json.loads(str(f0.read()),  encoding = 'UTF-8'))\n",
    "                    references.append(json.loads(str(f1.read()),  encoding = 'UTF-8'))\n",
    "\n",
    "            except IOError as e:\n",
    "                print('References file %f not found.' % file)\n",
    "    \n",
    "    return tex_ref_mappings, references\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.5/dist-packages/ipykernel_launcher.py:32: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0\n",
      "Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "3bed3d099c1a4e14834f5b5d414a25fd",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=0, max=16393), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "tex_ref_mappings, references = read_files()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'fnm': \"['V.']\",\n",
       " 'journal': 'Nature',\n",
       " 'snm': \"['Courtillot']\",\n",
       " 'title': 'Deccan flood basalts on Cretaceous/Tertiary boundary',\n",
       " 'year': '1988'}"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "references[1]['references']['b3']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.5/dist-packages/ipykernel_launcher.py:5: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0\n",
      "Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`\n",
      "  \"\"\"\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "5f3d7f49aad440faa847b34668ada336",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=0, max=16393), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "references_w_text = onur_json(tex_ref_mappings, references)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "nature_references_info = pd.read_hdf('data/nature/preprocessed/NatureReferencesInformation.hdf').set_index('DOI')\n",
    "nature2references      = pd.read_hdf('data/nature/preprocessed/Nature2References.hdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CitingArticleID</th>\n",
       "      <th>CitedArticleID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pub.1007899685</td>\n",
       "      <td>pub.1050061940</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pub.1007899685</td>\n",
       "      <td>pub.1033662532</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pub.1007899685</td>\n",
       "      <td>pub.1055023195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pub.1007899685</td>\n",
       "      <td>pub.1024775385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pub.1026639700</td>\n",
       "      <td>pub.1040537791</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  CitingArticleID  CitedArticleID\n",
       "0  pub.1007899685  pub.1050061940\n",
       "1  pub.1007899685  pub.1033662532\n",
       "2  pub.1007899685  pub.1055023195\n",
       "3  pub.1007899685  pub.1024775385\n",
       "4  pub.1026639700  pub.1040537791"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nature2references.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ArticleID</th>\n",
       "      <th>PubYear</th>\n",
       "      <th>Title</th>\n",
       "      <th>Volume</th>\n",
       "      <th>Issue</th>\n",
       "      <th>Cites</th>\n",
       "      <th>Journal</th>\n",
       "      <th>NatureID</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DOI</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10.1038/166444b0</th>\n",
       "      <td>pub.1007899685</td>\n",
       "      <td>1950</td>\n",
       "      <td>Detection of Sugars on Paper Chromatograms</td>\n",
       "      <td>166</td>\n",
       "      <td>4219</td>\n",
       "      <td>3143</td>\n",
       "      <td>Nature</td>\n",
       "      <td>166444b0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10.1038/166117b0</th>\n",
       "      <td>pub.1026639700</td>\n",
       "      <td>1950</td>\n",
       "      <td>The Soil-Plate Method for Isolation of Fungi f...</td>\n",
       "      <td>166</td>\n",
       "      <td>4211</td>\n",
       "      <td>394</td>\n",
       "      <td>Nature</td>\n",
       "      <td>166117b0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10.1038/165514a0</th>\n",
       "      <td>pub.1039717355</td>\n",
       "      <td>1950</td>\n",
       "      <td>Reversible Dilation and Contraction by Changin...</td>\n",
       "      <td>165</td>\n",
       "      <td>4196</td>\n",
       "      <td>359</td>\n",
       "      <td>Nature</td>\n",
       "      <td>165514a0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10.1038/166780a0</th>\n",
       "      <td>pub.1011096023</td>\n",
       "      <td>1950</td>\n",
       "      <td>Cell Specificity of Histones</td>\n",
       "      <td>166</td>\n",
       "      <td>4227</td>\n",
       "      <td>280</td>\n",
       "      <td>Nature</td>\n",
       "      <td>166780a0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10.1038/166288a0</th>\n",
       "      <td>pub.1000041177</td>\n",
       "      <td>1950</td>\n",
       "      <td>Synthesis of Peptides in Enzymic Reactions inv...</td>\n",
       "      <td>166</td>\n",
       "      <td>4216</td>\n",
       "      <td>259</td>\n",
       "      <td>Nature</td>\n",
       "      <td>166288a0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       ArticleID PubYear  \\\n",
       "DOI                                        \n",
       "10.1038/166444b0  pub.1007899685    1950   \n",
       "10.1038/166117b0  pub.1026639700    1950   \n",
       "10.1038/165514a0  pub.1039717355    1950   \n",
       "10.1038/166780a0  pub.1011096023    1950   \n",
       "10.1038/166288a0  pub.1000041177    1950   \n",
       "\n",
       "                                                              Title Volume  \\\n",
       "DOI                                                                          \n",
       "10.1038/166444b0         Detection of Sugars on Paper Chromatograms    166   \n",
       "10.1038/166117b0  The Soil-Plate Method for Isolation of Fungi f...    166   \n",
       "10.1038/165514a0  Reversible Dilation and Contraction by Changin...    165   \n",
       "10.1038/166780a0                       Cell Specificity of Histones    166   \n",
       "10.1038/166288a0  Synthesis of Peptides in Enzymic Reactions inv...    166   \n",
       "\n",
       "                 Issue Cites Journal  NatureID  \n",
       "DOI                                             \n",
       "10.1038/166444b0  4219  3143  Nature  166444b0  \n",
       "10.1038/166117b0  4211   394  Nature  166117b0  \n",
       "10.1038/165514a0  4196   359  Nature  165514a0  \n",
       "10.1038/166780a0  4227   280  Nature  166780a0  \n",
       "10.1038/166288a0  4216   259  Nature  166288a0  "
      ]
     },
     "execution_count": 212,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nature_references_info.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'10.1038/nature20106'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "references[0]['metadata']['doi']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'pub.1027267416'"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nature_references_info[nature_references_info.index == references[0]['metadata']['doi']]['ArticleID'].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CitingArticleID</th>\n",
       "      <th>CitedArticleID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6737792</th>\n",
       "      <td>pub.1027267416</td>\n",
       "      <td>pub.1030609599</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        CitingArticleID  CitedArticleID\n",
       "6737792  pub.1027267416  pub.1030609599"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nature2references[nature2references['CitingArticleID'] == nature_references_info[nature_references_info.index == references[0]['metadata']['doi']]['ArticleID'].values[0]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_reference_map_0(nature2references, nature_references_info, doi):\n",
    "    \n",
    "    try:\n",
    "        article_id = nature_references_info[nature_references_info.index == doi]['ArticleID'].values[0]\n",
    "    \n",
    "        references = nature2references[nature2references['CitingArticleID'] == article_id]\n",
    "    \n",
    "        references = pd.merge(references, nature_references_info,\n",
    "                              how = 'left', left_on = 'CitedArticleID', right_on = 'ArticleID')\n",
    "    \n",
    "    except Exception as e:\n",
    "        \n",
    "        print('doi: %s' % doi)\n",
    "        print(e)\n",
    "        references = pd.DataFrame()\n",
    "    \n",
    "    return references"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_reference_map_1(get_reference_map_1_input): \n",
    "    \n",
    "    \"\"\"\n",
    "    NOTE: Improvements Todo: Force 1-1 mapping, add year and journal to increase map accuracy\n",
    "    \"\"\"\n",
    "    \n",
    "    reference, tex_ref_mapping = get_reference_map_1_input\n",
    "    \n",
    "    if 'doi' in reference['metadata']:\n",
    "        doi = reference['metadata']['doi']\n",
    "\n",
    "        references_0 = get_reference_map_0(nature2references, nature_references_info, doi).fillna('')\n",
    "\n",
    "        references_1 = pd.DataFrame.from_dict(reference['references'], orient = 'index').fillna('')\n",
    "\n",
    "        if (len(references_1) > 0) and (len(references_0) > 0):\n",
    "\n",
    "            references_1['text'] = references_1.index.map(build_ref_index(tex_ref_mapping, reference))\n",
    "\n",
    "            vect = TfidfVectorizer(min_df=1, stop_words=\"english\")\n",
    "            tfidf = vect.fit(list(references_0['Title']) + list(references_1['title']))\n",
    "\n",
    "            vect_0 = vect.transform(list(references_0['Title']))\n",
    "            vect_1 = vect.transform(list(references_1['title']))\n",
    "\n",
    "            pairwise_similarity = vect_0 * vect_1.T \n",
    "\n",
    "            references_0['match_index'] = [i[0] for i in np.argmax(pairwise_similarity, axis = 1).tolist()]\n",
    "\n",
    "            test_match = pd.merge(references_0,\n",
    "                                  references_1.reset_index(),\n",
    "                                  left_on = 'match_index',\n",
    "                                  right_index = True, how = 'left').rename(columns = {'Title': 'title0',\n",
    "                                                                                     'title': 'title1'})\n",
    "\n",
    "            return test_match\n",
    "\n",
    "    else:\n",
    "\n",
    "        return pd.DataFrame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [],
   "source": [
    "matched_df = get_reference_map_1((references[1], tex_ref_mappings[1]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_columns', None)  \n",
    "pd.set_option('display.expand_frame_repr', False)\n",
    "pd.set_option('max_colwidth', -1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ArticleID</th>\n",
       "      <th>title0</th>\n",
       "      <th>title1</th>\n",
       "      <th>text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pub.1011515456</td>\n",
       "      <td>Kerguelen Hotspot Magma Output since 130 Ma</td>\n",
       "      <td>Kerguelen hotspot magma output since 130 Ma</td>\n",
       "      <td>[But recent geochronological data&lt;bibr rid=\"b4...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pub.1029752757</td>\n",
       "      <td>Deccan flood basalts and the Cretaceous/Tertia...</td>\n",
       "      <td>Deccan flood basalts on Cretaceous/Tertiary bo...</td>\n",
       "      <td>[The Siberian Traps at the Permian&amp;ndash;Trias...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pub.1033167081</td>\n",
       "      <td>Mineralogy of subducted basaltic crust (MORB) ...</td>\n",
       "      <td>Mineralogy of subducted basaltic crust (MORB) ...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pub.1062550794</td>\n",
       "      <td>Synchrony and Causal Relations Between Permian...</td>\n",
       "      <td>Synchrony and causal relations between Permian...</td>\n",
       "      <td>[The Siberian Traps at the Permian&amp;ndash;Trias...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pub.1000115762</td>\n",
       "      <td>Discrete alternating hotspot islands formed by...</td>\n",
       "      <td>Discrete alternating hotspot islands formed by...</td>\n",
       "      <td>[Other models, such as solitary waves travelli...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>pub.1003389384</td>\n",
       "      <td>Whole-mantle versus layered mantle convection ...</td>\n",
       "      <td>Whole-mantle versus layered mantle convection ...</td>\n",
       "      <td>[\\r\\n&lt;p&gt;We performed the calculations using th...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>pub.1011327470</td>\n",
       "      <td>Implications of mantle plume structure for the...</td>\n",
       "      <td>Implications of mantle plume structure for the...</td>\n",
       "      <td>[&lt;/descrip&gt;&lt;/suppobj&gt;\\r\\n&lt;/suppinfo&gt;\\r\\n&lt;doi&gt;1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>pub.1098792031</td>\n",
       "      <td>Mantle Plumes and their Record in Earth History</td>\n",
       "      <td>Excess temperature of mantle plumes: The role ...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>pub.1047915559</td>\n",
       "      <td>A comparison of methods for the modeling of th...</td>\n",
       "      <td>A comparison of methods for the modeling of th...</td>\n",
       "      <td>[The distribution of the compositionally disti...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>pub.1062549234</td>\n",
       "      <td>Double Flood Basalts and Plume Head Separation...</td>\n",
       "      <td>Double flood basalts and plume head separation...</td>\n",
       "      <td>[Plume head separation at the 660-km discontin...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>pub.1041060613</td>\n",
       "      <td>Plume formation in the D″-layer and the roughn...</td>\n",
       "      <td>Plume formation in the D″-layer and the roughn...</td>\n",
       "      <td>[Other models, such as solitary waves travelli...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>pub.1015876480</td>\n",
       "      <td>Mantle plumes and flood basalts: Enhanced melt...</td>\n",
       "      <td>Mantle plumes and flood basalts: Enhanced melt...</td>\n",
       "      <td>[The presence of the eclogite in the plume sou...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>pub.1017008712</td>\n",
       "      <td>3-D, 40Ar39Ar geochronology in the Paraná con...</td>\n",
       "      <td>Implications of mantle plume structure for the...</td>\n",
       "      <td>[&lt;/descrip&gt;&lt;/suppobj&gt;\\r\\n&lt;/suppinfo&gt;\\r\\n&lt;doi&gt;1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>pub.1046700962</td>\n",
       "      <td>Excess temperature of mantle plumes: The role ...</td>\n",
       "      <td>Excess temperature of mantle plumes: The role ...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>pub.1021749846</td>\n",
       "      <td>Evidence from episodic seamount volcanism for ...</td>\n",
       "      <td>Evidence from episodic seamount volcanism for ...</td>\n",
       "      <td>[But recent geochronological data&lt;bibr rid=\"b4...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>pub.1062567990</td>\n",
       "      <td>Rutile-Bearing Refractory Eclogites: Missing L...</td>\n",
       "      <td>Rutile-bearing refractory eclogites: missing l...</td>\n",
       "      <td>[Studies of the eclogites&lt;bibr rid=\"b19\"/&gt; and...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>pub.1016983940</td>\n",
       "      <td>Geochemical Study of Ultramafic Volcanic and P...</td>\n",
       "      <td>Geochemical study of ultramafic volcanic and p...</td>\n",
       "      <td>[But recent geochronological data&lt;bibr rid=\"b4...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>pub.1031113292</td>\n",
       "      <td>Evolution of starting mantle plumes: a compari...</td>\n",
       "      <td>Evolution of starting mantle plumes: a compari...</td>\n",
       "      <td>[Plume head separation at the 660-km discontin...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>pub.1042181131</td>\n",
       "      <td>Rheological control of oceanic crust separatio...</td>\n",
       "      <td>Rheological control of oceanic crust separatio...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>pub.1021008680</td>\n",
       "      <td>Phase transformations in subducted oceanic cru...</td>\n",
       "      <td>Phase transformations in subducted oceanic cru...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>pub.1039580896</td>\n",
       "      <td>Evidence for deep mantle circulation from glob...</td>\n",
       "      <td>Evidence for deep mantle circulation from glob...</td>\n",
       "      <td>[Seismic studies strongly suggest that many sl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>pub.1038343477</td>\n",
       "      <td>Evolution of the Walvis Ridge‐Rio Grande Rise ...</td>\n",
       "      <td>Evolution of the Walvis Ridge-Rio Grande Rise ...</td>\n",
       "      <td>[But recent geochronological data&lt;bibr rid=\"b4...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         ArticleID                                             title0  \\\n",
       "0   pub.1011515456        Kerguelen Hotspot Magma Output since 130 Ma   \n",
       "1   pub.1029752757  Deccan flood basalts and the Cretaceous/Tertia...   \n",
       "2   pub.1033167081  Mineralogy of subducted basaltic crust (MORB) ...   \n",
       "3   pub.1062550794  Synchrony and Causal Relations Between Permian...   \n",
       "4   pub.1000115762  Discrete alternating hotspot islands formed by...   \n",
       "5   pub.1003389384  Whole-mantle versus layered mantle convection ...   \n",
       "6   pub.1011327470  Implications of mantle plume structure for the...   \n",
       "7   pub.1098792031    Mantle Plumes and their Record in Earth History   \n",
       "8   pub.1047915559  A comparison of methods for the modeling of th...   \n",
       "9   pub.1062549234  Double Flood Basalts and Plume Head Separation...   \n",
       "10  pub.1041060613  Plume formation in the D″-layer and the roughn...   \n",
       "11  pub.1015876480  Mantle plumes and flood basalts: Enhanced melt...   \n",
       "12  pub.1017008712  3-D, 40Ar39Ar geochronology in the Paraná con...   \n",
       "13  pub.1046700962  Excess temperature of mantle plumes: The role ...   \n",
       "14  pub.1021749846  Evidence from episodic seamount volcanism for ...   \n",
       "15  pub.1062567990  Rutile-Bearing Refractory Eclogites: Missing L...   \n",
       "16  pub.1016983940  Geochemical Study of Ultramafic Volcanic and P...   \n",
       "17  pub.1031113292  Evolution of starting mantle plumes: a compari...   \n",
       "18  pub.1042181131  Rheological control of oceanic crust separatio...   \n",
       "19  pub.1021008680  Phase transformations in subducted oceanic cru...   \n",
       "20  pub.1039580896  Evidence for deep mantle circulation from glob...   \n",
       "21  pub.1038343477  Evolution of the Walvis Ridge‐Rio Grande Rise ...   \n",
       "\n",
       "                                               title1  \\\n",
       "0         Kerguelen hotspot magma output since 130 Ma   \n",
       "1   Deccan flood basalts on Cretaceous/Tertiary bo...   \n",
       "2   Mineralogy of subducted basaltic crust (MORB) ...   \n",
       "3   Synchrony and causal relations between Permian...   \n",
       "4   Discrete alternating hotspot islands formed by...   \n",
       "5   Whole-mantle versus layered mantle convection ...   \n",
       "6   Implications of mantle plume structure for the...   \n",
       "7   Excess temperature of mantle plumes: The role ...   \n",
       "8   A comparison of methods for the modeling of th...   \n",
       "9   Double flood basalts and plume head separation...   \n",
       "10  Plume formation in the D″-layer and the roughn...   \n",
       "11  Mantle plumes and flood basalts: Enhanced melt...   \n",
       "12  Implications of mantle plume structure for the...   \n",
       "13  Excess temperature of mantle plumes: The role ...   \n",
       "14  Evidence from episodic seamount volcanism for ...   \n",
       "15  Rutile-bearing refractory eclogites: missing l...   \n",
       "16  Geochemical study of ultramafic volcanic and p...   \n",
       "17  Evolution of starting mantle plumes: a compari...   \n",
       "18  Rheological control of oceanic crust separatio...   \n",
       "19  Phase transformations in subducted oceanic cru...   \n",
       "20  Evidence for deep mantle circulation from glob...   \n",
       "21  Evolution of the Walvis Ridge-Rio Grande Rise ...   \n",
       "\n",
       "                                                 text  \n",
       "0   [But recent geochronological data<bibr rid=\"b4...  \n",
       "1   [The Siberian Traps at the Permian&ndash;Trias...  \n",
       "2                                                 NaN  \n",
       "3   [The Siberian Traps at the Permian&ndash;Trias...  \n",
       "4   [Other models, such as solitary waves travelli...  \n",
       "5   [\\r\\n<p>We performed the calculations using th...  \n",
       "6   [</descrip></suppobj>\\r\\n</suppinfo>\\r\\n<doi>1...  \n",
       "7                                                 NaN  \n",
       "8   [The distribution of the compositionally disti...  \n",
       "9   [Plume head separation at the 660-km discontin...  \n",
       "10  [Other models, such as solitary waves travelli...  \n",
       "11  [The presence of the eclogite in the plume sou...  \n",
       "12  [</descrip></suppobj>\\r\\n</suppinfo>\\r\\n<doi>1...  \n",
       "13                                                NaN  \n",
       "14  [But recent geochronological data<bibr rid=\"b4...  \n",
       "15  [Studies of the eclogites<bibr rid=\"b19\"/> and...  \n",
       "16  [But recent geochronological data<bibr rid=\"b4...  \n",
       "17  [Plume head separation at the 660-km discontin...  \n",
       "18                                                NaN  \n",
       "19                                                NaN  \n",
       "20  [Seismic studies strongly suggest that many sl...  \n",
       "21  [But recent geochronological data<bibr rid=\"b4...  "
      ]
     },
     "execution_count": 166,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matched_df[['ArticleID', 'title0', 'title1', 'text']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "matched_df = pd.DataFrame()\n",
    "\n",
    "for i in tqdm(range(500)):\n",
    "    \n",
    "    try:\n",
    "        matched_df = matched_df.append(get_reference_map_1((references[i], tex_ref_mappings[i])))\n",
    "    except Exception as e:\n",
    "        pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "doi: 10.1038/nature11182b\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature11182b\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature.2017.21620\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature22808\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature01206\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature11870_techniques\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature21729\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature11847_techniques\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature11942_techniques\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature11893_techniques\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature11977_techniques\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature21898\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature11047a\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature14522\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature11363d\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature11363c\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature07825\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature11898_techniques\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature16134\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature10829aa\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature_techniques12367\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature10829b\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature18281\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature11363\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature08547\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature16470\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature11581a\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature11182b\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature.2017.21462\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature.2017.21339\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature11717a\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature11182b\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature17890\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature11192h\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature11192a\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature.2016.20649\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature11192a\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "CPU times: user 26.8 s, sys: 5.58 s, total: 32.4 s\n",
      "Wall time: 3h 1min 16s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "mp_input = zip(references, tex_ref_mappings)\n",
    "p        = Pool(8)\n",
    "\n",
    "matched_dfs = p.map(get_reference_map_1, mp_input)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "16393"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(matched_dfs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('data/intermediate/matched_dfs_raw.pkl', 'wb') as f:\n",
    "    pkl.dump(matched_dfs, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('data/intermediate/matched_dfs_raw.pkl', 'rb') as f:\n",
    "    matched_dfs = pkl.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [],
   "source": [
    "matched_dfs = pd.concat(matched_dfs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "matched_dfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "matched_dfs = pd.DataFrame(matched_dfs.groupby('ArticleID')\\\n",
    "                                 .apply(lambda x: list(x['text'].values))).rename(columns = {0: 'text'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CitingArticleID</th>\n",
       "      <th>CitedArticleID</th>\n",
       "      <th>ArticleID</th>\n",
       "      <th>PubYear</th>\n",
       "      <th>title0</th>\n",
       "      <th>Volume</th>\n",
       "      <th>Issue</th>\n",
       "      <th>Cites</th>\n",
       "      <th>Journal</th>\n",
       "      <th>NatureID</th>\n",
       "      <th>match_index</th>\n",
       "      <th>index</th>\n",
       "      <th>title1</th>\n",
       "      <th>journal</th>\n",
       "      <th>fnm</th>\n",
       "      <th>year</th>\n",
       "      <th>snm</th>\n",
       "      <th>text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pub.1003700634</td>\n",
       "      <td>pub.1011515456</td>\n",
       "      <td>pub.1011515456</td>\n",
       "      <td>2002</td>\n",
       "      <td>Kerguelen Hotspot Magma Output since 130 Ma</td>\n",
       "      <td>43</td>\n",
       "      <td>7</td>\n",
       "      <td>224</td>\n",
       "      <td>Journal of Petrology</td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td>b7</td>\n",
       "      <td>Kerguelen hotspot magma output since 130 Ma</td>\n",
       "      <td>J. Petrol.</td>\n",
       "      <td>['M. F.']</td>\n",
       "      <td>2002</td>\n",
       "      <td>['Coffin']</td>\n",
       "      <td>[But recent geochronological data&lt;bibr rid=\"b4...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pub.1003700634</td>\n",
       "      <td>pub.1029752757</td>\n",
       "      <td>pub.1029752757</td>\n",
       "      <td>1988</td>\n",
       "      <td>Deccan flood basalts and the Cretaceous/Tertia...</td>\n",
       "      <td>333</td>\n",
       "      <td>6176</td>\n",
       "      <td>333</td>\n",
       "      <td>Nature</td>\n",
       "      <td>333843a0</td>\n",
       "      <td>13</td>\n",
       "      <td>b3</td>\n",
       "      <td>Deccan flood basalts on Cretaceous/Tertiary bo...</td>\n",
       "      <td>Nature</td>\n",
       "      <td>['V.']</td>\n",
       "      <td>1988</td>\n",
       "      <td>['Courtillot']</td>\n",
       "      <td>[The Siberian Traps at the Permian&amp;ndash;Trias...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pub.1003700634</td>\n",
       "      <td>pub.1033167081</td>\n",
       "      <td>pub.1033167081</td>\n",
       "      <td>2001</td>\n",
       "      <td>Mineralogy of subducted basaltic crust (MORB) ...</td>\n",
       "      <td>190</td>\n",
       "      <td>1-2</td>\n",
       "      <td>185</td>\n",
       "      <td>Earth and Planetary Science Letters</td>\n",
       "      <td></td>\n",
       "      <td>5</td>\n",
       "      <td>b24</td>\n",
       "      <td>Mineralogy of subducted basaltic crust (MORB) ...</td>\n",
       "      <td>Earth Planet. Sci. Lett.</td>\n",
       "      <td>['S.', 'E.', 'T.']</td>\n",
       "      <td>2001</td>\n",
       "      <td>['Ono', 'Ito', 'Katsura']</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pub.1003700634</td>\n",
       "      <td>pub.1062550794</td>\n",
       "      <td>pub.1062550794</td>\n",
       "      <td>1995</td>\n",
       "      <td>Synchrony and Causal Relations Between Permian...</td>\n",
       "      <td>269</td>\n",
       "      <td>5229</td>\n",
       "      <td>425</td>\n",
       "      <td>Science</td>\n",
       "      <td></td>\n",
       "      <td>14</td>\n",
       "      <td>b2</td>\n",
       "      <td>Synchrony and causal relations between Permian...</td>\n",
       "      <td>Science</td>\n",
       "      <td>['P. R.', 'Z.', 'M. A.', 'M. T.', 'A. R.']</td>\n",
       "      <td>1995</td>\n",
       "      <td>['Renne', 'Zichao', 'Richards', 'Black', 'Basu']</td>\n",
       "      <td>[The Siberian Traps at the Permian&amp;ndash;Trias...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pub.1003700634</td>\n",
       "      <td>pub.1000115762</td>\n",
       "      <td>pub.1000115762</td>\n",
       "      <td>1999</td>\n",
       "      <td>Discrete alternating hotspot islands formed by...</td>\n",
       "      <td>397</td>\n",
       "      <td>6720</td>\n",
       "      <td>59</td>\n",
       "      <td>Nature</td>\n",
       "      <td>17584</td>\n",
       "      <td>20</td>\n",
       "      <td>b15</td>\n",
       "      <td>Discrete alternating hotspot islands formed by...</td>\n",
       "      <td>Nature</td>\n",
       "      <td>['C. F.', 'D.']</td>\n",
       "      <td>1999</td>\n",
       "      <td>['Hieronymus', 'Bercovici']</td>\n",
       "      <td>[Other models, such as solitary waves travelli...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  CitingArticleID  CitedArticleID       ArticleID PubYear  \\\n",
       "0  pub.1003700634  pub.1011515456  pub.1011515456    2002   \n",
       "1  pub.1003700634  pub.1029752757  pub.1029752757    1988   \n",
       "2  pub.1003700634  pub.1033167081  pub.1033167081    2001   \n",
       "3  pub.1003700634  pub.1062550794  pub.1062550794    1995   \n",
       "4  pub.1003700634  pub.1000115762  pub.1000115762    1999   \n",
       "\n",
       "                                              title0 Volume Issue Cites  \\\n",
       "0        Kerguelen Hotspot Magma Output since 130 Ma     43     7   224   \n",
       "1  Deccan flood basalts and the Cretaceous/Tertia...    333  6176   333   \n",
       "2  Mineralogy of subducted basaltic crust (MORB) ...    190   1-2   185   \n",
       "3  Synchrony and Causal Relations Between Permian...    269  5229   425   \n",
       "4  Discrete alternating hotspot islands formed by...    397  6720    59   \n",
       "\n",
       "                               Journal  NatureID  match_index index  \\\n",
       "0                 Journal of Petrology                      2    b7   \n",
       "1                               Nature  333843a0           13    b3   \n",
       "2  Earth and Planetary Science Letters                      5   b24   \n",
       "3                              Science                     14    b2   \n",
       "4                               Nature     17584           20   b15   \n",
       "\n",
       "                                              title1  \\\n",
       "0        Kerguelen hotspot magma output since 130 Ma   \n",
       "1  Deccan flood basalts on Cretaceous/Tertiary bo...   \n",
       "2  Mineralogy of subducted basaltic crust (MORB) ...   \n",
       "3  Synchrony and causal relations between Permian...   \n",
       "4  Discrete alternating hotspot islands formed by...   \n",
       "\n",
       "                    journal                                         fnm  year  \\\n",
       "0                J. Petrol.                                   ['M. F.']  2002   \n",
       "1                    Nature                                      ['V.']  1988   \n",
       "2  Earth Planet. Sci. Lett.                          ['S.', 'E.', 'T.']  2001   \n",
       "3                   Science  ['P. R.', 'Z.', 'M. A.', 'M. T.', 'A. R.']  1995   \n",
       "4                    Nature                             ['C. F.', 'D.']  1999   \n",
       "\n",
       "                                                snm  \\\n",
       "0                                        ['Coffin']   \n",
       "1                                    ['Courtillot']   \n",
       "2                         ['Ono', 'Ito', 'Katsura']   \n",
       "3  ['Renne', 'Zichao', 'Richards', 'Black', 'Basu']   \n",
       "4                       ['Hieronymus', 'Bercovici']   \n",
       "\n",
       "                                                text  \n",
       "0  [But recent geochronological data<bibr rid=\"b4...  \n",
       "1  [The Siberian Traps at the Permian&ndash;Trias...  \n",
       "2                                                NaN  \n",
       "3  [The Siberian Traps at the Permian&ndash;Trias...  \n",
       "4  [Other models, such as solitary waves travelli...  "
      ]
     },
     "execution_count": 158,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matched_dfs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "matched_dfs['count'] = matched_dfs.apply(lambda x: len(x['text']), axis = 1 )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[<matplotlib.axes._subplots.AxesSubplot object at 0x7fcb481d3550>]],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXoAAAEICAYAAABRSj9aAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvnQurowAAEqJJREFUeJzt3X+MXFd1wPHvISEJZNVQCFqBbdhEa6W1kgLNKCGif6wRKA7BoUWo2I1aUlmxaBtEK1qw1agKqlrSSoE2Ii1dNZFVRL24adQ6jqWUXyMaKYLEhRI7lsEkTrMuxASB23GhrcPpH/PWHRav583uzM7s3e9HWmXenTvv3T1xjm/Ou3NfZCaSpHK9aNgDkCQNlolekgpnopekwpnoJalwJnpJKpyJXpIKZ6KXpMKZ6KU+i4hjEfGWYY9DmmOil6TCmehVvIhYFxEPRMR3IuK7EfHxiHhRRNweEc9ExImI+JuIuKTqPxURs/POcWaWHhF3RMSe6jP/GRGHIqJRvfdJ4DXAgxHRiogPLvfvK81nolfRIuI8YB/wDDABrAFmgFuqn43A5cAY8PEeTn1TdZ6XAXvnPpuZvwr8G7A5M8cy80/78GtIS2KiV+muAV4N/F5mnsrMH2bmI8DNwEcz86nMbAE7gS0RcX7N8z6Smfsz8wXgk8DrBjJ6qQ9M9CrdOuCZzDw9r/3VtGf5c54BzgfGa5732x2v/wu4qIe/JKRlZaJX6Z4FXnOWJPzvwGs7jl8DnAaeA04BL517oyr/vLKHa7olrEaKiV6l+zLwLeDOiLg4Ii6KiDcBu4HfiYjLImIM+GPg09XM/+u0Z+g3RsSLgduBC3u45nO06/7SSDDRq2hVDX0zMEn7Juks8G7gPtq19S8CTwM/BN5XfeYk8JvAXwPHac/wZ+ef+xw+AtweEd+PiN/tz28iLV744BFJKpszekkqnIlekgpnopekwpnoJalwI/EFj0svvTQnJiZ6/typU6e4+OKL+z+gwhineoxTd8aonuWK04EDB57PzK7f8eh7oo+IKeAPgUPATGY2u31mYmKCxx9/vOdrNZtNpqamev7camOc6jFO3RmjepYrThHxTPdeNUs3EXFftcPfwXntmyLiSEQcjYgdVXMCLeAielt7LEkagLo1+l3Aps6G6mvh9wA3ABuArRGxAfjnzLwB+BDw4f4NVZK0GLVKN5n5xYiYmNd8DXA0M58CiIgZ4B2Z+WT1/vc4x9fGI2I7sB1gfHycZrPZ08ABWq3Woj632hineoxTd8aonlGL01Jq9Gtobxg1Zxa4NiLeCVxPe5/uBff3zsxpYBqg0WjkYupZ1gvrMU71GKfujFE9oxanvt+MzcwHgAfq9I2IzcDmycnJfg9DklRZyjr647T3+p6ztmqrLTMfzMztl1xyyRKGIUk6l6Uk+seA9dU2rxcAW2g/Uq22iNgcEdMnT55cwjAkSedSd3nlbuBR4IqImI2IbdW+3bcBDwOHgT2ZeaiXizujl6TBq7vqZusC7fuB/X0dUY8mdjx05vWxO28c4kgkaTQNda8bSzeSNHhDTfSWbiRp8JzRS1LhnNFLUuHcj16SCmeil6TCWaOXpMJZo5ekwlm6kaTCWbqRpMJZupGkwlm6kaTCmeglqXAmekkqnIlekgrnqhtJKpyrbiSpcJZuJKlwJnpJKpyJXpIKZ6KXpMKZ6CWpcC6vlKTCubxSkgpn6UaSCmeil6TCmeglqXAmekkqnIlekgpnopekwpnoJalwA0n0EXFxRDweEW8fxPklSfXVSvQRcV9EnIiIg/PaN0XEkYg4GhE7Ot76ELCnnwOVJC1O3Rn9LmBTZ0NEnAfcA9wAbAC2RsSGiHgr8CRwoo/jlCQtUmRmvY4RE8C+zLyyOr4OuCMzr6+Od1Zdx4CLaSf/HwC/lJk/Osv5tgPbAcbHx6+emZnpefCtVounT75w5viqNW6lcDatVouxsbFhD2PkGafujFE9yxWnjRs3HsjMRrd+5y/hGmuAZzuOZ4FrM/M2gIi4BXj+bEkeIDOngWmARqORU1NTPQ+g2Wxy1yOnzhwfu7n3c6wGzWaTxcR3tTFO3RmjekYtTktJ9OeUmbu69YmIzcDmycnJQQ1Dkla9pay6OQ6s6zheW7XV5u6VkjR4S0n0jwHrI+KyiLgA2ALs7eUE7kcvSYNXd3nlbuBR4IqImI2IbZl5GrgNeBg4DOzJzEO9XNwZvSQNXq0afWZuXaB9P7B/sRe3Ri9Jg+cTpiSpcO51I0mF8+HgklQ4SzeSVDhLN5JUOEs3klQ4SzeSVDhLN5JUOBO9JBXOGr0kFc4avSQVztKNJBVuYA8eGYaJHQ+deX3szhuHOBJJGh3O6CWpcN6MlaTCeTNWkgpn6UaSCmeil6TCmeglqXAmekkqnIlekgrn8kpJKpzLKyWpcJZuJKlwJnpJKpyJXpIKZ6KXpMKZ6CWpcCZ6SSqciV6SCtf3RB8RPxsRn4iI+yPiN/p9fklSb2ol+oi4LyJORMTBee2bIuJIRByNiB0AmXk4M98L/DLwpv4PWZLUi7oz+l3Aps6GiDgPuAe4AdgAbI2IDdV7NwEPAfv7NlJJ0qLUejh4Zn4xIibmNV8DHM3MpwAiYgZ4B/BkZu4F9kbEQ8Df9m+49fmgcElqq5XoF7AGeLbjeBa4NiKmgHcCF3KOGX1EbAe2A4yPj9NsNnseQKvV4gNXvdC132LOXZJWq7XqY1CHcerOGNUzanFaSqI/q8xsAs0a/aaBaYBGo5FTU1M9X6vZbHLXI6e69jt2c+/nLkmz2WQx8V1tjFN3xqieUYvTUlbdHAfWdRyvrdpqc5tiSRq8pST6x4D1EXFZRFwAbAH29nICtymWpMGru7xyN/AocEVEzEbEtsw8DdwGPAwcBvZk5qFeLu6MXpIGr+6qm60LtO9nCUsoM/NB4MFGo3HrYs8hSTo3HyUoSYXzUYKSVDg3NZOkwlm6kaTCWbqRpMJZupGkwlm6kaTCWbqRpMJZupGkwpnoJalw1uglqXDW6CWpcH1/8Mgo8rGCklYza/SSVDgTvSQVzpuxklQ4b8ZKUuEs3UhS4Uz0klQ4E70kFc5EL0mFM9FLUuFcXilJhRvqFgiZ+SDwYKPRuHW5rul2CJJWG0s3klQ4E70kFc5EL0mFM9FLUuFM9JJUOBO9JBXORC9JhRvIOvqI+EXgRuCngHsz858GcR1JUne1Z/QRcV9EnIiIg/PaN0XEkYg4GhE7ADLzHzLzVuC9wLv7O2RJUi96Kd3sAjZ1NkTEecA9wA3ABmBrRGzo6HJ79b4kaUgiM+t3jpgA9mXmldXxdcAdmXl9dbyz6npn9fOZzPzsAufaDmwHGB8fv3pmZqbnwbdaLZ4++ULPn5tz1ZrV8WSrVqvF2NjYsIcx8oxTd8aonuWK08aNGw9kZqNbv6XW6NcAz3YczwLXAu8D3gJcEhGTmfmJ+R/MzGlgGqDRaOTU1FTPF282m9z1yKlFDLvt2M29X3MlajabLCa+q41x6s4Y1TNqcRrIzdjMvBu4u1u/iNgMbJ6cnBzEMLrq3OAM3ORMUpmWurzyOLCu43ht1VaLDweXpMFbaqJ/DFgfEZdFxAXAFmBv3Q+7H70kDV4vyyt3A48CV0TEbERsy8zTwG3Aw8BhYE9mHqp7Tmf0kjR4tWv0mbl1gfb9wP6+jUiS1Fc+SlCSCjfURG/pRpIGz03NJKlwlm4kqXCWbiSpcAP5ZuxK1flNWb8lK6kUlm4kqXCWbiSpcJZuFmAZR1IpXF4pSYWzRi9JhbNGL0mFs3QjSYUz0UtS4Vx1U4MrcCStZN6MlaTCeTNWkgpnjV6SCmeil6TCmeglqXAmekkqnIlekgrn8kpJKpzLKyWpcJZuJKlwJnpJKpx73fSoc9+bTu6BI2lUOaOXpMKZ6CWpcCZ6SSqcNfo+cc96SaOq7zP6iLg8Iu6NiPv7fW5JUu9qJfqIuC8iTkTEwXntmyLiSEQcjYgdAJn5VGZuG8RgJUm9qzuj3wVs6myIiPOAe4AbgA3A1ojY0NfRSZKWLDKzXseICWBfZl5ZHV8H3JGZ11fHOwEy8yPV8f2Z+a5znG87sB1gfHz86pmZmZ4H32q1ePrkCz1/bjldtWb42zu0Wi3GxsaGPYyRZ5y6M0b1LFecNm7ceCAzG936LeVm7Brg2Y7jWeDaiHgF8EfAGyJi51ziny8zp4FpgEajkVNTUz0PoNlsctcjp3r+3HI6dvPUsIdAs9lkMfFdbYxTd8aonlGLU99X3WTmd4H31ukbEZuBzZOTk/0ehiSpspRVN8eBdR3Ha6u22ty9UpIGbykz+seA9RFxGe0EvwX4lV5OsNpm9K61lzQMdZdX7gYeBa6IiNmI2JaZp4HbgIeBw8CezDzUy8Wd0UvS4NWa0Wfm1gXa9wP7F3vx1TajX4gzfUmD5BOmJKlwbmomSYXz4eCSVDhLN5JUOEs3klQ4SzeSVDhLN5JUOEs3klQ4E70kFW6oz4xdDd+M7fzW6yDO6TdpJXVjjV6SCmfpRpIKZ6KXpMKZ6CWpcN6MHZI6N2m96SqpH7wZK0mFs3QjSYUz0UtS4Uz0klQ4E70kFc5EL0mFM9FLUuFcRz9iBrEJWgn8TsHwTex4iA9cdZpbdjzkv4MVxnX0klQ4SzeSVDgTvSQVzkQvSYUz0UtS4Uz0klQ4E70kFc5EL0mF6/sXpiLiYuAvgP8Bmpn5qX5fQ5JUX60ZfUTcFxEnIuLgvPZNEXEkIo5GxI6q+Z3A/Zl5K3BTn8crSepR3dLNLmBTZ0NEnAfcA9wAbAC2RsQGYC3wbNXthf4MU5K0WJGZ9TpGTAD7MvPK6vg64I7MvL463ll1nQW+l5n7ImImM7cscL7twHaA8fHxq2dmZnoefKvV4umT/l0y56o1Z99KotVqMTY2BsATx0+etf9C7fPf6/XaC52nTv86Y6j7O9QZT2echqVfMer1s3XO88Txk4y/BJ77wdLOf65rDMJyXmtOtz9L/RrTxo0bD2Rmo1u/pdTo1/D/M3doJ/hrgbuBj0fEjcCDC304M6eBaYBGo5FTU1M9D6DZbHLXI6d6/lypjt08ddb2ZrPJXHxv6dwcrKP/Qu3z3+v12gudp07/OmOo+zvUGU9nnIalXzHq9bN1znNLtanZXU+cv6Tzn+sag7Cc15rT7c/Sco+p7zdjM/MU8Ot1+rp7pSQN3lKWVx4H1nUcr63aanP3SkkavKUk+seA9RFxWURcAGwB9vZygojYHBHTJ092rwFLkhan7vLK3cCjwBURMRsR2zLzNHAb8DBwGNiTmYd6ubgzekkavFo1+szcukD7fmD/Yi9ujV6SBs8nTElS4dzrRpIKN9RE781YSRq82t+MHeggIr4DPLOIj14KPN/n4ZTIONVjnLozRvUsV5xem5mv7NZpJBL9YkXE43W+/rvaGad6jFN3xqieUYuTNXpJKpyJXpIKt9IT/fSwB7BCGKd6jFN3xqiekYrTiq7RS5K6W+kzeklSFyZ6SSrcikz0CzyrdlU62/N8I+LlEfGZiPhG9c+frtojIu6u4va1iPj54Y18eUXEuoj4QkQ8GRGHIuL9Vbux6hARF0XElyPiX6s4fbhqvywivlTF49PVjrVExIXV8dHq/Ylhjn85RcR5EfGViNhXHY9sjFZcoj/Hs2pXq13Me54vsAP4XGauBz5XHUM7Zuurn+3AXy7TGEfBaeADmbkBeCPwW9WfG2P14/4beHNmvg54PbApIt4I/AnwscycBL4HbKv6b6P96NBJ4GNVv9Xi/bR37p0zujHKzBX1A1wHPNxxvBPYOexxDTkmE8DBjuMjwKuq168CjlSv/wrYerZ+q+0H+EfgrcbqnDF6KfAvtB8R+jxwftV+5r9B2tuUX1e9Pr/qF8Me+zLEZi3ticGbgX1AjHKMVtyMnrM/q3bNkMYyqsYz81vV628D49VrY8eZB92/AfgSxuonVCWJrwIngM8A3wS+n+1nUMCPx+JMnKr3TwKvWN4RD8WfAR8EflQdv4IRjtFKTPTqQbanEa6hrUTEGPD3wG9n5n90vmes2jLzhcx8Pe1Z6zXAzwx5SCMlIt4OnMjMA8MeS10rMdEv+Vm1q8BzEfEqgOqfJ6r2VR27iHgx7ST/qcx8oGo2VgvIzO8DX6BdhnhZRMw9qKgzFmfiVL1/CfDdZR7qcnsTcFNEHANmaJdv/pwRjtFKTPRLflbtKrAXeE/1+j2069Fz7b9WrSh5I3Cyo2xRtIgI4F7gcGZ+tOMtY9UhIl4ZES+rXr+E9n2Mw7QT/ruqbvPjNBe/dwGfr/7PqFiZuTMz12bmBO388/nMvJlRjtGwb2os8kbI24Cv064d/v6wxzPkWOwGvgX8L+264Dba9b/PAd8APgu8vOobtFcsfRN4AmgMe/zLGKdfoF2W+Rrw1ernbcbqJ+L0c8BXqjgdBP6gar8c+DJwFPg74MKq/aLq+Gj1/uXD/h2WOV5TwL5Rj5FbIEhS4VZi6UaS1AMTvSQVzkQvSYUz0UtS4Uz0klQ4E70kFc5EL0mF+z8kT6NxSthvtAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "matched_dfs[matched_dfs['count'] > 1].hist(bins = 100, log = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ArticleID</th>\n",
       "      <th>PubYear</th>\n",
       "      <th>Title</th>\n",
       "      <th>Volume</th>\n",
       "      <th>Issue</th>\n",
       "      <th>Cites</th>\n",
       "      <th>Journal</th>\n",
       "      <th>NatureID</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DOI</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10.1107/s0907444904019158</th>\n",
       "      <td>pub.1020281098</td>\n",
       "      <td>2004</td>\n",
       "      <td>Coot: model-building tools for molecular graphics</td>\n",
       "      <td>60</td>\n",
       "      <td>12</td>\n",
       "      <td>18563</td>\n",
       "      <td>Acta Crystallographica Section D, Structural B...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                ArticleID PubYear  \\\n",
       "DOI                                                 \n",
       "10.1107/s0907444904019158  pub.1020281098    2004   \n",
       "\n",
       "                                                                       Title  \\\n",
       "DOI                                                                            \n",
       "10.1107/s0907444904019158  Coot: model-building tools for molecular graphics   \n",
       "\n",
       "                          Volume Issue  Cites  \\\n",
       "DOI                                             \n",
       "10.1107/s0907444904019158     60    12  18563   \n",
       "\n",
       "                                                                     Journal  \\\n",
       "DOI                                                                            \n",
       "10.1107/s0907444904019158  Acta Crystallographica Section D, Structural B...   \n",
       "\n",
       "                          NatureID  \n",
       "DOI                                 \n",
       "10.1107/s0907444904019158     None  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nature_references_info[nature_references_info['ArticleID'] == 'pub.1020281098']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "def flatten(l):\n",
    "    flat_list = []\n",
    "    for sublist in l:\n",
    "        if type(sublist) == list:\n",
    "            for item in sublist:\n",
    "                flat_list.append(item)\n",
    "    return flat_list\n",
    "\n",
    "matched_dfs['clean_text'] = matched_dfs['text'].apply(flatten)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "318577"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(matched_dfs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "matched_dfs['clean_text'] = matched_dfs['clean_text'].apply(lambda x: [BeautifulSoup(i, \"lxml\").text for i in x])\n",
    "matched_dfs['count'] = matched_dfs['clean_text'].apply(lambda x: len(x))\n",
    "matched_dfs = matched_dfs[matched_dfs['count'] != 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "matched_dfs.to_csv('data/intermediate/matched_dfs.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>text</th>\n",
       "      <th>count</th>\n",
       "      <th>clean_text</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ArticleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>pub.1027893469</th>\n",
       "      <td>[[\\n&lt;p&gt;&amp;agr;-Synuclein is an abundant presynap...</td>\n",
       "      <td>1</td>\n",
       "      <td>[&amp;agr;-Synuclein is an abundant presynaptic pr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pub.1025563456</th>\n",
       "      <td>[[Here we show that, in both humans and &lt;i&gt;Xen...</td>\n",
       "      <td>1</td>\n",
       "      <td>[Here we show that, in both humans and Xenopus...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pub.1045348105</th>\n",
       "      <td>[[The strong increase of the soliton period is...</td>\n",
       "      <td>1</td>\n",
       "      <td>[The strong increase of the soliton period is ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pub.1045347959</th>\n",
       "      <td>[[This technique, originally suggested by Butc...</td>\n",
       "      <td>1</td>\n",
       "      <td>[This technique, originally suggested by Butch...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pub.1025564511</th>\n",
       "      <td>[[GPR68 is expressed in many tissues and has b...</td>\n",
       "      <td>1</td>\n",
       "      <td>[GPR68 is expressed in many tissues and has be...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pub.1019779527</th>\n",
       "      <td>[[Using this phase information, the structural...</td>\n",
       "      <td>365</td>\n",
       "      <td>[Using this phase information, the structural ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pub.1012753217</th>\n",
       "      <td>[[Molecular replacement was performed using th...</td>\n",
       "      <td>372</td>\n",
       "      <td>[Molecular replacement was performed using the...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pub.1020777046</th>\n",
       "      <td>[[Structure determination was performed with P...</td>\n",
       "      <td>452</td>\n",
       "      <td>[Structure determination was performed with PH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <td>[[\\r\\n&lt;sec level=\"1\"&gt;\\r\\n&lt;sectitle&gt;Methods&lt;/se...</td>\n",
       "      <td>496</td>\n",
       "      <td>[\\nMethods\\n\\n\\nFossil fuel definitions\\nA ‘Mc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pub.1020281098</th>\n",
       "      <td>[[This model was combined into a single protei...</td>\n",
       "      <td>520</td>\n",
       "      <td>[This model was combined into a single protein...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>304447 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                             text  count  \\\n",
       "ArticleID                                                                  \n",
       "pub.1027893469  [[\\n<p>&agr;-Synuclein is an abundant presynap...      1   \n",
       "pub.1025563456  [[Here we show that, in both humans and <i>Xen...      1   \n",
       "pub.1045348105  [[The strong increase of the soliton period is...      1   \n",
       "pub.1045347959  [[This technique, originally suggested by Butc...      1   \n",
       "pub.1025564511  [[GPR68 is expressed in many tissues and has b...      1   \n",
       "...                                                           ...    ...   \n",
       "pub.1019779527  [[Using this phase information, the structural...    365   \n",
       "pub.1012753217  [[Molecular replacement was performed using th...    372   \n",
       "pub.1020777046  [[Structure determination was performed with P...    452   \n",
       "                [[\\r\\n<sec level=\"1\">\\r\\n<sectitle>Methods</se...    496   \n",
       "pub.1020281098  [[This model was combined into a single protei...    520   \n",
       "\n",
       "                                                       clean_text  \n",
       "ArticleID                                                          \n",
       "pub.1027893469  [&agr;-Synuclein is an abundant presynaptic pr...  \n",
       "pub.1025563456  [Here we show that, in both humans and Xenopus...  \n",
       "pub.1045348105  [The strong increase of the soliton period is ...  \n",
       "pub.1045347959  [This technique, originally suggested by Butch...  \n",
       "pub.1025564511  [GPR68 is expressed in many tissues and has be...  \n",
       "...                                                           ...  \n",
       "pub.1019779527  [Using this phase information, the structural ...  \n",
       "pub.1012753217  [Molecular replacement was performed using the...  \n",
       "pub.1020777046  [Structure determination was performed with PH...  \n",
       "                [\\nMethods\\n\\n\\nFossil fuel definitions\\nA ‘Mc...  \n",
       "pub.1020281098  [This model was combined into a single protein...  \n",
       "\n",
       "[304447 rows x 3 columns]"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matched_dfs.sort_values('count')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "matched_dfs = pd.read_csv('data/intermediate/matched_dfs.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "304447"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(matched_dfs[matched_dfs['text'].str.contains('rid=')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA4MAAAHoCAYAAADzOzpKAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvnQurowAAIABJREFUeJzs3XlcVXXCx/HvBSFSwwXr4iBpLqnjMlohbknh1gyuqNmm6ZNjY4YZNRNWQ4VL1lPWtGe8mrTVMZdKmjRxoU3QnlF6EsuNxAUqBUmY64XLff7o1X1iBD3IOZflfN6vV6+X9/yu53wv/BK/nuXn8Hq9XgEAAAAAbCWgrgMAAAAAAPyPMggAAAAANkQZBAAAAAAbogwCAAAAgA1RBgEAAADAhiiDAAAAAGBDlEEAAAAAsCHKIAAAAADYEGUQAAAAAGyoSV0HMFt0dLQiIiLqOkYlbrdbwcHBdR0DDRzzCGZhLsEMzCOYhbkEszCX/t+RI0eUmZl5zvc1ujIYERGh1atX13WMSnJyctS9e/e6joEGjnkEszCXYAbmEczCXIJZmEv/Lz4+3tD7uEwUAAAAAGyIMggAAAAANkQZBAAAAAAbogwCAAAAgA1RBgEAAADAhiiDAAAAAGBDlEEAAAAAsCHKIAAAAADYEGUQAAAAAGyIMggAAAAANkQZBAAAAAAbogwCAAAAgA1RBgEAAADAhiiDAAAAAGBDlEEAAAAAsCHKIAAAAADYEGUQAAAAAGyIMggAAAAANkQZ9BNXmadW4wAAAABgpiZ1HcAuQoIC1SEprdrx3MVxfkwDAAAAwO44MwgAAAAANkQZBAAAAAAbogwCAAAAgA1RBgEAAADAhiiDAAAAAGBDlEEAAAAAsCHKIAAAAADYEGUQAAAAAGyIMggAAAAANkQZBAAAAAAbogwCAAAAgA1RBgEAAADAhiiDAAAAAGBDlEEAAAAAsCHKIAAAAADYEGUQAAAAAGyIMggAAAAANkQZBAAAAAAbogwCAAAAgA1RBgEAAADAhiiDAAAAAGBDlEEAAAAAsCHKIAAAAADYEGUQAAAAAGyIMggAAAAANkQZBAAAAAAbogwCAAAAgA3V6zKYmZmpm266ScnJycrMzKzrOAAAAADQaPi9DM6bN08DBgzQqFGjKm3PyMjQyJEjNXz4cC1dulSS5HA41LRpU7ndboWHh/s7KgAAAAA0Wn4vg/Hx8UpNTa20zePxKCUlRampqUpLS9O6deu0b98+XXXVVUpNTdW9996rZ555xt9RAQAAAKDR8nsZjIqKUosWLSpty87OVvv27RUZGang4GDFxcUpPT1dAQE/xwsNDVVZWZm/owIAAABAo9WkrgNIUkFBQaXLQJ1Op7Kzs7VhwwZ9+umnKi4u1s0332xoX263Wzk5OVZFPS8ul8vQ++pbbtQvLpeLOQJTMJdgBuYRzMJcglmYSzVXL8pgdUaMGKERI0bU6PcEBwere/fuFiU6P0YnZX3LjfolJyeHOQJTMJdgBuYRzMJcglmYSzVXL54m6nQ6lZ+f73tdUFAgp9NZh4kAAAAAoHGrF2WwV69eys3NVV5entxut9LS0hQbG1vXsQAAAACg0fL7ZaKJiYnKyspSYWGhhgwZooSEBE2aNEnJycmaMWOGPB6PJkyYoC5duvg7GgAAAADYht/L4JIlS6rcHhMTo5iYGD+nAQAAAAB7qheXiQIAAAAA/IsyCAAAAAA2RBkEAAAAABuiDAIAAACADVEGAQAAAMCGKIMAAAAAYEOUQQAAAACwIcogAAAAANgQZRAAAAAAbIgyCAAAAAA2RBkEAAAAABuiDAIAAACADVEGAQAAAMCGKIMAAAAAYEOUQQAAAACwIcogAAAAANgQZRAAAAAAbIgyCAAAAAA2RBkEAAAAABuiDAIAAACADVEGAQAAAMCGKIMAAAAAYEOUQQAAAACwIcogAAAAANgQZRAAAAAAbIgyCAAAAAA2dN5lcN++fVq/fr0KCgrMzAMAAAAA8IMmRt6UkpKi8vJypaSkSJI2bNigu+++Wx6PR82bN9err76q3r17WxoUAAAAAGAeQ2cGMzIydMUVV/heP/vss7rmmmv03nvvqXfv3nr++ectCwgAAAAAMJ+hMvjDDz8oIiJCkpSfn6+9e/fq9ttvV9euXTVlyhR99dVXloYEAAAAAJjLUBkMCQlRaWmpJCkrK0vNmzdXz549JUlNmzZVSUmJdQkBAAAAAKYzdM9gjx499Oabb6pt27Z66623NHDgQAUE/NwjDx8+rIsvvtjSkAAAAAAAcxk6Mzh37lzt2rVLY8eO1cGDB3XHHXf4xjZu3MjDYwAAAACggTF0ZrB3797avHmzDhw4oA4dOqh58+a+scmTJ6t9+/aWBQQAAAAAmM9QGZR+vjfwl/sEf+2aa64xMw8AAAAAwA8MLzq/e/du3XnnnYqOjtZvf/tbff3115KkJUuWKCMjw7KAAAAAAADzGSqDO3bs0OTJk3XgwAGNHj1aFRUVvjGHw6F33nnHsoAAAAAAAPMZKoNPPvmkBg8erLS0NCUlJVUa69Gjh3bv3m1JOAAAAACANQyVwd27d+vGG2+Uw+GQw+GoNNaqVSudOHHCknAAAAAAAGsYKoMXXHCBXC5XlWM//PCDLrroIlNDAQAAAACsZagMXnHFFVq2bJk8Ho9v2y9nCN99913179/fmnQAAAAAAEsYXnR+9+7dGjt2rF544QU5HA6tWbNGU6ZM0c6dOzV79myrcwIAAAAATGSoDHbr1k1vvPGGwsLC9NJLL8nr9erNN9+UJL3xxhvq2LGjpSEBAAAAAOYyvOh8jx49tGzZMp0+fVpFRUUKDQ3VhRdeaGU2AAAAAIBFDJ0ZLCsrU2lpqaSfHybjdDp9RbC0tFRlZWXWJQQAAAAAmM7QmcEHH3xQ5eXlevLJJ88YS05OVlBQkB599FHTwwEAAAAArGHozGBmZqaGDh1a5VhsbKy2bdtmaigAAAAAgLUMlcHjx4+rdevWVY61bt1aP/74o6mhAAAAAADWMlQGw8LC9O2331Y59u2336ply5amhgIAAAAAWMtQGbzmmmv0wgsvaM+ePZW2f/PNN3rppZd07bXXWhIOAAAAAGANQw+QmTNnjj7//HNNmDBBvXr1ktPpVEFBgb766iu1a9dOc+fOtTonAAAAAMBEhs4Mtm7dWu+++65mzpwpr9frO0P4pz/9Se+++2619xMCAAAAAOonw4vOh4aG6q677tJdd91lZR4AAAAAgB8YOjMIAAAAAGhcDJ8ZXLNmjdatW6djx47p9OnTlcYcDoc2btxoejgAAAAAgDUMlcHnn39ezz77rLp06aLu3bsrODjY6lwAAAAAAAsZKoOrVq3S1KlTdf/991udBwAAAADgB4buGSwsLGQtQYu5yjznNQYAAAAA58PQmcF+/frpm2++0YABA6zOY1shQYHqkJRW5Vju4jg/pwEAAADQ2Bk6M3j//fdr1apVWrt2rU6cOKGKiooz/gMAAAAANByGzgyOHDlSkjRv3rwqxx0Oh3bv3m1eKgAAAACApQyVwdmzZ8vhcFidBQAAAADgJ4bKYEJCgtU5AAAAAAB+ZOiewV8rKSnRkSNHVFZWZkWeM5SWlio+Pl6bN2/2y/EAAAAAwA4Ml8HNmzdr/PjxuuqqqzRs2DB9++23kqQHHnhAH3zwgeEDzps3TwMGDNCoUaMqbc/IyNDIkSM1fPhwLV261Lf9lVde0e9//3vD+wcAAAAAnJuhMrhx40bdcccdatWqle699155vV7fWLt27bR27VrDB4yPj1dqamqlbR6PRykpKUpNTVVaWprWrVunffv26bPPPlPnzp0VFhZmeP8AAAAAgHMzdM/gc889p/j4eC1cuFDl5eX67//+b99Yly5d9NZbbxk+YFRUlA4fPlxpW3Z2ttq3b6/IyEhJUlxcnNLT01VaWqrS0lLt379fF1xwgWJiYhQQcPb+6na7lZOTYziPP7hcrlrvo759Jvify+ViHsAUzCWYgXkEszCXYBbmUs0ZKoP79+/Xn//8Z0k646miLVq0UFFRUa1CFBQUKDw83Pfa6XQqOztbycnJkqTVq1erVatW5yyCkhQcHKzu3bvXKo/ZzJiU9e0zwf9ycnKYBzAFcwlmYB7BLMwlmIW5VHOGymDz5s1VWFhY5diRI0fUunVrU0P9p/j4eEv3DwAAAAB2Y+iewYEDB+rll19WcXGxb5vD4ZDb7dYbb7yhIUOG1CqE0+lUfn6+73VBQYGcTmet9gkAAAAAqJ6hMnj33Xfrxx9/1HXXXacHH3xQDodDS5cu1dixY5Wfn68777yzViF69eql3Nxc5eXlye12Ky0tTbGxsbXaJwAAAACgeobKYLt27bRmzRpdc801+uyzzxQYGKgdO3aoT58+WrlyZY3O4iUmJuqGG27QwYMHNWTIEK1cuVJNmjRRcnKyZsyYoT/84Q/6/e9/ry5dupz3hwIAAAAAnJ2hewYlKTw8XIsWLar1AZcsWVLl9piYGMXExNR6/wAAAACAczN0ZnDq1Knav39/lWMHDx7U1KlTTQ0FAAAAALCWoTKYlZWlkpKSKsdKSkq0fft2U0MBAAAAAKxlqAyezaFDh9S0aVMzsgAAAAAA/KTaewZXrVql1atXS/p5GYnk5GQ1a9as0ntcLpf27t2rAQMGWJsSAAAAAGCqastgQECAAgJ+PnHo9Xorvf5Fy5YtdeONN+qPf/yjtSkBAAAAAKaqtgyOHz9e48ePlyRNmTJFDz/8sDp16uS3YAAAAAAA6xhaWuL111+3OgcAAAAAwI8MrzN46tQpbd26VUePHtXp06crjTkcDs2ePdv0cAAAAAAAaxgqg19++aVmzZql4uLiKscpgwAAAADQsBgqg4sWLVJERIReffVVXX755QoODrY6FwAAAADAQobK4IEDB/T000+rZ8+eVucBAAAAAPiBoUXn27ZtK7fbbXUWAAAAAICfGCqDd955p5YuXapTp05ZnQcAAAAA4AeGLhPdvHmzjh8/rqFDh6pPnz5q0aJFpXGHw6HHHnvMkoAAAAAAAPMZfpqoJDVr1kx79+49Y9zhcJibCgAAAABgKUNlcNOmTVbnAAAAAAD4kaF7BgEAAAAAjYvhMlhaWqrly5drzpw5mjJlinJzcyVJaWlp2r9/v1X5AAAAAAAWMHSZ6LFjxzRlyhTl5+erY8eO2rt3r0pKSiRJmZmZ+vzzz7Vw4UJLgwIAAAAAzGPozODixYsVHBys9evXa/Xq1fJ6vb6xqKgo7dixw7KAAAAAAADzGSqDn3/+uRISEhQREXHGk0OdTqe+//57S8IBAAAAAKxhqAyWlZWpWbNmVY799NNPCgwMNDUUAAAAAMBahspg165dtWHDhirHMjIy1KNHD1NDAQAAAACsZegBMrfddpvmzJkjSRo1apQkad++fUpPT9eqVav0wgsvWJcQcpV5FBJU/dnXc40DAAAAwH8yVAZHjBihhx56SE8++aRWrVolSbrvvvvUrFkz/fWvf9WQIUMsDWl3IUGB6pCUVu147uI4P6YBAAAA0BgYKoOSdOONN2rs2LHauXOnjh8/rpYtW6pv375q3ry5lfkAAAAAABYwXAYlqWnTpho4cKBVWQAAAAAAfmLoATIbN270XR4qSUeOHNHkyZPVt29fzZkzx7cAPQAAAACgYTBUBl988UWdOHHC93rx4sXKz8/X5MmTtX37dj333HOWBQQAAAAAmM9QGczLy1PXrl0lSS6XS1u3blVSUpKSkpKUmJiojz/+2NKQAAAAAABzGSqDp0+fVkhIiCTpX//6lzwejwYPHixJuuyyy/T9999blxAAAAAAYDpDZTAiIkJffvmlJCk9PV09evTQRRddJEk6fvy479cAAAAAgIbB0NNEJ0+erMcff1wff/yx9uzZo4cfftg3tnPnTnXq1MmqfAAAAAAACxgqg7feeqtatWqlXbt2aerUqRo3bpxvrKSkRPHx8ZYFBAAAAACYz/A6g2PGjNGYMWPO2J6SkmJqIAAAAACA9QzdMwgAAAAAaFwogwAAAABgQ5RBAAAAALAhyiAAAAAA2BBlEAAAAABsyFAZ3Lp161nHU1NTTQkDAAAAAPAPQ2Xw9ttv16JFi1RWVlZp+w8//KDp06frqaeesiQcAAAAAMAahspgSkqKVq5cqeuvv14HDhyQJG3atEljxozRwYMH9dprr1mZEQAAAABgMkNl8Prrr9e7774rj8ejCRMm6M4779Ts2bMVFRWl9957T1FRUVbnBAAAAACYyPADZDp16qTFixfL4/Fo48aN6tmzp5YsWaIWLVpYmQ8AAAAAYAHDZXDFihW6+eab1bFjRyUmJmrv3r2aPHmyvvvuOyvzAQAAAAAsYKgMJiQk6KGHHtLEiRP1j3/8QzNnztTKlSvldrs1btw4rVq1yuqcAAAAAAATGSqDX375pV5++WU98MADCg4OliR16dJFq1atUnx8vB588EFLQwIAAAAAzNXEyJvef/99tWnT5oztwcHB+utf/6qrr77a9GAAAAAAAOsYOjNYVRH8tWuuucaMLAAAAAAAPzH8ABkAAAAAQONBGQQAAAAAG6IMNgKuMk+txgEAAADYj6EHyKB+CwkKVIektGrHcxfH+TENAAAAgIaAM4MAAAAAYEOGzwxWVFQoOztbR48eldvtPmN83LhxpgYDAAAAAFjHUBnct2+fZs+erUOHDsnr9Z4x7nA4KIMAAAAA0IAYKoOPPPKIysvL9fTTT+vyyy9XcHCw1bkAAAAAABYyVAa//vprLV68WCNGjLA6DwAAAADADww9QKZVq1YKCgqyOgsAAAAAwE8MlcFp06bpzTfflMfDenUAAAAA0BhUe5no3/72t0qvDxw4oD/84Q8aNGiQWrRoUWnM4XBozpw51iQEAAAAAJiu2jL44osvVrn9u+++O2ObVWVw//79WrZsmYqKitS/f3/ddNNNph8DAAAAAOyo2jK4Z88eSw44b948bdmyRWFhYVq3bp1ve0ZGhhYuXKiKigpNmjRJM2fOVKdOnZSSkqKKigr95S9/oQwCAAAAgEkM3TN49OhRlZWVVTlWXl6uo0ePGj5gfHy8UlNTK23zeDxKSUlRamqq0tLStG7dOu3bt0+SlJ6erpkzZyomJsbwMQAAAAAAZ2eoDA4dOlQ5OTlVju3Zs0dDhw41fMCoqKgz7jnMzs5W+/btFRkZqeDgYMXFxSk9Pd137NTUVH3wwQeGjwEAAAAAODtD6wx6vd5qx8rLyxUQYKhTVqugoEDh4eG+106nU9nZ2crMzNTHH38st9tt+Myg2+2utrjWFZfLVdcR6t3XBDXncrn4PsIUzCWYgXkEszCXYBbmUs1VWwaLi4t18uRJ3+uCggLl5eVVeo/L5dKaNWvUpk0bS8JFR0crOjq6Rr8nODhY3bt3tyTP+aoPk7K+fU1Qczk5OXwfYQrmEszAPIJZmEswC3Op5qotg8uXL9dzzz0nh8Nx1qeFer1eJSQk1CqE0+lUfn6+73VBQYGcTmet9gkAAAAAqF61ZXDYsGGKiIiQ1+vV/fffr1mzZunSSy+t9J7g4GB16tRJ3bp1q1WIXr16KTc3V3l5eXI6nUpLS9OTTz5Zq30CAAAAAKpXbRns1q2br+Q5HA7FxMSodevWtT5gYmKisrKyVFhYqCFDhighIUGTJk1ScnKyZsyYIY/HowkTJqhLly61PhYAAAAAoGqGHiAzfvx40w64ZMmSKrfHxMSwfAQAAAAA+Em1ZXDq1Kl66KGH1KlTJ02dOvWsO3E4HFq2bJnp4QAAAAAA1qi2DP56OYmzLS1hZBwAAAAAUL9UWwZff/31Kn8NAAAAAGj4ardaPAAAAACgQTL0AJlBgwb5FoCPjo5Whw4dLI4FAAAAALCSoTI4ceJEZWZmav78+fJ4PLrkkkt8xbB///6KiIiwOidqwVXmUUhQYI3HAAAAADRehsrg3XffLUkqLS3V9u3blZmZqW3btiktLU0VFRWKiIjQxo0bLQ2K8xcSFKgOSWlVjuUujvNzGgAAAAD1gaEy+IumTZsqJiZG7dq1U0REhDZs2KDMzEwdPXrUqnwAAAAAAAsYKoN5eXnatm2btm3bpszMTB0/flydOnVSdHS0br75ZkVHR1udEwAAAABgIkNlcPjw4brwwgsVFxenefPmKTo6Wm3atLE6GwAAAADAIobK4G9/+1vt2bNHH3/8sYqKinTixAkNGDBAnTt3tjofAAAAAMAChsrg6tWrVVxcrKysLG3btk3/+Mc/tGjRIrVu3Vr9+vXTgAEDdP3111udFQAAAABgEsMPkAkNDdWwYcM0bNgwSdK//vUvPfPMM/rnP/+pjz76iDIIAAAAAA2I4TKYn5/ve4BMZmamjh07poCAAPXq1YsHyAAAAABAA2OoDI4YMUJ5eXmSpG7dumnEiBGKjo5WVFSUmjdvbmlAAAAAAID5DJXBIUOGKDo6Wv369VOLFi2szgQAAAAAsJihMvjggw9anQMAAAAA4EcBdR0AAAAAAOB/lEEAAAAAsCHKIAAAAADYEGXQ5lxlnlqNAwAAAGiYDK8ziMYpJChQHZLSqh3PXRznxzQAAAAA/MVQGTx48KB++ukn9e7dW5Lkcrn0/PPPa+/evRo8eLBuueUWS0MCAAAAAMxl6DLR+fPn66OPPvK9fuqpp/T3v/9d33//vR599FG9+eablgUEAAAAAJjPUBncs2ePrrjiCklSRUWF1q5dq3vvvVerV6/WrFmztGLFCktDAgAAAADMZagM/vTTT2rZsqUkaffu3SouLtbIkSMlSf369VNeXp51CQEAAAAApjNUBtu0aaNDhw5Jkj777DNdeumlatu2rSSptLRUTZrwHBoAAAAAaEgMtbjY2FgtWbJEe/fu1erVq3XDDTf4xr799ltFRkZaFhAAAAAAYD5DZfCee+7R6dOn9emnnyo2NlZ/+tOffGObNm3SoEGDLAsIAAAAADDfOcugx+PRoUOHdPfddyssLOyM8XfeeceSYAAAAAAA65zznkGHw6EJEyYoJyfHH3kAAAAAAH5wzjIYEBCg8PBw/fvf//ZHHgAAAACAHxh6mujkyZO1bNkyud1uq/MAAAAAAPzA0ANkSkpKdOjQIQ0bNkxXX321Lr74YjkcDt+4w+HQnDlzLAsJAAAAADCXoTL48ssv+369atWqM8YpgwAAAADQsBgqg3v27LE6BwAAAADAjwzdMwgAAAAAaFwMnRmUJK/Xq02bNmnHjh0qKirSnXfeqYiICGVlZal9+/ZyOp1W5gQAAAAAmMhQGTx58qRmzpypXbt2qVmzZiotLdUtt9yiiIgI/eMf/1DLli314IMPWp0VAAAAAGASQ5eJPv744zp27JjefvttZWZmyuv1+sYGDhyoL774wrKAAAAAAADzGSqD6enpuvvuu9W3b99KS0pIUtu2bXXs2DFLwgEAAAAArGGoDJaWllZ7T6Db7a50phAAAAAAUP8ZKoOXXXaZPv300yrHsrKy1LVrV1NDAQAAAACsZagM3nTTTVq+fLlefPFFHT16VJJUXFysVatW6c0339RNN91kaUgAAAAAgLkMPU108uTJysvL07PPPqtnnnlGkvRf//VfCggI0IwZMzRmzBhLQwIAAAAAzGV4ncF7771XN954oz777DOdOHFCLVu21KBBgxQZGWllPgAAAACABQyXQUmKiIjQ9ddfb1UWAAAAAICfGC6DHo9Ha9eu1c6dO1VQUCCn06m+fftq7NixCgwMtDIjAAAAAMBkhsrgkSNHdNtttyk3N1fh4eEKCwvTt99+q5UrV+qVV15RamqqIiIirM6KOuAq8ygkqOqyf7YxAAAAAPWboTI4f/58nTp1Sm+99ZauuOIK3/Yvv/xSd911l+bPn6+XXnrJspCoOyFBgeqQlFblWO7iOD+nAQAAAGAWQ0tLbNu2Tffcc0+lIihJV155pRITE5WZmWlJOAAAAACANQyVwaZNm6p169ZVjoWFhSkkJMTUUAAAAAAAaxkqg2PGjNE777xT5dg777yjcePGmRoKAAAAAGAtQ/cMXnrppfroo480evRojRgxQmFhYTp+/LjWr1+vkpISDRkyRO+++67v/RMnTrQsMAAAAACg9gyVwZSUFElSfn6+9u7de8b4I4884vu1w+GgDAIAAABAPWeoDKanp1udAwAAAADgR4bKIGsIAgAAAEDjYugBMgAAAACAxsXQmUFJ+vTTT/X222/r4MGDOn369BnjXEoKAAAAAA2HoTODW7du1R//+Ee5XC4dOHBAHTt21G9+8xvl5+crICBA/fr1szonAAAAAMBEhsrgCy+8oJtvvllLly6VJM2dO1evv/661q1bJ4/Ho6uvvtrSkAAAAAAAcxkqgwcOHNC1116rgIAAORwOeTweSdJll12mhIQEvfjii5aGBAAAAACYy1AZDAgIUGBgoBwOh1q3bq2jR4/6xi655BIdOnTIsoAAAAAAAPMZeoDMZZddpiNHjkiSevbsqWXLlumKK65QkyZN9Oqrr1q69MTGjRu1ZcsWnTp1ShMnTtTgwYMtOxYAAAAA2IWhMjh69Gjt379fkpSQkKDp06crJiZGkhQYGKgnnniiRgedN2+etmzZorCwMK1bt863PSMjQwsXLlRFRYUmTZqkmTNnatiwYRo2bJhOnjypxx57jDIIAAAAACYwVAZvvvlm36979uypDz74QBkZGXK5XBo4cKA6d+5co4PGx8frlltu0X333efb5vF4lJKSor///e9yOp2aOHGiYmNjfft+8cUXK+UAAAAAAJw/w+sM/lp4eLiuv/768z5oVFSUDh8+XGlbdna22rdvr8jISElSXFyc0tPT1alTJz3xxBMaMmSIevTocc59u91u5eTknHc2K7hcrrqOYAlXmUchQYHVjpf8+7QO5R7wY6LGzeVy1bu5jYaJuQQzMI9gFuYSzMJcqrkalcFt27Zp586dKigokNPpVN++fRUdHW1KkIKCAoWHh/teO51OZWdn6/XXX9cXX3yhn376Sd99951uvPHGs+4nODhY3bt3NyWTWRrrpAwJClSHpLRqx3MXx9W770VDlpOTw9cTpmAuwQzMI5iFuQTiSIA1AAAgAElEQVSzMJdqzlAZLCoq0l133aWsrCw5HA6FhoaquLhYXq9X0dHRevrpp9WyZUtLAk6dOlVTp061ZN8AAAAAYFeGlpZYsGCBvvrqKz3++OPatWuXtm3bpl27dumxxx7TV199pYULF9Y6iNPpVH5+vu/1L2cfAQAAAADmM1QGN2/erHvuuUejR49WUFCQJCkoKEhjxozR3LlztWnTploH6dWrl3Jzc5WXlye32620tDTFxsbWer8AAAAAgDMZukw0MDBQ7du3r3LssssuU2Bg9Q8RqUpiYqKysrJUWFioIUOGKCEhQZMmTVJycrJmzJghj8ejCRMmqEuXLjXaLwAAAADAGENlcOjQofrwww+rXOMvLS1NQ4cOrdFBlyxZUuX2mJgY3/qFAAAAAADrGCqD1157rR599FHNnDlT1113ncLCwnT8+HH985//1L59+/TAAw/oiy++8L1/wIABlgVGw3GupSfONQ4AAADAOobK4Jw5cyRJx44dU0ZGxhnjCQkJkiSv1yuHw9Fol1JAzRhZegIAAABA3TBUBpcvX251DgAAAACAHxkqg/369bM6BwAAAADAjwwtLQEAAAAAaFwogwAAAABgQ5RBAAAAALAhyiDqjKvMc15jAAAAAGqv2gfILF++XHFxcQoLC9PRo0d18cUXKygoyJ/Z0MidbekJlp0AAAAArFXtmcFHH31UR44ckSQNHTqUtQMBAAAAoBGptgyGhobqhx9+kPTzYvIAAAAAgMaj2stE+/btq6SkJHXr1k2S9PDDD6t58+ZVvtfhcGjZsmXWJAQAAAAAmK7aM4MLFixQXFycHA6HHA6HPB6PysvLq/yvrKzMn5kBAAAAALVU7ZnBNm3a6OGHH5YkdevWTfPnz1fv3r39lQsAAAAAYKFqy+Cvpaen6+KLL7Y6CwAAAADATwyVwYiICEnS5s2blZWVpZMnT6pFixbq37+/YmJiLA0IAAAAADCfoTJ46tQp/elPf9KOHTvUpEkTtWzZUkVFRXrttdd01VVX6aWXXlKzZs2szgoAAAAAMEm1D5D5taeeekpff/21Hn/8ce3atUuffvqpdu3apccee0xff/21nnrqKatzAgAAAABMZKgMrl+/XnPnztWYMWMUGBgoSQoMDNSYMWN01113af369ZaGhP24yjy1GgcAAABwdoYuEy0qKlLnzp2rHOvcubOKiopMDQWEBAWqQ1JateO5i+P8mAYAAABofAydGWzXrp02b95c5djWrVvVrl07U0MBAAAAAKxl6MzgDTfcoMWLF6u0tFSjR4/WJZdcoh9++EEffvihVq5cqaSkJKtzAgAAAABMZKgMTps2TSdOnNDf//53rVmzRpLk9XoVFBSkmTNn6tZbb7U0JAAAAADAXIbKoCQlJibqtttu086dO33rDPbp00ctWrSwMh8AAAAAwAKGy6AktWjRgkXmAQAAAKARMPQAGQAAAABA40IZBAAAAAAbogwCAAAAgA1RBgEAAADAhgyVwe3bt6ukpKTKsZKSEm3fvt3UUAAAAAAAaxkqg1OnTtX+/furHDt48KCmTp1qaigAAAAAgLUMlUGv11vtmNvtVmBgoGmBAAAAAADWq3adwcOHDysvL8/3+quvvjrjUtHTp09r1apVatu2rXUJgSq4yjwKCar6HyHONgYAAADgZ9WWwbVr1+q5556Tw+GQw+HQggULKp0hdDgc8nq9atKkiZKTk/0SFvhFSFCgOiSlVTmWuzjOz2kAAACAhqfaMjh+/Hj169dPXq9Xt956q5KTk9W5c+dK7wkODlaHDh3UsmVLy4MCAAAAAMxTbRmMiIhQRESEJGn58uX67W9/q+bNm/stGAAAAADAOtWWwV/r16+f1TkAAAAAAH5kqAy63W4tXbpU69at07Fjx+R2uyuNOxwO7d6925KAAAAAAADzGSqDjz/+uN566y0NGTJEI0aMUHBwsNW5AAAAAAAWMlQG169fr4SEBM2aNcvqPAAAAAAAPzC06Hxpaan69OljdRYAAAAAgJ8YKoPXXnutduzYYXUWAAAAAICfGLpMdMqUKfrLX/4ih8OhmJiYKtcVjIyMND0cAAAAAMAahsrg5MmTJUnPPfecnn/++Srfk5OTY14qAAAAAIClDJXBRYsWyeFwWJ0FAAAAAOAnhspgfHy81TkAAAAAAH5k6AEyAAAAAIDGxdCZwXnz5p113OFwaNGiRaYEAgAAAABYz1AZzMzMPGPbyZMnVVJSotDQUF100UWmBwMAAAAAWMdQGdy0aVOV27dv366HHnpITzzxhKmhAAAAAADWqtU9g1FRUZo2bZrmz59vVh4AAAAAgB/U+gEykZGRrDEIAAAAAA1MrcpgeXm5Vq9eLafTaVYeAAAAAIAfGLpncOrUqWdsKysrU25uroqKivTII4+YHgwAAAAAYB1DZdDr9Z6xrVmzZho+fLji4uIUHR1tejAAAAAAgHUMlcHXX3/d6hwAAAAAAD+q9QNkAAAAAAANj6Ezg5L0zTff6Pnnn1dWVpaKi4sVGhqq6Oho3XHHHeratauVGYEacZV5FBIUeN7jAAAAgB0YKoPZ2dmaMmWKQkJCFBsbqzZt2ujHH3/Upk2btHXrVr3xxhvq2bOn1VkBQ0KCAtUhKa3a8dzFcX5MAwAAANRPhsrgkiVL1KVLF7322mtq3ry5b/upU6c0ffp0LVmyRK+++qplIQEAAAAA5jJ0z+CuXbt0++23VyqCktS8eXP98Y9/1L/+9S9LwgH1javMU6txAAAAoL4wfM/g2TgcDjN2A9R7XIIKAACAxsLQmcHf/e53eumll3Tq1KlK20tLS/XKK6+oT58+loTLy8vT/fffrzlz5liyfwAAAACwK0NnBhMTEzVlyhTFxsbqmmuu0cUXX6wff/xRW7dulcvl0vLlyw0fcN68edqyZYvCwsK0bt063/aMjAwtXLhQFRUVmjRpkmbOnKnIyEgtWrSIMggAAAAAJjN0ZrB3795asWKF+vfvr08//VSvvfaaPvnkE0VHR2vFihXq3bu34QPGx8crNTW10jaPx6OUlBSlpqYqLS1N69at0759+2r2SQAAAAAAhhm+Z7Bbt2565plnan3AqKgoHT58uNK27OxstW/fXpGRkZKkuLg4paenq3PnzrU+HgAAAADgTNWWwYqKCm3ZskXt2rXT5ZdfXuV7vvnmGx05ckSxsbG1ClFQUKDw8HDfa6fTqezsbBUWFuqpp57S7t279fLLL+v2228/577cbrdycnJqlcdsLperriPgP5xtjlzaoaOaXXiBJfuuDZfLVe/mNhom5hLMwDyCWZhLMAtzqeaqLYPvv/++HnnkEb3//vvV/uZmzZrpnnvu0fz58zVq1CjTw7Vq1UopKSk1+j3BwcHq3r276Vlqg0lZ/5xrjlT3xFAjTwu1av7l5OTUu7mNhom5BDMwj2AW5hLMwlyquWrvGXz//fcVHx/vu3SzKu3atdOECRO0Zs2aWoVwOp3Kz8/3vS4oKJDT6azVPgEAAAAA1au2DH799dcaNGjQOXcwcOBA/e///m+tQvTq1Uu5ubnKy8uT2+1WWlparS89BQAAAABUr9rLREtKShQaGnrOHYSGhqqkpMTwARMTE5WVlaXCwkINGTJECQkJmjRpkpKTkzVjxgx5PB5NmDBBXbp0MbxPAAAAAEDNVFsGW7VqpaNHj55zB8eOHVOrVq0MH3DJkiVVbo+JiVFMTIzh/QAAAAAAzl+1l4leeeWVWrt27Tl3sGbNGl155ZWmhgIAAAAAWKvaMnjrrbfqiy++0KJFi+R2u88YLysr08KFC7Vt2zZNmzbNyowA9POSF9VxlXn8mAQAAACNQbWXifbt21f33XefHnvsMX3wwQcaNGiQIiIiJElHjhzR559/rqKiIt13333q06eP3wIDdtXswgtqteQFAAAA8GvVlkFJmjZtmnr06KFXXnlFGzdu9C2eHhISon79+mnmzJm66qqr/BIUAAAAAGCes5ZBSYqKilJUVJQqKipUWFgoSWrZsqUCAwMtDwcAAAAAsMY5y+AvAgICFBYWZmUWoFFzlXkUElT9P6KcaxwAAAAwk+EyCKB2QoICq73nT+K+PwAAAPhXtU8TBQAAAAA0XpRBAAAAALAhyiAAAAAA2BBlELbDAu0AAAAAD5CBDfEgFwAAAIAzgwAAAABgS5RBAAAAALAhyiAAAAAA2BBlEKgneLANAAAA/IkHyAD1BA+2AQAAgD9xZhAAAAAAbIgyCAAAAAA2RBkEAAAAABuiDAIAAACADVEGAQAAAMCGKIMAAAAAYEOUQQAAAACwIcogAAAAANgQZRAAAAAAbIgyCAAAAAA2RBkEAAAAABuiDAIAAACADVEGAQAAAMCGKIMAAAAAYEOUQcBErjJPXUeo0tly1dfMVjrXZ66rr0l9zQUAABqnJnUdAGhMQoIC1SEprcqx3MVxfk7z/+prrrpytq+HVHdfk/qaCwAANE6cGQQAAAAAG6IMAgAAAIANUQYBAAAAwIYogwAAAABgQ5RBAAAAALAhyiAAAAAA2BBlEAAAAABsiDIIAAAAADZEGQQAAAAAG6IMAgAAAIANUQYBAAAAwIYog4DNuco8tRpvqMcGAACwuyZ1HQBA3QoJClSHpLRqx3MXxzXKYwMAANgdZwYBAAAAwIYogwAAAABgQ5RBAAAAALAhyiAAAAAA2BBlEAAAAABsiDIIAAAAADZEGQQAAAAAG6IMAgAAAIANUQYBAAAAwIYogwAAAABgQ5RBAAAAALAhyiAAAAAA2BBlEAAAAABsiDIIAAAAADZEGQQAAAAAG6IMAgAAAIANUQYBAAAAwIaa1HWAsyktLdUjjzyioKAg9evXT2PGjKnrSAAAAADQKPj9zOC8efM0YMAAjRo1qtL2jIwMjRw5UsOHD9fSpUslSRs2bNDIkSO1YMECbdq0yd9RAQAAAKDR8nsZjI+PV2pqaqVtHo9HKSkpSk1NVVpamtatW6d9+/apoKBAbdu2lSQFBgb6OyoAAAAANFp+v0w0KipKhw8frrQtOztb7du3V2RkpCQpLi5O6enpcjqdys/PV/fu3VVRUWFo/263Wzk5Oabnrg2Xy1XXEWADZ5v33bt3t2zfZ3Nph45qduEFDfLYVh235N+ndSj3QJVj5/o+uco8Cgmq/h/G/u0u14XBVf+x/uvjulyuGn2+2nwm1C9n+17W9Pv463lU2zliZq7G4Fxfz7P9v362Mal+fj1r+meSHfDn7vmxYi419u9FvbhnsKCgQOHh4b7XTqdT2dnZmjJliubPn68tW7bo2muvNbSv4ODgWv/F12z8AQd/sHLe12bfHZLSqh3LXRxXJ8euq+P+cuzz3XdIUOA59322z/zLcXNycmqcwarPBP8zMkeM+M95VNs5YlauxqI2/683tP9fz+fPJDtoaN/H+sCqudSYvxf1ogxWp2nTpnr00UfrOgYAAAAANDr1YmmJXy4H/UVBQYGcTmcdJgIAAACAxq1elMFevXopNzdXeXl5crvdSktLU2xsbF3HAgAAAIBGy++XiSYmJiorK0uFhYUaMmSIEhISNGnSJCUnJ2vGjBnyeDyaMGGCunTp4u9oAAAAAGAbfi+DS5YsqXJ7TEyMYmJi/JwGAAAAAOypXlwmCgAAAADwL8ogAAAAANgQZRAAAAAAbIgyCAAAAAA2RBkEAAAAABuiDAIAAACADVEGAQAAAMCGKIMAAAAAYEOUQQAAAACwIcogAAAAANgQZRAAAAAAbIgyCAAAAAA2RBkEAAAAABtyeL1eb12HMFN0dLQiIiLqOgYAAAAA1IkjR44oMzPznO9rdGUQAAAAAHBuXCYKAAAAADZEGQQAAAAAG6IMAgAAAIANUQYBAAAAwIYogwAAAABgQ5RBAAAAALAhyqCFMjIyNHLkSA0fPlxLly6t6zio5+bNm6cBAwZo1KhRvm1FRUWaPn26RowYoenTp+vkyZOSJK/XqwULFmj48OEaPXq0vv7667qKjXrm2LFjmjJliv7whz8oLi5Oy5Ytk8RcQs2dPn1aEydO1JgxYxQXF6dnnnlGkpSXl6dJkyZp+PDhmjt3rtxutyTJ7XZr7ty5Gj58uCZNmqTDhw/XZXzUMx6PR+PGjdPtt98uiXmE8xMbG6vRo0dr7Nixio+Pl8TPt9qiDFrE4/EoJSVFqampSktL07p167Rv3766joV6LD4+XqmpqZW2LV26VAMGDNCGDRs0YMAA3z8qZGRkKDc3Vxs2bND8+fP18MMP10Fi1EeBgYFKSkrShx9+qBUrVuitt97Svn37mEuoseDgYC1btkzvv/++1q5dq08++UQ7d+7UE088oWnTpunjjz9WaGio3n33XUnSypUrFRoaqo8//ljTpk3TE088UcefAPXJ8uXL1alTJ99r5hHO17Jly/Tee+9p9erVkvi7Um1RBi2SnZ2t9u3bKzIyUsHBwYqLi1N6enpdx0I9FhUVpRYtWlTalp6ernHjxkmSxo0bp40bN1ba7nA41KdPHxUXF+v777/3e2bUP5dccol69OghSWrevLk6duyogoIC5hJqzOFwqFmzZpKk8vJylZeXy+FwaNu2bRo5cqQkafz48b6fbZs2bdL48eMlSSNHjtQXX3whr9dbN+FRr+Tn52vLli2aOHGipJ/P2DCPYBZ+vtUOZdAiBQUFCg8P9712Op0qKCiow0RoiI4fP65LLrlEknTxxRfr+PHjks6cX+Hh4cwvnOHw4cPKycnR7373O+YSzovH49HYsWM1cOBADRw4UJGRkQoNDVWTJk0kVZ4vBQUFatu2rSSpSZMmuuiii1RYWFhn2VF/LFq0SH/+858VEPDzXzsLCwuZRzhvt912m+Lj47VixQpJ/F2ptprUdQAAxjgcDjkcjrqOgQaipKREc+bM0f3336/mzZtXGmMuwajAwEC99957Ki4u1uzZs3XgwIG6joQGZvPmzWrdurV69uypzMzMuo6DBu7tt9+W0+nU8ePHNX36dHXs2LHSOD/fao4yaBGn06n8/Hzf64KCAjmdzjpMhIYoLCxM33//vS655BJ9//33at26taQz51d+fj7zCz5lZWWaM2eORo8erREjRkhiLqF2QkNDFR0drZ07d6q4uFjl5eVq0qRJpfnidDp17NgxhYeHq7y8XD/99JNatWpVx8lR1/7nf/5HmzZtUkZGhk6fPq1Tp05p4cKFzCOcl1/mSVhYmIYPH67s7Gx+vtUSl4lapFevXsrNzVVeXp7cbrfS0tIUGxtb17HQwMTGxmrt2rWSpLVr12ro0KGVtnu9Xu3cuVMXXXSR7xIJ2JvX69UDDzygjh07avr06b7tzCXU1IkTJ1RcXCxJcrlc+vzzz9WpUydFR0dr/fr1kqQ1a9b4frbFxsZqzZo1kqT169erf//+/As9dM899ygjI0ObNm3SkiVL1L9/fz355JPMI9RYaWmpTp065fv1Z599pi5duvDzrZYcXu7KtczWrVu1aNEieTweTZgwQbNmzarrSKjHEhMTlZWVpcLCQoWFhSkhIUHDhg3T3LlzdezYMf3mN7/R008/rZYtW8rr9SolJUWffPKJLrzwQi1atEi9evWq64+AemDHjh26+eabdfnll/vuz0lMTFTv3r2ZS6iRPXv2KCkpSR6PR16vV9ddd53uvPNO5eXl6e6779bJkyfVvXt3PfHEEwoODtbp06f15z//WTk5OWrRooWeeuopRUZG1vXHQD2SmZmpV199VS+//DLzCDWWl5en2bNnS/r5fuZRo0Zp1qxZKiws5OdbLVAGAQAAAMCGuEwUAAAAAGyIMggAAAAANkQZBAAAAAAbogwCAAAAgA1RBgEAAADAhiiDAGBjq1evVteuXXXVVVfp5MmTlcbKy8vVtWtXPfvss37P9eyzz6pr164qLy/3+7FroqKiQgsXLtTgwYPVrVs33XHHHdW+98cff9SCBQs0cuRI9e7dW9HR0YqPj9eCBQvkdrsty/jaa69pw4YNlu0fANBwNanrAACAuvfTTz/plVde0b333lvXURqUjz76SMuXL1dSUpL69Omjli1bVvm+U6dO6frrr5fD4dBtt92mjh076uTJk8rJydEHH3ygOXPmKDg42JKMy5cv1xVXXKERI0ZYsn8AQMNFGQQAaPDgwXrjjTc0bdo0tWnTpq7j+IXb7a51ATtw4IAk6dZbb1VAQPUX23z00Uc6cuSI3nvvPXXr1s23feTIkbrrrrtqlQEAgPPFZaIAAM2aNUuS9OKLL571fb9cvvmfkpKSFBsb63t9+PBhde3aVW+//baefPJJDRo0SH379tW9996rf//73/ruu+902223qW/fvho+fLjWrFlT5fH279+vKVOm6He/+50GDx6sv/3tb6qoqKj0nhMnTig5OVlXX321evbsqeuuu04rVqyo9J5fLofdvn275syZo6uuukqTJk0662fNyMjQ5MmT1bt3b1155ZW64447fOVPkmJjY32X0Hbv3l1du3bV6tWrq9zXL5fgVlW0HQ6HHA5HpW0rVqzQmDFj1KtXL0VHR+v+++9XUdH/tXe3MU1dfxzAv3QWBSxCZwURH0q2iwjyOBsQrTKHgs4ZtpFgCwoSqZR1QwFBkWgEJ5morNRGFDWbTAjbzJjZMHuIm+BMtIZXY0Z0ogwZDEIVU6Eo5/+CcP9eWnQo+f9f8PskfXHPOfc89dUv5+GaBWV8fX1x+PBhfP7553jzzTcREhKCxMRENDc3C/rY1taGc+fOwdfXF76+vsjLy+Pzr1+/ji1btmDRokUIDAxEQkICTCaToJ28vDwolUo0NTVBpVIhKCgIK1euRFVVlc1YWltbkZOTg8jISAQEBGDFihUoKioSlLly5Qo2btyIkJAQBAcHIzU1FTdu3BCUqa+vR0JCAsLCwhASEoJVq1bBYDDYnVtCCCEvjlYGCSGEQCaTQa1W47PPPsOmTZswa9ascan32LFjUCgUKC4uxq1bt3DgwAGIRCL88ccfiI+Px6ZNm1BVVYUdO3YgICAAr7/+uuD9jIwMvPfee9BoNGhoaIDRaIRIJIJOpwMwtP1y/fr16O/vh06ng7e3N+rr67Fnzx5YrVYkJSUJ6svOzsaaNWug1+ufeR7x4sWL0Gg0CA8Px+HDh2GxWKDX66FSqVBbWwsPDw8YDAacPn0aZ8+e5YPPOXPm2K0vMDAQALB161akpaUhLCwMzs7OdsuWlJTg1KlTSEpKwvbt29HR0YHS0lI0Nzejuroar7zyCl/23LlzkMvlyM/Px8DAAD755BNotVrU1dVh0qRJMBgMSEtLg6+vLz9nUqkUAPD7779DrVbDz88PhYWFcHJyQlVVFZKTk1FdXY2AgAC+nYcPHyIrKwsbN25ERkYGzp49iz179kAulyM8PBzAUCAYHx8PJycnfPjhh5g7dy7a29vR0NDA1/PLL79Aq9Vi2bJlOHDgAACgoqICarUa3377LWbOnInW1lakp6dj1apV0Gq1EIvFuHPnDlpbW0f9vwghhLwgRgghZML6+uuvGcdxrKWlhfX09LCwsDCWl5fHGGNsYGCAcRzH9Ho9X16v1zOO42zqyc3NZVFRUfxza2sr4ziOJSUlCcplZGQwjuPYN998w6eZzWbm5+fHysrKbNopLy8XvJ+fn8+Cg4PZ/fv3GWOMGQwGFhAQwG7fvm1TTqFQsIGBAcE49+3b96/mJS4ujkVHR/PvM8bY3bt32YIFC9jHH3/Mpx06dMjufNhTVlbG/P39GcdxzM/Pj8XFxTG9Xs+PhbGheZs/f75gLhhjzGQyMY7j2I8//sincRzHoqOjmdVq5dPq6uoYx3Hs2rVrfFpUVBTLysqy6c+GDRtYTEwM6+/v59MeP37MYmJiWHp6Op+Wm5vLOI5jly9f5tP6+/uZQqFgu3bt4tNycnJYcHAw+/vvv0edg7feeott2LBBkNbb28sUCgUrKioSjKG3t3fUegghhIwP2iZKCCEEAODm5oaUlBTU1tYKtkO+DKVSKXj28fEBACxdupRPmzZtGqRSKdrb223ej42NFTyvWbMGFouF31ZYX1+PoKAgeHt74/Hjx/xvyZIlMJvNuHnzpuD96Ojo5/bZYrGgqakJsbGxmDTpvxtoZs+ejdDQUFy9evW5ddjzwQcf4MKFCygqKsK6detgNpthMBjw9ttvo6urCwDw22+/YXBwEO+8845gPEFBQXBxcbFpe/HixRCLxfwzx3EAYHcun9bX14erV68iJiYGIpGIb4cxhsWLF9tsFXVycuJXAAHA0dER8+bNw7179/i0S5cuYfny5fDw8LDbZktLC+7evYu1a9cKxjZlyhSEhITwbfr5+UEsFmPr1q04f/48uru7nze1hBBCXhBtEyWEEMJLTk5GZWUl9Ho9SkpKXrq+adOmCZ6HAxdXV1dBuqOjI/r7+23ef/XVV+0+d3Z2Ahg6L3jnzh34+/vbbX/kOTuZTPbcPj948ACMMcyYMcMmb/r06Whra3tuHaORyWSIj4/nzyt+8cUX2Lt3L06cOIHc3Fw+8BktaB05npHzO3whjr25fNr9+/fx5MkTGI1GGI1Gu2UGBwf5S3FG/l/DbT39SQyz2QxPT89R2xweW35+PvLz823yvby8AABz585FRUUFjh8/ju3bt8NqtSIwMBDZ2dlQKBTPHBchhJCxoWCQEEIIz8XFBRqNBsXFxUhNTbXJnzx5MgDbmzhHBinjpbu7W3C2bjigGA7U3NzcIJVK7QYXACCXywXPIy9qscfV1RUODg74559/bPK6urpG/XzEi1Cr1fj000/5Fczhuk+ePGk3ABuvtiUSCUQiEdRqNdatW2e3zLNuR7XH3d0dHR0do+YP9z0rKwsRERE2+U+vcIaHhyM8PBxWqxXXrl2DXq+HRqPBzz//zJ95JIQQ8vIoGCSEECKgUqlw6tQplJaW2uQNr940Nzfzq3EPHjxAY2MjXPN0mhIAAAPFSURBVFxcxr0vdXV1SEtL45+/++47ODs78zeaLl26FJWVlfDy8rJZRXxRzs7O8Pf3x/nz56HT6fgLW9ra2tDY2IjExMQx19nV1QWpVGoTYHV2dqK3t5dfsYyMjIRIJMK9e/cQGRn58oPBUJA1cqXQ2dkZb7zxBq5fv46dO3eOOfCzJzIyEj/88AM6Ozvtrqr6+Phg1qxZaG5uFvynz+Lo6IiIiAhYLBZotVr89ddfFAwSQsg4omCQEEKIgKOjIzIyMlBQUGCTp1QqIZFIUFBQAJ1OB6vVioqKilFvxnxZNTU1GBwcxMKFC9HQ0IAvv/wSOp0OEokEwNC21u+//x4qlQrJycmQy+V49OgR/vzzT5hMpud+KmM0H330ETQaDTQaDVQqFSwWC8rKyjB16lSkpKSMub7a2lrU1NRg7dq1WLhwIZycnNDS0oKTJ09CLBZDrVYDGLqNdPPmzSgsLMTt27ehUCgwefJktLe349KlS4iPjxec3fs3XnvtNZhMJly4cAHTp0+Hu7s7vL29kZeXh8TERKSmpuL999+HTCZDT08Pmpqa8OTJE2RnZ4+pHZ1Oh19//RUJCQnYsmUL5syZg46ODtTX16OkpAQODg7YvXs3tFotBgYGEBsbC3d3d3R1daGxsRFeXl5ISUlBVVUVTCYTlEolZs6ciZ6eHpSXl2PGjBn8mUhCCCHjg4JBQgghNt59912cOHECLS0tgnRXV1ccPXoU+/fvR2ZmJjw9PaHVanH58mVcuXJl3PthNBpRWFgIo9EIiUSC9PR0aLVaPl8ikaC6uhpHjhzB8ePH0dnZCYlEArlcjpUrV75wu0qlEuXl5Thy5AgyMzMhFouhUCiQk5Mz6gUpz7J8+XJ0dHTgp59+wunTp/Hw4UO4u7sjNDQUBw8eFJx53LZtG3x8fHDmzBmcOXMGDg4O8PT0REREBObNmzfmtrdt24aCggJkZmair68PcXFxKC4uhr+/P7766isYDAYUFRWht7cXUqkUCxYswPr168fcjre3N2pqalBaWoqDBw/CYrHAw8MDK1as4MssW7YMlZWVOHr0KHbt2oW+vj7IZDIEBQVh9erVAID58+fj4sWLOHToELq7u+Hm5obQ0FCUlJRgypQpY+4XIYSQ0Tkwxtj/uxOEEEIIIYQQQv636NMShBBCCCGEEDIBUTBICCGEEEIIIRMQBYOEEEIIIYQQMgFRMEgIIYQQQgghExAFg4QQQgghhBAyAVEwSAghhBBCCCETEAWDhBBCCCGEEDIBUTBICCGEEEIIIRPQfwBe/L5Nw69E+AAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 1080x576 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.figure(figsize=(15,8))\n",
    "plt.hist(matched_dfs['count'], bins=100, log = True)\n",
    "#plt.xticks(range(1,21))\n",
    "plt.xlabel('Number of Sentences', fontsize=16)\n",
    "plt.ylabel('Count of papers with x sentences', fontsize=16)\n",
    "plt.grid()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ArticleID</th>\n",
       "      <th>text</th>\n",
       "      <th>count</th>\n",
       "      <th>clean_text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pub.1000000277</td>\n",
       "      <td>[['A detailed comparison of residues involved ...</td>\n",
       "      <td>1</td>\n",
       "      <td>['A detailed comparison of residues involved i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pub.1000000280</td>\n",
       "      <td>[['Also, the epigenetic stability of human ES ...</td>\n",
       "      <td>1</td>\n",
       "      <td>['Also, the epigenetic stability of human ES c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pub.1000000588</td>\n",
       "      <td>[['Hospital and clinic HTC had similar costs (...</td>\n",
       "      <td>1</td>\n",
       "      <td>['Hospital and clinic HTC had similar costs (&amp;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pub.1000000718</td>\n",
       "      <td>[['&lt;b&gt;e&lt;/b&gt;, Allele-based pleiotropy of &lt;i&gt;IL6...</td>\n",
       "      <td>2</td>\n",
       "      <td>['e, Allele-based pleiotropy of IL6R 358Asp (r...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>pub.1000000866</td>\n",
       "      <td>[['However, structural&lt;bibr rid=\"b39 b40 b41 b...</td>\n",
       "      <td>1</td>\n",
       "      <td>['However, structural and biochemical studies ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304442</th>\n",
       "      <td>pub.1121819253</td>\n",
       "      <td>[['Lineage-affiliated transcription factors&lt;bi...</td>\n",
       "      <td>4</td>\n",
       "      <td>['Lineage-affiliated transcription factors wit...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304443</th>\n",
       "      <td>pub.1121819356</td>\n",
       "      <td>[['\\r\\n&lt;p&gt;In order to investigate a yolk-sac c...</td>\n",
       "      <td>1</td>\n",
       "      <td>['In order to investigate a yolk-sac contribut...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304444</th>\n",
       "      <td>pub.1121824901</td>\n",
       "      <td>[['Despite the clear upper limit in observed a...</td>\n",
       "      <td>1</td>\n",
       "      <td>['Despite the clear upper limit in observed ax...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304445</th>\n",
       "      <td>pub.1122267945</td>\n",
       "      <td>[['The so-called photon-dominated or photodiss...</td>\n",
       "      <td>7</td>\n",
       "      <td>['The so-called photon-dominated or photodisso...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304446</th>\n",
       "      <td>pub.1122288893</td>\n",
       "      <td>[['Instead, transformations of reactive N in t...</td>\n",
       "      <td>2</td>\n",
       "      <td>['Instead, transformations of reactive N in th...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>304446 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             ArticleID                                               text  \\\n",
       "1       pub.1000000277  [['A detailed comparison of residues involved ...   \n",
       "2       pub.1000000280  [['Also, the epigenetic stability of human ES ...   \n",
       "3       pub.1000000588  [['Hospital and clinic HTC had similar costs (...   \n",
       "4       pub.1000000718  [['<b>e</b>, Allele-based pleiotropy of <i>IL6...   \n",
       "5       pub.1000000866  [['However, structural<bibr rid=\"b39 b40 b41 b...   \n",
       "...                ...                                                ...   \n",
       "304442  pub.1121819253  [['Lineage-affiliated transcription factors<bi...   \n",
       "304443  pub.1121819356  [['\\r\\n<p>In order to investigate a yolk-sac c...   \n",
       "304444  pub.1121824901  [['Despite the clear upper limit in observed a...   \n",
       "304445  pub.1122267945  [['The so-called photon-dominated or photodiss...   \n",
       "304446  pub.1122288893  [['Instead, transformations of reactive N in t...   \n",
       "\n",
       "        count                                         clean_text  \n",
       "1           1  ['A detailed comparison of residues involved i...  \n",
       "2           1  ['Also, the epigenetic stability of human ES c...  \n",
       "3           1  ['Hospital and clinic HTC had similar costs (&...  \n",
       "4           2  ['e, Allele-based pleiotropy of IL6R 358Asp (r...  \n",
       "5           1  ['However, structural and biochemical studies ...  \n",
       "...       ...                                                ...  \n",
       "304442      4  ['Lineage-affiliated transcription factors wit...  \n",
       "304443      1  ['In order to investigate a yolk-sac contribut...  \n",
       "304444      1  ['Despite the clear upper limit in observed ax...  \n",
       "304445      7  ['The so-called photon-dominated or photodisso...  \n",
       "304446      2  ['Instead, transformations of reactive N in th...  \n",
       "\n",
       "[304446 rows x 4 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matched_dfs[~matched_dfs['ArticleID'].isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "matched_dfs = pd.merge(nature_references_info.reset_index(), matched_dfs, right_on = 'ArticleID', left_on = 'ArticleID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "matched_dfs.to_csv('data/intermediate/matched_dfs_merged.csv', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ArticleID</th>\n",
       "      <th>PubYear</th>\n",
       "      <th>Title</th>\n",
       "      <th>Volume</th>\n",
       "      <th>Issue</th>\n",
       "      <th>Cites</th>\n",
       "      <th>Journal</th>\n",
       "      <th>NatureID</th>\n",
       "      <th>text</th>\n",
       "      <th>count</th>\n",
       "      <th>clean_text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pub.1002147798</td>\n",
       "      <td>1950</td>\n",
       "      <td>Deformations in the Crystal Structures of Anti...</td>\n",
       "      <td>166</td>\n",
       "      <td>4221</td>\n",
       "      <td>105</td>\n",
       "      <td>Nature</td>\n",
       "      <td>166523a0</td>\n",
       "      <td>[['It is well known that the crystallographic ...</td>\n",
       "      <td>1</td>\n",
       "      <td>['It is well known that the crystallographic d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pub.1053335653</td>\n",
       "      <td>1950</td>\n",
       "      <td>Upper Limit of Frequency for Human Hearing</td>\n",
       "      <td>166</td>\n",
       "      <td>4222</td>\n",
       "      <td>75</td>\n",
       "      <td>Nature</td>\n",
       "      <td>166571b0</td>\n",
       "      <td>[['\\r\\n&lt;p&gt;In humans, ultrasound can be detecte...</td>\n",
       "      <td>1</td>\n",
       "      <td>['In humans, ultrasound can be detected via bo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pub.1037944905</td>\n",
       "      <td>1950</td>\n",
       "      <td>Haploid Plants of Solanum demissum</td>\n",
       "      <td>166</td>\n",
       "      <td>4227</td>\n",
       "      <td>28</td>\n",
       "      <td>Nature</td>\n",
       "      <td>166795a0</td>\n",
       "      <td>[['Second, haploids can be induced from rare i...</td>\n",
       "      <td>1</td>\n",
       "      <td>['Second, haploids can be induced from rare in...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pub.1046899436</td>\n",
       "      <td>1950</td>\n",
       "      <td>The Liquid–Solid Transition in Helium Near Abs...</td>\n",
       "      <td>165</td>\n",
       "      <td>4204</td>\n",
       "      <td>26</td>\n",
       "      <td>Nature</td>\n",
       "      <td>165829a0</td>\n",
       "      <td>[['The only exception is helium, which can rem...</td>\n",
       "      <td>1</td>\n",
       "      <td>['The only exception is helium, which can rema...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pub.1003128813</td>\n",
       "      <td>1950</td>\n",
       "      <td>Odour Trails of Acanthomyops fuliginosus</td>\n",
       "      <td>166</td>\n",
       "      <td>4212</td>\n",
       "      <td>25</td>\n",
       "      <td>Nature</td>\n",
       "      <td>166154a0</td>\n",
       "      <td>[['However, previous research has found no evi...</td>\n",
       "      <td>1</td>\n",
       "      <td>['However, previous research has found no evid...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        ArticleID PubYear                                              Title  \\\n",
       "0  pub.1002147798    1950  Deformations in the Crystal Structures of Anti...   \n",
       "1  pub.1053335653    1950         Upper Limit of Frequency for Human Hearing   \n",
       "2  pub.1037944905    1950                 Haploid Plants of Solanum demissum   \n",
       "3  pub.1046899436    1950  The Liquid–Solid Transition in Helium Near Abs...   \n",
       "4  pub.1003128813    1950           Odour Trails of Acanthomyops fuliginosus   \n",
       "\n",
       "  Volume Issue Cites Journal  NatureID  \\\n",
       "0    166  4221   105  Nature  166523a0   \n",
       "1    166  4222    75  Nature  166571b0   \n",
       "2    166  4227    28  Nature  166795a0   \n",
       "3    165  4204    26  Nature  165829a0   \n",
       "4    166  4212    25  Nature  166154a0   \n",
       "\n",
       "                                                text  count  \\\n",
       "0  [['It is well known that the crystallographic ...      1   \n",
       "1  [['\\r\\n<p>In humans, ultrasound can be detecte...      1   \n",
       "2  [['Second, haploids can be induced from rare i...      1   \n",
       "3  [['The only exception is helium, which can rem...      1   \n",
       "4  [['However, previous research has found no evi...      1   \n",
       "\n",
       "                                          clean_text  \n",
       "0  ['It is well known that the crystallographic d...  \n",
       "1  ['In humans, ultrasound can be detected via bo...  \n",
       "2  ['Second, haploids can be induced from rare in...  \n",
       "3  ['The only exception is helium, which can rema...  \n",
       "4  ['However, previous research has found no evid...  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matched_dfs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7f523cfece10>"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0UAAAJQCAYAAABB688pAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvnQurowAAIABJREFUeJzs3XmYXHd95/vPr/buWlq9a2ljyba8YgsZYptdNjdYLImBMRmWgGF48CQhkDtzZyZwc59wk8BNeJiBDCEkYwYDTkjYhsWsjgerAQMO2HiVJVuLZatbanW3eqmq7trrd/+oU6223JK6pao651S9X8+jh+5Tp6u+reM/9OH7+31/xlorAAAAAOhUAbcLAAAAAAA3EYoAAAAAdDRCEQAAAICORigCAAAA0NEIRQAAAAA6GqEIAAAAQEcjFAEAAADoaIQiAAAAAB2NUAQAAACgo4XcLuBsrVu3zl500UVul4HTWFhYUDwed7sMnAbPyPt4Rt7HM/I+npH38Yy8z6/P6IEHHpi21g6e6T7fhqLh4WHdf//9bpeB0xgdHdWOHTvcLgOnwTPyPp6R9/GMvI9n5H08I+/z6zMyxjy9mvtYPgcAAACgoxGKAAAAAHQ0QhEAAACAjubbPUUAAAAATiiVShobG1M+n2/4e/f09GjPnj0Nf99GicViGhkZUTgcPqufJxQBAAAAbWBsbEzJZFKbN2+WMaah753JZJRMJhv6no1irdXx48c1NjamLVu2nNV7sHwOAAAAaAP5fF79/f0ND0ReZ4xRf3//OXXICEUAAABAm+i0QFR3rr83oQgAAABARyMUAQAAAPC8v/7rv9bi4mJT3ptQBAAAAMDzCEUAAAAAPO+OO+7QVVddpW3btukd73iHDh06pBtuuEFXXXWVXvWqV+mZZ56RJL3rXe/S17/+9aWfSyQSkqTR0VHt2LFDN998sy699FK9/e1vl7VWn/rUp3TkyBFdf/31uv766xteNyO5AQAAgDbzZ9/ZrcePpBv2fpVKRVee16sP/9YVp7xn9+7d+shHPqKf//znGhgY0MzMjG655ZalP7fffrs+8IEP6Fvf+tZpP+vBBx/U7t27tXHjRr30pS/Vz372M33gAx/QJz7xCe3atUsDAwMN+73q6BQBAAAAOGf33HOP3vzmNy+Flr6+Pv3iF7/Q2972NknSO97xDt17771nfJ9rrrlGIyMjCgQCesELXqBDhw41s2xJdIoAAACAtnO6js7ZaPThraFQSNVqVZJUrVZVLBaXXotGo0tfB4NBlcvlhn3uqdApAgAAAHDObrjhBn3ta1/T8ePHJUkzMzN6yUteoi9/+cuSpC996Ut6+ctfLknavHmzHnjgAUnSnXfeqVKpdMb3TyaTymQyTamdThEAAACAc3bFFVfoT/7kT/TKV75SwWBQ27dv19/8zd/o3e9+tz7+8Y9rcHBQn//85yVJ733ve3XTTTdp27Zt2rlzp+Lx+Bnf/9Zbb9XOnTu1ceNG7dq1q6G1E4oAAAAANER9qMJy99xzz3PuGx4e1n333bf0/cc+9jFJ0o4dO7Rjx46l65/+9KeXvn7/+9+v97///Q2uuIblcwAAAAA6GqEIAAAAQEcjFAEAAABtwlrrdgmuONffm1AEAAAAtIFYLKbjx493XDCy1ur48eOKxWJn/R4MWgAAAADawMjIiMbGxjQ1NdXw987n8+cUOpotFotpZGTkrH+eUAQAAAC0gXxF+vCuKf35b1+hzQNnHnG9FqOjo9q+fXtD39NLWD4HAAAAtIFHx+f1kyen9KO9k26X4juEIgAAAKANjM3mJEn7jmVcrsR/CEUAAABAGxivh6LJrMuV+A+hCAAAAGgD43O1UPTksUzHTaA7V4QiAAAAoA2MzS5KkjL5so6lCy5X4y+EIgAAAKANjM/ltD5VG5v9JPuK1oRQBAAAAPhcpWp1dC6vHZcMSiIUrRWhCAAAAPC5Y+m8ylWrK0d61B+PaN8xhi2sBaEIAAAA8Ln6kIWR3m5tHU7oyUk6RWtBKAIAAAB8rj5kYdO6Lm0dSmr/sSwT6NaAUAQAAAD4XP2MopHeLl08nFCmUNbR+bzLVfkHoQgAAADwufG5nAYSEcXCQW0dTkpi2MJaEIoAAAAAnxubzWnTui5J0sVOKGLYwuoRigAAAACfG5/NaaS3W5LUF49oIBHRPoYtrBqhCAAAAPAxa63G53La1Nu1dG3rUFJPrqJTVKpUlcmXmlmeLxCKAAAAAB+byhZUKFeXls9J0sXDCe2fPPMEuo/f9YRe96l7m12i5xGKAAAAAB9bPnmubutwUtlCWUfOMIHuX3ZP6JmZRc0vdna3iFAEAAAA+Fj94Nbly+cuXsUEusMzizp0vHa+0dMzC02s0PsIRQAAAICPjTmdopOXz0nSvtOEop/tn176+qlpQhEAAAAAnxqfzamnK6xkLLx0bV13RIPJ6GmHLfx0/7T64xFJ0tNOx6hTEYoAAAAAHxubXXxWl6hu61DilJ2iatXq5/un9cqLB7WhJ6ZDx+kUAQAAAPCpk8dx1108nNS+yayq1edOoHv8aFqziyW9bOuANvfH6RS5XQAAAACAs2OtdQ5uXaFTNJzQYrGyNIhhuZ/uq+0netlFA9o80K2n6RQBAAAA8KO5xZIWipUVl8/VJ9Dtm3zuErqf7Z/WJcNJDaViOr8/rulssaMPcSUUAQAAAD5V7wKN9HY/57WLh2qh6LHx9LOu50sV/fLQjF560YAkaXN/7Wc7eQkdoQgAAADwqbEVDm6t6+kO67oL+vTZnx7UxLJDXH91aEbFclUv3+qEooG4JHX0sAVCEQAAAOBTY7O17s5Ky+ck6a/edJVKlao++I1HZG1t4MK9+6cVDhpde0GfJOl5fXSKCEUAAACAT43P5RSPBLWuO7zi65sH4vrgzks1+sSUvvbAmCTp3n3Tuvp5veqOhCRJ3ZGQhlPRjj7AlVAEAAAA+NTYbG0ctzHmlPe888Wbde2WPv3Fdx7XY+Pz2n0krZc5+4nqzu+Pd/QEOkIRAAAA4FPjs7lTLp2rCwSMPn7zNlWs1bs+/0tJ0su2PjsUbemP6xDL5wAAAAD4zfhcbsXJcyd7Xn+3PviaSzWdLSoVC+mqkXXPev38gW5NZQpaKJSbVaqnhdwuAAAAAMDaZfIlzedK2rTC5LmV/O615+vefdPauK5LwcCzl9tt7j8xge6KjT0Nr9XrCEUAAACAD9XPKDrT8rm6QMDotne+aMXXzl92VhGhCAAAAICnZfIl7Z3I6IePTUha+YyitTq/v7PPKiIUAQAAAD5w1+4JfeR7j+vwTG7p2qZ1XbpoKHHO752IhjSYjOrp6c4ctkAoAgAAAHzgrscmNLdY0n++8RJdtiGpyzaktD4VO+047rXY3N+tp+gUAQAAAPCqqWxBFw4m9L7rL2rK+5/fH9dP90015b29jpHcAAAAgA9MpgsaSkab9v6b+7t1LF3QYrHzxnITigAAAAAfmMoWNNjMUDRQG7bwzEzn7SsiFAEAAAAeVyxXNbNQ1FAy1rTPWDqraLrz9hURigAAAACPO75QkKSmdoqe55xVdOg4nSIAAAAAHjOZroWiZu4pSsXC6o9H9HQHTqAjFAEAAAAeN5VpfqdIqu0rOtSBZxURigAAAACPm3RC0VCquaHo/P5uHaJTBAAAAMBr6p2i/niTO0X9cR2dzytfqjT1c7yGUAQAAAB43GQmr754RJFQc//5fun6pCTpl0/NNPVzvIZQBAAAAHjcVKagwURzu0SS9IqLB5WMhfStB8eb/lleQigCAAAAPG4yU2j6fiJJioWDet2VG/TD3RNaLJab/nleQSgCAAAAPK5VnSJJeuP2TVosVvQvu4+15PO8YNWhyBgTNMY8aIz5rvP9FmPMvxpj9htjvmKMiTjXo873+53XNy97jw85158wxty47PpO59p+Y8wHG/frAQAAAP5mra2FohZ0iiTpNzb3adO6Ln2zg5bQraVT9EeS9iz7/mOSPmmtvUjSrKT3ONffI2nWuf5J5z4ZYy6X9BZJV0jaKekzTtAKSvpbSa+RdLmktzr3AgAAAB0vnSurWKm2rFMUCBi9YftG/XTflCYz+ZZ8pttWFYqMMSOSXifpfzrfG0k3SPq6c8sXJb3B+fom53s5r7/Kuf8mSV+21hastU9J2i/pGufPfmvtQWttUdKXnXsBAACAjlcPJkOpWMs+843bN6lqpe88fLRln+mm1XaK/lrSf5FUdb7vlzRnra3vvhqTtMn5epOkw5LkvD7v3L90/aSfOdV1AAAAoOPVzyhqVadIki4aSurKTT365oNjLftMN4XOdIMx5vWSJq21DxhjdjS/pNPWcqukWyVpcHBQo6OjbpaDM8hmszwjj+MZeR/PyPt4Rt7HM/I+ntHp/fxIrQ9xaM/DKhxu3Zy05ydL+ue9RX3pu/eoR4tt/YzOGIokvVTSbxtjXispJikl6b9LWmeMCTndoBFJ9Z1Y45LOkzRmjAlJ6pF0fNn1uuU/c6rrz2KtvU3SbZJ0ySWX2B07dqyifLhldHRUPCNv4xl5H8/I+3hG3scz8j6e0ent+8lB6ZE9et3/8XKlYuGWfe4VmYK++pc/0nhoozbFJtr6GZ0xalprP2StHbHWblZtUMI91tq3S9ol6Wbntlskfdv5+k7nezmv32Ottc71tzjT6bZI2irpl5J+JWmrM80u4nzGnQ357QAAAACfm8zkFQsHlIyupp/ROIPJqF520YC+/dARVa1t6We32rn03/5Y0n80xuxXbc/Q55zrn5PU71z/j5I+KEnW2t2SvirpcUk/lPQ+a23F6TT9oaS7VJtu91XnXgAAAKDjTWUKGkxGVZtd1lpvunqTxudyemKmeuabfWxNcdNaOypp1Pn6oGqT406+Jy/pzaf4+Y9K+ugK178v6ftrqQUAAADoBJOZgoaSrZs8t9yrL1+vvnhEdx0q6fddqaA1WrdTCwAAAMCaTWUKLZ08t1xXJKh3XHe+HpqqaP9kxpUaWoFQBAAAAHjYZKagoZQ7oUiS3vni8xUOSJ/9yVOu1dBshCIAAADAowrliuZzJdc6RZLUn4jq5ZtC+uaD45pM512ro5kIRQAAAIBH1Q9udbNTJEk3bg6rVK3qCz8/5GodzUIoAgAAADyqHooGk+6GouF4QDuvWK9/vO9pZQtlV2tpBkIRAAAA4FGT9U6RS9Pnlrv1FRconS/rK7867HYpDUcoAgAAADzKK50iSdr+vF5ds7lPt9/7lEqV9jq3iFAEAAAAeNRkpiBjpP54xO1SJNW6ReNzOf3gsQm3S2koQhEAAADgUVOZgvrjEYWC3vhn+w2XDikZC+mBQzNul9JQ3vjbBQAAAPAcU5m8Bj2wn6guEDDatK5LR+fbazQ3oQgAAADwqKlMwRP7iZZb3xPTRJudV0QoAgAAADxqMlPQkMdC0YaeGJ0iAAAAAM1XrVpNZz3YKUp1aTpbULHcPhPoCEUAAACAB83lSipVrCc7RdZKk5n26RYRigAAAAAP8tIZRcsN99QGP0y00RI6QhEAAADgQfVOzGDCW6FogxOK2mlfEaEIAAAA8KB6p2go5Z2R3FJt+pxEpwgAAABAk016dPlcMhpSPBKkUwQAAACguaYyBXVHgkpEQ26X8izGGOesopzbpTQMoQgAAADwoEkPHtxat6Gni04RAAAAgOaayuQ9N2Shbn1PjD1FAAAAAJprbrGkvnjE7TJWtKEnpslMQeVKexzgSigCAAAAPGg+V1KqK+x2GSta3xNTpWo1nS26XUpDEIoAAAAAD5rPldTj0VB04qyi9hi2QCgCAAAAPKZUqWqxWPFsKFqf6pLUPmcVEYoAAAAAj0nnSpKkVMxb47jrTnSKCEUAAAAAmmDeCUU93d7sFK3rDisaCmgiTSgCAAAA0ATpfFmSPLt8zhijDT0xOkUAAAAAmmN+afmcN0ORVD+riEELAAAAAJpgafmcRztFkrShp4tOEQAAAIDmSPsgFK3vielYOq9q1bpdyjkjFAEAAAAes7R8zsOhaENPTKWK1fEF/x/gSigCAAAAPCadKykSCigWDrpdyimtT9XGcrfDWUWEIgAAAMBj0vmSp5fOSbU9RZJ0tA2GLRCKAAAAAI+Zz5U8e3Br3XrnANd2OKuIUAQAAAB4zHzO+52i/nhE4aBpiwl0hCIAAADAY9K5sudDUSBgNJyKsacIAAAAQOPN50qenjxXt6Enxp4iAAAAAI3nh+VzkrS+p4tOEQAAAIDGqlatMj6YPifVO0V5WevvA1wJRQAAAICHZItlVa2Uink/FK1PxVQoVzW3WHK7lHNCKAIAAAA8ZN4JGH7pFEny/QQ6QhEAAADgIel8LRT5YdDCibOK/D1sgVAEAAAAeMh8rh6KvH14qyRt6OmSRKcIAAAAQAOlc/5ZPjeYjCoYML6fQEcoAgAAADwknStL8kcoCgaMhpJROkUAAAAAGufE8jnvhyJJGunt0u4jaV+P5SYUAQAAAB4ynyspYKRExPt7iiTpt7dt1J6jaT10eM7tUs4aoQgAAADwkHS+pFRXWIGAcbuUVXnj1SNKREP6h1887XYpZ41QBAAAAHjIfK7ki4Nb6xLRkN509SZ995GjOp4tuF3OWSEUAQAAAB4ynyv5YsjCcu+47nwVK1V99f4xt0s5K4QiAAAAwEPSPgxFW4eTuu6CPv3jfU+rUvXfwAVCEQAAAOAh87mSLw5uPdk7X7xZ43M57do76XYpa0YoAgAAADwknS/7rlMkSb95+bCGU1HdcZ//Bi4QigAAAAAPqXWK/BeKwsGA3nbN+frJk1N6anrB7XLWhFAEAAAAeES+VFGxXPXV9Lnl3nrNeQoFjP7RZ90iQhEAAADgEelcSZJ8uXxOkoZSMb36imF9+6EjbpeyJoQiAAAAwCPmfR6KJOnS9SlNZwsqlCtul7JqhCIAAADAI+qhyI97iuqGU1FJ0lTGPwe5EooAAAAAj0jn/d8pGkrGJEmThCIAAAAAa9UOy+cGk7VO0WSaUAQAAABgjeYXneVzMf8d3lo3tLR8Lu9yJatHKAIAAAA8Ip0vS/L3nqL+eFQBw/I5AAAAAGdhPldSPBJUOOjff6YHA0YDiSjL5wAAAACs3Xyu5OsuUd1QKqpJls8BAAAAWKt0ruTrIQt1Q8kYy+cAAAAArF3bdIqSUUIRAAAAgLWbz5WUirVHKJrOFlSuVN0uZVUIRQAAAIBHZPLltlg+N5iKyVrp+ELR7VJWhVAEAAAAeMR82+wp8tcBroQiAAAAwAPKlaqyhbJSXf49uLVuOBWTJN9MoCMUAQAAAB6QcQ5ubatOkU+GLRCKAAAAAA+Yz5UktUcoGkiwfA4AAADAGtVDUTtMn4uEAuqLR1g+BwAAAGD10nmnU9Tt/1Ak+eusIkIRAAAA4AHttHxOkgYJRQAAAADWop2Wz0nSUDKmqTTL5wAAAACsUjrXPtPnJGkoFdVUtiBrrdulnBGhCAAAAPCA+VxJkWBAsXB7/BN9KBlVqWI1u1hyu5Qzao+/cQAAAMDn5nMlpbpCMsa4XUpDDCX9c4AroQgAAADwgHS+pFSbLJ2TasvnJH+cVUQoAgAAADwgnSu1zX4iqbZ8TpKO+WDYAqEIAAAA8ID5XKltJs9Jy5fP0SkCAAAAsArt1inqigSVjIU0RSgCAAAAsBrzbRaKpNoSOgYtAAAAADgja63S+bJSXSG3S2mooWSMQQsAAAAAzixfqqpStUq20Z4iqTaBjj1FAAAAAM4ok68dcJqItlunqLZ8zlrrdimnRSgCAAAAXJYplCVJyVi7haKY8qXq0u/nVYQiAAAAwGXZfJuGIp8c4EooAgAAAFyWcUJRItpee4oGnQNcvT6BjlAEAAAAuCxbaNc9RbUDXL1+VhGhCAAAAHBZhuVzriIUAQAAAC7LtumghWQ0pFg4wPI5AAAAAKdX7xTF22z5nDFGw6mY588qIhQBAAAALssWyoqFAwoH2++f50PJqI6l6RQBAAAAOI1MvqxkrL0mz9UNJekUAQAAADiDTL6kZJstnasbTEY1xaAFAAAAAKeTLZSVaLMhC3VDqagyhbJyxYrbpZwSoQgAAABwWTZfbrvJc3WDidpY7umsd7tFhCIAAADAZZl8ue0Obq1LddX2SqXzJZcrObUzhiJjTMwY80tjzMPGmN3GmD9zrm8xxvyrMWa/MeYrxpiIcz3qfL/feX3zsvf6kHP9CWPMjcuu73Su7TfGfLDxvyYAAADgXdlCWYloew5aqHfA0rmyy5Wc2mo6RQVJN1hrt0l6gaSdxpjrJH1M0iettRdJmpX0Huf+90iada5/0rlPxpjLJb1F0hWSdkr6jDEmaIwJSvpbSa+RdLmktzr3AgAAAB0hky+17fK5lDNVL+PnTpGtyTrfhp0/VtINkr7uXP+ipDc4X9/kfC/n9VcZY4xz/cvW2oK19ilJ+yVd4/zZb609aK0tSvqycy8AAADQ9qy1yhbad09R/feqH1DrRavaU+R0dB6SNCnpbkkHJM1Za+u/2ZikTc7XmyQdliTn9XlJ/cuvn/Qzp7oOAAAAtL3FYkVVq/bdU+SDTtGq/uattRVJLzDGrJP0TUmXNrWqUzDG3CrpVkkaHBzU6OioG2VglbLZLM/I43hG3scz8j6ekffxjLyv05/RbL4qSTry9EGNjh4+w93uOJdnVK5aSdIje/dptPR0A6tqnDXFUWvtnDFml6QXS1pnjAk53aARSePObeOSzpM0ZowJSeqRdHzZ9brlP3Oq6yd//m2SbpOkSy65xO7YsWMt5aPFRkdHxTPyNp6R9/GMvI9n5H08I+/r9Ge0fzIjjf5EL9x2hXZs2+h2OSs612fUteuH6l8/oh07vDk6YDXT5wadDpGMMV2SflPSHkm7JN3s3HaLpG87X9/pfC/n9Xustda5/hZnOt0WSVsl/VLSryRtdabZRVQbxnBnI345AAAAwOvqe22Sbbp8TqrtK/LynqLV/M1vkPRFZ0pcQNJXrbXfNcY8LunLxpiPSHpQ0uec+z8n6R+MMfslzagWcmSt3W2M+aqkxyWVJb3PWZYnY8wfSrpLUlDS7dba3Q37DQEAAAAPyxZqYSHRpoMWpDYIRdbaRyRtX+H6QdUmx518PS/pzad4r49K+ugK178v6furqBcAAABoK9l6p6itQ1HY34e3AgAAAGieegelXafPSVKqK+zpThGhCAAAAHBRplDfUxR2uZLmqS2fo1MEAAAAYAX15XPtvKcoFQspTacIAAAAwEoy+ZK6I0EFA8btUpomGQvTKQIAAACwsmyh3Nb7iaTauPF8qapSpep2KSsiFAEAAAAuyhTKbT15TjoxWc+rwxYIRQAAAICLMvmyErH2HbIg1abPSfLsEjpCEQAAAOCibL6kZLsvn3NCXzpHpwgAAADASTpiT9HS8jk6RQAAAABOksl3zp4ir47lJhQBAAAALsrmy219RpEkpWLsKQIAAACwgmrVKlsst/2eohOhiE4RAAAAgGUWimVZe2IQQbtKLC2fo1MEAAAAYJlsodY5afflc8GAUTwSpFMEAAAA4NmyTkho9+lzUq0bxp4iAAAAAM9Sn8bW7tPnpNrvSKcIAAAAwLPUl88RitxFKAIAAABccmL5XHsPWpCkVFeYQQsAAAAAni1bqIWEzugUhekUAQAAAHi2ekho9+lzUn35HJ0iAAAAAMvUQ1E80hmhKE2nCAAAAMBy2UJZ8UhQwYBxu5SmS8XCKparKpQrbpfyHIQiAAAAwCWZfEnJWPsPWZCklLNE0Iv7ighFAAAAgEuyhXJH7CeStBT+0jnv7SsiFAEAAAAuyeTLSkQ7JRTRKQIAAABwkky+3BHjuKUTnSJCEQAAAIAl2UInhaJ6p4jlcwAAAAAcWZbPeQKhCAAAAHBJR02f63IGLdApAgAAACBJlarVQrHSMZ2iRCQkY+TJA1wJRQAAAIALFoq1cNApe4oCAaNEJMSeIgAAAAA19b01ndIpkmoBkD1FAAAAACTVhixI6pg9RVLtd6VTBAAAAECSlC3UwkGiQ5bPSVKqK6R0jk4RAAAAAHXq8rmwMgU6RQAAAAB0IhSlOqhTxJ4iAAAAAEuyBadTRChyHaEIAAAAcEG2U5fP5Uuy1rpdyrMQigAAAAAXZPIlGSPFI50UikIqVazyparbpTwLoQgAAABwQaZQViISUiBg3C6lZVLO+HGvjeUmFAEAAAAuyObLHbWfSKp1iiQp7bF9RYQiAAAAwAWZfHkpJHQKOkUAAAAAlmQL5Y4asiCd6BR5bQIdoQgAAABwQaZQVsLpnHSK5FKniFAEAAAAdLxMvqRkh3aK0iyfAwAAAJDtxD1FXewpAgAAAODoxD1F8UhQAcPyOQAAAKDjlStVLRYrHTeS2xijRDREKAIAAAA63UKhIunE4IFOkoyF2VMEAAAAdLpMoRYKOm3QglQbtpDO0SkCAAAAOtp8zglFHbZ8TqoNW2DQAgAAANDhjs7lJUkb1nW5XEnrpWLsKQIAAAA63vhcTpK0cV3M5UpaLxkLLy0f9ApCEQAAANBiR+ZyioQCGohH3S6l5ZJ0igAAAACMzeW0sSemQMC4XUrL1UORtdbtUpYQigAAAIAWOzKX06bezttPJNWWz1WqVovFitulLCEUAQAAAC02PpvTxp7ODEUp52wmLy2hIxQBAAAALVQoVzSZKXRwp6g2htxLY7kJRQAAAEALTczXxnFv6sBx3NKJUJSmUwQAAAB0pvHZ2jjuzg1FteVzaTpFAAAAQGeqn1HUqcvnerqcUJQjFAEAAAAdqR6K1vd03sGtktQXj0iSZheKLldyAqEIAAAAaKEjczkNJaOKhoJul+KKnq6wjJFmFukUAQAAAB1pfC6njR26n0iSggGjdV1hOkUAAABApzoyl+/Y/UR1vfGIZhYJRQAAAEDHqVatxudyGungTpEk9XZH6BQBAAAAnWh6oaBiudrRy+ckJxSxpwgAAADoPEfmOvvg1rq+OHuKAAAAgI5UP7i14ztFzp4ia63bpUgiFAEAAAAtc6TDD26t6+uOqFiuarFYcbsUSYQiAAAAoGXG53JKRENKxUJul+KqXucA1xmPLKEjFAEAAAAtMj6X06ahVxnEAAAgAElEQVR1XTLGuF2Kq/q6a6Fo1iNjuQlFAAAAQIuMz+Y6fumcRKcIAAAA6FjjczltXBdzuwzX9cXpFAEAAAAdJ1soaz5X0qZ13W6X4rr68rmZBW+cVUQoAgAAAFqgPnmOTpGUjIUUDBjPnFVEKAIAAABaYNwJRSPsKVIgYNTbHdYMy+cAAACAzsHBrc/W2x2hUwQAAAB0kiNzOYUCRkNJls9JtQl0TJ8DAAAAOsj4XE4b1sUUDHT2GUV1fd0Rps8BAAAAneTIXE4be1g6V1frFDF9DgAAAOgYHNz6bH3xsGYXi7LWul0KoQgAAABotlKlqol0XpsYsrCktzuiStUqnS+7XQqhCAAAAGi2Y+m8qlaEomV6nQNcvTCBjlAEAAAANBnjuJ+rL+6EIg8MWyAUAQAAAE12ZJ5QdLJeQhEAAADQOaYyBUnScCrqciXe0ecsn/PCBDpCEQAAANBkU5mCoqGAEtGQ26V4Rm88LIk9RQAAAEBHmM4WNZCIyhgObq1LREMKB41mWD4HAAAAtL/pbEGDSZbOLWeMUW93hE4RAAAA0AmmMgUNJAhFJ+uLRzRDKAIAAADa33S2qMFkxO0yPKe3O8L0OQAAAKDdVapWMwt0ilZCpwgAAADoADMLRVWtCEUr6I2HNbvISG4AAACgrU1na2cUEYqeq687ornFoipV62odhCIAAACgieqhiOlzz9Ubj6hqpXTO3W4RoQgAAABoohOdIgYtnKwvXvs7cfusIkIRAAAA0ETTmdo/+AfoFD1Hb3ctFLl9VhGhCAAAAGiiqWxBkVBAyWjI7VI8Z6lTRCgCAAAA2td0pqDBRFTGGLdL8ZxeJxS5fVYRoQgAAABooqlsgf1Ep9DXXe8UMWgBAAAAaFvT2SKT506hKxJUNBTQnNc7RcaY84wxu4wxjxtjdhtj/si53meMudsYs8/5317nujHGfMoYs98Y84gx5upl73WLc/8+Y8wty66/0BjzqPMznzL0FgEAANAmprMFzig6jb54xBd7isqS/i9r7eWSrpP0PmPM5ZI+KOlH1tqtkn7kfC9Jr5G01flzq6S/k2ohStKHJV0r6RpJH64HKeee9y77uZ3n/qsBAAAA7qpUrWYWioSi0+jtjnh/T5G19qi19tfO1xlJeyRtknSTpC86t31R0hucr2+SdIetuU/SOmPMBkk3SrrbWjtjrZ2VdLeknc5rKWvtfdZaK+mOZe8FAAAA+NbsYlGVqmVP0Wl4oVO0prmAxpjNkrZL+ldJw9bao85LE5KGna83STq87MfGnGunuz62wvWVPv9W1bpPGhwc1Ojo6FrKR4tls1mekcfxjLyPZ+R9PCPv4xl5Xzs/o8OZqiRp8pkDGi097XI1Z6+Zz6iUzWs8XXX1v4FVhyJjTELS/5L0f1pr08u3/VhrrTHGNqG+Z7HW3ibpNkm65JJL7I4dO5r9kTgHo6Oj4hl5G8/I+3hG3scz8j6ekfe18zO6d9+09LN/1Suv3a5rL+h3u5yz1sxntGv+Me196Iir/w2savqcMSasWiD6krX2G87lY87SNzn/O+lcH5d03rIfH3Gune76yArXAQAAAF+bzhYkSQNMnzul3nhE87mSypWqazWsZvqckfQ5SXustZ9Y9tKdkuoT5G6R9O1l19/pTKG7TtK8s8zuLkmvNsb0OgMWXi3pLue1tDHmOuez3rnsvQAAAADfWgpFDFo4pT7nANe5nHtnFa1m+dxLJb1D0qPGmIeca/+3pL+S9FVjzHskPS3pd5zXvi/ptZL2S1qU9G5JstbOGGP+QtKvnPv+3Fo743z9B5K+IKlL0g+cPwAAAICvTWUKigQDSsXWtJW/o/Q6B7jOujil74xPx1p7r6RTnRv0qhXut5Led4r3ul3S7Stcv1/S889UCwAAAOAnU9mCBhIRcQznqdU7RW5OoFvVniIAAAAAazedLbKf6AyWOkUunlVEKAIAAACaZDpT0CD7iU7rRKfIvT1FhCIAAACgSaazBYYsnMG67rAkOkUAAABA26lWrY4vFDWQjLhdiqfFwkHFwgHNuzh9jlAEAAAANMHsYlGVqqVTtAqJaEgLhbJrn08oAgAAAJpgOltbDkYoOrPuCKEIAAAAaDsc3Lp68WhIC8WKa59PKAIAAACaoB6KBhnJfUbxSJBOEQAAANBupjJOKKJTdEZ0igAAAIA2NJ0tKhIMKNUVcrsUz4tH6RQBAAAAbWcqU1B/IiJjjNuleF48EtIioQgAAABoLxzcunrxaEhZQhEAAADQXmqhiINbV6M7EtRisSJrrSufTygCAAAAmmA6W2Dy3CrFoyGVq1aFctWVzycUAQAAAA1WrVodzxZZPrdK8UhQkrTo0gQ6QhEAAADQYHO5kspVSyhapXi0NqHPrQl0hCIAAACgweoHtw6wfG5VlkJRkVAEAAAAtIVp5+BWBi2szolOEcvnAAAAgLYw5XSKhugUrUp9TxHL5wAAAIA2MZ0tShJ7ilap3ilaZPkcAAAA0B6mswWFg0Y9XWG3S/GFeKQWirIsnwMAAADaw8R8XkPJmIwxbpfiC/FofSQ3nSIAAACgLYzP5rRpXZfbZfhGfflclj1FAAAAQHsYn8tpUy+haLWioYCCAaNFls8BAAAA/leuVDWRztMpWgNjjLojQTpFAAAAQDuYSOdVqVo6RWuUiIbYUwQAAAC0g/HZnCTRKVqj7kiQw1sBAACAdjA+54QiOkVrEo+GtECnCAAAAPA/OkVnJx4JaYE9RQAAAID/jc/lNJCIKBYOul2Kr8SjLJ8DAAAA2sLYbE6bervdLsN34gxaAAAAANrD+FxOIyydW7PuSEhZOkUAAACAv1WrloNbz1IiGqRTBAAAAPjd9EJBxXKVIQtnoTsS0mKxomrVtvyzCUUAAABAgzB57uwloiFJ0mKp9UvoCEUAAABAg3BG0dnrjtam9bkxlptQBAAAADTIUqeIULRm9U4RoQgAAADwsfG5nJKxkFKxsNul+E53pB6KWD4HAAAA+Nb4bI79RGcpXl8+58IEOkIRAAAA0CDjczmNsHTurMQjLJ8DAAAAfM1aqzE6RWctXt9TVGT5HAAAAOBL6VxZ2UJZI73dbpfiS3GmzwEAAAD+Nja3KInJc2crzvQ5AAAAwN84uPXcdIfrnSKWzwEAAAC+xMGt5yYUDCgaCmiR6XMAAACAP43P5hQLB9Qfj7hdim8loiFGcgMAAAB+NT6X08Z1XTLGuF2Kb3VHgyyfAwAAAPxqfI5x3OcqHgkxaAEAAADwq/FZDm49V3GWzwEAAAD+tFgs6/hCkU7ROYpHQyyfAwAAAPzoCJPnGiIeCbJ8DgAAAPCjsaUzirpdrsTf4tGQFot0igAAAADfqZ9RxJ6icxOPBJWlUwQAAAD4z/hsTqGA0XAq5nYpvlbrFBGKAAAAAN8Zn8tpfU9MwQBnFJ2LeDSkUsWqUG7tEjpCEQAAAHCOxmY5o6gR4pGgJGmxxRPoCEUAAADAOXpqekFbBuJul+F73dGQJLV8XxGhCAAAADgHswtFzSwUdeFgwu1SfC/hhKJWT6AjFAEAAADn4OB0VpJ04RCdonPV7Syfo1MEAAAA+MiByQVJ0gUDdIrO1YlOEaEIAAAA8I0D01lFggHOKGqA7kgtFC0waAEAAADwjwOTCzq/v1uhIP+0PlfxaG353ALL5wAAAAD/ODiVZchCg8RZPgcAAAD4S6lS1TMziwxZaJB4pD6Sm+VzAAAAgC88fXxR5aplyEKDxMIBBQydIgAAAMA3Dk7Vx3ETihrBGKN4JMRIbgAAAMAvDkw547gHWT7XKPFoSIssnwMAAAD84cBUVoPJqFKxsNultI3uaFBZls8BAAAA/lCbPEeXqJES0ZAWWT4HAAAAeJ+1VgemFnQB47gbqjsS5PBWAAAAwA9mFoqaz5U4o6jBEtGQFlg+BwAAAHgfQxaaozsS0gLL5wAAAADvq4/jvohOUUPFoyEtFFk+BwAAAHjegamsoqGANq7rcruUthKPBOkUAQAAAH5wYGpBWwbiCgaM26W0lXg0pMViRdWqbdlnEooAAACAs1Abx83SuUaLR4OSpFypdUvoCEUAAADAGhXKFT0zs8iQhSaIR0OS1NIldIQiAAAAYI2eOb6oqhWdoiaIR5xQ1MJhC4QiAAAAYI0OOJPnCEWN1x2pLZ+jUwQAAAB4WP2Moi0sn2u4BMvnAAAAAO87MJXV+lRs6R/waJzueigqEooAAAAAzzowtaALh+gSNUMiWl8+x54iAAAAwJOstTo4ldUFA+wnaobuCMvnAAAAAE87Op9XJl/WxeuTbpfSlpZGcjN9DgAAAPCmvRNpSdJlhKKmiDN9DgAAAPC2PUczkkSnqElCwYCioQCDFgAAAACv2juR0Uhvl1KxsNultK14NESnCAAAAPCqvUfTunR9yu0y2lo8GtQi0+cAAAAA78mXKjo4vaDLNrB0rpnikZCydIoAAAAA79k/mVWlaukUNVk8GmJPEQAAAOBFe47WJs9dSqeoqWp7ilg+BwAAAHjO3omMYuGANvfH3S6lrSWiQZbPAQAAAF60dyKtS4aTCgaM26W0tXiE6XMAAACA51hrtedohv1ELRCPMmgBAAAA8JypbEEzC0X2E7VAwjmnyFrbks8jFAEAAACrsPdoRpLoFLVAPBpS1Ur5UrUln0coAgAAAFZh74QzeW49naJmS0SDktSyJXSEIgAAAGAV9h7NaH0qpt54xO1S2l48GpKklg1bIBQBAAAAq7BnIsN+ohZJOKGIThEAAADgEaVKVfsnmTzXKoQiAAAAwGMOTi2oVLG6jE5RS3hu+Zwx5nZjzKQx5rFl1/qMMXcbY/Y5/9vrXDfGmE8ZY/YbYx4xxly97Gduce7fZ4y5Zdn1FxpjHnV+5lPGGE7CAgAAgKfsOVobsnDZBjpFrRD3YKfoC5J2nnTtg5J+ZK3dKulHzveS9BpJW50/t0r6O6kWoiR9WNK1kq6R9OF6kHLuee+ynzv5swAAAABX7ZlIKxIMaMtA3O1SOkJiqVNUacnnnTEUWWt/ImnmpMs3Sfqi8/UXJb1h2fU7bM19ktYZYzZIulHS3dbaGWvtrKS7Je10XktZa++ztZOZ7lj2XgAAAIAn7D2a0UVDCYWD7D5phbgzktszy+dOYdhae9T5ekLSsPP1JkmHl9035lw73fWxFa4DAAAAnrF3Is3kuRaKR1q7fC50rm9grbXGGNuIYs7EGHOrasvyNDg4qNHR0VZ8LM5SNpvlGXkcz8j7eEbexzPyPp6R93n9GWWLVsfSBYUXpjxdZzO58YyiQWnvgac0Gj7S9M8621B0zBizwVp71FkCN+lcH5d03rL7Rpxr45J2nHR91Lk+ssL9K7LW3ibpNkm65JJL7I4dO051KzxgdHRUPCNv4xl5H8/I+3hG3scz8j6vP6PvPHxE0oO6+foX6jc297ldjivceEY9P/vf6h0c0o4dVzX9s852+dydkuoT5G6R9O1l19/pTKG7TtK8s8zuLkmvNsb0OgMWXi3pLue1tDHmOmfq3DuXvRcAAADguu8/elSDyaiufl7vmW9GwySiIe8snzPG/LNqXZ4BY8yYalPk/krSV40x75H0tKTfcW7/vqTXStovaVHSuyXJWjtjjPkLSb9y7vtza219eMMfqDbhrkvSD5w/AAAAgOsWi2XtemJSv/Oi8xQMcHJMK8WjoZYNWjhjKLLWvvUUL71qhXutpPed4n1ul3T7Ctfvl/T8M9UBAAAAtNo9eyeVL1X12is3uF1Kx4lHg94ZyQ0AAAB0qu8/elQDiWjH7iVyUyuXzxGKAAAAgBUsFsu6Z++kXvP89Sydc0E8GtJCkVAEAAAAuGbX3imWzrmolXuKCEUAAADACupL567ZwtI5N7B8DgAAAHBRrljRPXsntfP5wyydc0kiGlK+VFW5Um36ZxGKAAAAgJPsemJSuVKFpXMuikdrg7JbMYGOUAQAAACc5HuPHNVAIqJrt/S7XUrHSkSDkqRsC4YtEIoAAACAZepL5268gqlzbjrRKSIUAQAAAC314yenlCtV9DqWzrmqHopaMWyBUAQAAAAs88DTM4qEAnoRB7a6KkGnCAAAAHDHw2PzunxDSpEQ/1R2UzxCKAIAAABarlyp6tGxeb3gvHVul9LxEkvL55g+BwAAALTM/qmscqWKtp3X43YpHS9enz6XLzX9swhFAAAAgOORw/OSpKtG6BS5LRFzls8V6RQBAAAALfPQ2JySsZC29MfdLqXjRUNBhYOG6XMAAABAKz18eE7bRtYpwPlEnhCPhhi0AAAAALRKvlTRExMZXTXCfiKviEdCdIoAAACAVtl9JK1y1Wobk+c8I0GnCAAAAGidhw/PSZK2MWTBM+LRoBYYyQ0AAAC0xiNjcxpORbW+J+Z2KXDEoyyfAwAAAFrm4bF5ukQekyAUAQAAAK0xv1jSU9ML7CfyGPYUAQAAAC3yyDj7ibyI5XMAAABAi9SHLFzJOG5PqXeKrLVN/RxCEQAAADrew2PzumAgrp6usNulYJl4NKSqlfKlalM/h1AEAACAjmat1UOH59hP5EGJaFCSmr6EjlAEAACAjjaRzmsqU9BVLJ3znHg0JElNH7ZAKAIAAEBHe/jwvCTRKfKgeiiiUwQAAAA00f2HZhQKGF2+IeV2KThJglAEAAAANNdisayv/3pM1186pFg46HY5OEmC5XMAAABAc33t/jHNLZb0e6+8wO1SsAKWzwEAAABNVK5U9dmfHtQLz+/VC8/vc7scrOBEp6jS1M8hFAEAAKAj/eCxCY3N5nTrK+gSeVXcGcnN8jkAAACgway1uu0nB3XBQFy/edmw2+XgFOIRls8BAAAATfGLA8f16Pi83vuKCxQIGLfLwSkEAkbdkSCdIgAAAKDR/sdPDmogEdUbt29yuxScQTwa0kKRUAQAAAA0zN6JtH785JTe/dLNjOH2gUQ0pEyeUAQAAAA0zGd/8pS6I0H97rXnu10KViEeZfkcAAAA0DCZfEnffeSI3nT1JvV0h90uB6uQiIYYyQ0AAAA0yg8em1ChXNWbrh5xuxSsUiIaYvocAAAA0Cjf+PWYtgzEtf28dW6XglVi0AIAAADQIONzOd13cEZv3L5JxjCG2y/i0RB7igAAAIBG+NaD45LEGG6fYfkcAAAA0ADWWn3j12O6ZnOfzuvrdrscrEE8ElK+VFW5Um3aZxCKAAAA0PYeHZ/XgakFvfFqukR+E4/WzpJq5gQ6QhEAAADa3jd+Pa5IKKDXXrnB7VKwRoloSJKUbeKwBUIRAAAA2lqpUtWdDx/Rb142rJ4uzibym0SsFoqaOWyBUAQAAIC29uMnpjSzUNSbWDrnS/F6p4hQBAAAAJydbz44rv54RK+4eNDtUnAW6svn6BQBAAAAZ+GZ44u6+/Fj+q1tGxUO8k9fP4pHCEUAAADAWfvLH+xRMGD0e6+80O1ScJaWBi0wfQ4AAABYm/sOHtcPHpvQH+y4UOt7Ym6Xg7NUH8mdzZea9hmEIgAAALSdStXqz77zuDat69J7X3GB2+XgHNQHLSwU6RQBAAAAq/a1+w9rz9G0PviaSxULB90uB+cgGgooFDBMnwMAAABWK5Mv6b/+yxP6jc29ev1VHNbqd8YYJWIhBi0AAAAAq/XpXft1fKGoP339FTLGuF0OGiAeCdEpAgAAAFbjiYmMPn/vIf2bq0d05UiP2+WgQRJROkUAAADAGWULZf3+lx5QqiusP955qdvloIHi0aAWGMkNAAAAnJq1Vn/8vx7RoekF/c1bt2swGXW7JDRQPMryOQAAAOC07vjF0/reI0f1n268RC++sN/tctBgCUIRAAAAcGoPHZ7TR773uF516ZB+7xUXul0OmiDOniIAAABgZcezBb3vS7/WUDKm//Y72xQIMG2uHfV0hTW3WJK1tinvTygCAACALz15LKM3fOZnmsoW9Jm3X6113RG3S0KTDKeiypUqyjSpW0QoAgAAgO/878eP6Y1/+zPlS1V95dbrtO28dW6XhCYaTsUkScfm8015f0IRAAAAfMNaq78bPaD3/sP9umAwoTv/8KXa/rxet8tCk613QtFEujmhKNSUdwUAAACa4L/9y5P69K79ev1VG/Txm7epKxJ0uyS0wPoeJxQ1qVNEKAIAAIAvPDY+r8+M7te/uXpE//XNV8kYhip0iqXlc03qFLF8DgAAAJ5XqlT1X77+iPoTUf3pb11OIOowsXBQPV1hls8BAACgc33u3qf0+NG0/v53r1ZPV9jtcuCC9amYjqULTXlvOkUAAADwtEPTC/rk3U/qxiuGtfP5G9wuBy4Z7omxfA4AAACdx1qrD33jUUWCAf35Tc93uxy4aH0q2rRBC4QiAAAAeNY//fIZ/eLgcX3otZctbbZHZ1qfimk6W1C5Um34e7OnCAAAAJ4zt1jU//f9Pfrq/WO67oI+veU3znO7JLhsuCemqpWmsgVt6Olq6HsTigAAAOAZ1lp979Gj+n/v3K3ZxZJ+f8eF+qNXbVUgwLS5Trd0gOt8nlAEAACA9pHJl7R/rqLJ+w/rwGRWDz4zp18emtGVm3r0xX93ja7Y2ON2ifCIZp5VRCgCAABAy80vlvTff7RPd/zikMpVK+kRRYIBbRmI6/953WV610s2KxRk+ztOWN9zolPUaIQiAAAAtEy5UtU///IZfeLuJzWXK+nfvug8DVcm9YYbXqzzersIQjilvu6IwkGjiSacVUQoAgAAQFM8Mjanu3ZPKFesqlCuKF+q6pGxOe2bzOq6C/r0p6+/QpdvTGl0dFRbBuJulwuPCwSMhpLNOauIUAQAAICGqlSt/v7HB/SJu5+UJHWFg4qGAoqFg+pPRPT3v/tC3XjFsIxheALWZn1PjOVzAAAA8LaJ+bz+w1ce0i8OHtfrr9qgj77xSvV0hd0uC21ifSqmPUfTDX9fQhEAAAAaYtcTk/oPX3lIxXJVH7/5Kt38whG6QWio4VRMu56YlLW2of9tEYoAAABwzh54ekb//o4HdNFQQp9+23ZdMJhwuyS0ofU9US0WK8oUykrFGteBJBQBAADgnByeWdStdzygTb1d+qf3Xqt13RG3S0KbWjqraD7f0FDEzEMAAACctXS+pH/3hV+pXLX63C0vIhChqdY7oWiiwRPoCEUAAAA4K+VKVe//pwf11PSC/u7tV7NkDk1XP8D1WIPPKmL5HAAAAFZtPlfS4ZlFPTOzqB8+NqEfPzmlv3zTlXrJRQNul4YOsLR8rsGdIkIRAAAATut4tqAv/PyQvvyrw5rKPPv/of+DHRfqrdc8z6XK0Gli4aDWdYcbflYRoQgAAAArOjyzqM/+9KC+8qvD+v/bu/coOcrzzuPfp2/T06MZzYzu6A4E4YABCRCwtgm+BmNnbUK8OMaOHbzLsY2ztne9WZzdZLPO7jl2NuuN18Sb49jYcDbGuTleXwgYHyJDgs1FICQEEkgCg4TuSHOf6duzf7xv97QGzUgjzUy3pn+fc+p09dtV1W/V09X9PlVvVY8Uy7ztdYtYv7qLFd05VnS3sby7lfYpvNhd5GQsas9O+TVFSopEREREmtzz+/v4+kMv8MgLhxkplsNQKDGQL5FOGtevXcotV5/DuQt1zZDU36K5WXWfExEREZHT5+78fNer/MVDu3hg2wGy6QRvXrOQ9myKllSSllSCrrYMN6xbVr24XaQRLO5oYdve3ildppIiERERkVnK3dlzdIiNvzjCppePsq9nmFcH8hwdLHB4YIRD/XnmtWX4zNvO40NXraS7TbfTlsa3uCPLof4RiqUyqeTU3ExbSZGIiIjILHKgb5gN2w7y0+cO8vgvXq3eurg1nWRpVyvduQwr5+VYu6KTi5d3cv3apWTTyTrXWuTkLZqbpexwsH+EJXNbp2SZSopEREREznAvHR7ku0/u5oFtB9i8uwcIR9OvPHsel67sYt2KLs5f3D5lR9VF6qn6B649w0qKRERERJqZu/PQ84e48+EXeWD7AQDWLu/ks+84j7ecv4jXLWnHzOpcS5GpNx3/VaSkSEREROQMMFIssfPAANv397JtXx/3P7OfXQcHmD8nwyfffC43XbFSN0SQplD5nE/lfxUpKRIRERFpEKWy89z+Pp586SgvvTrI/t5h9vYMsa9nmN1HhiiWHYB00rhoWSf/68Zzue71S2hJ6ZogaR7duQzppLGvd+TEE58kJUUiIiIidZIvltn08lH+eceh6h3i+keKAGSSCRZ2tLBkbpYLl87lXRctYc3iDs5f3M7q+W2kdX2QNKlEwljYPrX/VaSkSERERGSaFEpl9vUMc7B/hKF8KQyFEvt7h3l452F+vuswg/kSCYPzF3dw/dqlrFvZyboVXSzvypFI6JogkeNZPDer7nMiIiIiM61nqMCPNu/lgW37yaQSdLdlmNfWQndbhkKpzNHBAkeHwn8A7e8dZs+RIfb1DhN7vL3G6vlt3LBuGW84dz5XnTOPua3pmV0hkTPY4o4sz07hH7gqKRIREREhdGXbebCfA30jpBJGOpkglTQO9+f53qY93P/MfvLFMivn5UgnExzuH+HIYKE6f8KgM5ehszXN/PYWrjxnHsu6cizrbGVBRwu5dJLWTJLWdJK5uTQL23VTBJFTtagjyz9uP4C7T8ldFpUUiYiISNMplso8s7eXR3a9yuY9PWzf18uugwPVGxmM1d2W4QPrV3DDumVcuLSj2ggrlsocHSqQSSWYk0mpu5vIDFk8t4XBfImjgwW62jKnvTwlRSIiIjLjDvePsLdnmJXzcrRnT67bWLns7DzYz+O/OMLGOPQMFbjgrA4uWd7Jxcs6OW9RO/lSmcF8kYGREoP5Io/tLbLv0ZcYyJfoHSqwZU8Pj73wKn3xhgZLO1s5f3E7b3vdItYsbmdpZyvFslMsOYVymZZkgstWdZNJvfbGBqlkgvlzWqZ024jIiV26souEwUfvfIw7PvFmyCYAABKoSURBVHI5nbnTS4waJikys2uBLwNJ4Ovu/oU6V0lEREQid6dYdkplp1Aqx8djnxfLTrFcplgK4yOFEoP5EgP5IoMjJfb2DPP0Kz1s3dPDKzUXSC/tbGXN4nbOWdBGoeT0DRfpHS7QO1Sgd7hIXxzvGyni8UROd1uGdSu6mNeWYfOeHr66YSel8S7eAXhqS3X07AVt/NolZ3Hl2fO4cnU3CzvUjU3kTHPpym6+etM6/u3dm3jfn/+Muz66niVzW095eQ2RFJlZEvgz4O3AbuAxM/u+uz9T35qJiEi9uIcGd8mdchlK8Xk5Nr4TFo7SZ+J1H6mEvaZfubszmC9VG9l9w6GR3TtUoG+4SMKMbDpBazpJNpMknUhgRhgw0kkjm06Sy4RrQVKJBP0jRXqGKg32Ao+9VODZDTvpGQrLz6aTdLdl6Mpl6G5L05JKxmSiXE0qijGZqCQRhZJTLJWrZyfasyk6c2m6chk6c2EZqaSRsLCeyYSRSITxhBnFcpkjA6MX+Q/lS3TPybBgTgsL28ONAIplZ2CkyGD+2ERlMD9aVj27UijSE28WsL93hP29wxweyJ92TM3g7PltXL66mwvO6mBZV44XDg3w3P4+tu/r4592HKIlmaCjNU17NkVHa5qlna10tLbTkU3TkU2xrDvHZSu7WD2/7Zh4D+aLbH2ll10H+8mmk7RlUuRawuPWp57gmjdeRVtLilwmqVtZi8wS1164hG/dnOaWuzZyw1cf5q6PXsG5C+ec0rIaIikC1gM73H0XgJl9B3gPMG5SdHDQufUvnzjua84ER4qgepRpsq+daNknnvfU33eiuU/vfU+wrSacd+L3PfzqMHfsevSU33siE8bwNOM/VCjRP1Kkf7hI/0gRg+qPc3s2RWsmRcLAADOLjwAWG1LhecKs2rAqe+WIarnaXz1Z06BxYCg/2jjJl8qx4ZMgHRt7qWSi+pg+TuOv9qmNUx5rWR0/eGiYu19+vLpNKpsmPHfcQ1nZnbKHmLmH57WPTni9+rwyPU4qkSCbTpBNJ2mJXU9GimVGCmVGiqXq9qjWykZraDXbuTJNZX0Mq840trw6bqPrO1p2vGVRrX8pNlqr4x667FRilk7GBqkZQ4XR2+sO5UvVuCesMk34L4VkLEskjGTiONPY6DLHbtsjR4f4yrMPHzcGlYPiidrPm1n185moXPNQc/S+9gh/7RH9YqlMIT6ahfVMJ0KykU4m4hA/f9XP4mgDvRy3U0hgPCYwxysbTWqO3d7UJEDh8VS+IsL+EupqZvSPFCc+ezBVntlGOmm0Z9OMFEoM5EvT/57TqCWVoD2bZvHc8B85Fy/vZP6cTExAE9W4h/1hNCmtfk/Fz0dLKklbS4q2TJJcS4quXJpcZnqaHrlMistXdXP5qu7XvHZkZ4KzOk/9CLKINK5/cc58vnPLlXzkm49yw/95mAvO6jil5djpNEynipn9BnCtu//r+PxDwBXu/skx090C3AKQXXT2pWtv/coECz3Be57iaydyOne/ONGcp3NjjdNa31PclqVSiWRy4n/YnrBe07S+J1p2JmFkU5BLG9lkSAqGijBYcAaLTr4UyioJA68ZD2O101QbvhbGK/OUCQ1BM6MlCS1JyCSNdCI00ksOpcoR8up4GGp33fH24on2bwfK5TLJRCJukzFJFhyT5Nl4ZdUGONXlJOICjFDnfNkplKFQGv0n9nQC0glI1ryv166LH5vgVrbn2HUebzvUxuS4z495n5jIUonV6FBJMoDRsxVx+2eS0JIMsUsnrfo+ZYcyPjpeTS7jeLXMX/P62G1aKpVIp5LVZG/stq9dt2MfvTqetJCMVT5/qcRoWcJCeaqawIVlFqtnZqBY8zks1nwGq0mjj9a3dpsdb3smKq/VvG42Wo/K68ePx+hyK+teqcdoHUMSWKlXa8rIpSGXMnIpozXu260pI5cK2ytfgnwZRkphfWs/byUP5SMlyJfC+ufSYd62dFim5wdZ0NlGJjG6H+VLzkDB6c2HZSaMahKcPGZ9j41PKhG25VAR+gsehnx438pnpfK9Ua6pY9JCfeakjTmZ8JnszTs9I2HoyzupxOjnNZuqfOeE77va8kwi1HU26e/vZ86cUzt6LDNDMWp8jR6j/QNlvr0tz1Dx2LbP/bddt9HdLzvR/I1ypuikuPvXgK8BrFmzxh/+/evqXCOZyIYNG7jmmmvqXQ2ZgGLU+BSjxqcYNT7FqPEpRo3vTIjRje96bZnddnLzNkqn2j3A8prny2KZiIiIiIjItGqUpOgx4JfMbLWZZYD3A9+vc51ERERERKQJNET3OXcvmtkngfsIt+S+w9231rlaIiIiIiLSBBoiKQJw93uAe+pdDxERERERaS6N0n1ORERERESkLpQUiYiIiIhIU1NSJCIiIiIiTU1JkYiIiIiINDUlRSIiIiIi0tSUFImIiIiISFNTUiQiIiIiIk1NSZGIiIiIiDQ1JUUiIiIiItLUlBSJiIiIiEhTU1IkIiIiIiJNTUmRiIiIiIg0NSVFIiIiIiLS1JQUiYiIiIhIU1NSJCIiIiIiTU1JkYiIiIiINDUlRSIiIiIi0tSUFImIiIiISFNTUiQiIiIiIk3N3L3edTglZtYHbK93PWRC84FD9a6ETEgxanyKUeNTjBqfYtT4FKPGd6bGaKW7LzjRRKmZqMk02e7ul9W7EjI+M3tcMWpsilHjU4wan2LU+BSjxqcYNb7ZHiN1nxMRERERkaampEhERERERJramZwUfa3eFZATUowan2LU+BSjxqcYNT7FqPEpRo1vVsfojL3RgoiIiIiIyFQ4k88UiYiIiIiInLaGSYrM7A4zO2BmT48p/x0z22ZmW83sj2vKLzKzn8XyLWaWjeUbzGy7mW2Kw8KZXpfZ6ngxMrNLzOzncVs/bmbrY/k1ZtZTE4c/qJnnxRizTWb2eD3WZbaaZIzmmtkPzOypuB/99phldZjZbjO7fabXYzYbJ0YXx++zLTEmHTWvfc7MdsTvtV+tKf+UmT0dY/fpmV6P2WwyMTKz9TXfc0+Z2fVjlpU0syfN7IczvR6z2SRjtMrMhmri9Oc189xoZpvjfvTFeqzLbDXJGKXN7M5Y/qyZfW7MsrQfTQMzW25m/2hmz8R94FOxvNvM7jez5+NjVyw3M/vf8Tdps5mtq1nWF+Nv0tNmdmO91um0uHtDDMDVwDrg6ZqyNwM/AVri84XxMQVsBi6Oz+cByTi+Abis3uszG4dxYvRj4J1x/DpgQxy/BvjhOMt5EZhf7/WZjcMkY/R7wBfj+ALgVSBTM9+XgW8Dt9d7vWbTME6MHgN+JY7fDPxRHP9l4CmgBVgN7ASSwIXA00Aufh/+BDi33us2W4ZJxigHpOL4EuBA5Xks+3dxPzru96GGGYnRqtrpaqafB7wELIjP7wTeWu91my3DJGP0AeA7cTwX2wmraubTfjQ9MVoCrIvj7cBz8Xfnj4HbYvltNW2F64B/AAy4Engklr8LuD/+HrXFOHfUe/0mOzTMmSJ3f5DQKKv1ceAL7j4SpzkQy98BbHb3p2L5YXcvzVhlm9Q4MXKgclR7LvDKjFZKjjHJGDnQbmYGzInzFQHM7FJgESGhkik0TozOAx6M4/cDN8Tx9xAaCiPu/gKwA1gPvI7wYzTo7kXgp8CvT3vlm8RkYlQTA4AsYb8CwMyWERoLX5/WCjehSe5H4zkbeN7dD8bnPzmJeeQkTTJGDrSZWQpoBfJAL2g/mk7uvtfdn4jjfcCzwFLCb8+dcbI7gffG8fcAd3nwc6DTzJYQEqkH3b3o7gOEExfXzuCqTImGSYrGcR7wJjN7xMx+amaX15S7md1nZk+Y2e+Ome+b8RT578cGn0yfTwP/w8xeBv4EqD3lfVXsTvIPZnZBTbkDPzazjWZ2y0xWtkmNF6PbCY3rV4AtwKfcvWxmCeB/Ap+tR2Wb1FbCjw3A+4DlcXwp8HLNdLtj2dOE78Z5ZpYjHL1bjkyn8WKEmV1hZlsJ+9HHapKkPwV+FyjPZEWb2LgxAlbH7lc/NbM3xbIdwJrYvS5FaPhpP5pe48Xob4EBYC/h7N2fuHslodJ+NAPMbBWwFngEWOTue+NL+wgHSWH836SngGvNLGdm8wk9vc64fanRk6IU0E04RfcfgL+OSU4KeCNwU3y83szeGue5yd1fD7wpDh+a8Vo3l48Dn3H35cBngG/E8ieAle5+MfAV4Hs187zR3dcB7wRuNbOrZ7LCTWi8GP0qsAk4C7gEuD327/4EcI+7765HZZvUzcAnzGwjoQtDfqKJ3f1Z4IuEM3n3EuKos+XTa9wYufsj7n4BcDnwOTPLmtm7gQPuvrE+1W1K48VoL7DC3dcSu2GZWYe7HyF8P/4V8BChy5b2o+k1XozWE7b9WYSuwv/ezM7WfjQzzGwO8HfAp929t/Y1D/3jJrxVtbv/GLgHeBi4G/gZZ+C+1OhJ0W7gu/E03aOEowTzY/mD7n7I3QcJgVgH4O574mMfof/p+rrUvHl8GPhuHP8b4vZ2915374/j9wDpePSgNkYHgL9HMZpux40R8NuM7l87gBeA84GrgE+a2YuEM0u/ZWZfmNkqNxd33+bu73D3Swk/KDvjS3s49mjbsliGu3/D3S9196uBI4S+4DJNJohR7TTPAv2Ea77eAPzLuB99B3iLmf3fGaxy0xkvRrH76eE4vjGWnxef/8Ddr3D3q4DtaD+aVhPsRx8A7nX3Qmwb/DNwGdqPpp2ZpQkJ0V+6e6WtsD92iyM+Vi5fmeg36b+7+yXu/nbCNUdn3L7U6EnR9win4DCz84AMcAi4D3h9PE2XAn4FeMbMUpWGdwzyuwndTGT6vELY/gBvAZ4HMLPFla6LFu52lgAOm1mbmbXH8jbC9WGK0fQ6bowIXRTeCmBmi4A1wC53v8ndV7j7KkIXurvc/baZrXJzsXiXzNh18T8DlbtjfR94v5m1mNlq4JeAR8fMs4JwPdG3Z7rezWS8GJnZ6vg7hJmtJBxYeNHdP+fuy+J+9H7gAXf/YF0q3yQmiNECM0vG8bMJ+9GuMfN0Ec6S67qVaTTBd91LhN+nStvgSmCb9qPpFdtp3wCedfcv1bz0fcIBVeLj/6sp/614F7orgR5332vh7oDz4jIvAi7iDLwmOVXvClSY2d2EO5bNN7PdwH8B7gDusHA7xzzw4Xga74iZfYlwdwsndPX5UdyR7osJUZJw0eRfzPzazE7jxOjfAF+OjYJhoHKN0G8AHzezIjAEvN/dPTa+/z7mSyng2+5+78yuyew1yRj9EfAtM9tCOKrzH9390MzXurmME6M5ZnZrnOS7wDcB3H2rmf018AzhJhi31txU5u/ij1Ahlh+dwdWY1SYTI0IX7tvMrEDozfAJ7UfTb5Ixuhr4fE2MPlZzvcqXzeziOP55dz/jjm43qknG6M8I14NvJfwefdPdN89wlZvRGwiXmWwxs02x7PeALxAuWfko8AvgX8XX7iFcw7oDGCT0OAFIAw/Ftl0v8MGaayvPGBZyDBERERERkebU6N3nREREREREppWSIhERERERaWpKikREREREpKkpKRIRERERkaampEhERERERJqakiIREakbMyuZ2SYze9rM/sbMcieYfoOZXTam7AIze87MWmvKfmRmvzld9RYRkdlFSZGIiNTTUPwX9AsJ/0f3sckuwN23Ev7z5D8BmNl7gbS73306Fav8KauIiMx+SopERKRRPASca2ar4p92A2BmnzWzP6yZ7kM1Z5fWx7LPA+8zs0sIfzx4a5x3jpl9y8weNbMnzezXYvk5ZvZQLNtoZlfE8rfFs1E/BLZM/yqLiEgj0FEwERGpu3hW5p3AvScxec7dLzGzq4E7gAvdfdDMPgs8CHzJ3Z+P0/4BcK+7f8TMuoBHzOx+YC/wdncfNrPzgTuBK+I8lwG/7O4vTd0aiohII1NSJCIi9dRqZpvi+EPAN4CzTjDP3QDu/qCZdZhZp7sfdfcfmNlR4Ks1074DeKeZ3RafZ4EVwEHgdjO7GCgC59TM8zMlRCIizUVJkYiI1NOQu19SW2BmRY7t3p0dM49P8Lwch+rigPe6+84x7/HfgJeBDwJpoL/m5YGTrr2IiMwKuqZIREQazX5goZnNM7MW4N1jXr8RwMzeCPS4e88Ey7oP+J3KEzNbG0fnAnvd3YEPE5InERFpUjpTJCIiDcXdC2b2eeBRYA+wbcwkw2b2JOEMz80nWNx/Bf7UzLYQDgTuAN4D3A78rZndDPwIGJnCVRARkTOMhYNkIiIiIiIizUnd50REREREpKkpKRIRERERkaampEhERERERJqakiIREREREWlqSopERERERKSpKSkSEREREZGmpqRIRERERESampIiERERERFpav8fnwZQ28ofuJcAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 1008x720 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "matched_dfs[['count', 'PubYear']].groupby('PubYear').sum().plot(figsize = (14, 10), grid = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[\"['This model was combined into a single protein chain\",\n",
       " 'built and refined iteratively against native data using ARP-WARP',\n",
       " 'COOT',\n",
       " \"and PHENIX refine.'\",\n",
       " \"'This model was built and refined iteratively using COOT and PHENIX refine.'\",\n",
       " \"'Additional &agr;-helical density for the M domain was observed and &agr;M1–&agr;M1b (residues 327–352) and &agr;M2–&agr;M4 (residues 374–431) were built by several rounds of manual model rebuilding in COOT and refinement (see below) in CNS.'\",\n",
       " \"'An initial model of HEN1 was manually built with the programs O and Coot using the locations of SeMet positions as guides.'\",\n",
       " \"'The structure was solved by molecular replacement\",\n",
       " 'subsequent model building and refinement (including TLS refinement) was performed with COOT and CCP4 (refs 31',\n",
       " \"32).'\",\n",
       " \"'The atomic model was built manually with Coot and refined with PHENIX.'\",\n",
       " \"'The initial structure model was completed and improved through iterative refinement in Phenix and manual rebuilding of all the loops and several parts in the transmembrane region in Coot.'\",\n",
       " \"'The models were rebuilt using COOT and iterative structure refinement with restrained and TLS options was performed using PHENIX.'\",\n",
       " \"'The final models were built in O and Coot\",\n",
       " \"and refined with TLS parameters in REFMAC.'\",\n",
       " \"'Then\",\n",
       " \"a crude model was manually built in the Coot program.'\",\n",
       " \"'The remaining structure was traced by Coot independently into the electron density map\",\n",
       " \"which was calculated with partial phases.'\",\n",
       " \"'The model was refined by manual adjustment in Coot until rigid body refinement in Phenix yielded R-factors of 30.8&percnt; and 34.7&percnt; for the working and the free R-factor\",\n",
       " \"respectively.'\",\n",
       " \"'Model building and refinement were performed with Coot and Refmac5 (ref.'\",\n",
       " \"'Model building was performed in Coot\",\n",
       " \"and refinement was performed in phenix.refine.'\",\n",
       " \"'Rounds of model building with Coot and refinement with REFMAC5 produced a complete model\",\n",
       " \"which was used to phase the native data set to 2.4\\\\u2009Å resolution.'\",\n",
       " \"'The electron density allowed us to trace the polypeptide chain of the CSR using the graphics program COOT.'\",\n",
       " \"'After several iterative rounds of simulated annealing\",\n",
       " 'individual B factor refinement',\n",
       " 'and manual adjustments using COOT',\n",
       " 'the UDP and waters were added',\n",
       " \"and the structure refined to an Rwork of 21% and an Rfree of 24%.'\",\n",
       " \"'Some loops and all the RNA were rebuilt manually in COOT and the complex was refined with REFMAC.'\",\n",
       " \"'The experimentally phased NCS averaged electron density map permitted further manual rebuilding of one of the four monomers using COOT.'\",\n",
       " \"'Iterative rounds of model building and refinement were performed using COOT\",\n",
       " \"CNS and REFMAC5.'\",\n",
       " \"'These regions were manually fit in ‘Coot’ followed by simulated annealing in real space using Rsref2000 (ref.'\",\n",
       " \"'The model was built with BUCCANEER and Coot.'\",\n",
       " \"'Additional missing residues in the automatically built model were added manually in Coot.'\",\n",
       " \"'Molecular modelling and visualization was done using the Coot and Chimera software packages.'\",\n",
       " \"'Fitting and manual adjustments of the models was done using UCSF Chimera and Coot.'\",\n",
       " \"'The structure of the Ago ternary complex was manually built with COOT and O\",\n",
       " 'cycled with CNS refinement',\n",
       " \"to a final Rwork of 22.5% and Rfree of 28.2%.'\",\n",
       " \"'Iterative model building was performed in Coot and the X-ray structure was refined in PHENIX.'\",\n",
       " \"'Xfit and Coot were used for graphical map inspection and manual rebuilding of atomic models.'\",\n",
       " \"'Iterative rounds of coordinate and B-factor refinement were done with Refmac\",\n",
       " \"interspersed with manual model building in Coot.'\",\n",
       " \"'Model building was done in Coot.'\",\n",
       " \"'Model building and refinement were accomplished using Coot and PHENIX.refine.'\",\n",
       " \"'Model building was performed with Coot.'\",\n",
       " \"'The model was refined by alternating cycles of building in COOT and automatic refinement in Phenix or Refmac.'\",\n",
       " \"'All modelling building was carried out using Coot.'\",\n",
       " \"'Model building was done using the program COOT and initial structure refinement performed with the program CNS.'\",\n",
       " \"'Following the determination of the initial structure by molecular replacement\",\n",
       " 'rigid body refinement and simulated annealing were performed in Phenix and BUSTER',\n",
       " \"followed by restrained refinement and manual rebuilding in Coot.'\",\n",
       " \"'The electron density calculated to 2.38\\\\u2009Å was sufficient for model building with the program COOT.'\",\n",
       " \"'A further 32 residues were added manually in Coot\",\n",
       " \"and the improved model was used to calculate an improved map using combined experimental and model phases in PHASER.'\",\n",
       " \"'Further refinement was conducted using multidomain\",\n",
       " 'non-crystallographic-symmetry (NCS)-restrained',\n",
       " 'simulated annealing in PHENIX',\n",
       " 'fourfold multidomain NCS averaging with a custom solvent mask (including the region of DNA binding)',\n",
       " 'density modification using resolve',\n",
       " \"and manual model building in COOT.'\",\n",
       " \"'Both models were subjected to iterative rounds of building with COOT and refinement with REFMAC.'\",\n",
       " \"'(beja@tx.technion.ac.il).\\\\r\\\\n\\\\r\\\\n&e090910-8; &nature08284-s1;\\\\r\\\\n\\\\n\\\\nCyanobacteria of the Synechococcus and Prochlorococcus genera are important contributors to photosynthetic productivity in the open oceans.'\",\n",
       " \"'Structure refinement used REFMAC and models were rebuilt with COOT.'\",\n",
       " \"'The model was further rebuilt in COOT and refined with PHENIX.'\",\n",
       " \"'The initial model was built using Buccaneer\",\n",
       " 'and the model was finished using COOT',\n",
       " \"and the structure was iteratively refined with REFMAC5.'\",\n",
       " \"'The crystallographic model was built using COOT\",\n",
       " 'refined using REFMAC5 and Crystallography and NMR system (CNS)',\n",
       " \"and deposited in the RCSB Protein Data Bank (PDB code 4OUD).'\",\n",
       " \"'An initial model was built manually using COOT.'\",\n",
       " \"'An initial model was built into the experimental electron density map using COOT.'\",\n",
       " \"'The missing side chains were built manually in COOT\",\n",
       " \"and the missing loops were built with the program SLOOP by using fragments from the Richardson’s Top500 library of structures to fill the gaps.'\",\n",
       " \"'Final models were obtained after several cycles of refinement with REFMAC and PHENIX and manual re-building using COOT.'\",\n",
       " \"'Model building was done using program COOT.'\",\n",
       " \"'The model was completed by iterative cycles of manual model building and real space refinement using the program Coot and crystallographic refinement using Phenix.refine.'\",\n",
       " \"'Once the subunit arrangement was confirmed\",\n",
       " 'iterative cycles of manual rebuilding in Coot',\n",
       " \"jelly body refinement in Refmac and further restrained refinement in Phenix were performed.'\",\n",
       " \"'The RNA model was built in COOT and refined in PHENIX and REFMAC using 2.3\\\\u2009Å native data set of the ligand-bound fluoride riboswitch.'\",\n",
       " \"'The manual model building and refinement were performed in COOT and PHENIX in an iterative manner.'\",\n",
       " \"'The model was refined through iterative rounds of model building in COOT and refinement in Phenix.'\",\n",
       " \"'This map was used to place VPS29 in the subcomplex structure and the VPS35 C terminus was built by hand in COOT.'\",\n",
       " \"'Refinement of the structure was undertaken with Phenix\",\n",
       " \"with iterative model building using Coot.'\",\n",
       " \"'The model was traced using COOT and refined using CNS and SHELX.'\",\n",
       " \"'The model was built by using COOT and refined with the program REFMAC5 from the CCP4 suite.'\",\n",
       " \"'The initial model of the extracellular domain was rebuilt and refined using programs in the CCP4\",\n",
       " \"COOT and PHENIX packages.'\",\n",
       " \"'Final models were obtained using a combination of ERRASER for automated model building\",\n",
       " 'manual placement of residues in COOT',\n",
       " \"and iterative cycles of restrained refinement in PHENIX and Refmac5 (ref.'\",\n",
       " \"'The remainder of the molecule was traced manually with COOT and then by full refinement using PHENIX.'\",\n",
       " \"'The superposition of the model with their respective templates was made using Coot\",\n",
       " \"and the figure was created using Chimera.'\",\n",
       " \"'The resulting model was completed through iterative rounds of rebuilding in Coot and refinement in PHENIX.'\",\n",
       " \"'Manual model building and refinement were carried out using Coot and phenix.refine\",\n",
       " \"and Molprobity was used to monitor and improve protein geometry.'\",\n",
       " \"'The structure of the ADD–CD–CDNMT3L-H3 complex was determined by molecular replacement using the ADD domain (PDB accession number 3A1B) and CD–CDNMT3L (PDB accession number 2QRV) as searching models in the PHASER program\",\n",
       " \"and was then manually built by COOT.'\",\n",
       " \"'The remainder of the model was built manually using the program Coot.'\",\n",
       " \"'The final model rebuilding was performed using COOT and the protein structure was refined with PHENIX against the high-resolution native data using non-crystallographic symmetry and stereochemistry information as restraints.'\",\n",
       " \"'Model building was performed with XtalView followed by refinement with REFMAC and iterative rebuilding and model improvement with COOT.'\",\n",
       " \"'The models giving the top two solutions (as judged by both TF-Z and LLG metrics) were used as a single ensemble for phase calculation followed by manual building using COOT and reciprocal space refinement using phenix.refine.'\",\n",
       " \"'Density modification and two-fold symmetry averaging produced experimental maps that permitted most secondary structural elements to be manually assigned with COOT.'\",\n",
       " \"'The structures were completed by refinement in Refmac5 and rebuilding in Coot.'\",\n",
       " \"'Four of the copies were fitted using Molrep\",\n",
       " \"and the remaining two were fitted manually in Coot.'\",\n",
       " \"'The model was built using COOT into 2Fo\\\\u2009−\\\\u2009Fc maps\",\n",
       " 'cross-validated using simulated annealing composite omit maps and refined with PHENIX using non crystallographic symmetry',\n",
       " \"individual atomic displacement parameters and translation/libration/screw (TLS) subdomains.'\",\n",
       " \"'Further changes were made to the LeuT portion using COOT.'\",\n",
       " \"'This part of the model was rebuilt in Coot using the apo structure as a reference.'\",\n",
       " \"'Standard refinement was performed with the programs Phenix and Coot.'\",\n",
       " \"'Model-building was subsequently completed manually using COOT (version 0.3.3) with subsequent refinement (REFMAC5) using NCS and B-factor restraints.'\",\n",
       " \"'Further model improvements were performed using Coot and refined using the Refmac program in CCP4 (ref.'\",\n",
       " \"'Refinement of the structure was undertaken with Phenix\",\n",
       " \"with iterative model building using Coot.'\",\n",
       " \"'The built model was refined through alternating cycles using the Coot and CNS programs.'\",\n",
       " \"'The electron density calculated to 3.0\\\\u2009Å was sufficient for building the model of AcrIIA4 with the program COOT.'\",\n",
       " \"'Model building was carried out using COOT.'\",\n",
       " \"'Additional missing residues in the auto-built model were manually added in COOT.'\",\n",
       " \"'Additional missing residues in the auto-built model were manually added in COOT with the aid of the map sharpening utility.'\",\n",
       " \"'Manual model building was carried out for the Fab using the program Coot\",\n",
       " \"followed by structure refinement using the program Refmac.'\",\n",
       " \"'Subsequent model building and refinement were performed with Coot and REFMAC5.'\",\n",
       " \"'Coordinates\",\n",
       " 'atomic displacement parameters',\n",
       " \"and occupancy (where appropriate) were refined using phenix.refine with manual rebuilding using COOT.'\",\n",
       " \"'The electron density was fitted and refined in multiple iterations using Phenix.refine and Coot.'\",\n",
       " \"'After tracing the initial model manually with the program Coot\",\n",
       " 'the model was refined against the native data at 3.52\\\\u2009Å resolution with TLS refinement techniques adopting a single TLS body as implemented in PHENIX',\n",
       " \"leaving 5% of reflections in the Rfree set.'\",\n",
       " \"'Iterations of refinement with PHENIX and CNS and model building in Coot led to the current model\",\n",
       " \"which consists of 1,025 residues with excellent geometrical characteristics (Supplementary Table 1).'\",\n",
       " \"'Coot was used for structure visualization and calculation of NCS-averaged maps.'\",\n",
       " \"'The initial Cα model was remodelled manually with Coot and O and refined initially using PHENIX.'\",\n",
       " \"'The identification and analysis of the protein–DNA–RNA hydrogen bonds and van der Waals contacts were performed with the Protein Interfaces\",\n",
       " \"Surfaces and Assemblies service (PISA) as implemented in the CCP4 suite and Coot.'\",\n",
       " \"'Model building was done using COOT\",\n",
       " \"and refinement was done with CNS.'\",\n",
       " \"'Further model building was done manually in COOT\",\n",
       " \"followed by refinement in CNS 1.2 (ref.'\",\n",
       " \"'The initial output model (∼70% completeness) from Autobuild was then manually corrected and completed with COOT.'\",\n",
       " \"'We improved the initial model by iteratively building regions of the receptor in Coot and refining in Phenix.'\",\n",
       " \"'Model building was performed with the program Coot\",\n",
       " \"and refinement was done with REFMAC5 with TLS (CCP4i) or PHENIX (Extended Data Table 1).'\",\n",
       " \"'All structural models in the figures were superimposed with the bridge helix region (Rpb1 822–840) near the active site using Coot and PyMOL.'\",\n",
       " \"'Manual rebuilding of the MBH model and electron density interpretation was performed after each refinement cycle using the program COOT.'\",\n",
       " \"'This model was manually edited in COOT and a poorly modelled loop spanning residues 821–838 was removed.'\",\n",
       " \"'The model was built with program COOT and was improved by several cycles of manual rebuilding and refinement with CNS and REFMAC.'\",\n",
       " \"'The crystallographic model was extended and completed by repeated cycles of manual building in Coot and crystallographic refinement with PHENIX Refine.'\",\n",
       " \"'The homology models were placed in the asymmetric unit followed by iterative rounds of refinement in REFMAC against the Native-1 data set\",\n",
       " 'employing the ‘Jelly-Body’ and ‘secondary structure restrains’ refinement options',\n",
       " \"and manual rebuilding in COOT.'\",\n",
       " \"'Iterative model building and refinement were performed with JLigand\",\n",
       " 'COOT',\n",
       " \"and PHENIX.'\",\n",
       " \"'Manual model building was with O and Coot.'\",\n",
       " \"'Restraints for 5-OP-RU and 5-OE-RU were generated using the Grade Web Server\",\n",
       " \"with model building performed in COOT using MolProbity for validation.'\",\n",
       " \"'The resulting structure was iteratively refined in Phenix and manually rebuilt in Coot.'\",\n",
       " \"'The model was built manually in COOT and the structure was refined with PHENIX.'\",\n",
       " \"'Consequently\",\n",
       " \"alanine helices were initially placed in the clearly visible rods of electron density using COOT and refined individually as rigid bodies in REFMAC.'\",\n",
       " \"'We superposed the kinase structure from PI(3)K&ggr; on the head/crown domain of the DNA-PKcs crystal structure\",\n",
       " 'using COOT',\n",
       " 'and obtained a convincing fit to the &bgr;-strands of the N-lobe',\n",
       " 'which fit into the flat density present',\n",
       " \"as well as the helices that dominate the C-lobe as shown in Fig.'\",\n",
       " \"'Highly idealized alanine &agr;-helices of various lengths and curvatures could be fitted into the electron density maps using COOT.'\",\n",
       " \"'34) and manual model-building using COOT.'\",\n",
       " \"'Subsequent rounds of manual building of BT2264 and the N-terminal domain of BT2262 within COOT depended critically on using density-modified maps produced by RESOLVE\",\n",
       " \"in particular after application of two-fold non-crystallographic symmetry (NCS).'\",\n",
       " \"'The model was built using COOT and iteratively refined using CNS1.3 with a deformable elastic network\",\n",
       " \"followed by 10 cycles of TLS refinement using 1 TLS group per domain in refmac5 (ref.'\",\n",
       " \"'The model was refined by a rigid body fit of the transmembrane domain in PHENIX and Coot applying ten-fold NCS operations to the receptor chains.'\",\n",
       " \"'The PHENIX-model was further manually rebuilt using COOT and refined using REFMAC5 in successive cycles.'\",\n",
       " \"'The initial model was traced automatically using the program BUCCANEER\",\n",
       " \"and was manually rebuilt in COOT.'\",\n",
       " \"'Model building was guided by inspection of &sgr;&Agr;-weighted 2Fo\\\\u2009−\\\\u2009Fc and Fo\\\\u2009−\\\\u2009Fc omit maps\",\n",
       " \"and was carried out using COOT.'\",\n",
       " \"'Structures were built with Coot and refined with Refmac.'\",\n",
       " \"'Structures were built with Coot and refined with Refmac.'\",\n",
       " \"'Modification and further building of the structure was carried out in O and COOT.'\",\n",
       " \"'An initial model was built into the high-resolution P42 native data using COOT.'\",\n",
       " \"'The atomic models were manually built using Coot and iteratively refined using REFMAC5 (REFMAC5 and PHENIX were used for refinement of eV1).'\",\n",
       " \"'An initial model was built into the experimental map by using COOT.'\",\n",
       " \"'An initial model was built into the experimental map with COOT.'\",\n",
       " \"'The final cycles of model building and refinement\",\n",
       " 'including TLS refinement',\n",
       " \"were performed with Coot and REFMAC5 (ref.'\",\n",
       " \"'Refinement in CNS\",\n",
       " 'automatic rebuilding in Phenix',\n",
       " 'and manual rebuilding in Coot were carried out',\n",
       " \"followed by placing of ligands in Coot.'\",\n",
       " \"'Manual rebuilding of the Meta\\\\u2009II models and electron density interpretation were performed after each refinement cycle using the program COOT.'\",\n",
       " \"'Automatic model building was done in Arp/wArp and completed manually in the program COOT.'\",\n",
       " \"'Density modification using programs in the CCP4 suite improved the initial phases\",\n",
       " \"and the structure was refined using REFMAC and COOT.'\",\n",
       " \"'Iterative model building and refinement was performed using COOT and REFMAC5 (ref.'\",\n",
       " \"'The model was built using COOT and was refined using PHENIX.'\",\n",
       " \"'Structure refinement was carried out with the programs CNS and Refmac\",\n",
       " \"and programs O and Coot were used for manual model rebuilding.'\",\n",
       " \"'Model building with COOT was alternated with successive rounds of restrained refinement using REFMAC.'\",\n",
       " \"'Model building with COOT was alternated with successive rounds of restrained refinement with REFMAC and solvent flattening with DM.'\",\n",
       " \"'Model building and refinement of IDE-substrate complexes were done with COOT and CNS.'\",\n",
       " \"'Iterative model building and refinement was conducted with COOT and PHENIX.'\",\n",
       " \"'The ZP-N model was built manually with Coot and the backrub tool of KiNG\",\n",
       " 'using 3.2\\\\u2009Å resolution SigmaA-weighted difference Fourier maps',\n",
       " 'composite annealed omit maps calculated with CNS using a starting temperature of 4,500\\\\u2009K to minimize model bias',\n",
       " \"and X-ray holographic image reconstruction in Eden.'\",\n",
       " \"'The models were adjusted according to features of the electron density using Assemble\",\n",
       " \"molecular dynamic flexible fitting (MDFF) in visual molecular dynamics (VMD) and Coot.'\",\n",
       " \"'Using UCSF Chimera and Coot\",\n",
       " \"they were fitted rigidly and adjusted into the densities.'\",\n",
       " \"'Iterative rounds of restrained refinement in Refmac5 and manual rebuilding in Coot were used to improve the model.'\",\n",
       " \"'The resulting model was then iteratively refined by building regions of &bgr;-arrestin-1\",\n",
       " \"V2Rpp and Fab30 in Coot and refining in Phenix.'\",\n",
       " \"'The structure was determined by molecular replacement using Phaser with duck 2CARD and monoubiquitin (from Protein Data Bank (PDB): 4A2W and 1UBQ\",\n",
       " 'respectively)',\n",
       " \"and refined using Coot and Phenix Refine.'\",\n",
       " \"'The structure was refined by an iterative process of manual model building using Coot and simulated annealing\",\n",
       " 'individual site',\n",
       " \"TLS refinement with torsion NCS and bulk solvent scaling using Phenix Refine.'\",\n",
       " \"'The structure was solved in PHENIX\",\n",
       " \"built in COOT and refined using REFMAC.'\",\n",
       " \"'Model building was carried out using COOT\",\n",
       " \"and the structure was refined with PHENIX iteratively.'\",\n",
       " \"'Model building of the Lsm–RNA complex was carried out using COOT\",\n",
       " \"and the structure was refined with PHENIX iteratively.'\",\n",
       " \"'The best solution was used as input into autoSHARP for phase determination (phasing power for anomalous data 0.838\",\n",
       " 'and FOMacentric 0.33 before solvent flattening)',\n",
       " 'which resulted in an excellent electron density map',\n",
       " \"into which one monomer of the protein was built manually using Coot.'\",\n",
       " \"'To complete the model\",\n",
       " \"manual building was carried out in COOT.'\",\n",
       " \"'For water picking and electron density and structure visualization\",\n",
       " \"the software COOT was used.'\",\n",
       " \"'The model was built using Coot and refined in PHENIX.'\",\n",
       " \"'The resulting map was in good quality and the partial model was manually rebuilt in COOT.'\",\n",
       " \"'Using UVR8(W285A) coordinates as the initial search model\",\n",
       " \"crystal structures of W285F and wild-type UVR8 were solved by molecular replacement using PHASER and refined with COOT and PHENIX.'\",\n",
       " \"'The resulting map was in good quality and the partial model was manually rebuilt in COOT.'\",\n",
       " \"'Using UVR8(W285A) coordinates as an initial model\",\n",
       " \"crystal structures of W285F and wild-type UVR8 were solved by molecular replacement using PHASER and manually refined with COOT and PHENIX.'\",\n",
       " \"'The initial structure for DAXX–H3.3–H4 was solved by molecular replacement program BALBES with a final search model based on Protein Data Bank 2HUE and manually refined and built using Coot and Refmac5 in CCP4 program suite.'\",\n",
       " \"'\\\\n'\",\n",
       " \"'The model was further built manually using COOT and refined using PHENIX.'\",\n",
       " \"'The models were subjected to repeated rounds of building with Coot and refinement with CNS.'\",\n",
       " \"'RNA nucleotides were modelled using COOT and the RCrane plugin.'\",\n",
       " \"'The resulting phases were of sufficient quality to manually build protein and RNA in Coot.'\",\n",
       " \"'Structure determination and representation made use of XDS\",\n",
       " 'MOSFLM',\n",
       " 'PHYRE',\n",
       " 'PHASER',\n",
       " 'PHENIX',\n",
       " 'RAVE',\n",
       " 'COOT',\n",
       " 'Diffraction Anisotropy Server',\n",
       " 'PYMOL',\n",
       " 'STRAP',\n",
       " 'ESPrit',\n",
       " \"Hollow and CCP4 programs.'\",\n",
       " \"'The structure was finalized with cycles of model building with COOT and refinement with PHENIX and REFMAC (CCP4 package).'\",\n",
       " \"'Finally\",\n",
       " 'we added step-by-step MG8',\n",
       " 'MG7',\n",
       " '&agr;′NT',\n",
       " 'anchor and C345C from C3c and CUB from C3',\n",
       " 'respectively',\n",
       " \"using molecular graphics in Coot with rigid-body refinement in Phaser.'\",\n",
       " \"'Model building was done using Coot and restrained translation–libration–screw refinement using REFMAC5 (ref.'\",\n",
       " \"'The atomic model was built with the program Coot.'\",\n",
       " \"'The structure of the full-length GluA2cryst was iteratively built and refined using the computer graphics program Coot and Phenix\",\n",
       " \"respectively (Supplementary Table 2).'\",\n",
       " \"'Model building and refinement were performed with COOT\",\n",
       " \"PHENIX and BUSTER.'\",\n",
       " \"'Structure was manually rebuilt and modified in Coot and refined in Refmac5.'\",\n",
       " \"'Each round of refinement was interspersed with electron density inspection and manual model building with COOT.'\",\n",
       " \"'Most of the residues were built automatically by the program Resolve\",\n",
       " \"and the model was completed by manual building with the programs O and Coot.'\",\n",
       " \"'Model building of the BoNT/B–Syt-II complex was done with Coot and refinement with REFMAC5 using six TLS (translation\",\n",
       " 'libration',\n",
       " \"screw) groups.'\",\n",
       " \"'These starting fragments were extended manually using Coot (v. 0.3.3).'\",\n",
       " \"'The structure was refined using REFMAC5 and manual alteration of the model carried out using COOT.'\",\n",
       " \"'The model was built in Coot and includes all residues with the exception of the terminal residues 1–12 and 760–788 of BcsA and 19–53 and 721–725 of BcsB.'\",\n",
       " \"'The model was built in Coot via an iterative process involving careful model building\",\n",
       " 'combination of experimental and model phases',\n",
       " \"density modification and restraint refinement in Refmac-5 (including TLS parameters for the final model).'\",\n",
       " \"'The models were built using COOT and structural refinement was performed using the program PHENIX.'\",\n",
       " \"'To avoid model bias from the molecular replacement solution\",\n",
       " 'the initial model was adjusted to fit the experimentally phased map',\n",
       " 'then subjected to several cycles of rebuilding and refinement performed with the programs Coot',\n",
       " \"O and Refmac5 (ref.'\",\n",
       " \"'The initial models were adjusted to cryo-EM density (in cases when homology models were useful) or built manually in COOT.'\",\n",
       " \"'\\\\n'\",\n",
       " \"'The coordinates of the ELOB–ELOC complex structure from the top solution of molecular replacement were combined with those of an nCUL5 molecule that was shown to interact with ELOB–ELOC in COOT.'\",\n",
       " \"'The remainder of the model was built using XTALVIEW/XFIT and COOT.'\",\n",
       " \"'Successive rounds of model building using COOT and TURBO-FRODO were followed by refinement using CNS and REFMAC.'\",\n",
       " \"'The rest of the model was built and fitted using the program COOT.'\",\n",
       " \"'The 143H282 structure was built manually using the program Coot and refined with the program Phenix to 2.4\\\\u2009Å resolution.'\",\n",
       " \"'The program COOT was used for model building and for making adjustments.'\",\n",
       " \"'Model building and local real-space refinement were performed in Coot\",\n",
       " \"PyMol (DeLano Scientific) and RSRef.'\",\n",
       " \"'\\\\n'\",\n",
       " \"'The side chains were manually exchanged with COOT by using the metagenome sequence D1JBK2-4 (Supplementary Fig.'\",\n",
       " \"'Model building was performed in COOT and MAIN.'\",\n",
       " \"'Phases were extended to the native eIF3d data set processed to ~1.40\\\\u2009Å using SOLVE/RESOLVE\",\n",
       " \"and model building was completed in Coot before refinement with PHENIX.'\",\n",
       " \"'The atomic model of the complex was built using Buccaneer and Coot\",\n",
       " \"and refined using Refmac 5 (ref.'\",\n",
       " \"'The rest of the model was built manually\",\n",
       " 'using Coot and Refmac',\n",
       " \"where experimental phases were used as an additional restraint.'\",\n",
       " \"'Several cycles of manual rebuilding\",\n",
       " 'using COOT',\n",
       " \"and refinement using REFMAC5 and PHENIX were carried out to improve the electron density map.'\",\n",
       " \"'Model building was carried out using COOT.'\",\n",
       " \"'Additional missing residues in the auto-built model were manually added in COOT.'\",\n",
       " \"'The model of protein and DNA was built and refined up to 3\\\\u2009Å by multiple rounds of manual fitting in O or COOT and positional refinement in CNS.'\",\n",
       " \"'The initial structural model was built with Xtalview and COOT.'\",\n",
       " \"'The initial model was completed through several cycles of manual model rebuilding in COOT and refinement in Refmac5 for the structure obtained with partially proteolysed proteins\",\n",
       " \"whereas the high-resolution structure obtained with engineered proteins was subjected to automatic model rebuilding using the program ARP/wARP.'\",\n",
       " \"'Electron density maps were generated using CNS and the model was built using COOT.'\",\n",
       " \"'Model building was performed with COOT.'\",\n",
       " \"'Model building and refinement was performed using COOT and PHENIX\",\n",
       " \"respectively (see Extended Data Fig.'\",\n",
       " \"'Model building and refinement were performed using COOT and PHENIX\",\n",
       " \"respectively.'\",\n",
       " \"'The structure of the Topo II&Dgr;NC&dot;DNA complex was built using Coot and refined using PHENIX to a final Rwork of 22.9% and Rfree of 27.7%.'\",\n",
       " \"'The model was further rebuilt in COOT and refined with PHENIX.'\",\n",
       " \"'The model was refined iteratively in Phenix with manual rebuilding in Coot with and without averaging of non-crystallographic symmetry.'\",\n",
       " \"'The model was built using Coot\",\n",
       " \"and refinement was done using Refmac after excluding 5% of the data for the R-free calculation.'\",\n",
       " \"'Iterative cycles of model building using program COOT and refinement using CNS and a final TLS refinement\",\n",
       " 'all using the same test data set',\n",
       " 'produced a final model (Rfree 26.9%',\n",
       " 'Rwork 23.9%) that contains CRM1 residues 53–57',\n",
       " '62–91',\n",
       " '96–115',\n",
       " '121–179',\n",
       " '187–387',\n",
       " \"402–429 and 447–1061 and SNUPN residues 1–66 and 94–293.'\",\n",
       " \"'The structure model was then further built through successive rounds with COOT and refinement with phenix.refine.'\",\n",
       " \"'\\\\n'\",\n",
       " \"'The peptide was then modified to the mutant form using COOT.'\",\n",
       " \"'A final model was obtained after iterative cycles of real space model building in COOT and TLS refinement in Phenix\",\n",
       " \"and was refined to Rwork/Rfree values of 0.21/0.23 at 2.3 Å resolution with excellent geometry (Ramachandran favoured/outliers: 97.8%/0.2%) (Extended Data Table 1a).'\",\n",
       " \"'Final models were obtained after iterative cycles of model building in COOT\",\n",
       " 'and refinement in BUSTER (P21) and Phenix (P21212)',\n",
       " \"yielding excellent geometry (Ramachandran favoured/outliers: P21\\\\u2009=\\\\u200998.2%/0.0%; P21212\\\\u2009=\\\\u200998.2%/0.2%) and Rwork/Rfree values of 0.18/0.20 (P21) and 0.15/0.18 (P21212) (Extended Data Table 1a).'\",\n",
       " \"'Best-defined regions of the electron density maps were used for rebuilding of every domain type using Coot and O\",\n",
       " 'respectively',\n",
       " 'symmetry expanded',\n",
       " \"and recombined into 18 chains.'\",\n",
       " \"'1b)\",\n",
       " \"and the structure was completed in alternating cycles of model building in COOT and restrained TLS refinement in phenix.refine (http://www.phenix-online.org).'\",\n",
       " \"'XDS\",\n",
       " \"Coot and REFMAC programs were used for crystallographic analysis.'\",\n",
       " \"'The model was completed by manual model building of CSD1 into the electron density and was refined by cyclic rounds of model building using COOT and maximum-likelihood refinement using REFMAC.'\",\n",
       " \"'The model was constructed using Coot and was refined in PHENIX to 3.1\\\\u2009Å with Rwork of 23.8% and Rfree of 28.9%\",\n",
       " 'and contained residues 330–616',\n",
       " \"684–833 and 872–1059.'\",\n",
       " \"'Final models were obtained after several cycles of refinement with REFMAC and PHENIX plus manual re-building using COOT.'\",\n",
       " \"'CTD1 was built manually using Coot.'\",\n",
       " \"'The structure was built in Coot\",\n",
       " \"and refined in Phenix to an R factor of 0.229 and Rfree of 0.305.'\",\n",
       " \"'There were no homologues with known structure for CTD1 of FimD and this domain was built manually using Coot when all other structural units were located.'\",\n",
       " \"'Model building of the FimD–FimC–FimH complex was carried out manually in Coot.'\",\n",
       " \"'The FimD usher translocation domain was manually rebuilt in Coot and refined (no &sgr; cutoff applied) in Phenix.'\",\n",
       " \"'The model was built in Coot and refined using Phenix and Refmac.'\",\n",
       " \"'\\\\n'\",\n",
       " \"'Data processing and scaling were done with XDS and Scala.'\",\n",
       " \"'The model from MR was built with the program COOT and subsequently subjected to refinement by the program Phenix.'\",\n",
       " \"'The structures were iteratively rebuilt and refined using the programs Coot\",\n",
       " 'CNS1.3',\n",
       " \"and phenix.refine (Extended Data Table 1).'\",\n",
       " \"'The structure was iteratively built and refined using the program Coot\",\n",
       " \"and phenix.refine (Extended Data Table 1).'\",\n",
       " \"'Structures were built with Coot and generally have good geometry.'\",\n",
       " \"'\\\\n'\",\n",
       " \"'The atomic model was built and refined with COOT and PHENIX.'\",\n",
       " \"'The cycles of adjustments to the electron density were done using graphical program COOT.'\",\n",
       " \"'Water molecules were modelled using COOT and included in the refinement cycles at the later stage.'\",\n",
       " \"'The model was manually completed and rebuilt with Coot and O\",\n",
       " \"using a three-fold symmetry-averaged map that was sharpened with a B factor of -100\\\\u2009Å2.'\",\n",
       " \"'Protein side chains were manually readjusted in COOT followed by simulated annealing and grouped B factor refinement.'\",\n",
       " \"'\\\\n'\",\n",
       " \"'All subsequent\",\n",
       " \"lower resolution models were generated by manually positioning the higher resolution structure onto the available density and rigid-body refining the individual domains to their final position using Coot.'\",\n",
       " \"'\\\\n'\",\n",
       " \"'Superpositions were performed with the SSM superposition routine of COOT.'\",\n",
       " \"'Adjustments to the model\",\n",
       " 'including addition of solvent molecules',\n",
       " \"using Coot were guided by manual inspection of 2Fo\\\\u2009− Fc and Fo\\\\u2009−\\\\u2009Fc electron density maps and were judged successful by a decrease in Rfree during refinement.'\",\n",
       " \"'After several rounds of refinement in PHENIX package with manual rebuilding in COOT\",\n",
       " \"the final model has good stereochemistry with an R value of 18.0% and an Rfree of 22.9%.'\",\n",
       " \"'Model building was performed with Coot and refinement with CNS.'\",\n",
       " \"'The structure was refined in PHENIX with manual corrections made in COOT.'\",\n",
       " \"'Further modelling of mORC1BAH domain was carried out using COOT\",\n",
       " \"and was then subject to refinement using PHENIX.'\",\n",
       " \"'Pol II\",\n",
       " 'DNA',\n",
       " \"and the Tfg1 charged helix were slightly adjusted manually using COOT to accurately reflect the density.'\",\n",
       " \"'The Tfg1 charged helix was placed manually using COOT according to the chemical environment.'\",\n",
       " \"'This density was modelled with canonical duplex B-DNA\",\n",
       " \"generated in COOT.'\",\n",
       " \"'The DNA sequence of all nucleic acid models was edited in COOT to match the sequence of the employed nucleic acid scaffold.'\",\n",
       " \"'cerevisiae head module (PDB code 4GWP)\",\n",
       " \"using secondary structure matching in COOT.'\",\n",
       " \"'The geometry of the obtained head module model was regularized in COOT and with PHENIX.'\",\n",
       " \"'The model was built with COOT.'\",\n",
       " \"'The initial model was built in COOT manually.'\",\n",
       " \"'Manual model building was done with COOT\",\n",
       " \"and structure refinement was done using PHENIX with four-fold NCS restraints.'\",\n",
       " \"'Several domains\",\n",
       " 'including the CSD2 and S1 domains of Rrp44',\n",
       " 'all domains of Csl4 and Rrp4 subunits',\n",
       " 'and the N-terminal domain of Rrp40',\n",
       " 'Rrp6 and the RNA were placed manually as phases were slowly improved by several cycles of model building (Coot)',\n",
       " 'rigid-body',\n",
       " 'group B-factor',\n",
       " \"and TLS refinements (PHENIX).'\",\n",
       " \"'Crystallographic refinement involved repeated cycles of conjugate-gradient energy minimization and temperature-factor refinement\",\n",
       " \"and was performed using PHENIX followed by iterative model building in Coot.'\",\n",
       " \"'The structure was refined using heavy non-crystallographic symmetry (NCS) restraints for a large subset of the residues with the program CNS (version 1.0) and rebuilt using XtalView and Coot.'\",\n",
       " \"'The structural model was iteratively rebuilt in Coot and refined using Phenix.'\",\n",
       " \"'Alternating cycles of model building with COOT and refinement with REFMAC led to a complete model of Argos217 and SpitzEGF with Rcryst and Rfree values of 0.20 and 0.24\",\n",
       " 'respectively',\n",
       " \"to 1.6\\\\u2009Å resolution (Supplementary Table 1).'\",\n",
       " \"'Alternating cycles of model building with COOT and refinement with REFMAC led to a complete model of Argos217 and SpitzEGF.'\",\n",
       " \"'The resulting phase-combined electron density map was of sufficient quality to manually build a model using COOT.'\",\n",
       " \"'The structure was built into the initial electron density map\",\n",
       " \"and was then subjected to alternating rounds of rebuilding in COOT and refinement with REFMAC5.'\",\n",
       " \"'Rebuilding of the model between refinement cycles and adding waters was performed in Coot.'\",\n",
       " \"'The quality of the models was assessed using Coot\",\n",
       " \"Procheck and the Molprobity server.'\",\n",
       " \"'The structural model was revised in real space with the program COOT using sigma-A weighted 2Fo\\\\u2009−\\\\u2009Fc and Fo\\\\u2009−\\\\u2009Fc electron density maps.'\",\n",
       " \"'\\\\n'\",\n",
       " \"'For the IR tails of CDH1 and APC10\",\n",
       " \"the models were refined with COOT.'\",\n",
       " \"'All structures were subjected to multiple rounds of manual correction and refinement using COOT and Phenix.'\",\n",
       " \"'Docked as a rigid body\",\n",
       " 'the derived TARP homology model was refined in real space against the density map using COOT guided by the resolved helical density of the TARP TMs',\n",
       " \"and the density consistent with the conserved β-sheet on the extracellular side of the micelle.'\",\n",
       " \"'\\\\n'\",\n",
       " \"'The structure was real-space refined in Coot constrained to ideal geometry and secondary structure where appropriate.'\",\n",
       " \"'The model was modified manually in Coot\",\n",
       " \"and refined by PHENIX against the native data set.'\",\n",
       " \"'The structure was refined with AutoBUSTER alternated with model rebuilding using COOT.'\",\n",
       " \"'An initial atomic model of &Dgr;MTBD was built on the density-modified electron-density map using COOT.'\",\n",
       " \"'Initial models for subunits e and f were generated manually in Coot.'\",\n",
       " \"'Non-crystallographic symmetry (NCS) restraints were used throughout the refinement performed by alternate cycles of manual building in COOT and automatic refinement using REFMAC and BUSTER.'\",\n",
       " \"'Iterative model building and refinement were performed using the COOT and PHENIX package.'\",\n",
       " \"'The model was optimized by alternating manual building with Coot and refinement using Phenix against a 2.0\\\\u2009Å native data set.'\",\n",
       " \"'The model was built manually with COOT and refined with REFMAC5 (ref.'\",\n",
       " \"'The structure was refined with PHENIX and COOT.'\",\n",
       " \"'Manual adjustment of the model was carried out in Coot.'\",\n",
       " \"'\\\\n'\",\n",
       " \"'The ArpWarp-model was completed manually using Coot and refined with Refmac.'\",\n",
       " \"'An initial model was manually built with COOT\",\n",
       " \"where the positions of the 16 selenium atoms were used to assign the sequence.'\",\n",
       " \"'Model building was done using COOT\",\n",
       " \"and refinement was done with CNS and PHENIX.'\",\n",
       " \"'The resulting map was readily interpretable and a structural model was manually built with Coot.'\",\n",
       " \"'Refinement was performed with PHENIX followed by manual examination and rebuilding of the refined coordinates using COOT.'\",\n",
       " \"'The structure was solved by molecular replacement using the N-terminal and C-terminal bundles of rat GLUT5 as independent search models in PHASER\",\n",
       " 'and refined using REFMAC5 and BUSTER against data up to 3.0 Å resolution after anisotropy correction',\n",
       " \"with rounds of re-building in COOT.'\",\n",
       " \"'Model building was carried out with Coot.'\",\n",
       " \"'The initial model was rebuilt with COOT\",\n",
       " \"cycled with structure refinement by CNS.'\",\n",
       " \"'An initial model was built using COOT.'\",\n",
       " \"'An initial model was built into the experimental electron density map using COOT.'\",\n",
       " \"'Structures for the rest of the data sets were solved by molecular replacement using this structure as a search model followed by numerous cycles of refinement in PHENIX and manual model building in Coot.'\",\n",
       " \"'A Fab homology model was made\",\n",
       " 'using PDB 1NGQ for the light chain and 1F3D for the heavy chain',\n",
       " \"and Coot to overlay the two modelled chains to make a single Fab molecule.'\",\n",
       " \"'Computer programs used were XDS\",\n",
       " 'CCP4-SCALA',\n",
       " 'HKL2MAP',\n",
       " 'SHARP',\n",
       " 'SOLOMON',\n",
       " \"COOT and PHENIX.'\",\n",
       " \"'The model was built with COOT and refined with PHENIX applying anisotropic scaling and bulk solvent correction during rigid body\",\n",
       " 'multiple translation-libration-screw rotation (TLS) refinement',\n",
       " 'coordinate and individual restrained B-factor refinement',\n",
       " \"and simulated annealing.'\",\n",
       " \"'Model building using COOT was initiated by manually fitting the ISW1a(&Dgr;ATPase) structure to the electron density and refinement was performed with PHENIX.'\",\n",
       " \"'Model building was performed using COOT and the final model was refined by the CCP4 program REFMAC5; graphics were produced using PyMOL.'\",\n",
       " \"'All model building was carried out in COOT\",\n",
       " \"followed by refinement using PHENIX.'\",\n",
       " \"'The structure was built with the use of COOT and refined to final Rwork\\\\u2009=\\\\u200919.9% and Rfree\\\\u2009=\\\\u200924.5% at 2.15\\\\u2009Å resolution with CNS.'\",\n",
       " \"'Model building with Coot was interspersed with either Refmac5 or BUSTER refinement.'\",\n",
       " \"'Model superpositions were done with programs from the CCP4 suite\",\n",
       " \"lsqman and Coot.'\",\n",
       " \"'Corresponding residues were mutated and fit manually into the electron density maps using Coot.'\",\n",
       " \"'All models were docked into the EM density map using Chimera\",\n",
       " \"followed by iterative manual adjustment and real-space refinement using COOT and fragment-based refinement with Rosetta.'\",\n",
       " \"'\\\\n'\",\n",
       " \"'2)\",\n",
       " \"so that the model could be manually updated and completed in O and COOT.'\",\n",
       " \"'The complete atomic model of the ArfA-RF2-SRC was manually adjusted using Coot and refined using phenix.real_space_refine\",\n",
       " \"with restraints obtained by phenix.secondary_structure_restraints.'\",\n",
       " \"'The obtained model was of sufficient quality for molecular replacement using Phaser with the native data set\",\n",
       " \"after which the model was manually completed in Coot and refined in Phenix.refine using TLS parameters.'\",\n",
       " \"'Manual model building was carried out using the program Coot.'\",\n",
       " \"'The model was optimized by rounds of manual rebuilding in Coot and refinement in Refmac5 with TLS.'\",\n",
       " \"'An initial model was built with ARP-wARP and refinement and rebuilding was carried out with BUSTER and Coot\",\n",
       " \"respectively.'\",\n",
       " \"'Model building was performed with COOT and illustrated with PYMOL.'\",\n",
       " \"'Substantial model rebuilding was carried out manually with TURBO-FRODO and COOT.'\",\n",
       " \"'The initial model was refined with deformable elastic network (DEN) constraints in CNS\",\n",
       " \"and then improved by cycles of manual building in Coot and refinement in PHENIX.'\",\n",
       " \"'Refinement of the structure was performed by PHENIX Refine and the model was built using COOT.'\",\n",
       " \"'All diffraction data were processed with the HKL2000 suite and model building and refinement were performed in COOT and PHENIX\",\n",
       " \"respectively.'\",\n",
       " \"'The two loop regions\",\n",
       " 'Thr\\\\u200948 to Val\\\\u200953 and Gly\\\\u200989 to Leu\\\\u2009101',\n",
       " \"were rebuilt according to the electron density in program Coot.'\",\n",
       " \"'The initial model was built in COOT manually.'\",\n",
       " \"'Crystal structures were solved by molecular replacement with PHASER and refined using PHENIX and COOT.'\",\n",
       " \"'\\\\n'\",\n",
       " \"'Model building was carried out using COOT\",\n",
       " \"and refinement was performed with PHENIX.'\",\n",
       " \"'A molecular replacement solution consisting of two complexes per asymmetric unit was obtained using PHASER\",\n",
       " 'and COOT and PHENIX were used for model building and refinement',\n",
       " \"respectively.'\",\n",
       " \"'Model building and refinement were performed using COOT and PHENIX\",\n",
       " \"respectively.'\",\n",
       " \"'Model building was carried out using COOT\",\n",
       " \"and refinement was performed with PHENIX.'\",\n",
       " \"'Refinement of the structure was performed using a combination of CNS\",\n",
       " \"CCP4 and COOT.'\",\n",
       " \"'Model building and refinement were performed with COOT and REFMAC5 in which 5% of the reflections were excluded from the refinement.'\",\n",
       " \"'Then a crude helical model was manually built in the program Coot.'\",\n",
       " \"'Iterative rounds of model building and refinement (simulated annealing\",\n",
       " 'rigid-body',\n",
       " 'B-factor refinement',\n",
       " \"phase combination using COOT and PHENIX REFINE) led to the final model.'\",\n",
       " \"'We improved the initial model by iteratively building regions of the receptor in Coot and refining in Phenix.'\",\n",
       " \"'The initial model was built using RESOLVE and further manual model building was carried out using COOT.'\",\n",
       " \"'Subsequent manual model building was carried out using COOT.'\",\n",
       " \"'With over 7,000 AdoMet radical enzymes now annotated as Cbl-dependent\",\n",
       " \"these dual-cofactor enzymes are emerging as a new superfamily.'\",\n",
       " \"'Iterative cycles of refinement and manual model building were carried out using PHENIX and COOT\",\n",
       " 'respectively',\n",
       " \"until the models converged to acceptable levels of R-factors and stereochemistry.'\",\n",
       " \"'The remaining structural elements were built manually using the Coot program.'\",\n",
       " \"'After tracing the initial model manually using the program Coot\",\n",
       " 'the model was refined against the native data at 2.90-Å resolution using translation',\n",
       " 'libration and screw-rotation (TLS) refinement adopting a single TLS body as implemented in the program PHENIX',\n",
       " \"leaving 5% of the reflections in the free R set.'\",\n",
       " \"'Iterations of refinement using PHENIX and the software suite CNS and model building in Coot led to the current model\",\n",
       " 'which consists of 1,686 amino acid residues (residues 4–1,043 of CusA',\n",
       " \"residues 79–400 of molecule 1 of CusB and residues 79–402 of molecule 2 of CusB).'\",\n",
       " \"'After tracing the initial model manually using the program Coot\",\n",
       " 'the model was refined against the native data at 2.90-Å resolution using translation',\n",
       " 'libration and screw-rotation (TLS) refinement adopting a single TLS body as implemented in the program PHENIX',\n",
       " \"leaving 5% of the reflections in the free R set.'\",\n",
       " \"'Iterations of refinement using PHENIX and the software suite CNS and model building in Coot led to the current model\",\n",
       " 'which consists of 1,686 amino acid residues (residues 4–1,043 of CusA',\n",
       " \"residues 79–400 of molecule 1 of CusB and residues 79–402 of molecule 2 of CusB).'\",\n",
       " \"'Model building and structural refinement were performed with the PHENIX software and COOT\",\n",
       " \"respectively.'\",\n",
       " \"'Molecular replacement was performed with Molrep\",\n",
       " 'refinements were performed with PHENIX and intermittent model building was performed with COOT',\n",
       " \"as described above.'\",\n",
       " \"'Iterative rounds of manual and automatic structure refinement were carried out with Coot and phenix.refine\",\n",
       " 'respectively',\n",
       " \"and protein geometry was monitored using Molprobity.'\",\n",
       " \"'All models were iteratively improved by manual building in Coot and refined using REFMAC5 (ref.'\",\n",
       " \"'The initial model was further built using Coot and used for molecular replacement with Phaser of a higher resolution apo structure and the GSH bound structure.'\",\n",
       " \"'The model was further built using Coot and used for molecular replacement with Phaser of a higher resolution apo structure and the GSH bound structure.'\",\n",
       " \"'The resulting model was further optimized manually using Coot (Fig.'\",\n",
       " \"'The models were further optimized by Coot (Extended Data Figs 6b\",\n",
       " \"c and 8).'\",\n",
       " \"'The model was completed by rounds of manual building in COOT followed by refinement in REFMAC5 (ref.'\",\n",
       " \"'\\\\n'\",\n",
       " \"'Model building and refinement were accomplished iteratively using COOT and Refmac5 in CCP4.'\",\n",
       " \"'This solution was subjected to one round of simulated annealing refinement in PHENIX and subsequently re-built automatically by ARP/wARP and completed by manual rebuilding in COOT and refinement in PHENIX.'\",\n",
       " \"'PHENIX and Coot were used to refine the structure.'\",\n",
       " \"'Between rounds of refinement\",\n",
       " \"the model was manually checked and corrected against the corresponding electron-density maps in COOT.'\",\n",
       " \"'Subsequent model building and refinement were performed using REFMAC5 and Coot\",\n",
       " 'respectively',\n",
       " \"with the validity of the model monitored using MolProbity before deposition (PDB accession 3ZMR).'\",\n",
       " \"'The resulting map was of sufficient quality for building the model of the CED in Coot.'\",\n",
       " \"'The model was built in COOT and refined using PHENIX and BUSTER using non-crystallographic symmetry (NCS) and TLS refinement restraints.'\",\n",
       " \"'\\\\n'\",\n",
       " \"'The structure was further refined by iterative cycles of positional refinement and translation-libration-screw refinement with Refmac5 and Phenix and model building with COOT\",\n",
       " \"and the final model was refined to 2.0\\\\u2009Å resolution with Rwork/Rfree of 20.4%/22.7%.'\",\n",
       " \"'\\\\n'\",\n",
       " \"'Superposition of maps was performed in COOT using “transformation by LSQ model fit”.'\",\n",
       " \"'The apo-SMOΔC polypeptide chain was traced using iterative rounds of BUCCANEER\",\n",
       " \"manual building in COOT and refinement in autoBUSTER and PHENIX.'\",\n",
       " \"'Superpositions were calculated using the program COOT\",\n",
       " 'electrostatic potentials were generated using APBS and hydrophobicity was calculated according to the Eisenberg hydrophobicity scale',\n",
       " \"as implemented in PyMOL.'\",\n",
       " \"'Subsequent model building and inspection were carried out using the program Coot.'\",\n",
       " \"'\\\\n'\",\n",
       " \"'Then a crude helical model was built manually using the program COOT.'\",\n",
       " \"'A crude model was built automatically using the CCP4 program buccaneer\",\n",
       " \"improved by manual building using Coot and using the riboflavin-binding protein structure as a reference.'\",\n",
       " \"'A crude model was built automatically using the CCP4 program buccaneer and improved by manual building using Coot.'\",\n",
       " \"'Taking advantage of the defined geometry of &agr;-helices and clear bumps for C&agr; atoms\",\n",
       " \"a polyalanine model was first built in Coot.'\",\n",
       " \"'Model building was completed in COOT.'\",\n",
       " \"'The target DNA strands of the complexes were superimposed using a least-squares algorithm in Coot and the complexes are shown in identical orientations.'\",\n",
       " \"'A combination of simulated-annealing composite omit and sigmaA-weighted 3Fobs\\\\u2009-\\\\u20092Fcalc\",\n",
       " \"2Fobs\\\\u2009-\\\\u2009Fcalc and Fobs\\\\u2009-\\\\u2009Fcalc maps were used for model building in O and COOT.'\",\n",
       " \"'Models were built and refined using tools in the CCP4\",\n",
       " 'COOT',\n",
       " \"and PHENIX packages.'\",\n",
       " \"'Further model building was performed using COOT and subsequent refinement done in PHENIX and BUSTER-TNT.'\",\n",
       " \"'An initial model was generated by placing a high-resolution structure of the neuronal core complex into the experimental electron density using MOLREP and manually extending the structure into the linkers and TMRs with COOT.'\",\n",
       " \"'Initial phases were improved by iterative steps of manual model building\",\n",
       " 'refinement and maximum-likelihood density modification using Coot',\n",
       " 'Phenix Refine and Phenix Phase and Build',\n",
       " \"respectively.'\",\n",
       " \"'The crystal structure of 6HDNO (PDB 2BVG) was used as a search model\",\n",
       " 'and ARP/wARP',\n",
       " 'Coot and Refmac were used for automatic model building',\n",
       " 'for visual inspection and manual rebuilding of the model',\n",
       " 'and for several rounds of energy minimization and individual B-factor refinement',\n",
       " \"respectively.'\",\n",
       " \"'Iterative rounds of model building and refinement were carried out using Coot and the Phenix.Refine module of PHENIX.'\",\n",
       " \"'Manual rebuilding of the opsin model and electron density interpretation were performed using the program COOT.'\",\n",
       " \"'33) and CNS together with manual model building with O and Coot.'\",\n",
       " \"'Further modelling of the covalent DNMT3A–DNMT3L–DNA complexes was performed using COOT and then subject to refinement using the PHENIX software package.'\",\n",
       " \"'The three component proteins TraFCT\",\n",
       " \"TraOCT and TraN were traced manually with the program Coot.'\",\n",
       " \"'A partial model was built by manually placing idealized poly-Ala helices into the solvent-flattened electron density map by using COOT.'\",\n",
       " \"'Despite the presence of twinning\",\n",
       " 'initial phases were found by molecular replacement using PHASER with PDB 1U7G placing six molecules in the asymmetric unit cell',\n",
       " \"and followed automated model building using PHENIX and manual model building using COOT.'\",\n",
       " \"'The published fragments of FluPolA (PDB accessions 4IUJ\",\n",
       " '4AWH',\n",
       " '4CB4',\n",
       " '3A1G and 2VY7) were fitted by eye using Coot',\n",
       " \"which was used for all model building.'\",\n",
       " \"'Model building and refinement was conducted in Coot and PHENIX\",\n",
       " \"respectively.'\",\n",
       " \"'Phasing and initial model building of the IZUMO1 crystal structure were performed using autoSHARP and ARP/wARP\",\n",
       " 'respectively',\n",
       " \"followed by iterative cycles of manual model building using COOT program and restrained refinement using REFMAC until the R factor was converged.'\",\n",
       " \"'Initial phases were obtained using Phaser\",\n",
       " \"and the protein model built manually in Coot and refined using REFMAC with TLS and three-fold NCS restraints.'\",\n",
       " \"'Model building was performed using the programs O and coot\",\n",
       " \"and structural refinement was carried out under a tight restraint of non-crystallographic six-fold symmetry with the programs Crystallographic and NMR System (CNS) and REFMAC.'\",\n",
       " \"'Models were manually fitted using O and Coot and further refined using CNS and Refmac5.'\",\n",
       " \"'Programs O and Coot were used to manually fit the protein model.'\",\n",
       " \"'The resulting solution was refined using iterative cycles of model building in COOT and refinement (rigid body\",\n",
       " 'energy minimization',\n",
       " 'simulated annealing',\n",
       " \"individual B-factor refinement) with the PHENIX program suite.'\",\n",
       " \"'PHENIX was used for refinement\",\n",
       " \"Coot for model building and MolProbity for model validation.'\",\n",
       " \"'Structure interpretation was performed using COOT\",\n",
       " \"with refinement in PHENIX.'\",\n",
       " \"'Improved electron density (both calculated and density-modified) allowed for cycles of manual building of subunit Nqo8 and re-building of some regions in other subunits in Coot\",\n",
       " 'followed by refinement in PHENIX with secondary structure/NCS (non-crystallographic symmetry) restraints and using TLS (translation',\n",
       " \"libration and screw-motion parameterization).'\",\n",
       " \"'The initial model was built using RESOLVE\",\n",
       " \"Buccaneer and Coot.'\",\n",
       " \"'The model building was carried out using COOT.'\",\n",
       " \"'35) program suite REFMAC and all the model building was done using COOT.'\",\n",
       " \"'\\\\n'\",\n",
       " \"'The model was built in COOT and refined with PHENIX.refine with six translation/libration/screw (TLS) and non-crystallographic symmetry (NCS) groups per protomer in the final cycles of refinement.'\",\n",
       " \"'The preliminary model was rebuilt by the software BUCCANEER and manually completed and corrected in COOT.'\",\n",
       " \"'\\\\n'\",\n",
       " \"'The atomic models of the transcribing state were built by fitting the atomic structures of RdRP and VP4 at quiescent state into the density\",\n",
       " 'manually adjusting the changed residues with Coot',\n",
       " \"and refining the models with Phenix.'\",\n",
       " \"'Models were built with O and COOT.'\",\n",
       " \"'The structures were iteratively rebuilt and refined using the programs Coot and phenix.refine (Extended Data Table 1).'\",\n",
       " \"'COOT and REFMAC5 were used for manual building and refinement respectively to produce a complete model.'\",\n",
       " \"'40) and PHENIX together with manual model building in COOT.'\",\n",
       " \"'40) and PHENIX together with manual model building in COOT.'\",\n",
       " \"'Refinement used PHENIX and model building used Coot.'\",\n",
       " \"'The models were subsequently constructed and refined with the Coot toolkit (Supplementary Information).'\",\n",
       " \"'The model was built by iterative modelling in COOT and refined using CNS and REFMAC5.'\",\n",
       " \"'Iterative model building and refinement were carried out in Phenix with manual adjustments using COOT.'\",\n",
       " \"'The resulting structure was iteratively refined in Phenix and manually rebuilt in Coot.'\",\n",
       " \"'Loops were then regularized and additional residues added as needed in Coot.'\",\n",
       " \"'Model building was performed in COOT with the aid of the map sharpening utilities.'\",\n",
       " \"'The manual model building was performed in COOT with the aid of the map sharpening utilities.'\",\n",
       " \"'The final SbsB structure was  refined against data from a native  SbsB32–920:NbKB6 crystal collected to  2.4\\\\u2009Å resolution (phased by molecular substitution with  the SbsB32–920,T744HA:NbKB6 model)\",\n",
       " 'providing a model  of 18.6% and 23.6% R-factor and free R-factor',\n",
       " 'respectively',\n",
       " \" containing two copies of NbKB6 (chains B and D) and two SbsB molecules  encompassing residues 202–920 (chain A) or residues  295–920 (chain C) and having 98% of residues with  &phgr; and &psgr; angles in favoured regions of  the Ramachandran diagram (Supplementary Table 1).'\",\n",
       " \"'The model from MR was built with the program COOT and subsequently subjected to refinement by the program PHENIX.'\",\n",
       " \"'Model rebuilding was performed using COOT\",\n",
       " \"and model refinement was performed using CNS and REFMAC.'\",\n",
       " \"'Structural models were built in COOT unless indicated otherwise.'\",\n",
       " \"'To model closed promoter DNA\",\n",
       " \"upstream DNA was extended with canonical duplex B-form DNA in COOT and rigid-body fitted in UCSF Chimera to reflect the density.'\",\n",
       " \"'Model building of the TFIIF N-terminal residues in COOT was guided by a selenium anomalous difference Fourier peak (M27).'\",\n",
       " \"'The model was built iteratively in O and Coot and refined with CNS.'\",\n",
       " \"'The resulting maps showed clear unbiased density for protospacer DNA\",\n",
       " \"and subsequent model building was performed through iterative rounds of building in Coot and refinement in PHENIX with NCS restraints on the protein subunits.'\",\n",
       " \"'Models were built using PHENIX Autobuild and manually adjusted using COOT.'\",\n",
       " \"'Manual model building and local real-space refinement was performed in COOT.'\",\n",
       " \"'Automated real-space refinement cycles were alternated with manual model building in COOT.'\",\n",
       " \"'The model was manually rebuilt in Coot and iteratively refined in PHENIX.'\",\n",
       " \"'COOT was used for model building and adjusting.'\",\n",
       " \"'In all cases\",\n",
       " \"model rebuilding was done using Coot.'\",\n",
       " \"'The resulting map was readily interpretable and an initial model was built using the interactive graphics program COOT.'\",\n",
       " \"'Model correction and refinement were undertaken using the programs COOT and REFMAC5 (ref.'\",\n",
       " \"'Model correction and refinement were undertaken using the programs COOT and PHENIX.'\",\n",
       " \"'Model correction and refinement were undertaken using the programs COOT and PHENIX.'\",\n",
       " \"'The NMR model for the C-terminal barrel of fHbp could not be placed in this density (the strands were seen to be distorted) so a model was built by hand using program Coot.'\",\n",
       " \"'Rebuilding and refinement using data in the range 40–2.35\\\\u2009Å proceeded using Coot and Buster-TNT with non crystallographic symmetry used to restrain the core of each domain while more variation was allowed between the loop region until the R/Rfree were 25.5/27.1 with no residues in the disallowed regions of the Ramachandran plot (94.6% in the most favoured\",\n",
       " \"as determined by MolProbity).'\",\n",
       " \"'Minor rebuilding of the loop regions and refinement (90–2.35\\\\u2009Å) using Coot/Buster-TNT led to R/Rfree of 23.2/26.3 once again with no residues in the disallowed regions of the Ramachandran plot (95.3% in the most favoured).'\",\n",
       " \"'33) and Coot\",\n",
       " \"respectively (Supplementary Table 3).'\",\n",
       " \"'All water molecules were modelled in Coot.'\",\n",
       " \"'Finally\",\n",
       " \"25 protein sequences (Supplementary Table 2) were identified and confirmed by high agreement of secondary structural features and side chain information between the predicted models and the density maps in COOT.'\",\n",
       " \"'The structure was refined with Phenix.refine\",\n",
       " 'with default tight non-crystallographic symmetry restraints between the two molecules',\n",
       " \"and rebuilt with Coot.'\",\n",
       " \"'Missing residues were manually build in COOT after initial rounds of refinement.'\",\n",
       " \"'Because in all electron densities the secondary structure of proteins was visible\",\n",
       " \"a highly reliable initial rigid body fit for aPelota/Dom34 aABCE1/Rli1 could be performed using Coot and UCSF Chimera.'\",\n",
       " \"'This allowed extension of the initial model with Coot and subsequent refinement with Refmac.'\",\n",
       " \"'The E2–E1 model was built by using the program COOT with the guidance from the (2Fo\\\\u2009−\\\\u2009Fc)\",\n",
       " '(Fo\\\\u2009−\\\\u2009Fc)',\n",
       " 'omit',\n",
       " \"and B-factor sharpened (2Fo\\\\u2009−\\\\u2009Fc) maps.'\",\n",
       " \"'The model was built manually using COOT in the resulting electron density map (Fo\\\\xa0×\\\\xa0exp (i&agr;DM)).'\",\n",
       " \"'The model was built by using COOT and refined with the program REFMAC5 from the CCP4 suite.'\",\n",
       " \"'The structures were determined by molecular replacement using the program Phaser\",\n",
       " 'model building was performed using the program Coot',\n",
       " \"and structure refinement was accomplished using the programs CNS and PHENIX.'\",\n",
       " \"'The model was corrected and further refined by the programs Coot and CNS.'\",\n",
       " \"'The structure missing in the search models was manually constructed\",\n",
       " \"and the entire model was further corrected and refined by the programs Coot and CNS.'\",\n",
       " \"'The models were rebuilt using ARP/wARP and/or manually in Coot and refined in Phenix and/or Refmac.'\",\n",
       " \"'The complete intasome model was iteratively relaxed with Rosetta and then adjusted manually using Coot.'\",\n",
       " \"'Density modification with SOLOMON resulted in a readily interpretable map\",\n",
       " \"into which the structures of all three subunits could be built using Coot.'\",\n",
       " \"'The model of the hydroxylamine complex was prepared by manual docking in COOT\",\n",
       " \"using an iron–nitrogen bond length between haem &agr;I and hydroxylamine as observed in crystal structures of catalase–hydroxylamine complexes.'\",\n",
       " \"'This model was subsequently extended by iterative model building and refinement cycles using COOT and REFMAC5 (ref.'\",\n",
       " \"'Models were built with COOT and refined with Refmac and Phenix.'\",\n",
       " \"'Refinement and model building was done using PHENIX and COOT.'\",\n",
       " \"'Model building and refinement were carried out in PHENIX and COOT.'\",\n",
       " \"'The model was further completed and refined using alternating cycles of manual building in the programs Coot and REFMAC version 5 (ref.'\",\n",
       " \"'Rotamers for each residue within each of these ten models were then assigned manually using COOT based on visual inspection of the fit of the trial rotamer to the density at the corresponding site in Mobs.'\",\n",
       " \"'Refinement and model building were carried out with PHENIX and COOT.'\",\n",
       " \"'A native data set to 2.0\\\\u2009Å was used for model rebuilding and refinement with COOT and PHENIX.'\",\n",
       " \"'The atomic model was built into the electron density map manually with the program Coot.'\",\n",
       " \"'The model was completed by iterative manual building of loops\",\n",
       " \"regions of helices and replacement of side chains in the computer program Coot and refinement in Refmac.'\",\n",
       " \"'Coot and PHENIX were used to refine the structures.'\",\n",
       " \"'Additional missing residues were added manually in Coot.'\",\n",
       " \"'The atomic model was completed with Coot and refined with phenix.refine.'\",\n",
       " \"'All models were visualized and docked into the density in Chimera\",\n",
       " \"followed by manual adjustment and real-space refinement using COOT.'\",\n",
       " \"'Manual tracing and model building were carried out using COOT-0.5.1 and XFIT in XTALVIEW 4.0.'\",\n",
       " \"'Iterative rounds of refinement were carried out using the programs COOT\",\n",
       " 'PHENIX',\n",
       " \"and REFMAC.'\",\n",
       " \"'Model building and refinement were performed using Coot and Refmac\",\n",
       " \"respectively.'\",\n",
       " \"'The initial models were manually built with COOT and refined using PHENIX package.'\",\n",
       " \"'Iterative model building and refinements were carried out with COOT and PHENIX\",\n",
       " \"respectively.'\",\n",
       " \"'Visual inspection and manual correction of models were performed with the program COOT.'\",\n",
       " \"'Following iterative cycles of model building using COOT\",\n",
       " \"the model was refined using Refmac5 (Supplementary Table 1).'\",\n",
       " \"'Cycles of manual model building and crystallographic refinement were carried out using the computer graphic program Coot and the crystallographic refinement software package Phenix.'\",\n",
       " \"'Starting with torsion-angle simulated annealing with slow cooling\",\n",
       " 'iterative manual model building was carried out on COOT with maps generated from combinations of standard positional',\n",
       " 'individual B-factor',\n",
       " \"TLS (translation/libration/screw) refinement algorithms and non-crystallographic symmetry (NCS) restraints.'\",\n",
       " \"'We further remodelled the coordinates of both the proteins and the nucleic acids with the program Coot.'\",\n",
       " \"'Several of the rigid bodies deviated substantially from the electron density\",\n",
       " \"and they were manually adjusted to the density with the program Coot.'\",\n",
       " \"'Iterative cycles of model building and structure refinement were performed using COOT\",\n",
       " \"REFMAC5 and phenix.refine in the PHENIX program package.'\",\n",
       " \"'Model building was carried out in COOT.'\",\n",
       " \"'The cryo-EM structure of the U2 3′ domain from C complex\",\n",
       " 'where the Sm ring was replaced with the adjusted MODELLER model',\n",
       " 'and the SF3a crystal structure',\n",
       " 'where selenomethionine residues were changed to the original protein sequence in COOT',\n",
       " \"were fitted automatically into the B3 density using a global 6D search in Situs.'\",\n",
       " \"'The homology model was fitted into the B7 density and adjusted and extended in COOT.'\",\n",
       " \"'The resulting model was subject to iterative refinement with NCS restrains in Phenix and manual rebuilding in Coot.'\",\n",
       " \"'Model building was performed with the program COOT.'\",\n",
       " \"'Model building was carried out in O and COOT and was facilitated by the positions of the mercury and selenomethionine peaks.'\",\n",
       " \"'The initial structure for the TONSL-ARD–MCM2-HBD–H3–H4 tetramer complex was solved by molecular replacement in PHASER with our previous structure of the MCM2-HBD–H3–H4 tetramer complex as a search model and manually refined and built using Coot.'\",\n",
       " \"'Manual rebuilding of the opsin model and electron density interpretation were performed after each refinement cycle using the program COOT.'\",\n",
       " \"'These maps were of sufficient quality for model building which was carried out in COOT.'\",\n",
       " \"'Model building was performed using COOT\",\n",
       " \"and refinement was carried out by CNS and REFMAC.'\",\n",
       " \"'The resultant density-modified experimental map was used to build manually a partial C&agr; trace with COOT.'\",\n",
       " \"'The C&agr; backbone of this model was then adjusted to fit the cryo-EM map density using Coot and Gorgon (http://www.cs.wustl.edu/~ssa1/gorgon).'\",\n",
       " \"'The models were improved by iterative manual building in Coot\",\n",
       " \"simulated annealing in Phenix and maximum likelihood restrained positional refinement in Refmac.'\",\n",
       " \"'Atomic models were corrected by using COOT with careful inspection.'\",\n",
       " \"'All structures were refined using COOT and CNS.'\",\n",
       " \"'Then the entire atomic model was manually improved in COOT.'\",\n",
       " \"'An independent effort of atomic model building used stereochemical refinement tools in Coot and REFMAC that were originally designed for X-ray crystallography\",\n",
       " \"but later adapted for cryo-EM.'\",\n",
       " \"'The structures of DpNCT (Protein Data Bank accession number 4R12) and PSH (Protein Data Bank accession number 4HYC) were used as a guide\",\n",
       " 'while APH-1 and PEN-2 were built de novo in Coot',\n",
       " \"starting from idealized &agr;-helices.'\",\n",
       " \"'Model building in COOT was based on multi-crystal\",\n",
       " \"multi-domain averaged electron density maps and 2Fo\\\\u2009−\\\\u2009Fc maps.'\",\n",
       " \"'The structure was refined by iterative cycles of manual model fitting with the Coot program\",\n",
       " \"followed by positional and group B-factor refinements with the CNS and Phenix programs.'\",\n",
       " \"'The structure was refined by iterative cycles of manual model fitting with the Coot program\",\n",
       " \"followed by positional and group B-factor refinements with the CNS and Phenix programs.'\",\n",
       " \"'The model was further refined with stepwise cycles of manual model building using the COOT program and restrained refinement using REFMAC or phenix.refine until the R factors converged.'\",\n",
       " \"'Iterative model building was carried out in COOT.'\",\n",
       " \"'Recursive cycles of model building in coot and refinement in refmac5 (ref.'\",\n",
       " \"'For all models solvent molecules were added using coot and checked manually; 5% of the observations were randomly selected for the Rfree set.'\",\n",
       " \"'All models were validated using coot and molprobity.'\",\n",
       " \"'Model building was performed with the program Coot and structure refinement with Refmac 5 (ref.'\",\n",
       " \"'Model building was performed with the program Coot and structure refinement with Refmac 5 (ref.'\",\n",
       " \"'The homology model\",\n",
       " 'generated from NorM-VC (Protein Data Bank accession number 3MKU) with the program MODELLER',\n",
       " \"was fitted to the initial experimental electron density map and manually modified using the program Coot.'\",\n",
       " \"'Completion of the polyalanine trace of DCAF1-CtD and placement of the protein backbone of SAMHD1-CtD was then undertaken manually in Coot.'\",\n",
       " \"'Iterative rounds of manual model building in COOT\",\n",
       " 'coupled with refinement using PHENIX and stereochemical validation using MOLPROBITY',\n",
       " 'resulted in final models displaying excellent geometry and Rwork/Rfree values for their respective resolutions',\n",
       " \"as well as no Ramachandran outliers.'\",\n",
       " \"'Further refinement was done with PHENIX and Coot.'\",\n",
       " \"'Structural refinement was performed with Phenix and the model was manually modified with COOT.'\",\n",
       " \"'coli BamA and Saccharomyces cerevisiae Sam50 was performed manually using CHAINSAW/CCP4 (refs 42\",\n",
       " '43)',\n",
       " \"COOT and Chiron.'\",\n",
       " \"'All model building was performed using COOT and subsequent refinement done in PHENIX.'\",\n",
       " \"'Insertions were then added manually within COOT and a disulphide bond restraint added within loop 6 between residues Cys\\\\u2009690 and Cys\\\\u2009700 (not present in either HdBamA or NgBamA).'\",\n",
       " \"'Insertions were then added manually within COOT.'\",\n",
       " \"'Here\",\n",
       " \"the model for the ScSam50 POTRA domain was built manually within COOT and secondary structure elements aligned to POTRA 1 of EcBamA to retain the conserved overall fold.'\",\n",
       " \"'Several rounds of manual rebuilding and ligand and water fitting within the molecular graphics program COOT followed by refinement in PHENIX were performed to finalize the structure.']\"]"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matched_dfs.loc[220033]['clean_text']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "citation_graph = nx.from_pandas_edgelist(nature2references, source = 'CitingArticleID', target = 'CitedArticleID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "citation_node_degree = pd.DataFrame(citation_graph.degree)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "citation_node_degree = citation_node_degree.rename(columns = {0:'ArticleID', 1: 'Degree'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "matched_dfs = pd.merge(citation_node_degree, matched_dfs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Creating the JSON in the format the Onur requires"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def onur_json(tex_ref_mappings, references):\n",
    "    \n",
    "    references_w_text = references.copy()\n",
    "    \n",
    "    for i in tqdm(range(len(references_w_text))):\n",
    "        \n",
    "        paper = tex_ref_mappings[i]\n",
    "\n",
    "        #print (paper)\n",
    "        for tex_ref_mapping in paper:\n",
    "\n",
    "            #print(tex_ref_mapping.values())\n",
    "\n",
    "            for ref_id in list(tex_ref_mapping.values())[0]:\n",
    "\n",
    "                if 'references' in references_w_text[i]:\n",
    "                    \n",
    "                    if ref_id in references_w_text[i]['references']:\n",
    "\n",
    "                        references_w_text[i]['references'][ref_id].update({'text' :list(tex_ref_mapping.keys())[0]})\n",
    "                        \n",
    "                    else:\n",
    "                        references_w_text[i]['references'][ref_id] = {'text' :list(tex_ref_mapping.keys())[0]}\n",
    "\n",
    "    return references"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_zips_parallel_reducer(arg):\n",
    "    \n",
    "    try:\n",
    "        finfo, content = arg[0], zlib.decompress(arg[1])\n",
    "        del arg\n",
    "        \n",
    "        get_reference_text(finfo.filename, content)\n",
    "        get_reference_mapping(finfo.filename, content)\n",
    "        \n",
    "        return True\n",
    "    \n",
    "    except Exception as e:\n",
    "        \n",
    "        #print('Could not extract references from %s' % finfo)\n",
    "        return str(Exception)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def flatten(l):\n",
    "    flat_list = []\n",
    "    for sublist in l:\n",
    "        if type(sublist) == list:\n",
    "            for item in sublist:\n",
    "                flat_list.append(item)\n",
    "    return flat_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ArticleID</th>\n",
       "      <th>PubYear</th>\n",
       "      <th>Title</th>\n",
       "      <th>Volume</th>\n",
       "      <th>Issue</th>\n",
       "      <th>Cites</th>\n",
       "      <th>Journal</th>\n",
       "      <th>NatureID</th>\n",
       "      <th>DOI</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DOI</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10.1038/166444b0</th>\n",
       "      <td>pub.1007899685</td>\n",
       "      <td>1950</td>\n",
       "      <td>Detection of Sugars on Paper Chromatograms</td>\n",
       "      <td>166</td>\n",
       "      <td>4219</td>\n",
       "      <td>3143</td>\n",
       "      <td>Nature</td>\n",
       "      <td>166444b0</td>\n",
       "      <td>10.1038/166444b0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10.1038/166117b0</th>\n",
       "      <td>pub.1026639700</td>\n",
       "      <td>1950</td>\n",
       "      <td>The Soil-Plate Method for Isolation of Fungi f...</td>\n",
       "      <td>166</td>\n",
       "      <td>4211</td>\n",
       "      <td>394</td>\n",
       "      <td>Nature</td>\n",
       "      <td>166117b0</td>\n",
       "      <td>10.1038/166117b0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10.1038/165514a0</th>\n",
       "      <td>pub.1039717355</td>\n",
       "      <td>1950</td>\n",
       "      <td>Reversible Dilation and Contraction by Changin...</td>\n",
       "      <td>165</td>\n",
       "      <td>4196</td>\n",
       "      <td>359</td>\n",
       "      <td>Nature</td>\n",
       "      <td>165514a0</td>\n",
       "      <td>10.1038/165514a0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10.1038/166780a0</th>\n",
       "      <td>pub.1011096023</td>\n",
       "      <td>1950</td>\n",
       "      <td>Cell Specificity of Histones</td>\n",
       "      <td>166</td>\n",
       "      <td>4227</td>\n",
       "      <td>280</td>\n",
       "      <td>Nature</td>\n",
       "      <td>166780a0</td>\n",
       "      <td>10.1038/166780a0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10.1038/166288a0</th>\n",
       "      <td>pub.1000041177</td>\n",
       "      <td>1950</td>\n",
       "      <td>Synthesis of Peptides in Enzymic Reactions inv...</td>\n",
       "      <td>166</td>\n",
       "      <td>4216</td>\n",
       "      <td>259</td>\n",
       "      <td>Nature</td>\n",
       "      <td>166288a0</td>\n",
       "      <td>10.1038/166288a0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       ArticleID PubYear  \\\n",
       "DOI                                        \n",
       "10.1038/166444b0  pub.1007899685    1950   \n",
       "10.1038/166117b0  pub.1026639700    1950   \n",
       "10.1038/165514a0  pub.1039717355    1950   \n",
       "10.1038/166780a0  pub.1011096023    1950   \n",
       "10.1038/166288a0  pub.1000041177    1950   \n",
       "\n",
       "                                                              Title Volume  \\\n",
       "DOI                                                                          \n",
       "10.1038/166444b0         Detection of Sugars on Paper Chromatograms    166   \n",
       "10.1038/166117b0  The Soil-Plate Method for Isolation of Fungi f...    166   \n",
       "10.1038/165514a0  Reversible Dilation and Contraction by Changin...    165   \n",
       "10.1038/166780a0                       Cell Specificity of Histones    166   \n",
       "10.1038/166288a0  Synthesis of Peptides in Enzymic Reactions inv...    166   \n",
       "\n",
       "                 Issue Cites Journal  NatureID               DOI  \n",
       "DOI                                                               \n",
       "10.1038/166444b0  4219  3143  Nature  166444b0  10.1038/166444b0  \n",
       "10.1038/166117b0  4211   394  Nature  166117b0  10.1038/166117b0  \n",
       "10.1038/165514a0  4196   359  Nature  165514a0  10.1038/165514a0  \n",
       "10.1038/166780a0  4227   280  Nature  166780a0  10.1038/166780a0  \n",
       "10.1038/166288a0  4216   259  Nature  166288a0  10.1038/166288a0  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nature_references_info['DOI'] = nature_references_info.index\n",
    "nature_references_info.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Test on two papers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "matched_df = get_reference_map_onur((references[1], tex_ref_mappings[1]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "matched_dfs = matched_df.append(get_reference_map_onur((references[2], tex_ref_mappings[2]))).set_index('citing_doi')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CitingArticleID</th>\n",
       "      <th>CitedArticleID</th>\n",
       "      <th>ArticleID</th>\n",
       "      <th>PubYear</th>\n",
       "      <th>Title - matched_to</th>\n",
       "      <th>Volume</th>\n",
       "      <th>Issue</th>\n",
       "      <th>Cites</th>\n",
       "      <th>Journal</th>\n",
       "      <th>NatureID</th>\n",
       "      <th>DOI</th>\n",
       "      <th>index</th>\n",
       "      <th>snm</th>\n",
       "      <th>Title - matched_from</th>\n",
       "      <th>year</th>\n",
       "      <th>journal</th>\n",
       "      <th>fnm</th>\n",
       "      <th>text</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>citing_doi</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10.1038/nature03697</th>\n",
       "      <td>pub.1003700634</td>\n",
       "      <td>pub.1011515456</td>\n",
       "      <td>pub.1011515456</td>\n",
       "      <td>2002</td>\n",
       "      <td>Kerguelen Hotspot Magma Output since 130 Ma</td>\n",
       "      <td>43</td>\n",
       "      <td>7</td>\n",
       "      <td>224</td>\n",
       "      <td>Journal of Petrology</td>\n",
       "      <td></td>\n",
       "      <td>10.1093/petrology/43.7.1121</td>\n",
       "      <td>b7</td>\n",
       "      <td>['Coffin']</td>\n",
       "      <td>Kerguelen hotspot magma output since 130 Ma</td>\n",
       "      <td>2002</td>\n",
       "      <td>J. Petrol.</td>\n",
       "      <td>['M. F.']</td>\n",
       "      <td>[But recent geochronological data&lt;bibr rid=\"b4...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10.1038/nature03697</th>\n",
       "      <td>pub.1003700634</td>\n",
       "      <td>pub.1029752757</td>\n",
       "      <td>pub.1029752757</td>\n",
       "      <td>1988</td>\n",
       "      <td>Deccan flood basalts and the Cretaceous/Tertia...</td>\n",
       "      <td>333</td>\n",
       "      <td>6176</td>\n",
       "      <td>333</td>\n",
       "      <td>Nature</td>\n",
       "      <td>333843a0</td>\n",
       "      <td>10.1038/333843a0</td>\n",
       "      <td>b3</td>\n",
       "      <td>['Courtillot']</td>\n",
       "      <td>Deccan flood basalts on Cretaceous/Tertiary bo...</td>\n",
       "      <td>1988</td>\n",
       "      <td>Nature</td>\n",
       "      <td>['V.']</td>\n",
       "      <td>[The Siberian Traps at the Permian&amp;ndash;Trias...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10.1038/nature03697</th>\n",
       "      <td>pub.1003700634</td>\n",
       "      <td>pub.1033167081</td>\n",
       "      <td>pub.1033167081</td>\n",
       "      <td>2001</td>\n",
       "      <td>Mineralogy of subducted basaltic crust (MORB) ...</td>\n",
       "      <td>190</td>\n",
       "      <td>1-2</td>\n",
       "      <td>185</td>\n",
       "      <td>Earth and Planetary Science Letters</td>\n",
       "      <td></td>\n",
       "      <td>10.1016/s0012-821x(01)00375-2</td>\n",
       "      <td>b24</td>\n",
       "      <td>['Ono', 'Ito', 'Katsura']</td>\n",
       "      <td>Mineralogy of subducted basaltic crust (MORB) ...</td>\n",
       "      <td>2001</td>\n",
       "      <td>Earth Planet. Sci. Lett.</td>\n",
       "      <td>['S.', 'E.', 'T.']</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10.1038/nature03697</th>\n",
       "      <td>pub.1003700634</td>\n",
       "      <td>pub.1062550794</td>\n",
       "      <td>pub.1062550794</td>\n",
       "      <td>1995</td>\n",
       "      <td>Synchrony and Causal Relations Between Permian...</td>\n",
       "      <td>269</td>\n",
       "      <td>5229</td>\n",
       "      <td>425</td>\n",
       "      <td>Science</td>\n",
       "      <td></td>\n",
       "      <td>10.1126/science.269.5229.1413</td>\n",
       "      <td>b2</td>\n",
       "      <td>['Renne', 'Zichao', 'Richards', 'Black', 'Basu']</td>\n",
       "      <td>Synchrony and causal relations between Permian...</td>\n",
       "      <td>1995</td>\n",
       "      <td>Science</td>\n",
       "      <td>['P. R.', 'Z.', 'M. A.', 'M. T.', 'A. R.']</td>\n",
       "      <td>[The Siberian Traps at the Permian&amp;ndash;Trias...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10.1038/nature03697</th>\n",
       "      <td>pub.1003700634</td>\n",
       "      <td>pub.1000115762</td>\n",
       "      <td>pub.1000115762</td>\n",
       "      <td>1999</td>\n",
       "      <td>Discrete alternating hotspot islands formed by...</td>\n",
       "      <td>397</td>\n",
       "      <td>6720</td>\n",
       "      <td>59</td>\n",
       "      <td>Nature</td>\n",
       "      <td>17584</td>\n",
       "      <td>10.1038/17584</td>\n",
       "      <td>b15</td>\n",
       "      <td>['Hieronymus', 'Bercovici']</td>\n",
       "      <td>Discrete alternating hotspot islands formed by...</td>\n",
       "      <td>1999</td>\n",
       "      <td>Nature</td>\n",
       "      <td>['C. F.', 'D.']</td>\n",
       "      <td>[Other models, such as solitary waves travelli...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10.1038/nature01599</th>\n",
       "      <td>pub.1047617530</td>\n",
       "      <td>pub.1046838763</td>\n",
       "      <td>pub.1046838763</td>\n",
       "      <td>2000</td>\n",
       "      <td>A dynamic duo of microtubule modulators</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>Nature Cell Biology</td>\n",
       "      <td>71394</td>\n",
       "      <td>10.1038/71394</td>\n",
       "      <td>b34</td>\n",
       "      <td>['Heald']</td>\n",
       "      <td>A dynamic duo of microtubule regulators</td>\n",
       "      <td>1999</td>\n",
       "      <td>Nature Cell Biol.</td>\n",
       "      <td>['R.']</td>\n",
       "      <td>[These proteins could regulate the length of s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10.1038/nature01599</th>\n",
       "      <td>pub.1047617530</td>\n",
       "      <td>pub.1018459115</td>\n",
       "      <td>pub.1018459115</td>\n",
       "      <td>2002</td>\n",
       "      <td>Mad2 and BubR1 Function in a Single Checkpoint...</td>\n",
       "      <td>13</td>\n",
       "      <td>10</td>\n",
       "      <td>80</td>\n",
       "      <td>Molecular Biology of the Cell</td>\n",
       "      <td></td>\n",
       "      <td>10.1091/mbc.e02-03-0137</td>\n",
       "      <td>b61</td>\n",
       "      <td>['Shannon', 'Canman', 'Salmon']</td>\n",
       "      <td>Mad2 and BubR1 function in a single checkpoint...</td>\n",
       "      <td>2002</td>\n",
       "      <td>Mol. Biol. Cell</td>\n",
       "      <td>['K. B.', 'J. C.', 'E. D.']</td>\n",
       "      <td>[Accordingly, different experiments have led t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10.1038/nature01599</th>\n",
       "      <td>pub.1047617530</td>\n",
       "      <td>pub.1032265523</td>\n",
       "      <td>pub.1032265523</td>\n",
       "      <td>2002</td>\n",
       "      <td>Nuclear Envelope Breakdown Proceeds by Microtu...</td>\n",
       "      <td>108</td>\n",
       "      <td>1</td>\n",
       "      <td>310</td>\n",
       "      <td>Cell</td>\n",
       "      <td></td>\n",
       "      <td>10.1016/s0092-8674(01)00627-4</td>\n",
       "      <td>b63</td>\n",
       "      <td>['Beaudouin', 'Gerlich', 'Daigle', 'Eils', 'El...</td>\n",
       "      <td>Nuclear envelope breakdown proceeds by microtu...</td>\n",
       "      <td>2002</td>\n",
       "      <td>Cell</td>\n",
       "      <td>['J.', 'D.', 'N.', 'R.', 'J.']</td>\n",
       "      <td>[We speculate that, as in metaphase, these thr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10.1038/nature01599</th>\n",
       "      <td>pub.1047617530</td>\n",
       "      <td>pub.1000914405</td>\n",
       "      <td>pub.1000914405</td>\n",
       "      <td>2002</td>\n",
       "      <td>Cytokinesis in Eukaryotes</td>\n",
       "      <td>66</td>\n",
       "      <td>2</td>\n",
       "      <td>204</td>\n",
       "      <td>Microbiology and Molecular Biology Reviews</td>\n",
       "      <td></td>\n",
       "      <td>10.1128/mmbr.66.2.155-178.2002</td>\n",
       "      <td>b74</td>\n",
       "      <td>['Guertin', 'Trautmann', 'McCollum']</td>\n",
       "      <td>Cytokinesis in eukaryotes</td>\n",
       "      <td>2002</td>\n",
       "      <td>Microbiol. Mol. Biol. Rev.</td>\n",
       "      <td>['D. A.', 'S.', 'D.']</td>\n",
       "      <td>[The myosin-dependent equatorial contraction m...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10.1038/nature01599</th>\n",
       "      <td>pub.1047617530</td>\n",
       "      <td>pub.1038946241</td>\n",
       "      <td>pub.1038946241</td>\n",
       "      <td>2001</td>\n",
       "      <td>Mitosis: a history of division</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>182</td>\n",
       "      <td>Nature Cell Biology</td>\n",
       "      <td>35050656</td>\n",
       "      <td>10.1038/35050656</td>\n",
       "      <td>b2</td>\n",
       "      <td>['Mitchison', 'Salmon']</td>\n",
       "      <td>Mitosis: a history of division</td>\n",
       "      <td>2001</td>\n",
       "      <td>Nature Cell Biol.</td>\n",
       "      <td>['T. J.', 'E. D.']</td>\n",
       "      <td>[&lt;/abs&gt;\\r\\n&lt;/fm&gt;\\r\\n&lt;bdy&gt;\\r\\n&lt;p&gt;During the nin...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>97 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    CitingArticleID  CitedArticleID       ArticleID  PubYear  \\\n",
       "citing_doi                                                                     \n",
       "10.1038/nature03697  pub.1003700634  pub.1011515456  pub.1011515456     2002   \n",
       "10.1038/nature03697  pub.1003700634  pub.1029752757  pub.1029752757     1988   \n",
       "10.1038/nature03697  pub.1003700634  pub.1033167081  pub.1033167081     2001   \n",
       "10.1038/nature03697  pub.1003700634  pub.1062550794  pub.1062550794     1995   \n",
       "10.1038/nature03697  pub.1003700634  pub.1000115762  pub.1000115762     1999   \n",
       "...                             ...             ...             ...      ...   \n",
       "10.1038/nature01599  pub.1047617530  pub.1046838763  pub.1046838763     2000   \n",
       "10.1038/nature01599  pub.1047617530  pub.1018459115  pub.1018459115     2002   \n",
       "10.1038/nature01599  pub.1047617530  pub.1032265523  pub.1032265523     2002   \n",
       "10.1038/nature01599  pub.1047617530  pub.1000914405  pub.1000914405     2002   \n",
       "10.1038/nature01599  pub.1047617530  pub.1038946241  pub.1038946241     2001   \n",
       "\n",
       "                                                    Title - matched_to Volume  \\\n",
       "citing_doi                                                                      \n",
       "10.1038/nature03697        Kerguelen Hotspot Magma Output since 130 Ma     43   \n",
       "10.1038/nature03697  Deccan flood basalts and the Cretaceous/Tertia...    333   \n",
       "10.1038/nature03697  Mineralogy of subducted basaltic crust (MORB) ...    190   \n",
       "10.1038/nature03697  Synchrony and Causal Relations Between Permian...    269   \n",
       "10.1038/nature03697  Discrete alternating hotspot islands formed by...    397   \n",
       "...                                                                ...    ...   \n",
       "10.1038/nature01599            A dynamic duo of microtubule modulators      2   \n",
       "10.1038/nature01599  Mad2 and BubR1 Function in a Single Checkpoint...     13   \n",
       "10.1038/nature01599  Nuclear Envelope Breakdown Proceeds by Microtu...    108   \n",
       "10.1038/nature01599                          Cytokinesis in Eukaryotes     66   \n",
       "10.1038/nature01599                     Mitosis: a history of division      3   \n",
       "\n",
       "                    Issue  Cites                                     Journal  \\\n",
       "citing_doi                                                                     \n",
       "10.1038/nature03697     7    224                        Journal of Petrology   \n",
       "10.1038/nature03697  6176    333                                      Nature   \n",
       "10.1038/nature03697   1-2    185         Earth and Planetary Science Letters   \n",
       "10.1038/nature03697  5229    425                                     Science   \n",
       "10.1038/nature03697  6720     59                                      Nature   \n",
       "...                   ...    ...                                         ...   \n",
       "10.1038/nature01599     1     15                         Nature Cell Biology   \n",
       "10.1038/nature01599    10     80               Molecular Biology of the Cell   \n",
       "10.1038/nature01599     1    310                                        Cell   \n",
       "10.1038/nature01599     2    204  Microbiology and Molecular Biology Reviews   \n",
       "10.1038/nature01599     1    182                         Nature Cell Biology   \n",
       "\n",
       "                     NatureID                             DOI index  \\\n",
       "citing_doi                                                            \n",
       "10.1038/nature03697               10.1093/petrology/43.7.1121    b7   \n",
       "10.1038/nature03697  333843a0                10.1038/333843a0    b3   \n",
       "10.1038/nature03697             10.1016/s0012-821x(01)00375-2   b24   \n",
       "10.1038/nature03697             10.1126/science.269.5229.1413    b2   \n",
       "10.1038/nature03697     17584                   10.1038/17584   b15   \n",
       "...                       ...                             ...   ...   \n",
       "10.1038/nature01599     71394                   10.1038/71394   b34   \n",
       "10.1038/nature01599                   10.1091/mbc.e02-03-0137   b61   \n",
       "10.1038/nature01599             10.1016/s0092-8674(01)00627-4   b63   \n",
       "10.1038/nature01599            10.1128/mmbr.66.2.155-178.2002   b74   \n",
       "10.1038/nature01599  35050656                10.1038/35050656    b2   \n",
       "\n",
       "                                                                   snm  \\\n",
       "citing_doi                                                               \n",
       "10.1038/nature03697                                         ['Coffin']   \n",
       "10.1038/nature03697                                     ['Courtillot']   \n",
       "10.1038/nature03697                          ['Ono', 'Ito', 'Katsura']   \n",
       "10.1038/nature03697   ['Renne', 'Zichao', 'Richards', 'Black', 'Basu']   \n",
       "10.1038/nature03697                        ['Hieronymus', 'Bercovici']   \n",
       "...                                                                ...   \n",
       "10.1038/nature01599                                          ['Heald']   \n",
       "10.1038/nature01599                    ['Shannon', 'Canman', 'Salmon']   \n",
       "10.1038/nature01599  ['Beaudouin', 'Gerlich', 'Daigle', 'Eils', 'El...   \n",
       "10.1038/nature01599               ['Guertin', 'Trautmann', 'McCollum']   \n",
       "10.1038/nature01599                            ['Mitchison', 'Salmon']   \n",
       "\n",
       "                                                  Title - matched_from  year  \\\n",
       "citing_doi                                                                     \n",
       "10.1038/nature03697        Kerguelen hotspot magma output since 130 Ma  2002   \n",
       "10.1038/nature03697  Deccan flood basalts on Cretaceous/Tertiary bo...  1988   \n",
       "10.1038/nature03697  Mineralogy of subducted basaltic crust (MORB) ...  2001   \n",
       "10.1038/nature03697  Synchrony and causal relations between Permian...  1995   \n",
       "10.1038/nature03697  Discrete alternating hotspot islands formed by...  1999   \n",
       "...                                                                ...   ...   \n",
       "10.1038/nature01599            A dynamic duo of microtubule regulators  1999   \n",
       "10.1038/nature01599  Mad2 and BubR1 function in a single checkpoint...  2002   \n",
       "10.1038/nature01599  Nuclear envelope breakdown proceeds by microtu...  2002   \n",
       "10.1038/nature01599                          Cytokinesis in eukaryotes  2002   \n",
       "10.1038/nature01599                     Mitosis: a history of division  2001   \n",
       "\n",
       "                                        journal  \\\n",
       "citing_doi                                        \n",
       "10.1038/nature03697                  J. Petrol.   \n",
       "10.1038/nature03697                      Nature   \n",
       "10.1038/nature03697    Earth Planet. Sci. Lett.   \n",
       "10.1038/nature03697                     Science   \n",
       "10.1038/nature03697                      Nature   \n",
       "...                                         ...   \n",
       "10.1038/nature01599           Nature Cell Biol.   \n",
       "10.1038/nature01599             Mol. Biol. Cell   \n",
       "10.1038/nature01599                        Cell   \n",
       "10.1038/nature01599  Microbiol. Mol. Biol. Rev.   \n",
       "10.1038/nature01599           Nature Cell Biol.   \n",
       "\n",
       "                                                            fnm  \\\n",
       "citing_doi                                                        \n",
       "10.1038/nature03697                                   ['M. F.']   \n",
       "10.1038/nature03697                                      ['V.']   \n",
       "10.1038/nature03697                          ['S.', 'E.', 'T.']   \n",
       "10.1038/nature03697  ['P. R.', 'Z.', 'M. A.', 'M. T.', 'A. R.']   \n",
       "10.1038/nature03697                             ['C. F.', 'D.']   \n",
       "...                                                         ...   \n",
       "10.1038/nature01599                                      ['R.']   \n",
       "10.1038/nature01599                 ['K. B.', 'J. C.', 'E. D.']   \n",
       "10.1038/nature01599              ['J.', 'D.', 'N.', 'R.', 'J.']   \n",
       "10.1038/nature01599                       ['D. A.', 'S.', 'D.']   \n",
       "10.1038/nature01599                          ['T. J.', 'E. D.']   \n",
       "\n",
       "                                                                  text  \n",
       "citing_doi                                                              \n",
       "10.1038/nature03697  [But recent geochronological data<bibr rid=\"b4...  \n",
       "10.1038/nature03697  [The Siberian Traps at the Permian&ndash;Trias...  \n",
       "10.1038/nature03697                                                NaN  \n",
       "10.1038/nature03697  [The Siberian Traps at the Permian&ndash;Trias...  \n",
       "10.1038/nature03697  [Other models, such as solitary waves travelli...  \n",
       "...                                                                ...  \n",
       "10.1038/nature01599  [These proteins could regulate the length of s...  \n",
       "10.1038/nature01599  [Accordingly, different experiments have led t...  \n",
       "10.1038/nature01599  [We speculate that, as in metaphase, these thr...  \n",
       "10.1038/nature01599  [The myosin-dependent equatorial contraction m...  \n",
       "10.1038/nature01599  [</abs>\\r\\n</fm>\\r\\n<bdy>\\r\\n<p>During the nin...  \n",
       "\n",
       "[97 rows x 18 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matched_dfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "papers = {k: g.to_dict(orient='records') for k, g in matched_dfs.groupby(level=0)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "for reference in references: \n",
    "    \n",
    "    if 'doi' in reference['metadata']:\n",
    "        \n",
    "        if reference['metadata']['doi'] in papers:\n",
    "            \n",
    "            papers[reference['metadata']['doi']] = {'metadata' : reference['metadata'],\n",
    "                                                'references': papers[reference['metadata']['doi']]}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'metadata': {'doi': '10.1038/nature06234',\n",
       "  'id': 'nature06234',\n",
       "  'jtl': 'Nature',\n",
       "  'language': 'eng',\n",
       "  'publish': 'issue',\n",
       "  'relation': 'no',\n",
       "  'title': 'Controlling cavity reflectivity with a single quantum dot',\n",
       "  'vol': '450'},\n",
       " 'references': {'b1': {'fnm': \"['D.']\",\n",
       "   'journal': 'Phys. Rev. Lett.',\n",
       "   'snm': \"['Englund']\",\n",
       "   'title': 'Controlling the spontaneous emission rate of single quantum dots in a two-dimensional photonic crystal',\n",
       "   'year': '2005'},\n",
       "  'b10': {'fnm': \"['K. M.']\",\n",
       "   'journal': 'Nature',\n",
       "   'snm': \"['Birnbaum']\",\n",
       "   'title': 'Photon blockade in an optical cavity with one trapped atom',\n",
       "   'year': '2005'},\n",
       "  'b11': {'fnm': \"['A.']\",\n",
       "   'journal': 'Phys. Rev. Lett.',\n",
       "   'snm': \"['Rauschenbeutel']\",\n",
       "   'title': 'Coherent operation of a tunable quantum phase gate in cavity QED',\n",
       "   'year': '1999'},\n",
       "  'b12': {'fnm': \"['G.']\",\n",
       "   'journal': 'Nature',\n",
       "   'snm': \"['Nogues']\",\n",
       "   'title': 'Seeing a single photon without destroying it',\n",
       "   'year': '1999'},\n",
       "  'b13': {'fnm': \"['D. I.']\",\n",
       "   'journal': 'Nature',\n",
       "   'snm': \"['Schuster']\",\n",
       "   'title': 'Resolving photon number states in a superconducting circuit',\n",
       "   'year': '2007'},\n",
       "  'b15': {'fnm': \"['Y.', 'T.', 'B.-S.', 'S.']\",\n",
       "   'journal': 'Nature',\n",
       "   'snm': \"['Akahane', 'Asano', 'Song', 'Noda']\",\n",
       "   'title': 'High-Q photonic nanocavity in a two-dimensional photonic crystal',\n",
       "   'year': '2003'},\n",
       "  'b16': {'fnm': \"['A.']\",\n",
       "   'journal': 'Appl. Phys. Lett.',\n",
       "   'snm': \"['Faraon']\",\n",
       "   'title': 'Local quantum dot tuning on photonic crystal chips',\n",
       "   'year': '2007'},\n",
       "  'b17': {'fnm': \"['B. D.']\",\n",
       "   'journal': 'Appl. Phys. Lett.',\n",
       "   'snm': \"['Gerardot']\",\n",
       "   'title': 'Contrast in transmission spectroscopy of a single quantum dot',\n",
       "   'year': '2007'},\n",
       "  'b18': {'fnm': \"['S. M.']\",\n",
       "   'journal': 'J. Opt. B',\n",
       "   'snm': \"['Tan']\",\n",
       "   'title': 'A computational toolbox for quantum and atomic physics',\n",
       "   'year': '1999'},\n",
       "  'b19': {'fnm': \"['C. J.', 'M. S.', 'T. W.', 'H. J.']\",\n",
       "   'journal': 'Phys. Rev. Lett.',\n",
       "   'snm': \"['Hood', 'Chapman', 'Lynn', 'Kimble']\",\n",
       "   'title': 'Real-time cavity QED with single atoms',\n",
       "   'year': '1998'},\n",
       "  'b2': {'fnm': \"['T.']\",\n",
       "   'journal': 'Nature',\n",
       "   'snm': \"['Yoshie']\",\n",
       "   'title': 'Vacuum Rabi splitting with a single quantum dot in a photonic crystal nanocavity',\n",
       "   'year': '2004'},\n",
       "  'b20': {'fnm': \"['A.', 'C.', 'J. M.', 'J.-P.']\",\n",
       "   'journal': 'Phys. Rev. A',\n",
       "   'snm': \"['Auffeves-Garnier', 'Simon', 'Gerard', 'Poizat']\",\n",
       "   'title': 'Giant optical nonlinearity induced by a single two-level system interacting with a cavity in the Purcell regime',\n",
       "   'year': '2007'},\n",
       "  'b21': {'fnm': \"['D.', 'A.', 'B.', 'Y.', 'J.']\",\n",
       "   'journal': 'Opt. Express',\n",
       "   'snm': \"['Englund', 'Faraon', 'Zhang', 'Yamamoto', 'Vu&ccaron;kovi&cacute;']\",\n",
       "   'title': 'Generation and transfer of single photons on a photonic crystal chip',\n",
       "   'year': '2007'},\n",
       "  'b22': {'fnm': \"['A.', 'E.', 'D.', 'I.', 'J.']\",\n",
       "   'journal': 'Appl. Phys. Lett.',\n",
       "   'snm': \"['Faraon', 'Waks', 'Englund', 'Fushman', 'Vu&ccaron;kovi&cacute;']\",\n",
       "   'title': 'Efficient photonic crystal cavity-waveguide couplers',\n",
       "   'year': '2007'},\n",
       "  'b23': {'fnm': \"['J. E.', 'W. P.', 'L. A.', 'H. J.', 'P. R.']\",\n",
       "   'journal': 'J. Opt. Soc. Am. B',\n",
       "   'snm': \"['Reiner', 'Smith', 'Orozco', 'Carmichael', 'Rice']\",\n",
       "   'title': 'Time evolution and squeezing of the field amplitude in cavity QED',\n",
       "   'year': '2001'},\n",
       "  'b24': {'fnm': \"['N.', 'H. A.', 'Y.']\",\n",
       "   'journal': 'Phys. Rev. A',\n",
       "   'snm': \"['Imoto', 'Haus', 'Yamamoto']\",\n",
       "   'title': 'Quantum nondemolition measurement of the photon number via the optical Kerr effect',\n",
       "   'year': '1985'},\n",
       "  'b25': {'fnm': \"['J.', 'D.', 'D.', 'E.', 'Y.']\",\n",
       "   'journal': 'Physica E',\n",
       "   'snm': \"['Vu&ccaron;kovi&cacute;', 'Englund', 'Fattal', 'Waks', 'Yamamoto']\",\n",
       "   'title': 'Generation and manipulation of nonclassical light using photonic crystals',\n",
       "   'year': '2006'},\n",
       "  'b3': {'fnm': \"['K.']\",\n",
       "   'journal': 'Nature',\n",
       "   'snm': \"['Hennessy']\",\n",
       "   'title': 'Quantum nature of a strongly coupled single quantum dot-cavity system',\n",
       "   'year': '2007'},\n",
       "  'b4': {'fnm': \"['J. I.', 'P.', 'H. J.', 'H.']\",\n",
       "   'journal': 'Phys. Rev. Lett.',\n",
       "   'snm': \"['Cirac', 'Zoller', 'Kimble', 'Mabuchi']\",\n",
       "   'title': 'Quantum state transfer and entanglement distribution among distant nodes in a quantum network',\n",
       "   'year': '1997'},\n",
       "  'b5': {'fnm': \"['A.']\",\n",
       "   'journal': 'Phys. Rev. Lett.',\n",
       "   'snm': \"['Imamoglu']\",\n",
       "   'title': 'Quantum information processing using quantum dot spins and cavity QED',\n",
       "   'year': '1999'},\n",
       "  'b6': {'fnm': \"['L. M.', 'H. J.']\",\n",
       "   'journal': 'Phys. Rev. Lett.',\n",
       "   'snm': \"['Duan', 'Kimble']\",\n",
       "   'title': 'Scalable photonic quantum computation through cavity-assisted interactions',\n",
       "   'year': '2004'},\n",
       "  'b7': {'fnm': \"['L.', 'J. M.', 'A. S.', 'M. D.']\",\n",
       "   'journal': 'Phys. Rev. A',\n",
       "   'snm': \"['Childress', 'Taylor', 'Sorensen', 'Lukin']\",\n",
       "   'title': 'Fault-tolerant quantum repeaters with minimal physical resources and implementations based on single-photon emitters',\n",
       "   'year': '2005'},\n",
       "  'b8': {'fnm': \"['E.', 'J.']\",\n",
       "   'journal': 'Phys. Rev. Lett.',\n",
       "   'snm': \"['Waks', 'Vu&ccaron;kovi&cacute;']\",\n",
       "   'title': 'Dipole induced transparency in drop-filter cavity-waveguide systems',\n",
       "   'year': '2006'},\n",
       "  'b9': {'fnm': \"['T. D.', 'P. K.', 'K.', 'W. J.', 'Y.']\",\n",
       "   'journal': 'N. J. Phys.',\n",
       "   'snm': \"['Ladd', 'van Loock', 'Nemoto', 'Munro', 'Yamamoto']\",\n",
       "   'title': 'Hybrid quantum repeater based on dispersive CQED interactions between matter qubits and bright coherent light',\n",
       "   'year': '2006'}}}"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "14274"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(papers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ArticleID         pub.1047617530\n",
       "PubYear                     2003\n",
       "Title              Cell division\n",
       "Volume                       422\n",
       "Issue                       6933\n",
       "Cites                        216\n",
       "Journal                   Nature\n",
       "NatureID             nature01599\n",
       "DOI          10.1038/nature01599\n",
       "Name: 10.1038/nature01599, dtype: object"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nature_references_info.loc['10.1038/nature01599']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "doi: 10.1038/nature11182b\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature11182b\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature.2017.21620\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature22808\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature01206\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature11870_techniques\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature21729\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature11847_techniques\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature11942_techniques\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature11893_techniques\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature11977_techniques\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature21898\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature11047a\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature14522\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature11363d\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature11363c\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature07825\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature11898_techniques\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature16134\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature10829aa\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature_techniques12367\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature10829b\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature18281\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature11363\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature08547\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature16470\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature11581a\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature11182b\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature.2017.21462\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature.2017.21339\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature11717a\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature11182b\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature17890\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature11192h\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature11192a\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature.2016.20649\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "doi: 10.1038/nature11192a\n",
      "index 0 is out of bounds for axis 0 with size 0\n",
      "CPU times: user 30.7 s, sys: 6.63 s, total: 37.4 s\n",
      "Wall time: 3h 5min 27s\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Process ForkPoolWorker-7:\n",
      "Process ForkPoolWorker-8:\n",
      "Process ForkPoolWorker-5:\n",
      "Process ForkPoolWorker-1:\n",
      "Process ForkPoolWorker-4:\n",
      "Process ForkPoolWorker-3:\n",
      "Process ForkPoolWorker-2:\n",
      "Process ForkPoolWorker-6:\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/lib/python3.5/multiprocessing/process.py\", line 249, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/process.py\", line 249, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/process.py\", line 249, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/process.py\", line 249, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/process.py\", line 249, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/process.py\", line 249, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/process.py\", line 93, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/lib/python3.5/multiprocessing/process.py\", line 93, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/lib/python3.5/multiprocessing/process.py\", line 249, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/process.py\", line 93, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/lib/python3.5/multiprocessing/process.py\", line 93, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/lib/python3.5/multiprocessing/process.py\", line 93, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/lib/python3.5/multiprocessing/pool.py\", line 108, in worker\n",
      "    task = get()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/pool.py\", line 108, in worker\n",
      "    task = get()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/pool.py\", line 108, in worker\n",
      "    task = get()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/process.py\", line 93, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/lib/python3.5/multiprocessing/pool.py\", line 108, in worker\n",
      "    task = get()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/process.py\", line 93, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/lib/python3.5/multiprocessing/pool.py\", line 108, in worker\n",
      "    task = get()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/queues.py\", line 342, in get\n",
      "    with self._rlock:\n",
      "  File \"/usr/lib/python3.5/multiprocessing/queues.py\", line 342, in get\n",
      "    with self._rlock:\n",
      "  File \"/usr/lib/python3.5/multiprocessing/pool.py\", line 108, in worker\n",
      "    task = get()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/queues.py\", line 342, in get\n",
      "    with self._rlock:\n",
      "  File \"/usr/lib/python3.5/multiprocessing/queues.py\", line 342, in get\n",
      "    with self._rlock:\n",
      "  File \"/usr/lib/python3.5/multiprocessing/pool.py\", line 108, in worker\n",
      "    task = get()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/synchronize.py\", line 96, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/synchronize.py\", line 96, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/queues.py\", line 342, in get\n",
      "    with self._rlock:\n",
      "  File \"/usr/lib/python3.5/multiprocessing/synchronize.py\", line 96, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/queues.py\", line 342, in get\n",
      "    with self._rlock:\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/lib/python3.5/multiprocessing/synchronize.py\", line 96, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/queues.py\", line 342, in get\n",
      "    with self._rlock:\n",
      "  File \"/usr/lib/python3.5/multiprocessing/synchronize.py\", line 96, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "KeyboardInterrupt\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/lib/python3.5/multiprocessing/synchronize.py\", line 96, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/synchronize.py\", line 96, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "KeyboardInterrupt\n",
      "KeyboardInterrupt\n",
      "KeyboardInterrupt\n",
      "KeyboardInterrupt\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/lib/python3.5/multiprocessing/process.py\", line 249, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/process.py\", line 93, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/lib/python3.5/multiprocessing/pool.py\", line 108, in worker\n",
      "    task = get()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/queues.py\", line 343, in get\n",
      "    res = self._reader.recv_bytes()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/connection.py\", line 216, in recv_bytes\n",
      "    buf = self._recv_bytes(maxlength)\n",
      "  File \"/usr/lib/python3.5/multiprocessing/connection.py\", line 407, in _recv_bytes\n",
      "    buf = self._recv(4)\n",
      "  File \"/usr/lib/python3.5/multiprocessing/connection.py\", line 379, in _recv\n",
      "    chunk = read(handle, remaining)\n",
      "KeyboardInterrupt\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "mp_input = zip(references, tex_ref_mappings)\n",
    "p        = Pool(8)\n",
    "\n",
    "matched_dfs = p.map(get_reference_map_onur, mp_input)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('data/intermediate/matched_dfs_raw.pkl', 'rb') as f:\n",
    "    matched_dfs = pkl.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "matched_dfs = pd.concat(matched_dfs).set_index('citing_doi')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "matched_dfs['text'] = matched_dfs['text'].fillna('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "matched_dfs['clean_text'] = matched_dfs['text'].apply(lambda x: [BeautifulSoup(i, \"lxml\").text for i in x])\n",
    "matched_dfs['count'] = matched_dfs['clean_text'].apply(lambda x: len(x))\n",
    "#matched_dfs = matched_dfs[matched_dfs['count'] != 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.5/dist-packages/ipykernel_launcher.py:1: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0\n",
      "Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`\n",
      "  \"\"\"Entry point for launching an IPython kernel.\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "82af958ba18846d3a12c612c3bf485c1",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=0, max=14274), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "papers = {k: g.to_dict(orient='records') for k, g in tqdm(matched_dfs.groupby(level=0))}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "for reference in references: \n",
    "    \n",
    "    if 'doi' in reference['metadata']:\n",
    "        \n",
    "        if reference['metadata']['doi'] in papers:\n",
    "            \n",
    "            papers[reference['metadata']['doi']] = {'metadata' : reference['metadata'],\n",
    "                                                'references': papers[reference['metadata']['doi']]}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('data/intermediate/matched.jsons', 'w') as f:\n",
    "\n",
    "        for index, paper in papers.items():\n",
    "\n",
    "            f.write('\\n')\n",
    "            f.write(json.dumps({index:paper}))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "papers['10.1038/nature06452']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Expecting value: line 2 column 1 (char 1)\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "with open('data/intermediate/matched.jsons', 'r') as f:\n",
    "\n",
    "    for i, line in enumerate(f.readlines()):\n",
    "        \n",
    "        try:\n",
    "            unit_paper_dict = json.loads(line)\n",
    "        \n",
    "        except json.JSONDecodeError as e:\n",
    "            \n",
    "            print(e)\n",
    "            print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'10.1038/nature09207': {'metadata': {'doi': '10.1038/nature09207',\n",
       "   'id': 'nature09207',\n",
       "   'jtl': 'Nature',\n",
       "   'language': 'eng',\n",
       "   'relation': 'no',\n",
       "   'title': 'Microtubule nucleating -TuSC assembles structures with 13-fold microtubule-like symmetry',\n",
       "   'vol': '466'},\n",
       "  'references': [{'ArticleID': 'pub.1010121551',\n",
       "    'CitedArticleID': 'pub.1010121551',\n",
       "    'Cites': 165,\n",
       "    'CitingArticleID': 'pub.1030133211',\n",
       "    'DOI': '10.1083/jcb.100.4.1185',\n",
       "    'Issue': '4',\n",
       "    'Journal': 'The Journal of Cell Biology',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 1985,\n",
       "    'Title - matched_from': 'Influence of the centrosome on the structure of nucleated microtubules',\n",
       "    'Title - matched_to': 'Influence of the centrosome on the structure of nucleated microtubules.',\n",
       "    'Volume': '100',\n",
       "    'clean_text': ['However, microtubules nucleated in cells mostly have 13 protofilaments, suggesting that &ggr;-tubulin complexes constrain microtubule geometry.',\n",
       "     'We propose that a &ggr;-TuSC assembly very similar to a single ring from the filament provides the constraint that limits microtubules to 13 protofilaments in all eukaryotes in vivo.'],\n",
       "    'count': 2,\n",
       "    'fnm': \"['L.', 'T.', 'M.']\",\n",
       "    'index': 'b7',\n",
       "    'journal': 'J. Cell Biol.',\n",
       "    'snm': \"['Evans', 'Mitchison', 'Kirschner']\",\n",
       "    'text': ['However, microtubules nucleated in cells mostly have 13 protofilaments<bibr rid=\"b7\"/>, suggesting that &ggr;-tubulin complexes constrain microtubule geometry.',\n",
       "     'We propose that a &ggr;-TuSC assembly very similar to a single ring from the filament provides the constraint that limits microtubules to 13 protofilaments in all eukaryotes <i>in vivo</i><bibr rid=\"b7\"/>.'],\n",
       "    'year': '1985'},\n",
       "   {'ArticleID': 'pub.1048438939',\n",
       "    'CitedArticleID': 'pub.1048438939',\n",
       "    'Cites': 23,\n",
       "    'CitingArticleID': 'pub.1030133211',\n",
       "    'DOI': '10.1016/j.jsb.2009.08.012',\n",
       "    'Issue': '3',\n",
       "    'Journal': 'Journal of Structural Biology',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 2009,\n",
       "    'Title - matched_from': 'Localization and orientation of the &ggr;-tubulin small complex components using protein tags as labels for single particle EM',\n",
       "    'Title - matched_to': 'Localization and orientation of the γ-Tubulin Small Complex components using protein tags as labels for single particle EM',\n",
       "    'Volume': '168',\n",
       "    'clean_text': ['&ggr;-TuSC in the filament is remarkably similar to free &ggr;-TuSC (Fig.',\n",
       "     'We previously predicted that bending at the flexible kink in Spc98 is required to bring the intra-&ggr;-TuSC &ggr;-tubulins to the microtubule spacing.'],\n",
       "    'count': 2,\n",
       "    'fnm': \"['R. M.', 'J. M.', 'A.', 'T. N.', 'D. A.']\",\n",
       "    'index': 'b18',\n",
       "    'journal': 'J. Struct. Biol.',\n",
       "    'snm': \"['Choy', 'Kollman', 'Zelter', 'Davis', 'Agard']\",\n",
       "    'text': ['\\r\\n<p>&ggr;-TuSC in the filament is remarkably similar to free &ggr;-TuSC<bibr rid=\"b13 b18\"/> (<figr rid=\"f2\">Fig.',\n",
       "     'We previously predicted that bending at the flexible kink in Spc98 is required to bring the intra-&ggr;-TuSC &ggr;-tubulins to the microtubule spacing<bibr rid=\"b13 b18\"/>.'],\n",
       "    'year': '2009'},\n",
       "   {'ArticleID': 'pub.1053279391',\n",
       "    'CitedArticleID': 'pub.1053279391',\n",
       "    'Cites': 113,\n",
       "    'CitingArticleID': 'pub.1030133211',\n",
       "    'DOI': '10.1038/nature03586',\n",
       "    'Issue': '7041',\n",
       "    'Journal': 'Nature',\n",
       "    'NatureID': 'nature03586',\n",
       "    'PubYear': 2005,\n",
       "    'Title - matched_from': 'Insights into microtubule nucleation from the crystal structure of human &ggr;-tubulin',\n",
       "    'Title - matched_to': 'Insights into microtubule nucleation from the crystal structure of human γ-tubulin',\n",
       "    'Volume': '435',\n",
       "    'clean_text': ['Our previous 25-Å structure of Saccharomyces &ggr;-TuSC was determined in buffer conditions that yielded predominantly monomeric complexes.'],\n",
       "    'count': 1,\n",
       "    'fnm': \"['H.', 'L. M.', 'T.', 'D. A.']\",\n",
       "    'index': 'b12',\n",
       "    'journal': 'Nature',\n",
       "    'snm': \"['Aldaz', 'Rice', 'Stearns', 'Agard']\",\n",
       "    'text': ['\\r\\n<p>Our previous 25-&Aring; structure of <i>Saccharomyces</i> &ggr;-TuSC was determined in buffer conditions that yielded predominantly monomeric complexes<bibr rid=\"b12 b13\"/>.'],\n",
       "    'year': '2005'},\n",
       "   {'ArticleID': 'pub.1015051998',\n",
       "    'CitedArticleID': 'pub.1015051998',\n",
       "    'Cites': 24,\n",
       "    'CitingArticleID': 'pub.1030133211',\n",
       "    'DOI': '10.1038/35014084',\n",
       "    'Issue': '6',\n",
       "    'Journal': 'Nature Cell Biology',\n",
       "    'NatureID': '35014084',\n",
       "    'PubYear': 2000,\n",
       "    'Title - matched_from': '&ggr;-Tubulin nucleation: template or protofilament?',\n",
       "    'Title - matched_to': 'γ-tubulin nucleation: template or protofilament?',\n",
       "    'Volume': '2',\n",
       "    'clean_text': ['The template model for microtubule nucleation predicts that &ggr;-tubulins interact laterally to form a ring that makes longitudinal contacts with &agr;/&bgr;-tubulin minus ends; alternative models that predict lateral interactions between &ggr;-tubulin and &agr;/&bgr;-tubulin have not been definitively ruled out.'],\n",
       "    'count': 1,\n",
       "    'fnm': \"['H. P.']\",\n",
       "    'index': 'b11',\n",
       "    'journal': 'Nature Cell Biol.',\n",
       "    'snm': \"['Erickson']\",\n",
       "    'text': ['The template model for microtubule nucleation predicts that &ggr;-tubulins interact laterally to form a ring that makes longitudinal contacts with &agr;/&bgr;-tubulin<bibr rid=\"b2 b3 b4 b5\"/> minus ends; alternative models that predict lateral interactions between &ggr;-tubulin and &agr;/&bgr;-tubulin<bibr rid=\"b11\"/> have not been definitively ruled out.'],\n",
       "    'year': '2000'},\n",
       "   {'ArticleID': 'pub.1028952493',\n",
       "    'CitedArticleID': 'pub.1028952493',\n",
       "    'Cites': 431,\n",
       "    'CitingArticleID': 'pub.1030133211',\n",
       "    'DOI': '10.1251/bpo70',\n",
       "    'Issue': '1',\n",
       "    'Journal': 'Biological Procedures Online',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 2004,\n",
       "    'Title - matched_from': 'Negative staining and image classification—powerful tools in modern electron microscopy',\n",
       "    'Title - matched_to': 'Negative staining and image classification — powerful tools in modern electron microscopy',\n",
       "    'Volume': '6',\n",
       "    'clean_text': ['Negative-stain samples were prepared as described in 0.75% uranyl formate, and cryo-electron microscopy samples were prepared on C-FLAT holey carbon grids with a Vitrobot (FEI Co.).'],\n",
       "    'count': 1,\n",
       "    'fnm': \"['M.', 'Y.', 'Y.', 'T.']\",\n",
       "    'index': 'b26',\n",
       "    'journal': 'Biol. Proced. Online',\n",
       "    'snm': \"['Ohi', 'Li', 'Cheng', 'Walz']\",\n",
       "    'text': ['Negative-stain samples were prepared as described<bibr rid=\"b26\"/> in 0.75% uranyl formate, and cryo-electron microscopy samples were prepared on C-FLAT holey carbon grids<bibr rid=\"b27\"/> with a <prod id=\"p1\" companyid=\"cop1\">Vitrobot</prod> (<company id=\"cop1\">FEI Co.</company>).'],\n",
       "    'year': '2004'},\n",
       "   {'ArticleID': 'pub.1016383987',\n",
       "    'CitedArticleID': 'pub.1016383987',\n",
       "    'Cites': 1493,\n",
       "    'CitingArticleID': 'pub.1030133211',\n",
       "    'DOI': '10.1038/34465',\n",
       "    'Issue': '6663',\n",
       "    'Journal': 'Nature',\n",
       "    'NatureID': '34465',\n",
       "    'PubYear': 1998,\n",
       "    'Title - matched_from': 'Structure of the &agr;&bgr; tubulin dimer by electron crystallography',\n",
       "    'Title - matched_to': 'Structure of the αβ tubulin dimer by electron crystallography',\n",
       "    'Volume': '391',\n",
       "    'clean_text': ['The microtubule lattice consists of lateral and longitudinal contacts; the stronger longitudinal contacts define microtubule polarity, with plus and minus ends (Supplementary Fig.'],\n",
       "    'count': 1,\n",
       "    'fnm': \"['E.', 'S. G.', 'K. H.']\",\n",
       "    'index': 'b9',\n",
       "    'journal': 'Nature',\n",
       "    'snm': \"['Nogales', 'Wolf', 'Downing']\",\n",
       "    'text': ['The microtubule lattice consists of lateral and longitudinal contacts<bibr rid=\"b9 b10\"/>; the stronger longitudinal contacts define microtubule polarity, with plus and minus ends (<sir rid=\"s1\">Supplementary Fig.'],\n",
       "    'year': '1998'},\n",
       "   {'ArticleID': 'pub.1046234181',\n",
       "    'CitedArticleID': 'pub.1046234181',\n",
       "    'Cites': 1091,\n",
       "    'CitingArticleID': 'pub.1030133211',\n",
       "    'DOI': '10.1016/s1047-8477(03)00069-8',\n",
       "    'Issue': '3',\n",
       "    'Journal': 'Journal of Structural Biology',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 2003,\n",
       "    'Title - matched_from': 'Accurate determination of local defocus and specimen tilt in electron microscopy',\n",
       "    'Title - matched_to': 'Accurate determination of local defocus and specimen tilt in electron microscopy',\n",
       "    'Volume': '142',\n",
       "    'clean_text': ['Defocus was determined with CTFFIND, and each micrograph was corrected by application of a Wiener filter.'],\n",
       "    'count': 1,\n",
       "    'fnm': \"['J. A.', 'N.']\",\n",
       "    'index': 'b28',\n",
       "    'journal': 'J. Struct. Biol.',\n",
       "    'snm': \"['Mindell', 'Grigorieff']\",\n",
       "    'text': ['Defocus was determined with CTFFIND<bibr rid=\"b28\"/>, and each micrograph was corrected by application of a Wiener filter.'],\n",
       "    'year': '2003'},\n",
       "   {'ArticleID': 'pub.1046438435',\n",
       "    'CitedArticleID': 'pub.1046438435',\n",
       "    'Cites': 231,\n",
       "    'CitingArticleID': 'pub.1030133211',\n",
       "    'DOI': '10.1083/jcb.200711053',\n",
       "    'Issue': '3',\n",
       "    'Journal': 'The Journal of Cell Biology',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 2008,\n",
       "    'Title - matched_from': 'Augmin: a protein complex required for centrosome-independent microtubule generation within the spindle',\n",
       "    'Title - matched_to': 'Augmin: a protein complex required for centrosome-independent microtubule generation within the spindle',\n",
       "    'Volume': '181',\n",
       "    'clean_text': ['Although not required as a scaffold, the &ggr;-TuRC-specific proteins may serve to stabilize the ring structure or fully activate nucleation activity, and they are essential for &ggr;-TuRC localization at non-centrosomal sites, as in augmin-dependent binding within the mitotic spindle (Fig.'],\n",
       "    'count': 1,\n",
       "    'fnm': \"['G.', 'M.', 'N.', 'N.', 'R. D.']\",\n",
       "    'index': 'b24',\n",
       "    'journal': 'J. Cell Biol.',\n",
       "    'snm': \"['Goshima', 'Mayer', 'Zhang', 'Stuurman', 'Vale']\",\n",
       "    'text': ['Although not required as a scaffold, the &ggr;-TuRC-specific proteins may serve to stabilize the ring structure or fully activate nucleation activity, and they are essential for &ggr;-TuRC localization at non-centrosomal sites, as in augmin-dependent binding within the mitotic spindle<bibr rid=\"b24\"/> (<figr rid=\"f5\">Fig.'],\n",
       "    'year': '2008'},\n",
       "   {'ArticleID': 'pub.1036914647',\n",
       "    'CitedArticleID': 'pub.1036914647',\n",
       "    'Cites': 171,\n",
       "    'CitingArticleID': 'pub.1030133211',\n",
       "    'DOI': '10.1038/35014051',\n",
       "    'Issue': '6',\n",
       "    'Journal': 'Nature Cell Biology',\n",
       "    'NatureID': '35014051',\n",
       "    'PubYear': 2000,\n",
       "    'Title - matched_from': 'A new function for the &ggr;-tubulin ring complex as a microtubule minus-end cap',\n",
       "    'Title - matched_to': 'A new function for the γ -tubulin ring complex as a microtubule minus-end cap',\n",
       "    'Volume': '2',\n",
       "    'clean_text': ['1c), in a similar manner to microtubules nucleated in vivo or from &ggr;-TuRCs in vitro.'],\n",
       "    'count': 1,\n",
       "    'fnm': \"['C.', 'Y.']\",\n",
       "    'index': 'b16',\n",
       "    'journal': 'Nature Cell Biol.',\n",
       "    'snm': \"['Wiese', 'Zheng']\",\n",
       "    'text': ['1c</figr>), in a similar manner to microtubules nucleated <i>in vivo</i><bibr rid=\"b14 b15\"/> or from &ggr;-TuRCs <i>in vitro</i><bibr rid=\"b3 b4 b16\"/>.'],\n",
       "    'year': '2000'},\n",
       "   {'ArticleID': 'pub.1008779181',\n",
       "    'CitedArticleID': 'pub.1008779181',\n",
       "    'Cites': 404,\n",
       "    'CitingArticleID': 'pub.1030133211',\n",
       "    'DOI': '10.1126/science.1141314',\n",
       "    'Issue': '5823',\n",
       "    'Journal': 'Science',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 2007,\n",
       "    'Title - matched_from': 'Genes required for mitotic spindle assembly in ',\n",
       "    'Title - matched_to': 'Genes Required for Mitotic Spindle Assembly in <i>Drosophila</i> S2 Cells',\n",
       "    'Volume': '316',\n",
       "    'clean_text': ['This is consistent with the observation that all of the &ggr;-TuRC-specific proteins can be depleted without affecting centrosomal microtubule nucleation of 13-protofilament microtubules.'],\n",
       "    'count': 1,\n",
       "    'fnm': \"['G.']\",\n",
       "    'index': 'b22',\n",
       "    'journal': 'Science',\n",
       "    'snm': \"['Goshima']\",\n",
       "    'text': ['This is consistent with the observation that all of the &ggr;-TuRC-specific proteins can be depleted without affecting centrosomal microtubule nucleation of 13-protofilament microtubules<bibr rid=\"b22 b23\"/>.'],\n",
       "    'year': '2007'},\n",
       "   {'ArticleID': 'pub.1003177362',\n",
       "    'CitedArticleID': 'pub.1003177362',\n",
       "    'Cites': 79,\n",
       "    'CitingArticleID': 'pub.1030133211',\n",
       "    'DOI': '10.1083/jcb.200511071',\n",
       "    'Issue': '4',\n",
       "    'Journal': 'The Journal of Cell Biology',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 2006,\n",
       "    'Title - matched_from': '<i>Drosophila melanogaster</i>',\n",
       "    'Title - matched_to': 'Drosophila melanogaster γ-TuRC is dispensable for targeting γ-tubulin to the centrosome and microtubule nucleation',\n",
       "    'Volume': '172',\n",
       "    'clean_text': ['This is consistent with the observation that all of the &ggr;-TuRC-specific proteins can be depleted without affecting centrosomal microtubule nucleation of 13-protofilament microtubules.'],\n",
       "    'count': 1,\n",
       "    'fnm': \"['C.']\",\n",
       "    'index': 'b23',\n",
       "    'journal': 'J. Cell Biol.',\n",
       "    'snm': \"['Verollet']\",\n",
       "    'text': ['This is consistent with the observation that all of the &ggr;-TuRC-specific proteins can be depleted without affecting centrosomal microtubule nucleation of 13-protofilament microtubules<bibr rid=\"b22 b23\"/>.'],\n",
       "    'year': '2006'},\n",
       "   {'ArticleID': 'pub.1007176321',\n",
       "    'CitedArticleID': 'pub.1007176321',\n",
       "    'Cites': 84,\n",
       "    'CitingArticleID': 'pub.1030133211',\n",
       "    'DOI': '10.1083/jcb.151.7.1513',\n",
       "    'Issue': '7',\n",
       "    'Journal': 'The Journal of Cell Biology',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 2000,\n",
       "    'Title - matched_from': 'Characterization and reconstitution of ',\n",
       "    'Title - matched_to': 'Characterization and Reconstitution of Drosophila γ-Tubulin Ring Complex Subunits',\n",
       "    'Volume': '151',\n",
       "    'clean_text': ['The masses of the domains determined from the cryo-electron microscopy map provide a rough estimate of their boundaries in each sequence, indicating the positions of the grip1 and grip2 motifs, conserved in all &ggr;-tubulin complex proteins (Supplementary Fig.'],\n",
       "    'count': 1,\n",
       "    'fnm': \"['R. N.']\",\n",
       "    'index': 'b19',\n",
       "    'journal': 'J. Cell Biol.',\n",
       "    'snm': \"['Gunawardane']\",\n",
       "    'text': ['The masses of the domains determined from the cryo-electron microscopy map provide a rough estimate of their boundaries in each sequence, indicating the positions of the grip1 and grip2 motifs, conserved in all &ggr;-tubulin complex proteins<bibr rid=\"b19\"/> (<sir rid=\"s1\">Supplementary Fig.'],\n",
       "    'year': '2000'},\n",
       "   {'ArticleID': 'pub.1019629707',\n",
       "    'CitedArticleID': 'pub.1019629707',\n",
       "    'Cites': 175,\n",
       "    'CitingArticleID': 'pub.1030133211',\n",
       "    'DOI': '10.1093/emboj/16.23.6985',\n",
       "    'Issue': '23',\n",
       "    'Journal': 'The EMBO Journal',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 1997,\n",
       "    'Title - matched_from': 'Spc98p and Spc97p of the yeast &ggr;-tubulin complex mediate binding to the spindle pole body via their interaction with Spc110p',\n",
       "    'Title - matched_to': 'Spc98p and Spc97p of the yeast γ‐tubulin complex mediate binding to the spindle pole body via their interaction with Spc110p',\n",
       "    'Volume': '16',\n",
       "    'clean_text': ['The 300-kDa &ggr;-tubulin small complex (&ggr;-TuSC), consisting of two molecules of &ggr;-tubulin and one copy each of the accessory proteins Spc97 and Spc98, is the conserved, essential core of the microtubule nucleating machinery.'],\n",
       "    'count': 1,\n",
       "    'fnm': \"['M.', 'E.']\",\n",
       "    'index': 'b1',\n",
       "    'journal': 'EMBO J.',\n",
       "    'snm': \"['Knop', 'Schiebel']\",\n",
       "    'text': ['The 300-kDa &ggr;-tubulin small complex (&ggr;-TuSC), consisting of two molecules of &ggr;-tubulin and one copy each of the accessory proteins Spc97 and Spc98, is the conserved, essential core of the microtubule nucleating machinery<bibr rid=\"b1 b2\"/>.'],\n",
       "    'year': '1997'},\n",
       "   {'ArticleID': 'pub.1022647309',\n",
       "    'CitedArticleID': 'pub.1022647309',\n",
       "    'Cites': 222,\n",
       "    'CitingArticleID': 'pub.1030133211',\n",
       "    'DOI': '10.1083/jcb.144.4.721',\n",
       "    'Issue': '4',\n",
       "    'Journal': 'The Journal of Cell Biology',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 1999,\n",
       "    'Title - matched_from': 'Characterization of two related ',\n",
       "    'Title - matched_to': 'Characterization of Two Related Drosophila γ-tubulin Complexes that Differ in Their Ability to Nucleate Microtubules',\n",
       "    'Volume': '144',\n",
       "    'clean_text': ['The 300-kDa &ggr;-tubulin small complex (&ggr;-TuSC), consisting of two molecules of &ggr;-tubulin and one copy each of the accessory proteins Spc97 and Spc98, is the conserved, essential core of the microtubule nucleating machinery.',\n",
       "     'The structure of &ggr;-TuRC indicated that it functions as a microtubule template.',\n",
       "     'The template model for microtubule nucleation predicts that &ggr;-tubulins interact laterally to form a ring that makes longitudinal contacts with &agr;/&bgr;-tubulin minus ends; alternative models that predict lateral interactions between &ggr;-tubulin and &agr;/&bgr;-tubulin have not been definitively ruled out.',\n",
       "     'Here we show that buffer conditions that promote microtubule growth (BRB80: low salt, pH\\u20096.9) also promote the spontaneous assembly of &ggr;-TuSCs into rings similar to Drosophila &ggr;-TuRCs (Fig.',\n",
       "     'The levels of nucleation observed are consistent with previous measurements for &ggr;-TuSC but are less robust than seen with &ggr;-TuRC, suggesting that assembly alone is insufficient to fully activate &ggr;-TuSC nucleating activity.',\n",
       "     'The levels of nucleation observed are consistent with previous measurements for &ggr;-TuSC but are less robust than seen with &ggr;-TuRC, suggesting that assembly alone is insufficient to fully activate &ggr;-TuSC nucleating activity.',\n",
       "     '&ggr;-TuRCs were purified from Drosophila embryos as described.'],\n",
       "    'count': 7,\n",
       "    'fnm': \"['K.']\",\n",
       "    'index': 'b2',\n",
       "    'journal': 'J. Cell Biol.',\n",
       "    'snm': \"['Oegema']\",\n",
       "    'text': ['The 300-kDa &ggr;-tubulin small complex (&ggr;-TuSC), consisting of two molecules of &ggr;-tubulin and one copy each of the accessory proteins Spc97 and Spc98, is the conserved, essential core of the microtubule nucleating machinery<bibr rid=\"b1 b2\"/>.',\n",
       "     'The structure of &ggr;-TuRC indicated that it functions as a microtubule template<bibr rid=\"b2 b3 b4 b5\"/>.',\n",
       "     'The template model for microtubule nucleation predicts that &ggr;-tubulins interact laterally to form a ring that makes longitudinal contacts with &agr;/&bgr;-tubulin<bibr rid=\"b2 b3 b4 b5\"/> minus ends; alternative models that predict lateral interactions between &ggr;-tubulin and &agr;/&bgr;-tubulin<bibr rid=\"b11\"/> have not been definitively ruled out.',\n",
       "     'Here we show that buffer conditions that promote microtubule growth (BRB80: low salt, pH&thinsp;6.9) also promote the spontaneous assembly of &ggr;-TuSCs into rings similar to <i>Drosophila</i> &ggr;-TuRCs<bibr rid=\"b2 b4\"/> (<figr rid=\"f1\">Fig.',\n",
       "     'The levels of nucleation observed are consistent with previous measurements for &ggr;-TuSC<bibr rid=\"b2 b21\"/> but are less robust than seen with &ggr;-TuRC<bibr rid=\"b2 b5\"/>, suggesting that assembly alone is insufficient to fully activate &ggr;-TuSC nucleating activity.',\n",
       "     'The levels of nucleation observed are consistent with previous measurements for &ggr;-TuSC<bibr rid=\"b2 b21\"/> but are less robust than seen with &ggr;-TuRC<bibr rid=\"b2 b5\"/>, suggesting that assembly alone is insufficient to fully activate &ggr;-TuSC nucleating activity.',\n",
       "     '&ggr;-TuRCs were purified from <i>Drosophila</i> embryos as described<bibr rid=\"b2\"/>.'],\n",
       "    'year': '1999'},\n",
       "   {'ArticleID': 'pub.1023089819',\n",
       "    'CitedArticleID': 'pub.1023089819',\n",
       "    'Cites': 189,\n",
       "    'CitingArticleID': 'pub.1030133211',\n",
       "    'DOI': '10.1016/0248-4900(91)90062-r',\n",
       "    'Issue': '1‐2',\n",
       "    'Journal': 'Biology of the Cell',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 1991,\n",
       "    'Title - matched_from': 'New data on the microtubule surface lattice',\n",
       "    'Title - matched_to': 'New data on the microtubule surface lattice',\n",
       "    'Volume': '71',\n",
       "    'clean_text': [],\n",
       "    'count': 0,\n",
       "    'fnm': \"['D.', 'R. H.']\",\n",
       "    'index': 'b6',\n",
       "    'journal': 'Biol. Cell',\n",
       "    'snm': \"['Chretien', 'Wade']\",\n",
       "    'text': '',\n",
       "    'year': '1991'},\n",
       "   {'ArticleID': 'pub.1043606668',\n",
       "    'CitedArticleID': 'pub.1043606668',\n",
       "    'Cites': 159,\n",
       "    'CitingArticleID': 'pub.1030133211',\n",
       "    'DOI': '10.1083/jcb.130.5.1149',\n",
       "    'Issue': '5',\n",
       "    'Journal': 'The Journal of Cell Biology',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 1995,\n",
       "    'Title - matched_from': 'Three-dimensional structural characterization of centrosomes from early ',\n",
       "    'Title - matched_to': 'Three-dimensional structural characterization of centrosomes from early Drosophila embryos.',\n",
       "    'Volume': '130',\n",
       "    'clean_text': ['1c), in a similar manner to microtubules nucleated in vivo or from &ggr;-TuRCs in vitro.'],\n",
       "    'count': 1,\n",
       "    'fnm': \"['M.']\",\n",
       "    'index': 'b15',\n",
       "    'journal': 'J. Cell Biol.',\n",
       "    'snm': \"['Moritz']\",\n",
       "    'text': ['1c</figr>), in a similar manner to microtubules nucleated <i>in vivo</i><bibr rid=\"b14 b15\"/> or from &ggr;-TuRCs <i>in vitro</i><bibr rid=\"b3 b4 b16\"/>.'],\n",
       "    'year': '1995'},\n",
       "   {'ArticleID': 'pub.1031690894',\n",
       "    'CitedArticleID': 'pub.1031690894',\n",
       "    'Cites': 182,\n",
       "    'CitingArticleID': 'pub.1030133211',\n",
       "    'DOI': '10.1016/j.jmb.2007.05.088',\n",
       "    'Issue': '3',\n",
       "    'Journal': 'Journal of Molecular Biology',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 2007,\n",
       "    'Title - matched_from': 'High-resolution electron microscopy of helical specimens: a fresh look at tobacco mosaic virus',\n",
       "    'Title - matched_to': 'High-resolution Electron Microscopy of Helical Specimens: A Fresh Look at Tobacco Mosaic Virus',\n",
       "    'Volume': '371',\n",
       "    'clean_text': ['The cryo-electron microscopic reconstruction was performed essentially as described.',\n",
       "     'Iterative helical real-space reconstruction was performed essentially as described, using SPIDER.'],\n",
       "    'count': 2,\n",
       "    'fnm': \"['C.']\",\n",
       "    'index': 'b25',\n",
       "    'journal': 'J. Mol. Biol.',\n",
       "    'snm': \"['Sachse']\",\n",
       "    'text': ['The cryo-electron microscopic reconstruction was performed essentially as described<bibr rid=\"b17 b25\"/>.',\n",
       "     '\\r\\n<p>Iterative helical real-space reconstruction was performed essentially as described<bibr rid=\"b17 b25\"/>, using SPIDER<bibr rid=\"b29\"/>.'],\n",
       "    'year': '2007'},\n",
       "   {'ArticleID': 'pub.1079779609',\n",
       "    'CitedArticleID': 'pub.1079779609',\n",
       "    'Cites': 87,\n",
       "    'CitingArticleID': 'pub.1030133211',\n",
       "    'DOI': '',\n",
       "    'Issue': '',\n",
       "    'Journal': 'Journal of Cell Science',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 1978,\n",
       "    'Title - matched_from': 'The role of spindle pole bodies and modified microtubule ends in the initiation of microtubule assembly in ',\n",
       "    'Title - matched_to': 'The role of spindle pole bodies and modified microtubule ends in the initiation of microtubule assembly in Saccharomyces cerevisiae.',\n",
       "    'Volume': '30',\n",
       "    'clean_text': ['1c), in a similar manner to microtubules nucleated in vivo or from &ggr;-TuRCs in vitro.'],\n",
       "    'count': 1,\n",
       "    'fnm': \"['B.', 'K.', 'L.']\",\n",
       "    'index': 'b14',\n",
       "    'journal': 'J. Cell Sci.',\n",
       "    'snm': \"['Byers', 'Shriver', 'Goetsch']\",\n",
       "    'text': ['1c</figr>), in a similar manner to microtubules nucleated <i>in vivo</i><bibr rid=\"b14 b15\"/> or from &ggr;-TuRCs <i>in vitro</i><bibr rid=\"b3 b4 b16\"/>.'],\n",
       "    'year': '1978'},\n",
       "   {'ArticleID': 'pub.1031126661',\n",
       "    'CitedArticleID': 'pub.1031126661',\n",
       "    'Cites': 793,\n",
       "    'CitingArticleID': 'pub.1030133211',\n",
       "    'DOI': '10.1016/s0092-8674(00)80961-7',\n",
       "    'Issue': '1',\n",
       "    'Journal': 'Cell',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 1999,\n",
       "    'Title - matched_from': 'High-resolution model of the microtubule',\n",
       "    'Title - matched_to': 'High-Resolution Model of the Microtubule',\n",
       "    'Volume': '96',\n",
       "    'clean_text': ['The microtubule lattice consists of lateral and longitudinal contacts; the stronger longitudinal contacts define microtubule polarity, with plus and minus ends (Supplementary Fig.'],\n",
       "    'count': 1,\n",
       "    'fnm': \"['E.', 'M.', 'R. A.', 'K. H.']\",\n",
       "    'index': 'b10',\n",
       "    'journal': 'Cell',\n",
       "    'snm': \"['Nogales', 'Whittaker', 'Milligan', 'Downing']\",\n",
       "    'text': ['The microtubule lattice consists of lateral and longitudinal contacts<bibr rid=\"b9 b10\"/>; the stronger longitudinal contacts define microtubule polarity, with plus and minus ends (<sir rid=\"s1\">Supplementary Fig.'],\n",
       "    'year': '1999'},\n",
       "   {'ArticleID': 'pub.1016142128',\n",
       "    'CitedArticleID': 'pub.1016142128',\n",
       "    'Cites': 61,\n",
       "    'CitingArticleID': 'pub.1030133211',\n",
       "    'DOI': '10.1091/mbc.02-01-0607',\n",
       "    'Issue': '4',\n",
       "    'Journal': 'Molecular Biology of the Cell',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 2002,\n",
       "    'Title - matched_from': 'Reconstitution and characterization of budding yeast &ggr;-tubulin complex',\n",
       "    'Title - matched_to': 'Reconstitution and Characterization of Budding Yeast γ-Tubulin Complex',\n",
       "    'Volume': '13',\n",
       "    'clean_text': ['The levels of nucleation observed are consistent with previous measurements for &ggr;-TuSC but are less robust than seen with &ggr;-TuRC, suggesting that assembly alone is insufficient to fully activate &ggr;-TuSC nucleating activity.',\n",
       "     '\\nMethods Summary\\n&ggr;-TuSC was co-expressed with glutathione S-transferase (GST)–Spc1101–220 as described, except that complexes were eluted by cleavage of the GST tag with tobacco etch virus (TEV) protease as the final purification step.',\n",
       "     'Nucleation assays were performed essentially as described.',\n",
       "     '\\n\\nSample preparation\\n&ggr;-TuSC was co-expressed with GST–Spc1101–220 in Sf9 cells and purified as described, in H100 buffer (40\\u2009mM Hepes pH\\u20097.6, 100\\u2009mM KCl, 1\\u2009mM EGTA, 1\\u2009mM MgCl2), except that complexes were eluted from glutathione resin by cleavage of the GST tag with TEV protease as the final purification step.',\n",
       "     '\\n'],\n",
       "    'count': 5,\n",
       "    'fnm': \"['D. B.', 'J. W.', 'W. O.', 'J.', 'T. N.']\",\n",
       "    'index': 'b21',\n",
       "    'journal': 'Mol. Biol. Cell',\n",
       "    'snm': \"['Vinh', 'Kern', 'Hancock', 'Howard', 'Davis']\",\n",
       "    'text': ['The levels of nucleation observed are consistent with previous measurements for &ggr;-TuSC<bibr rid=\"b2 b21\"/> but are less robust than seen with &ggr;-TuRC<bibr rid=\"b2 b5\"/>, suggesting that assembly alone is insufficient to fully activate &ggr;-TuSC nucleating activity.',\n",
       "     '\\r\\n<sec level=\"1\">\\r\\n<sectitle>Methods Summary</sectitle>\\r\\n<p>&ggr;-TuSC was co-expressed with glutathione <i>S</i>-transferase (GST)&ndash;Spc110<super>1&ndash;220</super> as described<bibr rid=\"b13 b21\"/>, except that complexes were eluted by cleavage of the GST tag with tobacco etch virus (TEV) protease as the final purification step.',\n",
       "     'Nucleation assays were performed essentially as described<bibr rid=\"b21\"/>.',\n",
       "     '\\r\\n<online-methods>\\r\\n<sec level=\"2\">\\r\\n<sectitle>Sample preparation</sectitle>\\r\\n<p>&ggr;-TuSC was co-expressed with GST&ndash;Spc110<super>1&ndash;220</super> in Sf9 cells and purified as described, in H100 buffer (40&thinsp;mM Hepes pH&thinsp;7.6, 100&thinsp;mM KCl, 1&thinsp;mM EGTA, 1&thinsp;mM MgCl<sub>2</sub>)<bibr rid=\"b13 b21\"/>, except that complexes were eluted from glutathione resin by cleavage of the GST tag with TEV protease as the final purification step.',\n",
       "     '\\r\\n</sec>\\r\\n<sec level=\"2\">\\r\\n<sectitle>Microtubule nucleation assays</sectitle>\\r\\n<p>The solution nucleation assays were performed essentially as described<bibr rid=\"b21\"/>.'],\n",
       "    'year': '2002'},\n",
       "   {'ArticleID': 'pub.1008225564',\n",
       "    'CitedArticleID': 'pub.1008225564',\n",
       "    'Cites': 16323,\n",
       "    'CitingArticleID': 'pub.1030133211',\n",
       "    'DOI': '10.1002/jcc.20084',\n",
       "    'Issue': '13',\n",
       "    'Journal': 'Journal of Computational Chemistry',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 2004,\n",
       "    'Title - matched_from': 'UCSF Chimera—a visualization system for exploratory research and analysis',\n",
       "    'Title - matched_to': 'UCSF Chimera—A visualization system for exploratory research and analysis',\n",
       "    'Volume': '25',\n",
       "    'clean_text': ['Volumes were viewed and segmented with Chimera.'],\n",
       "    'count': 1,\n",
       "    'fnm': \"['E. F.']\",\n",
       "    'index': 'b30',\n",
       "    'journal': 'J. Comput. Chem.',\n",
       "    'snm': \"['Pettersen']\",\n",
       "    'text': ['Volumes were viewed and segmented with Chimera<bibr rid=\"b30\"/>.'],\n",
       "    'year': '2004'},\n",
       "   {'ArticleID': 'pub.1008960932',\n",
       "    'CitedArticleID': 'pub.1008960932',\n",
       "    'Cites': 195,\n",
       "    'CitingArticleID': 'pub.1030133211',\n",
       "    'DOI': '10.1038/35014058',\n",
       "    'Issue': '6',\n",
       "    'Journal': 'Nature Cell Biology',\n",
       "    'NatureID': '35014058',\n",
       "    'PubYear': 2000,\n",
       "    'Title - matched_from': 'Structure of the &ggr;-tubulin ring complex: a template for microtubule nucleation',\n",
       "    'Title - matched_to': 'Structure of the γ-tubulin ring complex: a template for microtubule nucleation',\n",
       "    'Volume': '2',\n",
       "    'clean_text': ['The structure of &ggr;-TuRC indicated that it functions as a microtubule template.',\n",
       "     'The template model for microtubule nucleation predicts that &ggr;-tubulins interact laterally to form a ring that makes longitudinal contacts with &agr;/&bgr;-tubulin minus ends; alternative models that predict lateral interactions between &ggr;-tubulin and &agr;/&bgr;-tubulin have not been definitively ruled out.',\n",
       "     'Here we show that buffer conditions that promote microtubule growth (BRB80: low salt, pH\\u20096.9) also promote the spontaneous assembly of &ggr;-TuSCs into rings similar to Drosophila &ggr;-TuRCs (Fig.',\n",
       "     '1c), in a similar manner to microtubules nucleated in vivo or from &ggr;-TuRCs in vitro.',\n",
       "     'Each turn of helix forms a lock-washer shape similar to that of &ggr;-TuRC (Fig.'],\n",
       "    'count': 5,\n",
       "    'fnm': \"['M.', 'M. B.', 'V.', 'J.', 'D. A.']\",\n",
       "    'index': 'b4',\n",
       "    'journal': 'Nature Cell Biol.',\n",
       "    'snm': \"['Moritz', 'Braunfeld', 'Guenebaut', 'Heuser', 'Agard']\",\n",
       "    'text': ['The structure of &ggr;-TuRC indicated that it functions as a microtubule template<bibr rid=\"b2 b3 b4 b5\"/>.',\n",
       "     'The template model for microtubule nucleation predicts that &ggr;-tubulins interact laterally to form a ring that makes longitudinal contacts with &agr;/&bgr;-tubulin<bibr rid=\"b2 b3 b4 b5\"/> minus ends; alternative models that predict lateral interactions between &ggr;-tubulin and &agr;/&bgr;-tubulin<bibr rid=\"b11\"/> have not been definitively ruled out.',\n",
       "     'Here we show that buffer conditions that promote microtubule growth (BRB80: low salt, pH&thinsp;6.9) also promote the spontaneous assembly of &ggr;-TuSCs into rings similar to <i>Drosophila</i> &ggr;-TuRCs<bibr rid=\"b2 b4\"/> (<figr rid=\"f1\">Fig.',\n",
       "     '1c</figr>), in a similar manner to microtubules nucleated <i>in vivo</i><bibr rid=\"b14 b15\"/> or from &ggr;-TuRCs <i>in vitro</i><bibr rid=\"b3 b4 b16\"/>.',\n",
       "     'Each turn of helix forms a lock-washer shape similar to that of &ggr;-TuRC<bibr rid=\"b4\"/> (<figr rid=\"f2\">Fig.'],\n",
       "    'year': '2000'},\n",
       "   {'ArticleID': 'pub.1045086717',\n",
       "    'CitedArticleID': 'pub.1045086717',\n",
       "    'Cites': 419,\n",
       "    'CitingArticleID': 'pub.1030133211',\n",
       "    'DOI': '10.1016/0092-8674(90)90693-9',\n",
       "    'Issue': '7',\n",
       "    'Journal': 'Cell',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 1990,\n",
       "    'Title - matched_from': '&ggr;-Tubulin is a component of the spindle pole body that is essential for microtubule function in Aspergillus nidulans',\n",
       "    'Title - matched_to': 'γ-tubulin is a component of the spindle pole body that is essential for microtubule function in Aspergillus nidulans',\n",
       "    'Volume': '61',\n",
       "    'clean_text': ['The sequence and structural similarity between &ggr;-tubulin and &agr;/&bgr;-tubulin indicated that nucleation results from microtubule-like contacts between &ggr;-tubulin and &agr;/&bgr;-tubulin.'],\n",
       "    'count': 1,\n",
       "    'fnm': \"['B. R.', 'C. E.', 'Y.', 'M. K.']\",\n",
       "    'index': 'b8',\n",
       "    'journal': 'Cell',\n",
       "    'snm': \"['Oakley', 'Oakley', 'Yoon', 'Jung']\",\n",
       "    'text': ['\\r\\n<p>The sequence and structural similarity between &ggr;-tubulin and &agr;/&bgr;-tubulin indicated that nucleation results from microtubule-like contacts between &ggr;-tubulin and &agr;/&bgr;-tubulin<bibr rid=\"b8\"/>.'],\n",
       "    'year': '1990'},\n",
       "   {'ArticleID': 'pub.1015930233',\n",
       "    'CitedArticleID': 'pub.1015930233',\n",
       "    'Cites': 39,\n",
       "    'CitingArticleID': 'pub.1030133211',\n",
       "    'DOI': '10.1017/s1431927607070791',\n",
       "    'Issue': '5',\n",
       "    'Journal': 'Microscopy and Microanalysis',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 2007,\n",
       "    'Title - matched_from': 'An improved holey carbon film for cryo-electron microscopy',\n",
       "    'Title - matched_to': 'An Improved Holey Carbon Film for Cryo-Electron Microscopy',\n",
       "    'Volume': '13',\n",
       "    'clean_text': ['Negative-stain samples were prepared as described in 0.75% uranyl formate, and cryo-electron microscopy samples were prepared on C-FLAT holey carbon grids with a Vitrobot (FEI Co.).'],\n",
       "    'count': 1,\n",
       "    'fnm': \"['J.']\",\n",
       "    'index': 'b27',\n",
       "    'journal': 'Microsc. Microanal.',\n",
       "    'snm': \"['Quispe']\",\n",
       "    'text': ['Negative-stain samples were prepared as described<bibr rid=\"b26\"/> in 0.75% uranyl formate, and cryo-electron microscopy samples were prepared on C-FLAT holey carbon grids<bibr rid=\"b27\"/> with a <prod id=\"p1\" companyid=\"cop1\">Vitrobot</prod> (<company id=\"cop1\">FEI Co.</company>).'],\n",
       "    'year': '2007'},\n",
       "   {'ArticleID': 'pub.1049175268',\n",
       "    'CitedArticleID': 'pub.1049175268',\n",
       "    'Cites': 598,\n",
       "    'CitingArticleID': 'pub.1030133211',\n",
       "    'DOI': '10.1038/378578a0',\n",
       "    'Issue': '6557',\n",
       "    'Journal': 'Nature',\n",
       "    'NatureID': '378578a0',\n",
       "    'PubYear': 1995,\n",
       "    'Title - matched_from': 'Nucleation of microtubule assembly by a &ggr;-tubulin-containing ring complex',\n",
       "    'Title - matched_to': 'Nucleation of microtubule assembly by a γ-tubulin-containing ring complex',\n",
       "    'Volume': '378',\n",
       "    'clean_text': ['The structure of &ggr;-TuRC indicated that it functions as a microtubule template.',\n",
       "     'The template model for microtubule nucleation predicts that &ggr;-tubulins interact laterally to form a ring that makes longitudinal contacts with &agr;/&bgr;-tubulin minus ends; alternative models that predict lateral interactions between &ggr;-tubulin and &agr;/&bgr;-tubulin have not been definitively ruled out.',\n",
       "     'The levels of nucleation observed are consistent with previous measurements for &ggr;-TuSC but are less robust than seen with &ggr;-TuRC, suggesting that assembly alone is insufficient to fully activate &ggr;-TuSC nucleating activity.'],\n",
       "    'count': 3,\n",
       "    'fnm': \"['Y.', 'M. L.', 'B.', 'T.']\",\n",
       "    'index': 'b5',\n",
       "    'journal': 'Nature',\n",
       "    'snm': \"['Zheng', 'Wong', 'Alberts', 'Mitchison']\",\n",
       "    'text': ['The structure of &ggr;-TuRC indicated that it functions as a microtubule template<bibr rid=\"b2 b3 b4 b5\"/>.',\n",
       "     'The template model for microtubule nucleation predicts that &ggr;-tubulins interact laterally to form a ring that makes longitudinal contacts with &agr;/&bgr;-tubulin<bibr rid=\"b2 b3 b4 b5\"/> minus ends; alternative models that predict lateral interactions between &ggr;-tubulin and &agr;/&bgr;-tubulin<bibr rid=\"b11\"/> have not been definitively ruled out.',\n",
       "     'The levels of nucleation observed are consistent with previous measurements for &ggr;-TuSC<bibr rid=\"b2 b21\"/> but are less robust than seen with &ggr;-TuRC<bibr rid=\"b2 b5\"/>, suggesting that assembly alone is insufficient to fully activate &ggr;-TuSC nucleating activity.'],\n",
       "    'year': '1995'},\n",
       "   {'ArticleID': 'pub.1039952769',\n",
       "    'CitedArticleID': 'pub.1039952769',\n",
       "    'Cites': 138,\n",
       "    'CitingArticleID': 'pub.1030133211',\n",
       "    'DOI': '10.1073/pnas.0801155105',\n",
       "    'Issue': '14',\n",
       "    'Journal': 'Proceedings of the National Academy of Sciences',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 2008,\n",
       "    'Title - matched_from': 'The lattice as allosteric effector: structural studies of &agr;&bgr;- and &ggr;-tubulin clarify the role of GTP in microtubule assembly',\n",
       "    'Title - matched_to': 'The lattice as allosteric effector: Structural studies of αβ- and γ-tubulin clarify the role of GTP in microtubule assembly',\n",
       "    'Volume': '105',\n",
       "    'clean_text': ['The human &ggr;-tubulin crystal structure was fitted into the density in the regions previously assigned to &ggr;-tubulin (Fig.'],\n",
       "    'count': 1,\n",
       "    'fnm': \"['L. M.', 'E. A.', 'D. A.']\",\n",
       "    'index': 'b20',\n",
       "    'journal': 'Proc. Natl Acad. Sci. USA',\n",
       "    'snm': \"['Rice', 'Montabana', 'Agard']\",\n",
       "    'text': ['\\r\\n<p>The human &ggr;-tubulin crystal structure<bibr rid=\"b20\"/> was fitted into the density in the regions previously assigned to &ggr;-tubulin<bibr rid=\"b13\"/> (<figr rid=\"f3\">Fig.'],\n",
       "    'year': '2008'},\n",
       "   {'ArticleID': 'pub.1047520979',\n",
       "    'CitedArticleID': 'pub.1047520979',\n",
       "    'Cites': 65,\n",
       "    'CitingArticleID': 'pub.1030133211',\n",
       "    'DOI': '10.1091/mbc.e07-09-0879',\n",
       "    'Issue': '1',\n",
       "    'Journal': 'Molecular Biology of the Cell',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 2008,\n",
       "    'Title - matched_from': 'The structure of the &ggr;-tubulin small complex: implications of its architecture and flexibility for microtubule nucleation',\n",
       "    'Title - matched_to': 'The Structure of the γ-Tubulin Small Complex: Implications of Its Architecture and Flexibility for Microtubule Nucleation',\n",
       "    'Volume': '19',\n",
       "    'clean_text': ['Our previous 25-Å structure of Saccharomyces &ggr;-TuSC was determined in buffer conditions that yielded predominantly monomeric complexes.',\n",
       "     '&ggr;-TuSC in the filament is remarkably similar to free &ggr;-TuSC (Fig.',\n",
       "     'The central domain of Spc98 is kinked, at a position previously shown to be the site of limited hinge-like flexibility.',\n",
       "     'The human &ggr;-tubulin crystal structure was fitted into the density in the regions previously assigned to &ggr;-tubulin (Fig.',\n",
       "     '3b, c), as observed in the free &ggr;-TuSC structure.',\n",
       "     'We previously predicted that bending at the flexible kink in Spc98 is required to bring the intra-&ggr;-TuSC &ggr;-tubulins to the microtubule spacing.',\n",
       "     '\\nMethods Summary\\n&ggr;-TuSC was co-expressed with glutathione S-transferase (GST)–Spc1101–220 as described, except that complexes were eluted by cleavage of the GST tag with tobacco etch virus (TEV) protease as the final purification step.',\n",
       "     '\\n\\nSample preparation\\n&ggr;-TuSC was co-expressed with GST–Spc1101–220 in Sf9 cells and purified as described, in H100 buffer (40\\u2009mM Hepes pH\\u20097.6, 100\\u2009mM KCl, 1\\u2009mM EGTA, 1\\u2009mM MgCl2), except that complexes were eluted from glutathione resin by cleavage of the GST tag with TEV protease as the final purification step.'],\n",
       "    'count': 8,\n",
       "    'fnm': \"['J. M.']\",\n",
       "    'index': 'b13',\n",
       "    'journal': 'Mol. Biol. Cell',\n",
       "    'snm': \"['Kollman']\",\n",
       "    'text': ['\\r\\n<p>Our previous 25-&Aring; structure of <i>Saccharomyces</i> &ggr;-TuSC was determined in buffer conditions that yielded predominantly monomeric complexes<bibr rid=\"b12 b13\"/>.',\n",
       "     '\\r\\n<p>&ggr;-TuSC in the filament is remarkably similar to free &ggr;-TuSC<bibr rid=\"b13 b18\"/> (<figr rid=\"f2\">Fig.',\n",
       "     'The central domain of Spc98 is kinked, at a position previously shown to be the site of limited hinge-like flexibility<bibr rid=\"b13\"/>.',\n",
       "     '\\r\\n<p>The human &ggr;-tubulin crystal structure<bibr rid=\"b20\"/> was fitted into the density in the regions previously assigned to &ggr;-tubulin<bibr rid=\"b13\"/> (<figr rid=\"f3\">Fig.',\n",
       "     '3b, c</figr>), as observed in the free &ggr;-TuSC structure<bibr rid=\"b13\"/>.',\n",
       "     'We previously predicted that bending at the flexible kink in Spc98 is required to bring the intra-&ggr;-TuSC &ggr;-tubulins to the microtubule spacing<bibr rid=\"b13 b18\"/>.',\n",
       "     '\\r\\n<sec level=\"1\">\\r\\n<sectitle>Methods Summary</sectitle>\\r\\n<p>&ggr;-TuSC was co-expressed with glutathione <i>S</i>-transferase (GST)&ndash;Spc110<super>1&ndash;220</super> as described<bibr rid=\"b13 b21\"/>, except that complexes were eluted by cleavage of the GST tag with tobacco etch virus (TEV) protease as the final purification step.',\n",
       "     '\\r\\n<online-methods>\\r\\n<sec level=\"2\">\\r\\n<sectitle>Sample preparation</sectitle>\\r\\n<p>&ggr;-TuSC was co-expressed with GST&ndash;Spc110<super>1&ndash;220</super> in Sf9 cells and purified as described, in H100 buffer (40&thinsp;mM Hepes pH&thinsp;7.6, 100&thinsp;mM KCl, 1&thinsp;mM EGTA, 1&thinsp;mM MgCl<sub>2</sub>)<bibr rid=\"b13 b21\"/>, except that complexes were eluted from glutathione resin by cleavage of the GST tag with TEV protease as the final purification step.'],\n",
       "    'year': '2008'},\n",
       "   {'ArticleID': 'pub.1024098120',\n",
       "    'CitedArticleID': 'pub.1024098120',\n",
       "    'Cites': 82,\n",
       "    'CitingArticleID': 'pub.1030133211',\n",
       "    'DOI': '10.1038/35014045',\n",
       "    'Issue': '6',\n",
       "    'Journal': 'Nature Cell Biology',\n",
       "    'NatureID': '35014045',\n",
       "    'PubYear': 2000,\n",
       "    'Title - matched_from': 'Immunostructural evidence for the template mechanism of microtubule nucleation',\n",
       "    'Title - matched_to': 'Immunostructural evidence for the template mechanism of microtubule nucleation',\n",
       "    'Volume': '2',\n",
       "    'clean_text': ['The structure of &ggr;-TuRC indicated that it functions as a microtubule template.',\n",
       "     'The template model for microtubule nucleation predicts that &ggr;-tubulins interact laterally to form a ring that makes longitudinal contacts with &agr;/&bgr;-tubulin minus ends; alternative models that predict lateral interactions between &ggr;-tubulin and &agr;/&bgr;-tubulin have not been definitively ruled out.',\n",
       "     '1c), in a similar manner to microtubules nucleated in vivo or from &ggr;-TuRCs in vitro.',\n",
       "     'Microtubules were grown in the presence of &ggr;-TuSC rings and prepared for electron microscopy as described, with the exception that microtubules were crosslinked for 3\\u2009min in 1% glutaraldehyde before sedimentation.'],\n",
       "    'count': 4,\n",
       "    'fnm': \"['T. J.', 'G. G.']\",\n",
       "    'index': 'b3',\n",
       "    'journal': 'Nature Cell Biol.',\n",
       "    'snm': \"['Keating', 'Borisy']\",\n",
       "    'text': ['The structure of &ggr;-TuRC indicated that it functions as a microtubule template<bibr rid=\"b2 b3 b4 b5\"/>.',\n",
       "     'The template model for microtubule nucleation predicts that &ggr;-tubulins interact laterally to form a ring that makes longitudinal contacts with &agr;/&bgr;-tubulin<bibr rid=\"b2 b3 b4 b5\"/> minus ends; alternative models that predict lateral interactions between &ggr;-tubulin and &agr;/&bgr;-tubulin<bibr rid=\"b11\"/> have not been definitively ruled out.',\n",
       "     '1c</figr>), in a similar manner to microtubules nucleated <i>in vivo</i><bibr rid=\"b14 b15\"/> or from &ggr;-TuRCs <i>in vitro</i><bibr rid=\"b3 b4 b16\"/>.',\n",
       "     'Microtubules were grown in the presence of &ggr;-TuSC rings and prepared for electron microscopy as described<bibr rid=\"b3\"/>, with the exception that microtubules were crosslinked for 3&thinsp;min in 1% glutaraldehyde before sedimentation.'],\n",
       "    'year': '2000'},\n",
       "   {'ArticleID': 'pub.1007617772',\n",
       "    'CitedArticleID': 'pub.1007617772',\n",
       "    'Cites': 351,\n",
       "    'CitingArticleID': 'pub.1030133211',\n",
       "    'DOI': '10.1016/s0304-3991(00)00062-0',\n",
       "    'Issue': '4',\n",
       "    'Journal': 'Ultramicroscopy',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 2000,\n",
       "    'Title - matched_from': 'A robust algorithm for the reconstruction of helical filaments using single-particle methods',\n",
       "    'Title - matched_to': 'A robust algorithm for the reconstruction of helical filaments using single-particle methods',\n",
       "    'Volume': '85',\n",
       "    'clean_text': ['2a), using iterative helical real-space reconstruction, a single-particle approach to helical structure determination.',\n",
       "     'The cryo-electron microscopic reconstruction was performed essentially as described.',\n",
       "     'Iterative helical real-space reconstruction was performed essentially as described, using SPIDER.',\n",
       "     'Initially, 18 rounds of projection matching were performed with 2× binned particles, with the helical symmetry parameters refined and imposed at each step with the programs hsearch_lorentz and himpose.'],\n",
       "    'count': 4,\n",
       "    'fnm': \"['E. H.']\",\n",
       "    'index': 'b17',\n",
       "    'journal': 'Ultramicroscopy',\n",
       "    'snm': \"['Egelman']\",\n",
       "    'text': ['2a</figr>), using iterative helical real-space reconstruction, a single-particle approach to helical structure determination<bibr rid=\"b17\"/>.',\n",
       "     'The cryo-electron microscopic reconstruction was performed essentially as described<bibr rid=\"b17 b25\"/>.',\n",
       "     '\\r\\n<p>Iterative helical real-space reconstruction was performed essentially as described<bibr rid=\"b17 b25\"/>, using SPIDER<bibr rid=\"b29\"/>.',\n",
       "     'Initially, 18 rounds of projection matching were performed with 2&times; binned particles, with the helical symmetry parameters refined and imposed at each step with the programs hsearch_lorentz and himpose<bibr rid=\"b17\"/>.'],\n",
       "    'year': '2000'}]}}"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unit_paper_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Getting original text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_text_mapping(filename, content):\n",
    "    \n",
    "    \"\"\"\n",
    "    Given XML filename and XML file, extract rid mappings and attribute data\n",
    "    \"\"\"\n",
    "    \n",
    "    mappings  = {}\n",
    "    extracted = {}\n",
    "    references = {}\n",
    "    parsed    = xmltodict.parse(content.decode('UTF-8'))\n",
    "    soup      = BeautifulSoup(content)\n",
    "\n",
    "    if '@id' in parsed['article']:\n",
    "        extracted['id']       = str(parsed['article']['@id'])\n",
    "    if '@language' in parsed['article']:\n",
    "        extracted['language'] = str(parsed['article']['@language'])\n",
    "    if '@publish' in parsed['article']:\n",
    "        extracted['publish']  = str(parsed['article']['@publish'])\n",
    "    if '@relation' in parsed['article']:\n",
    "        extracted['relation'] = str(parsed['article']['@relation'])\n",
    "    \n",
    "    if ('article' in parsed):\n",
    "        if ('pubfm' in parsed['article']):\n",
    "            if ('jtl' in parsed['article']['pubfm']):\n",
    "                extracted['jtl']   = str(parsed['article']['pubfm']['jtl'])\n",
    "    \n",
    "    if ('article' in parsed):\n",
    "        if ('pubfm' in parsed['article']):\n",
    "            if ('vol' in parsed['article']['pubfm']):\n",
    "                extracted['vol']   = str(parsed['article']['pubfm']['vol'])\n",
    "    \n",
    "    if ('article' in parsed):\n",
    "        if ('pubfm' in parsed['article']):\n",
    "            if ('issue' in parsed['article']['pubfm']):\n",
    "                extracted['issue'] = str(parsed['article']['pubfm']['issue'])\n",
    "\n",
    "    if ('article' in parsed):\n",
    "        if ('pubfm' in parsed['article']):\n",
    "            if ('vol' in parsed['article']['pubfm']):\n",
    "                extracted['doi']   = str(parsed['article']['pubfm']['doi'])\n",
    "    \n",
    "    if ('article' in parsed):\n",
    "        if ('fm' in parsed['article']):\n",
    "            if ('atl' in parsed['article']['fm']):\n",
    "                extracted['title']   = str(parsed['article']['fm']['atl'])\n",
    "     \n",
    "    \n",
    "    del parsed\n",
    "    \n",
    "    #print(len(soup.find_all(\"bdy\")))\n",
    "    \n",
    "    text = ''\n",
    "    \n",
    "    for bib in soup.find_all(\"bdy\"):\n",
    "        \n",
    "        #print(bib.text)\n",
    "        \n",
    "        text += bib.text\n",
    "            \n",
    "    \n",
    "    references = {'metadata': papers[extracted['doi']], 'text': text}\n",
    "    \n",
    "    try:\n",
    "        with open('data/intermediate/full_text_mappings/%s' % (filename + '.json'), 'w') as f:\n",
    "            f.write(json.dumps(references))\n",
    "        return True\n",
    "    \n",
    "    except TypeError as e:\n",
    "        #print('Some contents of the file %s is not serializable' % filename)\n",
    "        raise e\n",
    "        \n",
    "        \n",
    "def get_zips_parallel_reducer(arg):\n",
    "    \n",
    "    try:\n",
    "        finfo, content = arg[0], zlib.decompress(arg[1])\n",
    "        del arg\n",
    "\n",
    "        get_text_mapping(finfo.filename, content)\n",
    "        \n",
    "        return True\n",
    "    \n",
    "    except Exception as e:\n",
    "        \n",
    "        print('Could not extract references from %s' % finfo)\n",
    "        return str(Exception)\n",
    "\n",
    "    \n",
    "def get_zips_parallel_mapper(directory):\n",
    "    \n",
    "    \"\"\"\n",
    "    Function to extract relevant files from the filebase\n",
    "    \"\"\"\n",
    "    \n",
    "    try:\n",
    "        \n",
    "        zip_list  = [i for i in os.listdir(directory) if 'supp_xml' not in i]\n",
    "        errored   = []\n",
    "        finfos    = []\n",
    "        contents  = []\n",
    "\n",
    "        for zfile in tqdm(zip_list):\n",
    "\n",
    "            zfile     = zipfile.ZipFile(os.path.join(directory, zfile))\n",
    "\n",
    "            for finfo in zfile.infolist():\n",
    "\n",
    "                if 'nature' in finfo.filename:\n",
    "                    \n",
    "                    ifile = zfile.open(finfo)\n",
    "                    content = ifile.read()\n",
    "\n",
    "                    finfos.append(finfo)\n",
    "                    contents.append(zlib.compress(content))\n",
    "\n",
    "        return finfos, contents\n",
    "    \n",
    "    except Exception as e:\n",
    "        \n",
    "        print('Could not read file from zip %s, file %s' % (zfile, finfo.name))\n",
    "              "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DEBUG = False\n",
    "directory     = 'data/nature/raw xml/'\n",
    "#finfos, contents = get_zips_parallel_mapper(directory)\n",
    "get_zips_parallel_reducer((finfos[2], contents[2]))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'metadata': {'doi': '10.1038/nature04206',\n",
       "  'id': 'nature04206',\n",
       "  'jtl': 'Nature',\n",
       "  'language': 'eng',\n",
       "  'publish': 'issue',\n",
       "  'relation': 'no',\n",
       "  'title': 'Observation of spin Coulomb drag in a two-dimensional electron gas',\n",
       "  'vol': '437'},\n",
       " 'references': [{'ArticleID': 'pub.1005015903',\n",
       "   'CitedArticleID': 'pub.1005015903',\n",
       "   'Cites': 889,\n",
       "   'CitingArticleID': 'pub.1023679984',\n",
       "   'DOI': '10.1007/978-3-662-05003-3',\n",
       "   'Issue': '',\n",
       "   'Journal': '',\n",
       "   'NatureID': '',\n",
       "   'PubYear': 2002,\n",
       "   'Title - matched_from': 'Experimental observation of the spin Hall effect in a two-dimensional spin-orbit coupled semiconductor system',\n",
       "   'Title - matched_to': 'Semiconductor Spintronics and Quantum Computation',\n",
       "   'Volume': '',\n",
       "   'clean_text': ['(The recently observed spin Hall effect also involves the counterflow of spin populations, and so should be damped by SCD.)'],\n",
       "   'count': 1,\n",
       "   'fnm': \"['J.', 'B.', 'J.', 'T.']\",\n",
       "   'index': 'b12',\n",
       "   'journal': 'Phys. Rev. Lett.',\n",
       "   'snm': \"['Wunderlich', 'Kaestner', 'Sinova', 'Jungwirth']\",\n",
       "   'text': ['(The recently observed<bibr rid=\"b11 b12\"/> spin Hall effect also involves the counterflow of spin populations, and so should be damped by SCD.)'],\n",
       "   'year': '2005'},\n",
       "  {'ArticleID': 'pub.1060552109',\n",
       "   'CitedArticleID': 'pub.1060552109',\n",
       "   'Cites': 38,\n",
       "   'CitingArticleID': 'pub.1023679984',\n",
       "   'DOI': '10.1103/physrevb.40.5432',\n",
       "   'Issue': '8',\n",
       "   'Journal': 'Physical Review B',\n",
       "   'NatureID': '',\n",
       "   'PubYear': 1989,\n",
       "   'Title - matched_from': 'Spin susceptibility in a two-dimensional electron gas',\n",
       "   'Title - matched_to': 'Spin susceptibility in a two-dimensional electron gas',\n",
       "   'Volume': '40',\n",
       "   'clean_text': ['The ratio is far greater than can be accounted for by many-body enhancement of the spin susceptibility, as the factor χs/χ0 is less than 1.4 in this range of electron density.',\n",
       "    'The fact that the slope is slightly greater than unity is consistent with the expectation that the many-body enhancement of χs relative to χ0 is small in this density regime.'],\n",
       "   'count': 2,\n",
       "   'fnm': \"['S.', 'G. F.']\",\n",
       "   'index': 'b8',\n",
       "   'journal': 'Phys. Rev. B',\n",
       "   'snm': \"['Yarlagadda', 'Giuliani']\",\n",
       "   'text': ['The ratio is far greater than can be accounted for by many-body enhancement of the spin susceptibility, as the factor <i>&chi;</i><sub>s</sub>/<i>&chi;</i><sub>0</sub> is less than 1.4 in this range of electron density<bibr rid=\"b8 b9\"/>.',\n",
       "    'The fact that the slope is slightly greater than unity is consistent with the expectation that the many-body enhancement of <i>&chi;</i><sub>s</sub> relative to <i>&chi;</i><sub>0</sub> is small in this density regime<bibr rid=\"b8 b9\"/>.'],\n",
       "   'year': '1989'},\n",
       "  {'ArticleID': 'pub.1043133383',\n",
       "   'CitedArticleID': 'pub.1043133383',\n",
       "   'Cites': 67,\n",
       "   'CitingArticleID': 'pub.1023679984',\n",
       "   'DOI': '10.1103/physrevb.64.245308',\n",
       "   'Issue': '24',\n",
       "   'Journal': 'Physical Review B',\n",
       "   'NatureID': '',\n",
       "   'PubYear': 2001,\n",
       "   'Title - matched_from': 'Diffusion equation and spin drag in spin-polarized transport',\n",
       "   'Title - matched_to': 'Diffusion equation and spin drag in spin-polarized transport',\n",
       "   'Volume': '64',\n",
       "   'clean_text': [\"D'Amico and Vignale, and Flensberg and Jensen, have calculated ρ↑↓(T) for a 2DEG using the random phase approximation, obtaining results that depend only on the electron density of the quantum well.\"],\n",
       "   'count': 1,\n",
       "   'fnm': \"['K.', 'T. S.', 'N. A.']\",\n",
       "   'index': 'b13',\n",
       "   'journal': 'Phys. Rev. B',\n",
       "   'snm': \"['Flensberg', 'Jensen', 'Mortensen']\",\n",
       "   'text': ['D&apos;Amico and Vignale, and Flensberg and Jensen<bibr rid=\"b13\"/>, have calculated <f><italic>&rho;</italic><inf>&uarr;&darr;</inf>(<italic>T</italic>)</f> for a 2DEG using the random phase approximation, obtaining results that depend only on the electron density of the quantum well.'],\n",
       "   'year': '2001'},\n",
       "  {'ArticleID': 'pub.1060573261',\n",
       "   'CitedArticleID': 'pub.1060573261',\n",
       "   'Cites': 102,\n",
       "   'CitingArticleID': 'pub.1023679984',\n",
       "   'DOI': '10.1103/physrevb.50.1684',\n",
       "   'Issue': '3',\n",
       "   'Journal': 'Physical Review B',\n",
       "   'NatureID': '',\n",
       "   'PubYear': 1994,\n",
       "   'Title - matched_from': 'Quantum Monte Carlo calculation of the Fermi liquid parameters in the two-dimensional electron gas',\n",
       "   'Title - matched_to': 'Quantum Monte Carlo calculation of the Fermi-liquid parameters in the two-dimensional electron gas',\n",
       "   'Volume': '50',\n",
       "   'clean_text': ['The ratio is far greater than can be accounted for by many-body enhancement of the spin susceptibility, as the factor χs/χ0 is less than 1.4 in this range of electron density.',\n",
       "    'The fact that the slope is slightly greater than unity is consistent with the expectation that the many-body enhancement of χs relative to χ0 is small in this density regime.'],\n",
       "   'count': 2,\n",
       "   'fnm': \"['Y.', 'D. M.', 'R. M.']\",\n",
       "   'index': 'b9',\n",
       "   'journal': 'Phys. Rev. B',\n",
       "   'snm': \"['Kwon', 'Ceperley', 'Martin']\",\n",
       "   'text': ['The ratio is far greater than can be accounted for by many-body enhancement of the spin susceptibility, as the factor <i>&chi;</i><sub>s</sub>/<i>&chi;</i><sub>0</sub> is less than 1.4 in this range of electron density<bibr rid=\"b8 b9\"/>.',\n",
       "    'The fact that the slope is slightly greater than unity is consistent with the expectation that the many-body enhancement of <i>&chi;</i><sub>s</sub> relative to <i>&chi;</i><sub>0</sub> is small in this density regime<bibr rid=\"b8 b9\"/>.'],\n",
       "   'year': '1994'},\n",
       "  {'ArticleID': 'pub.1062451238',\n",
       "   'CitedArticleID': 'pub.1062451238',\n",
       "   'Cites': 1611,\n",
       "   'CitingArticleID': 'pub.1023679984',\n",
       "   'DOI': '10.1126/science.1105514',\n",
       "   'Issue': '5703',\n",
       "   'Journal': 'Science',\n",
       "   'NatureID': '',\n",
       "   'PubYear': 2004,\n",
       "   'Title - matched_from': 'Observation of the spin Hall effect in semiconductors',\n",
       "   'Title - matched_to': 'Observation of the Spin Hall Effect in Semiconductors',\n",
       "   'Volume': '306',\n",
       "   'clean_text': ['(The recently observed spin Hall effect also involves the counterflow of spin populations, and so should be damped by SCD.)'],\n",
       "   'count': 1,\n",
       "   'fnm': \"['Y. K.', 'R. C.', 'A. C.', 'D. D.']\",\n",
       "   'index': 'b11',\n",
       "   'journal': 'Science',\n",
       "   'snm': \"['Kato', 'Myers', 'Gossard', 'Awschalom']\",\n",
       "   'text': ['(The recently observed<bibr rid=\"b11 b12\"/> spin Hall effect also involves the counterflow of spin populations, and so should be damped by SCD.)'],\n",
       "   'year': '2004'},\n",
       "  {'ArticleID': 'pub.1060813403',\n",
       "   'CitedArticleID': 'pub.1060813403',\n",
       "   'Cites': 102,\n",
       "   'CitingArticleID': 'pub.1023679984',\n",
       "   'DOI': '10.1103/physrevlett.76.4793',\n",
       "   'Issue': '25',\n",
       "   'Journal': 'Physical Review Letters',\n",
       "   'NatureID': '',\n",
       "   'PubYear': 1996,\n",
       "   'Title - matched_from': 'Spin gratings and the measurement of electron drift mobility in multiple quantum well semiconductors',\n",
       "   'Title - matched_to': 'Spin Gratings and the Measurement of Electron Drift Mobility in Multiple Quantum Well Semiconductors',\n",
       "   'Volume': '76',\n",
       "   'clean_text': ['\\n',\n",
       "    'In this description, the decay rate varies with q quadratically, as γq &equals; τs - 1 &plus; Dsq2, where Ds is the spin diffusion coefficient and τs is the spin relaxation time.'],\n",
       "   'count': 2,\n",
       "   'fnm': \"['A. R.', 'P.', 'A.']\",\n",
       "   'index': 'b3',\n",
       "   'journal': 'Phys. Rev. Lett.',\n",
       "   'snm': \"['Cameron', 'Riblet', 'Miller']\",\n",
       "   'text': ['</fp>\\r\\n<p>In this work, spin diffusion is characterized by the transient spin grating technique<bibr rid=\"b3\"/>, which is based on the optical injection of spin-polarized electrons.',\n",
       "    'In this description, the decay rate varies with <i>q</i> quadratically, as <f><italic>&gamma;</italic><inf><italic>q</italic></inf> &equals; <italic>&tau;</italic><inf location=\"post\" arrange=\"compact\">s</inf><sup location=\"post\" arrange=\"compact\"> - 1</sup> &plus; <italic>D</italic><inf>s</inf><italic>q</italic><sup>2</sup></f>, where <i>D</i><sub>s</sub> is the spin diffusion coefficient and <i>&tau;</i><sub>s</sub> is the spin relaxation time<bibr rid=\"b3\"/>.'],\n",
       "   'year': '1996'},\n",
       "  {'ArticleID': 'pub.1055649471',\n",
       "   'CitedArticleID': 'pub.1055649471',\n",
       "   'Cites': 41,\n",
       "   'CitingArticleID': 'pub.1023679984',\n",
       "   'DOI': '10.1021/j100020a004',\n",
       "   'Issue': '20',\n",
       "   'Journal': 'The Journal of Physical Chemistry',\n",
       "   'NatureID': '',\n",
       "   'PubYear': 1995,\n",
       "   'Title - matched_from': 'Optical heterodyne-detection of impulsive stimulated Raman-scattering in liquids',\n",
       "   'Title - matched_to': 'Optical heterodyne detection of impulsive stimulated Raman scattering in liquids',\n",
       "   'Volume': '99',\n",
       "   'clean_text': ['To detect the induced spin grating, we mixed diffracted and transmitted probes to produce a photodetector current linear in the diffracted field.'],\n",
       "   'count': 1,\n",
       "   'fnm': \"['Y. J.', 'P.', 'J. D.']\",\n",
       "   'index': 'b17',\n",
       "   'journal': 'J. Phys. Chem.',\n",
       "   'snm': \"['Chang', 'Cong', 'Simon']\",\n",
       "   'text': ['To detect the induced spin grating, we mixed diffracted and transmitted probes to produce a photodetector current linear in the diffracted field<bibr rid=\"b16 b17 b18\"/>.'],\n",
       "   'year': '1995'},\n",
       "  {'ArticleID': 'pub.1019726809',\n",
       "   'CitedArticleID': 'pub.1019726809',\n",
       "   'Cites': 58,\n",
       "   'CitingArticleID': 'pub.1023679984',\n",
       "   'DOI': '10.1209/epl/i2001-00453-y',\n",
       "   'Issue': '4',\n",
       "   'Journal': 'EPL (Europhysics Letters)',\n",
       "   'NatureID': '',\n",
       "   'PubYear': 2001,\n",
       "   'Title - matched_from': 'Spin diffusion in doped semiconductors: The role of Coulomb interactions',\n",
       "   'Title - matched_to': 'Spin diffusion in doped semiconductors: The role of Coulomb interactions',\n",
       "   'Volume': '55',\n",
       "   'clean_text': ['Seen macroscopically, e–e collisions transfer momentum between the spin-up and spin-down populations, creating a force damping their relative motion that these authors term ‘spin Coulomb drag’ (SCD).',\n",
       "    \"According to D'Amico and Vignale, the reduction of Ds relative to Dc0 is:DsDc0 &equals; χ0χs11 &plus; &verbar;ρ↑↓&verbar;/ρ (2)\\r\\nwhere ρ &equals; 1/σc is the charge resistance and ρ↑↓ is the spin drag resistance, parameterizing the rate of momentum exchange between spin ↑ and ↓ electrons.\"],\n",
       "   'count': 2,\n",
       "   'fnm': \"['I.', 'G.']\",\n",
       "   'index': 'b10',\n",
       "   'journal': 'Europhys. Lett.',\n",
       "   'snm': '[\"D\\'Amico\", \\'Vignale\\']',\n",
       "   'text': ['Seen macroscopically, <i>e</i>&ndash;<i>e</i> collisions transfer momentum between the spin-up and spin-down populations, creating a force damping their relative motion that these authors term &lsquo;spin Coulomb drag&rsquo; (SCD)<bibr rid=\"b10\"/>.',\n",
       "    'According to D&apos;Amico and Vignale<bibr rid=\"b10\"/>, the reduction of <i>D</i><sub>s</sub> relative to <i>D</i><sub>c0</sub> is:<fd id=\"m2\"><fraction><num><italic>D</italic><inf>s</inf></num><den><italic>D</italic><inf>c0</inf></den></fraction> &equals; <fence lpost=\"(\" rpost=\")\"><fraction><num><italic>&chi;</italic><inf>0</inf></num><den><italic>&chi;</italic><inf>s</inf></den></fraction></fence><fraction><num>1</num><den>1 &plus; &verbar;<italic>&rho;</italic><inf>&uarr;&darr;</inf>&verbar;/<italic>&rho;</italic></den></fraction> (2)<mathimg entname=\"eqn2\"/></fd>\\r\\nwhere <f><italic>&rho;</italic> &equals; 1/<italic>&sigma;</italic><inf>c</inf></f> is the charge resistance and <f><italic>&rho;</italic><inf>&uarr;&darr;</inf></f> is the spin drag resistance, parameterizing the rate of momentum exchange between spin &uarr; and &darr; electrons.'],\n",
       "   'year': '2001'},\n",
       "  {'ArticleID': 'pub.1060804770',\n",
       "   'CitedArticleID': 'pub.1060804770',\n",
       "   'Cites': 233,\n",
       "   'CitingArticleID': 'pub.1023679984',\n",
       "   'DOI': '10.1103/physrevlett.68.349',\n",
       "   'Issue': '3',\n",
       "   'Journal': 'Physical Review Letters',\n",
       "   'NatureID': '',\n",
       "   'PubYear': 1992,\n",
       "   'Title - matched_from': 'Exciton spin dynamics in GaAs heterostructures',\n",
       "   'Title - matched_to': 'Exciton spin dynamics in GaAs heterostructures',\n",
       "   'Volume': '68',\n",
       "   'clean_text': ['Here 1/τs is independently determined from the decay rate of the circular dichroism induced by a circularly polarized pump beam (see Supplementary Information).'],\n",
       "   'count': 1,\n",
       "   'fnm': \"['S.', 'I.']\",\n",
       "   'index': 'b5',\n",
       "   'journal': 'Phys. Rev. Lett.',\n",
       "   'snm': \"['Bar-Ad', 'Bar-Joseph']\",\n",
       "   'text': ['Here <f>1/<italic>&tau;</italic><inf>s</inf></f> is independently determined from the decay rate of the circular dichroism induced by a circularly polarized pump beam<bibr rid=\"b5\"/> (see <sir rid=\"s1\">Supplementary Information</sir>).'],\n",
       "   'year': '1992'},\n",
       "  {'ArticleID': 'pub.1065222098',\n",
       "   'CitedArticleID': 'pub.1065222098',\n",
       "   'Cites': 23,\n",
       "   'CitingArticleID': 'pub.1023679984',\n",
       "   'DOI': '10.1364/ol.29.002109',\n",
       "   'Issue': '18',\n",
       "   'Journal': 'Optics Letters',\n",
       "   'NatureID': '',\n",
       "   'PubYear': 2004,\n",
       "   'Title - matched_from': 'Absolute phase measurement in heterodyne detection of transient gratings',\n",
       "   'Title - matched_to': 'Absolute phase measurement in heterodyne detection of transient gratings.',\n",
       "   'Volume': '29',\n",
       "   'clean_text': ['To detect the induced spin grating, we mixed diffracted and transmitted probes to produce a photodetector current linear in the diffracted field.'],\n",
       "   'count': 1,\n",
       "   'fnm': \"['N.', 'J.']\",\n",
       "   'index': 'b18',\n",
       "   'journal': 'Opt. Lett.',\n",
       "   'snm': \"['Gedik', 'Orenstein']\",\n",
       "   'text': ['To detect the induced spin grating, we mixed diffracted and transmitted probes to produce a photodetector current linear in the diffracted field<bibr rid=\"b16 b17 b18\"/>.'],\n",
       "   'year': '2004'},\n",
       "  {'ArticleID': 'pub.1098793363',\n",
       "   'CitedArticleID': 'pub.1098793363',\n",
       "   'Cites': 570,\n",
       "   'CitingArticleID': 'pub.1023679984',\n",
       "   'DOI': '10.1093/acprof:oso/9780198507796.001.0001',\n",
       "   'Issue': '',\n",
       "   'Journal': '',\n",
       "   'NatureID': '',\n",
       "   'PubYear': 2001,\n",
       "   'Title - matched_from': 'Spin diffusion in doped semiconductors: The role of Coulomb interactions',\n",
       "   'Title - matched_to': 'Electrons and Phonons',\n",
       "   'Volume': '',\n",
       "   'clean_text': ['Seen macroscopically, e–e collisions transfer momentum between the spin-up and spin-down populations, creating a force damping their relative motion that these authors term ‘spin Coulomb drag’ (SCD).',\n",
       "    \"According to D'Amico and Vignale, the reduction of Ds relative to Dc0 is:DsDc0 &equals; χ0χs11 &plus; &verbar;ρ↑↓&verbar;/ρ (2)\\r\\nwhere ρ &equals; 1/σc is the charge resistance and ρ↑↓ is the spin drag resistance, parameterizing the rate of momentum exchange between spin ↑ and ↓ electrons.\"],\n",
       "   'count': 2,\n",
       "   'fnm': \"['I.', 'G.']\",\n",
       "   'index': 'b10',\n",
       "   'journal': 'Europhys. Lett.',\n",
       "   'snm': '[\"D\\'Amico\", \\'Vignale\\']',\n",
       "   'text': ['Seen macroscopically, <i>e</i>&ndash;<i>e</i> collisions transfer momentum between the spin-up and spin-down populations, creating a force damping their relative motion that these authors term &lsquo;spin Coulomb drag&rsquo; (SCD)<bibr rid=\"b10\"/>.',\n",
       "    'According to D&apos;Amico and Vignale<bibr rid=\"b10\"/>, the reduction of <i>D</i><sub>s</sub> relative to <i>D</i><sub>c0</sub> is:<fd id=\"m2\"><fraction><num><italic>D</italic><inf>s</inf></num><den><italic>D</italic><inf>c0</inf></den></fraction> &equals; <fence lpost=\"(\" rpost=\")\"><fraction><num><italic>&chi;</italic><inf>0</inf></num><den><italic>&chi;</italic><inf>s</inf></den></fraction></fence><fraction><num>1</num><den>1 &plus; &verbar;<italic>&rho;</italic><inf>&uarr;&darr;</inf>&verbar;/<italic>&rho;</italic></den></fraction> (2)<mathimg entname=\"eqn2\"/></fd>\\r\\nwhere <f><italic>&rho;</italic> &equals; 1/<italic>&sigma;</italic><inf>c</inf></f> is the charge resistance and <f><italic>&rho;</italic><inf>&uarr;&darr;</inf></f> is the spin drag resistance, parameterizing the rate of momentum exchange between spin &uarr; and &darr; electrons.'],\n",
       "   'year': '2001'},\n",
       "  {'ArticleID': 'pub.1041009713',\n",
       "   'CitedArticleID': 'pub.1041009713',\n",
       "   'Cites': 45,\n",
       "   'CitingArticleID': 'pub.1023679984',\n",
       "   'DOI': '10.1103/physrevb.68.045307',\n",
       "   'Issue': '4',\n",
       "   'Journal': 'Physical Review B',\n",
       "   'NatureID': '',\n",
       "   'PubYear': 2003,\n",
       "   'Title - matched_from': 'Spin Coulomb drag in the two-dimensional electron liquid',\n",
       "   'Title - matched_to': 'Spin Coulomb drag in the two-dimensional electron liquid',\n",
       "   'Volume': '68',\n",
       "   'clean_text': [],\n",
       "   'count': 0,\n",
       "   'fnm': \"['I.', 'G.']\",\n",
       "   'index': 'b14',\n",
       "   'journal': 'Phys. Rev. B',\n",
       "   'snm': '[\"D\\'Amico\", \\'Vignale\\']',\n",
       "   'text': '',\n",
       "   'year': '2003'},\n",
       "  {'ArticleID': 'pub.1033723558',\n",
       "   'CitedArticleID': 'pub.1033723558',\n",
       "   'Cites': 710,\n",
       "   'CitingArticleID': 'pub.1023679984',\n",
       "   'DOI': '10.1038/16420',\n",
       "   'Issue': '6715',\n",
       "   'Journal': 'Nature',\n",
       "   'NatureID': '16420',\n",
       "   'PubYear': 1999,\n",
       "   'Title - matched_from': 'Lateral drag of spin coherence in gallium arsenide',\n",
       "   'Title - matched_to': 'Lateral drag of spin coherence in gallium arsenide',\n",
       "   'Volume': '397',\n",
       "   'clean_text': ['Finally, we note that SCD can be highly advantageous for spintronic applications, as it increases the distance that a spin packet can be dragged by an electric field, E, before it spreads owing to diffusion.'],\n",
       "   'count': 1,\n",
       "   'fnm': \"['J. M.', 'D. D.']\",\n",
       "   'index': 'b15',\n",
       "   'journal': 'Nature',\n",
       "   'snm': \"['Kikkawa', 'Awschalom']\",\n",
       "   'text': ['\\r\\n<p>Finally, we note that SCD can be highly advantageous for spintronic applications, as it increases the distance that a spin packet can be dragged by an electric field, <i>E</i>, before it spreads owing to diffusion<bibr rid=\"b15\"/>.'],\n",
       "   'year': '1999'},\n",
       "  {'ArticleID': 'pub.1005002745',\n",
       "   'CitedArticleID': 'pub.1005002745',\n",
       "   'Cites': 1066,\n",
       "   'CitingArticleID': 'pub.1023679984',\n",
       "   'DOI': '10.1103/physrevlett.94.047204',\n",
       "   'Issue': '4',\n",
       "   'Journal': 'Physical Review Letters',\n",
       "   'NatureID': '',\n",
       "   'PubYear': 2005,\n",
       "   'Title - matched_from': 'Experimental observation of the spin Hall effect in a two-dimensional spin-orbit coupled semiconductor system',\n",
       "   'Title - matched_to': 'Experimental Observation of the Spin-Hall Effect in a Two-Dimensional Spin-Orbit Coupled Semiconductor System',\n",
       "   'Volume': '94',\n",
       "   'clean_text': ['(The recently observed spin Hall effect also involves the counterflow of spin populations, and so should be damped by SCD.)'],\n",
       "   'count': 1,\n",
       "   'fnm': \"['J.', 'B.', 'J.', 'T.']\",\n",
       "   'index': 'b12',\n",
       "   'journal': 'Phys. Rev. Lett.',\n",
       "   'snm': \"['Wunderlich', 'Kaestner', 'Sinova', 'Jungwirth']\",\n",
       "   'text': ['(The recently observed<bibr rid=\"b11 b12\"/> spin Hall effect also involves the counterflow of spin populations, and so should be damped by SCD.)'],\n",
       "   'year': '2005'},\n",
       "  {'ArticleID': 'pub.1055648769',\n",
       "   'CitedArticleID': 'pub.1055648769',\n",
       "   'Cites': 94,\n",
       "   'CitingArticleID': 'pub.1023679984',\n",
       "   'DOI': '10.1021/j100009a027',\n",
       "   'Issue': '9',\n",
       "   'Journal': 'The Journal of Physical Chemistry',\n",
       "   'NatureID': '',\n",
       "   'PubYear': 1995,\n",
       "   'Title - matched_from': 'Transient grating optical heterodyne detected impulsive stimulated Raman-scattering in simple liquids',\n",
       "   'Title - matched_to': 'Transient Grating Optical Heterodyne Detected Impulsive Stimulated Raman Scattering in Simple Liquids',\n",
       "   'Volume': '99',\n",
       "   'clean_text': ['To detect the induced spin grating, we mixed diffracted and transmitted probes to produce a photodetector current linear in the diffracted field.'],\n",
       "   'count': 1,\n",
       "   'fnm': \"['P.', 'N. F.']\",\n",
       "   'index': 'b16',\n",
       "   'journal': 'J. Phys. Chem.',\n",
       "   'snm': \"['Vohringer', 'Scherer']\",\n",
       "   'text': ['To detect the induced spin grating, we mixed diffracted and transmitted probes to produce a photodetector current linear in the diffracted field<bibr rid=\"b16 b17 b18\"/>.'],\n",
       "   'year': '1995'},\n",
       "  {'ArticleID': 'pub.1060796185',\n",
       "   'CitedArticleID': 'pub.1060796185',\n",
       "   'Cites': 55,\n",
       "   'CitingArticleID': 'pub.1023679984',\n",
       "   'DOI': '10.1103/physrevlett.59.477',\n",
       "   'Issue': '4',\n",
       "   'Journal': 'Physical Review Letters',\n",
       "   'NatureID': '',\n",
       "   'PubYear': 1987,\n",
       "   'Title - matched_from': 'Thermal conductivity in disordered interacting-electron systems',\n",
       "   'Title - matched_to': 'Thermal conductivity in disordered interacting-electron systems',\n",
       "   'Volume': '59',\n",
       "   'clean_text': ['Physically, Dc0 is the quasiparticle diffusion coefficient, approaching µEF/e and µkBT/e in the degenerate and non-degenerate regimes, respectively.'],\n",
       "   'count': 1,\n",
       "   'fnm': \"['C.', 'C.', 'G.', 'P. A.', 'G.']\",\n",
       "   'index': 'b7',\n",
       "   'journal': 'Phys. Rev. Lett.',\n",
       "   'snm': \"['Castellani', 'DiCastro', 'Kotliar', 'Lee', 'Strinati']\",\n",
       "   'text': ['Physically, <i>D</i><sub>c0</sub> is the quasiparticle diffusion coefficient<bibr rid=\"b7\"/>, approaching <f><italic>&micro;E</italic><inf>F</inf>/<italic>e</italic></f> and <f><italic>&micro;k</italic><inf>B</inf><italic>T</italic>/<italic>e</italic></f> in the degenerate and non-degenerate regimes, respectively.'],\n",
       "   'year': '1987'}]}"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "papers['10.1038/nature04206']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Could not extract references from <ZipInfo filename='nature02811.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=1321 compress_size=691>\n",
      "Could not extract references from <ZipInfo filename='nature15633.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=11935 compress_size=4443>\n",
      "Could not extract references from <ZipInfo filename='nature.2015.18351.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=8812 compress_size=3774>\n",
      "Could not extract references from <ZipInfo filename='nature.2015.18431.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=10598 compress_size=4411>\n",
      "Could not extract references from <ZipInfo filename='nature.2015.18432.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=9529 compress_size=4061>\n",
      "Could not extract references from <ZipInfo filename='nature.2015.18440.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=8319 compress_size=3696>\n",
      "Could not extract references from <ZipInfo filename='nature.2015.18441.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=8948 compress_size=3844>\n",
      "Could not extract references from <ZipInfo filename='nature.2015.18448.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=10273 compress_size=4484>\n",
      "Could not extract references from <ZipInfo filename='nature12288.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=3646 compress_size=1520>\n",
      "Could not extract references from <ZipInfo filename='nature12317.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=4448 compress_size=1637>\n",
      "Could not extract references from <ZipInfo filename='nature13336.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=12906 compress_size=4738>\n",
      "Could not extract references from <ZipInfo filename='nature13341.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=12047 compress_size=4524>\n",
      "Could not extract references from <ZipInfo filename='nature13342.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=11219 compress_size=4064>\n",
      "Could not extract references from <ZipInfo filename='nature21381.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=2582 compress_size=1088>\n",
      "Could not extract references from <ZipInfo filename='nature12316.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=2409 compress_size=1135>\n",
      "Could not extract references from <ZipInfo filename='nature12324.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=2117 compress_size=989>\n",
      "Could not extract references from <ZipInfo filename='nature12325.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=1590 compress_size=874>\n",
      "Could not extract references from <ZipInfo filename='nature13211.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=11477 compress_size=4423>\n",
      "Could not extract references from <ZipInfo filename='nature13214.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=12409 compress_size=4791>\n",
      "Could not extract references from <ZipInfo filename='nature13215.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=12787 compress_size=4962>\n",
      "Could not extract references from <ZipInfo filename='nature12412.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=13638 compress_size=4992>\n",
      "Could not extract references from <ZipInfo filename='nature12413.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=11006 compress_size=4337>\n",
      "Could not extract references from <ZipInfo filename='nature12454.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=12129 compress_size=4532>\n",
      "Could not extract references from <ZipInfo filename='nature12455.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=13734 compress_size=4900>\n",
      "Could not extract references from <ZipInfo filename='nature05713.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=9186 compress_size=3667>\n",
      "Could not extract references from <ZipInfo filename='nature15253.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=4574 compress_size=1593>\n",
      "Could not extract references from <ZipInfo filename='nature15255.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=5265 compress_size=1712>\n",
      "Could not extract references from <ZipInfo filename='nature15370.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=7699 compress_size=3079>\n",
      "Could not extract references from <ZipInfo filename='nature12461.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=11114 compress_size=4454>\n",
      "Could not extract references from <ZipInfo filename='nature12462.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=11567 compress_size=4520>\n",
      "Could not extract references from <ZipInfo filename='nature02183.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=3205 compress_size=1204>\n",
      "Could not extract references from <ZipInfo filename='nature08600.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=1598 compress_size=842>\n",
      "Could not extract references from <ZipInfo filename='nature15638.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=12012 compress_size=4562>\n",
      "Could not extract references from <ZipInfo filename='nature15639.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=12133 compress_size=4654>\n",
      "Could not extract references from <ZipInfo filename='nature15704.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=3043 compress_size=1340>\n",
      "Could not extract references from <ZipInfo filename='nature02586.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=4619 compress_size=2030>\n",
      "Could not extract references from <ZipInfo filename='nature02587.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=4956 compress_size=2156>\n",
      "Could not extract references from <ZipInfo filename='nature21398.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=3006 compress_size=1384>\n",
      "Could not extract references from <ZipInfo filename='nature21404.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=1895 compress_size=936>\n",
      "Could not extract references from <ZipInfo filename='nature21502.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=11641 compress_size=4208>\n",
      "Could not extract references from <ZipInfo filename='nature08440.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=1666 compress_size=841>\n",
      "Could not extract references from <ZipInfo filename='nature08464.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=2012 compress_size=900>\n",
      "Could not extract references from <ZipInfo filename='nature08491.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=1868 compress_size=1007>\n",
      "Could not extract references from <ZipInfo filename='nature21505.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=14266 compress_size=5118>\n",
      "Could not extract references from <ZipInfo filename='nature21506.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=13351 compress_size=4881>\n",
      "Could not extract references from <ZipInfo filename='nature21507.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=12383 compress_size=4543>\n",
      "Could not extract references from <ZipInfo filename='nature.2017.21527.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=8821 compress_size=3915>\n",
      "Could not extract references from <ZipInfo filename='nature.2017.21536.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=8392 compress_size=3769>\n",
      "Could not extract references from <ZipInfo filename='nature.2017.21539.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=7705 compress_size=3328>\n",
      "Could not extract references from <ZipInfo filename='nature.2017.21550.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=11623 compress_size=4475>\n",
      "Could not extract references from <ZipInfo filename='nature11735.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=9385 compress_size=3290>\n",
      "Could not extract references from <ZipInfo filename='nature11758.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=15476 compress_size=5468>\n",
      "Could not extract references from <ZipInfo filename='nature11761.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=11324 compress_size=4335>\n",
      "Could not extract references from <ZipInfo filename='nature11762.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=11319 compress_size=4364>\n",
      "Could not extract references from <ZipInfo filename='nature11422.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=2263 compress_size=1162>\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Could not extract references from <ZipInfo filename='nature11426.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=4955 compress_size=1512>\n",
      "Could not extract references from <ZipInfo filename='nature14199.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=13562 compress_size=5336>\n",
      "Could not extract references from <ZipInfo filename='nature14201.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=11130 compress_size=4109>\n",
      "Could not extract references from <ZipInfo filename='nature11454.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=2189 compress_size=1028>\n",
      "Could not extract references from <ZipInfo filename='nature11455.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=2754 compress_size=1136>\n",
      "Could not extract references from <ZipInfo filename='nature04354.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=9073 compress_size=3555>\n",
      "Could not extract references from <ZipInfo filename='nature04355.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=8259 compress_size=3163>\n",
      "Could not extract references from <ZipInfo filename='nature04356.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=6195 compress_size=2354>\n",
      "Could not extract references from <ZipInfo filename='nature12835.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=11885 compress_size=4699>\n",
      "Could not extract references from <ZipInfo filename='nature12836.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=8959 compress_size=3762>\n",
      "Could not extract references from <ZipInfo filename='nature12837.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=10568 compress_size=3982>\n",
      "Could not extract references from <ZipInfo filename='nature12838.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=10723 compress_size=4316>\n",
      "Could not extract references from <ZipInfo filename='nature06871.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=1972 compress_size=904>\n",
      "Could not extract references from <ZipInfo filename='nature06872.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=1457 compress_size=746>\n",
      "Could not extract references from <ZipInfo filename='nature06898.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=1654 compress_size=841>\n",
      "Could not extract references from <ZipInfo filename='nature13183.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=2053 compress_size=1108>\n",
      "Could not extract references from <ZipInfo filename='nature15531.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=1736 compress_size=884>\n",
      "Could not extract references from <ZipInfo filename='nature15532.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=2558 compress_size=1274>\n",
      "Could not extract references from <ZipInfo filename='nature15534.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=2559 compress_size=1168>\n",
      "Could not extract references from <ZipInfo filename='nature11489.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=11744 compress_size=4533>\n",
      "Could not extract references from <ZipInfo filename='nature13219.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=13734 compress_size=4919>\n",
      "Could not extract references from <ZipInfo filename='nature13221.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=15895 compress_size=5483>\n",
      "Could not extract references from <ZipInfo filename='nature13222.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=10473 compress_size=4143>\n",
      "Could not extract references from <ZipInfo filename='nature15631.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=12241 compress_size=4704>\n",
      "Could not extract references from <ZipInfo filename='nature15632.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=14674 compress_size=5168>\n",
      "Could not extract references from <ZipInfo filename='nature15635.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=12574 compress_size=4964>\n",
      "Could not extract references from <ZipInfo filename='nature15640.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=13793 compress_size=4934>\n",
      "Could not extract references from <ZipInfo filename='nature06845.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=1709 compress_size=843>\n",
      "Could not extract references from <ZipInfo filename='nature.2015.18631.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=7578 compress_size=3317>\n",
      "Could not extract references from <ZipInfo filename='nature.2015.18638.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=10446 compress_size=4594>\n",
      "Could not extract references from <ZipInfo filename='nature.2015.18642.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=7595 compress_size=3254>\n",
      "Could not extract references from <ZipInfo filename='nature13003.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=5332 compress_size=2248>\n",
      "Could not extract references from <ZipInfo filename='nature13041.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=3296 compress_size=1251>\n",
      "Could not extract references from <ZipInfo filename='nature13042.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=2474 compress_size=1156>\n",
      "Could not extract references from <ZipInfo filename='nature13048.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=12994 compress_size=4815>\n",
      "Could not extract references from <ZipInfo filename='nature13051.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=11018 compress_size=4387>\n",
      "Could not extract references from <ZipInfo filename='nature13052.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=11890 compress_size=4394>\n",
      "Could not extract references from <ZipInfo filename='nature13053.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=9720 compress_size=4078>\n",
      "Could not extract references from <ZipInfo filename='nature05789.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=2196 compress_size=1101>\n",
      "Could not extract references from <ZipInfo filename='nature05882.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=2159 compress_size=1142>\n",
      "Could not extract references from <ZipInfo filename='nature13635.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=2070 compress_size=1015>\n",
      "Could not extract references from <ZipInfo filename='nature05888.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=12089 compress_size=4426>\n",
      "Could not extract references from <ZipInfo filename='nature13655.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=12762 compress_size=4699>\n",
      "Could not extract references from <ZipInfo filename='nature13661.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=3451 compress_size=1467>\n",
      "Could not extract references from <ZipInfo filename='nature25470.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=7323 compress_size=2285>\n",
      "Could not extract references from <ZipInfo filename='nature25472.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=1639 compress_size=903>\n",
      "Could not extract references from <ZipInfo filename='nature13740.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=13442 compress_size=4653>\n",
      "Could not extract references from <ZipInfo filename='nature25498.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=3522 compress_size=1428>\n",
      "Could not extract references from <ZipInfo filename='nature07824.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=2082 compress_size=976>\n",
      "Could not extract references from <ZipInfo filename='nature14520.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=11355 compress_size=4445>\n",
      "Could not extract references from <ZipInfo filename='nature10264.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=1769 compress_size=934>\n",
      "Could not extract references from <ZipInfo filename='nature10265.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=13434 compress_size=4979>\n",
      "Could not extract references from <ZipInfo filename='nature14523.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=11593 compress_size=4349>\n",
      "Could not extract references from <ZipInfo filename='nature14529.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=12117 compress_size=4516>\n",
      "Could not extract references from <ZipInfo filename='nature14530.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=16304 compress_size=6002>\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Could not extract references from <ZipInfo filename='nature14531.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=13349 compress_size=4939>\n",
      "Could not extract references from <ZipInfo filename='nature.2015.17651.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=9615 compress_size=4174>\n",
      "Could not extract references from <ZipInfo filename='nature12456.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=11127 compress_size=4479>\n",
      "Could not extract references from <ZipInfo filename='nature12457.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=12580 compress_size=4679>\n",
      "Could not extract references from <ZipInfo filename='nature12458.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=13834 compress_size=5166>\n",
      "Could not extract references from <ZipInfo filename='nature12459.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=10880 compress_size=4060>\n",
      "Could not extract references from <ZipInfo filename='nature22807.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=2331 compress_size=1063>\n",
      "Could not extract references from <ZipInfo filename='nature21106.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=11958 compress_size=4589>\n",
      "Could not extract references from <ZipInfo filename='nature21110.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=13787 compress_size=5147>\n",
      "Could not extract references from <ZipInfo filename='nature21111.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=12444 compress_size=4593>\n",
      "Could not extract references from <ZipInfo filename='nature21112.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=12183 compress_size=4693>\n",
      "Could not extract references from <ZipInfo filename='nature.2017.21253.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=8197 compress_size=3286>\n",
      "Could not extract references from <ZipInfo filename='nature.2017.21256.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=10883 compress_size=4678>\n",
      "Could not extract references from <ZipInfo filename='nature13330.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=13145 compress_size=4753>\n",
      "Could not extract references from <ZipInfo filename='nature22985.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=2705 compress_size=1283>\n",
      "Could not extract references from <ZipInfo filename='nature07826.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=1598 compress_size=797>\n",
      "Could not extract references from <ZipInfo filename='nature07833.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=1327 compress_size=726>\n",
      "Could not extract references from <ZipInfo filename='nature07835.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=1808 compress_size=904>\n",
      "Could not extract references from <ZipInfo filename='nature23012.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=5074 compress_size=1686>\n",
      "Could not extract references from <ZipInfo filename='nature23088.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=13905 compress_size=5383>\n",
      "Could not extract references from <ZipInfo filename='nature23090.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=15722 compress_size=5560>\n",
      "Could not extract references from <ZipInfo filename='nature.2017.22167.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=9539 compress_size=4086>\n",
      "Could not extract references from <ZipInfo filename='nature.2017.22173.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=8228 compress_size=3561>\n",
      "Could not extract references from <ZipInfo filename='nature.2017.22190.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=10334 compress_size=4189>\n",
      "Could not extract references from <ZipInfo filename='nature07346.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=7211 compress_size=1968>\n",
      "Could not extract references from <ZipInfo filename='nature10156.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=1958 compress_size=997>\n",
      "Could not extract references from <ZipInfo filename='nature10186.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=2658 compress_size=1233>\n",
      "Could not extract references from <ZipInfo filename='nature11384.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=11358 compress_size=4327>\n",
      "Could not extract references from <ZipInfo filename='nature08359.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=3487 compress_size=1601>\n",
      "Could not extract references from <ZipInfo filename='nature11749.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=14626 compress_size=5325>\n",
      "Could not extract references from <ZipInfo filename='nature11751.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=16624 compress_size=5989>\n",
      "Could not extract references from <ZipInfo filename='nature11752.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=12081 compress_size=4376>\n",
      "Could not extract references from <ZipInfo filename='nature11753.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=13721 compress_size=5142>\n",
      "Could not extract references from <ZipInfo filename='nature05548.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=18579 compress_size=5657>\n",
      "Could not extract references from <ZipInfo filename='nature05549.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=14109 compress_size=4597>\n",
      "Could not extract references from <ZipInfo filename='nature05550.xml' compress_type=deflate filemode='-rwsrwxr-x' file_size=5129 compress_size=2136>\n",
      "Could not extract references from <ZipInfo filename='nature05565.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=14542 compress_size=5369>\n",
      "Could not extract references from <ZipInfo filename='nature11029.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=2496 compress_size=1235>\n",
      "Could not extract references from <ZipInfo filename='nature04898.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=15611 compress_size=5427>\n",
      "Could not extract references from <ZipInfo filename='nature04899.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=9628 compress_size=3740>\n",
      "Could not extract references from <ZipInfo filename='nature11036.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=13546 compress_size=4867>\n",
      "Could not extract references from <ZipInfo filename='nature04900.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=10218 compress_size=3684>\n",
      "Could not extract references from <ZipInfo filename='nature24008.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=3377 compress_size=1590>\n",
      "Could not extract references from <ZipInfo filename='nature24013.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=2413 compress_size=1043>\n",
      "Could not extract references from <ZipInfo filename='nature01308.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=1848 compress_size=1019>\n",
      "Could not extract references from <ZipInfo filename='nature01329.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=2293 compress_size=1008>\n",
      "Could not extract references from <ZipInfo filename='nature07964.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=1497 compress_size=826>\n",
      "Could not extract references from <ZipInfo filename='nature05893.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=13539 compress_size=4878>\n",
      "Could not extract references from <ZipInfo filename='nature01561.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=1927 compress_size=993>\n",
      "Could not extract references from <ZipInfo filename='nature01562.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=1434 compress_size=816>\n",
      "Could not extract references from <ZipInfo filename='nature01565.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=2003 compress_size=932>\n",
      "Could not extract references from <ZipInfo filename='nature24140.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=13420 compress_size=4983>\n",
      "Could not extract references from <ZipInfo filename='nature24142.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=12922 compress_size=4649>\n",
      "Could not extract references from <ZipInfo filename='nature24144.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=12180 compress_size=4707>\n",
      "Could not extract references from <ZipInfo filename='nature.2017.22686.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=8760 compress_size=3732>\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Could not extract references from <ZipInfo filename='nature.2017.22694.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=10838 compress_size=4430>\n",
      "Could not extract references from <ZipInfo filename='nature.2017.22731.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=6069 compress_size=2748>\n",
      "Could not extract references from <ZipInfo filename='nature.2017.22736.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=8487 compress_size=3639>\n",
      "Could not extract references from <ZipInfo filename='nature.2017.22737.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=13236 compress_size=5323>\n",
      "Could not extract references from <ZipInfo filename='nature08846.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=8569 compress_size=1924>\n",
      "Could not extract references from <ZipInfo filename='nature08847.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=1627 compress_size=824>\n",
      "Could not extract references from <ZipInfo filename='nature10884.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=8471 compress_size=3667>\n",
      "Could not extract references from <ZipInfo filename='nature13066.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=11489 compress_size=4387>\n",
      "Could not extract references from <ZipInfo filename='nature13067.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=15716 compress_size=5347>\n",
      "Could not extract references from <ZipInfo filename='nature21494.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=12288 compress_size=5196>\n",
      "Could not extract references from <ZipInfo filename='nature21495.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=12001 compress_size=4985>\n",
      "Could not extract references from <ZipInfo filename='nature21499.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=12598 compress_size=4641>\n",
      "Could not extract references from <ZipInfo filename='nature21500.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=13885 compress_size=5118>\n",
      "Could not extract references from <ZipInfo filename='nature.2017.21439.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=11863 compress_size=4662>\n",
      "Could not extract references from <ZipInfo filename='nature.2017.21440.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=7261 compress_size=3239>\n",
      "Could not extract references from <ZipInfo filename='nature.2017.21462.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=11756 compress_size=4410>\n",
      "Could not extract references from <ZipInfo filename='nature.2017.21466.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=9699 compress_size=4171>\n",
      "Could not extract references from <ZipInfo filename='nature.2017.21470.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=9023 compress_size=4017>\n",
      "Could not extract references from <ZipInfo filename='nature13126.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=2797 compress_size=1160>\n",
      "Could not extract references from <ZipInfo filename='nature13143.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=3541 compress_size=1547>\n",
      "Could not extract references from <ZipInfo filename='nature13208.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=11698 compress_size=4587>\n",
      "Could not extract references from <ZipInfo filename='nature13209.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=11602 compress_size=4375>\n",
      "Could not extract references from <ZipInfo filename='nature11766.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=12570 compress_size=4656>\n",
      "Could not extract references from <ZipInfo filename='nature25163.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=2321 compress_size=1199>\n",
      "Could not extract references from <ZipInfo filename='nature25445.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=3760 compress_size=1596>\n",
      "Could not extract references from <ZipInfo filename='nature25471.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=2150 compress_size=994>\n",
      "Could not extract references from <ZipInfo filename='nature22493.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=12854 compress_size=4857>\n",
      "Could not extract references from <ZipInfo filename='nature22494.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=14371 compress_size=5048>\n",
      "Could not extract references from <ZipInfo filename='nature.2017.21975.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=9104 compress_size=3989>\n",
      "Could not extract references from <ZipInfo filename='nature.2017.21977.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=8789 compress_size=4000>\n",
      "Could not extract references from <ZipInfo filename='nature.2017.21985.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=6413 compress_size=2998>\n",
      "Could not extract references from <ZipInfo filename='nature.2017.21987.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=7026 compress_size=3222>\n",
      "Could not extract references from <ZipInfo filename='nature09034.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=1589 compress_size=853>\n",
      "Could not extract references from <ZipInfo filename='nature09046.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=1556 compress_size=799>\n",
      "Could not extract references from <ZipInfo filename='nature723.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=14282 compress_size=4522>\n",
      "Could not extract references from <ZipInfo filename='nature725.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=13907 compress_size=5176>\n",
      "Could not extract references from <ZipInfo filename='nature05002.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=11040 compress_size=4338>\n",
      "Could not extract references from <ZipInfo filename='nature05003.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=9912 compress_size=3773>\n",
      "Could not extract references from <ZipInfo filename='nature05212.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=14136 compress_size=4691>\n",
      "Could not extract references from <ZipInfo filename='nature05213.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=9830 compress_size=3252>\n",
      "Could not extract references from <ZipInfo filename='nature12409.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=10393 compress_size=4227>\n",
      "Could not extract references from <ZipInfo filename='nature12410.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=13100 compress_size=4797>\n",
      "Could not extract references from <ZipInfo filename='nature12411.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=14177 compress_size=4986>\n",
      "Could not extract references from <ZipInfo filename='nature13333.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=11871 compress_size=4401>\n",
      "Could not extract references from <ZipInfo filename='nature13337.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=11817 compress_size=4530>\n",
      "Could not extract references from <ZipInfo filename='nature13338.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=12579 compress_size=4647>\n",
      "Could not extract references from <ZipInfo filename='nature13339.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=13888 compress_size=5074>\n",
      "Could not extract references from <ZipInfo filename='nature23548.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=15411 compress_size=5362>\n",
      "Could not extract references from <ZipInfo filename='nature01896.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=1732 compress_size=830>\n",
      "Could not extract references from <ZipInfo filename='nature04218.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=2616 compress_size=1135>\n",
      "Could not extract references from <ZipInfo filename='nature04685.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=37497 compress_size=6823>\n",
      "Could not extract references from <ZipInfo filename='nature04686.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=14571 compress_size=4216>\n",
      "Could not extract references from <ZipInfo filename='nature01924.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=1396 compress_size=785>\n",
      "Could not extract references from <ZipInfo filename='nature01952.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=1792 compress_size=971>\n",
      "Could not extract references from <ZipInfo filename='nature25991.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=3931 compress_size=1502>\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Could not extract references from <ZipInfo filename='nature25994.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=1696 compress_size=874>\n",
      "Could not extract references from <ZipInfo filename='nature24139.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=13590 compress_size=5072>\n",
      "Could not extract references from <ZipInfo filename='nature24141.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=12505 compress_size=4800>\n",
      "Could not extract references from <ZipInfo filename='nature.2017.22648.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=8629 compress_size=3669>\n",
      "Could not extract references from <ZipInfo filename='nature.2017.22655.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=9432 compress_size=4034>\n",
      "Could not extract references from <ZipInfo filename='nature.2017.22667.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=7429 compress_size=3359>\n",
      "Could not extract references from <ZipInfo filename='nature26164.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=12144 compress_size=2985>\n",
      "Could not extract references from <ZipInfo filename='nature05712.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=12105 compress_size=4759>\n",
      "Could not extract references from <ZipInfo filename='nature05753.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=12633 compress_size=4357>\n",
      "Could not extract references from <ZipInfo filename='nature05754.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=10859 compress_size=3948>\n",
      "Could not extract references from <ZipInfo filename='nature05791.xml' compress_type=deflate filemode='-rw-rw-rw-' file_size=1529 compress_size=839>\n",
      "Could not extract references from <ZipInfo filename='nature13331.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=13626 compress_size=5019>\n",
      "Could not extract references from <ZipInfo filename='nature13332.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=12618 compress_size=4649>\n",
      "Could not extract references from <ZipInfo filename='nature14303.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=2919 compress_size=1184>\n",
      "Could not extract references from <ZipInfo filename='nature14304.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=3147 compress_size=1352>\n",
      "Could not extract references from <ZipInfo filename='nature01168.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=1794 compress_size=932>\n",
      "Could not extract references from <ZipInfo filename='nature14377.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=12590 compress_size=4734>\n",
      "Could not extract references from <ZipInfo filename='nature14381.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=9857 compress_size=4177>\n",
      "Could not extract references from <ZipInfo filename='nature14382.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=12734 compress_size=4771>\n",
      "Could not extract references from <ZipInfo filename='nature14384.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=11660 compress_size=4701>\n",
      "Could not extract references from <ZipInfo filename='nature11956.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=12611 compress_size=4903>\n",
      "Could not extract references from <ZipInfo filename='nature12084.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=11677 compress_size=4807>\n",
      "Could not extract references from <ZipInfo filename='nature12085.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=14840 compress_size=5065>\n",
      "Could not extract references from <ZipInfo filename='nature01221.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=2026 compress_size=1110>\n",
      "Could not extract references from <ZipInfo filename='nature11754.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=12365 compress_size=4849>\n",
      "Could not extract references from <ZipInfo filename='nature11757.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=11425 compress_size=4551>\n",
      "Could not extract references from <ZipInfo filename='nature11759.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=14791 compress_size=5421>\n",
      "Could not extract references from <ZipInfo filename='nature11760.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=11358 compress_size=4570>\n",
      "Could not extract references from <ZipInfo filename='nature06365.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=11274 compress_size=4372>\n",
      "Could not extract references from <ZipInfo filename='nature06540.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=1709 compress_size=882>\n",
      "Could not extract references from <ZipInfo filename='nature07044.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=1989 compress_size=1025>\n",
      "Could not extract references from <ZipInfo filename='nature10952.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=14112 compress_size=5083>\n",
      "Could not extract references from <ZipInfo filename='nature10976.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=2469 compress_size=1169>\n",
      "Could not extract references from <ZipInfo filename='nature11034.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=11781 compress_size=4720>\n",
      "Could not extract references from <ZipInfo filename='nature22501.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=12455 compress_size=4854>\n",
      "Could not extract references from <ZipInfo filename='nature22502.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=14469 compress_size=5205>\n",
      "Could not extract references from <ZipInfo filename='nature22503.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=14773 compress_size=5344>\n",
      "Could not extract references from <ZipInfo filename='nature.2017.22096.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=10103 compress_size=4181>\n",
      "Could not extract references from <ZipInfo filename='nature13758.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=15268 compress_size=5285>\n",
      "Could not extract references from <ZipInfo filename='nature24756.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=12132 compress_size=4880>\n",
      "Could not extract references from <ZipInfo filename='nature.2017.22998.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=8528 compress_size=3579>\n",
      "Could not extract references from <ZipInfo filename='nature.2017.23000.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=8306 compress_size=3888>\n",
      "Could not extract references from <ZipInfo filename='nature.2017.23003.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=7710 compress_size=3286>\n",
      "Could not extract references from <ZipInfo filename='nature.2017.23023.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=10821 compress_size=4570>\n",
      "Could not extract references from <ZipInfo filename='nature01349.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=1191 compress_size=702>\n",
      "Could not extract references from <ZipInfo filename='nature05366.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=3755 compress_size=1684>\n",
      "Could not extract references from <ZipInfo filename='nature05377.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=2221 compress_size=1055>\n",
      "Could not extract references from <ZipInfo filename='nature05400.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=1053 compress_size=627>\n",
      "Could not extract references from <ZipInfo filename='nature05417.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=9272 compress_size=3593>\n",
      "Could not extract references from <ZipInfo filename='nature05418.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=9639 compress_size=3445>\n",
      "Could not extract references from <ZipInfo filename='nature05423.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=1793 compress_size=988>\n",
      "Could not extract references from <ZipInfo filename='nature05460.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=2361 compress_size=1124>\n",
      "Could not extract references from <ZipInfo filename='nature14671.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=4005 compress_size=1497>\n",
      "Could not extract references from <ZipInfo filename='nature14871.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=1846 compress_size=945>\n",
      "Could not extract references from <ZipInfo filename='nature14955.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=1943 compress_size=979>\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Could not extract references from <ZipInfo filename='nature15216.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=11917 compress_size=4750>\n",
      "Could not extract references from <ZipInfo filename='nature15217.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=12449 compress_size=4876>\n",
      "Could not extract references from <ZipInfo filename='nature15630.xml' compress_type=deflate filemode='-rwxrwxr-x' file_size=16612 compress_size=5918>\n",
      "Could not extract references from <ZipInfo filename='nature.2015.18381.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=7195 compress_size=3131>\n",
      "Could not extract references from <ZipInfo filename='nature.2015.18391.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=8019 compress_size=3607>\n",
      "Could not extract references from <ZipInfo filename='nature00951.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=1466 compress_size=749>\n",
      "Could not extract references from <ZipInfo filename='nature11191.xml' compress_type=deflate filemode='-rw-rw-r--' file_size=9696 compress_size=3781>\n",
      "0.871452 of all processed files Succeeded\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Process ForkPoolWorker-8:\n",
      "Process ForkPoolWorker-12:\n",
      "Process ForkPoolWorker-9:\n",
      "Process ForkPoolWorker-10:\n",
      "Process ForkPoolWorker-14:\n",
      "Process ForkPoolWorker-13:\n",
      "Process ForkPoolWorker-11:\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/lib/python3.5/multiprocessing/process.py\", line 249, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/process.py\", line 249, in _bootstrap\n",
      "    self.run()\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/lib/python3.5/multiprocessing/process.py\", line 249, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/process.py\", line 249, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/process.py\", line 249, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/process.py\", line 249, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/process.py\", line 93, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/lib/python3.5/multiprocessing/process.py\", line 93, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/lib/python3.5/multiprocessing/process.py\", line 93, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/lib/python3.5/multiprocessing/process.py\", line 249, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/process.py\", line 93, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/lib/python3.5/multiprocessing/process.py\", line 93, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/lib/python3.5/multiprocessing/process.py\", line 93, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/lib/python3.5/multiprocessing/pool.py\", line 108, in worker\n",
      "    task = get()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/pool.py\", line 108, in worker\n",
      "    task = get()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/process.py\", line 93, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/lib/python3.5/multiprocessing/pool.py\", line 108, in worker\n",
      "    task = get()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/pool.py\", line 108, in worker\n",
      "    task = get()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/pool.py\", line 108, in worker\n",
      "    task = get()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/pool.py\", line 108, in worker\n",
      "    task = get()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/queues.py\", line 342, in get\n",
      "    with self._rlock:\n",
      "  File \"/usr/lib/python3.5/multiprocessing/queues.py\", line 342, in get\n",
      "    with self._rlock:\n",
      "  File \"/usr/lib/python3.5/multiprocessing/queues.py\", line 342, in get\n",
      "    with self._rlock:\n",
      "  File \"/usr/lib/python3.5/multiprocessing/queues.py\", line 342, in get\n",
      "    with self._rlock:\n",
      "  File \"/usr/lib/python3.5/multiprocessing/pool.py\", line 108, in worker\n",
      "    task = get()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/synchronize.py\", line 96, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/synchronize.py\", line 96, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/synchronize.py\", line 96, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/queues.py\", line 343, in get\n",
      "    res = self._reader.recv_bytes()\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/lib/python3.5/multiprocessing/queues.py\", line 342, in get\n",
      "    with self._rlock:\n",
      "  File \"/usr/lib/python3.5/multiprocessing/synchronize.py\", line 96, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/connection.py\", line 216, in recv_bytes\n",
      "    buf = self._recv_bytes(maxlength)\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/lib/python3.5/multiprocessing/connection.py\", line 407, in _recv_bytes\n",
      "    buf = self._recv(4)\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/lib/python3.5/multiprocessing/synchronize.py\", line 96, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/queues.py\", line 342, in get\n",
      "    with self._rlock:\n",
      "  File \"/usr/lib/python3.5/multiprocessing/connection.py\", line 379, in _recv\n",
      "    chunk = read(handle, remaining)\n",
      "KeyboardInterrupt\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/lib/python3.5/multiprocessing/synchronize.py\", line 96, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "KeyboardInterrupt\n",
      "KeyboardInterrupt\n"
     ]
    }
   ],
   "source": [
    "%time\n",
    "finfos, contents = get_zips_parallel_mapper(directory)\n",
    "p = Pool(7)\n",
    "success = p.map(get_zips_parallel_reducer, zip(finfos, contents))\n",
    "error_rate = sum([1 if i == True else 0 for i in success])/len(success)\n",
    "print ('%f of all processed files Succeeded' % error_rate)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "def combine_jsons(directory = 'data/intermediate/full_text_mappings',\n",
    "                  target = 'data/intermediate/matched_full_text.jsons'):\n",
    "\n",
    "    \n",
    "    with open (target, 'w') as f_to:\n",
    "        for file in tqdm(os.listdir(directory)):\n",
    "            with open(os.path.join(directory, file), 'r') as f_from:\n",
    "                f_to.write(f_from.read() + '\\n')\n",
    "                \n",
    "            \n",
    "    \n",
    "    return True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.5/dist-packages/ipykernel_launcher.py:6: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0\n",
      "Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`\n",
      "  \n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6e4479c78bed41a4afde1f1093f7fe7d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=0, max=14286), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combine_jsons()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14286\n"
     ]
    }
   ],
   "source": [
    "with open('data/intermediate/matched_full_text.jsons', 'r') as f:\n",
    "    \n",
    "    print(len(f.readlines()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "line = json.loads(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'metadata': {'metadata': {'doi': '10.1038/nature16166',\n",
       "   'id': 'nature16166',\n",
       "   'jtl': 'Nature',\n",
       "   'language': 'eng',\n",
       "   'relation': 'no',\n",
       "   'title': 'Substantial contribution of extrinsic risk factors to cancer development',\n",
       "   'vol': '529'},\n",
       "  'references': [{'ArticleID': 'pub.1035931121',\n",
       "    'CitedArticleID': 'pub.1035931121',\n",
       "    'Cites': 646,\n",
       "    'CitingArticleID': 'pub.1037786036',\n",
       "    'DOI': '10.1038/bjc.1991.210',\n",
       "    'Issue': '6',\n",
       "    'Journal': 'British Journal of Cancer',\n",
       "    'NatureID': 'bjc.1991.210',\n",
       "    'PubYear': 1991,\n",
       "    'Title - matched_from': 'Cancers of the prostate and breast among Japanese and white immigrants in Los Angeles County',\n",
       "    'Title - matched_to': 'Cancers of the prostate and breast among Japanese and white immigrants in Los Angeles County',\n",
       "    'Volume': '63',\n",
       "    'clean_text': ['Particularly, for breast and prostate cancers, it has long been observed that large international geographical variations exist in their incidence rates (for example, Western Europe has the highest incidence of breast cancer, which is almost 5 times higher than areas such as Eastern Asia or Middle Africa; Australia/New Zealand has the highest incidence of prostate cancer, which is almost 25 times higher than areas such as South-Central Asia), and immigrants moving from countries with lower cancer incidence to countries with higher cancer rates soon acquire the higher risk of their new country.',\n",
       "     'More importantly, cancers known to have substantial environmental risk proportions, for example, breast cancer, prostate cancer, colorectal cancer, melanoma, head and neck cancer, oesophageal cancer, cervical cancer, liver cancer and stomach cancer, all harbour large percentages of total extrinsic mutational signatures.'],\n",
       "    'count': 2,\n",
       "    'fnm': \"['H.']\",\n",
       "    'index': 'b16',\n",
       "    'journal': 'Br. J. Cancer',\n",
       "    'snm': \"['Shimizu']\",\n",
       "    'text': ['Particularly, for breast and prostate cancers, it has long been observed that large international geographical variations exist in their incidence rates (for example, Western Europe has the highest incidence of breast cancer, which is almost 5 times higher than areas such as Eastern Asia or Middle Africa; Australia/New Zealand has the highest incidence of prostate cancer, which is almost 25 times higher than areas such as South-Central Asia)<bibr rid=\"b14\"/>, and immigrants moving from countries with lower cancer incidence to countries with higher cancer rates soon acquire the higher risk of their new country<bibr rid=\"b15 b16\"/>.',\n",
       "     'More importantly, cancers known to have substantial environmental risk proportions, for example, breast cancer<bibr rid=\"b15\"/>, prostate cancer<bibr rid=\"b16\"/>, colorectal cancer<bibr rid=\"b18\"/>, melanoma<bibr rid=\"b19\"/>, head and neck cancer<bibr rid=\"b21\"/>, oesophageal cancer<bibr rid=\"b22\"/>, cervical cancer<bibr rid=\"b23\"/>, liver cancer<bibr rid=\"b26\"/> and stomach cancer<bibr rid=\"b27\"/>, all harbour large percentages of total extrinsic mutational signatures.'],\n",
       "    'year': '1991'},\n",
       "   {'ArticleID': 'pub.1050004918',\n",
       "    'CitedArticleID': 'pub.1050004918',\n",
       "    'Cites': 3682,\n",
       "    'CitingArticleID': 'pub.1037786036',\n",
       "    'DOI': '10.1038/nature12477',\n",
       "    'Issue': '7463',\n",
       "    'Journal': 'Nature',\n",
       "    'NatureID': 'nature12477',\n",
       "    'PubYear': 2013,\n",
       "    'Title - matched_from': 'Signatures of mutational processes in human cancer',\n",
       "    'Title - matched_to': 'Signatures of mutational processes in human cancer',\n",
       "    'Volume': '500',\n",
       "    'clean_text': ['These are regarded as ‘fingerprints’ left on cancer genomes by different mutagenic processes, revealing ~30 distinct signatures among various cancers.',\n",
       "     '31), demonstrated strong positive correlations with age in the majority of cancers, suggesting that they are acquired at a relatively constant rate over the lifetime of cancer patients and thus probably result from intrinsic processes; however, all other signature mutations (~30) lack the consistent correlations with age, suggesting that they are acquired at different rates in life and thus are probably a consequence of extrinsic carcinogen exposures.',\n",
       "     'Indeed, several mutational signatures have been linked to known factors such as UV radiation and smoking.'],\n",
       "    'count': 3,\n",
       "    'fnm': \"['L. B.']\",\n",
       "    'index': 'b31',\n",
       "    'journal': 'Nature',\n",
       "    'snm': \"['Alexandrov']\",\n",
       "    'text': ['These are regarded as &#x02018;fingerprints&#x02019; left on cancer genomes by different mutagenic processes<bibr rid=\"b30\"/>, revealing &#x0007E;30 distinct signatures among various cancers<bibr rid=\"b31\"/>.',\n",
       "     '<bibrinl rid=\"b31\">31</bibrinl>), demonstrated strong positive correlations with age in the majority of cancers, suggesting that they are acquired at a relatively constant rate over the lifetime of cancer patients and thus probably result from intrinsic processes; however, all other signature mutations (&#x0007E;30) lack the consistent correlations with age, suggesting that they are acquired at different rates in life and thus are probably a consequence of extrinsic carcinogen exposures<bibr rid=\"b31\"/>.',\n",
       "     'Indeed, several mutational signatures have been linked to known factors such as UV radiation and smoking<bibr rid=\"b31\"/>.'],\n",
       "    'year': '2013'},\n",
       "   {'ArticleID': 'pub.1009493519',\n",
       "    'CitedArticleID': 'pub.1009493519',\n",
       "    'Cites': 234,\n",
       "    'CitingArticleID': 'pub.1037786036',\n",
       "    'DOI': '10.1016/j.gtc.2009.01.004',\n",
       "    'Issue': '1',\n",
       "    'Journal': 'Gastroenterology Clinics of North America',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 2009,\n",
       "    'Title - matched_from': 'Environmental causes of esophageal cancer',\n",
       "    'Title - matched_to': 'Environmental Causes of Esophageal Cancer',\n",
       "    'Volume': '38',\n",
       "    'clean_text': ['At least 75% of oesophageal cancer, or head and neck cancer, is caused by tobacco and alcohol.',\n",
       "     'More importantly, cancers known to have substantial environmental risk proportions, for example, breast cancer, prostate cancer, colorectal cancer, melanoma, head and neck cancer, oesophageal cancer, cervical cancer, liver cancer and stomach cancer, all harbour large percentages of total extrinsic mutational signatures.'],\n",
       "    'count': 2,\n",
       "    'fnm': \"['F.', 'W.-H.', 'C.', 'S.']\",\n",
       "    'index': 'b22',\n",
       "    'journal': 'Gastroenterology Clin. North Am.',\n",
       "    'snm': \"['Kamangar', 'Chow', 'Abnet', 'Dawsey']\",\n",
       "    'text': ['At least 75% of oesophageal cancer, or head and neck cancer, is caused by tobacco and alcohol<bibr rid=\"b21 b22\"/>.',\n",
       "     'More importantly, cancers known to have substantial environmental risk proportions, for example, breast cancer<bibr rid=\"b15\"/>, prostate cancer<bibr rid=\"b16\"/>, colorectal cancer<bibr rid=\"b18\"/>, melanoma<bibr rid=\"b19\"/>, head and neck cancer<bibr rid=\"b21\"/>, oesophageal cancer<bibr rid=\"b22\"/>, cervical cancer<bibr rid=\"b23\"/>, liver cancer<bibr rid=\"b26\"/> and stomach cancer<bibr rid=\"b27\"/>, all harbour large percentages of total extrinsic mutational signatures.'],\n",
       "    'year': '2009'},\n",
       "   {'ArticleID': 'pub.1035322968',\n",
       "    'CitedArticleID': 'pub.1035322968',\n",
       "    'Cites': 946,\n",
       "    'CitingArticleID': 'pub.1037786036',\n",
       "    'DOI': '10.1055/s-0029-1242458',\n",
       "    'Issue': '04',\n",
       "    'Journal': 'Clinics in Colon and Rectal Surgery',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 2009,\n",
       "    'Title - matched_from': 'Colorectal cancer epidemiology: incidence, mortality, survival, and risk factors',\n",
       "    'Title - matched_to': 'Colorectal Cancer Epidemiology: Incidence, Mortality, Survival, and Risk Factors',\n",
       "    'Volume': '22',\n",
       "    'clean_text': ['Colorectal cancer is a high-incidence cancer that is widely considered to be an environmental disease, with an estimated 75% or more of colorectal cancer risk attributable to diet.'],\n",
       "    'count': 1,\n",
       "    'fnm': \"['F. A.', 'R. P.']\",\n",
       "    'index': 'b17',\n",
       "    'journal': 'Clin. Colon Rectal Surg.',\n",
       "    'snm': \"['Haggar', 'Boushey']\",\n",
       "    'text': ['Colorectal cancer is a high-incidence cancer that is widely considered to be an environmental disease<bibr rid=\"b17\"/>, with an estimated 75% or more of colorectal cancer risk attributable to diet<bibr rid=\"b18\"/>.'],\n",
       "    'year': '2009'},\n",
       "   {'ArticleID': 'pub.1026701420',\n",
       "    'CitedArticleID': 'pub.1026701420',\n",
       "    'Cites': 223,\n",
       "    'CitingArticleID': 'pub.1037786036',\n",
       "    'DOI': '10.1073/pnas.1221068110',\n",
       "    'Issue': '6',\n",
       "    'Journal': 'Proceedings of the National Academy of Sciences',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 2013,\n",
       "    'Title - matched_from': 'Half or more of the somatic mutations in cancers of self-renewing tissues originate prior to tumor initiation',\n",
       "    'Title - matched_to': 'Half or more of the somatic mutations in cancers of self-renewing tissues originate prior to tumor initiation',\n",
       "    'Volume': '110',\n",
       "    'clean_text': ['The somatic mutation rate in tumours is estimated to be 5\\u2009×\\u200910−10 per nucleotide site per cell division.'],\n",
       "    'count': 1,\n",
       "    'fnm': \"['C.', 'B.', 'G.']\",\n",
       "    'index': 'b33',\n",
       "    'journal': 'Proc. Natl Acad. Sci. USA',\n",
       "    'snm': \"['Tomasetti', 'Vogelstein', 'Parmigiani']\",\n",
       "    'text': ['The somatic mutation rate in tumours is estimated to be 5&#x02009;&#x000D7;&#x02009;10<super>&#x02212;10</super> per nucleotide site per cell division<bibr rid=\"b32 b33 b34\"/>.'],\n",
       "    'year': '2013'},\n",
       "   {'ArticleID': 'pub.1003387870',\n",
       "    'CitedArticleID': 'pub.1003387870',\n",
       "    'Cites': 87,\n",
       "    'CitingArticleID': 'pub.1037786036',\n",
       "    'DOI': '10.1111/j.1365-2036.2007.03371.x',\n",
       "    'Issue': '2',\n",
       "    'Journal': 'Alimentary Pharmacology & Therapeutics',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 2007,\n",
       "    'Title - matched_from': 'Review article: nutrition, obesity and colorectal cancer',\n",
       "    'Title - matched_to': 'Review article: nutrition, obesity and colorectal cancer',\n",
       "    'Volume': '26',\n",
       "    'clean_text': ['Colorectal cancer is a high-incidence cancer that is widely considered to be an environmental disease, with an estimated 75% or more of colorectal cancer risk attributable to diet.',\n",
       "     'More importantly, cancers known to have substantial environmental risk proportions, for example, breast cancer, prostate cancer, colorectal cancer, melanoma, head and neck cancer, oesophageal cancer, cervical cancer, liver cancer and stomach cancer, all harbour large percentages of total extrinsic mutational signatures.'],\n",
       "    'count': 2,\n",
       "    'fnm': \"['I. T.', 'E. K.']\",\n",
       "    'index': 'b18',\n",
       "    'journal': 'Aliment. Pharmacol. Ther.',\n",
       "    'snm': \"['Johnson', 'Lund']\",\n",
       "    'text': ['Colorectal cancer is a high-incidence cancer that is widely considered to be an environmental disease<bibr rid=\"b17\"/>, with an estimated 75% or more of colorectal cancer risk attributable to diet<bibr rid=\"b18\"/>.',\n",
       "     'More importantly, cancers known to have substantial environmental risk proportions, for example, breast cancer<bibr rid=\"b15\"/>, prostate cancer<bibr rid=\"b16\"/>, colorectal cancer<bibr rid=\"b18\"/>, melanoma<bibr rid=\"b19\"/>, head and neck cancer<bibr rid=\"b21\"/>, oesophageal cancer<bibr rid=\"b22\"/>, cervical cancer<bibr rid=\"b23\"/>, liver cancer<bibr rid=\"b26\"/> and stomach cancer<bibr rid=\"b27\"/>, all harbour large percentages of total extrinsic mutational signatures.'],\n",
       "    'year': '2007'},\n",
       "   {'ArticleID': 'pub.1024390439',\n",
       "    'CitedArticleID': 'pub.1024390439',\n",
       "    'Cites': 73,\n",
       "    'CitingArticleID': 'pub.1037786036',\n",
       "    'DOI': '10.1179/oeh.2009.15.1.43',\n",
       "    'Issue': '1',\n",
       "    'Journal': 'International Journal of Occupational and Environmental Health',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 2009,\n",
       "    'Title - matched_from': 'State of the evidence: the connection between breast cancer and the environment',\n",
       "    'Title - matched_to': 'State of the Evidence: The Connection Between Breast Cancer and the Environment',\n",
       "    'Volume': '15',\n",
       "    'clean_text': ['Particularly, for breast and prostate cancers, it has long been observed that large international geographical variations exist in their incidence rates (for example, Western Europe has the highest incidence of breast cancer, which is almost 5 times higher than areas such as Eastern Asia or Middle Africa; Australia/New Zealand has the highest incidence of prostate cancer, which is almost 25 times higher than areas such as South-Central Asia), and immigrants moving from countries with lower cancer incidence to countries with higher cancer rates soon acquire the higher risk of their new country.',\n",
       "     'More importantly, cancers known to have substantial environmental risk proportions, for example, breast cancer, prostate cancer, colorectal cancer, melanoma, head and neck cancer, oesophageal cancer, cervical cancer, liver cancer and stomach cancer, all harbour large percentages of total extrinsic mutational signatures.'],\n",
       "    'count': 2,\n",
       "    'fnm': \"['J.', 'N.', 'B.', 'J.', 'M.']\",\n",
       "    'index': 'b15',\n",
       "    'journal': 'Int. J. Occup. Environ. Health',\n",
       "    'snm': \"['Gray', 'Evans', 'Taylor', 'Rizzo', 'Walker']\",\n",
       "    'text': ['Particularly, for breast and prostate cancers, it has long been observed that large international geographical variations exist in their incidence rates (for example, Western Europe has the highest incidence of breast cancer, which is almost 5 times higher than areas such as Eastern Asia or Middle Africa; Australia/New Zealand has the highest incidence of prostate cancer, which is almost 25 times higher than areas such as South-Central Asia)<bibr rid=\"b14\"/>, and immigrants moving from countries with lower cancer incidence to countries with higher cancer rates soon acquire the higher risk of their new country<bibr rid=\"b15 b16\"/>.',\n",
       "     'More importantly, cancers known to have substantial environmental risk proportions, for example, breast cancer<bibr rid=\"b15\"/>, prostate cancer<bibr rid=\"b16\"/>, colorectal cancer<bibr rid=\"b18\"/>, melanoma<bibr rid=\"b19\"/>, head and neck cancer<bibr rid=\"b21\"/>, oesophageal cancer<bibr rid=\"b22\"/>, cervical cancer<bibr rid=\"b23\"/>, liver cancer<bibr rid=\"b26\"/> and stomach cancer<bibr rid=\"b27\"/>, all harbour large percentages of total extrinsic mutational signatures.'],\n",
       "    'year': '2009'},\n",
       "   {'ArticleID': 'pub.1062665194',\n",
       "    'CitedArticleID': 'pub.1062665194',\n",
       "    'Cites': 26,\n",
       "    'CitingArticleID': 'pub.1037786036',\n",
       "    'DOI': '10.1126/science.aaa6094',\n",
       "    'Issue': '6223',\n",
       "    'Journal': 'Science',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 2015,\n",
       "    'Title - matched_from': 'Cancer risk: many factors contribute',\n",
       "    'Title - matched_to': 'Cancer risk: Many factors contribute',\n",
       "    'Volume': '347',\n",
       "    'clean_text': ['Many arguments against the ‘bad luck’ hypothesis have been made, yet none of these have offered specific alternatives to quantitatively evaluate the contribution of extrinsic risk factors in cancer development.'],\n",
       "    'count': 1,\n",
       "    'fnm': \"['M.', 'E. L.']\",\n",
       "    'index': 'b10',\n",
       "    'journal': 'Science',\n",
       "    'snm': \"['Song', 'Giovannucci']\",\n",
       "    'text': ['Many arguments against the &#x02018;bad luck&#x02019; hypothesis have been made<bibr rid=\"b5 b6 b7 b8 b9 b10 b11 b12 b13\"/>, yet none of these have offered specific alternatives to quantitatively evaluate the contribution of extrinsic risk factors in cancer development.'],\n",
       "    'year': '2015'},\n",
       "   {'ArticleID': 'pub.1014822447',\n",
       "    'CitedArticleID': 'pub.1014822447',\n",
       "    'Cites': 451,\n",
       "    'CitingArticleID': 'pub.1037786036',\n",
       "    'DOI': '10.1016/j.critrevonc.2004.04.007',\n",
       "    'Issue': '1',\n",
       "    'Journal': 'Critical Reviews in Oncology/Hematology',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 2004,\n",
       "    'Title - matched_from': 'Stem cell origin of cancer and differentiation therapy',\n",
       "    'Title - matched_to': 'Stem cell origin of cancer and differentiation therapy',\n",
       "    'Volume': '51',\n",
       "    'clean_text': ['\\n'],\n",
       "    'count': 1,\n",
       "    'fnm': \"['S.']\",\n",
       "    'index': 'b1',\n",
       "    'journal': 'Crit. Rev. Oncol. Hematol.',\n",
       "    'snm': \"['Sell']\",\n",
       "    'text': ['\\r\\n</abs>\\r\\n<websumm type=\"toc\">Recent analyses have suggested that the intrinsic behaviour of tissue stem cells may be responsible for malignant transformation and cancer progression, raising questions regarding the influence of extrinsic factors on tumourigenesis; here, both data-driven and model-driven evidence show that such intrinsic risk factors contribute only marginally to cancer development, indicating that cancer risk is heavily influenced by extrinsic factors.</websumm>\\r\\n&e160107-2;</fm>\\r\\n<bdy>\\r\\n<p>Cancers were once thought to originate from mature tissue cells that underwent dedifferentiation in response to cancer progression<bibr rid=\"b1\"/>.'],\n",
       "    'year': '2004'},\n",
       "   {'ArticleID': 'pub.1079706328',\n",
       "    'CitedArticleID': 'pub.1079706328',\n",
       "    'Cites': 1183,\n",
       "    'CitingArticleID': 'pub.1037786036',\n",
       "    'DOI': '',\n",
       "    'Issue': '11',\n",
       "    'Journal': 'Cancer Research',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 1988,\n",
       "    'Title - matched_from': 'Smoking and drinking in relation to oral and pharyngeal cancer',\n",
       "    'Title - matched_to': 'Smoking and drinking in relation to oral and pharyngeal cancer.',\n",
       "    'Volume': '48',\n",
       "    'clean_text': ['At least 75% of oesophageal cancer, or head and neck cancer, is caused by tobacco and alcohol.',\n",
       "     'More importantly, cancers known to have substantial environmental risk proportions, for example, breast cancer, prostate cancer, colorectal cancer, melanoma, head and neck cancer, oesophageal cancer, cervical cancer, liver cancer and stomach cancer, all harbour large percentages of total extrinsic mutational signatures.'],\n",
       "    'count': 2,\n",
       "    'fnm': \"['W. J.']\",\n",
       "    'index': 'b21',\n",
       "    'journal': 'Cancer Res.',\n",
       "    'snm': \"['Blot']\",\n",
       "    'text': ['At least 75% of oesophageal cancer, or head and neck cancer, is caused by tobacco and alcohol<bibr rid=\"b21 b22\"/>.',\n",
       "     'More importantly, cancers known to have substantial environmental risk proportions, for example, breast cancer<bibr rid=\"b15\"/>, prostate cancer<bibr rid=\"b16\"/>, colorectal cancer<bibr rid=\"b18\"/>, melanoma<bibr rid=\"b19\"/>, head and neck cancer<bibr rid=\"b21\"/>, oesophageal cancer<bibr rid=\"b22\"/>, cervical cancer<bibr rid=\"b23\"/>, liver cancer<bibr rid=\"b26\"/> and stomach cancer<bibr rid=\"b27\"/>, all harbour large percentages of total extrinsic mutational signatures.'],\n",
       "    'year': '1988'},\n",
       "   {'ArticleID': 'pub.1006633917',\n",
       "    'CitedArticleID': 'pub.1006633917',\n",
       "    'Cites': 845,\n",
       "    'CitingArticleID': 'pub.1037786036',\n",
       "    'DOI': '10.1038/nature09781',\n",
       "    'Issue': '7330',\n",
       "    'Journal': 'Nature',\n",
       "    'NatureID': 'nature09781',\n",
       "    'PubYear': 2011,\n",
       "    'Title - matched_from': 'Cells of origin in cancer',\n",
       "    'Title - matched_to': 'Cells of origin in cancer',\n",
       "    'Volume': '469',\n",
       "    'clean_text': ['Today, cancers are proposed to originate from the malignant transformation of normal tissue progenitor and stem cells, although this is not wholly accepted.',\n",
       "     'This approach allows for every dividing cell to be a potential cancer-initiating cell, which would be an application of another cell-of-origin theory of cancer whereby tumours may originate from a hierarchy of cells, from stem cells to committed progenitor cells to differentiated cells.'],\n",
       "    'count': 2,\n",
       "    'fnm': \"['J. E.']\",\n",
       "    'index': 'b4',\n",
       "    'journal': 'Nature',\n",
       "    'snm': \"['Visvader']\",\n",
       "    'text': ['Today, cancers are proposed to originate from the malignant transformation of normal tissue progenitor and stem cells<bibr rid=\"b2 b3\"/>, although this is not wholly accepted<bibr rid=\"b4\"/>.',\n",
       "     'This approach allows for every dividing cell to be a potential cancer-initiating cell, which would be an application of another cell-of-origin theory of cancer whereby tumours may originate from a hierarchy of cells, from stem cells to committed progenitor cells to differentiated cells<bibr rid=\"b4\"/>.'],\n",
       "    'year': '2011'},\n",
       "   {'ArticleID': 'pub.1024338922',\n",
       "    'CitedArticleID': 'pub.1024338922',\n",
       "    'Cites': 472,\n",
       "    'CitingArticleID': 'pub.1037786036',\n",
       "    'DOI': '10.1056/nejm199711063371904',\n",
       "    'Issue': '19',\n",
       "    'Journal': 'New England Journal of Medicine',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 1997,\n",
       "    'Title - matched_from': 'Sexually transmitted infection as a cause of anal cancer',\n",
       "    'Title - matched_to': 'Sexually Transmitted Infection as a Cause of Anal Cancer',\n",
       "    'Volume': '337',\n",
       "    'clean_text': ['For instance, human papilloma virus may cause ~90% of cervical cancer cases, ~90% of anal cancer cases and ~70% of oropharyngeal cancer cases; hepatitis B and C may account for ~80% of hepatocellular carcinoma cases; and Helicobacter pylori may be responsible for 65–80% of gastric cancer cases.'],\n",
       "    'count': 1,\n",
       "    'fnm': \"['M.']\",\n",
       "    'index': 'b24',\n",
       "    'journal': 'N. Engl. J. Med.',\n",
       "    'snm': \"['Frisch']\",\n",
       "    'text': ['For instance, human papilloma virus may cause &#x0007E;90% of cervical cancer cases<bibr rid=\"b23\"/>, &#x0007E;90% of anal cancer cases<bibr rid=\"b24\"/> and &#x0007E;70% of oropharyngeal cancer cases<bibr rid=\"b25\"/>; hepatitis B and C may account for &#x0007E;80% of hepatocellular carcinoma cases<bibr rid=\"b26\"/>; and <i>Helicobacter pylori</i> may be responsible for 65&#x02013;80% of gastric cancer cases<bibr rid=\"b27\"/>.'],\n",
       "    'year': '1997'},\n",
       "   {'ArticleID': 'pub.1024281881',\n",
       "    'CitedArticleID': 'pub.1024281881',\n",
       "    'Cites': 90,\n",
       "    'CitingArticleID': 'pub.1037786036',\n",
       "    'DOI': '10.1038/bjc.2011.486',\n",
       "    'Issue': 'Suppl 2',\n",
       "    'Journal': 'British Journal of Cancer',\n",
       "    'NatureID': 'bjc.2011.486',\n",
       "    'PubYear': 2011,\n",
       "    'Title - matched_from': '13. Cancers attributable to solar (ultraviolet) radiation exposure in the UK in 2010',\n",
       "    'Title - matched_to': '13. Cancers attributable to solar (ultraviolet) radiation exposure in the UK in 2010',\n",
       "    'Volume': '105',\n",
       "    'clean_text': ['For example, for melanoma the risk ascribed to sun exposure is around 65–86%, and for non-melanoma basal and squamous skin cancers ~90% is attributable to UV radiation.',\n",
       "     'More importantly, cancers known to have substantial environmental risk proportions, for example, breast cancer, prostate cancer, colorectal cancer, melanoma, head and neck cancer, oesophageal cancer, cervical cancer, liver cancer and stomach cancer, all harbour large percentages of total extrinsic mutational signatures.'],\n",
       "    'count': 2,\n",
       "    'fnm': \"['D. M.', 'D.', 'P.']\",\n",
       "    'index': 'b19',\n",
       "    'journal': 'Br. J. Cancer',\n",
       "    'snm': \"['Parkin', 'Mesher', 'Sasieni']\",\n",
       "    'text': ['For example, for melanoma the risk ascribed to sun exposure is around 65&#x02013;86%<bibr rid=\"b19\"/>, and for non-melanoma basal and squamous skin cancers &#x0007E;90% is attributable to UV radiation<bibr rid=\"b20\"/>.',\n",
       "     'More importantly, cancers known to have substantial environmental risk proportions, for example, breast cancer<bibr rid=\"b15\"/>, prostate cancer<bibr rid=\"b16\"/>, colorectal cancer<bibr rid=\"b18\"/>, melanoma<bibr rid=\"b19\"/>, head and neck cancer<bibr rid=\"b21\"/>, oesophageal cancer<bibr rid=\"b22\"/>, cervical cancer<bibr rid=\"b23\"/>, liver cancer<bibr rid=\"b26\"/> and stomach cancer<bibr rid=\"b27\"/>, all harbour large percentages of total extrinsic mutational signatures.'],\n",
       "    'year': '2011'},\n",
       "   {'ArticleID': 'pub.1049957935',\n",
       "    'CitedArticleID': 'pub.1049957935',\n",
       "    'Cites': 22,\n",
       "    'CitingArticleID': 'pub.1037786036',\n",
       "    'DOI': '10.1126/science.aaa6462',\n",
       "    'Issue': '6223',\n",
       "    'Journal': 'Science',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 2015,\n",
       "    'Title - matched_from': 'Cancer risk: prevention is crucial',\n",
       "    'Title - matched_to': 'Cancer risk: Prevention is crucial',\n",
       "    'Volume': '347',\n",
       "    'clean_text': ['Many arguments against the ‘bad luck’ hypothesis have been made, yet none of these have offered specific alternatives to quantitatively evaluate the contribution of extrinsic risk factors in cancer development.'],\n",
       "    'count': 1,\n",
       "    'fnm': \"['C.', 'T.', 'J.']\",\n",
       "    'index': 'b9',\n",
       "    'journal': 'Science',\n",
       "    'snm': \"['Gotay', 'Dummer', 'Spinelli']\",\n",
       "    'text': ['Many arguments against the &#x02018;bad luck&#x02019; hypothesis have been made<bibr rid=\"b5 b6 b7 b8 b9 b10 b11 b12 b13\"/>, yet none of these have offered specific alternatives to quantitatively evaluate the contribution of extrinsic risk factors in cancer development.'],\n",
       "    'year': '2015'},\n",
       "   {'ArticleID': 'pub.1021613879',\n",
       "    'CitedArticleID': 'pub.1021613879',\n",
       "    'Cites': 1663,\n",
       "    'CitingArticleID': 'pub.1037786036',\n",
       "    'DOI': '10.1053/j.gastro.2011.12.061',\n",
       "    'Issue': '6',\n",
       "    'Journal': 'Gastroenterology',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 2012,\n",
       "    'Title - matched_from': 'Epidemiology of viral hepatitis and hepatocellular carcinoma',\n",
       "    'Title - matched_to': 'Epidemiology of Viral Hepatitis and Hepatocellular Carcinoma',\n",
       "    'Volume': '142',\n",
       "    'clean_text': ['For instance, human papilloma virus may cause ~90% of cervical cancer cases, ~90% of anal cancer cases and ~70% of oropharyngeal cancer cases; hepatitis B and C may account for ~80% of hepatocellular carcinoma cases; and Helicobacter pylori may be responsible for 65–80% of gastric cancer cases.',\n",
       "     'More importantly, cancers known to have substantial environmental risk proportions, for example, breast cancer, prostate cancer, colorectal cancer, melanoma, head and neck cancer, oesophageal cancer, cervical cancer, liver cancer and stomach cancer, all harbour large percentages of total extrinsic mutational signatures.'],\n",
       "    'count': 2,\n",
       "    'fnm': \"['H. B.']\",\n",
       "    'index': 'b26',\n",
       "    'journal': 'Gastroenterology',\n",
       "    'snm': \"['El-Serag']\",\n",
       "    'text': ['For instance, human papilloma virus may cause &#x0007E;90% of cervical cancer cases<bibr rid=\"b23\"/>, &#x0007E;90% of anal cancer cases<bibr rid=\"b24\"/> and &#x0007E;70% of oropharyngeal cancer cases<bibr rid=\"b25\"/>; hepatitis B and C may account for &#x0007E;80% of hepatocellular carcinoma cases<bibr rid=\"b26\"/>; and <i>Helicobacter pylori</i> may be responsible for 65&#x02013;80% of gastric cancer cases<bibr rid=\"b27\"/>.',\n",
       "     'More importantly, cancers known to have substantial environmental risk proportions, for example, breast cancer<bibr rid=\"b15\"/>, prostate cancer<bibr rid=\"b16\"/>, colorectal cancer<bibr rid=\"b18\"/>, melanoma<bibr rid=\"b19\"/>, head and neck cancer<bibr rid=\"b21\"/>, oesophageal cancer<bibr rid=\"b22\"/>, cervical cancer<bibr rid=\"b23\"/>, liver cancer<bibr rid=\"b26\"/> and stomach cancer<bibr rid=\"b27\"/>, all harbour large percentages of total extrinsic mutational signatures.'],\n",
       "    'year': '2012'},\n",
       "   {'ArticleID': 'pub.1021100179',\n",
       "    'CitedArticleID': 'pub.1021100179',\n",
       "    'Cites': 6390,\n",
       "    'CitingArticleID': 'pub.1037786036',\n",
       "    'DOI': '10.1038/35102167',\n",
       "    'Issue': '6859',\n",
       "    'Journal': 'Nature',\n",
       "    'NatureID': '35102167',\n",
       "    'PubYear': 2001,\n",
       "    'Title - matched_from': 'Stem cells, cancer, and cancer stem cells',\n",
       "    'Title - matched_to': 'Stem cells, cancer, and cancer stem cells',\n",
       "    'Volume': '414',\n",
       "    'clean_text': ['Today, cancers are proposed to originate from the malignant transformation of normal tissue progenitor and stem cells, although this is not wholly accepted.'],\n",
       "    'count': 1,\n",
       "    'fnm': \"['T.', 'S. J.', 'M. F.', 'I. L.']\",\n",
       "    'index': 'b2',\n",
       "    'journal': 'Nature',\n",
       "    'snm': \"['Reya', 'Morrison', 'Clarke', 'Weissman']\",\n",
       "    'text': ['Today, cancers are proposed to originate from the malignant transformation of normal tissue progenitor and stem cells<bibr rid=\"b2 b3\"/>, although this is not wholly accepted<bibr rid=\"b4\"/>.'],\n",
       "    'year': '2001'},\n",
       "   {'ArticleID': 'pub.1029275657',\n",
       "    'CitedArticleID': 'pub.1029275657',\n",
       "    'Cites': 51,\n",
       "    'CitingArticleID': 'pub.1037786036',\n",
       "    'DOI': '10.1126/science.aaa6592',\n",
       "    'Issue': '6223',\n",
       "    'Journal': 'Science',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 2015,\n",
       "    'Title - matched_from': 'Cancer risk: role of environment',\n",
       "    'Title - matched_to': 'Cancer risk: Role of environment—Response',\n",
       "    'Volume': '347',\n",
       "    'clean_text': ['However, there has been controversy regarding the conclusion that this correlation implies a very high unavoidable risk for many cancers that is due solely to the intrinsic baseline population size of tissue-specific stem cells.',\n",
       "     'Many arguments against the ‘bad luck’ hypothesis have been made, yet none of these have offered specific alternatives to quantitatively evaluate the contribution of extrinsic risk factors in cancer development.'],\n",
       "    'count': 2,\n",
       "    'fnm': \"['N. A.']\",\n",
       "    'index': 'b6',\n",
       "    'journal': 'Science',\n",
       "    'snm': \"['Ashford']\",\n",
       "    'text': ['However, there has been controversy regarding the conclusion that this correlation implies a very high unavoidable risk for many cancers that is due solely to the intrinsic baseline population size of tissue-specific stem cells<bibr rid=\"b6 b7\"/>.',\n",
       "     'Many arguments against the &#x02018;bad luck&#x02019; hypothesis have been made<bibr rid=\"b5 b6 b7 b8 b9 b10 b11 b12 b13\"/>, yet none of these have offered specific alternatives to quantitatively evaluate the contribution of extrinsic risk factors in cancer development.'],\n",
       "    'year': '2015'},\n",
       "   {'ArticleID': 'pub.1062665200',\n",
       "    'CitedArticleID': 'pub.1062665200',\n",
       "    'Cites': 24,\n",
       "    'CitingArticleID': 'pub.1037786036',\n",
       "    'DOI': '10.1126/science.aaa6212',\n",
       "    'Issue': '6223',\n",
       "    'Journal': 'Science',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 2015,\n",
       "    'Title - matched_from': 'Cancer risk: accuracy of literature',\n",
       "    'Title - matched_to': 'Cancer risk: Accuracy of literature',\n",
       "    'Volume': '347',\n",
       "    'clean_text': ['Many arguments against the ‘bad luck’ hypothesis have been made, yet none of these have offered specific alternatives to quantitatively evaluate the contribution of extrinsic risk factors in cancer development.'],\n",
       "    'count': 1,\n",
       "    'fnm': \"['M.']\",\n",
       "    'index': 'b11',\n",
       "    'journal': 'Science',\n",
       "    'snm': \"['O’Callaghan']\",\n",
       "    'text': ['Many arguments against the &#x02018;bad luck&#x02019; hypothesis have been made<bibr rid=\"b5 b6 b7 b8 b9 b10 b11 b12 b13\"/>, yet none of these have offered specific alternatives to quantitatively evaluate the contribution of extrinsic risk factors in cancer development.'],\n",
       "    'year': '2015'},\n",
       "   {'ArticleID': 'pub.1003288389',\n",
       "    'CitedArticleID': 'pub.1003288389',\n",
       "    'Cites': 70,\n",
       "    'CitingArticleID': 'pub.1037786036',\n",
       "    'DOI': '10.1001/archderm.1996.03890280098014',\n",
       "    'Issue': '4',\n",
       "    'Journal': 'JAMA Dermatology',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 1996,\n",
       "    'Title - matched_from': 'Prevention and early detection strategies for melanoma and skin cancer. Current status',\n",
       "    'Title - matched_to': 'Prevention and Early Detection Strategies for Melanoma and Skin Cancer: Current Status',\n",
       "    'Volume': '132',\n",
       "    'clean_text': ['For example, for melanoma the risk ascribed to sun exposure is around 65–86%, and for non-melanoma basal and squamous skin cancers ~90% is attributable to UV radiation.'],\n",
       "    'count': 1,\n",
       "    'fnm': \"['H. K.', 'A. C.', 'D. R.', 'T. A.', 'R. A.']\",\n",
       "    'index': 'b20',\n",
       "    'journal': 'Arch. Dermatol.',\n",
       "    'snm': \"['Koh', 'Geller', 'Miller', 'Grossbart', 'Lew']\",\n",
       "    'text': ['For example, for melanoma the risk ascribed to sun exposure is around 65&#x02013;86%<bibr rid=\"b19\"/>, and for non-melanoma basal and squamous skin cancers &#x0007E;90% is attributable to UV radiation<bibr rid=\"b20\"/>.'],\n",
       "    'year': '1996'},\n",
       "   {'ArticleID': 'pub.1062665201',\n",
       "    'CitedArticleID': 'pub.1062665201',\n",
       "    'Cites': 32,\n",
       "    'CitingArticleID': 'pub.1037786036',\n",
       "    'DOI': '10.1126/science.aaa6246',\n",
       "    'Issue': '6223',\n",
       "    'Journal': 'Science',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 2015,\n",
       "    'Title - matched_from': 'Cancer risk: role of environment',\n",
       "    'Title - matched_to': 'Cancer risk: Role of environment',\n",
       "    'Volume': '347',\n",
       "    'clean_text': ['However, there has been controversy regarding the conclusion that this correlation implies a very high unavoidable risk for many cancers that is due solely to the intrinsic baseline population size of tissue-specific stem cells.',\n",
       "     'Many arguments against the ‘bad luck’ hypothesis have been made, yet none of these have offered specific alternatives to quantitatively evaluate the contribution of extrinsic risk factors in cancer development.'],\n",
       "    'count': 2,\n",
       "    'fnm': \"['N. A.']\",\n",
       "    'index': 'b6',\n",
       "    'journal': 'Science',\n",
       "    'snm': \"['Ashford']\",\n",
       "    'text': ['However, there has been controversy regarding the conclusion that this correlation implies a very high unavoidable risk for many cancers that is due solely to the intrinsic baseline population size of tissue-specific stem cells<bibr rid=\"b6 b7\"/>.',\n",
       "     'Many arguments against the &#x02018;bad luck&#x02019; hypothesis have been made<bibr rid=\"b5 b6 b7 b8 b9 b10 b11 b12 b13\"/>, yet none of these have offered specific alternatives to quantitatively evaluate the contribution of extrinsic risk factors in cancer development.'],\n",
       "    'year': '2015'},\n",
       "   {'ArticleID': 'pub.1059819308',\n",
       "    'CitedArticleID': 'pub.1059819308',\n",
       "    'Cites': 2490,\n",
       "    'CitingArticleID': 'pub.1037786036',\n",
       "    'DOI': '10.1093/jnci/87.11.796',\n",
       "    'Issue': '11',\n",
       "    'Journal': 'Journal of the National Cancer Institute',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 1995,\n",
       "    'Title - matched_from': 'Prevalence of human papillomavirus in cervical cancer: a worldwide perspective',\n",
       "    'Title - matched_to': 'Prevalence of human papillomavirus in cervical cancer: a worldwide perspective. International biological study on cervical cancer (IBSCC) Study Group.',\n",
       "    'Volume': '87',\n",
       "    'clean_text': ['For instance, human papilloma virus may cause ~90% of cervical cancer cases, ~90% of anal cancer cases and ~70% of oropharyngeal cancer cases; hepatitis B and C may account for ~80% of hepatocellular carcinoma cases; and Helicobacter pylori may be responsible for 65–80% of gastric cancer cases.',\n",
       "     'More importantly, cancers known to have substantial environmental risk proportions, for example, breast cancer, prostate cancer, colorectal cancer, melanoma, head and neck cancer, oesophageal cancer, cervical cancer, liver cancer and stomach cancer, all harbour large percentages of total extrinsic mutational signatures.'],\n",
       "    'count': 2,\n",
       "    'fnm': \"['F. X.']\",\n",
       "    'index': 'b23',\n",
       "    'journal': 'J. Natl. Cancer Inst.',\n",
       "    'snm': \"['Bosch']\",\n",
       "    'text': ['For instance, human papilloma virus may cause &#x0007E;90% of cervical cancer cases<bibr rid=\"b23\"/>, &#x0007E;90% of anal cancer cases<bibr rid=\"b24\"/> and &#x0007E;70% of oropharyngeal cancer cases<bibr rid=\"b25\"/>; hepatitis B and C may account for &#x0007E;80% of hepatocellular carcinoma cases<bibr rid=\"b26\"/>; and <i>Helicobacter pylori</i> may be responsible for 65&#x02013;80% of gastric cancer cases<bibr rid=\"b27\"/>.',\n",
       "     'More importantly, cancers known to have substantial environmental risk proportions, for example, breast cancer<bibr rid=\"b15\"/>, prostate cancer<bibr rid=\"b16\"/>, colorectal cancer<bibr rid=\"b18\"/>, melanoma<bibr rid=\"b19\"/>, head and neck cancer<bibr rid=\"b21\"/>, oesophageal cancer<bibr rid=\"b22\"/>, cervical cancer<bibr rid=\"b23\"/>, liver cancer<bibr rid=\"b26\"/> and stomach cancer<bibr rid=\"b27\"/>, all harbour large percentages of total extrinsic mutational signatures.'],\n",
       "    'year': '1995'},\n",
       "   {'ArticleID': 'pub.1018833150',\n",
       "    'CitedArticleID': 'pub.1018833150',\n",
       "    'Cites': 526,\n",
       "    'CitingArticleID': 'pub.1037786036',\n",
       "    'DOI': '10.1073/pnas.0712345105',\n",
       "    'Issue': '11',\n",
       "    'Journal': 'Proceedings of the National Academy of Sciences',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 2008,\n",
       "    'Title - matched_from': 'Comparative lesion sequencing provides insights into tumor evolution',\n",
       "    'Title - matched_to': 'Comparative lesion sequencing provides insights into tumor evolution',\n",
       "    'Volume': '105',\n",
       "    'clean_text': ['The somatic mutation rate in tumours is estimated to be 5\\u2009×\\u200910−10 per nucleotide site per cell division.'],\n",
       "    'count': 1,\n",
       "    'fnm': \"['S.']\",\n",
       "    'index': 'b32',\n",
       "    'journal': 'Proc. Natl Acad. Sci. USA',\n",
       "    'snm': \"['Jones']\",\n",
       "    'text': ['The somatic mutation rate in tumours is estimated to be 5&#x02009;&#x000D7;&#x02009;10<super>&#x02212;10</super> per nucleotide site per cell division<bibr rid=\"b32 b33 b34\"/>.'],\n",
       "    'year': '2008'},\n",
       "   {'ArticleID': 'pub.1006621195',\n",
       "    'CitedArticleID': 'pub.1006621195',\n",
       "    'Cites': 183,\n",
       "    'CitingArticleID': 'pub.1037786036',\n",
       "    'DOI': '10.1073/pnas.1421839112',\n",
       "    'Issue': '1',\n",
       "    'Journal': 'Proceedings of the National Academy of Sciences',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 2015,\n",
       "    'Title - matched_from': 'Only three driver gene mutations are required for the development of lung and colorectal cancers',\n",
       "    'Title - matched_to': 'Only three driver gene mutations are required for the development of lung and colorectal cancers',\n",
       "    'Volume': '112',\n",
       "    'clean_text': ['4, if we consider the more reasonable case where three mutations are required, k\\u2009=\\u20093, almost all modelled intrinsic risks (both tLIRsc and tLIRtt) drop well below our earlier ‘intrinsic’ risk lines estimated conservatively from the observed data alone (red dashed lines, estimated based on observed data following the same mechanism as Fig.'],\n",
       "    'count': 1,\n",
       "    'fnm': \"['C.', 'L.', 'M. A.', 'G.', 'B.']\",\n",
       "    'index': 'b35',\n",
       "    'journal': 'Proc. Natl Acad. Sci. USA',\n",
       "    'snm': \"['Tomasetti', 'Marchionni', 'Nowak', 'Parmigiani', 'Vogelstein']\",\n",
       "    'text': ['4</figr>, if we consider the more reasonable case where three mutations are required<bibr rid=\"b35\"/>, <i>k</i>&#x02009;=&#x02009;3, almost all modelled intrinsic risks (both tLIRsc and tLIRtt) drop well below our earlier &#x02018;intrinsic&#x02019; risk lines estimated conservatively from the observed data alone (red dashed lines, estimated based on observed data following the same mechanism as <figr rid=\"f3\">Fig.'],\n",
       "    'year': '2015'},\n",
       "   {'ArticleID': 'pub.1013257560',\n",
       "    'CitedArticleID': 'pub.1013257560',\n",
       "    'Cites': 15285,\n",
       "    'CitingArticleID': 'pub.1037786036',\n",
       "    'DOI': '10.3322/caac.21262',\n",
       "    'Issue': '2',\n",
       "    'Journal': 'CA A Cancer Journal for Clinicians',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 2015,\n",
       "    'Title - matched_from': 'Global cancer statistics, 2012',\n",
       "    'Title - matched_to': 'Global cancer statistics, 2012',\n",
       "    'Volume': '65',\n",
       "    'clean_text': ['Particularly, for breast and prostate cancers, it has long been observed that large international geographical variations exist in their incidence rates (for example, Western Europe has the highest incidence of breast cancer, which is almost 5 times higher than areas such as Eastern Asia or Middle Africa; Australia/New Zealand has the highest incidence of prostate cancer, which is almost 25 times higher than areas such as South-Central Asia), and immigrants moving from countries with lower cancer incidence to countries with higher cancer rates soon acquire the higher risk of their new country.'],\n",
       "    'count': 1,\n",
       "    'fnm': \"['L. A.']\",\n",
       "    'index': 'b14',\n",
       "    'journal': 'CA Cancer J. Clin.',\n",
       "    'snm': \"['Torre']\",\n",
       "    'text': ['Particularly, for breast and prostate cancers, it has long been observed that large international geographical variations exist in their incidence rates (for example, Western Europe has the highest incidence of breast cancer, which is almost 5 times higher than areas such as Eastern Asia or Middle Africa; Australia/New Zealand has the highest incidence of prostate cancer, which is almost 25 times higher than areas such as South-Central Asia)<bibr rid=\"b14\"/>, and immigrants moving from countries with lower cancer incidence to countries with higher cancer rates soon acquire the higher risk of their new country<bibr rid=\"b15 b16\"/>.'],\n",
       "    'year': '2015'},\n",
       "   {'ArticleID': 'pub.1025675859',\n",
       "    'CitedArticleID': 'pub.1025675859',\n",
       "    'Cites': 473,\n",
       "    'CitingArticleID': 'pub.1037786036',\n",
       "    'DOI': '10.1016/j.ccm.2011.09.001',\n",
       "    'Issue': '4',\n",
       "    'Journal': 'Clinics in Chest Medicine',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 2011,\n",
       "    'Title - matched_from': 'Lung cancer: epidemiology, etiology, and prevention',\n",
       "    'Title - matched_to': 'Lung Cancer: Epidemiology, Etiology, and Prevention',\n",
       "    'Volume': '32',\n",
       "    'clean_text': [],\n",
       "    'count': 0,\n",
       "    'fnm': \"['C. S.', 'L. T.', 'R. A.']\",\n",
       "    'index': 'b29',\n",
       "    'journal': 'Clin. Chest Med.',\n",
       "    'snm': \"['Dela Cruz', 'Tanoue', 'Matthay']\",\n",
       "    'text': '',\n",
       "    'year': '2011'},\n",
       "   {'ArticleID': 'pub.1053009387',\n",
       "    'CitedArticleID': 'pub.1053009387',\n",
       "    'Cites': 1185,\n",
       "    'CitingArticleID': 'pub.1037786036',\n",
       "    'DOI': '10.1038/255197a0',\n",
       "    'Issue': '5505',\n",
       "    'Journal': 'Nature',\n",
       "    'NatureID': '255197a0',\n",
       "    'PubYear': 1975,\n",
       "    'Title - matched_from': 'Mutation selection and the natural history of cancer',\n",
       "    'Title - matched_to': 'Mutation selection and the natural history of cancer',\n",
       "    'Volume': '255',\n",
       "    'clean_text': ['Today, cancers are proposed to originate from the malignant transformation of normal tissue progenitor and stem cells, although this is not wholly accepted.'],\n",
       "    'count': 1,\n",
       "    'fnm': \"['J.']\",\n",
       "    'index': 'b3',\n",
       "    'journal': 'Nature',\n",
       "    'snm': \"['Cairns']\",\n",
       "    'text': ['Today, cancers are proposed to originate from the malignant transformation of normal tissue progenitor and stem cells<bibr rid=\"b2 b3\"/>, although this is not wholly accepted<bibr rid=\"b4\"/>.'],\n",
       "    'year': '1975'},\n",
       "   {'ArticleID': 'pub.1016140589',\n",
       "    'CitedArticleID': 'pub.1016140589',\n",
       "    'Cites': 918,\n",
       "    'CitingArticleID': 'pub.1037786036',\n",
       "    'DOI': '10.1126/science.1260825',\n",
       "    'Issue': '6217',\n",
       "    'Journal': 'Science',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 2015,\n",
       "    'Title - matched_from': 'Variation in cancer risk among tissues can be explained by the number of stem cell divisions',\n",
       "    'Title - matched_to': 'Variation in cancer risk among tissues can be explained by the number of stem cell divisions',\n",
       "    'Volume': '347',\n",
       "    'clean_text': ['Nevertheless, recent research has highlighted a strong correlation of 0.81 between tissue-specific cancer risk and the lifetime population size in cumulative number of cell divisions of tissue-specific stem cells.',\n",
       "     'Many arguments against the ‘bad luck’ hypothesis have been made, yet none of these have offered specific alternatives to quantitatively evaluate the contribution of extrinsic risk factors in cancer development.',\n",
       "     'However, if we conduct regression analyses on either the new hypothetical cancer risks or the current cancer risks as reported, against the number of stem-cell divisions, the correlations from both cases would be 0.81 (Fig.',\n",
       "     'In a data-driven approach, we first re-examined the quantitative relationship between the observed lifetime cancer risk and the divisions of the normal tissue stem cells as reported, with a distinct alternative method.',\n",
       "     'It is important to note that here we included breast and prostate cancers—two high-incidence cancers missing in the original stem-cell analysis.',\n",
       "     'To overcome the limitation of inaccurate estimation in the reported stem-cell numbers, we calculated tLIR using both the reported stem-cell number (tLIRsc) and the total tissue cell number (tLIRtt).',\n",
       "     'Information on the total rounds of symmetric and asymmetric divisions as well as the total number of stem cells in the steady state for various tissues discussed in this work has been extracted from supplementary table 1 of Tomasetti & Vogelstein.',\n",
       "     '\\n',\n",
       "     '\\n',\n",
       "     '',\n",
       "     ''],\n",
       "    'count': 11,\n",
       "    'fnm': \"['C.', 'B.']\",\n",
       "    'index': 'b5',\n",
       "    'journal': 'Science',\n",
       "    'snm': \"['Tomasetti', 'Vogelstein']\",\n",
       "    'text': ['Nevertheless, recent research has highlighted a strong correlation of 0.81 between tissue-specific cancer risk and the lifetime population size in cumulative number of cell divisions of tissue-specific stem cells<bibr rid=\"b5\"/>.',\n",
       "     'Many arguments against the &#x02018;bad luck&#x02019; hypothesis have been made<bibr rid=\"b5 b6 b7 b8 b9 b10 b11 b12 b13\"/>, yet none of these have offered specific alternatives to quantitatively evaluate the contribution of extrinsic risk factors in cancer development.',\n",
       "     'However, if we conduct regression analyses on either the new hypothetical cancer risks or the current cancer risks as reported, against the number of stem-cell divisions<bibr rid=\"b5\"/>, the correlations from both cases would be 0.81 (<figr rid=\"f2\">Fig.',\n",
       "     'In a data-driven approach, we first re-examined the quantitative relationship between the observed lifetime cancer risk and the divisions of the normal tissue stem cells as reported<bibr rid=\"b5\"/>, with a distinct alternative method.',\n",
       "     'It is important to note that here we included breast and prostate cancers&#x02014;two high-incidence cancers missing in the original stem-cell analysis<bibr rid=\"b5\"/>.',\n",
       "     'To overcome the limitation of inaccurate estimation in the reported stem-cell numbers<bibr rid=\"b5\"/>, we calculated tLIR using both the reported stem-cell number (tLIRsc) and the total tissue cell number (tLIRtt).',\n",
       "     'Information on the total rounds of symmetric and asymmetric divisions as well as the total number of stem cells in the steady state for various tissues discussed in this work has been extracted from supplementary table 1 of Tomasetti &amp; Vogelstein<bibr rid=\"b5\"/>.',\n",
       "     '\\r\\n</sec>\\r\\n<sec level=\"2\">\\r\\n<sectitle>Estimating cancer risk for different tissues</sectitle>\\r\\n<p>The rounds of symmetric and asymmetric divisions for different tissues were adopted from supplementary table 1 of Tomasetti &amp; Vogelstein<bibr rid=\"b5\"/>.',\n",
       "     '\\r\\n</sec>\\r\\n<sec level=\"2\">\\r\\n<sectitle>Data and statistical analysis</sectitle>\\r\\n<p>The observed lifetime cancer risks and the cumulative number of divisions (<i>n</i>) of all stem cells per lifetime are adopted from supplementary table 1 of Tomasetti &amp; Vogelstein<bibr rid=\"b5\"/>.',\n",
       "     '</caption></fig>\\r\\n<fig id=\"f2\" type=\"jpeg\" entref=\"figf2\"><slide entref=\"slidepf2\"/><figtl>Correlation analysis of stem-cell division and cancer risk does not distinguish contribution of extrinsic versus intrinsic factors to cancer risk.</figtl><caption><p>The black dots are data from figure 1(also shown in supplementary table 1) of Tomasetti &amp; Vogelstein<bibr rid=\"b5\"/>, and the black line shows their original regression line.',\n",
       "     '</caption></fig>\\r\\n<fig id=\"f3\" type=\"jpeg\" entref=\"figf3\"><slide entref=\"slidepf3\"/><figtl>Estimation of the proportion of lifetime cancer risk that is not due entirely to &#x02018;bad luck&#x02019;.</figtl><caption><p><b>a</b>, <b>b</b>, Estimations based on total tissue stem-cell divisions originally reported in Tomasetti &amp; Vogelstein<bibr rid=\"b5\"/> (<b>a</b>) and total tissue cell divisions (<b>b</b>).'],\n",
       "    'year': '2015'},\n",
       "   {'ArticleID': 'pub.1062665363',\n",
       "    'CitedArticleID': 'pub.1062665363',\n",
       "    'Cites': 31,\n",
       "    'CitingArticleID': 'pub.1037786036',\n",
       "    'DOI': '10.1126/science.aaa6799',\n",
       "    'Issue': '6223',\n",
       "    'Journal': 'Science',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 2015,\n",
       "    'Title - matched_from': 'Cancer risk: role of chance overstated',\n",
       "    'Title - matched_to': 'Cancer risk: Role of chance overstated',\n",
       "    'Volume': '347',\n",
       "    'clean_text': ['However, there has been controversy regarding the conclusion that this correlation implies a very high unavoidable risk for many cancers that is due solely to the intrinsic baseline population size of tissue-specific stem cells.',\n",
       "     'Many arguments against the ‘bad luck’ hypothesis have been made, yet none of these have offered specific alternatives to quantitatively evaluate the contribution of extrinsic risk factors in cancer development.'],\n",
       "    'count': 2,\n",
       "    'fnm': \"['C.']\",\n",
       "    'index': 'b7',\n",
       "    'journal': 'Science',\n",
       "    'snm': \"['Wild']\",\n",
       "    'text': ['However, there has been controversy regarding the conclusion that this correlation implies a very high unavoidable risk for many cancers that is due solely to the intrinsic baseline population size of tissue-specific stem cells<bibr rid=\"b6 b7\"/>.',\n",
       "     'Many arguments against the &#x02018;bad luck&#x02019; hypothesis have been made<bibr rid=\"b5 b6 b7 b8 b9 b10 b11 b12 b13\"/>, yet none of these have offered specific alternatives to quantitatively evaluate the contribution of extrinsic risk factors in cancer development.'],\n",
       "    'year': '2015'},\n",
       "   {'ArticleID': 'pub.1016297922',\n",
       "    'CitedArticleID': 'pub.1016297922',\n",
       "    'Cites': 671,\n",
       "    'CitingArticleID': 'pub.1037786036',\n",
       "    'DOI': '10.1136/gut.49.3.347',\n",
       "    'Issue': '3',\n",
       "    'Journal': 'Gut',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 2001,\n",
       "    'Title - matched_from': 'Gastric cancer and ',\n",
       "    'Title - matched_to': 'Gastric cancer and Helicobacter pylori: a combined analysis of 12 case control studies nested within prospective cohorts',\n",
       "    'Volume': '49',\n",
       "    'clean_text': ['For instance, human papilloma virus may cause ~90% of cervical cancer cases, ~90% of anal cancer cases and ~70% of oropharyngeal cancer cases; hepatitis B and C may account for ~80% of hepatocellular carcinoma cases; and Helicobacter pylori may be responsible for 65–80% of gastric cancer cases.',\n",
       "     'More importantly, cancers known to have substantial environmental risk proportions, for example, breast cancer, prostate cancer, colorectal cancer, melanoma, head and neck cancer, oesophageal cancer, cervical cancer, liver cancer and stomach cancer, all harbour large percentages of total extrinsic mutational signatures.'],\n",
       "    'count': 2,\n",
       "    'fnm': '[]',\n",
       "    'index': 'b27',\n",
       "    'journal': 'Gut',\n",
       "    'snm': '[]',\n",
       "    'text': ['For instance, human papilloma virus may cause &#x0007E;90% of cervical cancer cases<bibr rid=\"b23\"/>, &#x0007E;90% of anal cancer cases<bibr rid=\"b24\"/> and &#x0007E;70% of oropharyngeal cancer cases<bibr rid=\"b25\"/>; hepatitis B and C may account for &#x0007E;80% of hepatocellular carcinoma cases<bibr rid=\"b26\"/>; and <i>Helicobacter pylori</i> may be responsible for 65&#x02013;80% of gastric cancer cases<bibr rid=\"b27\"/>.',\n",
       "     'More importantly, cancers known to have substantial environmental risk proportions, for example, breast cancer<bibr rid=\"b15\"/>, prostate cancer<bibr rid=\"b16\"/>, colorectal cancer<bibr rid=\"b18\"/>, melanoma<bibr rid=\"b19\"/>, head and neck cancer<bibr rid=\"b21\"/>, oesophageal cancer<bibr rid=\"b22\"/>, cervical cancer<bibr rid=\"b23\"/>, liver cancer<bibr rid=\"b26\"/> and stomach cancer<bibr rid=\"b27\"/>, all harbour large percentages of total extrinsic mutational signatures.'],\n",
       "    'year': '2001'},\n",
       "   {'ArticleID': 'pub.1044347602',\n",
       "    'CitedArticleID': 'pub.1044347602',\n",
       "    'Cites': 17,\n",
       "    'CitingArticleID': 'pub.1037786036',\n",
       "    'DOI': '10.1126/science.1247887',\n",
       "    'Issue': '6161',\n",
       "    'Journal': 'Science',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 2013,\n",
       "    'Title - matched_from': 'Unwanted evolution',\n",
       "    'Title - matched_to': 'Unwanted Evolution',\n",
       "    'Volume': '342',\n",
       "    'clean_text': ['The somatic mutation rate in tumours is estimated to be 5\\u2009×\\u200910−10 per nucleotide site per cell division.'],\n",
       "    'count': 1,\n",
       "    'fnm': \"['I.', 'M. A.']\",\n",
       "    'index': 'b34',\n",
       "    'journal': 'Science',\n",
       "    'snm': \"['Bozic', 'Nowak']\",\n",
       "    'text': ['The somatic mutation rate in tumours is estimated to be 5&#x02009;&#x000D7;&#x02009;10<super>&#x02212;10</super> per nucleotide site per cell division<bibr rid=\"b32 b33 b34\"/>.'],\n",
       "    'year': '2013'},\n",
       "   {'ArticleID': 'pub.1062665355',\n",
       "    'CitedArticleID': 'pub.1062665355',\n",
       "    'Cites': 28,\n",
       "    'CitingArticleID': 'pub.1037786036',\n",
       "    'DOI': '10.1126/science.aaa6507',\n",
       "    'Issue': '6223',\n",
       "    'Journal': 'Science',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 2015,\n",
       "    'Title - matched_from': 'Cancer risk: tumors excluded',\n",
       "    'Title - matched_to': 'Cancer risk: Tumors excluded',\n",
       "    'Volume': '347',\n",
       "    'clean_text': ['Many arguments against the ‘bad luck’ hypothesis have been made, yet none of these have offered specific alternatives to quantitatively evaluate the contribution of extrinsic risk factors in cancer development.'],\n",
       "    'count': 1,\n",
       "    'fnm': \"['J. D.', 'R. L.']\",\n",
       "    'index': 'b8',\n",
       "    'journal': 'Science',\n",
       "    'snm': \"['Potter', 'Prentice']\",\n",
       "    'text': ['Many arguments against the &#x02018;bad luck&#x02019; hypothesis have been made<bibr rid=\"b5 b6 b7 b8 b9 b10 b11 b12 b13\"/>, yet none of these have offered specific alternatives to quantitatively evaluate the contribution of extrinsic risk factors in cancer development.'],\n",
       "    'year': '2015'},\n",
       "   {'ArticleID': 'pub.1022798904',\n",
       "    'CitedArticleID': 'pub.1022798904',\n",
       "    'Cites': 207,\n",
       "    'CitingArticleID': 'pub.1037786036',\n",
       "    'DOI': '10.1016/j.gde.2013.11.014',\n",
       "    'Issue': '100',\n",
       "    'Journal': 'Current Opinion in Genetics & Development',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 2014,\n",
       "    'Title - matched_from': 'Mutational signatures: the patterns of somatic mutations hidden in cancer genomes',\n",
       "    'Title - matched_to': 'Mutational signatures: the patterns of somatic mutations hidden in cancer genomes',\n",
       "    'Volume': '24',\n",
       "    'clean_text': ['These are regarded as ‘fingerprints’ left on cancer genomes by different mutagenic processes, revealing ~30 distinct signatures among various cancers.'],\n",
       "    'count': 1,\n",
       "    'fnm': \"['L. B.', 'M. R.']\",\n",
       "    'index': 'b30',\n",
       "    'journal': 'Curr. Opin. Genet. Dev.',\n",
       "    'snm': \"['Alexandrov', 'Stratton']\",\n",
       "    'text': ['These are regarded as &#x02018;fingerprints&#x02019; left on cancer genomes by different mutagenic processes<bibr rid=\"b30\"/>, revealing &#x0007E;30 distinct signatures among various cancers<bibr rid=\"b31\"/>.'],\n",
       "    'year': '2014'},\n",
       "   {'ArticleID': 'pub.1054099883',\n",
       "    'CitedArticleID': 'pub.1054099883',\n",
       "    'Cites': 37,\n",
       "    'CitingArticleID': 'pub.1037786036',\n",
       "    'DOI': '10.1001/archderm.132.4.436',\n",
       "    'Issue': '4',\n",
       "    'Journal': 'JAMA Dermatology',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 1996,\n",
       "    'Title - matched_from': 'Prevention and early detection strategies for melanoma and skin cancer. Current status',\n",
       "    'Title - matched_to': 'Prevention and early detection strategies for melanoma and skin cancer. Current status.',\n",
       "    'Volume': '132',\n",
       "    'clean_text': ['For example, for melanoma the risk ascribed to sun exposure is around 65–86%, and for non-melanoma basal and squamous skin cancers ~90% is attributable to UV radiation.'],\n",
       "    'count': 1,\n",
       "    'fnm': \"['H. K.', 'A. C.', 'D. R.', 'T. A.', 'R. A.']\",\n",
       "    'index': 'b20',\n",
       "    'journal': 'Arch. Dermatol.',\n",
       "    'snm': \"['Koh', 'Geller', 'Miller', 'Grossbart', 'Lew']\",\n",
       "    'text': ['For example, for melanoma the risk ascribed to sun exposure is around 65&#x02013;86%<bibr rid=\"b19\"/>, and for non-melanoma basal and squamous skin cancers &#x0007E;90% is attributable to UV radiation<bibr rid=\"b20\"/>.'],\n",
       "    'year': '1996'},\n",
       "   {'ArticleID': 'pub.1005834376',\n",
       "    'CitedArticleID': 'pub.1005834376',\n",
       "    'Cites': 1836,\n",
       "    'CitingArticleID': 'pub.1037786036',\n",
       "    'DOI': '10.1200/jco.2011.36.4596',\n",
       "    'Issue': '32',\n",
       "    'Journal': 'Journal of Clinical Oncology',\n",
       "    'NatureID': '',\n",
       "    'PubYear': 2011,\n",
       "    'Title - matched_from': 'Human papillomavirus and rising oropharyngeal cancer incidence in the United States',\n",
       "    'Title - matched_to': 'Human Papillomavirus and Rising Oropharyngeal Cancer Incidence in the United States',\n",
       "    'Volume': '29',\n",
       "    'clean_text': ['For instance, human papilloma virus may cause ~90% of cervical cancer cases, ~90% of anal cancer cases and ~70% of oropharyngeal cancer cases; hepatitis B and C may account for ~80% of hepatocellular carcinoma cases; and Helicobacter pylori may be responsible for 65–80% of gastric cancer cases.'],\n",
       "    'count': 1,\n",
       "    'fnm': \"['A. K.']\",\n",
       "    'index': 'b25',\n",
       "    'journal': 'J. Clin. Oncol.',\n",
       "    'snm': \"['Chaturvedi']\",\n",
       "    'text': ['For instance, human papilloma virus may cause &#x0007E;90% of cervical cancer cases<bibr rid=\"b23\"/>, &#x0007E;90% of anal cancer cases<bibr rid=\"b24\"/> and &#x0007E;70% of oropharyngeal cancer cases<bibr rid=\"b25\"/>; hepatitis B and C may account for &#x0007E;80% of hepatocellular carcinoma cases<bibr rid=\"b26\"/>; and <i>Helicobacter pylori</i> may be responsible for 65&#x02013;80% of gastric cancer cases<bibr rid=\"b27\"/>.'],\n",
       "    'year': '2011'}]},\n",
       " 'text': '\\nCancers were once thought to originate from mature tissue cells that underwent dedifferentiation in response to cancer progression. Today, cancers are proposed to originate from the malignant transformation of normal tissue progenitor and stem cells, although this is not wholly accepted. Nevertheless, recent research has highlighted a strong correlation of 0.81 between tissue-specific cancer risk and the lifetime population size in cumulative number of cell divisions of tissue-specific stem cells. However, there has been controversy regarding the conclusion that this correlation implies a very high unavoidable risk for many cancers that is due solely to the intrinsic baseline population size of tissue-specific stem cells. Many arguments against the ‘bad luck’ hypothesis have been made, yet none of these have offered specific alternatives to quantitatively evaluate the contribution of extrinsic risk factors in cancer development. Applying several distinct modelling approaches, here we provide strong evidence that unavoidable intrinsic risk factors contribute only modestly (less than ~10–30%) to the development of many common cancers.\\nWe made the conservative and yet conventional assumption that errors occurring during the division of cells, being routes of malignant transformation, can be influenced by both intrinsic processes as well as extrinsic factors (Fig. 1). ‘Intrinsic processes’ include those that result in mutations due to random errors in DNA replication, whereas ‘extrinsic factors’ are environmental factors that affect mutagenesis rates (such as ultraviolet (UV) radiation, ionizing radiation and carcinogens). For example, radiation can cause DNA damage, which would primarily result in deleterious mutations with functional consequences on cancer development only after cell division. Therefore, extrinsic factors may act through the accumulation of genetic alterations during cell division to increase cancer risk. Accordingly, cancer risk would result from those apparently uncontrollable intrinsic processes (Fig. 1, arrow 1) as well as from those highly modifiable and thus preventable extrinsic factors (Fig. 1, arrow 2).\\n\\nCorrelation cannot differentiate risks\\nAccording to the above hypothesis, both intrinsic and extrinsic factors can impart cancer risk through the accumulation of these errors, especially the ‘driver mutations’ (Fig. 1, arrow 3). As such, a correlational analysis between cancer risk and cell division, for either stem or non-stem cells, is unable to differentiate between the contributions of intrinsic and extrinsic factors. This is best illustrated through a thought experiment where we consider a hypothetical scenario of a sudden global emergence of a very potent mutagen, such as a strong radiation burst from a nuclear fallout, which quadruples the lifetime risks for all cancers. In this scenario, it transpires that the proportion of cancer risk caused by intrinsic random errors would be small (at most one-quarter if we assume all of the original risk was due to intrinsic processes). However, if we conduct regression analyses on either the new hypothetical cancer risks or the current cancer risks as reported, against the number of stem-cell divisions, the correlations from both cases would be 0.81 (Fig. 2). This thought experiment negates the ability of the correlation to detect solely the contribution of intrinsic factors as it cannot distinguish between intrinsic and extrinsic factors. Thus, it argues against the implication that around two-thirds of variation could be explained by division-related random intrinsic errors.\\n\\n\\nLower bound intrinsic risk line\\nThe above conclusion then raises the question of what proportion of total cancer risk is due to extrinsic versus intrinsic factors. In a data-driven approach, we first re-examined the quantitative relationship between the observed lifetime cancer risk and the divisions of the normal tissue stem cells as reported, with a distinct alternative method. Our rationale was that intrinsic risk, or indeed its upper bound, can be better estimated by the lowest boundary on the plots of cancer risk versus total tissue stem-cell divisions (Fig. 3a, red ‘intrinsic’ risk line), meaning that intrinsic cancer risk should be determined by the cancer incidence for those cancers with the least risk in the entire group controlling for total stem-cell divisions (Fig. 3a, red dots). The argument here is that cancers with the same number of stem-cell divisions should share the same base of intrinsic cancer risk (if the relationship is causal); if one or more cancers would feature a much higher cancer incidence, for example, lung cancer among smokers versus non-smokers, then this probably reflects additional (and probably extrinsic) risk factors (smoking in this case). One could argue that the low-incidence tumour types may have lower incidences because of additional genetic repair mechanisms that restrict evolving malignant cells from accumulating sufficient numbers of genetic alterations required to become fully tumorigenic; however, without more specific data on the operation of repair mechanisms, these could drive the risk up or down, depending on whether they are less or more efficient in any particular tissue. According to our hypothesis, intrinsic risk from stem-cell divisions would define the lowest bound for a given number of stem-cell divisions, therefore we define an ‘intrinsic’ risk line for stem-cell divisions by regressing the smallest cancer risks on any given number of stem-cell divisions (Fig. 3a, red line). The ‘intrinsic’ risk lines themselves are still probably overestimates for the intrinsic risk; however, we should suspect that any cancer risk above that line implies additional biologic determinants, on the basis of which we can compute the percentage of cancer risk not explained by intrinsic ‘randomness’. As shown in Fig. 3a, most cancer types have very high excess risks relative to the ‘intrinsic’ risk line, indicating large proportions of risks that are unaccounted for by the intrinsic factors, typically larger than 90%. Moreover, these estimated excess risks are very robust: with plausible measurement errors added to the total stem-cell divisions, the resulting excess risks remain essentially intact (Extended Data Table 1).\\n\\n\\nExtrinsic risks by tissue cell turnover\\nAlthough we performed the initial analysis from a ‘stem-cell theory’ point of view, we wanted to evaluate if our results are dependent on this specific theory or independent of it. Furthermore, the lack of reliable data on human tissue stem-cell dynamics is a notable concern (see Supplementary Information), rendering the analysis in Fig. 3a less determinate. Thus, we separately collected data for the total number of tissue cell divisions that is based on homeostatic tissue cell numbers and their turnover rates (see Supplementary Information), and analysed the relationship of cancer risk versus total tissue cell divisions (Fig. 3b). This approach allows for every dividing cell to be a potential cancer-initiating cell, which would be an application of another cell-of-origin theory of cancer whereby tumours may originate from a hierarchy of cells, from stem cells to committed progenitor cells to differentiated cells. Mathematically, this can also be considered as an extreme form of stem-cell theory where the fraction of stem cells is 1 (this latter formulation then provides an upper bound of the effects of the size of the stem-cell population on cancer risk and the role of extrinsic factors). The regression analysis between cancer risk and total tissue cell division shows a high correlation of 0.75, establishing a strong quantitative relationship between cancer risk and total cell division. To dissect the extrinsic versus intrinsic risks, we applied the same rationale and regressed the smallest cancer risks on any given number of cell divisions (Fig. 3b, red line). Although we could only find reliable turnover data for a subset of tissues, it is remarkable that the conclusion drawn here is nearly identical to that in Fig. 3a; that is, large proportions of risks that may not be attributable to intrinsic factors are mostly higher than 90%. It is important to note that here we included breast and prostate cancers—two high-incidence cancers missing in the original stem-cell analysis. Again, plausible measurement errors have been added to the total cell divisions, and the excess risks remained almost identical (Extended Data Table 1). In summary, irrespective of whether a subpopulation or all dividing cells contribute to cancer, these results indicate that intrinsic factors do not play a major causal role.\\n\\n\\nEpidemiological evidence\\nIn parallel, numerous epidemiological studies have established strong evidence that many cancers have substantial risk proportions attributed to environmental exposures (Extended Data Table 2). Particularly, for breast and prostate cancers, it has long been observed that large international geographical variations exist in their incidence rates (for example, Western Europe has the highest incidence of breast cancer, which is almost 5 times higher than areas such as Eastern Asia or Middle Africa; Australia/New Zealand has the highest incidence of prostate cancer, which is almost 25 times higher than areas such as South-Central Asia), and immigrants moving from countries with lower cancer incidence to countries with higher cancer rates soon acquire the higher risk of their new country. While several risk factors have been identified for these cancers, no single one can account for their substantial extrinsic risk proportions, suggesting complex mechanisms for their aetiologies. Colorectal cancer is a high-incidence cancer that is widely considered to be an environmental disease, with an estimated 75% or more of colorectal cancer risk attributable to diet. For many other cancers, known environmental risk factors have also been identified. For example, for melanoma the risk ascribed to sun exposure is around 65–86%, and for non-melanoma basal and squamous skin cancers ~90% is attributable to UV radiation. At least 75% of oesophageal cancer, or head and neck cancer, is caused by tobacco and alcohol. It is also well known that certain pathogens may markedly increase the risk of cancers. For instance, human papilloma virus may cause ~90% of cervical cancer cases, ~90% of anal cancer cases and ~70% of oropharyngeal cancer cases; hepatitis B and C may account for ~80% of hepatocellular carcinoma cases; and Helicobacter pylori may be responsible for 65–80% of gastric cancer cases. These, along with many other reports, provide direct evidence that environmental factors play important roles in cancer incidence and they are modifiable through lifestyle changes and/or vaccinations.\\nAdditionally, analyses of data from the Surveillance, Epidemiology, and End Results Program (SEER) in the USA between 1973–2012 demonstrate that while many cancers have declining or maintain relatively consistent age-adjusted incidence rates (for example, cervical, gallbladder and oesophageal cancers, Extended Data Fig. 1), incidences of some cancers (including melanoma, thyroid, kidney, liver, thymus, small intestine, extranodal non-Hodgkin lymphoma, testicular, anal and anorectal cancers) have been steadily increasing, and their current incidences are substantially higher than their historical minima in the past 40 years (Extended Data Fig. 1). Moreover, the mortality trend of lung cancer from 1930–2011 (ref. 29), which usually mirrors its incidence trend, shows a more than 15-fold increase for lung cancer risk. These substantial increases in incidence suggest that large risk proportions are attributable to changing environments (for example, smoking and air pollutants and their role in the risk of developing lung cancer). Collectively, nearly all major cancers have been covered in these epidemiological studies, further supporting the hypothesis of substantial extrinsic risks for most cancers. Notably, most of these cancers from the epidemiological and SEER results, except for small intestine, are located above the red ‘intrinsic’ risk lines in Fig. 3a, b (blue points). Accounting for the external factors would move them closer to the proposed ‘intrinsic’ line, further supporting the conjecture that the intrinsic line is mainly defined by cancers without compelling known epidemiological risk, whereas those above are at higher risks owing to extrinsic factors.\\n\\n\\nAnalysis of mutational signatures\\nIn addition to epidemiological studies, we evaluated recent studies on mutational signatures in cancer. These are regarded as ‘fingerprints’ left on cancer genomes by different mutagenic processes, revealing ~30 distinct signatures among various cancers. Analysis of these signatures was therefore used to shed light on the proportion of intrinsic versus extrinsic origins of cancer. Two signature mutations, 1A/1B (see ref. 31), demonstrated strong positive correlations with age in the majority of cancers, suggesting that they are acquired at a relatively constant rate over the lifetime of cancer patients and thus probably result from intrinsic processes; however, all other signature mutations (~30) lack the consistent correlations with age, suggesting that they are acquired at different rates in life and thus are probably a consequence of extrinsic carcinogen exposures. Indeed, several mutational signatures have been linked to known factors such as UV radiation and smoking. We therefore categorized the signatures into intrinsic (type 1A/1B) and extrinsic mutations with known or unknown factors, and summarized their corresponding percentages in Extended Data Table 3. Notably, many cancers have substantial extrinsic mutations with known factors. More importantly, cancers known to have substantial environmental risk proportions, for example, breast cancer, prostate cancer, colorectal cancer, melanoma, head and neck cancer, oesophageal cancer, cervical cancer, liver cancer and stomach cancer, all harbour large percentages of total extrinsic mutational signatures. This suggests that the percentages of total extrinsic mutational signatures can serve as a good surrogate for extrinsic cancer risks. While a few cancers have relatively large proportions of intrinsic mutations (>50%), the majority of cancers have large proportions of extrinsic mutations, for example, ~100% for myeloma, lung and thyroid cancers and ~80–90% for bladder, colorectal and uterine cancers, indicating substantial contributions of carcinogen exposures in the development of most cancers.\\n\\n\\nModelling theoretical lifetime intrinsic risk\\nFinally, in another independent model-driven approach to dissecting the risk contribution of the intrinsic processes, we modelled the potential lifetime cancer risk due to intrinsic stem-cell mutation errors by varying the number of hits (that is, driver gene mutations), denoted by k, required for cancer onset. We derived the probability distribution of the propagation of driver gene mutations from one generation to the next, and subsequently established the theoretical relationship between cell divisions and the degree of lifetime cancer risk due to intrinsic cell mutation errors alone, which we refer to as the theoretical lifetime intrinsic risk (tLIR). To overcome the limitation of inaccurate estimation in the reported stem-cell numbers, we calculated tLIR using both the reported stem-cell number (tLIRsc) and the total tissue cell number (tLIRtt). The latter is equivalent to assuming all homeostatic tissue cells to be stem cells, representing an extreme overestimation of tissue stem cells, which consequently leads to a conservative estimation of the upper bounds in tLIR. The somatic mutation rate in tumours is estimated to be 5\\u2009×\\u200910−10 per nucleotide site per cell division. On this basis, in our initial calculation we used an intrinsic mutation rate (r) of 1\\u2009×\\u200910−8 per cell division, which is equivalent to approximately 20 mutable nucleotide sites for each driver gene where the driver gene will mutate if at least one site mutates. As shown in Fig. 4a, b, if only one hit (that is, mutation of one designated driver gene) is required to develop cancer—that is, k\\u2009=\\u20091—the lifetime risk for almost all cancers is close to 100%. This confirms that one mutation is not enough for cancer onset (otherwise everyone would theoretically acquire each type of cancer). If two driver gene mutations are needed, k\\u2009=\\u20092, the modelled intrinsic risk becomes small for cancers with a small total number of stem-cell divisions; however, it is still very large for those with higher stem-cell divisions, and even unreasonably large for some cancers by surpassing the corresponding observed total lifetime cancer risks (adjusted basal cell carcinoma, colon adenocarcinoma, adjusted melanoma, small intestine cancer, acute myeloid leukaemia and duodenal cancer; Fig. 4a). It is therefore unlikely that, at least in these cancers, two hits will suffice to induce cancer. As shown in Fig. 4, if we consider the more reasonable case where three mutations are required, k\\u2009=\\u20093, almost all modelled intrinsic risks (both tLIRsc and tLIRtt) drop well below our earlier ‘intrinsic’ risk lines estimated conservatively from the observed data alone (red dashed lines, estimated based on observed data following the same mechanism as Fig. 3a). The lifetime risk drops even further for k\\u2009=\\u20094 and beyond. The extrinsic risks based on the tLIRsc and tLIRtt are further summarized in Extended Data Table 4. This modelling approach demonstrates that cancer risk due to intrinsic stem-cell mutation errors alone is low for almost all cancers that require over two mutations, indeed it is lower than the relatively conservative estimate based on data alone (red lines, Fig. 4). As the driver gene mutation rate in stem-cell division is a key parameter, we further conducted sensitivity analyses with different rates (r\\u2009=\\u20091\\u2009×\\u200910−10 to 1\\u2009×\\u200910−6) to examine how this may affect the tLIR (Extended Data Figs 2 and 3). The results show that for k\\u2009=\\u20093, when r\\u2009<\\u20091\\u2009×\\u200910−7 (~200 sites for each driver gene hit), almost all modelled intrinsic risks are below the observed ‘intrinsic’ risk line (red lines); when r\\u2009=\\u20091\\u2009×\\u200910−6 (~2,000 sites for each driver gene hit), the majority of modelled intrinsic risks are still well below the observed ‘intrinsic’ risk lines, particularly those with small total number of divisions (Extended Data Fig. 2). For k\\u2009=\\u20094, when r\\u2009<\\u20091\\u2009×\\u200910−6, almost all modelled intrinsic risks are below the observed ‘intrinsic’ risk lines estimated through the data-driven approach (Extended Data Fig. 3). These sensitivity analyses demonstrate that our conclusions are highly robust, and that the attribution of intrinsic mutations to lifetime cancer risk through stem-cell divisions, particularly for those cancers with low risk, is rather small, even using widely different intrinsic mutation rates.\\nIn summary, we find that a simple regression analysis cannot distinguish between intrinsic and extrinsic factors. We have provided a new framework to quantify the lifetime cancer risks from both intrinsic and extrinsic factors on the basis of four independent approaches that are data-driven and model-driven, with and without using the stem-cell estimations. Importantly, these four approaches provide a consistent estimate of contribution of extrinsic factors of >70–90% in most common cancer types. This is consistent with the overall conclusion regarding the role of extrinsic factors in cancer development.\\n\\n\\nMethods\\n\\n\\nData reporting\\nNo statistical methods were used to predetermine sample size. The experiments were not randomized and investigators were not blinded to allocation during experiments and outcome assessment.\\n\\n\\nDerivation of the probability of possessing k hits after n cell divisions for one cell\\nOn the basis of the theory of the clonal stem-cell origin of cancer, in a given tissue the stem cell would first go through m rounds of symmetric divisions (for each division, each stem cell would divide into two daughter stem cells) to reach a total of S stem cells (S\\u2009=\\u20092m) at the steady state. Subsequently, these S stem cells would go through a rounds of asymmetric divisions (for each division, each stem cell would yield only one daughter stem cell) throughout the lifetime of the tissue. This means that the total rounds of lifetime stem-cell divisions per generation is n\\u2009=\\u2009m\\u2009+\\u2009a. Information on the total rounds of symmetric and asymmetric divisions as well as the total number of stem cells in the steady state for various tissues discussed in this work has been extracted from supplementary table 1 of Tomasetti & Vogelstein. With k hits (mutations of k predetermined driver genes) on a stem cell required for cancer onset, the number of possible cell states of a given stem-cell generation would be k\\u2009+\\u20091, including a zero state with no hit. If we assume that once a hit occurs it cannot be reversed and therefore be carried to all progeny cells, then a cell state may only transition from lower to higher or equal levels from generation to generation. In Extended Data Fig. 4, we demonstrate with k\\u2009=\\u20093 the state transitions of accumulating driver gene mutations. Let Xg denote the number of driver gene mutations accumulated at generation g, and r be the intrinsic driver gene mutation rate due to random errors during DNA replication; the transition probabilities to generation g\\u2009+\\u20091 with i mutations from the previous generation g with j\\u2009≤\\u2009i mutations are derived as follows:\\n\\nIn particular, for the emission state i\\u2009=\\u20090:\\n\\nFor the absorbing state i\\u2009=\\u2009k:\\n\\nBased on these, the computing algorithm is derived as follows:\\nSet the initial cell state at generation 0:\\n\\nFor g\\u2009=\\u20091,…, n and 0\\u2009≤\\u2009i\\u2009≤\\u2009k, we compute the following probabilities iteratively:\\n\\nwhere n is the total number of divisions that one stem cell may experience during its lifetime.\\n\\n\\nDerivation of the theoretical lifetime intrinsic risk (tLIR) of cancer for a given tissue\\nAs mentioned previously, we assume stem cells in a specific tissue undergo two phases of divisions (Extended Data Fig. 5): (1) a total of m symmetric divisions before full tissue development, and (2) a total of a asymmetric divisions for normal tissue turnovers. So in a fully developed tissue, there is a total of S\\u2009=\\u20092m stem cells. For each stem cell, the probability of possessing all k hits for cancer onset after n\\u2009=\\u2009m\\u2009+\\u2009a rounds of divisions is P(Xn\\u2009=\\u2009k), which can be calculated from the previous part. Therefore, the theoretical lifetime intrinsic risk (tLIR) of developing cancer—that is, the probability of at least one stem cell containing k hits during its lifetime—can be expressed as:\\n\\n\\n\\nEstimating cancer risk for different tissues\\nThe rounds of symmetric and asymmetric divisions for different tissues were adopted from supplementary table 1 of Tomasetti & Vogelstein. In particular, the rounds of symmetric divisions, m, is equal to the integer part of log2S, where S is the number of normal stem cells in the tissue of origin (data from ref. 5), and the rounds of asymmetric divisions a was the column labelled ‘d’ in supplementary table 1 of ref. 5. Sensitivity analyses have been conducted for scenarios with a broad range of mutation rates, from 1\\u2009×\\u200910−10 to 1\\u2009×\\u200910−6, and several required hits (k\\u2009=\\u20091, 2, 3, 4).\\n\\n\\nLower-bound estimates of extrinsic risks with the SEER data\\nAs a program of the National Cancer Institute (NCI), SEER (Surveillance, Epidemiology, and End Results Program) is a source of information on cancer incidence and survival in the USA (http://seer.cancer.gov/). The age-adjusted cancer incidences were extracted from the database ‘SEER 9 Regs Research Data, Nov 2014 Sub (1973–2012) <Katrina/Rita Population Adjustment>’ using the SEER*Stat 8.2.1 (ref. 28). For several cancers, it has been observed that their incidence rates have increased markedly during the past 40 years (Extended Data Fig. 1). For these cancers, it is reasonable to assume that anything above the historical minimum incidence should be attributed to some environmental/extrinsic factors. Therefore, we can establish the following inequality:\\nExtrinsic risk\\u2009>\\u2009(1\\u2009−\\u2009historical minimum incidence rate/incidence rate in 2012).\\nCorrespondingly, the lower bounds of contributions by extrinsic factors for these cancers can be calculated. As shown in Extended Data Fig. 1, some cancers show substantial contributions from extrinsic factors.\\n\\n\\nData and statistical analysis\\nThe observed lifetime cancer risks and the cumulative number of divisions (n) of all stem cells per lifetime are adopted from supplementary table 1 of Tomasetti & Vogelstein. The total tissue cell divisions are from our evaluation of the data (Supplementary Information). For the robustness analysis of Fig. 3 as shown in Extended Data Table 1, error terms following the normal distribution with mean 0 and standard deviations of 1 or 0.4 were added to the log10(total stem-cell division) or log10(total cell division). These allow the number of total stem-cell and cell divisions to vary approximately within a range of ~1/100–100-fold or ~1/5–5-fold, respectively. On the basis of the new data set with measurement errors, the excess risks for each cancer were quantified. This process is repeated 1,000 times, and from this the mean, the 2.5 and the 97.5 percentiles (namely the 95% confidence intervals) of the excess risk for each cancer are tabulated. In calculating the percentage of intrinsic versus extrinsic mutations based on mutational signatures from cancer genome, we define the intrinsic mutation as those with signatures 1A/1B, and extrinsic mutation as all other mutational signatures (2–21, R1–R3, U1 and U2). The corresponding data were obtained from supplementary figures 59–88 of ref. 31. All statistical analyses and mathematical calculations were performed using R (version 3.1.2).\\n\\n\\n\\n\\nAuthor Contributions\\nY.A.H. formulated the hypothesis. S.W. and Y.A.H. designed the research. S.W. and W.Z. performed mathematical and statistical analysis. S.W., S.P., W.Z. and Y.A.H. performed research. S.W., S.P., W.Z. and Y.A.H. wrote the paper.\\n\\n'}"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "line"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
